// @stack-spot/portal-observability/dist/event-bus-events.js
import Bowser from 'bowser';
import { ulid } from 'ulid';
export const prepareEvent = ({ data, portal, type }) => {
    const isoDate = new Date().toISOString();
    const isoDateWithoutMs = isoDate.split('.')[0] + 'Z';
    const event = {
        id: ulid(),
        source: `/portal/${portal.toLowerCase()}`,
        specversion: '1.0',
        type: `portal.${type}`,
        time: isoDateWithoutMs,
        data,
    };
    return event;
};
export const getCommonProps = (userSession) => {
    if (!userSession)
        return;
    const { browser } = Bowser.parse(window.navigator.userAgent);
    return {
        browser: browser.name || '',
        browserVersion: browser.version,
        email: userSession.getTokenData().email,
        os: browser.name,
        company: userSession.getTokenData().tenant,
    };
};
export function prepareRouteDetails(routeDetails) {
    const { pathNameWithParams, searchParams, routeParams, searchParamsWithVariableName } = routeDetails;
    return {
        entirePath: searchParamsWithVariableName
            ? `${pathNameWithParams}?${searchParamsWithVariableName}`
            : pathNameWithParams,
        queryParams: searchParams,
        routeParams,
        path: pathNameWithParams,
        currentUrl: window.location.href,
    };
}
