// @stack-spot/portal-components/src/components/Table/HeaderItem.tsx
import { Button, Flex, IconBox, Styles, Text } from '@citric/core'
import { Props as FlexProps } from '@citric/core/dist/Flex'
import { SortAscending, SortDescending } from '@citric/icons'
import React, { ReactNode } from 'react'

export type SortTypes = 'ASC' | 'DESC' | undefined

interface HeaderSortProps extends FlexProps {
  children: ReactNode | string,
  sortType: SortTypes,
  onSort: (sortType: SortTypes) => void,
}

export const parseParams = ({ key, sort }: { key: string | undefined, sort: 'ASC' | 'DESC' | undefined }) =>
  key && sort ? `${key?.toUpperCase()}_${sort?.toUpperCase()}` : undefined

function next(current: SortTypes) {
  switch (current) {
    case undefined: return 'ASC'
    case 'ASC': return 'DESC'
    case 'DESC': return undefined
  }
}

export const HeaderSort: React.FC<HeaderSortProps> = ({ children, sortType, onSort, ...props }) => (
  <Flex
    as={Button}
    appearance="text"
    alignItems="center"
    flexWrap="nowrap"
    sx={{ '&:hover, &:active, &:focus': { borderColor: 'transparent' }, padding: 0 } as Styles}
    onClick={() => onSort(next(sortType))} {...props}>
    <Text appearance="microtext1" mr="2" colorScheme="light.700">
      {children}
    </Text>

    <IconBox colorIcon="inverse.500" size="xs">
      {sortType === 'ASC' && <SortAscending />}
      {sortType === 'DESC' && <SortDescending />}
    </IconBox>
  </Flex>
)

type FullFlexProps = Parameters<typeof Flex>[0]

export const Header: React.FC<FullFlexProps> = ({ children, ...props }) => (
  <Flex alignItems="center" {...props}>
    <Text appearance="microtext1" mr="2" colorScheme="light.700">
      {children}
    </Text>
  </Flex>
)
