// @stack-spot/portal-components/src/components/Table/SettingsVerticalMenu.tsx
import { Box, IconBox, OneOfIconSizes } from '@citric/core'
import { SxProperties } from '@citric/core/dist/sx'
import { EllipsisVertical } from '@citric/icons'
import { IconButton } from '@citric/ui'
import { SelectionList } from '@stack-spot/portal-components/SelectionList'
import { useState } from 'react'

interface SettingsVerticalMenuProps {
  id: string,
  buttonAriaLabel?: string,
  size?: OneOfIconSizes,
  sxIconButton?: SxProperties,
  buttonDataTestHint?: string,
  items: React.ComponentProps<typeof SelectionList>['items'],
}

export const SettingsVerticalMenu = ({ 
  id, 
  buttonAriaLabel, 
  size, 
  buttonDataTestHint, 
  sxIconButton, 
  items = [] }: SettingsVerticalMenuProps) => {
  const [visibleMenu, setVisibleMenu] = useState(false)
  if (items?.length === 0) {
    return null
  }

  return (
    <Box sx={{ position: 'relative' }}>
      <IconButton 
        role="button"
        title={buttonAriaLabel} 
        data-test-hint={buttonDataTestHint} 
        aria-label={buttonAriaLabel} 
        onClick={() => setVisibleMenu(state => !state)} >
        <IconBox size={size} sx={sxIconButton}>
          <EllipsisVertical />
        </IconBox>
      </IconButton>
      <SelectionList
        id={id}
        visible={visibleMenu}
        onHide={() => setVisibleMenu(false)}
        style={{ position: 'absolute', top: 40, right: 0, minWidth: '200px' }}
        items={items}
      />
    </Box>
  )
}
