// @stack-spot/auth/src/interceptors/InterceptorManager.ts
import { FetchRequest } from '../types'

export interface Interceptor {
  interceptRequest?(request: FetchRequest): Promise<FetchRequest>
  interceptResponse?(response: Response, request: FetchRequest): Promise<Response>
}

/**
 * Manages a collection of interceptors for HTTP requests and responses.
 */
export class InterceptorManager {
  private static next = 0
  private originalFetch: typeof fetch
  private interceptors: Record<number, Interceptor> = {}

  constructor(originalFetch: typeof fetch) {
    this.originalFetch = originalFetch
  }

  /**
   * Adds a new interceptor
   * @param interceptor The interceptor to add.
   * @return {Number} An ID used to remove interceptor later
   */  
  addInterceptor(interceptor: Interceptor): number {
    this.interceptors[InterceptorManager.next++] = interceptor
    return InterceptorManager.next - 1
  }

  /**
   * Remove an interceptor
   * @param {Number} id The ID that was returned by `addInterceptor`
   * @returns {Boolean} `true` if the interceptor was removed, `false` otherwise
   */
  removeInterceptor(id: number): boolean {
    if (this.interceptors[id]) {
      delete this.interceptors[id]
      return true
    }   
    return false
  }

  /**
   * Executes all request interceptors in sequence.
   * @param request The original request.
   * @returns The modified request.
   */
  private async executeRequestInterceptors(request: FetchRequest): Promise<FetchRequest> {
    let modifiedRequest = request
    for (const interceptor of Object.values(this.interceptors)) {
      if (interceptor.interceptRequest) {
        modifiedRequest = await interceptor.interceptRequest(modifiedRequest)
      }
    }
    return modifiedRequest
  }

  /**
   * Executes all response interceptors in sequence.
   * @param response The original response.
   * @param request The original request.
   * @returns The modified response.
   */
  private async executeResponseInterceptors(response: Response, request: FetchRequest): Promise<Response> {
    let modifiedResponse = response
    for (const interceptor of Object.values(this.interceptors)) {
      if (interceptor.interceptResponse) {
        modifiedResponse = await interceptor.interceptResponse(modifiedResponse, request)
      }
    }
    return modifiedResponse
  }

  /**
   * A custom implementation of {@link global.fetch} including all interceptors.
   * @param input The URL to the request.
   * @param init The request options.
   * @returns The response.
   */
  async fetch(input: RequestInfo | URL, init?: RequestInit): Promise<Response> {
    const request = await this.executeRequestInterceptors({ url: input, init })
    const response = await this.originalFetch(request.url, request.init)
    return this.executeResponseInterceptors(response, request)
  }
}
