// @stack-spot/auth-react/src/dictionary.ts
import { Dictionary, useTranslate } from '@stack-spot/portal-translate'

const dictionary = {
  en: {
    welcome: 'Welcome to StackSpot AI',
    loginWithEmail: 'Log in with your email.',
    loginWithSocialAccount1: 'Sign up or access your ',
    loginWithSocialAccount2: ' freemium account ',
    loginWithSocialAccount3: ' with a social login',
    label: 'Corporate email',
    placeholder: 'email@company.com',
    continue: 'Continue',
    or: 'Or',
    loginWith: 'Continue with $0',
    emailNotAllowedTitle: 'Your email is linked to an Enterprise account.',
    emailNotAllowedSubtitle: "Please log in with your corporate email.",
    socialLogin: 'Login or register with a social account',
    corporateLoginTitle: 'Already have a StackSpot Enterprise account?',
    corporateLoginButton: 'Enter Enterprise account',
    socialLoginTitle: 'Do you want to access another way?',
    emailNotFoundError: 'We couldn\'t find an account for this email.',
    errorMobileDesktop: "It looks like you are using a mobile device in desktop mode. We recommend adjusting this setting before continuing for a complete experience.",
    loginTemporarilyUnavailable: 'Login temporarily unavailable.'
  },
  pt: {
    welcome: 'Boas vindas à StackSpot AI',
    loginWithEmail: 'Faça login com seu e-mail.',
    loginWithSocialAccount1: 'Cadastre-se ou acesse sua ',
    loginWithSocialAccount2: ' conta freemium ',
    loginWithSocialAccount3: ' com uma conta social',
    label: 'Email corporativo',
    placeholder: 'email@empresa.com',
    continue: 'Continuar',
    or: 'Ou',
    loginWith: 'Continuar com $0',
    emailNotAllowedTitle: '"Este e-mail está vinculado a uma conta Enterprise.',
    emailNotAllowedSubtitle: "Faça login com seu email corporativo.",
    socialLogin: 'Entre ou cadastre-se com uma conta social',
    corporateLoginTitle: 'Já possui uma conta StackSpot Enterprise?',
    corporateLoginButton: 'Entrar na conta Enterprise',
    socialLoginTitle: 'Você quer entrar de outro jeito?',
    emailNotFoundError: 'Não encontramos uma conta para este e-mail.',
    errorMobileDesktop: "Parece que você está utilizando um dispositivo móvel na versão desktop. Recomendamos ajustar essa configuração antes de continuar para uma experiência completa.",
    loginTemporarilyUnavailable: 'Login indisponível temporariamente.'
  },
} satisfies Dictionary

export const useTranslation = () => useTranslate(dictionary)
