// @stack-spot/auth-react/src/IDPLogin.tsx
import { Box, Button, Flex, IconBox, Text } from '@citric/core'
import { ExclamationTriangle } from '@citric/icons'
import { LoadingCircular } from '@citric/ui'
import { interpolate } from '@stack-spot/portal-translate'
import { capitalize } from 'lodash'
import { useTranslation } from './dictionary'
import { Provider } from './hooks'
import { Github } from './provider-icons/Github'
import { Google } from './provider-icons/Google'
import { Microsoft } from './provider-icons/Microsoft'
import { LoginType } from './types'

interface Props {
  trialProviders: Provider[],
  loading: boolean,
  loginProvider: Provider | undefined,
  onSubmit: (type: LoginType, provider?: Provider) => void,
  onChangeMode: (mode: LoginType) => void,
}

interface ButtonProviderProps {
  provider: Provider,
  loading: boolean,
  disabled?: boolean,
  login: (type: LoginType, provider?: Provider) => void,
}

const providerIcons: Record<Provider, React.ReactElement> = {
  github: <Github />,
  google: <Google />,
  microsoft: <Microsoft />,
}

const ButtonProvider = ({ provider, login, loading, disabled }: ButtonProviderProps) => {
  const t = useTranslation()
  return (
    <Box>
      <Button
        colorScheme="light"
        type="button"
        size='md'
        sx={{ width: '100%' } as any}
        onClick={() => login('idp', provider)}
        disabled={loading || disabled}
      >
        {loading
          ? <LoadingCircular />
          : <Flex alignItems='center' style={{ gap: '4px' }}>
            {providerIcons[provider]}{interpolate(t.loginWith, capitalize(provider))}
          </Flex>
        }
      </Button>
    </Box>
  )
}

function hasTouchSupport() {
  return 'ontouchstart' in window || navigator.maxTouchPoints > 0;
}

function isMobileUserAgent() {
  return /Mobi|Android|iPhone|iPad|iPod|BlackBerry|IEMobile|Opera Mini/i.test(navigator.userAgent);
}

function isMobileWithDesktopView() {
  return !isMobileUserAgent() && hasTouchSupport();
}

export const IDPLogin = ({ trialProviders, loading, loginProvider, onSubmit, onChangeMode }: Props) => {
  const t = useTranslation()
  const showErrorMessage = isMobileWithDesktopView()

  return (
    <>
      {showErrorMessage && <Flex bg="warning" p={4} role="alert" flexWrap='nowrap'>
        <IconBox colorIcon="warning.contrastText">
          <ExclamationTriangle /> 
        </IconBox>
        <Text appearance='body2' ml={2} colorScheme="warning.contrastText">
          {t.errorMobileDesktop}
        </Text>
      </Flex>}

      <Flex flexDirection='column' gap>
        {trialProviders?.map((provider) => <ButtonProvider
          provider={provider}
          login={onSubmit}
          loading={loading && loginProvider === provider}
          disabled={loading}
          key={provider}
        />)
        }
      </Flex>
      <p className="separator">
        <Text appearance='microtext1' colorScheme='light.700'>{t.or}</Text>
      </p>
      <Text colorScheme="light.700" align="center">{t.corporateLoginTitle}</Text>
      <Button size='md' disabled={loading} colorScheme="light" onClick={() => onChangeMode('sso')}>
        {t.corporateLoginButton}
      </Button>
    </>
  )
}
