// @stack-spot/auth-react/src/SSOLogin.tsx
import { Button, Flex, Input, Label, Text } from '@citric/core'
import { LoadingCircular } from '@citric/ui'
import { useTranslation } from './dictionary'
import { LoginType } from './types'

interface Props {
  value: string,
  onChange: (value: string) => void,
  disabled: boolean,
  loading: boolean,
  hasProvider: boolean,
  onChangeMode: (mode: LoginType) => void,
  idpLoginEnabled: boolean,
}


export const SSOLogin = ({ value, onChange, loading, disabled, hasProvider, onChangeMode, idpLoginEnabled }: Props) => {
  const t = useTranslation()
  return (
    <>
      <Flex flexDirection='column' style={{ gap: '4px', marginTop: '4px' }}>
        <Label htmlFor='email'>{t.label}</Label>
        <Input id="email" type='email' name="email" value={value} onChange={e => onChange(e.target.value)} placeholder={t.placeholder} />
        <Button colorScheme="primary" size='md' style={{ marginTop: '12px' }} disabled={disabled || loading}>
          {loading && !hasProvider ? <LoadingCircular /> : <Text>{t.continue}</Text>}
        </Button>
      </Flex>
      {idpLoginEnabled && <>
        <p className="separator">
          <Text appearance='microtext1' colorScheme='light.700'>{t.or}</Text>
        </p>
        <Text colorScheme="light.700" align="center">{t.socialLoginTitle}</Text>
        <Button size='md' disabled={loading} colorScheme="light" onClick={() => onChangeMode('idp')}>
          {t.socialLogin}
        </Button>
      </>
      }
    </>
  )
}
