// @stack-spot/portal-layout/src/components/tour/StepContainer.tsx
import { theme } from '@stack-spot/portal-theme'
import React, { ReactNode } from 'react'
import { styled } from 'styled-components'
import { NavigationProps, StepNavigation } from './StepNavigation'
import { StepTitle } from './StepTitle'
import { useTour } from './hook'

interface StepContainerProps {
  children: ReactNode,
  /**
   * The unique identifier for the step.
   */
  stepKey: string,
  /**
   * The title for the step.
   */
  title: string,
  /**
   * The position of the tour overlay related to the content being explained.
   */
  position?: PointingArrowPosition,
  /**
   * The CSS 'right' property to adjust the horizontal position of the overlay.
   */
  right?: React.CSSProperties['right'],
  /**
   * The CSS 'top' property to adjust the vertical position of the overlay.
   */
  top?: React.CSSProperties['top'],
  /**
   * A customizable set of buttons for navigating the tour steps.
   */
  customNavigation?: Omit<NavigationProps, 'stepKey'>,
}

/**
 * Tutorial: the overlay component for showing a step on React Tour.
 * @param props the react props for the component {@link StepContainerProps}.
 */
export const StepContainer = ({ title, stepKey, customNavigation, position = 'bottom', top, right, children }: StepContainerProps) => {
  const { closeTour } = useTour()
  return (
    <BoxWithPointingArrow $position={position} $top={top} $right={right}>
      <StepTitle title={title} onClose={() => closeTour()} />
      {children}
      <StepNavigation stepKey={stepKey} {...(customNavigation || {})} />
    </BoxWithPointingArrow>
  )
}

export type PointingArrowPosition = 'right' | 'top' | 'left' | 'bottom'

const BoxWithPointingArrow = styled.div < {
  $position: PointingArrowPosition,
  $top?: React.CSSProperties['top'],
  $right?: React.CSSProperties['right'],
} > `
  position: relative;
  width: 100%;
  background-color: ${theme.color.inverse[500]};
  &::after {
    content: '';
    position: absolute;
    border-width: 10px;
    border-style: solid;
    border-color: transparent;
    border-right-color: ${theme.color.inverse[500]};
    ${({ $position, $top }) => $position === 'right' ?
    `
      top: ${$top || '16px'};
      right: 100%;
    ` : ''}
    ${({ $position, $right }) => $position === 'top' ?
    `
    top: 100%;
    right: ${$right || '16px'};
    transform: rotate(-90deg);
    ` : ''}
    ${({ $position, $top }) => $position === 'left' ?
    `
      top: ${$top || '16px'};
      left: 100%;
      transform: rotate(180deg);
    ` : ''}
    ${({ $position, $right }) => $position === 'bottom' ?
    `
    bottom: 100%;
    right: ${$right || '16px'};
    transform: rotate(90deg);
    ` : ''}
  }
`
