// @stack-spot/citric-react/dist/components/Tabs/TabController.js
import { ValueController } from '../../utils/ValueController.js';
export class TabController extends ValueController {
    constructor(tabOrder, value) {
        super(value);
        Object.defineProperty(this, "tabOrder", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: void 0
        });
        this.tabOrder = tabOrder;
    }
    getCurrentIndex() {
        return this.tabOrder.findIndex(t => t === this.value);
    }
    /**
     * @returns true if there's another tab after the one currently selected. False otherwise.
     */
    hasNext() {
        const current = this.getCurrentIndex();
        return current > -1 && current + 1 < this.tabOrder.length;
    }
    /**
     * @returns true if there's a tab before the one currently selected. False otherwise.
     */
    hasPrevious() {
        const current = this.getCurrentIndex();
        return current > -1 && current - 1 >= 0;
    }
    /**
     * Selects the tab on the right of the one currently selected. If there's no next tab, nothing happens.
     * @returns true if the tab is changed, false otherwise.
     */
    next() {
        if (this.hasNext()) {
            this.setValue(this.tabOrder[this.getCurrentIndex() + 1]);
            return true;
        }
        return false;
    }
    /**
     * Selects the tab on the left of the one currently selected. If there's no previous tab, nothing happens.
     * @returns true if the tab is changed, false otherwise.
     */
    previous() {
        if (this.hasPrevious()) {
            this.setValue(this.tabOrder[this.getCurrentIndex() - 1]);
            return true;
        }
        return false;
    }
}
