// @stack-spot/auth-react/out/index.mjs
import { jsxs, jsx, Fragment } from 'react/jsx-runtime';
import { Flex, IconBox, Text, Button, Box, Label, Input } from '@citric/core';
import { getCookieDomain, setCookie, getCookie, removeCookie, BannerWarning, useEffectOnce } from '@stack-spot/portal-components';
import { theme, CSSToCitricAdapter } from '@stack-spot/portal-theme';
import '@stack-spot/portal-theme/dist/theme.css';
import { useTranslate, interpolate, useLanguage, ptEn } from '@stack-spot/portal-translate';
import { useState, useEffect, forwardRef } from 'react';
import { LoadingCircular, Card } from '@citric/ui';
import { AuthManager, AuthMethodUnavailable } from '@stack-spot/auth';
import { MiniLogo } from '@stack-spot/portal-components/svg';
import { styled } from 'styled-components';
import { ExclamationTriangle } from '@citric/icons';
import { capitalize } from 'lodash';

const dictionary = {
  en: {
    welcome: "Welcome to StackSpot AI",
    loginWithEmail: "Log in with your email.",
    loginWithSocialAccount1: "Sign up or access your ",
    loginWithSocialAccount2: " freemium account ",
    loginWithSocialAccount3: " with a social login",
    label: "Corporate email",
    placeholder: "email@company.com",
    continue: "Continue",
    or: "Or",
    loginWith: "Continue with $0",
    emailNotAllowedTitle: "Your email is linked to an Enterprise account.",
    emailNotAllowedSubtitle: "Please log in with your corporate email.",
    socialLogin: "Login or register with a social account",
    corporateLoginTitle: "Already have a StackSpot Enterprise account?",
    corporateLoginButton: "Enter Enterprise account",
    socialLoginTitle: "Do you want to access another way?",
    emailNotFoundError: "We couldn't find an account for this email.",
    errorMobileDesktop: "It looks like you are using a mobile device in desktop mode. We recommend adjusting this setting before continuing for a complete experience.",
    loginTemporarilyUnavailable: "Login temporarily unavailable."
  },
  pt: {
    welcome: "Boas vindas \xE0 StackSpot AI",
    loginWithEmail: "Fa\xE7a login com seu e-mail.",
    loginWithSocialAccount1: "Cadastre-se ou acesse sua ",
    loginWithSocialAccount2: " conta freemium ",
    loginWithSocialAccount3: " com uma conta social",
    label: "Email corporativo",
    placeholder: "email@empresa.com",
    continue: "Continuar",
    or: "Ou",
    loginWith: "Continuar com $0",
    emailNotAllowedTitle: '"Este e-mail est\xE1 vinculado a uma conta Enterprise.',
    emailNotAllowedSubtitle: "Fa\xE7a login com seu email corporativo.",
    socialLogin: "Entre ou cadastre-se com uma conta social",
    corporateLoginTitle: "J\xE1 possui uma conta StackSpot Enterprise?",
    corporateLoginButton: "Entrar na conta Enterprise",
    socialLoginTitle: "Voc\xEA quer entrar de outro jeito?",
    emailNotFoundError: "N\xE3o encontramos uma conta para este e-mail.",
    errorMobileDesktop: "Parece que voc\xEA est\xE1 utilizando um dispositivo m\xF3vel na vers\xE3o desktop. Recomendamos ajustar essa configura\xE7\xE3o antes de continuar para uma experi\xEAncia completa.",
    loginTemporarilyUnavailable: "Login indispon\xEDvel temporariamente."
  }
};
const useTranslation = () => useTranslate(dictionary);

const sessionKey$1 = `stk-session${getCookieDomain()}`;
const sessionCookie = Object.freeze({
  /**
   * Sets the user's session cookie
   * @param data session to be saved in the cookie
   * @param customAttributes Accepted values: https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Set-Cookie#attributes.
   */
  set: (data, customAttributes = {}) => setCookie(sessionKey$1, JSON.stringify(data), customAttributes),
  get: () => {
    try {
      const cookie = getCookie(sessionKey$1);
      return cookie ? JSON.parse(cookie) : void 0;
    } catch (error) {
      console.error(error);
    }
  },
  delete: () => removeCookie(sessionKey$1)
});

const isValidDomain = (url) => {
  const portalDomainRegex = new RegExp(`^https?://[a-zA-Z0-9.-]*${getCookieDomain().replaceAll(".", ".")}(:[0-9]{2,4})*.*$`, "g");
  const platformDomainRegex = new RegExp(/^https:\/\/[a-zA-Z0-9.-]+\.stackspot\.com.*$/, "g");
  const result = portalDomainRegex.test(url) || platformDomainRegex.test(url);
  return result;
};
const redirect = async (url) => {
  if (!isValidDomain(url))
    throw new Error("Redirect URL invalid domain");
  window.location.href = url;
  await new Promise(() => "");
};

var __defProp$4 = Object.defineProperty;
var __defProps$4 = Object.defineProperties;
var __getOwnPropDescs$4 = Object.getOwnPropertyDescriptors;
var __getOwnPropSymbols$4 = Object.getOwnPropertySymbols;
var __hasOwnProp$4 = Object.prototype.hasOwnProperty;
var __propIsEnum$4 = Object.prototype.propertyIsEnumerable;
var __defNormalProp$4 = (obj, key, value) => key in obj ? __defProp$4(obj, key, { enumerable: true, configurable: true, writable: true, value }) : obj[key] = value;
var __spreadValues$4 = (a, b) => {
  for (var prop in b || (b = {}))
    if (__hasOwnProp$4.call(b, prop))
      __defNormalProp$4(a, prop, b[prop]);
  if (__getOwnPropSymbols$4)
    for (var prop of __getOwnPropSymbols$4(b)) {
      if (__propIsEnum$4.call(b, prop))
        __defNormalProp$4(a, prop, b[prop]);
    }
  return a;
};
var __spreadProps$4 = (a, b) => __defProps$4(a, __getOwnPropDescs$4(b));
var __publicField = (obj, key, value) => {
  __defNormalProp$4(obj, typeof key !== "symbol" ? key + "" : key, value);
  return value;
};
const sessionKey = "session";
const _SessionManager = class _SessionManager {
  constructor(config) {
    __publicField(this, "current");
    __publicField(this, "auth");
    __publicField(this, "config");
    __publicField(this, "changeListeners", []);
    __publicField(this, "logger");
    config.loginUrl || (config.loginUrl = location.origin);
    const redirectUrl = (config.redirectUrl || config.loginUrl).replace(/([^/])$/, "$1/");
    this.config = config;
    this.auth = new AuthManager(__spreadProps$4(__spreadValues$4({}, config), {
      redirectUrl,
      storage: localStorage,
      sessionPersistence: {
        load: () => localStorage.getItem(sessionKey),
        save: (session) => localStorage.setItem(sessionKey, session)
      }
    }));
    this.logger = this.auth.config.logger;
    _SessionManager.instance = this;
    addEventListener("focus", () => this.validateSharedSession());
  }
  static create(config) {
    var _a;
    return (_a = _SessionManager.instance) != null ? _a : new _SessionManager(config);
  }
  setSession(session) {
    this.current = session;
    this.changeListeners.forEach((l) => l(session));
    if (session)
      this.setSessionCookie(session);
  }
  async restoreSession() {
    const session = await this.auth.restoreSession();
    this.logger.log("Validating shared session.");
    const sessionValid = await this.validateSharedSession(session);
    this.setSession(sessionValid ? session : void 0);
  }
  async validateSharedSession(session = this.current) {
    var _a;
    if (this.urlHasThirdPartyLoginData()) {
      this.logger.log("Session is invalid because there's another authentication in progress.");
      return false;
    }
    const sharedSessionCookie = sessionCookie.get();
    if (!sharedSessionCookie) {
      this.logger.log("Session is invalid because no shared session cookie was found, i.e, a logout was performed in another portal. Forcing log off.");
      session && await this.logout();
      return false;
    }
    const isDifferentSessionActive = sharedSessionCookie.sub != (session == null ? void 0 : session.getTokenData().sub);
    const isSharedSessionTypeBlocked = (_a = this.config.blockedAuthTypes) == null ? void 0 : _a.includes(sharedSessionCookie.type);
    if (isSharedSessionTypeBlocked) {
      this.logger.log("Session is invalid because shared sessions have been blocked in the SessionManager's configuration (blockedAuthTypes).");
      return false;
    } else if (isDifferentSessionActive || !session) {
      this.logger.log(
        isDifferentSessionActive ? "Session is invalid because a different session is already active." : "Session is invalid because it's undefined."
      );
      this.logger.log("Starting login with tenant from the session cookie.");
      await this.startThirdPartyLoginUsingTenant(sharedSessionCookie);
      return false;
    }
    return true;
  }
  hasSession() {
    return !!this.current && !this.current.isExpired();
  }
  getSession() {
    if (!this.hasSession()) {
      this.endSession();
      throw new Error("Session is not available, redirecting to login.");
    }
    return this.current;
  }
  async endSession(redirectToLogin = true) {
    this.current = void 0;
    localStorage.removeItem(sessionKey);
    sessionCookie.delete();
    if (redirectToLogin && this.config.loginUrl)
      await redirect(this.config.loginUrl);
  }
  async restartSession() {
    await this.logout({ endSession: false });
    this.current = void 0;
    localStorage.removeItem(sessionKey);
    await this.restoreSession();
  }
  async logout({ endSession } = { endSession: true }) {
    var _a;
    try {
      await ((_a = this.current) == null ? void 0 : _a.logout());
    } catch (error) {
      console.error(`Could not logout from IDM.
${error}`);
    }
    if (!endSession)
      return;
    await this.endSession();
  }
  async startThirdPartyLogin(data) {
    const params = new URLSearchParams(location.search);
    const authUrl = await this.auth.startThirdPartyLogin(data, {
      from: location.href,
      finalRedirect: params.get("finalRedirect")
    });
    await redirect(authUrl);
  }
  urlHasThirdPartyLoginData() {
    const url = new URL(location.toString());
    return url.searchParams.has("state") && !url.searchParams.has("error");
  }
  async startThirdPartyLoginUsingTenant(data) {
    const params = new URLSearchParams(location.search);
    const cookie = sessionCookie.get();
    if (!cookie || !cookie.tenant) {
      this.logger.log("Login out because no tenant information is available in the following data:", JSON.stringify(data));
      await this.logout();
      return;
    }
    const authUrl = await this.auth.getThirdPartyLoginFromTenant(
      data,
      cookie.tenant,
      {
        from: location.href,
        finalRedirect: params.get("finalRedirect")
      }
    );
    await redirect(authUrl);
  }
  async completeThirdPartyLogin() {
    var _a;
    const url = new URL(location.toString());
    if (url.searchParams.has("error")) {
      throw new Error(`Error while signing in: ${url.searchParams.get("error_description")}`);
    }
    const { session, data: { from, finalRedirect } } = await this.auth.completeThirdPartyLogin(location.search);
    this.setSession(session);
    history.replaceState(null, "", from || location.toString().replace(/\?.*$/, ""));
    this.sendLoginEventRd((_a = this.current) == null ? void 0 : _a.getTokenData());
    if (finalRedirect)
      await redirect(finalRedirect);
  }
  getEmailForLogin() {
    const session = sessionCookie.get();
    return (session == null ? void 0 : session.type) == "sso" ? session.email : void 0;
  }
  async switchAccount(accountId) {
    var _a;
    this.logger.log("Switching accounts", accountId, (_a = this.current) == null ? void 0 : _a.getTokenData().account_id_v2);
    try {
      this.current && await this.auth.switchAccount(accountId, this.current);
    } catch (error) {
      this.logger.error("Error while switching accounts", error);
      throw error;
    }
    this.setSession(this.current);
  }
  onChange(listener) {
    this.changeListeners.push(listener);
    return () => {
      const index = this.changeListeners.indexOf(listener);
      if (index != -1)
        this.changeListeners.splice(index, 1);
    };
  }
  setSessionCookie(session) {
    const { email, account_type, sub, tenant } = session.getTokenData();
    const { provider, refresh_expires_in } = session.getSessionData();
    if (!email || !sub || !tenant)
      return;
    const isFreemium = account_type == "FREEMIUM";
    const cookieAttributes = { "Max-Age": refresh_expires_in, path: "/" };
    if (isFreemium) {
      sessionCookie.set({ type: "idp", provider, sub, tenant }, cookieAttributes);
    } else {
      sessionCookie.set({ email, type: "sso", sub, tenant }, cookieAttributes);
    }
  }
  async sendLoginEventRd(tokenData) {
    if (!this.config.rdUrl)
      return;
    if (!tokenData) {
      console.error("Unable to trigger login hook. No sessionEmail or name identified.");
      return;
    }
    const { email, name, account_type, client_id, account_name, trial_account_status } = tokenData;
    const isLoginAI = client_id === "stackspot-portal-ai";
    const isLoginEDP = client_id === "stackspot-portal";
    if (!isLoginAI && !isLoginEDP && trial_account_status === "PENDING")
      return;
    const leadType = account_type === "FREEMIUM" ? "TRIAL" : "ENTERPRISE";
    const rdObject = {
      event_type: "CONVERSION",
      event_family: "CDP",
      payload: {
        email,
        name,
        conversion_identifier: isLoginAI ? "login_ai" : "login_edp",
        cf_leadtype: leadType,
        cf_account_name: leadType === "TRIAL" ? leadType : account_name
      }
    };
    const response = await fetch(this.config.rdUrl, {
      method: "POST",
      body: JSON.stringify(rdObject),
      headers: {
        "content-type": "application/json"
      }
    });
    const data = await response.json();
    if (!response.ok) {
      console.error("Error while sending event to RD Station", data);
    }
  }
  async getTrialEnabledProviders() {
    try {
      const response = await fetch(`${this.config.accountUrl}/v1/accounts/trial/sso`);
      const trialProviders = await response.json();
      if (!response.ok) {
        console.error("Error while fetching available login providers", trialProviders);
      }
      const providerKeys = Object.keys(trialProviders || {});
      return providerKeys.filter((key) => trialProviders[key] === true);
    } catch (error) {
      console.error("Error while fetching available login providers", error);
      return [];
    }
  }
};
__publicField(_SessionManager, "instance");
let SessionManager = _SessionManager;

function useSession() {
  const manager = SessionManager.instance;
  const [session, setSession] = useState((manager == null ? void 0 : manager.hasSession()) ? manager.getSession() : void 0);
  useEffect(() => {
    return manager == null ? void 0 : manager.onChange(setSession);
  }, []);
  return session;
}
const useTrialProviders = ({ enabled = true }) => {
  const [isLoadingTrialProviders, setIsLoadingTrialProviders] = useState(enabled);
  const [trialProviders, setTrialProviders] = useState([]);
  useEffect(() => {
    (async () => {
      if (!SessionManager.instance || !enabled)
        return;
      try {
        const providers = await SessionManager.instance.getTrialEnabledProviders();
        setTrialProviders(providers);
        setIsLoadingTrialProviders(false);
      } catch (error) {
        console.error(error);
        setIsLoadingTrialProviders(false);
      }
    })();
  }, [SessionManager.instance]);
  return [trialProviders, isLoadingTrialProviders];
};

var __defProp$3 = Object.defineProperty;
var __defProps$3 = Object.defineProperties;
var __getOwnPropDescs$3 = Object.getOwnPropertyDescriptors;
var __getOwnPropSymbols$3 = Object.getOwnPropertySymbols;
var __hasOwnProp$3 = Object.prototype.hasOwnProperty;
var __propIsEnum$3 = Object.prototype.propertyIsEnumerable;
var __defNormalProp$3 = (obj, key, value) => key in obj ? __defProp$3(obj, key, { enumerable: true, configurable: true, writable: true, value }) : obj[key] = value;
var __spreadValues$3 = (a, b) => {
  for (var prop in b || (b = {}))
    if (__hasOwnProp$3.call(b, prop))
      __defNormalProp$3(a, prop, b[prop]);
  if (__getOwnPropSymbols$3)
    for (var prop of __getOwnPropSymbols$3(b)) {
      if (__propIsEnum$3.call(b, prop))
        __defNormalProp$3(a, prop, b[prop]);
    }
  return a;
};
var __spreadProps$3 = (a, b) => __defProps$3(a, __getOwnPropDescs$3(b));
const Github = forwardRef((props, ref) => /* @__PURE__ */ jsxs("svg", __spreadProps$3(__spreadValues$3({ ref }, props), { width: "25", height: "25", viewBox: "0 0 25 25", xmlns: "http://www.w3.org/2000/svg", children: [
  /* @__PURE__ */ jsxs("g", { "clip-path": "url(#clip0_1215_2072)", children: [
    /* @__PURE__ */ jsx("path", { "fill-rule": "evenodd", "clip-rule": "evenodd", d: "M12.0011 4.38525C10.1019 4.38624 8.26498 5.05889 6.81881 6.28294C5.37263 7.50698 4.4115 9.20259 4.10726 11.0666C3.80302 12.9306 4.17551 14.8414 5.15814 16.4574C6.14077 18.0733 7.66946 19.2891 9.47085 19.8872C9.86827 19.961 10.018 19.7147 10.018 19.5053C10.018 19.2958 10.01 18.6886 10.0074 18.0247C7.78183 18.5055 7.31155 17.0856 7.31155 17.0856C6.94858 16.1635 6.42399 15.9212 6.42399 15.9212C5.69804 15.4286 6.4783 15.4378 6.4783 15.4378C7.28241 15.4944 7.705 16.2584 7.705 16.2584C8.4177 17.4741 9.57683 17.1225 10.0325 16.917C10.1041 16.402 10.312 16.0516 10.5412 15.8527C8.76345 15.6525 6.89559 14.9702 6.89559 11.9222C6.88457 11.1317 7.17958 10.3673 7.71957 9.78704C7.63744 9.58683 7.36321 8.77808 7.79772 7.67954C7.79772 7.67954 8.46936 7.46616 9.99809 8.49488C11.3093 8.13834 12.6928 8.13834 14.0041 8.49488C15.5315 7.46616 16.2018 7.67954 16.2018 7.67954C16.6376 8.77544 16.3634 9.58419 16.2813 9.78704C16.823 10.3674 17.1186 11.1332 17.1066 11.9248C17.1066 14.9794 15.2347 15.6525 13.4543 15.8487C13.7404 16.0964 13.9961 16.5798 13.9961 17.3227C13.9961 18.387 13.9868 19.2431 13.9868 19.5053C13.9868 19.7173 14.1312 19.9649 14.5366 19.8872C16.3382 19.289 17.867 18.0731 18.8496 16.4568C19.8323 14.8406 20.2046 12.9295 19.9 11.0653C19.5954 9.20112 18.6338 7.50549 17.1871 6.28166C15.7405 5.05782 13.9031 4.38561 12.0037 4.38525H12.0011Z", fill: "white" }),
    /* @__PURE__ */ jsx("path", { d: "M9.35091 17.3684C9.35091 17.4329 9.27673 17.4882 9.18135 17.4896C9.08597 17.4909 9.00781 17.4382 9.00781 17.3736C9.00781 17.3091 9.08199 17.2538 9.17737 17.2525C9.27275 17.2511 9.35091 17.3025 9.35091 17.3684Z", fill: "white" }),
    /* @__PURE__ */ jsx("path", { d: "M9.96094 17.2672C9.97286 17.3317 9.90662 17.3989 9.81124 17.4147C9.71586 17.4305 9.63241 17.3923 9.62048 17.3291C9.60856 17.2659 9.67745 17.1974 9.77018 17.1803C9.86291 17.1631 9.94902 17.2027 9.96094 17.2672Z", fill: "white" }),
    /* @__PURE__ */ jsx("path", { d: "M8.6968 17.324C8.67693 17.3859 8.58685 17.4136 8.49676 17.3872C8.40668 17.3609 8.34707 17.2871 8.36429 17.2239C8.38151 17.1607 8.47292 17.1317 8.56433 17.1607C8.65573 17.1897 8.71402 17.2595 8.6968 17.324Z", fill: "white" }),
    /* @__PURE__ */ jsx("path", { d: "M8.09774 17.0658C8.05402 17.1145 7.96527 17.1013 7.89241 17.0355C7.81955 16.9696 7.80232 16.88 7.84604 16.8326C7.88975 16.7852 7.97851 16.7984 8.05402 16.8629C8.12953 16.9274 8.1441 17.0183 8.09774 17.0658V17.0658Z", fill: "white" }),
    /* @__PURE__ */ jsx("path", { d: "M7.6686 16.6231C7.61959 16.6574 7.53612 16.6231 7.48976 16.5546C7.47694 16.5423 7.46674 16.5276 7.45978 16.5113C7.45281 16.495 7.44922 16.4775 7.44922 16.4598C7.44922 16.4421 7.45281 16.4246 7.45978 16.4083C7.46674 16.392 7.47694 16.3772 7.48976 16.3649C7.53877 16.332 7.62224 16.3649 7.6686 16.4321C7.71497 16.4993 7.71629 16.5889 7.6686 16.6231V16.6231Z", fill: "white" }),
    /* @__PURE__ */ jsx("path", { d: "M7.3535 16.1662C7.32607 16.1799 7.29468 16.1838 7.26472 16.177C7.23475 16.1703 7.20807 16.1534 7.18924 16.1293C7.13757 16.074 7.12697 15.9976 7.16671 15.9633C7.20645 15.9291 7.27799 15.9449 7.32966 16.0002C7.38132 16.0555 7.39324 16.1319 7.3535 16.1662Z", fill: "white" }),
    /* @__PURE__ */ jsx("path", { d: "M7.02905 15.8062C7.01183 15.8457 6.94825 15.8576 6.89658 15.8299C6.84492 15.8022 6.8065 15.7509 6.82505 15.71C6.8436 15.6692 6.90586 15.6587 6.95752 15.6863C7.00919 15.714 7.04893 15.7667 7.02905 15.8062Z", fill: "white" })
  ] }),
  /* @__PURE__ */ jsx("defs", { children: /* @__PURE__ */ jsx("clipPath", { id: "clip0_1215_2072", children: /* @__PURE__ */ jsx("rect", { width: "16", height: "16", fill: "white", transform: "translate(4 4.05176)" }) }) })
] })));

var __defProp$2 = Object.defineProperty;
var __defProps$2 = Object.defineProperties;
var __getOwnPropDescs$2 = Object.getOwnPropertyDescriptors;
var __getOwnPropSymbols$2 = Object.getOwnPropertySymbols;
var __hasOwnProp$2 = Object.prototype.hasOwnProperty;
var __propIsEnum$2 = Object.prototype.propertyIsEnumerable;
var __defNormalProp$2 = (obj, key, value) => key in obj ? __defProp$2(obj, key, { enumerable: true, configurable: true, writable: true, value }) : obj[key] = value;
var __spreadValues$2 = (a, b) => {
  for (var prop in b || (b = {}))
    if (__hasOwnProp$2.call(b, prop))
      __defNormalProp$2(a, prop, b[prop]);
  if (__getOwnPropSymbols$2)
    for (var prop of __getOwnPropSymbols$2(b)) {
      if (__propIsEnum$2.call(b, prop))
        __defNormalProp$2(a, prop, b[prop]);
    }
  return a;
};
var __spreadProps$2 = (a, b) => __defProps$2(a, __getOwnPropDescs$2(b));
const Google = forwardRef((props, ref) => /* @__PURE__ */ jsxs("svg", __spreadProps$2(__spreadValues$2({ ref }, props), { width: "25", height: "25", viewBox: "0 0 25 25", xmlns: "http://www.w3.org/2000/svg", children: [
  /* @__PURE__ */ jsxs("g", { "clip-path": "url(#clip0_1215_2403)", children: [
    /* @__PURE__ */ jsx("path", { d: "M20.3442 12.2359C20.3442 11.6921 20.3001 11.1454 20.206 10.6104H12.6602V13.691H16.9813C16.802 14.6846 16.2258 15.5635 15.3822 16.122V18.1209H17.9602C19.4741 16.7276 20.3442 14.6699 20.3442 12.2359Z", fill: "#4285F4" }),
    /* @__PURE__ */ jsx("path", { d: "M12.6607 20.0525C14.8184 20.0525 16.6379 19.344 17.9637 18.1212L15.3857 16.1223C14.6684 16.6103 13.7425 16.8866 12.6637 16.8866C10.5766 16.8866 8.80696 15.4785 8.17202 13.5854H5.51172V15.6461C6.86979 18.3475 9.63592 20.0525 12.6607 20.0525V20.0525Z", fill: "#34A853" }),
    /* @__PURE__ */ jsx("path", { d: "M8.16852 13.5856C7.83341 12.592 7.83341 11.5161 8.16852 10.5225V8.46191H5.51116C4.37649 10.7224 4.37649 13.3857 5.51116 15.6462L8.16852 13.5856V13.5856Z", fill: "#FBBC04" }),
    /* @__PURE__ */ jsx("path", { d: "M12.6607 7.2182C13.8013 7.20056 14.9036 7.62974 15.7296 8.41754L18.0136 6.1335C16.5674 4.77543 14.6479 4.02878 12.6607 4.0523C9.63592 4.0523 6.86979 5.75724 5.51172 8.46163L8.16908 10.5223C8.80108 8.62625 10.5736 7.2182 12.6607 7.2182V7.2182Z", fill: "#EA4335" })
  ] }),
  /* @__PURE__ */ jsx("defs", { children: /* @__PURE__ */ jsx("clipPath", { id: "clip0_1215_2403", children: /* @__PURE__ */ jsx("rect", { width: "16", height: "16", fill: "white", transform: "translate(4.5 4.05176)" }) }) })
] })));

var __defProp$1 = Object.defineProperty;
var __defProps$1 = Object.defineProperties;
var __getOwnPropDescs$1 = Object.getOwnPropertyDescriptors;
var __getOwnPropSymbols$1 = Object.getOwnPropertySymbols;
var __hasOwnProp$1 = Object.prototype.hasOwnProperty;
var __propIsEnum$1 = Object.prototype.propertyIsEnumerable;
var __defNormalProp$1 = (obj, key, value) => key in obj ? __defProp$1(obj, key, { enumerable: true, configurable: true, writable: true, value }) : obj[key] = value;
var __spreadValues$1 = (a, b) => {
  for (var prop in b || (b = {}))
    if (__hasOwnProp$1.call(b, prop))
      __defNormalProp$1(a, prop, b[prop]);
  if (__getOwnPropSymbols$1)
    for (var prop of __getOwnPropSymbols$1(b)) {
      if (__propIsEnum$1.call(b, prop))
        __defNormalProp$1(a, prop, b[prop]);
    }
  return a;
};
var __spreadProps$1 = (a, b) => __defProps$1(a, __getOwnPropDescs$1(b));
const Microsoft = forwardRef((props, ref) => /* @__PURE__ */ jsxs("svg", __spreadProps$1(__spreadValues$1({ ref }, props), { width: "25", height: "25", viewBox: "0 0 25 25", xmlns: "http://www.w3.org/2000/svg", children: [
  /* @__PURE__ */ jsx("path", { d: "M5.19531 4.74707H12.1518V11.7036H5.19531V4.74707Z", fill: "#F35325" }),
  /* @__PURE__ */ jsx("path", { d: "M12.8477 4.74707H19.8042V11.7036H12.8477V4.74707Z", fill: "#81BC06" }),
  /* @__PURE__ */ jsx("path", { d: "M5.19531 12.3994H12.1518V19.3559H5.19531V12.3994Z", fill: "#05A6F0" }),
  /* @__PURE__ */ jsx("path", { d: "M12.8477 12.3994H19.8042V19.3559H12.8477V12.3994Z", fill: "#FFBA08" })
] })));

const providerIcons = {
  github: /* @__PURE__ */ jsx(Github, {}),
  google: /* @__PURE__ */ jsx(Google, {}),
  microsoft: /* @__PURE__ */ jsx(Microsoft, {})
};
const ButtonProvider = ({ provider, login, loading, disabled }) => {
  const t = useTranslation();
  return /* @__PURE__ */ jsx(Box, { children: /* @__PURE__ */ jsx(
    Button,
    {
      colorScheme: "light",
      type: "button",
      size: "md",
      sx: { width: "100%" },
      onClick: () => login("idp", provider),
      disabled: loading || disabled,
      children: loading ? /* @__PURE__ */ jsx(LoadingCircular, {}) : /* @__PURE__ */ jsxs(Flex, { alignItems: "center", style: { gap: "4px" }, children: [
        providerIcons[provider],
        interpolate(t.loginWith, capitalize(provider))
      ] })
    }
  ) });
};
function hasTouchSupport() {
  return "ontouchstart" in window || navigator.maxTouchPoints > 0;
}
function isMobileUserAgent() {
  return /Mobi|Android|iPhone|iPad|iPod|BlackBerry|IEMobile|Opera Mini/i.test(navigator.userAgent);
}
function isMobileWithDesktopView() {
  return !isMobileUserAgent() && hasTouchSupport();
}
const IDPLogin = ({ trialProviders, loading, loginProvider, onSubmit, onChangeMode }) => {
  const t = useTranslation();
  const showErrorMessage = isMobileWithDesktopView();
  return /* @__PURE__ */ jsxs(Fragment, { children: [
    showErrorMessage && /* @__PURE__ */ jsxs(Flex, { bg: "warning", p: 4, role: "alert", flexWrap: "nowrap", children: [
      /* @__PURE__ */ jsx(IconBox, { colorIcon: "warning.contrastText", children: /* @__PURE__ */ jsx(ExclamationTriangle, {}) }),
      /* @__PURE__ */ jsx(Text, { appearance: "body2", ml: 2, colorScheme: "warning.contrastText", children: t.errorMobileDesktop })
    ] }),
    /* @__PURE__ */ jsx(Flex, { flexDirection: "column", gap: true, children: trialProviders == null ? void 0 : trialProviders.map((provider) => /* @__PURE__ */ jsx(
      ButtonProvider,
      {
        provider,
        login: onSubmit,
        loading: loading && loginProvider === provider,
        disabled: loading
      },
      provider
    )) }),
    /* @__PURE__ */ jsx("p", { className: "separator", children: /* @__PURE__ */ jsx(Text, { appearance: "microtext1", colorScheme: "light.700", children: t.or }) }),
    /* @__PURE__ */ jsx(Text, { colorScheme: "light.700", align: "center", children: t.corporateLoginTitle }),
    /* @__PURE__ */ jsx(Button, { size: "md", disabled: loading, colorScheme: "light", onClick: () => onChangeMode("sso"), children: t.corporateLoginButton })
  ] });
};

const lastLoginTypeKey = "lastLoginType";
const fallbackKeys = ["guided-tour", "@stack-spot/opa:user", "CHAT_AGENTS", "RATED_US_IN"];
function getLastLoginType() {
  const type = localStorage.getItem(lastLoginTypeKey);
  if (type === "idp" || type === "sso")
    return type;
  if (getCookie("stk-session.stackspot.com"))
    return "sso";
  for (const key of fallbackKeys) {
    if (localStorage.getItem(key))
      return "sso";
  }
  return "idp";
}
function setLastLoginType(type) {
  localStorage.setItem(lastLoginTypeKey, type);
}

const SSOLogin = ({ value, onChange, loading, disabled, hasProvider, onChangeMode, idpLoginEnabled }) => {
  const t = useTranslation();
  return /* @__PURE__ */ jsxs(Fragment, { children: [
    /* @__PURE__ */ jsxs(Flex, { flexDirection: "column", style: { gap: "4px", marginTop: "4px" }, children: [
      /* @__PURE__ */ jsx(Label, { htmlFor: "email", children: t.label }),
      /* @__PURE__ */ jsx(Input, { id: "email", type: "email", name: "email", value, onChange: (e) => onChange(e.target.value), placeholder: t.placeholder }),
      /* @__PURE__ */ jsx(Button, { colorScheme: "primary", size: "md", style: { marginTop: "12px" }, disabled: disabled || loading, children: loading && !hasProvider ? /* @__PURE__ */ jsx(LoadingCircular, {}) : /* @__PURE__ */ jsx(Text, { children: t.continue }) })
    ] }),
    idpLoginEnabled && /* @__PURE__ */ jsxs(Fragment, { children: [
      /* @__PURE__ */ jsx("p", { className: "separator", children: /* @__PURE__ */ jsx(Text, { appearance: "microtext1", colorScheme: "light.700", children: t.or }) }),
      /* @__PURE__ */ jsx(Text, { colorScheme: "light.700", align: "center", children: t.socialLoginTitle }),
      /* @__PURE__ */ jsx(Button, { size: "md", disabled: loading, colorScheme: "light", onClick: () => onChangeMode("idp"), children: t.socialLogin })
    ] })
  ] });
};

const LoginBox = styled.form`
  display: flex;
  flex-direction: column;
  justify-content: center;
  gap: 24px;

  header {
    display: flex;
    flex-direction: column;
    align-items: center;
    gap: 24px;
  }

  .separator {
    padding: 0 8px;
    display: flex;
    flex-direction: row;
    align-items: center;
    justify-content: center;
    gap: 8px;
    margin: 0;

    &:before, &:after {
      content: '';
      height: 1px;
      flex: 1;
      background-color: ${theme.color.light["500"]};
    }
  }

  .error {
    color: ${theme.color.danger["500"]};
    line-height: 1.5rem;
  }
`;
const EmailNotAllowed = () => {
  const t = useTranslation();
  return /* @__PURE__ */ jsx(Card, { children: /* @__PURE__ */ jsxs(Flex, { justifyContent: "center", children: [
    /* @__PURE__ */ jsx(Text, { appearance: "body2", children: t.emailNotAllowedTitle }),
    /* @__PURE__ */ jsx(Text, { appearance: "body2", colorScheme: "light.700", children: t.emailNotAllowedSubtitle })
  ] }) });
};
const stylesBanner = {
  position: "fixed",
  top: 0,
  left: 0,
  width: "100%"
};
const Login = ({ onSubmit, initialValue = "", showLogin = true, welcomeText, removeLoadingOnSuccess, className, style, banner, loginTypes = ["idp", "sso"] }) => {
  const t = useTranslation();
  const [trialProviders, isLoadingTrialProviders] = useTrialProviders({ enabled: loginTypes.includes("idp") });
  const searchParams = new URLSearchParams(location.search);
  const [error, setError] = useState(searchParams.get("error_description") || searchParams.get("error") || "");
  const [errorCode] = useState(searchParams.get("error_code") || "");
  const [loading, setLoading] = useState(false);
  const providerQueryParam = searchParams.get("provider");
  const [loginProvider, setLoginProvider] = useState();
  const [email, setEmail] = useState(initialValue || searchParams.get("email") || "");
  const disabled = !email.match(/\w+@\w+/);
  const idpLoginEnabled = loginTypes.includes("idp") && !!(trialProviders == null ? void 0 : trialProviders.length);
  const [mode, setMode] = useState();
  useEffect(() => {
    setMode(idpLoginEnabled ? getLastLoginType() : "sso");
  }, [idpLoginEnabled]);
  useEffect(() => {
    if (!providerQueryParam)
      return;
    if (providerQueryParam === "email")
      login("sso");
    else if (trialProviders == null ? void 0 : trialProviders.includes(providerQueryParam))
      login("idp", providerQueryParam);
  }, [trialProviders, isLoadingTrialProviders]);
  async function login(type, provider) {
    setError("");
    setLoading(true);
    provider !== loginProvider && setLoginProvider(provider);
    try {
      const data = type === "idp" && !!provider ? { type: "idp", provider: `external-idp:${provider}` } : { type: "sso", email };
      setLastLoginType(data.type);
      await onSubmit(data);
      if (removeLoadingOnSuccess)
        setLoading(false);
    } catch (error2) {
      setLoading(false);
      setLoginProvider(void 0);
      if (error2 instanceof AuthMethodUnavailable) {
        setError(t.emailNotFoundError);
      } else {
        setError(error2.message || error2.toString());
      }
    }
  }
  useEffect(() => {
    const meta = document.createElement("meta");
    meta.name = "adopt-website-id";
    meta.content = "a32fe656-2333-4dc5-8039-dddac6464587";
    document.head.appendChild(meta);
    const script = document.createElement("script");
    script.src = "//tag.goadopt.io/injector.js?website_code=a32fe656-2333-4dc5-8039-dddac6464587";
    script.className = "adopt-injector";
    script.async = true;
    document.body.appendChild(script);
    return () => {
      document.head.removeChild(meta);
      document.body.removeChild(script);
    };
  }, []);
  function submitForm(e) {
    e.preventDefault();
    if (disabled)
      return;
    login("sso");
  }
  if (isLoadingTrialProviders || !mode) {
    return /* @__PURE__ */ jsxs(Flex, { alignContent: "center", justifyContent: "center", my: 5, children: [
      /* @__PURE__ */ jsx(LoadingCircular, {}),
      " "
    ] });
  }
  const loginWithSocialAccount = /* @__PURE__ */ jsxs(Fragment, { children: [
    /* @__PURE__ */ jsx("span", { children: t.loginWithSocialAccount1 }),
    /* @__PURE__ */ jsx("span", { style: { fontWeight: "bold" }, children: t.loginWithSocialAccount2 }),
    /* @__PURE__ */ jsx("span", { children: t.loginWithSocialAccount3 })
  ] });
  return /* @__PURE__ */ jsxs(Fragment, { children: [
    banner && /* @__PURE__ */ jsx(Box, { style: stylesBanner, children: /* @__PURE__ */ jsx(BannerWarning, { showCloseButton: false, children: banner }) }),
    showLogin ? /* @__PURE__ */ jsxs(LoginBox, { onSubmit: submitForm, className, style, children: [
      /* @__PURE__ */ jsxs("header", { children: [
        /* @__PURE__ */ jsx(MiniLogo, {}),
        /* @__PURE__ */ jsxs(Flex, { flexDirection: "column", alignItems: "center", children: [
          /* @__PURE__ */ jsx(Text, { appearance: "body1", weight: "medium", children: welcomeText || t.welcome }),
          /* @__PURE__ */ jsx(Text, { appearance: "body2", colorScheme: "light.700", align: "center", children: mode === "idp" ? loginWithSocialAccount : t.loginWithEmail })
        ] })
      ] }),
      errorCode && errorCode === "EMAIL_IS_NOT_ALLOWED" && /* @__PURE__ */ jsx(EmailNotAllowed, {}),
      mode === "sso" ? /* @__PURE__ */ jsx(
        SSOLogin,
        {
          disabled,
          loading,
          hasProvider: !loginProvider,
          value: email,
          onChange: setEmail,
          onChangeMode: setMode,
          idpLoginEnabled
        }
      ) : /* @__PURE__ */ jsx(
        IDPLogin,
        {
          loading,
          loginProvider,
          onSubmit: login,
          trialProviders,
          onChangeMode: setMode
        }
      ),
      error && /* @__PURE__ */ jsx(Text, { className: "error", align: "center", children: error })
    ] }) : /* @__PURE__ */ jsxs(Flex, { flexDirection: "column", alignItems: "center", children: [
      /* @__PURE__ */ jsx(MiniLogo, {}),
      /* @__PURE__ */ jsx(Text, { appearance: "body1", weight: "medium", children: welcomeText || t.welcome }),
      /* @__PURE__ */ jsx(Text, { mt: 4, children: t.loginTemporarilyUnavailable })
    ] })
  ] });
};

var __defProp = Object.defineProperty;
var __defProps = Object.defineProperties;
var __getOwnPropDescs = Object.getOwnPropertyDescriptors;
var __getOwnPropSymbols = Object.getOwnPropertySymbols;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __propIsEnum = Object.prototype.propertyIsEnumerable;
var __defNormalProp = (obj, key, value) => key in obj ? __defProp(obj, key, { enumerable: true, configurable: true, writable: true, value }) : obj[key] = value;
var __spreadValues = (a, b) => {
  for (var prop in b || (b = {}))
    if (__hasOwnProp.call(b, prop))
      __defNormalProp(a, prop, b[prop]);
  if (__getOwnPropSymbols)
    for (var prop of __getOwnPropSymbols(b)) {
      if (__propIsEnum.call(b, prop))
        __defNormalProp(a, prop, b[prop]);
    }
  return a;
};
var __spreadProps = (a, b) => __defProps(a, __getOwnPropDescs(b));
const Authenticated = ({ children, onLogin, onSession, customLoginProps, sessionManager, onChangeStatus }) => {
  const [authStatus, setAuthStatus] = useState("unknown");
  const language = useLanguage(ptEn);
  sessionManager != null ? sessionManager : sessionManager = SessionManager.instance;
  if (!sessionManager)
    throw new Error("Please, provide a sessionManager");
  useEffectOnce(() => {
    async function checkAuth() {
      if (!sessionManager)
        throw new Error("Please, provide a sessionManager");
      await sessionManager.restoreSession();
      if (sessionManager.urlHasThirdPartyLoginData()) {
        await sessionManager.completeThirdPartyLogin();
        onLogin == null ? void 0 : onLogin();
      }
      if (sessionManager.hasSession()) {
        setAuthStatus("authenticated");
        onSession == null ? void 0 : onSession();
        onChangeStatus == null ? void 0 : onChangeStatus("authenticated");
      } else {
        setAuthStatus("unauthenticated");
        onChangeStatus == null ? void 0 : onChangeStatus("unauthenticated");
      }
    }
    checkAuth();
  });
  if (authStatus === "unknown")
    return null;
  if (authStatus === "authenticated")
    return children;
  return /* @__PURE__ */ jsx(CSSToCitricAdapter, { children: /* @__PURE__ */ jsx(Flex, { justifyContent: "center", alignItems: "center", flex: 1, style: { height: "100%" }, children: /* @__PURE__ */ jsx(
    Login,
    __spreadValues({
      style: { width: "360px" },
      onSubmit: (data) => sessionManager.startThirdPartyLogin(__spreadProps(__spreadValues({}, data), { locale: language })),
      initialValue: sessionManager.getEmailForLogin()
    }, customLoginProps || {})
  ) }) });
};

export { Authenticated, Login, SessionManager, useSession };
