// @stack-spot/portal-components/dist/components/Select/CloseItem.js
import { jsx as _jsx } from "react/jsx-runtime";
import { IconBox } from '@citric/core';
import { TimesMini } from '@citric/icons';
import { IconButton } from '@citric/ui';
import { interpolate, useTranslate } from '@stack-spot/portal-translate';
import { components } from 'react-select';
export const CloseItem = (props) => {
    const t = useTranslate(dictionary);
    return (_jsx(components.MultiValueRemove, { ...props, children: _jsx(IconButton, { type: "button", "aria-label": interpolate(t.ariaLabelRemoveItemButton, props.data.value), sx: {
                bg: 'transparent',
                border: 'none',
                px: 0,
                w: 'auto',
                height: 'auto',
                '&:hover, &:focus': { bg: 'transparent' },
            }, children: _jsx(IconBox, { size: "xs", sx: { px: 0 }, children: _jsx(TimesMini, {}) }) }) }));
};
const dictionary = {
    en: {
        ariaLabelRemoveItemButton: 'Remove item $0',
    },
    pt: {
        ariaLabelRemoveItemButton: 'Remover item $0',
    },
};
