// @stack-spot/portal-components/dist/components/Select/MultiValue.js
import { jsx as _jsx, Fragment as _Fragment } from "react/jsx-runtime";
import { TimesMini } from '@citric/icons';
import { Badge, IconButton } from '@citric/ui';
import { interpolate, useTranslate } from '@stack-spot/portal-translate';
import { components } from 'react-select';
export const MultiValue = (props) => {
    const t = useTranslate(dictionary);
    return (_jsx(_Fragment, { children: _jsx(components.MultiValue, { ...props, children: _jsx(Badge, { appearance: "square", className: "badge", palette: "cyan", sx: {
                    mr: 0,
                    my: 0,
                    '.badge-citric-container:not(:has(> button:first-child))': {
                        marginRight: 0,
                    },
                }, afterElement: _jsx(components.MultiValueRemove, { ...props, children: _jsx(IconButton, { type: "button", "aria-label": interpolate(t.ariaLabelRemoveItemButton, props.children), appearance: "square", colorIcon: "cyan.800", onClick: (e) => {
                            e.stopPropagation();
                            props.removeProps?.onClick?.(e);
                        }, children: _jsx(TimesMini, {}) }) }), children: props.children }) }) }));
};
const dictionary = {
    en: {
        ariaLabelRemoveItemButton: 'Remove item $0',
    },
    pt: {
        ariaLabelRemoveItemButton: 'Remover item $0',
    },
};
