// @stack-spot/portal-components/dist/components/Select/SelectInfiniteScroll.js
import { jsx as _jsx } from "react/jsx-runtime";
import { IconBox } from '@citric/core';
import { LoadingCircular } from '@citric/ui';
import InfiniteScroll from 'react-infinite-scroll-component';
import { components } from 'react-select';
import { SelectSearch } from './SelectSearch.js';
const MenuList = (props) => {
    const { children, selectProps } = props;
    const { loadMore, hasMore, options, isLoadingMore } = selectProps;
    return (_jsx(InfiniteScroll, { dataLength: options.length, next: loadMore, hasMore: hasMore, scrollableTarget: "list-item", scrollThreshold: 0.8, loader: isLoadingMore ? _jsx(IconBox, { w: 12, colorBg: "light.300", sx: { alignItems: 'center', justifyContent: 'center' }, children: _jsx(LoadingCircular, {}) }) : null, style: { overflow: 'hidden' }, children: _jsx(components.MenuList, { ...props, innerProps: { ...props.innerProps, id: 'list-item' }, children: children }) }));
};
export const SelectInfiniteScroll = ({ options, loadMore, hasMore, isLoadingMore, ...props }) => (_jsx(SelectSearch, { components: { MenuList }, options: options, closeMenuOnSelect: false, styles: {
        control: () => ({
            minHeight: '2.5rem',
            height: 'auto',
        }),
        menu: () => ({
            position: 'absolute',
        }),
        menuList: (base) => ({
            ...base,
            maxHeight: 300,
            overflowY: 'auto',
        }),
        input: () => ({
            position: 'relative',
        }),
    }, 
    // @ts-ignore: Essas props não existem no tipo, mas são acessíveis via selectProps no MenuList
    loadMore: loadMore, hasMore: hasMore, isLoadingMore: isLoadingMore, ...props }));
