// @stack-spot/portal-mfe-utils/src/index.ts
/* eslint-disable no-console */
import { PrefetchMetadata } from './types'

/**
 * The hashing algorithm was taken from:
 * https://stackoverflow.com/a/25105589/1147777
 */

interface Options {
  moduleUrl: string,
  prefetchFile?: string,
  localStoragePrefix?: string,
}

const MAX_CHAR_SUM = 65027
const LOCAL_STORAGE_PREFIX = 'prefetch'
const REQUEST_GROUP_MAX_SIZE = 10
const DELAY_BETWEEN_REQUEST_GROUPS_MS = 500

function charSum(str: string) {
  let i, sum = 0
  for (i = 0; i < str.length; i++) {
    sum += (str.charCodeAt(i) * (i+1))
  }
  return sum
}

function computeHash(strArray: string[]) {
  let i, sum = 0
  for (i = 0; i < strArray.length; i++) {
    const cs = charSum(strArray[i])
    sum = sum + (MAX_CHAR_SUM / cs)
  }
  return (`${sum}`).slice(0, 16)
}

function delay(ms: number) {
  return new Promise((resolve) => {
    setTimeout(resolve, ms)
  })
}

async function prefetchInGroups(base: string, files: string[]) {
  let result: PromiseSettledResult<Response>[] = []
  let group: Promise<Response>[] = []
  for (const file of files) {
    group.push(fetch(`${base}${file}`, { priority: 'low' }))
    if (group.length === REQUEST_GROUP_MAX_SIZE) {
      result = [...result, ...(await Promise.allSettled(group))]
      await delay(DELAY_BETWEEN_REQUEST_GROUPS_MS)
      group = []
    }
  }
  result = [...result, ...(await Promise.allSettled(group))]
  const success = result.filter(i => i.status === 'fulfilled').length
  return { success, failure: result.length - success, total: result.length }
}

export async function prefetch({
  moduleUrl,
  prefetchFile = 'prefetch.json',
  localStoragePrefix = LOCAL_STORAGE_PREFIX,
}: Options) {
  moduleUrl = moduleUrl.replace(/\/$/, '')
  const localStorageId = `${localStoragePrefix}-${moduleUrl}`
  try {
    const response = await fetch(`${moduleUrl}/${prefetchFile}`, { priority: 'low' })
    if (!response.ok) throw response
    const metadata: PrefetchMetadata = await response.json()
    const hash = computeHash([metadata.assetsDir, ...metadata.files])
    if (localStorage.getItem(localStorageId) === hash) return
    const base = metadata.assetsDir ? `${moduleUrl}/${metadata.assetsDir}/` : `${moduleUrl}/`
    // allows for most files to be prefetched even if the user doesn't usually stay connected long enough to prefetch everything
    const files = Math.random() > 0.5 ? metadata.files : metadata.files.reverse()
    const { failure, success, total } = await prefetchInGroups(base, files)
    if (success) console.log(`Prefetched ${success} files from ${moduleUrl}.`)
    if (success === total) localStorage.setItem(localStorageId, hash)
    else console.warn(`Failed to prefetch ${failure} files from ${moduleUrl}.`)
  } catch (error: any) {
    if (error instanceof Response) {
      console.warn(`Could not prefetch ${moduleUrl}: ${error.status} while loading json file.`)
    } else {
      console.warn(`Could not prefetch ${moduleUrl}.`, error.message || `${error}`)
    }
  }
}
