// @stack-spot/citric-react/dist/components/ErrorBoundary.js
import { jsx as _jsx } from "react/jsx-runtime";
import { Component } from 'react';
import { CitricContext } from '../context/CitricContext.js';
import { ErrorMessage } from './ErrorMessage.js';
/**
 * An Error Boundary that renders an error feedback instead of its content if any of its children throws.
 *
 * You can customize the appearance of the error through the function `renderError` of a CitricController.
 *
 * Errors can be watched through the function `onError` of a CitricController.
 *
 * Attention: if you're using React Suspense, consider using the component "FallbackBoundary" instead.
 *
 * @example
 * ```
 * <ErrorBoundary>
 *   {content}
 * </ErrorBoundary>
 * ```
 */
export class ErrorBoundary extends Component {
    constructor(props) {
        super(props);
        this.state = {};
    }
    static getDerivedStateFromError(error) {
        return { error };
    }
    componentDidCatch(error, errorInfo) {
        this.context?.onError?.(error);
        // eslint-disable-next-line no-console
        console.error(error, errorInfo);
    }
    componentDidUpdate(prevProps) {
        if (this.state.error && this.props.children !== prevProps.children) {
            this.setState({ error: null });
        }
    }
    renderCustomErrorUI() {
        return this.context?.renderError?.(this.state.error);
    }
    renderErrorUI() {
        if (typeof this.props.message === 'function')
            return this.props.message(this.state.error);
        return (this.props.message || !this.context?.renderError)
            ? _jsx(ErrorMessage, { error: this.props.message || this.state.error })
            : this.renderCustomErrorUI();
    }
    render() {
        return this.state.error ? this.renderErrorUI() : this.props.children;
    }
}
Object.defineProperty(ErrorBoundary, "contextType", {
    enumerable: true,
    configurable: true,
    writable: true,
    value: CitricContext
});
