// @stack-spot/citric-react/dist/components/ImageBox.js
import { jsx as _jsx } from "react/jsx-runtime";
import { listToClass } from '@stack-spot/portal-theme';
import { useCitricController } from '../context/hooks.js';
import { withRef } from '../utils/react.js';
import { asCitricProps, CitricComponent } from './CitricComponent.js';
/**
 * Renders a wrapper for its child (normally an image). The image will be resized and cropped to fit the container. The image is not cropped
 * if the property "feedback" is set.
 *
 * Hover and focus effects are applied if the ImageBox is focusable.
 *
 * This works exactly like an IconBox, but instead of an Icon, it can contain anything.
 *
 * @example
 * ```
 * <ImageBox><img src="https://images.com/myimage.png" /></ImageBox>
 * ```
 */
export const ImageBox = withRef(function ImageBox({ tag, appearance, size, className, metadata, onClick, feedback, ...props }) {
    props['aria-label'] ||= props.title; // accessibility
    const citric = useCitricController();
    function handleClick(e) {
        onClick?.(e);
        if (tag === 'button')
            citric?.onClickButton?.(e, metadata);
        else if (tag === 'a')
            citric?.onClickLink?.(e, metadata);
    }
    const componentProps = {
        tag: (tag || 'i'),
        component: 'icon-box',
        className: listToClass([appearance, size, className]),
        'data-feedback': feedback || undefined,
        onClick: ['button', 'a'].includes(tag ?? '') ? handleClick : onClick,
        ...props,
    };
    return tag === 'a' && citric?.renderLink
        ? citric.renderLink(asCitricProps(componentProps))
        : _jsx(CitricComponent, { ...componentProps });
});
/**
 * A shortcut for `<ImageBox tag="button">`.
 *
 * Whenever a button is clicked, the function `onClickButton` of the nearest CitricController is called with the event and the value of the
 * prop `metadata`.
 *
 * @example
 * ```
 * <ImageButton><img src="https://images.com/myimage.png" /></ImageButton>
 * ```
 */
export const ImageButton = withRef(function ImageButton(props) {
    return _jsx(ImageBox, { ...props, tag: "button", type: props.type || 'button' });
});
/**
 * A shortcut for `<ImageBox tag="a">`.
 *
 * Whenever a link is clicked, the function `onClickLink` of the nearest CitricController is called with the event and the value of the
 * prop `metadata`.
 *
 * @example
 * ```
 * <ImageLink href="#"><img src="https://images.com/myimage.png" /></ImageButton>
 * ```
 */
export const ImageLink = withRef(function ImageLink(props) {
    return _jsx(ImageBox, { ...props, tag: "a" });
});
