// @stack-spot/citric-react/dist/components/layout.js
import { listToClass } from '@stack-spot/portal-theme';
import { createElement } from 'react';
import { applyStyles } from '../utils/css.js';
import { withRef } from '../utils/react.js';
export const layout = {
    center: 'center',
    row: 'row',
    column: 'column',
};
/**
 * Centers the content both vertically and horizontally. Elements are placed in a column (flex).
 *
 * This element accepts many style shortcuts as attributes: bg, fg, border, radius, justifyContent, alignItems, flex, gap, m, mt, mb, ml,
 * mr, p, pt, pb, pl, pr, w, h.
 *
 * @example
 * ```
 * <Center w="200px" h="200px" gap="10px">
 *   <img src="/an-image.png" />
 *   <Text>This content is centered!</Text>
 * </Center>
 * ```
 */
export const Center = withRef(function Center({ tag, className, ...props }) {
    return createElement(tag || 'div', { className: listToClass(['center', className]), ...applyStyles(props) });
});
/**
 * Creates a row (flex). The elements in this row are horizontally centered unless `center = false`.
 *
 * @example
 * ```
 * <Row gap={3}>
 *   <Icon icon="Stack" />
 *   <Text>Stacks</Text>
 * </Row>
 * ```
 */
export const Row = withRef(function Row({ tag, className, center, ...props }) {
    return createElement(tag || 'div', { className: listToClass(['row', center === false && 'no-center', className]), ...applyStyles(props) });
});
/**
 * Creates a column (flex). The elements in this row are placed at the start of the horizontal axis, unless `center = true`, in this case,
 * they're centered.
 *
 * @example
 * ```
 * <Column gap={5}>
 *   <Text>Item 1</Text>
 *   <Text>Item 2</Text>
 *   <Text>Item 3</Text>
 * </Row>
 * ```
 */
export const Column = withRef(function Column({ tag, className, center, ...props }) {
    return createElement(tag || 'div', { className: listToClass(['column', center && 'center', className]), ...applyStyles(props) });
});
