// @stack-spot/citric-react/dist/components/Pagination.js
import { jsx as _jsx, jsxs as _jsxs } from "react/jsx-runtime";
import { interpolate, useTranslate } from '@stack-spot/portal-translate';
import { useMemo } from 'react';
import { withRef } from '../utils/react.js';
import { CitricComponent } from './CitricComponent.js';
import { IconButton } from './IconBox.js';
const DEFAULT_PAGE = 1;
const DEFAULT_PAGE_SIZE_OPTIONS = [10, 20, 30];
/**
 * Renders a pagination UI, letting the user chose one among multiple pages. This is generally rendered at the bottom of a table.
 *
 * @example
 * ```
 * const [pageData, setPageData] = useState({ page: 1, size: 10 })
 * return <Pagination
 *   value={pageData}
 *   totalPages={50}
 *   onChange={setPageData}
 * />
 * ```
 */
export const Pagination = withRef(({ value, onChange, totalPages, pageSizeOptions = DEFAULT_PAGE_SIZE_OPTIONS, ...props }) => {
    const t = useTranslate(dictionary);
    const sizeOptions = useMemo(() => pageSizeOptions.map(o => _jsx("option", { selected: value.size === o, children: o }, o)), [pageSizeOptions]);
    const pageOptions = useMemo(() => {
        const options = [];
        for (let i = 1; i <= totalPages; i++) {
            options.push(_jsx("option", { value: i, selected: value.page === i, children: i }, i));
        }
        return options;
    }, [value.page, totalPages]);
    return (_jsxs(CitricComponent, { tag: "div", component: "pagination", ...props, children: [_jsx("div", { className: "page-size", children: _jsxs("label", { children: [t.itemsPerPage, ":", _jsx("select", { name: "itemsPerPage", onChange: e => onChange({ page: DEFAULT_PAGE, size: parseInt(e.target.value) }), children: sizeOptions })] }) }), _jsxs("div", { className: "page-number", children: [_jsx("label", { children: _jsx("select", { name: "page", onChange: e => onChange({ page: parseInt(e.target.value), size: value.size }), value: value.page || DEFAULT_PAGE, children: pageOptions }) }), totalPages > 1 ? interpolate(t.ofTotalPlural, totalPages) : t.ofTotalSingular, _jsx(IconButton, { icon: "ChevronLeft", "aria-label": "previous", title: "previous", disabled: value.page === DEFAULT_PAGE, onClick: () => onChange({ page: Math.max(value.page - 1, DEFAULT_PAGE), size: value.size }) }), _jsx(IconButton, { icon: "ChevronRight", "aria-label": "next", title: "next", disabled: value.page === totalPages, onClick: () => onChange({ page: Math.min(totalPages, value.page + 1), size: value.size }) })] })] }));
});
const dictionary = {
    en: {
        itemsPerPage: 'Items per page',
        ofTotalSingular: 'of 1 page',
        ofTotalPlural: 'of $0 pages',
    },
    pt: {
        itemsPerPage: 'Itens por página',
        ofTotalSingular: 'de 1 página',
        ofTotalPlural: 'de $0 páginas',
    },
};
