// @stack-spot/citric-react/dist/components/SmartTable.js
import { jsx as _jsx, jsxs as _jsxs } from "react/jsx-runtime";
import { createElement as _createElement } from "react";
import { useMemo } from 'react';
import { withRef } from '../utils/react.js';
import { Table, Th, Tr } from './Table.js';
function createRowClickListener(elementQuerySelector) {
    return (event) => {
        const target = event.target instanceof HTMLElement ? event.target : undefined;
        const row = target instanceof HTMLTableRowElement ? target : target?.closest('tr');
        if (row) {
            const element = row.querySelector(elementQuerySelector);
            if (element instanceof HTMLElement)
                element.click();
        }
    };
}
/**
 * Renders a table. Always prefer using this component over the raw Table component.
 *
 * This receives a dataset and renders each item as a row. The columns are determined by the property "columns".
 *
 * To change the overall table style, use the property "appearance".
 *
 * To check more complex table examples, like tables with accordions and sorting, check the storybook.
 *
 * @example
 *
 * ```
 * const data = useMemo(() => [
 *   { id: 1, name: 'Yuri Tyson', phone: '1-801-475-4561', country: 'Poland', age: 21 },
 *   { id: 2, name: 'Amal Mcclure', phone: '(695) 948-4102', country: 'Ukraine', age: 32 },
 *   { id: 3, name: 'Levi Glass', phone: '1-976-544-4872', country: 'Colombia', age: 45 },
 * ], [])
 * const columns = useMemo(() => [
 *   { key: 'name', label: 'Name', render: item => `${item.name} (${item.age})` },
 *   { key: 'phone', label: 'Phone' },
 *   { key: 'country', label: 'Country' },
 *   {
 *      key: 'settings',
 *      render: item => (
 *        <Row gap={2}>
 *          <IconButton icon="ExternalLink" title="View" />
 *          <IconButton icon="Pencil" title="Edit" />
 *          <IconButton icon="Trash" title="Delete" colorScheme="danger" />
 *        </Row>
 *      ),
 *   },
 * ], [])
 * return <SmartTable data={data} columns={columns} />
 * ```
 */
export const SmartTable = withRef(function SmartTable({ data, keygen, columns, renderAccordion, accordionTrigger, accordionMaxHeight, id, rowClickElement, headless, ...props }) {
    id = useMemo(() => id || `${Math.random()}`, [id]);
    const headers = useMemo(() => headless ? [] : columns.map((c) => (_createElement(Th, { ...c.th, key: c.key, onSort: c.onSort, direction: c.direction, tabIndex: c.onSort ? 0 : undefined }, c.label ?? c.key))), [columns, headless]);
    const rows = useMemo(() => data.map((item, index) => {
        const key = keygen ? keygen(item) : index;
        const accordionContent = renderAccordion?.(item);
        const row = (_jsxs(Tr, { accordionTrigger: accordionContent ? accordionTrigger : undefined, onClick: rowClickElement ? createRowClickListener(rowClickElement) : undefined, children: [columns.map(c => _createElement("td", { ...c.td, key: c.key }, c.render ? c.render(item) : (item[c.key] || ''))), renderAccordion && (_jsx("td", { children: accordionContent && (_jsx("input", { type: "checkbox", "aria-controls": `${id}-${key}`, onKeyDown: e => e.key === 'Enter' && e.target instanceof HTMLElement && e.target.click() })) }))] }, key));
        return renderAccordion ? (_jsxs("tbody", { children: [row, _jsx(Tr, { id: `${id}-${key}`, accordion: true, accordionMaxHeight: accordionMaxHeight, children: _jsx("td", { colSpan: columns.length + 1, children: _jsx("div", { children: accordionContent }) }) })] }, `${key}-group`)) : row;
    }), [columns, data, !!renderAccordion, accordionTrigger]);
    return (_jsxs(Table, { id: id, accordionRows: !!renderAccordion, ...props, children: [!headless && _jsx("thead", { children: _jsxs("tr", { children: [headers, renderAccordion && _jsx("th", {})] }) }), renderAccordion ? rows : _jsx("tbody", { children: rows })] }));
});
