// @stack-spot/citric-react/dist/components/Stepper.js
import { jsx as _jsx, jsxs as _jsxs } from "react/jsx-runtime";
import { listToClass } from '@stack-spot/portal-theme';
import { useTranslate } from '@stack-spot/portal-translate';
import { useCallback, useEffect, useMemo, useState } from 'react';
import { withRef } from '../utils/react.js';
import { Button } from './Button.js';
import { Column, Row } from './layout.js';
import { Tabs } from './Tabs/index.js';
import { TabController } from './Tabs/TabController.js';
import { findSelectedIndex } from './Tabs/utils.js';
function getTabsWithDisabled(tabs, value) {
    let index = findSelectedIndex(tabs, value ?? '');
    if (index < 0)
        index = 0;
    return tabs.map((t, i) => ({ ...t, disabled: i > index }));
}
/**
 * A Stepper is a tab view with a different appearance. To control the current tab (step), retrieve the controller by calling
 * `useTabsController()` from within a tab (step) content.
 *
 * Besides all the properties in "Tabs", the "Stepper" can create a set of "Previous" and "Next" buttons through the property "buttons". By
 * default, the buttons are rendered, to disable them, pass `buttons = false`.
 *
 * @example
 *
 * ```
 * const steps: Tab[] = useMemo(() => [
 *   { key: 'step1', label: 'Step 1', content: <><h1>First step</h1><p>Content of the first step</p></> },
 *   { key: 'step2', label: 'Step 2', content: <><h1>Second step</h1><p>Content of the second step</p></> },
 *   { key: 'step3', label: 'Step 3', content: <><h1>Third step</h1><p>Content of the third step</p></> },
 * ], [])
 *
 * return <Stepper tabs={steps} />
 * ```
 */
export const Stepper = withRef(function Stepper({ tabs: initialTabs, controller: ctrl, value, onChange, buttons = true, className, ...props }) {
    const controller = useMemo(() => ctrl ?? new TabController(initialTabs.map(t => t.key), value || initialTabs[0]?.key), []);
    const [tabs, setTabs] = useState(getTabsWithDisabled(initialTabs, value));
    const [selectedIndex, setSelectedIndex] = useState(findSelectedIndex(tabs, controller.getValue()));
    const t = useTranslate(dictionary);
    useEffect(() => controller.onChange((v) => {
        setSelectedIndex(findSelectedIndex(tabs, v));
    }), [tabs]);
    useEffect(() => controller.onChange((value) => {
        setTabs(getTabsWithDisabled(initialTabs, value));
    }), []);
    const onPrevious = useCallback(() => {
        controller.previous();
        if (typeof buttons === 'object')
            buttons.onPrevious?.(controller.getValue());
    }, []);
    const onNext = useCallback(() => {
        controller.next();
        if (typeof buttons === 'object')
            buttons.onNext?.(controller.getValue());
    }, []);
    return buttons ? (_jsxs(Column, { ...props, className: className, gap: "20px", children: [_jsx(Tabs, { tabs: tabs, controller: controller, value: value, onChange: onChange, className: "stepper" }), _jsxs(Row, { justifyContent: (typeof buttons !== 'object' || !buttons.onCancel) && selectedIndex === 0 ? 'end' : 'space-between', children: [selectedIndex === 0 && typeof buttons === 'object' && buttons.onCancel && (_jsx(Button, { onClick: buttons.onCancel, colorScheme: "inverse", appearance: "outlined", children: buttons.cancel || t.cancel })), selectedIndex > 0 && buttons && (_jsx(Button, { onClick: onPrevious, colorScheme: "inverse", appearance: "outlined", children: (typeof buttons === 'object' && buttons.previous) || t.previous })), selectedIndex < tabs.length - 1 && buttons && (_jsx(Button, { onClick: onNext, children: (typeof buttons === 'object' && buttons.next) || t.next })), selectedIndex === tabs.length - 1 && typeof buttons === 'object' && buttons.onFinish && (_jsx(Button, { onClick: buttons.onFinish, children: buttons.finish || t.finish }))] })] })) : _jsx(Tabs, { tabs: tabs, controller: controller, value: value, onChange: onChange, className: listToClass([className, 'stepper']), ...props });
});
const dictionary = {
    en: {
        cancel: 'Cancel',
        previous: 'Previous',
        next: 'Next',
        finish: 'Finish',
    },
    pt: {
        cancel: 'Cancelar',
        previous: 'Anterior',
        next: 'Próximo',
        finish: 'Finalizar',
    },
};
