// @stack-spot/citric-react/dist/components/Table.js
import { jsx as _jsx } from "react/jsx-runtime";
import { listToClass } from '@stack-spot/portal-theme';
import { useTranslate } from '@stack-spot/portal-translate';
import { useEffect, useRef } from 'react';
import { applyCSSVariable } from '../utils/css.js';
import { withRef } from '../utils/react.js';
import { CitricComponent } from './CitricComponent.js';
/**
 * Renders an HTML table. Use its props for customizing the appearance.
 *
 * - If you need to use Accordion rows, use `<Tr>`instead of `<tr>`.
 * - If you need to sort columns, use `<Th>`instead of `<th>`.
 *
 * This works exactly like the HTML tag "table".
 *
 * Attention: prefer using the component "SmartTable". Use this only if you need full control over the table.
 */
export const Table = withRef(({ appearance, stripped, compressed, showBorders, showHeaderBorders, showRowBorders, rounded, roundedRows, uppercaseHeader, className, children, accordionRows, ...props }) => {
    const classes = listToClass([
        className,
        stripped && 'stripped',
        compressed && 'compressed',
        showBorders && 'bordered',
        showHeaderBorders === false && 'borderless-header',
        showRowBorders && 'bordered-rows',
        rounded === false && 'square',
        roundedRows === false && 'square-rows',
        uppercaseHeader && 'uppercase-header',
        accordionRows && 'accordion-rows',
    ]);
    return (_jsx(CitricComponent, { tag: "table", component: "table", "data-appearance": appearance, className: classes, ...props, children: children }));
});
export const Th = withRef(({ direction, onSort, children, className, ...props }) => {
    const t = useTranslate(dictionary);
    let tip;
    const label = typeof children === 'string' ? children : undefined;
    if (onSort && !direction)
        tip = label ? `${label}. ${t.sortAscending}` : t.sortAscending;
    if (onSort && direction === 'asc')
        tip = label ? `${label}. ${t.sortDescending}` : t.sortDescending;
    if (onSort && direction === 'desc')
        tip = label ? `${label}. ${t.sortNone}` : t.sortNone;
    function handleSort() {
        if (!direction)
            onSort?.('asc');
        else if (direction === 'asc')
            onSort?.('desc');
        else
            onSort?.(undefined);
    }
    return (_jsx("th", { className: listToClass([onSort && 'sortable', direction, className]), onClick: onSort ? handleSort : undefined, onKeyDown: onSort ? (e => e.key === 'Enter' && handleSort()) : undefined, "aria-label": tip, tabIndex: onSort ? 0 : undefined, ...props, children: children }));
});
export const Tr = withRef(({ ref: outerRef, accordion, accordionTrigger, accordionMaxHeight, children, className, style, onClick, ...props }) => {
    const innerRef = useRef(null);
    const ref = outerRef ?? innerRef;
    useEffect(() => {
        if (!accordion)
            return;
        const checkbox = ref.current?.closest('tbody')?.querySelector('td:last-child input[aria-controls]');
        if (checkbox instanceof HTMLElement) {
            const onChange = (e) => {
                if (!ref.current || !(e.target instanceof HTMLInputElement))
                    return;
                if (e.target.checked) {
                    ref.current.setAttribute('aria-hidden', 'false');
                    ref.current.removeAttribute('inert');
                }
                else {
                    ref.current.setAttribute('aria-hidden', 'true');
                    ref.current.setAttribute('inert', 'true');
                }
            };
            checkbox.addEventListener('change', onChange);
            return () => checkbox.removeEventListener('change', onChange);
        }
    }, [ref.current]);
    return (_jsx("tr", { ref: ref, ...props, className: listToClass([className, accordion && 'accordion', (accordionTrigger === 'row' || onClick) && 'clickable']), onClick: accordionTrigger === 'row'
            ? (e) => {
                const checkbox = ref.current?.querySelector('td:last-child input[aria-controls]');
                if (checkbox instanceof HTMLInputElement && e.target !== checkbox)
                    checkbox.click();
                onClick?.(e);
            }
            : onClick, style: applyCSSVariable(style, 'max-height', accordionMaxHeight), "aria-hidden": accordion, ...(accordion ? { inert: 'true' } : {}), children: children }));
});
const dictionary = {
    en: {
        sortAscending: 'Click to sort in ascending order.',
        sortDescending: 'Click to sort in descending order.',
        sortNone: 'Click to remove sorting.',
    },
    pt: {
        sortAscending: 'Clique para ordenar em ordem crescente.',
        sortDescending: 'Clique para ordenar em ordem decrescente.',
        sortNone: 'Clique para remover a ordenação.',
    },
};
