// @stack-spot/citric-react/dist/components/Text.js
import { listToClass } from '@stack-spot/portal-theme';
import { createElement } from 'react';
import { applyColor, applyTextAppearance } from '../utils/css.js';
import { withRef } from '../utils/react.js';
const tagToAppearance = {
    h1: 'h1',
    h2: 'h2',
    h3: 'h3',
    h4: 'h4',
    h5: 'h5',
    h6: 'h6',
    span: undefined,
    small: 'microtext1',
    p: 'body2',
    pre: 'code2',
};
const appearanceToTag = {
    h1: 'h1',
    h2: 'h2',
    h3: 'h3',
    h4: 'h4',
    h5: 'h5',
    h6: 'h6',
    body1: 'p',
    body2: 'p',
    code1: 'pre',
    code2: 'pre',
    display1: 'h1',
    microtext1: 'small',
    overheader1: 'p',
    overheader2: 'p',
    subtitle1: 'p',
    subtitle2: 'p',
    subtitle3: 'p',
    subtitle4: 'p',
};
/**
 * Renders a text. The tag can be set by the property "tag". If unset, the tag rendered will be based on the property "appearance". If
 * "appearance" is also unset, the tag "p" is rendered.
 *
 * Texts have no margin by default, to enable margins, use `showMargins = true`.
 *
 * If the appearance is not set, it will be inherited from the parent.
 *
 * A font can be applied to a text without this component, just use `{ font: theme.font.$textAppearance }` in your style.
 *
 * @example
 * ```
 * <Text appearance="body2">Hello World!</Text>
 * ```
 */
export const Text = withRef(function Text({ tag, appearance, color, showMargins, weight, align, nowrapEllipsis, style, className, children, ...props }) {
    const renderedTag = tag || appearanceToTag[appearance || 'body2'];
    const renderedAppearance = appearance || (tag ? tagToAppearance[tag] : undefined);
    if (!appearance && !showMargins)
        style = { margin: 0, ...style };
    if (weight)
        style = { fontWeight: weight, ...style };
    if (align)
        style = { textAlign: align, ...style };
    if (nowrapEllipsis)
        style = { whiteSpace: 'nowrap', overflow: 'hidden', textOverflow: 'ellipsis', ...style };
    return createElement(renderedTag, {
        className: applyTextAppearance(listToClass([className, showMargins && 'text-with-margins']), renderedAppearance),
        style: applyColor(style, color),
        ...props,
    }, children);
});
