// @stack-spot/portal-network/dist/api-addresses.js
import apisItauJson from './apis-itau.json' assert { type: 'json' };
import apisJson from './apis.json' assert { type: 'json' };
import { NetworkClient } from './network/NetworkClient.js';
let defaultApis = apisJson;
let overrideApis = {};
const apis = {
    'stackspot': apisJson,
    'itau': apisItauJson,
};
/**
 * Returns the API catalog for the given tenant.
 *
 * @param tenant The current tenant (e.g., 'stackspot', 'itau').
 * @returns API catalog containing the URLs per environment.
 */
export function getApisByTenant(tenant) {
    return apis[tenant];
}
function transformApisToApiAddress(apis, env) {
    const apiAddress = {};
    for (const api in apis) {
        const key = api;
        apiAddress[key] = apis[key].url[env];
    }
    return apiAddress;
}
/**
 * Searchs the default APIs configuration for the API object that matches the given baseURL.
 * @param baseUrl the base URL to search.
 * @param candidate the URL of the desired API to find.
 * @returns {boolean} true if the URL is found, otherwise false.
 */
function matchesBaseUrl(baseUrl, candidate) {
    const envs = ['dev', 'stg', 'prd'];
    return envs.every(e => {
        if (candidate[e]) {
            return baseUrl[e] === candidate[e];
        }
        return true;
    });
}
/**
 * Returns the base URL for the given environment and tenant, applying any overrides if available
 *
 * @param baseURL the default API base URL.
 * @param targetEnv the target environment
 * @param tenant the current tenant.
 * @returns {string} the base URL for the given environment and tenant, applying any overrides if available.
 */
export const getBaseUrlByTenantWithOverride = (baseUrl, targetEnv = 'prd', tenant = 'stackspot') => {
    let matchesApiName = null;
    for (const [apiName, config] of Object.entries(defaultApis)) {
        if (matchesBaseUrl(baseUrl, config.url)) {
            matchesApiName = apiName;
            break;
        }
    }
    if (!matchesApiName)
        return '';
    const overrideUrl = overrideApis[matchesApiName]?.url?.[targetEnv];
    if (overrideUrl) {
        return overrideUrl;
    }
    const apis = getApisByTenant(tenant);
    return apis[matchesApiName]?.url?.[targetEnv];
};
/**
 * Sets the APIs addresses for override.
 * @param customApis a JSON object containing APIs address to override.
 */
export const setApisOverride = (customApis) => {
    overrideApis = customApis;
};
/**
 * Sets the default APIs addresses for each environments.
 * @returns {Apis} an object containing APIs URLs grouped by api name and enviroments.
 */
export const getApiAddresses = () => {
    return defaultApis;
};
export const apiAddresses = () => {
    const env = NetworkClient.getEnv();
    return transformApisToApiAddress(defaultApis, env);
};
