// @stack-spot/portal-components/src/components/form/Form/Form.tsx
import { get } from 'lodash'
import { FormEvent, ReactNode } from 'react'
import { FieldValues, FormProvider, UseFormReturn, useForm, useFormContext } from 'react-hook-form'

export interface FormProps {
  id?: string,
  name: string,
  defaultValues?: FieldValues,
  autoComplete?: HTMLInputElement['autocomplete'],
  children?: ReactNode,
  onSubmit?: (params: FieldValues) => unknown,
  form?: UseFormReturn<any, any>,
  style?: React.CSSProperties,
  className?: string,
}


function handleGraphqlErrors(error: unknown, methods: UseFormReturn<FieldValues>) {
  const errors: any = get(error, 'cause.extensions.fields')

  if (errors) {
    Object.keys(errors).forEach((key) => {
      if (Object.keys(methods.control._fields).includes(key)) {
        methods.setError(key, { type: 'server', message: errors[`${key}`] })
      }
    })
  }
}

const handleFormSubmit = async (
  data: FormEvent<HTMLFormElement>,
  methods: UseFormReturn,
  onSubmit?: (params: Record<string, any>) => Promise<unknown> | unknown,
) => {

  if (!onSubmit) {
    return
  }

  const result = methods.handleSubmit(onSubmit)(data)
  const isPromise = result instanceof Promise

  if (!isPromise) {
    return result
  }

  try {
    return await result
  } catch (error: any) {
    const isGraphqlError = error.message === 'GraphqlError'
    if (!isGraphqlError) {
      throw error
    }
    handleGraphqlErrors(error, methods)
  }
}

const BasicForm = ({
  onSubmit,
  id,
  name,
  children,
  methods,
  autoComplete,
  ...props
}: FormProps & { methods: UseFormReturn<FieldValues> }) => (
  <form
    id={id}
    autoComplete={autoComplete}
    name={name}
    onSubmit={(data: FormEvent<HTMLFormElement>) => handleFormSubmit(data, methods, onSubmit)}
    noValidate
    style={props.style}
    className={props.className}
  >
    {children}
  </form>
)

const FormWithProvider = (props: FormProps) => {
  const methods = useForm({ defaultValues: props.defaultValues })
  const form = props.form || methods

  return (
    <FormProvider {...form}>
      <BasicForm {...props} methods={form} />
    </FormProvider>
  )
}

export const Form = (props: FormProps) => {
  const useFormMethods = useFormContext()
  const hasFormProvider = !!useFormMethods

  if (hasFormProvider) {
    return <BasicForm {...props} methods={useFormMethods} />
  }

  return <FormWithProvider {...props} />
}
