// @stack-spot/citric-react/src/components/MenuOverlay/Menu.tsx
import { listToClass } from '@stack-spot/portal-theme'
import { Dictionary, useTranslate } from '@stack-spot/portal-translate'
import { createElement, useEffect, useMemo } from 'react'
import { ValueController } from '../../utils/ValueController'
import { CitricComponent } from '../CitricComponent'
import { IconButton } from '../IconBox'
import { useOverlayController } from '../Overlay/context'
import { MenuProvider, useMenuController, useMenuState } from './context'
import { keyboardNavigation } from './keyboard'
import { MenuAction as MenuActionI, MenuCollapsible, MenuItem, MenuProps, MenuSection as MenuSectionI, MenuState } from './types'

// Arbitrary time (ms) to wait before running a function that needs the view to be updated with the next state value.
const RENDER_DELAY = 20

function Submenu({ children, label, className, icon, iconRight, style }: MenuCollapsible) {
  const controller = useMenuController()
  return (
    <button
      className={listToClass([className, 'submenu'])}
      style={style}
      onClick={(e) => {
        e.stopPropagation()
        const menu = e.target instanceof HTMLElement ? e.target.closest('[data-citric="menu"]') : undefined
        controller?.setValue({ items: children, label, parent: controller?.getValue() })
        const isOpenedWithMouse = e.detail > 0
        if (!isOpenedWithMouse) {
          setTimeout(() => {
            const firstFocusable = menu?.querySelector('a, button')
            if (firstFocusable instanceof HTMLElement) firstFocusable.focus()
          }, RENDER_DELAY) 
        }
      }}
    >
      {icon}
      <span>{label}</span>
      {iconRight}
    </button>
  )
}

function MenuSection({ children, label, className, style }: MenuSectionI) {
  return (
    <section className={className} style={style}>
      {label && <h6>{label}</h6>}
      <nav>
        <MenuItems items={children} />
      </nav>
      {!label && <hr />}
    </section>
  )
}

function MenuAction({ label, active, href, icon, iconRight, className, onClick, ...props }: MenuActionI) {
  const overlayController = useOverlayController()
  const children = <>
    {icon}
    {typeof label === 'string' ? <span>{label}</span> : label.element}
    {iconRight}
  </>
  return createElement(
    href ? 'a' : 'button',
    {
      href,
      'aria-label': typeof label === 'string' ? label : label.id,
      className: listToClass([className, active && 'active']),
      onClick: (e) => {
        overlayController?.close()
        onClick?.(e)
      },
      ...props,
    },
    children,
  )
}

function hasSections(items: MenuItem[]) {
  return items.some(i => 'children' in i && i.children.length && i.type === 'section')
}

function hasSubmenus(items: MenuItem[]) {
  return items.some(i => 'children' in i && i.children.length && i.type === 'collapsible')
}

function MenuItems({ items }: { items: MenuItem[] }) {
  return useMemo(() => items.map((item, index) => {
    if ('children' in item && item.type === 'section') return <MenuSection key={item.label || index} {...item} />
    if ('children' in item && item.type === 'collapsible') return <Submenu key={item.label || index} {...item} />
    return <MenuAction key={(typeof item.label === 'string' ? item.label : item.label?.id) || index} {...item} />
  }), [items])
}

/**
 * TODO: make the height changes animated.
 */
export function Menu(
  { items, appearance, bgLevel, header, roundedItems, showBorders, showShadows, spaced, className, onKeyDown, ...props }: MenuProps,
) {
  const controller = useMemo(() => new ValueController<MenuState>({ items }), [])
  const current = useMenuState(controller)
  const { sectioned, collapsible } = useMemo(
    () => ({ sectioned: hasSections(current.items), collapsible: hasSubmenus(current.items) }),
    [current.items],
  )
  const tag = header || sectioned || collapsible ? 'div' : 'nav'
  const t = useTranslate(dictionary)

  useEffect(() => {
    if (items !== controller.getValue().items) controller.setValue({ items })
  }, [items])
  
  return (
    <MenuProvider value={controller}>
      <CitricComponent
        tag={tag}
        component="menu"
        className={listToClass([
          className, appearance, roundedItems && 'rounded-items', showBorders && 'bordered', showShadows === false && 'no-shadow',
          spaced && 'spaced', bgLevel && `bg-${bgLevel}`,
        ])}
        onKeyDown={(e: any) => {
          keyboardNavigation(e)
          onKeyDown?.(e)
        }}
        {...props}
      >
        {header && <header>{header}</header>}
        {current.parent && (
          <div className="back-button">
            <IconButton
              icon="ArrowLeft"
              aria-label={t.goBack}
              onClick={(e) => {
                e.stopPropagation()
                const menu = e.target instanceof HTMLElement ? e.target.closest('[data-citric="menu"]') : undefined
                if (current.parent) controller.setValue(current.parent)
                setTimeout(() => {
                  const lastItem = Array.from(menu?.querySelectorAll('a, button') ?? []).find(el => el.textContent === current.label)
                  if (lastItem instanceof HTMLElement) lastItem.focus()
                }, RENDER_DELAY)
              }}
            />
            <span>{current.label || t.goBack}</span>
          </div>
        )}
        {sectioned ? <MenuItems items={current.items} /> : <nav><MenuItems items={current.items} /></nav>}
      </CitricComponent>
    </MenuProvider>
  )
}

const dictionary = {
  en: {
    goBack: 'Go back',
  },
  pt: {
    goBack: 'Voltar',
  },
} satisfies Dictionary
