// @stack-spot/citric-react/src/components/MenuOverlay/types.ts
import { WithStyle } from '@stack-spot/portal-theme'
import { BaseOverlayProps } from '../Overlay/types'

interface CustomLabel {
  /**
   * A unique identifier for this label. This is also used to read this label to screen readers.
   */
  id: string,
  /**
   * A custom label that can be made up of any React component.
   */
  element: React.ReactNode,
}

interface Action {
  /**
   * The label of the action.
   */
  label: string | CustomLabel,
  /**
   * Function to run on a click.
   */
  onClick?: (event?: React.MouseEvent) => void,
  /**
   * URL to open on a click.
   */
  href?: string,
  /**
   * Target of the URL to open.
   */
  target?: React.AnchorHTMLAttributes<HTMLAnchorElement>['target'],
  /**
   * Language of the linked document.
   */
  lang?: React.AnchorHTMLAttributes<HTMLAnchorElement>['lang'],
  /**
   * Language of the destination URL.
   */
  hrefLang?: React.AnchorHTMLAttributes<HTMLAnchorElement>['hrefLang'],
  /**
   * The title of the action. 
   */
  title?: string,
  /**
   * Whether or not this link/button is disabled.
   * 
   * @default false
   */
  disabled?: boolean,
}

interface ItemWithIcon {
  /**
   * An Icon to appear at the left of the item.
   */
  icon?: React.ReactElement,
  /**
   * An Icon to appear at the right of the item.
   */
  iconRight?: React.ReactElement,
}

export interface MenuAction extends ItemWithIcon, Action, WithStyle {
  /**
   * Whether or not this option is currently active.
   */
  active?: boolean,
}

interface MenuGroup extends WithStyle {
  /**
   * If this group is rendered as a section with its items right below it or a collapsible, which requires a click to open a submenu.
   */
  type?: 'section' | 'collapsible',
  /**
   * The items of this group.
   */
  children: MenuItem[],
}

export interface MenuSection extends MenuGroup {
  type: 'section',
  /**
   * The section's title.
   */
  label?: string,
}

export interface MenuCollapsible extends MenuGroup, ItemWithIcon {
  type?: 'collapsible',
  /**
   * The title of the collapsible menu.
   */
  label: string,
}

export type MenuItem = MenuSection | MenuCollapsible | MenuAction

export interface MenuState {
  items: MenuItem[],
  parent?: MenuState,
  label?: string,
}

export interface BaseMenuProps {
  /**
   * The options in the menu.
   */
  items: MenuItem[],
  /**
   * A header for the menu.
   */
  header?: React.ReactNode,
  /**
   * Whether or not to render spaces between the items and the sides of the menu.
   * 
   * @default false
   */
  spaced?: boolean,
  /**
   * Wheter or not to render borders.
   * 
   * @default false
   */
  showBorders?: boolean,
  /**
   * Whether or not to use rounded corners for each menu item.
   * 
   * @default false
   */
  roundedItems?: boolean,
  /**
   * Use 'square' for sharp corners and 'circle' for rounded corners.
   * 
   * @default 'circle'
   */
  appearance?: 'square' | 'circle',
  /**
   * The level of the background color. Use this to lighten or darken the background.
   * 
   * @default 300
   */
  bgLevel?: 300 | 400 | 500 | 600,
  /**
   * Whether or not to render shadows behind the menu.
   * 
   * @default true
   */
  showShadows?: boolean,
  /**
   * A class for the menu element.
   */
  menuClass?: string,
  /**
   * Styles for the menu element.
   */
  menuStyle?: React.CSSProperties,
}

export type MenuProps = React.JSX.IntrinsicElements['div'] & BaseMenuProps

export interface BaseMenuOverlayProps extends
  BaseMenuProps, Omit<BaseOverlayProps<'div'>, 'tag' | 'onRenderChild' | 'content' | 'triggerOn' | 'reference' | 'children'> {
  /**
   * A unique id for this menu. If not provided, a random number will be used.
   */
  id?: string,
  children?: React.ReactElement,
}

export type MenuOverlayProps = Omit<React.JSX.IntrinsicElements['div'], 'content' | 'children'> & BaseMenuOverlayProps
