// @stack-spot/citric-react/dist/utils/radio.js
import { useEffect, useMemo, useRef, useState } from 'react';
import { defaultRenderKey } from './options.js';
/**
 * Use this hook to easily implement filtering for a radio group.
 * @param params the parameters to create the controls.
 * @returns the checkbox controls.
 */
export function useRadioGroupControls(params) {
    const [value, setValue] = useState(params.initialValue);
    const [filter, setFilter] = useState();
    const previousValue = useRef(value);
    const renderKey = params.renderKey ?? defaultRenderKey;
    useEffect(() => {
        params.onChange?.(value, previousValue.current);
        previousValue.current = value;
    }, [value]);
    const { options, isUnfilteredButChecked } = useMemo(() => {
        if (!params.applyFilter || !filter)
            return { options: params.options, isUnfilteredButChecked: () => false };
        const valueKey = value ? renderKey(value) : undefined;
        const isValueFilteredOut = value && !params.applyFilter?.(filter, value);
        const filtered = params.options.filter(o => params.applyFilter?.(filter, o));
        return {
            options: isValueFilteredOut ? [value, ...filtered] : filtered,
            isUnfilteredButChecked: (o) => isValueFilteredOut && renderKey(o) === valueKey,
        };
    }, [params.options, filter]);
    return {
        /**
         * The current filter applied.
         */
        filter,
        /**
         * Apply a new filter.
         */
        setFilter: setFilter,
        /**
         * The options that should be passed to the radio group.
         */
        options,
        /**
         * The value that should be passed to the radio group.
         */
        value,
        /**
         * Changes the current value, should be passed to the property `onChange` of the radio group.
         */
        setValue,
        /**
         * A function to render a unique key for an option. Should be passed to the property `renderKey` of the radio group.
         */
        renderKey,
        /**
         * A function that returns true if the option is filtered out, but checked.
         * @param option the option to check.
         * @returns true if the option was filtered out, but is checked; false otherwise.
         */
        isUnfilteredButChecked,
    };
}
