// @stack-spot/portal-network/src/error/dictionary/cloud-platform.ts
import { Dictionary } from '@stack-spot/portal-translate'

export const scfDictionary = {
  en: {
    SCF_CERTIFICATE_ERROR: 'Certificate operation failed.',
    SCF_CERTIFICATE_NOT_FOUND_ERROR: 'Certificate with ID {1} not found.',
    SCF_CIDR_ERROR: 'CIDR operation failed.',
    SCF_CIDR_RANGE_CONFLICT: 'CIDR range {1} conflicts with an existing range.',
    SCF_CIDR_FOUNDATION_RESERVED: 'CIDR range {1} conflicts with the Foundation\'s reserved range.',
    SCF_CIDR_NOT_FOUND_ERROR: 'CIDR with ID {1} not found.',
    SCF_CONTROL_TOWER_BUSY_ERROR: 'Foundation\'s Control Tower is currently busy. Please try again later.',
    SCF_DNS_RECORD_ERROR: 'DNS Record operation failed.',
    SCF_DNS_RECORD_NOT_FOUND_ERROR: 'DNS Record with ID {1} not found.',
    SCF_DNS_ZONE_ERROR: 'DNS Zone operation failed.',
    SCF_DNS_ZONE_NOT_FOUND_ERROR: 'DNS Zone with ID {1} not found.',
    SCF_DNS_ZONE_NOT_READY_ERROR: 'DNS Zone with ID {1} not ready.',
    SCF_FOLDER_ALREADY_EXISTS_ERROR: 'A folder named {1} already exists.',
    SCF_FOLDER_ERROR: 'Folder operation failed.',
    SCF_FOLDER_NOT_FOUND_ERROR: 'Folder with ID {1} not found.',
    SCF_FOLDER_NOT_READY_ERROR: 'Folder with ID {1} not ready.',
    SCF_FOUNDATION_ERROR: 'Foundation operation failed.',
    SCF_FOUNDATION_NOT_FOUND_ERROR: 'Foundation with ID {1} not found.',
    SCF_FOUNDATION_NOT_READY_ERROR: 'Foundation with ID {1} not ready.',
    SCF_INBOUND_ERROR: 'Inbound operation failed.',
    SCF_INVALID_REGEX_ERROR: 'The value {1} is invalid.',
    SCF_MASTER_ACCOUNT_ERROR: 'Master Account operation failed.',
    SCF_MASTER_ACCOUNT_NOT_FOUND_ERROR: 'No Master Account available for Foundation creation.',
    SCF_NAMESPACE_CREATION_ERROR: 'An error occurred while creating the Kubernetes namespace.',
    SCF_NETWORK_CONNECTION_ERROR: 'Network Connection operation failed.',
    SCF_NETWORK_ERROR: 'Network operation failed.',
    SCF_NETWORK_NOT_FOUND_ERROR: 'Network with ID {1} not found.',
    SCF_NETWORK_NOT_READY_ERROR: 'Network with ID {1} not ready.',
    SCF_PARENT_FOLDER_NOT_READY_ERROR: 'Parent Folder with ID {1} not ready.',
    SCF_PROJECT_ERROR: 'Project operation failed.',
    SCF_PROJECT_NOT_FOUND_ERROR: 'Project with ID {1} not found.',
    SCF_PROJECT_NOT_READY_ERROR: 'Project with ID {1} not ready.',
    SCF_UNKNOWN_ERROR: 'An unexpected error has occurred.',
    SCF_VPN_CONFIGURATION_NOT_FOUND_ERROR: 'Configuration for VPN with ID {1} not found.',
    SCF_VPN_ERROR: 'VPN operation failed.',
    SCF_VPN_NOT_FOUND_ERROR: 'VPN with ID {1} not found.',
    SCF_VPN_NOT_READY_ERROR: 'VPN with ID {1} not ready.',
    SCF_DNS_ZONE_NOT_EMPTY_ERROR: 'DNS Zone with ID {1} is not empty.',
    SCF_FOLDER_NOT_EMPTY_ERROR: 'Folder with ID {1} is not empty.',
    SCF_PROJECT_NOT_EMPTY_ERROR: 'Project with ID {1} is not empty.',
  },
  pt: {
    SCF_CERTIFICATE_ERROR: 'Falha na operação de Certificado.',
    SCF_CERTIFICATE_NOT_FOUND_ERROR: 'Certificado com ID {1} não encontrado.',
    SCF_CIDR_ERROR: 'Falha na operação de CIDR.',
    SCF_CIDR_RANGE_CONFLICT: 'O intervalo CIDR {1} conflita com um intervalo existente.',
    SCF_CIDR_FOUNDATION_RESERVED: 'O intervalo CIDR {1} conflita com o intervalo reservado da Fundação.',
    SCF_CIDR_NOT_FOUND_ERROR: 'CIDR com ID {1} não encontrado.',
    SCF_CONTROL_TOWER_BUSY_ERROR: 'Control Tower da Fundação está ocupada no momento. Por favor, tente novamente mais tarde.',
    SCF_DNS_RECORD_ERROR: 'Falha na operação de DNS Record.',
    SCF_DNS_RECORD_NOT_FOUND_ERROR: 'DNS Record com ID {1} não encontrado.',
    SCF_DNS_ZONE_ERROR: 'Falha na operação de DNS Zone.',
    SCF_DNS_ZONE_NOT_FOUND_ERROR: 'DNS Zone com ID {1} não encontrada.',
    SCF_DNS_ZONE_NOT_READY_ERROR: 'DNS Zone com ID {1} não está pronta.',
    SCF_FOLDER_ALREADY_EXISTS_ERROR: 'Uma Pasta com o nome {1} já existe.',
    SCF_FOLDER_ERROR: 'Falha na operação de Pasta.',
    SCF_FOLDER_NOT_FOUND_ERROR: 'Pasta com ID {1} não encontrada.',
    SCF_FOLDER_NOT_READY_ERROR: 'Pasta com ID {1} não está pronta.',
    SCF_FOUNDATION_ERROR: 'Falha na operação da Fundação.',
    SCF_FOUNDATION_NOT_FOUND_ERROR: 'Fundação com ID {1} não encontrada.',
    SCF_FOUNDATION_NOT_READY_ERROR: 'Fundação com ID {1} não está pronta.',
    SCF_INBOUND_ERROR: 'Falha na operação de Inbound.',
    SCF_INVALID_REGEX_ERROR: 'O valor {1} é inválido.',
    SCF_MASTER_ACCOUNT_ERROR: 'Falha na operação da Master Account.',
    SCF_MASTER_ACCOUNT_NOT_FOUND_ERROR: 'Nenhuma Master Account disponível para criação da Fundação.',
    SCF_NAMESPACE_CREATION_ERROR: 'Ocorreu um erro ao criar o namespace no Kubernetes.',
    SCF_NETWORK_CONNECTION_ERROR: 'Falha na operação de Network Connection.',
    SCF_NETWORK_ERROR: 'Falha na operação de Network.',
    SCF_NETWORK_NOT_FOUND_ERROR: 'Network com ID {1} não encontrada.',
    SCF_NETWORK_NOT_READY_ERROR: 'Network com ID {1} não está pronta.',
    SCF_PARENT_FOLDER_NOT_READY_ERROR: 'Pasta pai com ID {1} não está pronta.',
    SCF_PROJECT_ERROR: 'Falha na operação do Projeto.',
    SCF_PROJECT_NOT_FOUND_ERROR: 'Projeto com ID {1} não encontrado.',
    SCF_PROJECT_NOT_READY_ERROR: 'Projeto com ID {1} não está pronto.',
    SCF_UNKNOWN_ERROR: 'Ocorreu um erro inesperado.',
    SCF_VPN_CONFIGURATION_NOT_FOUND_ERROR: 'Configuração para VPN com ID {1} não encontrada.',
    SCF_VPN_ERROR: 'Falha na operação de VPN.',
    SCF_VPN_NOT_FOUND_ERROR: 'VPN com ID {1} não encontrada.',
    SCF_VPN_NOT_READY_ERROR: 'VPN com ID {1} não está pronta.',
    SCF_DNS_ZONE_NOT_EMPTY_ERROR: 'DNS Zone com ID {1} não está vazia.',
    SCF_FOLDER_NOT_EMPTY_ERROR: 'Folder com ID {1} não está vazia.',
    SCF_PROJECT_NOT_EMPTY_ERROR: 'Project com ID {1} não está vazia.',
  },
} satisfies Dictionary
