// @stack-spot/portal-components/dist/components/BannerWarning.js
import { jsx as _jsx, jsxs as _jsxs } from "react/jsx-runtime";
import { Button, Flex, IconBox, Text } from '@citric/core';
import { ExclamationTriangle, TimesCircleFill } from '@citric/icons';
import { Alert } from '@citric/ui';
import { useState } from 'react';
import { getCookie, setCookie } from '../utils/cookie.js';
/**
 * A styled and dismissible alert box for showing important warnings for the user.
 *
 * This renders a yellow box with an exclamation icon. The content passed as parameter is placed on the right of the icon.
 * @param props an object containing the content (children) to show inside the alert box.
 */
export const BannerWarning = ({ children, showCloseButton = true }) => {
    const bannerKey = 'bannerClosed';
    const [isVisible, setIsVisible] = useState(() => {
        if (typeof window !== 'undefined') {
            const isBannerClosed = getCookie(bannerKey) === 'true';
            return !isBannerClosed;
        }
        return false;
    });
    const handleClose = () => {
        setCookie(bannerKey, 'true');
        setIsVisible(false);
    };
    if (!isVisible) {
        return null;
    }
    return (_jsx(Flex, { sx: { zIndex: '1', w: '12', mb: 7 }, children: _jsxs(Alert, { colorScheme: "warning", sx: {
                position: 'relative',
                px: 3,
                py: 1,
                alignItems: 'center',
                display: 'flex',
                justifyContent: 'space-between',
                height: 'auto',
                padding: `${showCloseButton ? '0.5rem' : '1rem'}`,
            }, children: [_jsxs(Flex, { sx: {
                        alignItems: 'center',
                        flexWrap: 'nowrap',
                    }, children: [_jsx(IconBox, { size: "sm", colorIcon: "light.500", children: _jsx(ExclamationTriangle, {}) }), _jsx(Text, { sx: {
                                ml: 5,
                            }, appearance: "body2", colorScheme: "warning.contrastText", children: children })] }), showCloseButton && _jsx(Button, { appearance: 'text', onClick: handleClose, sx: {
                        p: 0,
                        ml: 2,
                    }, children: _jsx(IconBox, { size: "lg", colorIcon: "light.500", children: _jsx(TimesCircleFill, {}) }) })] }) }));
};
