// @stack-spot/portal-components/dist/components/Placeholder.js
import { jsx as _jsx, jsxs as _jsxs } from "react/jsx-runtime";
import { Box, Flex, Text } from '@citric/core';
import { Card, CardContent } from '@citric/ui';
import { useTranslate } from '@stack-spot/portal-translate';
import { GenericPlaceholder } from '../svg/index.js';
export const Placeholder = ({ title, description, button, img, fullWidth, show = true, sx, style, className }) => {
    if (!show) {
        return null;
    }
    return (_jsx(Card, { sx: { ...(fullWidth ? { w: '100%' } : {}) }, style: style, className: className, children: _jsxs(Flex, { alignItems: "center", justifyContent: "center", py: 15, sx: { gap: '20px', ...sx }, children: [_jsx(Box, { sx: { justifySelf: 'end' }, children: img ?? _jsx(GenericPlaceholder, {}) }), _jsxs(Box, { children: [title && _jsx(Text, { as: "label", appearance: "h4", mb: "3", pt: 5, children: title }), _jsx(Text, { mb: "3", appearance: "body2", colorScheme: "light.700", children: description }), button] })] }) }));
};
export const PlaceholderCallToAction = ({ title, description, subDescription, img, button, fullWidth, show = true, sx, sxCard, style, className, titleAs, placeholderImageConfig, }) => {
    if (!show) {
        return null;
    }
    return (_jsx(Card, { sx: { w: fullWidth ? '100%' : 'auto', bg: 'transparent', ...sxCard }, style: style, className: className, children: _jsxs(CardContent, { as: Flex, py: 15, sx: { flexDirection: 'column', justifyContent: 'center', alignItems: 'center', gap: '92px', ...sx }, children: [_jsx(Box, { sx: { justifySelf: 'end' }, children: img ?? _jsx(GenericPlaceholder, { ...placeholderImageConfig }) }), _jsxs(Box, { children: [title && _jsx(Text, { as: titleAs, appearance: "h4", mb: "3", "aria-live": "assertive", children: title }), _jsx(Text, { mb: "3", appearance: "body2", colorScheme: "light.700", children: description }), _jsx(Text, { mb: "3", appearance: "body2", colorScheme: "light.700", children: subDescription }), button] })] }) }));
};
export const PlaceholderResourceEmptySearch = ({ title, description }) => {
    const t = useTranslate(dictionary);
    return (_jsx(PlaceholderCallToAction, { fullWidth: true, title: title || t.title, description: description || t.description }));
};
const dictionary = {
    en: {
        title: 'Sorry, your search did not match any results.',
        description: 'Please, refine your search and try again.',
    },
    pt: {
        title: 'Desculpe, sua busca não encontrou nenhum resultado.',
        description: 'Por favor, refine sua busca e tente novamente.',
    },
};
