// @stack-spot/citric-react/dist/components/MenuOverlay/keyboard.js
// TODO: scroll the element into view when it's focused. Very important for accessibility in scrollable menus.
function getCurrent(element) {
    return element instanceof HTMLElement
        ? element.querySelector('button:focus, a:focus')
        : null;
}
function getMenuItems(element) {
    return element instanceof HTMLElement ? element.querySelectorAll('button, a') : undefined;
}
function focusFirst(element) {
    const first = getMenuItems(element)?.[0];
    if (first instanceof HTMLElement)
        first.focus();
}
function focusLast(element) {
    const all = getMenuItems(element);
    const last = all?.[all.length];
    if (last instanceof HTMLElement)
        last.focus();
}
function indexOf(elements, element) {
    for (let i = 0; i < elements.length; i++) {
        if (elements[i] === element)
            return i;
    }
    return -1;
}
function focusPrevious(element, current) {
    const all = getMenuItems(element);
    const index = all ? indexOf(all, current) : -1;
    if (!all || index === -1)
        return;
    const prev = index === 0 ? all[all.length - 1] : all[index - 1];
    if (prev instanceof HTMLElement)
        prev.focus();
}
function focusNext(element, current) {
    const all = getMenuItems(element);
    const index = all ? indexOf(all, current) : -1;
    if (!all || index === -1)
        return;
    const next = index >= all.length - 1 ? all[0] : all[index + 1];
    if (next instanceof HTMLElement)
        next.focus();
}
export function keyboardNavigation(event) {
    if (!['ArrowUp', 'ArrowDown'].includes(event.key))
        return;
    event.stopPropagation();
    event.preventDefault();
    const menu = event.target instanceof HTMLElement ? event.target.closest('[data-citric="menu"]') : undefined;
    const current = getCurrent(menu);
    if (event.key === 'ArrowUp') {
        if (current)
            focusPrevious(menu, current);
        else
            focusLast(menu);
    }
    else {
        if (current)
            focusNext(menu, current);
        else
            focusFirst(menu);
    }
}
