// @stack-spot/portal-network/src/api/agent.ts
/**
 * ai-agent-mgmt-api API
 * 0.0.1
 * DO NOT MODIFY - This file has been generated using oazapfts.
 * See https://www.npmjs.com/package/oazapfts
 */
import * as Oazapfts from "@oazapfts/runtime";
import * as QS from "@oazapfts/runtime/query";
export const defaults: Oazapfts.Defaults<Oazapfts.CustomHeaders> = {
    headers: {},
    baseUrl: "/",
};
const oazapfts = Oazapfts.runtime(defaults);
export const servers = {};
export type AgentType = "CONVERSATIONAL" | "SINGLE_ANSWER";
export type LlmConfigRequestPart = {
    model_slug: string;
    temperature?: number | null;
};
export type SimilarityFunction = "COSINE" | "EUCLIDEAN" | "DOT_PRODUCT";
export type KnowledgeSourcesConfigRequestPart = {
    similarity_function?: SimilarityFunction;
    max_number_of_kos?: number | null;
    relevancy_threshold?: number | null;
    post_processing?: boolean | null;
    sealed?: boolean | null;
    /** List of knowledge source ids (ULID). */
    knowledge_sources: string[];
};
export type CustomToolRequest = {
    toolkit_id?: string;
    tools_ids?: string[];
};
export type NewAgentRequest = {
    name: string;
    slug: string;
    description?: string;
    avatar?: string;
    suggested_prompts?: string[];
    system_prompt: string;
    "type": AgentType;
    llm_config?: (LlmConfigRequestPart) | null;
    knowledge_sources_config?: (KnowledgeSourcesConfigRequestPart) | null;
    builtin_tools_ids?: string[];
    custom_tools?: CustomToolRequest[];
    detail_mode?: boolean;
    model_name?: string | null;
    model_id?: string | null;
    structured_output?: {
        [key: string]: any;
    } | null;
    llm_settings?: {
        [key: string]: any;
    } | null;
};
export type ViolationResponse = {
    field?: string;
    message?: string;
};
export type WorkspaceListRequest = {
    agentIds?: string[];
    name?: string;
};
export type Uuid = string;
export type WorkspaceForkRequest = {
    agentIds?: string[];
    memberId?: Uuid;
};
export type AgentItemFork = {
    agentId?: string;
    isBuiltIn?: boolean;
};
export type WorkspaceForkResponse = {
    agents?: AgentItemFork[];
    agentIds?: string[];
    agentIdsAccount?: string[];
    ksIds?: string[];
    ksIdsAccount?: string[];
    errorIds?: string[];
};
export type WorkspaceDeleteRequest = {
    agentIds?: string[];
    isRollback?: boolean;
};
export type LlmConfigResponse = {
    model_slug: string;
    temperature: number;
};
export type KnowledgeSourceDetailResponse = {
    id: string;
    slug: string;
    name: string;
    description: string;
    "type": string;
    creator: string;
    "default": boolean;
    visibility_level: string;
    model_name: string;
    username: string;
};
export type KnowledgeSourceResponse = {
    similarity_function: SimilarityFunction;
    max_number_of_kos: number;
    relevancy_threshold: number;
    post_processing: boolean;
    sealed: boolean;
    knowledge_sources: string[];
    knowledge_sources_details?: KnowledgeSourceDetailResponse[];
};
export type VisibilityLevel = "PERSONAL" | "SHARED" | "WORKSPACE" | "ACCOUNT" | "FAVORITE";
export type ZonedDateTime = string;
export type BuiltinToolResponse = {
    id?: string;
    name?: string;
    description?: string;
};
export type BuiltinToolkitResponse = {
    id?: string;
    name?: string;
    description?: string;
    image_url?: string;
    tools?: BuiltinToolResponse[];
};
export type CustomToolResponse = {
    id?: string;
    name?: string;
    description?: string;
    method?: string;
    url?: string;
};
export type CustomToolkitResponse = {
    id?: string;
    name?: string;
    description?: string;
    avatar?: string;
    secret_id?: string;
    tools?: CustomToolResponse[];
};
export type ToolkitsInAgentResponse = {
    builtins?: BuiltinToolkitResponse[];
    customs?: CustomToolkitResponse[];
};
export type ModifiedAgentRequest = {
    name: string;
    slug: string;
    description?: string | null;
    avatar?: string | null;
    suggested_prompts?: string[] | null;
    system_prompt?: string | null;
    "type"?: AgentType;
    llm_config?: (LlmConfigRequestPart) | null;
    knowledge_sources_config?: (KnowledgeSourcesConfigRequestPart) | null;
    use_only?: boolean | null;
    builtin_tools_ids?: string[];
    custom_tools?: CustomToolRequest[];
    detail_mode?: boolean | null;
    model_name?: string | null;
    model_id?: string | null;
    llm_settings?: {
        [key: string]: any;
    } | null;
    structured_output?: {
        [key: string]: any;
    } | null;
};
export type ForkAgentRequest = {
    slug?: string;
    name?: string | null;
};
export type PublishRequest = {
    level: VisibilityLevel;
    use_only?: boolean;
};
export type AgentResponse = {
    id: string;
    name: string;
    slug: string;
    description?: string;
    avatar?: string;
    system_prompt: string;
    "type": string;
    llm_config?: LlmConfigResponse;
    suggested_prompts?: string[];
    knowledge_sources_config?: KnowledgeSourceResponse;
    visibility_level: VisibilityLevel;
    use_only: boolean;
    creator_name: string;
    created_by: string;
    created_at: ZonedDateTime;
    updated_by?: string;
    updated_at?: ZonedDateTime;
    toolkits?: ToolkitsInAgentResponse;
    detail_mode?: boolean;
    model_id?: string;
    model_name?: string;
    structured_output?: {
        [key: string]: any;
    } | null;
    llm_settings?: {
        [key: string]: any;
    } | null;
};
/**
 * Creates a new AI Agent.
 */
export function postV1Agent({ newAgentRequest }: {
    newAgentRequest?: NewAgentRequest;
}, opts?: Oazapfts.RequestOpts) {
    return oazapfts.ok(oazapfts.fetchJson<{
        status: 201;
        data: {
            id?: string;
        };
    } | {
        status: 400;
        data: {
            code?: string;
            details?: string;
            violations?: ViolationResponse[];
        };
    } | {
        status: 422;
        data: {
            code?: string;
            details?: string;
            additionalInformation?: {
                [key: string]: any;
            } | null;
        };
    } | {
        status: 500;
        data: {
            code?: string;
            details?: string;
            additionalInformation?: {
                [key: string]: any;
            } | null;
        };
    }>("/v1/agent", oazapfts.json({
        ...opts,
        method: "POST",
        body: newAgentRequest
    })));
}
/**
 * List Workspace Fork AI agents.
 */
export function postV1AgentDependencies({ workspaceListRequest }: {
    workspaceListRequest?: WorkspaceListRequest;
}, opts?: Oazapfts.RequestOpts) {
    return oazapfts.ok(oazapfts.fetchJson<{
        status: 201;
        data: Blob;
    } | {
        status: 500;
        data: {
            code?: string;
            details?: string;
            additionalInformation?: {
                [key: string]: any;
            } | null;
        };
    }>("/v1/agent/dependencies", oazapfts.json({
        ...opts,
        method: "POST",
        body: workspaceListRequest
    })));
}
/**
 * List agents with specific KS
 */
export function getV1AgentKnowledgeSourceById({ id }: {
    id: string;
}, opts?: Oazapfts.RequestOpts) {
    return oazapfts.ok(oazapfts.fetchJson<{
        status: 200;
        data: Blob;
    } | {
        status: 500;
        data: {
            code?: string;
            details?: string;
            additionalInformation?: {
                [key: string]: any;
            } | null;
        };
    }>(`/v1/agent/knowledge-source/${encodeURIComponent(id)}`, {
        ...opts
    }));
}
/**
 * Disassociate KS from Agent
 */
export function deleteV1AgentKnowledgeSourceById({ id }: {
    id: string;
}, opts?: Oazapfts.RequestOpts) {
    return oazapfts.ok(oazapfts.fetchJson<{
        status: 204;
        data: Blob;
    } | {
        status: 500;
        data: {
            code?: string;
            details?: string;
            additionalInformation?: {
                [key: string]: any;
            } | null;
        };
    }>(`/v1/agent/knowledge-source/${encodeURIComponent(id)}`, {
        ...opts,
        method: "DELETE"
    }));
}
/**
 * Workspace Fork AI agents.
 */
export function postV1AgentWorkspaceFork({ workspaceForkRequest }: {
    workspaceForkRequest?: WorkspaceForkRequest;
}, opts?: Oazapfts.RequestOpts) {
    return oazapfts.ok(oazapfts.fetchJson<{
        status: 201;
        data: WorkspaceForkResponse[];
    } | {
        status: 500;
        data: {
            code?: string;
            details?: string;
            additionalInformation?: {
                [key: string]: any;
            } | null;
        };
    }>("/v1/agent/workspace/fork", oazapfts.json({
        ...opts,
        method: "POST",
        body: workspaceForkRequest
    })));
}
/**
 * Delete Workspace Fork AI agents.
 */
export function deleteV1AgentWorkspaceFork({ workspaceDeleteRequest }: {
    workspaceDeleteRequest?: WorkspaceDeleteRequest;
}, opts?: Oazapfts.RequestOpts) {
    return oazapfts.ok(oazapfts.fetchJson<{
        status: 201;
        data: Blob;
    } | {
        status: 500;
        data: {
            code?: string;
            details?: string;
            additionalInformation?: {
                [key: string]: any;
            } | null;
        };
    }>("/v1/agent/workspace/fork", oazapfts.json({
        ...opts,
        method: "DELETE",
        body: workspaceDeleteRequest
    })));
}
/**
 * List Workspace Fork AI agents.
 */
export function postV1AgentWorkspaceList({ workspaceListRequest }: {
    workspaceListRequest?: WorkspaceListRequest;
}, opts?: Oazapfts.RequestOpts) {
    return oazapfts.ok(oazapfts.fetchJson<{
        status: 201;
        data: Blob;
    } | {
        status: 500;
        data: {
            code?: string;
            details?: string;
            additionalInformation?: {
                [key: string]: any;
            } | null;
        };
    }>("/v1/agent/workspace/list", oazapfts.json({
        ...opts,
        method: "POST",
        body: workspaceListRequest
    })));
}
/**
 * Get agent by ID
 */
export function getV1AgentByAgentId({ agentId }: {
    agentId: string;
}, opts?: Oazapfts.RequestOpts) {
    return oazapfts.ok(oazapfts.fetchJson<{
        status: 200;
        data: {
            id: string;
            name: string;
            slug: string;
            description?: string;
            avatar?: string;
            system_prompt: string;
            "type": string;
            llm_config?: LlmConfigResponse;
            suggested_prompts?: string[];
            knowledge_sources_config?: KnowledgeSourceResponse;
            visibility_level: VisibilityLevel;
            use_only: boolean;
            creator_name: string;
            created_by: string;
            created_at: ZonedDateTime;
            updated_by?: string;
            updated_at?: ZonedDateTime;
            toolkits?: ToolkitsInAgentResponse;
            detail_mode?: boolean;
            model_id?: string;
            model_name?: string;
            structured_output?: {
                [key: string]: any;
            } | null;
            llm_settings?: {
                [key: string]: any;
            } | null;
        };
    } | {
        status: 404;
        data: {
            code?: string;
            details?: string;
            additionalInformation?: {
                [key: string]: any;
            } | null;
        };
    } | {
        status: 500;
        data: {
            code?: string;
            details?: string;
            additionalInformation?: {
                [key: string]: any;
            } | null;
        };
    }>(`/v1/agent/${encodeURIComponent(agentId)}`, {
        ...opts
    }));
}
/**
 * Update agent by ID
 */
export function putV1AgentByAgentId({ agentId, modifiedAgentRequest }: {
    agentId: string;
    modifiedAgentRequest?: ModifiedAgentRequest;
}, opts?: Oazapfts.RequestOpts) {
    return oazapfts.ok(oazapfts.fetchJson<{
        status: 204;
    } | {
        status: 400;
        data: {
            code?: string;
            details?: string;
            violations?: ViolationResponse[];
        };
    } | {
        status: 404;
        data: {
            code?: string;
            details?: string;
            additionalInformation?: {
                [key: string]: any;
            } | null;
        };
    } | {
        status: 422;
        data: {
            code?: string;
            details?: string;
            additionalInformation?: {
                [key: string]: any;
            } | null;
        };
    } | {
        status: 500;
        data: {
            code?: string;
            details?: string;
            additionalInformation?: {
                [key: string]: any;
            } | null;
        };
    }>(`/v1/agent/${encodeURIComponent(agentId)}`, oazapfts.json({
        ...opts,
        method: "PUT",
        body: modifiedAgentRequest
    })));
}
/**
 * Delete agent by ID
 */
export function deleteV1AgentByAgentId({ agentId }: {
    agentId: string;
}, opts?: Oazapfts.RequestOpts) {
    return oazapfts.ok(oazapfts.fetchJson<{
        status: 204;
    } | {
        status: 404;
        data: {
            code?: string;
            details?: string;
            additionalInformation?: {
                [key: string]: any;
            } | null;
        };
    } | {
        status: 422;
        data: {
            code?: string;
            details?: string;
            additionalInformation?: {
                [key: string]: any;
            } | null;
        };
    } | {
        status: 500;
        data: {
            code?: string;
            details?: string;
            additionalInformation?: {
                [key: string]: any;
            } | null;
        };
    }>(`/v1/agent/${encodeURIComponent(agentId)}`, {
        ...opts,
        method: "DELETE"
    }));
}
/**
 * Favorite an AI agent.
 */
export function postV1AgentByAgentIdFavorite({ agentId }: {
    agentId: string;
}, opts?: Oazapfts.RequestOpts) {
    return oazapfts.ok(oazapfts.fetchJson<{
        status: 204;
        data: Blob;
    } | {
        status: 500;
        data: {
            code?: string;
            details?: string;
            additionalInformation?: {
                [key: string]: any;
            } | null;
        };
    }>(`/v1/agent/${encodeURIComponent(agentId)}/favorite`, {
        ...opts,
        method: "POST"
    }));
}
/**
 * Delete favorite of an AI agent.
 */
export function deleteV1AgentByAgentIdFavorite({ agentId }: {
    agentId: string;
}, opts?: Oazapfts.RequestOpts) {
    return oazapfts.ok(oazapfts.fetchJson<{
        status: 204;
        data: Blob;
    } | {
        status: 500;
        data: {
            code?: string;
            details?: string;
            additionalInformation?: {
                [key: string]: any;
            } | null;
        };
    }>(`/v1/agent/${encodeURIComponent(agentId)}/favorite`, {
        ...opts,
        method: "DELETE"
    }));
}
/**
 * Fork an AI agent.
 */
export function postV1AgentByAgentIdFork({ agentId, forkAgentRequest }: {
    agentId: string;
    forkAgentRequest?: ForkAgentRequest;
}, opts?: Oazapfts.RequestOpts) {
    return oazapfts.ok(oazapfts.fetchJson<{
        status: 201;
        data: {
            id?: string;
        };
    } | {
        status: 500;
        data: {
            code?: string;
            details?: string;
            additionalInformation?: {
                [key: string]: any;
            } | null;
        };
    }>(`/v1/agent/${encodeURIComponent(agentId)}/fork`, oazapfts.json({
        ...opts,
        method: "POST",
        body: forkAgentRequest
    })));
}
/**
 * AI Agent update visibility.
 */
export function postV1AgentByAgentIdPublish({ agentId, publishRequest }: {
    agentId: string;
    publishRequest?: PublishRequest;
}, opts?: Oazapfts.RequestOpts) {
    return oazapfts.ok(oazapfts.fetchJson<{
        status: 204;
    } | {
        status: 400;
        data: {
            code?: string;
            details?: string;
            violations?: ViolationResponse[];
        };
    } | {
        status: 404;
        data: {
            code?: string;
            details?: string;
            additionalInformation?: {
                [key: string]: any;
            } | null;
        };
    } | {
        status: 422;
        data: {
            code?: string;
            details?: string;
            additionalInformation?: {
                [key: string]: any;
            } | null;
        };
    } | {
        status: 500;
        data: {
            code?: string;
            details?: string;
            additionalInformation?: {
                [key: string]: any;
            } | null;
        };
    }>(`/v1/agent/${encodeURIComponent(agentId)}/publish`, oazapfts.json({
        ...opts,
        method: "POST",
        body: publishRequest
    })));
}
/**
 * Share an AI agent.
 */
export function postV1AgentByAgentIdShare({ agentId }: {
    agentId: string;
}, opts?: Oazapfts.RequestOpts) {
    return oazapfts.ok(oazapfts.fetchJson<{
        status: 204;
        data: object;
    } | {
        status: 500;
        data: {
            code?: string;
            details?: string;
            additionalInformation?: {
                [key: string]: any;
            } | null;
        };
    }>(`/v1/agent/${encodeURIComponent(agentId)}/share`, {
        ...opts,
        method: "POST"
    }));
}
/**
 * Get agent by Slug
 */
export function getV1AgentByAgentSlugExists({ agentSlug }: {
    agentSlug: string;
}, opts?: Oazapfts.RequestOpts) {
    return oazapfts.ok(oazapfts.fetchJson<{
        status: 200;
        data: {
            id: string;
            name: string;
            slug: string;
            description?: string;
            avatar?: string;
            system_prompt: string;
            "type": string;
            llm_config?: LlmConfigResponse;
            suggested_prompts?: string[];
            knowledge_sources_config?: KnowledgeSourceResponse;
            visibility_level: VisibilityLevel;
            use_only: boolean;
            creator_name: string;
            created_by: string;
            created_at: ZonedDateTime;
            updated_by?: string;
            updated_at?: ZonedDateTime;
            toolkits?: ToolkitsInAgentResponse;
            detail_mode?: boolean;
            model_id?: string;
            model_name?: string;
            structured_output?: {
                [key: string]: any;
            } | null;
            llm_settings?: {
                [key: string]: any;
            } | null;
        };
    } | {
        status: 404;
        data: {
            code?: string;
            details?: string;
            additionalInformation?: {
                [key: string]: any;
            } | null;
        };
    } | {
        status: 500;
        data: {
            code?: string;
            details?: string;
            additionalInformation?: {
                [key: string]: any;
            } | null;
        };
    }>(`/v1/agent/${encodeURIComponent(agentSlug)}/exists`, {
        ...opts
    }));
}
/**
 * List built-in AI Agents filtered by visibility.
 */
export function getV1Agents({ visibility }: {
    visibility?: VisibilityLevel;
}, opts?: Oazapfts.RequestOpts) {
    return oazapfts.ok(oazapfts.fetchJson<{
        status: 200;
        data: AgentResponse[];
    } | {
        status: 500;
        data: {
            code?: string;
            details?: string;
            additionalInformation?: {
                [key: string]: any;
            } | null;
        };
    }>(`/v1/agents${QS.query(QS.explode({
        visibility
    }))}`, {
        ...opts
    }));
}
/**
 * Creates a new built-in AI Agent.
 */
export function postV1PublicAgent({ newAgentRequest }: {
    newAgentRequest?: NewAgentRequest;
}, opts?: Oazapfts.RequestOpts) {
    return oazapfts.ok(oazapfts.fetchJson<{
        status: 201;
        data: {
            id?: string;
        };
    } | {
        status: 400;
        data: {
            code?: string;
            details?: string;
            violations?: ViolationResponse[];
        };
    } | {
        status: 422;
        data: {
            code?: string;
            details?: string;
            additionalInformation?: {
                [key: string]: any;
            } | null;
        };
    } | {
        status: 500;
        data: {
            code?: string;
            details?: string;
            additionalInformation?: {
                [key: string]: any;
            } | null;
        };
    }>("/v1/public/agent", oazapfts.json({
        ...opts,
        method: "POST",
        body: newAgentRequest
    })));
}
/**
 * Get agent by ID
 */
export function getV1PublicAgentByAgentId({ agentId }: {
    agentId: string;
}, opts?: Oazapfts.RequestOpts) {
    return oazapfts.ok(oazapfts.fetchJson<{
        status: 200;
        data: {
            id: string;
            name: string;
            slug: string;
            description?: string;
            avatar?: string;
            system_prompt: string;
            "type": string;
            llm_config?: LlmConfigResponse;
            suggested_prompts?: string[];
            knowledge_sources_config?: KnowledgeSourceResponse;
            visibility_level: VisibilityLevel;
            use_only: boolean;
            creator_name: string;
            created_by: string;
            created_at: ZonedDateTime;
            updated_by?: string;
            updated_at?: ZonedDateTime;
            toolkits?: ToolkitsInAgentResponse;
            detail_mode?: boolean;
            model_id?: string;
            model_name?: string;
            structured_output?: {
                [key: string]: any;
            } | null;
            llm_settings?: {
                [key: string]: any;
            } | null;
        };
    } | {
        status: 404;
        data: {
            code?: string;
            details?: string;
            additionalInformation?: {
                [key: string]: any;
            } | null;
        };
    } | {
        status: 500;
        data: {
            code?: string;
            details?: string;
            additionalInformation?: {
                [key: string]: any;
            } | null;
        };
    }>(`/v1/public/agent/${encodeURIComponent(agentId)}`, {
        ...opts
    }));
}
/**
 * Update agent by ID
 */
export function putV1PublicAgentByAgentId({ agentId, modifiedAgentRequest }: {
    agentId: string;
    modifiedAgentRequest?: ModifiedAgentRequest;
}, opts?: Oazapfts.RequestOpts) {
    return oazapfts.ok(oazapfts.fetchJson<{
        status: 204;
    } | {
        status: 400;
        data: {
            code?: string;
            details?: string;
            violations?: ViolationResponse[];
        };
    } | {
        status: 404;
        data: {
            code?: string;
            details?: string;
            additionalInformation?: {
                [key: string]: any;
            } | null;
        };
    } | {
        status: 422;
        data: {
            code?: string;
            details?: string;
            additionalInformation?: {
                [key: string]: any;
            } | null;
        };
    } | {
        status: 500;
        data: {
            code?: string;
            details?: string;
            additionalInformation?: {
                [key: string]: any;
            } | null;
        };
    }>(`/v1/public/agent/${encodeURIComponent(agentId)}`, oazapfts.json({
        ...opts,
        method: "PUT",
        body: modifiedAgentRequest
    })));
}
/**
 * Delete agent by ID
 */
export function deleteV1PublicAgentByAgentId({ agentId }: {
    agentId: string;
}, opts?: Oazapfts.RequestOpts) {
    return oazapfts.ok(oazapfts.fetchJson<{
        status: 204;
    } | {
        status: 404;
        data: {
            code?: string;
            details?: string;
            additionalInformation?: {
                [key: string]: any;
            } | null;
        };
    } | {
        status: 422;
        data: {
            code?: string;
            details?: string;
            additionalInformation?: {
                [key: string]: any;
            } | null;
        };
    } | {
        status: 500;
        data: {
            code?: string;
            details?: string;
            additionalInformation?: {
                [key: string]: any;
            } | null;
        };
    }>(`/v1/public/agent/${encodeURIComponent(agentId)}`, {
        ...opts,
        method: "DELETE"
    }));
}
/**
 * AI Agent update visibility.
 */
export function postV1PublicAgentByAgentIdPublish({ agentId, publishRequest }: {
    agentId: string;
    publishRequest?: PublishRequest;
}, opts?: Oazapfts.RequestOpts) {
    return oazapfts.ok(oazapfts.fetchJson<{
        status: 204;
    } | {
        status: 400;
        data: {
            code?: string;
            details?: string;
            violations?: ViolationResponse[];
        };
    } | {
        status: 404;
        data: {
            code?: string;
            details?: string;
            additionalInformation?: {
                [key: string]: any;
            } | null;
        };
    } | {
        status: 422;
        data: {
            code?: string;
            details?: string;
            additionalInformation?: {
                [key: string]: any;
            } | null;
        };
    } | {
        status: 500;
        data: {
            code?: string;
            details?: string;
            additionalInformation?: {
                [key: string]: any;
            } | null;
        };
    }>(`/v1/public/agent/${encodeURIComponent(agentId)}/publish`, oazapfts.json({
        ...opts,
        method: "POST",
        body: publishRequest
    })));
}
/**
 * Get agent by Slug
 */
export function getV1PublicAgentByAgentSlugExists({ agentSlug }: {
    agentSlug: string;
}, opts?: Oazapfts.RequestOpts) {
    return oazapfts.ok(oazapfts.fetchJson<{
        status: 200;
        data: {
            id: string;
            name: string;
            slug: string;
            description?: string;
            avatar?: string;
            system_prompt: string;
            "type": string;
            llm_config?: LlmConfigResponse;
            suggested_prompts?: string[];
            knowledge_sources_config?: KnowledgeSourceResponse;
            visibility_level: VisibilityLevel;
            use_only: boolean;
            creator_name: string;
            created_by: string;
            created_at: ZonedDateTime;
            updated_by?: string;
            updated_at?: ZonedDateTime;
            toolkits?: ToolkitsInAgentResponse;
            detail_mode?: boolean;
            model_id?: string;
            model_name?: string;
            structured_output?: {
                [key: string]: any;
            } | null;
            llm_settings?: {
                [key: string]: any;
            } | null;
        };
    } | {
        status: 404;
        data: {
            code?: string;
            details?: string;
            additionalInformation?: {
                [key: string]: any;
            } | null;
        };
    } | {
        status: 500;
        data: {
            code?: string;
            details?: string;
            additionalInformation?: {
                [key: string]: any;
            } | null;
        };
    }>(`/v1/public/agent/${encodeURIComponent(agentSlug)}/exists`, {
        ...opts
    }));
}
/**
 * List built-in AI Agents filtered by visibility.
 */
export function getV1PublicAgents({ visibility }: {
    visibility?: VisibilityLevel;
}, opts?: Oazapfts.RequestOpts) {
    return oazapfts.ok(oazapfts.fetchJson<{
        status: 200;
        data: AgentResponse[];
    } | {
        status: 500;
        data: {
            code?: string;
            details?: string;
            additionalInformation?: {
                [key: string]: any;
            } | null;
        };
    }>(`/v1/public/agents${QS.query(QS.explode({
        visibility
    }))}`, {
        ...opts
    }));
}
