// @stack-spot/portal-network/src/api/apiManagement.ts
/**
 * OpenAPI definition
 * v0
 * DO NOT MODIFY - This file has been generated using oazapfts.
 * See https://www.npmjs.com/package/oazapfts
 */
import * as Oazapfts from "@oazapfts/runtime";
import * as QS from "@oazapfts/runtime/query";
export const defaults: Oazapfts.Defaults<Oazapfts.CustomHeaders> = {
    headers: {},
    baseUrl: "http://api-management-catalog-assets-api.stg.stackspot.com",
};
const oazapfts = Oazapfts.runtime(defaults);
export const servers = {
    generatedServerUrl: "http://api-management-catalog-assets-api.stg.stackspot.com"
};
export type CreateTagRequest = {
    key: string;
    value?: string;
};
export type JsonNode = object;
export type CreateSchemaRequest = {
    name: string;
    description: string;
    schema: JsonNode;
};
export type RequestParams = {
    page: number;
    size: number;
    search?: string;
    slugs?: string[];
    restrict: boolean;
    parentEntryId?: string;
    workspaceId?: string;
    presenter?: string;
    bySlug: boolean;
    isExact: boolean;
    attribute?: string;
    tag?: string[];
};
export type TagInputRequest = {
    key: string;
    value?: string;
};
export type CreateEntryRequest = {
    slug: string;
    tags?: TagInputRequest[];
    metadata: JsonNode;
    artifacts?: JsonNode;
    workspaceId: string;
};
export type CreateSubSchemaRequest = {
    name: string;
    description: string;
    is_dependency: boolean;
    schema: JsonNode;
};
export type CreateSubEntryRequest = {
    slug: string;
    tags?: {
        [key: string]: string;
    };
    parentEntryId: string;
    metadata: JsonNode;
    artifacts?: JsonNode;
    workspaceId: string;
};
export type UpdateEntryRequest = {
    tags?: TagInputRequest[];
    metadata?: JsonNode;
};
export type UpdateEntryTag = {
    key: string;
    value?: string;
};
export function list({ size, page, tag, search }: {
    size?: number;
    page?: number;
    tag?: string[];
    search?: string;
}, opts?: Oazapfts.RequestOpts) {
    return oazapfts.ok(oazapfts.fetchJson<{
        status: 200;
        data: object;
    }>(`/v1/tags${QS.query(QS.explode({
        size,
        page,
        tag,
        search
    }))}`, {
        ...opts
    }));
}
export function createTag({ createTagRequest }: {
    createTagRequest: CreateTagRequest;
}, opts?: Oazapfts.RequestOpts) {
    return oazapfts.ok(oazapfts.fetchJson<{
        status: 200;
        data: object;
    }>("/v1/tags", oazapfts.json({
        ...opts,
        method: "POST",
        body: createTagRequest
    })));
}
export function getObjectSchemas(opts?: Oazapfts.RequestOpts) {
    return oazapfts.ok(oazapfts.fetchText("/v1/schemas", {
        ...opts
    }));
}
export function createSchema({ createSchemaRequest }: {
    createSchemaRequest: CreateSchemaRequest;
}, opts?: Oazapfts.RequestOpts) {
    return oazapfts.ok(oazapfts.fetchJson<{
        status: 200;
        data: object;
    }>("/v1/schemas", oazapfts.json({
        ...opts,
        method: "POST",
        body: createSchemaRequest
    })));
}
export function uploadSchemaArtifact({ schemaId, versionId, contentLength }: {
    schemaId: string;
    versionId: string;
    contentLength: number;
}, opts?: Oazapfts.RequestOpts) {
    return oazapfts.ok(oazapfts.fetchJson<{
        status: 200;
        data: object;
    }>(`/v1/schemas/${encodeURIComponent(schemaId)}/versions/${encodeURIComponent(versionId)}/artifacts`, {
        ...opts,
        method: "POST",
        headers: oazapfts.mergeHeaders(opts?.headers, {
            "Content-Length": contentLength
        })
    }));
}
export function getEntries({ schemaId, schemaVersionId, queryParams, allRequestParams }: {
    schemaId: string;
    schemaVersionId: string;
    queryParams: RequestParams;
    allRequestParams: {
        [key: string]: string;
    };
}, opts?: Oazapfts.RequestOpts) {
    return oazapfts.ok(oazapfts.fetchJson<{
        status: 200;
        data: object;
    }>(`/v1/schemas/${encodeURIComponent(schemaId)}/versions/${encodeURIComponent(schemaVersionId)}/entries${QS.query(QS.explode({
        queryParams,
        allRequestParams
    }))}`, {
        ...opts
    }));
}
export function createEntry({ schemaId, schemaVersionId, createEntryRequest }: {
    schemaId: string;
    schemaVersionId: string;
    createEntryRequest: CreateEntryRequest;
}, opts?: Oazapfts.RequestOpts) {
    return oazapfts.ok(oazapfts.fetchJson<{
        status: 200;
        data: object;
    }>(`/v1/schemas/${encodeURIComponent(schemaId)}/versions/${encodeURIComponent(schemaVersionId)}/entries`, oazapfts.json({
        ...opts,
        method: "POST",
        body: createEntryRequest
    })));
}
export function getSubSchemas(opts?: Oazapfts.RequestOpts) {
    return oazapfts.ok(oazapfts.fetchText(`/v1/schemas/${encodeURIComponent(schemaId)}/subschemas`, {
        ...opts
    }));
}
export function createSubSchema({ schemaId, createSubSchemaRequest }: {
    schemaId: string;
    createSubSchemaRequest: CreateSubSchemaRequest;
}, opts?: Oazapfts.RequestOpts) {
    return oazapfts.ok(oazapfts.fetchJson<{
        status: 200;
        data: object;
    }>(`/v1/schemas/${encodeURIComponent(schemaId)}/subschemas`, oazapfts.json({
        ...opts,
        method: "POST",
        body: createSubSchemaRequest
    })));
}
export function uploadSchemaArtifact1({ schemaId, subSchemaId, versionId, contentLength }: {
    schemaId: string;
    subSchemaId: string;
    versionId: string;
    contentLength: number;
}, opts?: Oazapfts.RequestOpts) {
    return oazapfts.ok(oazapfts.fetchJson<{
        status: 200;
        data: object;
    }>(`/v1/schemas/${encodeURIComponent(schemaId)}/subschemas/${encodeURIComponent(subSchemaId)}/versions/${encodeURIComponent(versionId)}/artifacts`, {
        ...opts,
        method: "POST",
        headers: oazapfts.mergeHeaders(opts?.headers, {
            "Content-Length": contentLength
        })
    }));
}
export function getSubEntries({ schemaId, subSchemaId, subVersionId, allRequestParams, queryParams }: {
    schemaId: string;
    subSchemaId: string;
    subVersionId: string;
    allRequestParams: {
        [key: string]: string;
    };
    queryParams: RequestParams;
}, opts?: Oazapfts.RequestOpts) {
    return oazapfts.ok(oazapfts.fetchJson<{
        status: 200;
        data: object;
    }>(`/v1/schemas/${encodeURIComponent(schemaId)}/subschemas/${encodeURIComponent(subSchemaId)}/versions/${encodeURIComponent(subVersionId)}/entries${QS.query(QS.explode({
        allRequestParams,
        queryParams
    }))}`, {
        ...opts
    }));
}
export function createSubEntry({ schemaId, subSchemaId, subVersionId, createSubEntryRequest }: {
    schemaId: string;
    subSchemaId: string;
    subVersionId: string;
    createSubEntryRequest: CreateSubEntryRequest;
}, opts?: Oazapfts.RequestOpts) {
    return oazapfts.ok(oazapfts.fetchJson<{
        status: 200;
        data: object;
    }>(`/v1/schemas/${encodeURIComponent(schemaId)}/subschemas/${encodeURIComponent(subSchemaId)}/versions/${encodeURIComponent(subVersionId)}/entries`, oazapfts.json({
        ...opts,
        method: "POST",
        body: createSubEntryRequest
    })));
}
export function getObjectSchemaId({ schemaId }: {
    schemaId: string;
}, opts?: Oazapfts.RequestOpts) {
    return oazapfts.ok(oazapfts.fetchJson<{
        status: 200;
        data: object;
    }>(`/v1/schemas/${encodeURIComponent(schemaId)}`, {
        ...opts
    }));
}
export function createSchemaVersion({ schemaId, jsonNode }: {
    schemaId: string;
    jsonNode: JsonNode;
}, opts?: Oazapfts.RequestOpts) {
    return oazapfts.ok(oazapfts.fetchJson<{
        status: 200;
        data: object;
    }>(`/v1/schemas/${encodeURIComponent(schemaId)}`, oazapfts.json({
        ...opts,
        method: "PATCH",
        body: jsonNode
    })));
}
export function getEntryById({ schemaId, schemaVersionId, entryId, presenter }: {
    schemaId: string;
    schemaVersionId: string;
    entryId: string;
    presenter?: string;
}, opts?: Oazapfts.RequestOpts) {
    return oazapfts.ok(oazapfts.fetchJson<{
        status: 200;
        data: object;
    }>(`/v1/schemas/${encodeURIComponent(schemaId)}/versions/${encodeURIComponent(schemaVersionId)}/entries/${encodeURIComponent(entryId)}${QS.query(QS.explode({
        presenter
    }))}`, {
        ...opts
    }));
}
export function updateEntry({ schemaId, schemaVersionId, entryId, updateEntryRequest }: {
    schemaId: string;
    schemaVersionId: string;
    entryId: string;
    updateEntryRequest: UpdateEntryRequest;
}, opts?: Oazapfts.RequestOpts) {
    return oazapfts.ok(oazapfts.fetchJson<{
        status: 200;
        data: object;
    }>(`/v1/schemas/${encodeURIComponent(schemaId)}/versions/${encodeURIComponent(schemaVersionId)}/entries/${encodeURIComponent(entryId)}`, oazapfts.json({
        ...opts,
        method: "PATCH",
        body: updateEntryRequest
    })));
}
export function updateTag({ schemaId, schemaVersionId, entryId, oldTag, updateEntryTag }: {
    schemaId: string;
    schemaVersionId: string;
    entryId: string;
    oldTag: string;
    updateEntryTag: UpdateEntryTag;
}, opts?: Oazapfts.RequestOpts) {
    return oazapfts.ok(oazapfts.fetchJson<{
        status: 200;
        data: object;
    }>(`/v1/schemas/${encodeURIComponent(schemaId)}/versions/${encodeURIComponent(schemaVersionId)}/entries/${encodeURIComponent(entryId)}/tags/${encodeURIComponent(oldTag)}`, oazapfts.json({
        ...opts,
        method: "PATCH",
        body: updateEntryTag
    })));
}
export function downloadObjectArtifacts({ schemaId, schemaVersionId, entryId, name }: {
    schemaId: string;
    schemaVersionId: string;
    entryId: string;
    name?: string;
}, opts?: Oazapfts.RequestOpts) {
    return oazapfts.ok(oazapfts.fetchJson<{
        status: 200;
        data: object;
    }>(`/v1/schemas/${encodeURIComponent(schemaId)}/versions/${encodeURIComponent(schemaVersionId)}/entries/${encodeURIComponent(entryId)}/artifacts${QS.query(QS.explode({
        name
    }))}`, {
        ...opts
    }));
}
export function attachArtifacts({ schemaId, schemaVersionId, entryId, jsonNode }: {
    schemaId: string;
    schemaVersionId: string;
    entryId: string;
    jsonNode: JsonNode;
}, opts?: Oazapfts.RequestOpts) {
    return oazapfts.ok(oazapfts.fetchJson<{
        status: 200;
        data: object;
    }>(`/v1/schemas/${encodeURIComponent(schemaId)}/versions/${encodeURIComponent(schemaVersionId)}/entries/${encodeURIComponent(entryId)}/artifacts`, oazapfts.json({
        ...opts,
        method: "PATCH",
        body: jsonNode
    })));
}
export function downloadSchemaArtifact({ schemaId, schemaVersionId, entryId, artifactId }: {
    schemaId: string;
    schemaVersionId: string;
    entryId: string;
    artifactId: string;
}, opts?: Oazapfts.RequestOpts) {
    return oazapfts.ok(oazapfts.fetchJson<{
        status: 200;
        data: object;
    }>(`/v1/schemas/${encodeURIComponent(schemaId)}/versions/${encodeURIComponent(schemaVersionId)}/entries/${encodeURIComponent(entryId)}/artifacts/${encodeURIComponent(artifactId)}`, {
        ...opts
    }));
}
export function updatedArtifact({ artifactId, schemaId, schemaVersionId, entryId, contentLength }: {
    artifactId: string;
    schemaId: string;
    schemaVersionId: string;
    entryId: string;
    contentLength: number;
}, opts?: Oazapfts.RequestOpts) {
    return oazapfts.ok(oazapfts.fetchJson<{
        status: 200;
        data: object;
    }>(`/v1/schemas/${encodeURIComponent(schemaId)}/versions/${encodeURIComponent(schemaVersionId)}/entries/${encodeURIComponent(entryId)}/artifacts/${encodeURIComponent(artifactId)}`, {
        ...opts,
        method: "PATCH",
        headers: oazapfts.mergeHeaders(opts?.headers, {
            "Content-Length": contentLength
        })
    }));
}
export function getSubSchema({ schemaId, subSchemaId }: {
    schemaId: string;
    subSchemaId: string;
}, opts?: Oazapfts.RequestOpts) {
    return oazapfts.ok(oazapfts.fetchJson<{
        status: 200;
        data: object;
    }>(`/v1/schemas/${encodeURIComponent(schemaId)}/subschemas/${encodeURIComponent(subSchemaId)}`, {
        ...opts
    }));
}
export function newVersion({ schemaId, subSchemaId, jsonNode }: {
    schemaId: string;
    subSchemaId: string;
    jsonNode: JsonNode;
}, opts?: Oazapfts.RequestOpts) {
    return oazapfts.ok(oazapfts.fetchJson<{
        status: 200;
        data: object;
    }>(`/v1/schemas/${encodeURIComponent(schemaId)}/subschemas/${encodeURIComponent(subSchemaId)}`, oazapfts.json({
        ...opts,
        method: "PATCH",
        body: jsonNode
    })));
}
export function updateTag1({ schemaId, subSchemaId, subVersionId, subEntryId, oldTag, updateEntryTag }: {
    schemaId: string;
    subSchemaId: string;
    subVersionId: string;
    subEntryId: string;
    oldTag: string;
    updateEntryTag: UpdateEntryTag;
}, opts?: Oazapfts.RequestOpts) {
    return oazapfts.ok(oazapfts.fetchJson<{
        status: 200;
        data: object;
    }>(`/v1/schemas/${encodeURIComponent(schemaId)}/subschemas/${encodeURIComponent(subSchemaId)}/versions/${encodeURIComponent(subVersionId)}/entries/${encodeURIComponent(subEntryId)}/tags/${encodeURIComponent(oldTag)}`, oazapfts.json({
        ...opts,
        method: "PATCH",
        body: updateEntryTag
    })));
}
export function downloadObjectArtifacts1({ schemaId, subSchemaId, subVersionId, subEntryId, name }: {
    schemaId: string;
    subSchemaId: string;
    subVersionId: string;
    subEntryId: string;
    name?: string;
}, opts?: Oazapfts.RequestOpts) {
    return oazapfts.ok(oazapfts.fetchJson<{
        status: 200;
        data: object;
    }>(`/v1/schemas/${encodeURIComponent(schemaId)}/subschemas/${encodeURIComponent(subSchemaId)}/versions/${encodeURIComponent(subVersionId)}/entries/${encodeURIComponent(subEntryId)}/artifacts${QS.query(QS.explode({
        name
    }))}`, {
        ...opts
    }));
}
export function attachArtifacts1({ schemaId, subSchemaId, subVersionId, subEntryId, jsonNode }: {
    schemaId: string;
    subSchemaId: string;
    subVersionId: string;
    subEntryId: string;
    jsonNode: JsonNode;
}, opts?: Oazapfts.RequestOpts) {
    return oazapfts.ok(oazapfts.fetchJson<{
        status: 200;
        data: object;
    }>(`/v1/schemas/${encodeURIComponent(schemaId)}/subschemas/${encodeURIComponent(subSchemaId)}/versions/${encodeURIComponent(subVersionId)}/entries/${encodeURIComponent(subEntryId)}/artifacts`, oazapfts.json({
        ...opts,
        method: "PATCH",
        body: jsonNode
    })));
}
export function downloadSchemaArtifact1({ schemaId, subSchemaId, subVersionId, subEntryId, artifactId }: {
    schemaId: string;
    subSchemaId: string;
    subVersionId: string;
    subEntryId: string;
    artifactId: string;
}, opts?: Oazapfts.RequestOpts) {
    return oazapfts.ok(oazapfts.fetchJson<{
        status: 200;
        data: object;
    }>(`/v1/schemas/${encodeURIComponent(schemaId)}/subschemas/${encodeURIComponent(subSchemaId)}/versions/${encodeURIComponent(subVersionId)}/entries/${encodeURIComponent(subEntryId)}/artifacts/${encodeURIComponent(artifactId)}`, {
        ...opts
    }));
}
export function updatedArtifact1({ artifactId, schemaId, subSchemaId, subVersionId, subEntryId, contentLength }: {
    artifactId: string;
    schemaId: string;
    subSchemaId: string;
    subVersionId: string;
    subEntryId: string;
    contentLength: number;
}, opts?: Oazapfts.RequestOpts) {
    return oazapfts.ok(oazapfts.fetchJson<{
        status: 200;
        data: object;
    }>(`/v1/schemas/${encodeURIComponent(schemaId)}/subschemas/${encodeURIComponent(subSchemaId)}/versions/${encodeURIComponent(subVersionId)}/entries/${encodeURIComponent(subEntryId)}/artifacts/${encodeURIComponent(artifactId)}`, {
        ...opts,
        method: "PATCH",
        headers: oazapfts.mergeHeaders(opts?.headers, {
            "Content-Length": contentLength
        })
    }));
}
export function updateEntry1({ schemaId, subSchemaId, subVersionId, entryId, updateEntryRequest }: {
    schemaId: string;
    subSchemaId: string;
    subVersionId: string;
    entryId: string;
    updateEntryRequest: UpdateEntryRequest;
}, opts?: Oazapfts.RequestOpts) {
    return oazapfts.ok(oazapfts.fetchJson<{
        status: 200;
        data: object;
    }>(`/v1/schemas/${encodeURIComponent(schemaId)}/subschemas/${encodeURIComponent(subSchemaId)}/versions/${encodeURIComponent(subVersionId)}/entries/${encodeURIComponent(entryId)}`, oazapfts.json({
        ...opts,
        method: "PATCH",
        body: updateEntryRequest
    })));
}
export function getTag({ tag }: {
    tag: string;
}, opts?: Oazapfts.RequestOpts) {
    return oazapfts.ok(oazapfts.fetchJson<{
        status: 200;
        data: object;
    }>(`/v1/tags/${encodeURIComponent(tag)}`, {
        ...opts
    }));
}
export function getSchemaVersions(opts?: Oazapfts.RequestOpts) {
    return oazapfts.ok(oazapfts.fetchText(`/v1/schemas/${encodeURIComponent(schemaId)}/versions`, {
        ...opts
    }));
}
export function getSchemaVersionById({ schemaId, versionId }: {
    schemaId: string;
    versionId: string;
}, opts?: Oazapfts.RequestOpts) {
    return oazapfts.ok(oazapfts.fetchJson<{
        status: 200;
        data: object;
    }>(`/v1/schemas/${encodeURIComponent(schemaId)}/versions/${encodeURIComponent(versionId)}`, {
        ...opts
    }));
}
export function getSchemaVersions1(opts?: Oazapfts.RequestOpts) {
    return oazapfts.ok(oazapfts.fetchText(`/v1/schemas/${encodeURIComponent(schemaId)}/subschemas/${encodeURIComponent(subSchemaId)}/versions`, {
        ...opts
    }));
}
export function getSchemaVersionById1(opts?: Oazapfts.RequestOpts) {
    return oazapfts.ok(oazapfts.fetchText(`/v1/schemas/${encodeURIComponent(schemaId)}/subschemas/${encodeURIComponent(subSchemaId)}/versions/${encodeURIComponent(versionId)}`, {
        ...opts
    }));
}
export function getSubEntriesById({ schemaId, subSchemaId, subVersionId, subEntryId }: {
    schemaId: string;
    subSchemaId: string;
    subVersionId: string;
    subEntryId: string;
}, opts?: Oazapfts.RequestOpts) {
    return oazapfts.ok(oazapfts.fetchJson<{
        status: 200;
        data: object;
    }>(`/v1/schemas/${encodeURIComponent(schemaId)}/subschemas/${encodeURIComponent(subSchemaId)}/versions/${encodeURIComponent(subVersionId)}/entries/${encodeURIComponent(subEntryId)}`, {
        ...opts
    }));
}
export function removeTag({ schemaId, schemaVersionId, entryId, tag }: {
    schemaId: string;
    schemaVersionId: string;
    entryId: string;
    tag: string;
}, opts?: Oazapfts.RequestOpts) {
    return oazapfts.ok(oazapfts.fetchJson<{
        status: 200;
        data: object;
    }>(`/v1/schemas/${encodeURIComponent(schemaId)}/versions/${encodeURIComponent(schemaVersionId)}/entries/${encodeURIComponent(entryId)}/tags/${encodeURIComponent(tag)}`, {
        ...opts,
        method: "DELETE"
    }));
}
export function removeTag1({ schemaId, subSchemaId, subVersionId, subEntryId, tag }: {
    schemaId: string;
    subSchemaId: string;
    subVersionId: string;
    subEntryId: string;
    tag: string;
}, opts?: Oazapfts.RequestOpts) {
    return oazapfts.ok(oazapfts.fetchJson<{
        status: 200;
        data: object;
    }>(`/v1/schemas/${encodeURIComponent(schemaId)}/subschemas/${encodeURIComponent(subSchemaId)}/versions/${encodeURIComponent(subVersionId)}/entries/${encodeURIComponent(subEntryId)}/tags/${encodeURIComponent(tag)}`, {
        ...opts,
        method: "DELETE"
    }));
}
