// @stack-spot/portal-network/src/api/cloudPlatform.ts
/**
 * OpenAPI definition
 * v0
 * DO NOT MODIFY - This file has been generated using oazapfts.
 * See https://www.npmjs.com/package/oazapfts
 */
import * as Oazapfts from "@oazapfts/runtime";
import * as QS from "@oazapfts/runtime/query";
export const defaults: Oazapfts.Defaults<Oazapfts.CustomHeaders> = {
    headers: {},
    baseUrl: "https://cloud-cloud-platform-api.dev.stackspot.com",
};
const oazapfts = Oazapfts.runtime(defaults);
export const servers = {
    generatedServerUrl: "https://cloud-cloud-platform-api.dev.stackspot.com"
};
export type Tag = {
    key: string;
    value: string;
};
export type TunnelStatus = {
    tunnel1?: string;
    tunnel2?: string;
};
export type VpnDetails = {
    peerCustomerGatewayIp: string;
    destinationCidrBlock: string;
    tunnelStatus: TunnelStatus;
};
export type VpnResponse = {
    stackSpotAccountId: string;
    foundationId: string;
    vpnId: string;
    details: VpnDetails;
    tags: Tag[];
    status: "READY" | "PENDING" | "PENDING_APPROVAL" | "ERROR";
    createdAt: string;
    updatedAt: string;
    deletedAt?: string;
};
export type ProjectDetails = {
    name: string;
    folderName: string;
    providerAccountId?: string;
};
export type ProjectResponse = {
    stackSpotAccountId: string;
    foundationId: string;
    parentFolderId: string;
    projectId: string;
    details: ProjectDetails;
    tags: Tag[];
    status: "READY" | "PENDING" | "PENDING_APPROVAL" | "ERROR";
    createdAt: string;
    updatedAt: string;
    deletedAt?: string;
};
export type NetworkSubnetDetails = {
    az?: string;
    cidr?: string;
    id?: string;
    "type": "NAT" | "PRIVATE" | "PUBLIC";
};
export type NetworkDetails = {
    cidrs?: string[];
    name: string;
    "type": "WORKLOAD";
    subnets: NetworkSubnetDetails[];
};
export type NetworkResponse = {
    stackSpotAccountId: string;
    foundationId: string;
    projectId: string;
    networkId: string;
    details: NetworkDetails;
    tags: Tag[];
    status: "READY" | "PENDING" | "PENDING_APPROVAL" | "ERROR";
    createdAt: string;
    updatedAt: string;
    deletedAt?: string;
};
export type FolderContent = {
    "type": "FOLDER" | "PROJECT";
    id: string;
    name: string;
    status: "READY" | "PENDING" | "PENDING_APPROVAL" | "ERROR";
};
export type FolderDetails = {
    name: string;
    pendingResources: boolean;
    content: FolderContent[];
};
export type FolderResponse = {
    stackSpotAccountId: string;
    foundationId: string;
    parentFolderId: string;
    folderId: string;
    details: FolderDetails;
    tags: Tag[];
    status: "READY" | "PENDING" | "PENDING_APPROVAL" | "ERROR";
    createdAt: string;
    updatedAt: string;
    deletedAt?: string;
};
export type DnsZoneDetails = {
    domain: string;
    dnsZoneType: string;
};
export type DnsZoneResponse = {
    stackSpotAccountId: string;
    foundationId: string;
    dnsZoneId: string;
    details: DnsZoneDetails;
    tags: Tag[];
    status: "READY" | "PENDING" | "PENDING_APPROVAL" | "ERROR";
    createdAt: string;
    updatedAt: string;
    deletedAt?: string;
};
export type DnsRecordDetails = {
    recordName: string;
    records: string[];
    ttl: number;
    "type": "A" | "AAAA" | "CAA" | "CNAME" | "MX" | "PTR" | "SOA" | "SRV" | "TXT";
    dnsZoneName: string;
};
export type DnsRecordResponse = {
    stackSpotAccountId: string;
    foundationId: string;
    dnsZoneId: string;
    dnsRecordId: string;
    details: DnsRecordDetails;
    tags: Tag[];
    status: "READY" | "PENDING" | "PENDING_APPROVAL" | "ERROR";
    createdAt: string;
    updatedAt: string;
    deletedAt?: string;
};
export type CidrDetails = {
    cidrNotation: string;
    networkAddress: string;
    netmask: string;
    broadcastAddress: string;
    lowAddress: string;
    highAddress: string;
    addressCount: number;
};
export type CidrResponse = {
    stackSpotAccountId: string;
    foundationId: string;
    cidrId: string;
    details: CidrDetails;
    tags: Tag[];
    status: "READY" | "PENDING" | "PENDING_APPROVAL" | "ERROR";
    createdAt: string;
    updatedAt: string;
    deletedAt?: string;
};
export type CertificateDomainValidation = {
    domainName?: string;
    recordName?: string;
    recordType?: string;
    recordValue?: string;
};
export type CertificateDetails = {
    name: string;
    forInbound: "TRUE" | "FALSE";
    certificateStatus?: string;
    domainName?: string;
    domainValidation?: CertificateDomainValidation[];
    notAfter?: string;
    notBefore?: string;
    "type"?: string;
    validationMethod?: string;
};
export type CertificateResponse = {
    stackSpotAccountId: string;
    foundationId: string;
    certificateId: string;
    details: CertificateDetails;
    status: "READY" | "PENDING" | "PENDING_APPROVAL" | "ERROR";
    tags: Tag[];
    createdAt: string;
    updatedAt: string;
    deletedAt?: string;
};
export type FoundationDetails = {
    name: string;
    description: string;
    cloudProvider: string;
    region: string;
};
export type FoundationResponse = {
    stackSpotAccountId: string;
    foundationId: string;
    details: FoundationDetails;
    tags: Tag[];
    status: "READY" | "PENDING" | "PENDING_APPROVAL" | "ERROR";
    createdAt: string;
    updatedAt: string;
    deletedAt?: string;
};
export type ListFoundationResponse = {
    stackSpotAccountId: string;
    pendingResources: boolean;
    content: FoundationResponse[];
};
export type CreateFoundationRequest = {
    name: string;
    description: string;
    cloudProvider: string;
    region: string;
};
export type ListVpnResponse = {
    stackSpotAccountId: string;
    foundationId: string;
    pendingResources: boolean;
    content: VpnResponse[];
};
export type CreateVpnRequest = {
    peerCustomerGatewayIp: string;
    destinationCidrBlock: string;
    tags?: Tag[];
};
export type TenantDetails = {
    name: string;
};
export type TenantResponse = {
    stackSpotAccountId: string;
    foundationId: string;
    tenantId: string;
    details: TenantDetails;
    status: "READY" | "PENDING" | "PENDING_APPROVAL" | "ERROR";
    createdAt: string;
    updatedAt: string;
    deletedAt?: string;
};
export type ListTenantResponse = {
    stackSpotAccountId: string;
    foundationId: string;
    pendingResources: boolean;
    content: TenantResponse[];
};
export type CreateTenantRequest = {
    name: string;
    publicCidrs: string[];
};
export type RuntimeDetails = {
    name: string;
};
export type RuntimeResponse = {
    stackSpotAccountId: string;
    foundationId: string;
    tenantId: string;
    runtimeId: string;
    details: RuntimeDetails;
    status: "READY" | "PENDING" | "PENDING_APPROVAL" | "ERROR";
    createdAt: string;
    updatedAt: string;
    deletedAt?: string;
};
export type ListRuntimeResponse = {
    stackSpotAccountId: string;
    foundationId: string;
    projectId?: string;
    tenantId?: string;
    pendingResources: boolean;
    content: RuntimeResponse[];
};
export type CreateRuntimeRequest = {
    projectId: string;
    name: string;
    publicCidrs: string[];
    clusterAdminRoles: string[];
};
export type ListProjectResponse = {
    stackSpotAccountId: string;
    foundationId: string;
    parentFolderId?: string;
    pendingResources: boolean;
    content: ProjectResponse[];
};
export type CreateProjectRequest = {
    parentFolderId: string;
    name: string;
    description: string;
    tags?: Tag[];
};
export type ListNetworkResponse = {
    stackSpotAccountId: string;
    foundationId: string;
    projectId?: string;
    pendingResources: boolean;
    content: NetworkResponse[];
};
export type CreateNetworkRequest = {
    projectId: string;
    prefixLength: number;
    networkName?: string;
    networkType: "WORKLOAD";
    tags?: Tag[];
};
export type CreateFolderRequest = {
    parentFolderId: string;
    name: string;
    tags?: Tag[];
};
export type ListDnsZoneResponse = {
    stackSpotAccountId: string;
    foundationId: string;
    pendingResources: boolean;
    content: DnsZoneResponse[];
};
export type CreateDnsZoneRequest = {
    domain: string;
    "type": "PUBLIC" | "PRIVATE";
    projectId?: string;
    tags?: Tag[];
};
export type ListDnsRecordResponse = {
    stackSpotAccountId: string;
    foundationId: string;
    projectId?: string;
    dnsZoneId?: string;
    pendingResources: boolean;
    content: DnsRecordResponse[];
};
export type CreateDnsRecordRequest = {
    dnsZoneId: string;
    projectId?: string;
    recordName: string;
    records: string[];
    ttl: number;
    "type": "A" | "AAAA" | "CAA" | "CNAME" | "MX" | "PTR" | "SOA" | "SRV" | "TXT";
};
export type ListCidrResponse = {
    stackSpotAccountId: string;
    foundationId: string;
    pendingResources: boolean;
    content: CidrResponse[];
};
export type CreateCidrRequest = {
    address: string;
    prefix: number;
    tags?: Tag[];
};
export type ListCertificateResponse = {
    stackSpotAccountId: string;
    foundationId: string;
    pendingResources: boolean;
    content: CertificateResponse[];
};
export type CreateCertificateRequestBase = {
    tags?: Tag[];
    certificateName: string;
    forInbound: "TRUE" | "FALSE";
    "type": string;
};
export type CreatePublicCertificateRequest = {
    "type": "CreatePublicCertificateRequest";
} & CreateCertificateRequestBase & {
    domainName?: string;
};
export type ImportCertificateRequest = {
    "type": "ImportCertificateRequest";
} & CreateCertificateRequestBase & {
    certificateBody?: string;
    certificatePrivateKey?: string;
    certificateChain?: string;
};
export type Configuration = {
    ikev1: string;
    ikev2: string;
};
export type VpnConfigurationResponse = {
    vpnId: string;
    foundationId: string;
    configuration: Configuration;
};
export function putVpnTags({ authorization, xAccountId, foundationId, vpnId, body }: {
    authorization: string;
    xAccountId?: string;
    foundationId: string;
    vpnId: string;
    body: Tag[];
}, opts?: Oazapfts.RequestOpts) {
    return oazapfts.ok(oazapfts.fetchJson<{
        status: 202;
        data: VpnResponse;
    }>(`/v1/foundations/${encodeURIComponent(foundationId)}/vpns/${encodeURIComponent(vpnId)}/tags`, oazapfts.json({
        ...opts,
        method: "PUT",
        body,
        headers: oazapfts.mergeHeaders(opts?.headers, {
            Authorization: authorization,
            "x-account-id": xAccountId
        })
    })));
}
export function putProjectTags({ authorization, xAccountId, foundationId, projectId, body }: {
    authorization: string;
    xAccountId?: string;
    foundationId: string;
    projectId: string;
    body: Tag[];
}, opts?: Oazapfts.RequestOpts) {
    return oazapfts.ok(oazapfts.fetchJson<{
        status: 202;
        data: ProjectResponse;
    }>(`/v1/foundations/${encodeURIComponent(foundationId)}/projects/${encodeURIComponent(projectId)}/tags`, oazapfts.json({
        ...opts,
        method: "PUT",
        body,
        headers: oazapfts.mergeHeaders(opts?.headers, {
            Authorization: authorization,
            "x-account-id": xAccountId
        })
    })));
}
export function putNetworkTags({ authorization, xAccountId, foundationId, networkId, body }: {
    authorization: string;
    xAccountId?: string;
    foundationId: string;
    networkId: string;
    body: Tag[];
}, opts?: Oazapfts.RequestOpts) {
    return oazapfts.ok(oazapfts.fetchJson<{
        status: 202;
        data: NetworkResponse;
    }>(`/v1/foundations/${encodeURIComponent(foundationId)}/networks/${encodeURIComponent(networkId)}/tags`, oazapfts.json({
        ...opts,
        method: "PUT",
        body,
        headers: oazapfts.mergeHeaders(opts?.headers, {
            Authorization: authorization,
            "x-account-id": xAccountId
        })
    })));
}
export function getFolderTags({ authorization, xAccountId, foundationId, folderId }: {
    authorization: string;
    xAccountId?: string;
    foundationId: string;
    folderId: string;
}, opts?: Oazapfts.RequestOpts) {
    return oazapfts.ok(oazapfts.fetchJson<{
        status: 202;
        data: Tag[];
    }>(`/v1/foundations/${encodeURIComponent(foundationId)}/folders/${encodeURIComponent(folderId)}/tags`, {
        ...opts,
        headers: oazapfts.mergeHeaders(opts?.headers, {
            Authorization: authorization,
            "x-account-id": xAccountId
        })
    }));
}
export function putFolderTags({ authorization, xAccountId, foundationId, folderId, body }: {
    authorization: string;
    xAccountId?: string;
    foundationId: string;
    folderId: string;
    body: Tag[];
}, opts?: Oazapfts.RequestOpts) {
    return oazapfts.ok(oazapfts.fetchJson<{
        status: 202;
        data: FolderResponse;
    }>(`/v1/foundations/${encodeURIComponent(foundationId)}/folders/${encodeURIComponent(folderId)}/tags`, oazapfts.json({
        ...opts,
        method: "PUT",
        body,
        headers: oazapfts.mergeHeaders(opts?.headers, {
            Authorization: authorization,
            "x-account-id": xAccountId
        })
    })));
}
export function putDnsZoneTags({ authorization, xAccountId, foundationId, dnsZoneId, body }: {
    authorization: string;
    xAccountId?: string;
    foundationId: string;
    dnsZoneId: string;
    body: Tag[];
}, opts?: Oazapfts.RequestOpts) {
    return oazapfts.ok(oazapfts.fetchJson<{
        status: 202;
        data: DnsZoneResponse;
    }>(`/v1/foundations/${encodeURIComponent(foundationId)}/dns-zones/${encodeURIComponent(dnsZoneId)}/tags`, oazapfts.json({
        ...opts,
        method: "PUT",
        body,
        headers: oazapfts.mergeHeaders(opts?.headers, {
            Authorization: authorization,
            "x-account-id": xAccountId
        })
    })));
}
export function putDnsRecordRecords({ authorization, xAccountId, foundationId, dnsRecordId, body }: {
    authorization: string;
    xAccountId?: string;
    foundationId: string;
    dnsRecordId: string;
    body: string[];
}, opts?: Oazapfts.RequestOpts) {
    return oazapfts.ok(oazapfts.fetchJson<{
        status: 200;
        data: DnsRecordResponse;
    }>(`/v1/foundations/${encodeURIComponent(foundationId)}/dns-records/${encodeURIComponent(dnsRecordId)}/records`, oazapfts.json({
        ...opts,
        method: "PUT",
        body,
        headers: oazapfts.mergeHeaders(opts?.headers, {
            Authorization: authorization,
            "x-account-id": xAccountId
        })
    })));
}
export function putCidrTags({ authorization, xAccountId, foundationId, cidrId, body }: {
    authorization: string;
    xAccountId?: string;
    foundationId: string;
    cidrId: string;
    body: Tag[];
}, opts?: Oazapfts.RequestOpts) {
    return oazapfts.ok(oazapfts.fetchJson<{
        status: 202;
        data: CidrResponse;
    }>(`/v1/foundations/${encodeURIComponent(foundationId)}/cidrs/${encodeURIComponent(cidrId)}/tags`, oazapfts.json({
        ...opts,
        method: "PUT",
        body,
        headers: oazapfts.mergeHeaders(opts?.headers, {
            Authorization: authorization,
            "x-account-id": xAccountId
        })
    })));
}
export function putCertificateTags({ authorization, xAccountId, foundationId, certificateId, body }: {
    authorization: string;
    xAccountId?: string;
    foundationId: string;
    certificateId: string;
    body: Tag[];
}, opts?: Oazapfts.RequestOpts) {
    return oazapfts.ok(oazapfts.fetchJson<{
        status: 202;
        data: CertificateResponse;
    }>(`/v1/foundations/${encodeURIComponent(foundationId)}/certificates/${encodeURIComponent(certificateId)}/tags`, oazapfts.json({
        ...opts,
        method: "PUT",
        body,
        headers: oazapfts.mergeHeaders(opts?.headers, {
            Authorization: authorization,
            "x-account-id": xAccountId
        })
    })));
}
export function listFoundations({ authorization, xAccountId }: {
    authorization: string;
    xAccountId?: string;
}, opts?: Oazapfts.RequestOpts) {
    return oazapfts.ok(oazapfts.fetchJson<{
        status: 202;
        data: ListFoundationResponse;
    }>("/v1/foundations", {
        ...opts,
        headers: oazapfts.mergeHeaders(opts?.headers, {
            Authorization: authorization,
            "x-account-id": xAccountId
        })
    }));
}
export function createFoundation({ authorization, xAccountId, createFoundationRequest }: {
    authorization: string;
    xAccountId?: string;
    createFoundationRequest: CreateFoundationRequest;
}, opts?: Oazapfts.RequestOpts) {
    return oazapfts.ok(oazapfts.fetchJson<{
        status: 202;
        data: FoundationResponse;
    }>("/v1/foundations", oazapfts.json({
        ...opts,
        method: "POST",
        body: createFoundationRequest,
        headers: oazapfts.mergeHeaders(opts?.headers, {
            Authorization: authorization,
            "x-account-id": xAccountId
        })
    })));
}
export function listVpns({ authorization, xAccountId, foundationId }: {
    authorization: string;
    xAccountId?: string;
    foundationId: string;
}, opts?: Oazapfts.RequestOpts) {
    return oazapfts.ok(oazapfts.fetchJson<{
        status: 200;
        data: ListVpnResponse;
    }>(`/v1/foundations/${encodeURIComponent(foundationId)}/vpns`, {
        ...opts,
        headers: oazapfts.mergeHeaders(opts?.headers, {
            Authorization: authorization,
            "x-account-id": xAccountId
        })
    }));
}
export function createVpn({ authorization, xAccountId, foundationId, createVpnRequest }: {
    authorization: string;
    xAccountId?: string;
    foundationId: string;
    createVpnRequest: CreateVpnRequest;
}, opts?: Oazapfts.RequestOpts) {
    return oazapfts.ok(oazapfts.fetchJson<{
        status: 202;
        data: VpnResponse;
    }>(`/v1/foundations/${encodeURIComponent(foundationId)}/vpns`, oazapfts.json({
        ...opts,
        method: "POST",
        body: createVpnRequest,
        headers: oazapfts.mergeHeaders(opts?.headers, {
            Authorization: authorization,
            "x-account-id": xAccountId
        })
    })));
}
export function listTenant({ authorization, xAccountId, foundationId }: {
    authorization: string;
    xAccountId?: string;
    foundationId: string;
}, opts?: Oazapfts.RequestOpts) {
    return oazapfts.ok(oazapfts.fetchJson<{
        status: 200;
        data: ListTenantResponse;
    }>(`/v1/foundations/${encodeURIComponent(foundationId)}/tenants`, {
        ...opts,
        headers: oazapfts.mergeHeaders(opts?.headers, {
            Authorization: authorization,
            "x-account-id": xAccountId
        })
    }));
}
export function createTenant({ authorization, xAccountId, foundationId, createTenantRequest }: {
    authorization: string;
    xAccountId?: string;
    foundationId: string;
    createTenantRequest: CreateTenantRequest;
}, opts?: Oazapfts.RequestOpts) {
    return oazapfts.ok(oazapfts.fetchJson<{
        status: 202;
        data: TenantResponse;
    }>(`/v1/foundations/${encodeURIComponent(foundationId)}/tenants`, oazapfts.json({
        ...opts,
        method: "POST",
        body: createTenantRequest,
        headers: oazapfts.mergeHeaders(opts?.headers, {
            Authorization: authorization,
            "x-account-id": xAccountId
        })
    })));
}
export function listRuntime({ authorization, xAccountId, foundationId, tenantId, projectId }: {
    authorization: string;
    xAccountId?: string;
    foundationId: string;
    tenantId?: string;
    projectId?: string;
}, opts?: Oazapfts.RequestOpts) {
    return oazapfts.ok(oazapfts.fetchJson<{
        status: 200;
        data: ListRuntimeResponse;
    }>(`/v1/foundations/${encodeURIComponent(foundationId)}/runtimes${QS.query(QS.explode({
        tenantId,
        projectId
    }))}`, {
        ...opts,
        headers: oazapfts.mergeHeaders(opts?.headers, {
            Authorization: authorization,
            "x-account-id": xAccountId
        })
    }));
}
export function createRuntime({ authorization, xAccountId, foundationId, createRuntimeRequest }: {
    authorization: string;
    xAccountId?: string;
    foundationId: string;
    createRuntimeRequest: CreateRuntimeRequest;
}, opts?: Oazapfts.RequestOpts) {
    return oazapfts.ok(oazapfts.fetchJson<{
        status: 202;
        data: RuntimeResponse;
    }>(`/v1/foundations/${encodeURIComponent(foundationId)}/runtimes`, oazapfts.json({
        ...opts,
        method: "POST",
        body: createRuntimeRequest,
        headers: oazapfts.mergeHeaders(opts?.headers, {
            Authorization: authorization,
            "x-account-id": xAccountId
        })
    })));
}
export function listProject({ authorization, xAccountId, foundationId, parentFolderId }: {
    authorization: string;
    xAccountId?: string;
    foundationId: string;
    parentFolderId?: string;
}, opts?: Oazapfts.RequestOpts) {
    return oazapfts.ok(oazapfts.fetchJson<{
        status: 200;
        data: ListProjectResponse;
    }>(`/v1/foundations/${encodeURIComponent(foundationId)}/projects${QS.query(QS.explode({
        parentFolderId
    }))}`, {
        ...opts,
        headers: oazapfts.mergeHeaders(opts?.headers, {
            Authorization: authorization,
            "x-account-id": xAccountId
        })
    }));
}
export function createProject({ authorization, xAccountId, foundationId, createProjectRequest }: {
    authorization: string;
    xAccountId?: string;
    foundationId: string;
    createProjectRequest: CreateProjectRequest;
}, opts?: Oazapfts.RequestOpts) {
    return oazapfts.ok(oazapfts.fetchJson<{
        status: 202;
        data: ProjectResponse;
    }>(`/v1/foundations/${encodeURIComponent(foundationId)}/projects`, oazapfts.json({
        ...opts,
        method: "POST",
        body: createProjectRequest,
        headers: oazapfts.mergeHeaders(opts?.headers, {
            Authorization: authorization,
            "x-account-id": xAccountId
        })
    })));
}
export function listNetwork({ authorization, xAccountId, foundationId, projectId }: {
    authorization: string;
    xAccountId?: string;
    foundationId: string;
    projectId?: string;
}, opts?: Oazapfts.RequestOpts) {
    return oazapfts.ok(oazapfts.fetchJson<{
        status: 200;
        data: ListNetworkResponse;
    }>(`/v1/foundations/${encodeURIComponent(foundationId)}/networks${QS.query(QS.explode({
        projectId
    }))}`, {
        ...opts,
        headers: oazapfts.mergeHeaders(opts?.headers, {
            Authorization: authorization,
            "x-account-id": xAccountId
        })
    }));
}
export function createNetwork({ authorization, xAccountId, foundationId, createNetworkRequest }: {
    authorization: string;
    xAccountId?: string;
    foundationId: string;
    createNetworkRequest: CreateNetworkRequest;
}, opts?: Oazapfts.RequestOpts) {
    return oazapfts.ok(oazapfts.fetchJson<{
        status: 202;
        data: NetworkResponse;
    }>(`/v1/foundations/${encodeURIComponent(foundationId)}/networks`, oazapfts.json({
        ...opts,
        method: "POST",
        body: createNetworkRequest,
        headers: oazapfts.mergeHeaders(opts?.headers, {
            Authorization: authorization,
            "x-account-id": xAccountId
        })
    })));
}
export function getFolder({ authorization, xAccountId, foundationId, folderId }: {
    authorization: string;
    xAccountId?: string;
    foundationId: string;
    folderId: string;
}, opts?: Oazapfts.RequestOpts) {
    return oazapfts.ok(oazapfts.fetchJson<{
        status: 200;
        data: FolderResponse;
    }>(`/v1/foundations/${encodeURIComponent(foundationId)}/folders${QS.query(QS.explode({
        folderId
    }))}`, {
        ...opts,
        headers: oazapfts.mergeHeaders(opts?.headers, {
            Authorization: authorization,
            "x-account-id": xAccountId
        })
    }));
}
export function createFolder({ authorization, xAccountId, foundationId, createFolderRequest }: {
    authorization: string;
    xAccountId?: string;
    foundationId: string;
    createFolderRequest: CreateFolderRequest;
}, opts?: Oazapfts.RequestOpts) {
    return oazapfts.ok(oazapfts.fetchJson<{
        status: 202;
        data: FolderResponse;
    }>(`/v1/foundations/${encodeURIComponent(foundationId)}/folders`, oazapfts.json({
        ...opts,
        method: "POST",
        body: createFolderRequest,
        headers: oazapfts.mergeHeaders(opts?.headers, {
            Authorization: authorization,
            "x-account-id": xAccountId
        })
    })));
}
export function listDnsZone({ authorization, xAccountId, foundationId, projectId, privacy }: {
    authorization: string;
    xAccountId?: string;
    foundationId: string;
    projectId?: string;
    privacy?: "PUBLIC" | "PRIVATE";
}, opts?: Oazapfts.RequestOpts) {
    return oazapfts.ok(oazapfts.fetchJson<{
        status: 200;
        data: ListDnsZoneResponse;
    }>(`/v1/foundations/${encodeURIComponent(foundationId)}/dns-zones${QS.query(QS.explode({
        projectId,
        privacy
    }))}`, {
        ...opts,
        headers: oazapfts.mergeHeaders(opts?.headers, {
            Authorization: authorization,
            "x-account-id": xAccountId
        })
    }));
}
export function createDnsZone({ authorization, xAccountId, foundationId, createDnsZoneRequest }: {
    authorization: string;
    xAccountId?: string;
    foundationId: string;
    createDnsZoneRequest: CreateDnsZoneRequest;
}, opts?: Oazapfts.RequestOpts) {
    return oazapfts.ok(oazapfts.fetchJson<{
        status: 202;
        data: DnsZoneResponse;
    }>(`/v1/foundations/${encodeURIComponent(foundationId)}/dns-zones`, oazapfts.json({
        ...opts,
        method: "POST",
        body: createDnsZoneRequest,
        headers: oazapfts.mergeHeaders(opts?.headers, {
            Authorization: authorization,
            "x-account-id": xAccountId
        })
    })));
}
export function listDnsRecord({ authorization, xAccountId, foundationId, projectId, dnsZoneId }: {
    authorization: string;
    xAccountId?: string;
    foundationId: string;
    projectId?: string;
    dnsZoneId?: string;
}, opts?: Oazapfts.RequestOpts) {
    return oazapfts.ok(oazapfts.fetchJson<{
        status: 200;
        data: ListDnsRecordResponse;
    }>(`/v1/foundations/${encodeURIComponent(foundationId)}/dns-records${QS.query(QS.explode({
        projectId,
        dnsZoneId
    }))}`, {
        ...opts,
        headers: oazapfts.mergeHeaders(opts?.headers, {
            Authorization: authorization,
            "x-account-id": xAccountId
        })
    }));
}
export function createDnsRecord({ authorization, xAccountId, foundationId, createDnsRecordRequest }: {
    authorization: string;
    xAccountId?: string;
    foundationId: string;
    createDnsRecordRequest: CreateDnsRecordRequest;
}, opts?: Oazapfts.RequestOpts) {
    return oazapfts.ok(oazapfts.fetchJson<{
        status: 202;
        data: DnsRecordResponse;
    }>(`/v1/foundations/${encodeURIComponent(foundationId)}/dns-records`, oazapfts.json({
        ...opts,
        method: "POST",
        body: createDnsRecordRequest,
        headers: oazapfts.mergeHeaders(opts?.headers, {
            Authorization: authorization,
            "x-account-id": xAccountId
        })
    })));
}
export function listCidr({ authorization, xAccountId, foundationId }: {
    authorization: string;
    xAccountId?: string;
    foundationId: string;
}, opts?: Oazapfts.RequestOpts) {
    return oazapfts.ok(oazapfts.fetchJson<{
        status: 200;
        data: ListCidrResponse;
    }>(`/v1/foundations/${encodeURIComponent(foundationId)}/cidrs`, {
        ...opts,
        headers: oazapfts.mergeHeaders(opts?.headers, {
            Authorization: authorization,
            "x-account-id": xAccountId
        })
    }));
}
export function createCidr({ authorization, xAccountId, foundationId, createCidrRequest }: {
    authorization: string;
    xAccountId?: string;
    foundationId: string;
    createCidrRequest: CreateCidrRequest;
}, opts?: Oazapfts.RequestOpts) {
    return oazapfts.ok(oazapfts.fetchJson<{
        status: 202;
        data: CidrResponse;
    }>(`/v1/foundations/${encodeURIComponent(foundationId)}/cidrs`, oazapfts.json({
        ...opts,
        method: "POST",
        body: createCidrRequest,
        headers: oazapfts.mergeHeaders(opts?.headers, {
            Authorization: authorization,
            "x-account-id": xAccountId
        })
    })));
}
export function listCertificates({ authorization, xAccountId, foundationId, forInbound }: {
    authorization: string;
    xAccountId?: string;
    foundationId: string;
    forInbound?: "TRUE" | "FALSE";
}, opts?: Oazapfts.RequestOpts) {
    return oazapfts.ok(oazapfts.fetchJson<{
        status: 200;
        data: ListCertificateResponse;
    }>(`/v1/foundations/${encodeURIComponent(foundationId)}/certificates${QS.query(QS.explode({
        forInbound
    }))}`, {
        ...opts,
        headers: oazapfts.mergeHeaders(opts?.headers, {
            Authorization: authorization,
            "x-account-id": xAccountId
        })
    }));
}
export function createCertificate({ authorization, xAccountId, foundationId, body }: {
    authorization: string;
    xAccountId?: string;
    foundationId: string;
    body: CreatePublicCertificateRequest | ImportCertificateRequest;
}, opts?: Oazapfts.RequestOpts) {
    return oazapfts.ok(oazapfts.fetchJson<{
        status: 202;
        data: CertificateResponse;
    }>(`/v1/foundations/${encodeURIComponent(foundationId)}/certificates`, oazapfts.json({
        ...opts,
        method: "POST",
        body,
        headers: oazapfts.mergeHeaders(opts?.headers, {
            Authorization: authorization,
            "x-account-id": xAccountId
        })
    })));
}
export function getFoundation({ authorization, xAccountId, foundationId }: {
    authorization: string;
    xAccountId?: string;
    foundationId: string;
}, opts?: Oazapfts.RequestOpts) {
    return oazapfts.ok(oazapfts.fetchJson<{
        status: 202;
        data: FoundationResponse;
    }>(`/v1/foundations/${encodeURIComponent(foundationId)}`, {
        ...opts,
        headers: oazapfts.mergeHeaders(opts?.headers, {
            Authorization: authorization,
            "x-account-id": xAccountId
        })
    }));
}
export function getVpn({ authorization, xAccountId, foundationId, vpnId }: {
    authorization: string;
    xAccountId?: string;
    foundationId: string;
    vpnId: string;
}, opts?: Oazapfts.RequestOpts) {
    return oazapfts.ok(oazapfts.fetchJson<{
        status: 200;
        data: VpnResponse;
    }>(`/v1/foundations/${encodeURIComponent(foundationId)}/vpns/${encodeURIComponent(vpnId)}`, {
        ...opts,
        headers: oazapfts.mergeHeaders(opts?.headers, {
            Authorization: authorization,
            "x-account-id": xAccountId
        })
    }));
}
export function deleteVpn({ authorization, xAccountId, foundationId, vpnId }: {
    authorization: string;
    xAccountId?: string;
    foundationId: string;
    vpnId: string;
}, opts?: Oazapfts.RequestOpts) {
    return oazapfts.ok(oazapfts.fetchText(`/v1/foundations/${encodeURIComponent(foundationId)}/vpns/${encodeURIComponent(vpnId)}`, {
        ...opts,
        method: "DELETE",
        headers: oazapfts.mergeHeaders(opts?.headers, {
            Authorization: authorization,
            "x-account-id": xAccountId
        })
    }));
}
export function getVpnConfiguration({ authorization, xAccountId, foundationId, vpnId }: {
    authorization: string;
    xAccountId?: string;
    foundationId: string;
    vpnId: string;
}, opts?: Oazapfts.RequestOpts) {
    return oazapfts.ok(oazapfts.fetchJson<{
        status: 200;
        data: VpnConfigurationResponse;
    }>(`/v1/foundations/${encodeURIComponent(foundationId)}/vpns/${encodeURIComponent(vpnId)}/configuration`, {
        ...opts,
        headers: oazapfts.mergeHeaders(opts?.headers, {
            Authorization: authorization,
            "x-account-id": xAccountId
        })
    }));
}
export function getTenant({ authorization, xAccountId, foundationId, tenantId }: {
    authorization: string;
    xAccountId?: string;
    foundationId: string;
    tenantId: string;
}, opts?: Oazapfts.RequestOpts) {
    return oazapfts.ok(oazapfts.fetchJson<{
        status: 200;
        data: TenantResponse;
    }>(`/v1/foundations/${encodeURIComponent(foundationId)}/tenants/${encodeURIComponent(tenantId)}`, {
        ...opts,
        headers: oazapfts.mergeHeaders(opts?.headers, {
            Authorization: authorization,
            "x-account-id": xAccountId
        })
    }));
}
export function getRuntime({ authorization, xAccountId, foundationId, runtimeId }: {
    authorization: string;
    xAccountId?: string;
    foundationId: string;
    runtimeId: string;
}, opts?: Oazapfts.RequestOpts) {
    return oazapfts.ok(oazapfts.fetchJson<{
        status: 200;
        data: RuntimeResponse;
    }>(`/v1/foundations/${encodeURIComponent(foundationId)}/runtimes/${encodeURIComponent(runtimeId)}`, {
        ...opts,
        headers: oazapfts.mergeHeaders(opts?.headers, {
            Authorization: authorization,
            "x-account-id": xAccountId
        })
    }));
}
export function getProject({ authorization, xAccountId, foundationId, projectId }: {
    authorization: string;
    xAccountId?: string;
    foundationId: string;
    projectId: string;
}, opts?: Oazapfts.RequestOpts) {
    return oazapfts.ok(oazapfts.fetchJson<{
        status: 200;
        data: ProjectResponse;
    }>(`/v1/foundations/${encodeURIComponent(foundationId)}/projects/${encodeURIComponent(projectId)}`, {
        ...opts,
        headers: oazapfts.mergeHeaders(opts?.headers, {
            Authorization: authorization,
            "x-account-id": xAccountId
        })
    }));
}
export function deleteProject({ authorization, xAccountId, foundationId, projectId }: {
    authorization: string;
    xAccountId?: string;
    foundationId: string;
    projectId: string;
}, opts?: Oazapfts.RequestOpts) {
    return oazapfts.ok(oazapfts.fetchText(`/v1/foundations/${encodeURIComponent(foundationId)}/projects/${encodeURIComponent(projectId)}`, {
        ...opts,
        method: "DELETE",
        headers: oazapfts.mergeHeaders(opts?.headers, {
            Authorization: authorization,
            "x-account-id": xAccountId
        })
    }));
}
export function getNetwork({ authorization, xAccountId, foundationId, networkId }: {
    authorization: string;
    xAccountId?: string;
    foundationId: string;
    networkId: string;
}, opts?: Oazapfts.RequestOpts) {
    return oazapfts.ok(oazapfts.fetchJson<{
        status: 200;
        data: NetworkResponse;
    }>(`/v1/foundations/${encodeURIComponent(foundationId)}/networks/${encodeURIComponent(networkId)}`, {
        ...opts,
        headers: oazapfts.mergeHeaders(opts?.headers, {
            Authorization: authorization,
            "x-account-id": xAccountId
        })
    }));
}
export function deleteNetwork({ authorization, xAccountId, foundationId, networkId }: {
    authorization: string;
    xAccountId?: string;
    foundationId: string;
    networkId: string;
}, opts?: Oazapfts.RequestOpts) {
    return oazapfts.ok(oazapfts.fetchText(`/v1/foundations/${encodeURIComponent(foundationId)}/networks/${encodeURIComponent(networkId)}`, {
        ...opts,
        method: "DELETE",
        headers: oazapfts.mergeHeaders(opts?.headers, {
            Authorization: authorization,
            "x-account-id": xAccountId
        })
    }));
}
export function getDnsZone({ authorization, xAccountId, foundationId, dnsZoneId }: {
    authorization: string;
    xAccountId?: string;
    foundationId: string;
    dnsZoneId: string;
}, opts?: Oazapfts.RequestOpts) {
    return oazapfts.ok(oazapfts.fetchJson<{
        status: 200;
        data: DnsZoneResponse;
    }>(`/v1/foundations/${encodeURIComponent(foundationId)}/dns-zones/${encodeURIComponent(dnsZoneId)}`, {
        ...opts,
        headers: oazapfts.mergeHeaders(opts?.headers, {
            Authorization: authorization,
            "x-account-id": xAccountId
        })
    }));
}
export function deleteDnsZone({ authorization, xAccountId, foundationId, dnsZoneId }: {
    authorization: string;
    xAccountId?: string;
    foundationId: string;
    dnsZoneId: string;
}, opts?: Oazapfts.RequestOpts) {
    return oazapfts.ok(oazapfts.fetchText(`/v1/foundations/${encodeURIComponent(foundationId)}/dns-zones/${encodeURIComponent(dnsZoneId)}`, {
        ...opts,
        method: "DELETE",
        headers: oazapfts.mergeHeaders(opts?.headers, {
            Authorization: authorization,
            "x-account-id": xAccountId
        })
    }));
}
export function getDnsRecord({ authorization, xAccountId, foundationId, dnsRecordId }: {
    authorization: string;
    xAccountId?: string;
    foundationId: string;
    dnsRecordId: string;
}, opts?: Oazapfts.RequestOpts) {
    return oazapfts.ok(oazapfts.fetchJson<{
        status: 200;
        data: DnsRecordResponse;
    }>(`/v1/foundations/${encodeURIComponent(foundationId)}/dns-records/${encodeURIComponent(dnsRecordId)}`, {
        ...opts,
        headers: oazapfts.mergeHeaders(opts?.headers, {
            Authorization: authorization,
            "x-account-id": xAccountId
        })
    }));
}
export function deleteDnsRecord({ authorization, xAccountId, foundationId, dnsRecordId }: {
    authorization: string;
    xAccountId?: string;
    foundationId: string;
    dnsRecordId: string;
}, opts?: Oazapfts.RequestOpts) {
    return oazapfts.ok(oazapfts.fetchText(`/v1/foundations/${encodeURIComponent(foundationId)}/dns-records/${encodeURIComponent(dnsRecordId)}`, {
        ...opts,
        method: "DELETE",
        headers: oazapfts.mergeHeaders(opts?.headers, {
            Authorization: authorization,
            "x-account-id": xAccountId
        })
    }));
}
export function getCidr({ authorization, xAccountId, foundationId, cidrId }: {
    authorization: string;
    xAccountId?: string;
    foundationId: string;
    cidrId: string;
}, opts?: Oazapfts.RequestOpts) {
    return oazapfts.ok(oazapfts.fetchJson<{
        status: 200;
        data: CidrResponse;
    }>(`/v1/foundations/${encodeURIComponent(foundationId)}/cidrs/${encodeURIComponent(cidrId)}`, {
        ...opts,
        headers: oazapfts.mergeHeaders(opts?.headers, {
            Authorization: authorization,
            "x-account-id": xAccountId
        })
    }));
}
export function deleteCidr({ authorization, xAccountId, foundationId, cidrId }: {
    authorization: string;
    xAccountId?: string;
    foundationId: string;
    cidrId: string;
}, opts?: Oazapfts.RequestOpts) {
    return oazapfts.ok(oazapfts.fetchText(`/v1/foundations/${encodeURIComponent(foundationId)}/cidrs/${encodeURIComponent(cidrId)}`, {
        ...opts,
        method: "DELETE",
        headers: oazapfts.mergeHeaders(opts?.headers, {
            Authorization: authorization,
            "x-account-id": xAccountId
        })
    }));
}
export function getCertificate({ authorization, xAccountId, foundationId, certificateId }: {
    authorization: string;
    xAccountId?: string;
    foundationId: string;
    certificateId: string;
}, opts?: Oazapfts.RequestOpts) {
    return oazapfts.ok(oazapfts.fetchJson<{
        status: 200;
        data: CertificateResponse;
    }>(`/v1/foundations/${encodeURIComponent(foundationId)}/certificates/${encodeURIComponent(certificateId)}`, {
        ...opts,
        headers: oazapfts.mergeHeaders(opts?.headers, {
            Authorization: authorization,
            "x-account-id": xAccountId
        })
    }));
}
export function deleteCertificate({ authorization, xAccountId, foundationId, certificateId }: {
    authorization: string;
    xAccountId?: string;
    foundationId: string;
    certificateId: string;
}, opts?: Oazapfts.RequestOpts) {
    return oazapfts.ok(oazapfts.fetchText(`/v1/foundations/${encodeURIComponent(foundationId)}/certificates/${encodeURIComponent(certificateId)}`, {
        ...opts,
        method: "DELETE",
        headers: oazapfts.mergeHeaders(opts?.headers, {
            Authorization: authorization,
            "x-account-id": xAccountId
        })
    }));
}
export function getProvisionAvailability({ cloudProvider }: {
    cloudProvider: string;
}, opts?: Oazapfts.RequestOpts) {
    return oazapfts.ok(oazapfts.fetchJson<{
        status: 200;
        data: {
            [key: string]: boolean;
        };
    }>(`/v1/foundations/provision-avaliability${QS.query(QS.explode({
        cloudProvider
    }))}`, {
        ...opts
    }));
}
export function providers(opts?: Oazapfts.RequestOpts) {
    return oazapfts.ok(oazapfts.fetchJson<{
        status: 200;
        data: {
            [key: string]: any;
        }[];
    }>("/v1/foundations/providers", {
        ...opts
    }));
}
export function deleteFolder({ authorization, xAccountId, foundationId, folderId }: {
    authorization: string;
    xAccountId?: string;
    foundationId: string;
    folderId: string;
}, opts?: Oazapfts.RequestOpts) {
    return oazapfts.ok(oazapfts.fetchText(`/v1/foundations/${encodeURIComponent(foundationId)}/folders/${encodeURIComponent(folderId)}`, {
        ...opts,
        method: "DELETE",
        headers: oazapfts.mergeHeaders(opts?.headers, {
            Authorization: authorization,
            "x-account-id": xAccountId
        })
    }));
}
