// @stack-spot/portal-network/src/api/codeShift.ts
/**
 * Code shift API
 * 0.1.0
 * DO NOT MODIFY - This file has been generated using oazapfts.
 * See https://www.npmjs.com/package/oazapfts
 */
import * as Oazapfts from "@oazapfts/runtime";
import * as QS from "@oazapfts/runtime/query";
export const defaults: Oazapfts.Defaults<Oazapfts.CustomHeaders> = {
    headers: {},
    baseUrl: "/",
};
const oazapfts = Oazapfts.runtime(defaults);
export const servers = {};
export type CodeShiftRole = "list-applications" | "create-application" | "list-repository" | "create-repository" | "delete-repository" | "dispatch-module" | "create-module" | "account-settings" | "list-integration" | "create-integration" | "delete-integration" | "list-program-group" | "create-program-group" | "delete-program-group" | "view-analytics" | "account-discovery-repos" | "account-validate-repos";
export type RolesResponse = {
    result: boolean;
};
export type ValidationError = {
    loc: (string | number)[];
    msg: string;
    "type": string;
};
export type HttpValidationError = {
    detail?: ValidationError[];
};
export type ExceptionType = "CODE_SHIFT_API_0000_UNEXPECTED_ERROR" | "CODE_SHIFT_API_0001_INVALID_VALUE" | "CODE_SHIFT_API_0002_NOT_IMPLEMENTED" | "CODE_SHIFT_API_0003_NOT_FOUND" | "CODE_SHIFT_API_0004_VALIDATION_ERROR" | "CODE_SHIFT_API_0005_CONFLICT" | "CODE_SHIFT_API_0006_INVALID_REPO_URL" | "CODE_SHIFT_API_0007_FORBIDDEN" | "CODE_SHIFT_API_0008_DEFAULT_BRANCH" | "CODE_SHIFT_API_0009_INVALID_SPREADSHEET" | "CODE_SHIFT_API_0010_PR_LINK_NOT_EXISTS" | "CODE_SHIFT_API_0011_EMPTY_SPREADSHEET" | "CODE_SHIFT_API_0012_REPOSITORY_NOT_EXISTS_IN_SCM" | "CODE_SHIFT_API_0013_SCM_BRANCH_NOT_EXISTS_EXCEPTION" | "CODE_SHIFT_API_0014_ACCOUNT_NOT_FOUND_EXCEPTION" | "CODE_SHIFT_API_0015_INVALID_WORKFLOW_URL_EXCEPTION" | "CODE_SHIFT_API_0016_INVALID_ACTION_YAML_EXCEPTION" | "CODE_SHIFT_API_0017_FAIL_READING_ACTION_INPUTS_EXCEPTION" | "CODE_SHIFT_API_0018_MODULE_DISPATCHER_GENERIC_EXCEPTION" | "CODE_SHIFT_API_0019_GITHUB_DISPATCH_WORKFLOW_UNAUTHORIZED_EXCEPTION" | "CODE_SHIFT_API_0020_GITHUB_SAML_EXCEPTION" | "CODE_SHIFT_API_0021_GITHUB_AUTHORIZATION_ENCODING_EXCEPTION" | "CODE_SHIFT_API_0022_GITHUB_DISPATCH_WORKFLOW_NOT_FOUND_EXCEPTION" | "CODE_SHIFT_API_0023_GENERIC_FAILURE_SEARCHING_REPOSITORY_EXCEPTION" | "CODE_SHIFT_API_0024_COMPONENTS_MAX_SIZE_ALLOWED_EXCEPTION" | "CODE_SHIFT_API_0025_INVALID_SEARCH_REPOSITORY_STATUS_SAVE_EXCEPTION" | "CODE_SHIFT_API_0026_SEARCH_REPO_EMPTY_RESULT_EXCEPTION" | "CODE_SHIFT_API_0027_ACCOUNT_SCM_CONFIG_NOT_FOUND_EXCEPTION" | "CODE_SHIFT_API_0028_USER_SCM_CONFIG_NOT_FOUND_EXCEPTION" | "CODE_SHIFT_API_0029_USER_CUSTOMER_RATING_CREATION_FORBIDDEN" | "CODE_SHIFT_API_0030_BROKEN_STACKSPOT_WORKFLOW_SCM_SETUP_BY_INVALID_TRIGGER_EXCEPTION" | "CODE_SHIFT_API_0031_GITHUB_SECONDARY_INDEX" | "CODE_SHIFT_API_0032_GITHUB_MISSING_PAT_SCOPE" | "CODE_SHIFT_API_0033_FEATURE_NOT_IMPLEMENTED_TO_SCM_YET" | "CODE_SHIFT_API_0034_FAILURE_GETTING_READING_CREDENTIALS" | "CODE_SHIFT_API_0035_SCM_MODULE_DOCUMENTATION_NOT_FOUND_EXCEPTION" | "CODE_SHIFT_API_3000_WORKFLOW_API_DISPATCH_FAILURE" | "CODE_SHIFT_API_3001_WORKFLOW_API_DISPATCH_FORBIDDEN" | "CODE_SHIFT_API_3009_WORKFLOW_API_DISPATCH_PARSE_ERROR" | "CODE_SHIFT_API_3010_AWS_SECRET_MANAGER_GET_SECRET_FAILURE" | "CODE_SHIFT_API_3011_AWS_SECRET_MANAGER_PARSE_SECRET_ERROR" | "CODE_SHIFT_API_3100_SCM_GET_REPOSITORIES_FAILURE" | "CODE_SHIFT_API_3101_SCM_GET_REPOSITORIES_NOT_FOUND" | "CODE_SHIFT_API_3102_SCM_GET_REPOSITORIES_UNAUTHORIZED" | "CODE_SHIFT_API_3103_SCM_GET_REPOSITORIES_FORBIDDEN" | "CODE_SHIFT_API_3104_SCM_GET_REPOSITORIES_SAML_ERROR" | "CODE_SHIFT_API_3110_SCM_GET_PULL_REQUEST_FAILURE" | "CODE_SHIFT_API_3111_SCM_GET_PULL_REQUEST_NOT_FOUND" | "CODE_SHIFT_API_3112_SCM_GET_PULL_REQUEST_UNAUTHORIZED" | "CODE_SHIFT_API_3113_SCM_GET_PULL_REQUEST_FORBIDDEN" | "CODE_SHIFT_API_3114_SCM_GET_PULL_REQUEST_SAML_ERROR" | "CODE_SHIFT_API_3120_SCM_GET_BRANCHES_FAILURE" | "CODE_SHIFT_API_3121_SCM_GET_BRANCHES_NOT_FOUND" | "CODE_SHIFT_API_3122_SCM_GET_BRANCHES_UNAUTHORIZED" | "CODE_SHIFT_API_3123_SCM_GET_BRANCHES_FORBIDDEN" | "CODE_SHIFT_API_3124_SCM_GET_BRANCHES_SAML_ERROR" | "CODE_SHIFT_API_3130_SCM_GET_SCM_REPOSITORY_FETCH_FAILURE" | "CODE_SHIFT_API_3131_SCM_REPOSITORY_NOT_FOUND" | "CODE_SHIFT_API_3132_SCM_REPOSITORY_UNAUTHORIZED" | "CODE_SHIFT_API_3133_SCM_REPOSITORY_FORBIDDEN" | "CODE_SHIFT_API_3134_SCM_REPOSITORY_SAML_ERROR" | "CODE_SHIFT_API_3140_SCM_GET_BRANCH_FAILURE" | "CODE_SHIFT_API_3142_SCM_GET_BRANCH_NOT_FOUND" | "CODE_SHIFT_API_3143_SCM_GET_BRANCH_UNAUTHORIZED" | "CODE_SHIFT_API_3144_SCM_GET_BRANCH_FORBIDDEN" | "CODE_SHIFT_API_3145_SCM_GET_BRANCH_SAML_ERROR" | "CODE_SHIFT_API_3150_SCM_GET_REPOSITORY_FILE_FAILURE" | "CODE_SHIFT_API_3151_SCM_GET_REPOSITORY_FILE_NOT_FOUND" | "CODE_SHIFT_API_3152_SCM_GET_REPOSITORY_FILE_UNAUTHORIZED" | "CODE_SHIFT_API_3153_SCM_GET_REPOSITORY_FILE_FORBIDDEN" | "CODE_SHIFT_API_3154_SCM_GET_REPOSITORY_FILE_SAML_ERROR" | "CODE_SHIFT_API_3160_SCM_GET_REPOSITORY_FILE_FAILURE" | "CODE_SHIFT_API_3161_SCM_GET_REPOSITORY_FILE_NOT_FOUND" | "CODE_SHIFT_API_3162_SCM_GET_REPOSITORY_FILE_UNAUTHORIZED" | "CODE_SHIFT_API_3163_SCM_GET_REPOSITORY_FILE_FORBIDDEN" | "CODE_SHIFT_API_3164_SCM_GET_REPOSITORY_FILE_SAML_ERROR" | "CODE_SHIFT_API_3170_SCM_GET_OWNER_FAILURE" | "CODE_SHIFT_API_3172_SCM_GET_OWNER_NOT_FOUND" | "CODE_SHIFT_API_3173_SCM_GET_OWNER_UNAUTHORIZED" | "CODE_SHIFT_API_3174_SCM_GET_OWNER_FORBIDDEN" | "CODE_SHIFT_API_3175_SCM_GET_OWNER_SAML_ERROR" | "CODE_SHIFT_API_3500_CONTENT_API_GET_WORKFLOW_VERSION_FAILURE" | "CODE_SHIFT_API_3501_CONTENT_API_GET_WORKFLOW_VERSION_NOT_FOUND" | "CODE_SHIFT_API_3502_CONTENT_API_GET_WORKFLOW_VERSION_FORBIDDEN" | "CODE_SHIFT_API_3503_CONTENT_API_GET_WORKFLOW_VERSION_UNAUTHORIZED" | "CODE_SHIFT_API_3510_CONTENT_API_GET_WORKFLOW_FAILURE" | "CODE_SHIFT_API_3511_CONTENT_API_GET_WORKFLOW_NOT_FOUND" | "CODE_SHIFT_API_3512_CONTENT_API_GET_WORKFLOW_FORBIDDEN" | "CODE_SHIFT_API_3513_CONTENT_API_GET_WORKFLOW_UNAUTHORIZED" | "CODE_SHIFT_API_3520_CONTENT_API_GET_WORKFLOW_INPUTS_FAILURE" | "CODE_SHIFT_API_3521_CONTENT_API_GET_WORKFLOW_INPUTS_NOT_FOUND" | "CODE_SHIFT_API_3522_CONTENT_API_GET_WORKFLOW_INPUTS_FORBIDDEN" | "CODE_SHIFT_API_3523_CONTENT_API_GET_WORKFLOW_INPUTS_UNAUTHORIZED" | "CODE_SHIFT_API_3530_CONTENT_API_GET_WORKFLOW_DOCUMENTATION_FAILURE" | "CODE_SHIFT_API_3531_CONTENT_API_GET_WORKFLOW_DOCUMENTATION_NOT_FOUND" | "CODE_SHIFT_API_3532_CONTENT_API_GET_WORKFLOW_DOCUMENTATION_FORBIDDEN" | "CODE_SHIFT_API_3533_CONTENT_API_GET_WORKFLOW_DOCUMENTATION_UNAUTHORIZED" | "CODE_SHIFT_API_4010_GET_SCM_CREDENTIALS_FAILURE" | "CODE_SHIFT_API_4011_GET_SCM_CREDENTIALS_FORBIDDEN" | "CODE_SHIFT_API_4012_GET_SCM_CREDENTIALS_NOT_FOUND" | "CODE_SHIFT_API_4019_GET_SCM_CREDENTIALS_PARSE_ERROR" | "CODE_SHIFT_API_6000_IAM_GENERATE_TOKEN_FAILURE" | "CODE_SHIFT_API_6000_IAM_INTROSPECT_TOKEN_FAILURE";
export type BadRequestExceptionTypes = "CODE_SHIFT_API_1001_DECODE_JWT_ERROR" | "CODE_SHIFT_API_2001_BAD_REQUEST_REQUIRED_FIELD" | "CODE_SHIFT_API_2002_BAD_REQUEST_EMPTY_FIELD_NOT_ALLOWED" | "CODE_SHIFT_API_2003_BAD_REQUEST_TYPE_FIELD_NOT_ALLOWED" | "CODE_SHIFT_API_2004_BAD_REQUEST_VALUE_FIELD_NOT_ALLOWED" | "CODE_SHIFT_API_2005_BAD_REQUEST_VALUE_FIELD_MAX_LENGTH" | "CODE_SHIFT_API_2006_BAD_REQUEST_VALUE_FIELD_MIN_LENGTH" | "CODE_SHIFT_API_2007_BAD_REQUEST_VALUE_REGEX_DOESNT_MATCH" | "CODE_SHIFT_API_2008_BAD_REQUEST_VALUE_FIELD_NUMBER_NOT_LT" | "CODE_SHIFT_API_2100_CREATE_BODY_REPO_AND_REPO_CREATE_ACTION" | "CODE_SHIFT_API_2101_CREATE_BODY_NOT_REPO_AND_NOT_REPO_CREATE_ACTION" | "CODE_SHIFT_API_2102_BAD_REQUEST_DATE_MAX_INTERVAL" | "CODE_SHIFT_API_2103_BAD_REQUEST_INVALID_PAGE" | "CODE_SHIFT_API_2104_BAD_REQUEST_DUPLICATED_ALIAS" | "CODE_SHIFT_API_2105_BAD_REQUEST_INVALID_REPO_URL" | "CODE_SHIFT_API_2110_BAD_REQUEST_PUT_STEP_STATUS_COMPLETED_WITHOUT_CONCLUSION" | "CODE_SHIFT_API_2112_BAD_REQUEST_PUT_STEP_STATUS_NOT_COMPLETED_AND_CONCLUSION_DEFINED" | "CODE_SHIFT_API_2113_BAD_REQUEST_PUT_STEP_STATUS_COMPLETED_AND_COMPLETED_AT_NOT_DEFINED" | "CODE_SHIFT_API_2114_BAD_REQUEST_PUT_STEP_STATUS_IN_PROGRESS_AND_STARTED_AT_NOT_DEFINED" | "CODE_SHIFT_API_2115_BAD_REQUEST_PUT_STEP_COMPLETED_AT_DEFINED_AND_STATUS_NOT_COMPLETED" | "CODE_SHIFT_API_2116_BAD_REQUEST_PUT_STEP_STARTED_AT_DEFINED_AND_STATUS_PENDING" | "CODE_SHIFT_API_2117_BAD_REQUEST_PUT_STEP_LOG_DEFINED_AND_CONCLUSION_IS_NOT_FAILURE" | "CODE_SHIFT_API_2120_BAD_REQUEST_MULTIPLE_INTEGRATION" | "CODE_SHIFT_API_2121_BAD_REQUEST_EMPTY_INTEGRATION" | "CODE_SHIFT_API_2121_BAD_REQUEST_EMPTY_TARGET" | "CODE_SHIFT_API_2121_BAD_REQUEST_FILLED_TARGET" | "CODE_SHIFT_API_2999_BAD_REQUEST_UNMAPPED";
export type InvalidPayloadDetails = {
    code: BadRequestExceptionTypes | ExceptionType;
    field: string | null;
    message?: string | null;
};
export type SuggestedMessage = {
    ptBr: string | null;
    enUs: string | null;
};
export type InvalidPayloadDetailsRead = {
    code: BadRequestExceptionTypes | ExceptionType;
    field: string | null;
    message?: string | null;
    suggestedMessages: SuggestedMessage | null;
};
export type HttpErrorResponse = {
    status: string;
    code: ExceptionType;
    details: string;
    validationDetails?: InvalidPayloadDetails[] | null;
};
export type HttpErrorResponseRead = {
    status: string;
    code: ExceptionType;
    details: string;
    validationDetails?: InvalidPayloadDetailsRead[] | null;
    suggestedMessages: SuggestedMessage | null;
};
export type ModuleType = "repository" | "program_group";
export type InputRequest = {
    name: string;
    required: boolean;
    "default": string;
};
export type GitAction = {
    repositoryName: string;
    branch: string;
    directory: string;
};
export type ModuleVisibility = "Personal" | "Account";
export type UserRequest = {
    /** User ID */
    id: string;
    /** User username */
    username: string;
    /** User email */
    email: string;
};
export type ModulePermission = "read" | "write";
export type SharedUserRequest = {
    user: UserRequest;
    permission: ModulePermission;
};
export type CreateModuleRequest = {
    /** Module name */
    name: string;
    /** Module type */
    "type"?: ModuleType;
    /** Module description */
    description: string;
    /** Module workflow_name */
    workflowName?: string | null;
    /** Module version */
    version?: string | null;
    /** Module mode */
    modes: string[];
    /** Module tags */
    tags: string[];
    /** Inputs of module */
    inputs?: InputRequest[] | null;
    /** Module studio */
    studio?: string | null;
    /** Module studio */
    constants?: {
        [key: string]: any;
    } | null;
    gitAction?: GitAction | null;
    /** Module visibility */
    visibility?: ModuleVisibility;
    /** List of users to share the module with */
    sharedWith?: SharedUserRequest[] | null;
};
export type InputResponse = {
    name: string;
    required: boolean;
    "default": string;
};
export type GitAction2 = {
    repositoryName: string;
    branch: string;
    directory: string;
};
export type ModuleResponse = {
    /** Module ID */
    id: string;
    /** Module type */
    "type"?: ModuleType;
    /** Module name */
    name: string;
    /** Module description */
    description: string;
    /** Workflow name */
    workflowName: string | null;
    /** Studio */
    studio: string | null;
    /** Module modes */
    modes: string[];
    /** Module versions */
    versions: string[];
    /** Module tags */
    tags: string[];
    /** Module inputs */
    inputs: InputResponse[];
    /** Flag to internal developed modules */
    stackspotInternal: boolean;
    /** Module studio */
    constants?: {
        [key: string]: any;
    } | null;
    /** Module visibility */
    visibility: ModuleVisibility;
    gitAction?: GitAction2 | null;
    /** Flag to favor modules */
    isFavorite: boolean;
    /** Flag to edit modules */
    canEdit: boolean;
};
export type ListModuleResponse = {
    /** List of modules */
    items: ModuleResponse[];
    /** Number of modules */
    itemsCount: number;
    /** Last evaluated key */
    lastEvaluatedKey: string | null;
};
export type PutModuleRequest = {
    /** Module name */
    name?: string | null;
    /** Module type */
    "type"?: ModuleType | null;
    /** Module description */
    description?: string | null;
    /** Module workflow_name */
    workflowName?: string | null;
    /** Module version */
    version?: string | null;
    /** Module mode */
    modes?: string[] | null;
    /** Module tags */
    tags?: string[] | null;
    /** Inputs of module */
    inputs?: InputRequest[] | null;
    /** Module studio */
    studio?: string | null;
    /** Module Constants */
    constants?: {
        [key: string]: any;
    } | null;
    gitAction?: GitAction | null;
    /** Module visibility */
    visibility?: ModuleVisibility | null;
    /** List of users to share the module with */
    sharedWith?: SharedUserRequest[] | null;
};
export type SharedPermissionResponse = {
    /** User ID */
    userId: string;
    /** User email */
    email: string;
    /** User username */
    username: string;
    /** Permission type: 'read' (view/execute) or 'write' (edit/manage) */
    permission: ModulePermission;
};
export type UserResponse = {
    name: string;
};
export type GetModuleResponse = {
    /** Module ID */
    id: string;
    /** Module type */
    "type": ModuleType;
    /** Module name */
    name: string;
    /** Module description */
    description: string;
    /** Workflow name */
    workflowName: string | null;
    /** Studio */
    studio: string | null;
    /** Module modes */
    modes: string[];
    /** Module versions */
    version: string | null;
    /** Module studio */
    constants?: {
        [key: string]: any;
    } | null;
    gitAction?: GitAction2 | null;
    /** Module tags */
    tags?: string[] | null;
    /** Flag to internal developed modules */
    stackspotInternal: boolean;
    /** Module visibility */
    visibility: ModuleVisibility;
    /** List of users to share the module with */
    sharedWith?: SharedPermissionResponse[] | null;
    /** Flag to favor modules */
    isFavorite: boolean;
    /** Flag to edit modules */
    canEdit: boolean;
    /** Module usage */
    totalUsage: number;
    /** Module users */
    totalUsers: number;
    createdBy: UserResponse | null;
};
export type ExternalItemsModel = {
    source: string;
    value: string;
    label: string;
};
export type ConditionModel = {
    variable: string;
    operator: string;
    value: any;
};
export type OutputModel = {
    "from": string;
    to: string;
};
export type InputModel = {
    label: string;
    name: string | null;
    "type": string;
    required: boolean;
    pattern: string | null;
    help: string | null;
    "default": any | null;
    items: string[] | null;
    externalItems: ExternalItemsModel | null;
    itemsValues: string[] | null;
    condition: ConditionModel | null;
    connectionInterfaceType: string | null;
    outputs: OutputModel[] | null;
    input: InputModel | null;
    inputs: InputModel[] | null;
    scope: string | null;
};
export type ComputedInputModel = {
    name: string;
    expression: string;
};
export type GetModuleInputsResponse = {
    inputs: InputModel[];
    computedInputs: ComputedInputModel[] | null;
};
export type GetModuleDocsResponse = {
    content: string;
};
export type Mode = "scan" | "fix";
export type ModuleDispatchRequest = {
    /** Module mode */
    mode: Mode;
    /** Target branch */
    targetBranch?: string | null;
    /** Application ID */
    applicationId?: string | null;
    /** Program Group ID */
    programGroupId?: string | null;
    /** Module ID */
    moduleId: string;
    inputs?: {
        [key: string]: any;
    } | null;
};
export type DispatchModuleResponse = {
    batch_id: string;
};
export type DependsOnRequest = {
    sourceRef: string;
    targetRef: string;
    description: string;
};
export type IssuesRequest = {
    ref?: string | null;
    line?: number | null;
    description: string;
    severity?: string | null;
    recommendation?: string | null;
};
export type TargetFilesRequest = {
    fileName: string;
    filePath?: string | null;
    details?: string | null;
    dependsOn?: DependsOnRequest[] | null;
    issues?: IssuesRequest[] | null;
    totalTokenSent?: number | null;
    totalTokenReceived?: number | null;
    modified?: boolean | null;
};
export type LanguageInfoRequest = {
    bytes: number;
    absolutePercentage: number;
    relativePercentage: number;
};
export type PutReportRequest = {
    filesCount?: number;
    issuesCount: number;
    totalFilesCount?: number | null;
    analyzedFilesCount?: number | null;
    changedFilesCount?: number | null;
    pullRequestLink?: string | null;
    commitHash?: string | null;
    targetFiles?: TargetFilesRequest[];
    moduleVersion?: string | null;
    customOutputs?: {
        [key: string]: any;
    } | null;
    totalTokenSent?: number | null;
    totalTokenReceived?: number | null;
    languageBreakdown?: {
        [key: string]: LanguageInfoRequest;
    } | null;
};
export type ModuleResponse2 = {
    id: string;
    name: string;
};
export type ApplicationResponse = {
    id: string;
    name: string;
    url: string;
};
export type ExecutionResponse = {
    id: string;
    startedAt: string | null;
    completedAt: string | null;
    status: string | null;
    conclusion: string | null;
};
export type UserResponse2 = {
    id: string;
    name: string;
    email: string;
    ip?: string | null;
};
export type ProgramGroupResponse = {
    id: string;
    name: string;
};
export type DependsOnResponse = {
    sourceRef: string;
    targetRef: string;
    description: string;
};
export type IssuesResponse = {
    ref?: string | null;
    line?: number | null;
    description: string;
    lineLink?: string | null;
    severity?: string | null;
    recommendation?: string | null;
};
export type TargetFilesResponse = {
    fileName: string;
    filePath: string;
    details?: string | null;
    dependsOn?: DependsOnResponse[] | null;
    issues?: IssuesResponse[] | null;
    totalTokenSent?: number | null;
    totalTokenReceived?: number | null;
    modified?: boolean | null;
};
export type ReportStatus = "waiting_dispatch" | "waiting_runner" | "dispatch_failure" | "in_progress" | "suspended" | "error" | "success" | "cancelled";
export type PullRequestResponse = {
    state: string | null;
    merged: boolean | null;
    mergedBy: string | null;
    mergedAt: string | null;
    createdAt: string | null;
    updatedAt: string | null;
    lastSyncAt: string | null;
};
export type CustomerRatingResponse = {
    id: string;
    report_id: string;
    score: number;
    commentary: string | null;
    created_by: UserResponse2;
    created_at: string;
};
export type LanguageResponse = {
    name: string;
    extensions: string[];
    bytes: number;
    relativePercentage: number;
    absolutePercentage: number;
};
export type GetReportResponse = {
    id: string;
    "module": ModuleResponse2 | null;
    application: ApplicationResponse | null;
    execution: ExecutionResponse | null;
    mode: string;
    sourceBranch: string | null;
    targetBranch?: string | null;
    createdBy: UserResponse2;
    createdAt: string;
    updatedAt: string;
    filesCount: number | null;
    totalFilesCount: number | null;
    changedFilesCount: number | null;
    analyzedFilesCount: number | null;
    issuesFilesCount: number | null;
    issuesCount: number | null;
    commitHash: string | null;
    pullRequestLink?: string | null;
    programGroup: ProgramGroupResponse | null;
    moduleVersion?: string | null;
    customOutputs?: {
        [key: string]: any;
    } | null;
    totalTokenSent?: number | null;
    totalTokenReceived?: number | null;
    targetFiles: TargetFilesResponse[];
    errorLog?: string | null;
    exceptions: HttpErrorResponse[];
    status: ReportStatus | null;
    scmPullRequestData: PullRequestResponse | null;
    inputs?: {
        [key: string]: any;
    } | null;
    customer_rating?: CustomerRatingResponse | null;
    languages?: LanguageResponse[] | null;
};
export type GetReportResponseRead = {
    id: string;
    "module": ModuleResponse2 | null;
    application: ApplicationResponse | null;
    execution: ExecutionResponse | null;
    mode: string;
    sourceBranch: string | null;
    targetBranch?: string | null;
    createdBy: UserResponse2;
    createdAt: string;
    updatedAt: string;
    filesCount: number | null;
    totalFilesCount: number | null;
    changedFilesCount: number | null;
    analyzedFilesCount: number | null;
    issuesFilesCount: number | null;
    issuesCount: number | null;
    commitHash: string | null;
    pullRequestLink?: string | null;
    programGroup: ProgramGroupResponse | null;
    moduleVersion?: string | null;
    customOutputs?: {
        [key: string]: any;
    } | null;
    totalTokenSent?: number | null;
    totalTokenReceived?: number | null;
    targetFiles: TargetFilesResponse[];
    errorLog?: string | null;
    exceptions: HttpErrorResponseRead[];
    status: ReportStatus | null;
    scmPullRequestData: PullRequestResponse | null;
    inputs?: {
        [key: string]: any;
    } | null;
    customer_rating?: CustomerRatingResponse | null;
    languages?: LanguageResponse[] | null;
};
export type GetPullRequest = {
    content: string;
};
export type PutCustomerRatingReportRequest = {
    score: number;
    commentary?: string | null;
};
export type SearchRepoScmResponse = {
    id: string;
};
export type SearchRepositoryStatus = "pending" | "success" | "failure";
export type GetStatusSearchRepoScmResponse = {
    status: SearchRepositoryStatus;
    errorLog?: string | null;
};
export type ModuleResponse3 = {
    id: string;
    name: string;
};
export type UserResponse3 = {
    id: string;
    name: string;
    email: string;
    ip?: string | null;
};
export type ReportResponse = {
    id: string;
    "module": ModuleResponse3;
    execution: ExecutionResponse | null;
    mode: string;
    sourceBranch: string;
    targetBranch?: string | null;
    createdBy: UserResponse3;
    createdAt: string;
    updatedAt: string;
    filesCount: number | null;
    issuesCount: number | null;
    pullRequestLink?: string | null;
    errorLog?: string | null;
    exceptions: HttpErrorResponse[];
    status: ReportStatus | null;
};
export type ReportResponseRead = {
    id: string;
    "module": ModuleResponse3;
    execution: ExecutionResponse | null;
    mode: string;
    sourceBranch: string;
    targetBranch?: string | null;
    createdBy: UserResponse3;
    createdAt: string;
    updatedAt: string;
    filesCount: number | null;
    issuesCount: number | null;
    pullRequestLink?: string | null;
    errorLog?: string | null;
    exceptions: HttpErrorResponseRead[];
    status: ReportStatus | null;
};
export type ListRepositoryReportResponse = {
    /** List of reports */
    items: ReportResponse[];
    /** Number of reports */
    itemsCount: number;
    /** Last evaluated key */
    lastEvaluatedKey: string | null;
};
export type ListRepositoryReportResponseRead = {
    /** List of reports */
    items: ReportResponseRead[];
    /** Number of reports */
    itemsCount: number;
    /** Last evaluated key */
    lastEvaluatedKey: string | null;
};
export type BodyCreateReposBatchServiceV1ReposBatchPost = {
    file: Blob;
};
export type CreateRepositoryRequest = {
    url: string;
    defaultBranch: string;
    tags?: string[] | null;
};
export type UserResponse4 = {
    id: string;
    name: string;
    email: string;
    ip?: string | null;
};
export type ReportResponse2 = {
    id: string;
    execution: ExecutionResponse | null;
    mode: string;
    moduleId: string;
    sourceBranch: string;
    targetBranch?: string | null;
    createdBy: UserResponse4 | null;
    createdAt: string;
    updatedAt: string;
    filesCount: number | null;
    issuesCount: number | null;
    pullRequestLink?: string | null;
    errorLog?: string | null;
    exceptions: HttpErrorResponse[];
    status: ReportStatus | null;
};
export type ReportResponseRead2 = {
    id: string;
    execution: ExecutionResponse | null;
    mode: string;
    moduleId: string;
    sourceBranch: string;
    targetBranch?: string | null;
    createdBy: UserResponse4 | null;
    createdAt: string;
    updatedAt: string;
    filesCount: number | null;
    issuesCount: number | null;
    pullRequestLink?: string | null;
    errorLog?: string | null;
    exceptions: HttpErrorResponseRead[];
    status: ReportStatus | null;
};
export type RepositoryResponse = {
    /** Repository ID */
    id: string;
    /** Repository creation date */
    createdAt?: string | null;
    /** Repository creation user */
    createdBy?: string | null;
    lastModuleReport: ReportResponse2 | null;
    tags: string[];
    url: string;
    defaultBranch: string;
    languages?: LanguageResponse[] | null;
};
export type RepositoryResponseRead = {
    /** Repository ID */
    id: string;
    /** Repository creation date */
    createdAt?: string | null;
    /** Repository creation user */
    createdBy?: string | null;
    lastModuleReport: ReportResponseRead2 | null;
    tags: string[];
    url: string;
    defaultBranch: string;
    languages?: LanguageResponse[] | null;
};
export type ListRepositoryResponse = {
    /** List of applications */
    items: RepositoryResponse[];
    /** Number of applications */
    itemsCount: number;
    /** Last evaluated key */
    lastEvaluatedKey?: string | null;
};
export type ListRepositoryResponseRead = {
    /** List of applications */
    items: RepositoryResponseRead[];
    /** Number of applications */
    itemsCount: number;
    /** Last evaluated key */
    lastEvaluatedKey?: string | null;
};
export type ListBranchesResponse = {
    /** List of all branches */
    branches: string[];
};
export type PutRepositoryRequest = {
    url?: string | null;
    defaultBranch?: string | null;
    tags?: string[] | null;
};
export type ValidateScmUrlRequest = {
    url?: string | null;
};
export type ValidateScmUrlResponse = {
    url: string;
    name: string | null;
    id: string | null;
    defaultBranch: string | null;
    lastCommitDate: string | null;
    archived: boolean | null;
    visibility: string | null;
};
export type GetAccountSettingsRequest = {
    maxSimultaneousExecution: number | null;
    createRepositoryValidateBranch: boolean | null;
    dispatchModuleValidateBranch: boolean | null;
    createRepositoryUrlValidate?: boolean | null;
    dispatchRepositoryUrlValidate?: boolean | null;
    pullRequestSync: boolean | null;
    pullRequestSyncMinHours: number | null;
    pullRequestSyncMaxHours: number | null;
    pullRequestSyncReprocessTime: number | null;
};
export type CreateAccountSettingsRequest = {
    maxSimultaneousExecution?: number | null;
    createRepositoryUrlValidate?: boolean | null;
    dispatchRepositoryUrlValidate?: boolean | null;
    pullRequestSync?: boolean | null;
    createRepositoryValidateBranch?: boolean | null;
    dispatchModuleValidateBranch?: boolean | null;
    pullRequestSyncMinHours?: number | null;
    pullRequestSyncMaxHours?: number | null;
    pullRequestSyncReprocessTime?: number | null;
};
export type CreateIntegrationRequest = {
    name: string;
    description: string;
    value: {
        [key: string]: any;
    };
};
export type CreateIntegrationResponse = {
    id: string;
};
export type IntegrationResponse = {
    id: string;
    name: string;
    description?: string | null;
    value: {
        [key: string]: any;
    };
};
export type ListIntegrationResponse = {
    items: IntegrationResponse[];
    itemsCount: number;
};
export type PutIntegrationRequest = {
    /** Integration name */
    name?: string | null;
    /** Integration description */
    description?: string | null;
    /** Integration configuration or data */
    value?: {
        [key: string]: any;
    } | null;
};
export type BodyCreateProgramGroupServiceV1ProgramGroupsPost = {
    file?: Blob | null;
};
export type CreateProgramGroupResponse = {
    id: string;
};
export type UserResponse5 = {
    id: string;
    name: string;
    email: string;
    ip?: string | null;
};
export type ReportResponse3 = {
    id: string;
    execution: ExecutionResponse | null;
    mode: string;
    moduleId: string;
    createdBy: UserResponse5 | null;
    createdAt: string;
    updatedAt: string;
    filesCount: number | null;
    issuesCount: number | null;
    pullRequestLink?: string | null;
    errorLog?: string | null;
    exceptions: HttpErrorResponse[];
    status: ReportStatus | null;
};
export type ReportResponseRead3 = {
    id: string;
    execution: ExecutionResponse | null;
    mode: string;
    moduleId: string;
    createdBy: UserResponse5 | null;
    createdAt: string;
    updatedAt: string;
    filesCount: number | null;
    issuesCount: number | null;
    pullRequestLink?: string | null;
    errorLog?: string | null;
    exceptions: HttpErrorResponseRead[];
    status: ReportStatus | null;
};
export type ProgramGroupResponse2 = {
    id: string;
    name: string;
    integrationId: string | null;
    tags: string[];
    components: any[];
    accountId: string;
    createdBy: string | null;
    createdAt: string;
    updatedAt: string;
    lastModuleReport: ReportResponse3 | null;
};
export type ProgramGroupResponseRead = {
    id: string;
    name: string;
    integrationId: string | null;
    tags: string[];
    components: any[];
    accountId: string;
    createdBy: string | null;
    createdAt: string;
    updatedAt: string;
    lastModuleReport: ReportResponseRead3 | null;
};
export type ListProgramGroupResponse = {
    items: ProgramGroupResponse2[];
    itemsCount: number;
};
export type ListProgramGroupResponseRead = {
    items: ProgramGroupResponseRead[];
    itemsCount: number;
};
export type ProgramGroupResponse3 = {
    id: string;
    name: string;
    integrationId: string | null;
    tags: string[];
    components: any[];
    accountId: string;
    createdBy: string;
    createdAt: string;
    updatedAt: string;
};
export type BodyUpdateProgramGroupServiceV1ProgramGroupsProgramGroupIdPut = {
    file?: Blob | null;
};
export type PutProgramGroupComponentsRequest = {
    appl_name: string;
    "type": string;
    name: string;
};
export type ModuleResponse4 = {
    id: string;
    name: string;
};
export type UserResponse6 = {
    id: string;
    name: string;
    email: string;
    ip?: string | null;
};
export type ReportResponse4 = {
    id: string;
    "module": ModuleResponse4;
    execution: ExecutionResponse | null;
    mode: string;
    createdBy: UserResponse6;
    createdAt: string;
    updatedAt: string;
    filesCount: number | null;
    issuesCount: number | null;
    pullRequestLink?: string | null;
    errorLog?: string | null;
    exceptions: HttpErrorResponse[];
    status: ReportStatus | null;
};
export type ReportResponseRead4 = {
    id: string;
    "module": ModuleResponse4;
    execution: ExecutionResponse | null;
    mode: string;
    createdBy: UserResponse6;
    createdAt: string;
    updatedAt: string;
    filesCount: number | null;
    issuesCount: number | null;
    pullRequestLink?: string | null;
    errorLog?: string | null;
    exceptions: HttpErrorResponseRead[];
    status: ReportStatus | null;
};
export type ListProgramGroupReportResponse = {
    /** List of reports */
    items: ReportResponse4[];
    /** Number of reports */
    itemsCount: number;
};
export type ListProgramGroupReportResponseRead = {
    /** List of reports */
    items: ReportResponseRead4[];
    /** Number of reports */
    itemsCount: number;
};
export type TagResponse = {
    name: string;
};
export type ListTagResponse = {
    items: TagResponse[];
    itemsCount: number;
};
export type ReportsWithErrorsResponse = {
    user_id?: string | null;
    user_name?: string | null;
    user_email?: string | null;
    account_id?: string | null;
    account_slug?: string | null;
    error_log?: string | null;
    count_errors?: number | null;
};
export type GeneralReportSuccessAndErrorsResponses = {
    success_count?: number | null;
    error_count?: number | null;
    reports_with_errors?: ReportsWithErrorsResponse[];
};
export type RepositoryLastReportStatusResponse = {
    repositoryUrl: string;
    moduleName: string;
    mode: string;
    reportCreatedAt: string;
    reportCreatedBy: string;
    totalFiles: number | null;
    totalFilesAnalyzed: number | null;
    totalFilesIssues: number | null;
    totalFilesChanged: number | null;
    totalIssues: number | null;
    pullRequestLink: string | null;
    merged?: boolean | null;
    customerScore?: number | null;
    languages?: string | null;
};
export type AnalyticsRepositoryLastReportStatusResponse = {
    totalFiles?: number;
    totalFilesAnalyzed?: number;
    totalFilesIssues?: number;
    totalFilesChanged?: number;
    totalIssues?: number;
    items: RepositoryLastReportStatusResponse[];
    totalPullRequests?: number;
    totalMergedPullRequests?: number;
    totalItems?: number;
};
export type RepositoryUsageResponse = {
    repositoryUrl: string;
    totalScan?: number;
    totalScanSuccess?: number;
    totalScanFailure?: number;
    totalFix?: number;
    totalFixSuccess?: number;
    totalFixFailure?: number;
    totalPullRequests?: number;
    mergedPullRequests?: number;
    customerScore?: number | null;
};
export type AnalyticsRepositoryUsageResponse = {
    totalScan?: number;
    totalScanSuccess?: number;
    totalScanFailure?: number;
    totalFix?: number;
    totalFixSuccess?: number;
    totalFixFailure?: number;
    totalPullRequests?: number;
    mergedPullRequests?: number;
    items: RepositoryUsageResponse[];
    totalItems?: number;
};
export type UserUsageResponse = {
    name: string;
    email: string;
    totalScan?: number;
    totalScanSuccess?: number;
    totalScanFailure?: number;
    totalFix?: number;
    totalFixSuccess?: number;
    totalFixFailure?: number;
};
export type AnalyticsUserUsageResponse = {
    totalScan?: number;
    totalScanSuccess?: number;
    totalScanFailure?: number;
    totalFix?: number;
    totalFixSuccess?: number;
    totalFixFailure?: number;
    items: UserUsageResponse[];
    totalItems?: number;
};
export type ProgramGroupLastReportStatusResponse = {
    programName: string;
    moduleName: string;
    mode: string;
    reportCreatedAt: string;
    reportCreatedBy: string;
    totalFiles?: number;
    totalFilesAnalyzed?: number;
    totalFilesIssues?: number;
    totalFilesChanged?: number;
    totalIssues?: number;
    customerScore?: number | null;
};
export type AnalyticsProgramGroupsLastReportStatusResponse = {
    totalFiles?: number;
    totalFilesAnalyzed?: number;
    totalFilesIssues?: number;
    totalFilesChanged?: number;
    totalIssues?: number;
    items: ProgramGroupLastReportStatusResponse[];
    totalItems?: number;
};
export type ProgramGroupsUsageResponse = {
    name: string;
    totalScan?: number;
    totalScanSuccess?: number;
    totalScanFailure?: number;
    totalFix?: number;
    totalFixSuccess?: number;
    totalFixFailure?: number;
    customerScore?: number | null;
};
export type AnalyticsProgramGroupsUsageResponse = {
    totalScan?: number;
    totalScanSuccess?: number;
    totalScanFailure?: number;
    totalFix?: number;
    totalFixSuccess?: number;
    totalFixFailure?: number;
    items: ProgramGroupsUsageResponse[];
    totalItems?: number;
};
export type ProgramGroupsDetailsResponse = {
    programGroupName: string;
    programGroupTags?: string | null;
    integrationName?: string | null;
    programGroupCreatedByName?: string | null;
    programGroupCreatedByEmail?: string | null;
    moduleName: string;
    reportInputs?: string | null;
    mode: string;
    reportCreatedAt: string;
    reportCreatedByName?: string | null;
    reportCreatedByEmail?: string | null;
    reportOutputs?: string | null;
    executionStartedAt?: string | null;
    executionCompletedAt?: string | null;
    processingTimeSeconds?: number | null;
    totalFilesCount?: number;
    analyzedFilesCount?: number;
    issuesFilesCount?: number;
    changedFilesCount?: number;
    issuesCount?: number;
    status?: string;
    customerScore?: number | null;
};
export type AnalyticsProgramGroupsDetailsResponse = {
    programsCount?: number;
    totalFilesCount?: number;
    analyzedFilesCount?: number;
    issuesFilesCount?: number;
    changedFilesCount?: number;
    issuesCount?: number;
    processingTime95Percentile?: number | null;
    totalItems?: number;
    items?: ProgramGroupsDetailsResponse[];
};
export type RepositoryDetailedReportItemResponse = {
    analyzedFilesCount?: number;
    changedFilesCount?: number;
    errorLog?: string | null;
    executionCompletedAt?: string | null;
    executionConclusion?: string | null;
    executionStartedAt?: string | null;
    executionStatus?: string | null;
    inputs?: string;
    issuesCount?: number;
    issuesFilesCount?: number;
    mode?: string | null;
    moduleName?: string | null;
    moduleVersion?: string | null;
    processingTimeSeconds?: number | null;
    pullRequestLink?: string | null;
    merged?: boolean | null;
    reportCreatedAt?: string | null;
    reportCreatedByEmail?: string | null;
    reportCreatedByName?: string | null;
    repositoryCreatedByEmail?: string | null;
    repositoryCreatedByName?: string | null;
    repositoryTags?: string;
    repositoryUrl: string;
    sourceBranch?: string | null;
    targetBranch?: string | null;
    totalFilesCount?: number;
    status?: string | null;
    customerScore?: number | null;
    languages?: string | null;
};
export type AnalyticsRepositoryDetailedReportResponse = {
    reposCount?: number;
    totalFilesCount?: number;
    analyzedFilesCount?: number;
    issuesFilesCount?: number;
    changedFilesCount?: number;
    issuesCount?: number;
    processingTime95Percentile?: number | null;
    totalMergedPullRequests?: number;
    totalItems?: number;
    items?: RepositoryDetailedReportItemResponse[];
};
export type ProgramGroupsTargetDetailsResponse = {
    programGroupName: string;
    programGroupTags?: string[] | null;
    mode: string;
    moduleName: string;
    reportCreatedAt: string;
    executionStartedAt?: string | null;
    executionCompletedAt?: string | null;
    componentAppName: string;
    componentName: string;
    componentType: string;
    totalIssues?: number;
    changedFilesCount?: number;
    customOutputs?: {
        [key: string]: any;
    } | null;
    customerScore?: number | null;
    totalTokenSent: number | null;
    totalTokenReceived: number | null;
    modified: boolean | null;
};
export type AnalyticsProgramGroupsTargetDetailsResponse = {
    totalComponents?: number;
    totalIssues?: number;
    totalProgramGroups?: number;
    totalItems?: number;
    items?: ProgramGroupsTargetDetailsResponse[];
};
export type RepositoryTargetDetailsItemResponse = {
    repositoryUrl: string;
    sourceBranch: string | null;
    moduleName: string | null;
    targetName: string | null;
    moduleInputs: string | null;
    targetPath: string | null;
    targetDetails: string | null;
    issuesCount?: number | null;
    reportCreatedAt: string | null;
    reportId: string | null;
    customerScore?: number | null;
    totalTokenSent: number | null;
    totalTokenReceived: number | null;
    modified: boolean | null;
};
export type AnalyticsRepositoryTargetDetailsResponse = {
    totalFiles?: number;
    totalIssues?: number;
    totalItems?: number;
    items?: RepositoryTargetDetailsItemResponse[];
};
export type ModuleExecutionTimesDetailsResponse = {
    /** Name of the module */
    moduleName: string;
    /** Version of the module */
    moduleVersion: string;
    /** Average queue time in seconds */
    queueTimeAvg: number;
    /** 95th percentile queue time in seconds */
    queueTimeP95: number;
    /** Average execution time in seconds */
    executionTimeAvg: number;
    /** 95th percentile execution time in seconds */
    executionTimeP95: number;
    /** Average total time in seconds */
    totalTimeAvg: number;
    /** 95th percentile total time in seconds */
    totalTimeP95: number;
    /** Number of executions */
    executionCount: number | null;
};
export type AnalyticsModuleExecutionTimesResponse = {
    /** Total number of modules analyzed */
    modulesCount?: number;
    /** Overall average queue time */
    queueTimeAvg: number | null;
    /** Overall 95th percentile queue time */
    queueTimeP95: number | null;
    /** Overall average execution time */
    executionTimeAvg: number | null;
    /** Overall 95th percentile execution time */
    executionTimeP95: number | null;
    /** Overall average total time */
    totalTimeAvg: number | null;
    /** Overall 95th percentile total time */
    totalTimeP95: number | null;
    /** Detailed metrics per module */
    items?: ModuleExecutionTimesDetailsResponse[];
};
export type UserResponse7 = {
    id: string;
    name: string;
    email: string;
};
export type ListUserResponse = {
    items: UserResponse7[];
    itemsCount: number;
};
export type SearchReposInScmRequest = {
    repositoryName?: string | null;
    codeTerm?: string | null;
};
export type SearchReposInScmResponse = {
    search_id: string;
};
export type RepositoryPayload = {
    url: string;
    tags: string[];
};
export type ImportReposScmWithTagsRequest = {
    repositories: RepositoryPayload[];
};
export type SaveSearchReposRepositoryResponse = {
    id: string;
};
export type SaveSearchReposResponse = {
    search_id: string;
    repositories: SaveSearchReposRepositoryResponse[];
};
export type SearchRepoByIdResponse = {
    url: string;
    name: string | null;
    defaultBranch: string;
    lastCommitDate: string | null;
    archived: boolean | null;
    visibility?: ("public" | "private" | "internal") | null;
    alreadyCreated?: boolean;
};
export type ResponseSearchRepoByIdResponse = {
    total: number | null;
    status?: string;
    exceptions: HttpErrorResponse[];
    items?: SearchRepoByIdResponse[];
};
export type ResponseSearchRepoByIdResponseRead = {
    total: number | null;
    status?: string;
    exceptions: HttpErrorResponseRead[];
    items?: SearchRepoByIdResponse[];
};
export type BodyCreateReposBatchServiceV2ReposBatchPost = {
    file: Blob;
};
export type CreateRepoBatchResponse = {
    id: string;
};
export type ImportStatus = "COMPLETED" | "PENDING";
export type ValidateRepositoryStatus = "PENDING" | "SUCCESS" | "FAILURE" | "CONFLICT";
export type BatchRepositoryResponse = {
    repositoryId: string | null;
    repositoryUrl: string;
    repositoryBranch: string;
    repositoryTags: string[];
    status: ValidateRepositoryStatus;
    line: number;
    repositoryName: string | null;
    repositoryArchived: boolean | null;
    repositoryPushedAt: string | null;
    repositoryVisibility: string | null;
    exceptions: HttpErrorResponse[];
};
export type BatchRepositoryResponseRead = {
    repositoryId: string | null;
    repositoryUrl: string;
    repositoryBranch: string;
    repositoryTags: string[];
    status: ValidateRepositoryStatus;
    line: number;
    repositoryName: string | null;
    repositoryArchived: boolean | null;
    repositoryPushedAt: string | null;
    repositoryVisibility: string | null;
    exceptions: HttpErrorResponseRead[];
};
export type GetImportResultResponse = {
    status: ImportStatus;
    totalItems: number;
    items: BatchRepositoryResponse[];
    repositoriesIds: string[];
};
export type GetImportResultResponseRead = {
    status: ImportStatus;
    totalItems: number;
    items: BatchRepositoryResponseRead[];
    repositoriesIds: string[];
};
export type RepositoryResponse2 = {
    id: string;
    repositoryUrl: string | null;
    repositoryDefaultBranch: string | null;
    reportCreatedAt: string | null;
    reportIssuesCount: string | null;
    reportStatus: ReportStatus | null;
};
export type ListRepositoryResponse2 = {
    items: RepositoryResponse2[];
    itemsCount: number;
};
export type RepositoryRequest = {
    url: string;
    defaultBranch: string;
    tags?: string[] | null;
};
export type CreateRepositoryRequest2 = {
    repositories: RepositoryRequest[];
    raiseConflict?: boolean;
};
export type RepositoryResponse3 = {
    id: string;
};
export type CreateRepositoryResponse = {
    repositories: RepositoryResponse3[];
};
export type ModuleExecutionBatchStatus = "PENDING" | "IN_PROGRESS" | "COMPLETED";
export type ExecutionBatchResponse = {
    id: string;
    name: string | null;
    status: ModuleExecutionBatchStatus | null;
    created_at: string;
};
export type ListModuleExecutionBatchActivitiesResponse = {
    items: ExecutionBatchResponse[];
    itemsCount: number;
};
export type ModuleExecution = {
    repositoryUrl: string | null;
    programGroupName: string | null;
    reportId: string;
    executionId: string | null;
    status: ReportStatus | null;
    progress: number | null;
    executionCreatedAt: string | null;
    executionCompletedAt: string | null;
};
export type GetModuleExecutionBatchActivitiesById = {
    items: ModuleExecution[];
};
/**
 * Check Role Route
 */
export function checkRoleRouteV1RolesRoleGet({ role, authorization }: {
    role: CodeShiftRole;
    authorization: string;
}, opts?: Oazapfts.RequestOpts) {
    return oazapfts.ok(oazapfts.fetchJson<{
        status: 200;
        data: RolesResponse;
    } | {
        status: 422;
        data: HttpValidationError;
    }>(`/v1/roles/${encodeURIComponent(role)}`, {
        ...opts,
        headers: oazapfts.mergeHeaders(opts?.headers, {
            authorization
        })
    }));
}
/**
 * Module Favorite Service Add
 */
export function moduleFavoriteServiceAddV1ModulesModuleIdFavoritesPost({ moduleId, authorization }: {
    moduleId: string;
    authorization: string;
}, opts?: Oazapfts.RequestOpts) {
    return oazapfts.ok(oazapfts.fetchJson<{
        status: 200;
        data: any;
    } | {
        status: 400;
        data: HttpErrorResponseRead;
    } | {
        status: 401;
        data: HttpErrorResponseRead;
    } | {
        status: 404;
        data: HttpErrorResponseRead;
    } | {
        status: 422;
    } | {
        status: 500;
        data: HttpErrorResponseRead;
    } | {
        status: 503;
        data: HttpErrorResponseRead;
    }>(`/v1/modules/${encodeURIComponent(moduleId)}/favorites`, {
        ...opts,
        method: "POST",
        headers: oazapfts.mergeHeaders(opts?.headers, {
            authorization
        })
    }));
}
/**
 * Module Favorite Service Delete
 */
export function moduleFavoriteServiceDeleteV1ModulesModuleIdFavoritesDelete({ moduleId, authorization }: {
    moduleId: string;
    authorization: string;
}, opts?: Oazapfts.RequestOpts) {
    return oazapfts.ok(oazapfts.fetchJson<{
        status: 200;
        data: any;
    } | {
        status: 400;
        data: HttpErrorResponseRead;
    } | {
        status: 401;
        data: HttpErrorResponseRead;
    } | {
        status: 404;
        data: HttpErrorResponseRead;
    } | {
        status: 422;
    } | {
        status: 500;
        data: HttpErrorResponseRead;
    } | {
        status: 503;
        data: HttpErrorResponseRead;
    }>(`/v1/modules/${encodeURIComponent(moduleId)}/favorites`, {
        ...opts,
        method: "DELETE",
        headers: oazapfts.mergeHeaders(opts?.headers, {
            authorization
        })
    }));
}
/**
 * Create Module Service
 */
export function createModuleServiceV1ModulesPost({ authorization, xAccountSlug, createModuleRequest }: {
    authorization: string;
    xAccountSlug?: string;
    createModuleRequest: CreateModuleRequest;
}, opts?: Oazapfts.RequestOpts) {
    return oazapfts.ok(oazapfts.fetchJson<{
        status: 200;
        data: any;
    } | {
        status: 400;
        data: HttpErrorResponseRead;
    } | {
        status: 401;
        data: HttpErrorResponseRead;
    } | {
        status: 404;
        data: HttpErrorResponseRead;
    } | {
        status: 422;
    } | {
        status: 500;
        data: HttpErrorResponseRead;
    } | {
        status: 503;
        data: HttpErrorResponseRead;
    }>("/v1/modules", oazapfts.json({
        ...opts,
        method: "POST",
        body: createModuleRequest,
        headers: oazapfts.mergeHeaders(opts?.headers, {
            authorization,
            "x-account-slug": xAccountSlug
        })
    })));
}
/**
 * List Modules Service
 */
export function listModulesServiceV1ModulesGet({ pageSize, moduleType, page, lastEvaluatedKey, sharedWith, createdBy, visibility, name, tags, xAccountSlug, authorization }: {
    pageSize?: number;
    moduleType?: ModuleType | null;
    page?: number;
    lastEvaluatedKey?: string;
    sharedWith?: string | null;
    createdBy?: string | null;
    visibility?: ModuleVisibility | null;
    name?: string | null;
    tags?: string[] | null;
    xAccountSlug?: string;
    authorization: string;
}, opts?: Oazapfts.RequestOpts) {
    return oazapfts.ok(oazapfts.fetchJson<{
        status: 200;
        data: ListModuleResponse;
    } | {
        status: 400;
        data: HttpErrorResponseRead;
    } | {
        status: 401;
        data: HttpErrorResponseRead;
    } | {
        status: 404;
        data: HttpErrorResponseRead;
    } | {
        status: 422;
    } | {
        status: 500;
        data: HttpErrorResponseRead;
    } | {
        status: 503;
        data: HttpErrorResponseRead;
    }>(`/v1/modules${QS.query(QS.explode({
        pageSize,
        moduleType,
        page,
        lastEvaluatedKey,
        sharedWith,
        createdBy,
        visibility,
        name,
        tags
    }))}`, {
        ...opts,
        headers: oazapfts.mergeHeaders(opts?.headers, {
            "x-account-slug": xAccountSlug,
            authorization
        })
    }));
}
/**
 * Delete Module Service
 */
export function deleteModuleServiceV1ModulesModuleIdDelete({ moduleId, authorization, xAccountSlug }: {
    moduleId: string;
    authorization: string;
    xAccountSlug?: string;
}, opts?: Oazapfts.RequestOpts) {
    return oazapfts.ok(oazapfts.fetchJson<{
        status: 200;
        data: any;
    } | {
        status: 400;
        data: HttpErrorResponseRead;
    } | {
        status: 401;
        data: HttpErrorResponseRead;
    } | {
        status: 404;
        data: HttpErrorResponseRead;
    } | {
        status: 422;
    } | {
        status: 500;
        data: HttpErrorResponseRead;
    } | {
        status: 503;
        data: HttpErrorResponseRead;
    }>(`/v1/modules/${encodeURIComponent(moduleId)}`, {
        ...opts,
        method: "DELETE",
        headers: oazapfts.mergeHeaders(opts?.headers, {
            authorization,
            "x-account-slug": xAccountSlug
        })
    }));
}
/**
 * Update Module Service
 */
export function updateModuleServiceV1ModulesModuleIdPut({ moduleId, authorization, xAccountSlug, putModuleRequest }: {
    moduleId: string;
    authorization: string;
    xAccountSlug?: string;
    putModuleRequest: PutModuleRequest;
}, opts?: Oazapfts.RequestOpts) {
    return oazapfts.ok(oazapfts.fetchJson<{
        status: 200;
        data: any;
    } | {
        status: 400;
        data: HttpErrorResponseRead;
    } | {
        status: 401;
        data: HttpErrorResponseRead;
    } | {
        status: 404;
        data: HttpErrorResponseRead;
    } | {
        status: 422;
    } | {
        status: 500;
        data: HttpErrorResponseRead;
    } | {
        status: 503;
        data: HttpErrorResponseRead;
    }>(`/v1/modules/${encodeURIComponent(moduleId)}`, oazapfts.json({
        ...opts,
        method: "PUT",
        body: putModuleRequest,
        headers: oazapfts.mergeHeaders(opts?.headers, {
            authorization,
            "x-account-slug": xAccountSlug
        })
    })));
}
/**
 * Get Module
 */
export function getModuleV1ModulesModuleIdGet({ moduleId, authorization, xAccountSlug }: {
    moduleId: string;
    authorization: string;
    xAccountSlug?: string;
}, opts?: Oazapfts.RequestOpts) {
    return oazapfts.ok(oazapfts.fetchJson<{
        status: 200;
        data: GetModuleResponse;
    } | {
        status: 400;
        data: HttpErrorResponseRead;
    } | {
        status: 401;
        data: HttpErrorResponseRead;
    } | {
        status: 404;
        data: HttpErrorResponseRead;
    } | {
        status: 422;
    } | {
        status: 500;
        data: HttpErrorResponseRead;
    } | {
        status: 503;
        data: HttpErrorResponseRead;
    }>(`/v1/modules/${encodeURIComponent(moduleId)}`, {
        ...opts,
        headers: oazapfts.mergeHeaders(opts?.headers, {
            authorization,
            "x-account-slug": xAccountSlug
        })
    }));
}
/**
 * Get Module Inputs
 */
export function getModuleInputsV1ModulesModuleIdInputsGet({ moduleId, authorization }: {
    moduleId: string;
    authorization: string;
}, opts?: Oazapfts.RequestOpts) {
    return oazapfts.ok(oazapfts.fetchJson<{
        status: 200;
        data: GetModuleInputsResponse | null;
    } | {
        status: 400;
        data: HttpErrorResponseRead;
    } | {
        status: 401;
        data: HttpErrorResponseRead;
    } | {
        status: 404;
        data: HttpErrorResponseRead;
    } | {
        status: 422;
    } | {
        status: 500;
        data: HttpErrorResponseRead;
    } | {
        status: 503;
        data: HttpErrorResponseRead;
    }>(`/v1/modules/${encodeURIComponent(moduleId)}/inputs`, {
        ...opts,
        headers: oazapfts.mergeHeaders(opts?.headers, {
            authorization
        })
    }));
}
/**
 * Get Module Docs
 */
export function getModuleDocsV1ModulesModuleIdDocsGet({ moduleId, language, authorization }: {
    moduleId: string;
    language?: "pt-br" | "en-us";
    authorization: string;
}, opts?: Oazapfts.RequestOpts) {
    return oazapfts.ok(oazapfts.fetchJson<{
        status: 200;
        data: GetModuleDocsResponse | null;
    } | {
        status: 400;
        data: HttpErrorResponseRead;
    } | {
        status: 401;
        data: HttpErrorResponseRead;
    } | {
        status: 404;
        data: HttpErrorResponseRead;
    } | {
        status: 422;
    } | {
        status: 500;
        data: HttpErrorResponseRead;
    } | {
        status: 503;
        data: HttpErrorResponseRead;
    }>(`/v1/modules/${encodeURIComponent(moduleId)}/docs${QS.query(QS.explode({
        language
    }))}`, {
        ...opts,
        headers: oazapfts.mergeHeaders(opts?.headers, {
            authorization
        })
    }));
}
/**
 * Dispatch Module Service
 */
export function dispatchModuleServiceV1ModulesDispatchesPost({ batchName, authorization, body }: {
    batchName?: string | null;
    authorization: string;
    body: ModuleDispatchRequest[];
}, opts?: Oazapfts.RequestOpts) {
    return oazapfts.ok(oazapfts.fetchJson<{
        status: 200;
        data: DispatchModuleResponse;
    } | {
        status: 400;
        data: HttpErrorResponseRead;
    } | {
        status: 401;
        data: HttpErrorResponseRead;
    } | {
        status: 404;
        data: HttpErrorResponseRead;
    } | {
        status: 422;
    } | {
        status: 500;
        data: HttpErrorResponseRead;
    } | {
        status: 503;
        data: HttpErrorResponseRead;
    }>(`/v1/modules/dispatches${QS.query(QS.explode({
        batchName
    }))}`, oazapfts.json({
        ...opts,
        method: "POST",
        body,
        headers: oazapfts.mergeHeaders(opts?.headers, {
            authorization
        })
    })));
}
/**
 * Put Report
 */
export function putReportV1ReportsReportIdPut({ reportId, authorization, putReportRequest }: {
    reportId: string;
    authorization: string;
    putReportRequest: PutReportRequest;
}, opts?: Oazapfts.RequestOpts) {
    return oazapfts.ok(oazapfts.fetchJson<{
        status: 200;
        data: any;
    } | {
        status: 400;
        data: HttpErrorResponseRead;
    } | {
        status: 401;
        data: HttpErrorResponseRead;
    } | {
        status: 404;
        data: HttpErrorResponseRead;
    } | {
        status: 422;
    } | {
        status: 500;
        data: HttpErrorResponseRead;
    } | {
        status: 503;
        data: HttpErrorResponseRead;
    }>(`/v1/reports/${encodeURIComponent(reportId)}`, oazapfts.json({
        ...opts,
        method: "PUT",
        body: putReportRequest,
        headers: oazapfts.mergeHeaders(opts?.headers, {
            authorization
        })
    })));
}
/**
 * Get Report
 */
export function getReportV1ReportsReportIdGet({ reportId, authorization }: {
    reportId: string;
    authorization: string;
}, opts?: Oazapfts.RequestOpts) {
    return oazapfts.ok(oazapfts.fetchJson<{
        status: 200;
        data: GetReportResponseRead;
    } | {
        status: 400;
        data: HttpErrorResponseRead;
    } | {
        status: 401;
        data: HttpErrorResponseRead;
    } | {
        status: 404;
        data: HttpErrorResponseRead;
    } | {
        status: 422;
    } | {
        status: 500;
        data: HttpErrorResponseRead;
    } | {
        status: 503;
        data: HttpErrorResponseRead;
    }>(`/v1/reports/${encodeURIComponent(reportId)}`, {
        ...opts,
        headers: oazapfts.mergeHeaders(opts?.headers, {
            authorization
        })
    }));
}
/**
 * Get Report Pull Request Content
 */
export function getReportPullRequestContentV1ReportsReportIdPullRequestGet({ reportId, authorization }: {
    reportId: string;
    authorization: string;
}, opts?: Oazapfts.RequestOpts) {
    return oazapfts.ok(oazapfts.fetchJson<{
        status: 200;
        data: GetPullRequest;
    } | {
        status: 400;
        data: HttpErrorResponseRead;
    } | {
        status: 401;
        data: HttpErrorResponseRead;
    } | {
        status: 404;
        data: HttpErrorResponseRead;
    } | {
        status: 422;
    } | {
        status: 500;
        data: HttpErrorResponseRead;
    } | {
        status: 503;
        data: HttpErrorResponseRead;
    }>(`/v1/reports/${encodeURIComponent(reportId)}/pull-request`, {
        ...opts,
        headers: oazapfts.mergeHeaders(opts?.headers, {
            authorization
        })
    }));
}
/**
 * Download Report
 */
export function downloadReportV1ReportsReportIdDownloadGet({ reportId, extension, authorization }: {
    reportId: string;
    extension?: "csv" | "xlsx";
    authorization: string;
}, opts?: Oazapfts.RequestOpts) {
    return oazapfts.ok(oazapfts.fetchJson<{
        status: 200;
        data: any;
    } | {
        status: 400;
        data: HttpErrorResponseRead;
    } | {
        status: 401;
        data: HttpErrorResponseRead;
    } | {
        status: 404;
        data: HttpErrorResponseRead;
    } | {
        status: 422;
    } | {
        status: 500;
        data: HttpErrorResponseRead;
    } | {
        status: 503;
        data: HttpErrorResponseRead;
    }>(`/v1/reports/${encodeURIComponent(reportId)}/download${QS.query(QS.explode({
        extension
    }))}`, {
        ...opts,
        headers: oazapfts.mergeHeaders(opts?.headers, {
            authorization
        })
    }));
}
/**
 * Put Customer Rating Report
 */
export function putCustomerRatingReportV1ReportsReportIdCustomerRatingPut({ reportId, authorization, putCustomerRatingReportRequest }: {
    reportId: string;
    authorization: string;
    putCustomerRatingReportRequest: PutCustomerRatingReportRequest;
}, opts?: Oazapfts.RequestOpts) {
    return oazapfts.ok(oazapfts.fetchJson<{
        status: 200;
        data: any;
    } | {
        status: 400;
        data: HttpErrorResponseRead;
    } | {
        status: 401;
        data: HttpErrorResponseRead;
    } | {
        status: 404;
        data: HttpErrorResponseRead;
    } | {
        status: 422;
    } | {
        status: 500;
        data: HttpErrorResponseRead;
    } | {
        status: 503;
        data: HttpErrorResponseRead;
    }>(`/v1/reports/${encodeURIComponent(reportId)}/customer-rating`, oazapfts.json({
        ...opts,
        method: "PUT",
        body: putCustomerRatingReportRequest,
        headers: oazapfts.mergeHeaders(opts?.headers, {
            authorization
        })
    })));
}
/**
 * Search Repos Scm Service
 */
export function searchReposScmServiceV1ReposSearchScmGet({ repoName, authorization }: {
    repoName: string;
    authorization: string;
}, opts?: Oazapfts.RequestOpts) {
    return oazapfts.ok(oazapfts.fetchJson<{
        status: 200;
        data: SearchRepoScmResponse;
    } | {
        status: 422;
        data: HttpValidationError;
    }>(`/v1/repos/search-scm${QS.query(QS.explode({
        repoName
    }))}`, {
        ...opts,
        headers: oazapfts.mergeHeaders(opts?.headers, {
            authorization
        })
    }));
}
/**
 * Get Status Search Repos Scm Service
 */
export function getStatusSearchReposScmServiceV1ReposSearchScmSearchRepoIdStatusGet({ searchRepoId, authorization }: {
    searchRepoId: string;
    authorization: string;
}, opts?: Oazapfts.RequestOpts) {
    return oazapfts.ok(oazapfts.fetchJson<{
        status: 200;
        data: GetStatusSearchRepoScmResponse;
    } | {
        status: 422;
        data: HttpValidationError;
    }>(`/v1/repos/search-scm/${encodeURIComponent(searchRepoId)}/status`, {
        ...opts,
        headers: oazapfts.mergeHeaders(opts?.headers, {
            authorization
        })
    }));
}
/**
 * Download Search Repos Scm Service
 */
export function downloadSearchReposScmServiceV1ReposSearchScmSearchRepoIdDownloadGet({ searchRepoId, extension, authorization }: {
    searchRepoId: string;
    extension?: "csv" | "xlsx";
    authorization: string;
}, opts?: Oazapfts.RequestOpts) {
    return oazapfts.ok(oazapfts.fetchJson<{
        status: 200;
        data: any;
    } | {
        status: 422;
        data: HttpValidationError;
    }>(`/v1/repos/search-scm/${encodeURIComponent(searchRepoId)}/download${QS.query(QS.explode({
        extension
    }))}`, {
        ...opts,
        headers: oazapfts.mergeHeaders(opts?.headers, {
            authorization
        })
    }));
}
/**
 * List Repository Report
 */
export function listRepositoryReportV1ReposRepositoryIdReportsGet({ repositoryId, pageSize, page, lastEvaluatedKey, authorization }: {
    repositoryId: string;
    pageSize?: number;
    page?: number;
    lastEvaluatedKey?: string;
    authorization: string;
}, opts?: Oazapfts.RequestOpts) {
    return oazapfts.ok(oazapfts.fetchJson<{
        status: 200;
        data: ListRepositoryReportResponseRead;
    } | {
        status: 400;
        data: HttpErrorResponseRead;
    } | {
        status: 401;
        data: HttpErrorResponseRead;
    } | {
        status: 404;
        data: HttpErrorResponseRead;
    } | {
        status: 422;
    } | {
        status: 500;
        data: HttpErrorResponseRead;
    } | {
        status: 503;
        data: HttpErrorResponseRead;
    }>(`/v1/repos/${encodeURIComponent(repositoryId)}/reports${QS.query(QS.explode({
        pageSize,
        page,
        lastEvaluatedKey
    }))}`, {
        ...opts,
        headers: oazapfts.mergeHeaders(opts?.headers, {
            authorization
        })
    }));
}
/**
 * Create Repos Batch Service
 */
export function createReposBatchServiceV1ReposBatchPost({ tags, authorization, fastapiCompatV2BodyCreateReposBatchServiceV1ReposBatchPost }: {
    tags?: string[] | null;
    authorization: string;
    fastapiCompatV2BodyCreateReposBatchServiceV1ReposBatchPost: BodyCreateReposBatchServiceV1ReposBatchPost;
}, opts?: Oazapfts.RequestOpts) {
    return oazapfts.ok(oazapfts.fetchJson<{
        status: 200;
        data: any;
    } | {
        status: 207;
        data: HttpErrorResponseRead;
    } | {
        status: 400;
        data: HttpErrorResponseRead;
    } | {
        status: 401;
        data: HttpErrorResponseRead;
    } | {
        status: 404;
        data: HttpErrorResponseRead;
    } | {
        status: 422;
    } | {
        status: 500;
        data: HttpErrorResponseRead;
    } | {
        status: 503;
        data: HttpErrorResponseRead;
    }>(`/v1/repos/batch${QS.query(QS.explode({
        tags
    }))}`, oazapfts.multipart({
        ...opts,
        method: "POST",
        body: fastapiCompatV2BodyCreateReposBatchServiceV1ReposBatchPost,
        headers: oazapfts.mergeHeaders(opts?.headers, {
            authorization
        })
    })));
}
/**
 * Create Repository Service
 */
export function createRepositoryServiceV1ReposPost({ authorization, srcPresentationApiRestServicesV1ReposRequestsCreateRepositoryRequestCreateRepositoryRequest }: {
    authorization: string;
    srcPresentationApiRestServicesV1ReposRequestsCreateRepositoryRequestCreateRepositoryRequest: CreateRepositoryRequest;
}, opts?: Oazapfts.RequestOpts) {
    return oazapfts.ok(oazapfts.fetchJson<{
        status: 200;
        data: any;
    } | {
        status: 400;
        data: HttpErrorResponseRead;
    } | {
        status: 401;
        data: HttpErrorResponseRead;
    } | {
        status: 404;
        data: HttpErrorResponseRead;
    } | {
        status: 422;
    } | {
        status: 500;
        data: HttpErrorResponseRead;
    } | {
        status: 503;
        data: HttpErrorResponseRead;
    }>("/v1/repos", oazapfts.json({
        ...opts,
        method: "POST",
        body: srcPresentationApiRestServicesV1ReposRequestsCreateRepositoryRequestCreateRepositoryRequest,
        headers: oazapfts.mergeHeaders(opts?.headers, {
            authorization
        })
    })));
}
/**
 * List Repository Service
 */
export function listRepositoryServiceV1ReposGet({ pageSize, page, lastEvaluatedKey, filter, moduleId, createdByEmail, reportByEmail, orderBy, orderDirection, tags, authorization }: {
    pageSize?: number;
    page?: number;
    lastEvaluatedKey?: string;
    filter?: string;
    moduleId?: string;
    createdByEmail?: string | null;
    reportByEmail?: string | null;
    orderBy?: "repository.url" | "repository.created_at" | "repository.updated_at" | "report.issues_count" | "report.total_files_count" | "report.analyzed_files_count" | "report.issues_files_count" | "report.created_at" | "report.updated_at" | "execution.status";
    orderDirection?: "asc" | "desc" | "ASC" | "DESC";
    tags?: string[] | null;
    authorization: string;
}, opts?: Oazapfts.RequestOpts) {
    return oazapfts.ok(oazapfts.fetchJson<{
        status: 200;
        data: ListRepositoryResponseRead;
    } | {
        status: 400;
        data: HttpErrorResponseRead;
    } | {
        status: 401;
        data: HttpErrorResponseRead;
    } | {
        status: 404;
        data: HttpErrorResponseRead;
    } | {
        status: 422;
    } | {
        status: 500;
        data: HttpErrorResponseRead;
    } | {
        status: 503;
        data: HttpErrorResponseRead;
    }>(`/v1/repos${QS.query(QS.explode({
        pageSize,
        page,
        lastEvaluatedKey,
        filter,
        moduleId,
        createdByEmail,
        reportByEmail,
        orderBy,
        orderDirection,
        tags
    }))}`, {
        ...opts,
        headers: oazapfts.mergeHeaders(opts?.headers, {
            authorization
        })
    }));
}
/**
 * List Repository Download Service
 */
export function listRepositoryDownloadServiceV1ReposDownloadGet({ filter, moduleId, createdByEmail, reportByEmail, orderBy, orderDirection, tags, extension, authorization }: {
    filter?: string;
    moduleId?: string;
    createdByEmail?: string | null;
    reportByEmail?: string | null;
    orderBy?: "repository.url" | "repository.created_at" | "repository.updated_at" | "report.issues_count" | "report.total_files_count" | "report.analyzed_files_count" | "report.issues_files_count" | "report.created_at" | "report.updated_at" | "execution.status";
    orderDirection?: "asc" | "desc" | "ASC" | "DESC";
    tags?: string[] | null;
    extension?: "csv" | "xlsx";
    authorization: string;
}, opts?: Oazapfts.RequestOpts) {
    return oazapfts.ok(oazapfts.fetchJson<{
        status: 200;
        data: any;
    } | {
        status: 400;
        data: HttpErrorResponseRead;
    } | {
        status: 401;
        data: HttpErrorResponseRead;
    } | {
        status: 404;
        data: HttpErrorResponseRead;
    } | {
        status: 422;
    } | {
        status: 500;
        data: HttpErrorResponseRead;
    } | {
        status: 503;
        data: HttpErrorResponseRead;
    }>(`/v1/repos/download${QS.query(QS.explode({
        filter,
        moduleId,
        createdByEmail,
        reportByEmail,
        orderBy,
        orderDirection,
        tags,
        extension
    }))}`, {
        ...opts,
        headers: oazapfts.mergeHeaders(opts?.headers, {
            authorization
        })
    }));
}
/**
 * List Branches Service
 */
export function listBranchesServiceV1ReposBranchesGet({ repositoryUrl, authorization }: {
    repositoryUrl: string;
    authorization: string;
}, opts?: Oazapfts.RequestOpts) {
    return oazapfts.ok(oazapfts.fetchJson<{
        status: 200;
        data: ListBranchesResponse;
    } | {
        status: 400;
        data: HttpErrorResponseRead;
    } | {
        status: 401;
        data: HttpErrorResponseRead;
    } | {
        status: 404;
        data: HttpErrorResponseRead;
    } | {
        status: 422;
    } | {
        status: 500;
        data: HttpErrorResponseRead;
    } | {
        status: 503;
        data: HttpErrorResponseRead;
    }>(`/v1/repos/branches${QS.query(QS.explode({
        repositoryUrl
    }))}`, {
        ...opts,
        headers: oazapfts.mergeHeaders(opts?.headers, {
            authorization
        })
    }));
}
/**
 * Get Repository By Id Service
 */
export function getRepositoryByIdServiceV1ReposRepositoryIdGet({ repositoryId, authorization }: {
    repositoryId: string;
    authorization: string;
}, opts?: Oazapfts.RequestOpts) {
    return oazapfts.ok(oazapfts.fetchJson<{
        status: 200;
        data: RepositoryResponseRead;
    } | {
        status: 400;
        data: HttpErrorResponseRead;
    } | {
        status: 401;
        data: HttpErrorResponseRead;
    } | {
        status: 404;
        data: HttpErrorResponseRead;
    } | {
        status: 422;
    } | {
        status: 500;
        data: HttpErrorResponseRead;
    } | {
        status: 503;
        data: HttpErrorResponseRead;
    }>(`/v1/repos/${encodeURIComponent(repositoryId)}`, {
        ...opts,
        headers: oazapfts.mergeHeaders(opts?.headers, {
            authorization
        })
    }));
}
/**
 * Delete Repository Service
 */
export function deleteRepositoryServiceV1ReposRepositoryIdDelete({ repositoryId, authorization }: {
    repositoryId: string;
    authorization: string;
}, opts?: Oazapfts.RequestOpts) {
    return oazapfts.ok(oazapfts.fetchJson<{
        status: 200;
        data: any;
    } | {
        status: 400;
        data: HttpErrorResponseRead;
    } | {
        status: 401;
        data: HttpErrorResponseRead;
    } | {
        status: 404;
        data: HttpErrorResponseRead;
    } | {
        status: 422;
    } | {
        status: 500;
        data: HttpErrorResponseRead;
    } | {
        status: 503;
        data: HttpErrorResponseRead;
    }>(`/v1/repos/${encodeURIComponent(repositoryId)}`, {
        ...opts,
        method: "DELETE",
        headers: oazapfts.mergeHeaders(opts?.headers, {
            authorization
        })
    }));
}
/**
 * Update Repository Service
 */
export function updateRepositoryServiceV1ReposRepositoryIdPut({ repositoryId, authorization, putRepositoryRequest }: {
    repositoryId: string;
    authorization: string;
    putRepositoryRequest: PutRepositoryRequest;
}, opts?: Oazapfts.RequestOpts) {
    return oazapfts.ok(oazapfts.fetchJson<{
        status: 200;
        data: any;
    } | {
        status: 400;
        data: HttpErrorResponseRead;
    } | {
        status: 401;
        data: HttpErrorResponseRead;
    } | {
        status: 404;
        data: HttpErrorResponseRead;
    } | {
        status: 422;
    } | {
        status: 500;
        data: HttpErrorResponseRead;
    } | {
        status: 503;
        data: HttpErrorResponseRead;
    }>(`/v1/repos/${encodeURIComponent(repositoryId)}`, oazapfts.json({
        ...opts,
        method: "PUT",
        body: putRepositoryRequest,
        headers: oazapfts.mergeHeaders(opts?.headers, {
            authorization
        })
    })));
}
/**
 * Validate Scm Url Service
 */
export function validateScmUrlServiceV1ReposValidateScmUrlPost({ raiseConflict, authorization, validateScmUrlRequest }: {
    raiseConflict?: boolean;
    authorization: string;
    validateScmUrlRequest: ValidateScmUrlRequest;
}, opts?: Oazapfts.RequestOpts) {
    return oazapfts.ok(oazapfts.fetchJson<{
        status: 200;
        data: ValidateScmUrlResponse;
    } | {
        status: 400;
        data: HttpErrorResponseRead;
    } | {
        status: 401;
        data: HttpErrorResponseRead;
    } | {
        status: 404;
        data: HttpErrorResponseRead;
    } | {
        status: 422;
    } | {
        status: 500;
        data: HttpErrorResponseRead;
    } | {
        status: 503;
        data: HttpErrorResponseRead;
    }>(`/v1/repos/validate-scm-url${QS.query(QS.explode({
        raiseConflict
    }))}`, oazapfts.json({
        ...opts,
        method: "POST",
        body: validateScmUrlRequest,
        headers: oazapfts.mergeHeaders(opts?.headers, {
            authorization
        })
    })));
}
/**
 * Get Account Settings
 */
export function getAccountSettingsV1SettingsGet({ authorization }: {
    authorization: string;
}, opts?: Oazapfts.RequestOpts) {
    return oazapfts.ok(oazapfts.fetchJson<{
        status: 200;
        data: GetAccountSettingsRequest;
    } | {
        status: 422;
        data: HttpValidationError;
    }>("/v1/settings", {
        ...opts,
        headers: oazapfts.mergeHeaders(opts?.headers, {
            authorization
        })
    }));
}
/**
 * Create Account Settings
 */
export function createAccountSettingsV1SettingsPut({ authorization, createAccountSettingsRequest }: {
    authorization: string;
    createAccountSettingsRequest: CreateAccountSettingsRequest;
}, opts?: Oazapfts.RequestOpts) {
    return oazapfts.ok(oazapfts.fetchJson<{
        status: 200;
        data: any;
    } | {
        status: 422;
        data: HttpValidationError;
    }>("/v1/settings", oazapfts.json({
        ...opts,
        method: "PUT",
        body: createAccountSettingsRequest,
        headers: oazapfts.mergeHeaders(opts?.headers, {
            authorization
        })
    })));
}
/**
 * Create Integration Service
 */
export function createIntegrationServiceV1IntegrationsPost({ authorization, createIntegrationRequest }: {
    authorization: string;
    createIntegrationRequest: CreateIntegrationRequest;
}, opts?: Oazapfts.RequestOpts) {
    return oazapfts.ok(oazapfts.fetchJson<{
        status: 200;
        data: CreateIntegrationResponse;
    } | {
        status: 422;
        data: HttpValidationError;
    }>("/v1/integrations", oazapfts.json({
        ...opts,
        method: "POST",
        body: createIntegrationRequest,
        headers: oazapfts.mergeHeaders(opts?.headers, {
            authorization
        })
    })));
}
/**
 * List Integration Service
 */
export function listIntegrationServiceV1IntegrationsGet({ pageSize, page, name, authorization }: {
    pageSize?: number;
    page?: number;
    name?: string | null;
    authorization: string;
}, opts?: Oazapfts.RequestOpts) {
    return oazapfts.ok(oazapfts.fetchJson<{
        status: 200;
        data: ListIntegrationResponse;
    } | {
        status: 422;
        data: HttpValidationError;
    }>(`/v1/integrations${QS.query(QS.explode({
        pageSize,
        page,
        name
    }))}`, {
        ...opts,
        headers: oazapfts.mergeHeaders(opts?.headers, {
            authorization
        })
    }));
}
/**
 * Get Integration By Id Service
 */
export function getIntegrationByIdServiceV1IntegrationsIntegrationIdGet({ integrationId, authorization }: {
    integrationId: string;
    authorization: string;
}, opts?: Oazapfts.RequestOpts) {
    return oazapfts.ok(oazapfts.fetchJson<{
        status: 200;
        data: IntegrationResponse;
    } | {
        status: 422;
        data: HttpValidationError;
    }>(`/v1/integrations/${encodeURIComponent(integrationId)}`, {
        ...opts,
        headers: oazapfts.mergeHeaders(opts?.headers, {
            authorization
        })
    }));
}
/**
 * Update Integration Service
 */
export function updateIntegrationServiceV1IntegrationsIntegrationIdPut({ integrationId, authorization, putIntegrationRequest }: {
    integrationId: string;
    authorization: string;
    putIntegrationRequest: PutIntegrationRequest;
}, opts?: Oazapfts.RequestOpts) {
    return oazapfts.ok(oazapfts.fetchJson<{
        status: 200;
        data: any;
    } | {
        status: 422;
        data: HttpValidationError;
    }>(`/v1/integrations/${encodeURIComponent(integrationId)}`, oazapfts.json({
        ...opts,
        method: "PUT",
        body: putIntegrationRequest,
        headers: oazapfts.mergeHeaders(opts?.headers, {
            authorization
        })
    })));
}
/**
 * Delete Integration Service
 */
export function deleteIntegrationServiceV1IntegrationsIntegrationIdDelete({ integrationId, authorization }: {
    integrationId: string;
    authorization: string;
}, opts?: Oazapfts.RequestOpts) {
    return oazapfts.ok(oazapfts.fetchJson<{
        status: 200;
        data: any;
    } | {
        status: 422;
        data: HttpValidationError;
    }>(`/v1/integrations/${encodeURIComponent(integrationId)}`, {
        ...opts,
        method: "DELETE",
        headers: oazapfts.mergeHeaders(opts?.headers, {
            authorization
        })
    }));
}
/**
 * Create Program Group Service
 */
export function createProgramGroupServiceV1ProgramGroupsPost({ name, tags, integrationId, authorization, fastapiCompatV2BodyCreateProgramGroupServiceV1ProgramGroupsPost }: {
    name: string;
    tags?: string[] | null;
    integrationId?: string;
    authorization: string;
    fastapiCompatV2BodyCreateProgramGroupServiceV1ProgramGroupsPost?: BodyCreateProgramGroupServiceV1ProgramGroupsPost;
}, opts?: Oazapfts.RequestOpts) {
    return oazapfts.ok(oazapfts.fetchJson<{
        status: 200;
        data: CreateProgramGroupResponse;
    } | {
        status: 422;
        data: HttpValidationError;
    }>(`/v1/program-groups${QS.query(QS.explode({
        name,
        tags,
        integrationId
    }))}`, oazapfts.multipart({
        ...opts,
        method: "POST",
        body: fastapiCompatV2BodyCreateProgramGroupServiceV1ProgramGroupsPost,
        headers: oazapfts.mergeHeaders(opts?.headers, {
            authorization
        })
    })));
}
/**
 * List Program Group Service
 */
export function listProgramGroupServiceV1ProgramGroupsGet({ moduleId, createdByEmail, reportByEmail, tags, name, pageSize, page, orderBy, orderDirection, integrationId, authorization }: {
    moduleId?: string | null;
    createdByEmail?: string | null;
    reportByEmail?: string | null;
    tags?: string[] | null;
    name?: string | null;
    pageSize?: number;
    page?: number;
    orderBy?: "program_group.name" | "program_group.created_at" | "program_group.updated_at" | "report.issues_count" | "report.total_files_count" | "report.analyzed_files_count" | "report.issues_files_count" | "report.created_at" | "report.updated_at" | "execution.status";
    orderDirection?: "asc" | "desc" | "ASC" | "DESC";
    integrationId?: string;
    authorization: string;
}, opts?: Oazapfts.RequestOpts) {
    return oazapfts.ok(oazapfts.fetchJson<{
        status: 200;
        data: ListProgramGroupResponseRead;
    } | {
        status: 422;
        data: HttpValidationError;
    }>(`/v1/program-groups${QS.query(QS.explode({
        moduleId,
        createdByEmail,
        reportByEmail,
        tags,
        name,
        pageSize,
        page,
        orderBy,
        orderDirection,
        integrationId
    }))}`, {
        ...opts,
        headers: oazapfts.mergeHeaders(opts?.headers, {
            authorization
        })
    }));
}
/**
 * List Program Group Download Service
 */
export function listProgramGroupDownloadServiceV1ProgramGroupsDownloadGet({ moduleId, createdByEmail, reportByEmail, tags, name, orderBy, orderDirection, integrationId, extension, authorization }: {
    moduleId?: string | null;
    createdByEmail?: string | null;
    reportByEmail?: string | null;
    tags?: string[] | null;
    name?: string | null;
    orderBy?: "program_group.name" | "program_group.created_at" | "program_group.updated_at" | "report.issues_count" | "report.total_files_count" | "report.analyzed_files_count" | "report.issues_files_count" | "report.created_at" | "report.updated_at" | "execution.status";
    orderDirection?: "asc" | "desc" | "ASC" | "DESC";
    integrationId?: string;
    extension?: "csv" | "xlsx";
    authorization: string;
}, opts?: Oazapfts.RequestOpts) {
    return oazapfts.ok(oazapfts.fetchJson<{
        status: 200;
        data: any;
    } | {
        status: 422;
        data: HttpValidationError;
    }>(`/v1/program-groups/download${QS.query(QS.explode({
        moduleId,
        createdByEmail,
        reportByEmail,
        tags,
        name,
        orderBy,
        orderDirection,
        integrationId,
        extension
    }))}`, {
        ...opts,
        headers: oazapfts.mergeHeaders(opts?.headers, {
            authorization
        })
    }));
}
/**
 * Get Program Group By Id Service
 */
export function getProgramGroupByIdServiceV1ProgramGroupsProgramGroupIdGet({ programGroupId, authorization }: {
    programGroupId: string;
    authorization: string;
}, opts?: Oazapfts.RequestOpts) {
    return oazapfts.ok(oazapfts.fetchJson<{
        status: 200;
        data: ProgramGroupResponse3;
    } | {
        status: 422;
        data: HttpValidationError;
    }>(`/v1/program-groups/${encodeURIComponent(programGroupId)}`, {
        ...opts,
        headers: oazapfts.mergeHeaders(opts?.headers, {
            authorization
        })
    }));
}
/**
 * Update Program Group Service
 */
export function updateProgramGroupServiceV1ProgramGroupsProgramGroupIdPut({ programGroupId, name, tags, integrationId, authorization, fastapiCompatV2BodyUpdateProgramGroupServiceV1ProgramGroupsProgramGroupIdPut }: {
    programGroupId: string;
    name?: string | null;
    tags?: string[] | null;
    integrationId?: string | null;
    authorization: string;
    fastapiCompatV2BodyUpdateProgramGroupServiceV1ProgramGroupsProgramGroupIdPut?: BodyUpdateProgramGroupServiceV1ProgramGroupsProgramGroupIdPut;
}, opts?: Oazapfts.RequestOpts) {
    return oazapfts.ok(oazapfts.fetchJson<{
        status: 200;
        data: any;
    } | {
        status: 422;
        data: HttpValidationError;
    }>(`/v1/program-groups/${encodeURIComponent(programGroupId)}${QS.query(QS.explode({
        name,
        tags,
        integrationId
    }))}`, oazapfts.multipart({
        ...opts,
        method: "PUT",
        body: fastapiCompatV2BodyUpdateProgramGroupServiceV1ProgramGroupsProgramGroupIdPut,
        headers: oazapfts.mergeHeaders(opts?.headers, {
            authorization
        })
    })));
}
/**
 * Delete Program Group Service
 */
export function deleteProgramGroupServiceV1ProgramGroupsProgramGroupIdDelete({ programGroupId, authorization }: {
    programGroupId: string;
    authorization: string;
}, opts?: Oazapfts.RequestOpts) {
    return oazapfts.ok(oazapfts.fetchJson<{
        status: 200;
        data: any;
    } | {
        status: 422;
        data: HttpValidationError;
    }>(`/v1/program-groups/${encodeURIComponent(programGroupId)}`, {
        ...opts,
        method: "DELETE",
        headers: oazapfts.mergeHeaders(opts?.headers, {
            authorization
        })
    }));
}
/**
 * Update Program Group Components Service
 */
export function updateProgramGroupComponentsServiceV1ProgramGroupsProgramGroupIdComponentsPut({ programGroupId, authorization, body }: {
    programGroupId: string;
    authorization: string;
    body: PutProgramGroupComponentsRequest[];
}, opts?: Oazapfts.RequestOpts) {
    return oazapfts.ok(oazapfts.fetchJson<{
        status: 200;
        data: any;
    } | {
        status: 422;
        data: HttpValidationError;
    }>(`/v1/program-groups/${encodeURIComponent(programGroupId)}/components`, oazapfts.json({
        ...opts,
        method: "PUT",
        body,
        headers: oazapfts.mergeHeaders(opts?.headers, {
            authorization
        })
    })));
}
/**
 * List Program Group Report Service
 */
export function listProgramGroupReportServiceV1ProgramGroupsProgramGroupIdReportsGet({ programGroupId, pageSize, page, lastEvaluatedKey, authorization }: {
    programGroupId: string;
    pageSize?: number;
    page?: number;
    lastEvaluatedKey?: string;
    authorization: string;
}, opts?: Oazapfts.RequestOpts) {
    return oazapfts.ok(oazapfts.fetchJson<{
        status: 200;
        data: ListProgramGroupReportResponseRead;
    } | {
        status: 422;
        data: HttpValidationError;
    }>(`/v1/program-groups/${encodeURIComponent(programGroupId)}/reports${QS.query(QS.explode({
        pageSize,
        page,
        lastEvaluatedKey
    }))}`, {
        ...opts,
        headers: oazapfts.mergeHeaders(opts?.headers, {
            authorization
        })
    }));
}
/**
 * List Tags Service
 */
export function listTagsServiceV1TagsGet({ name, $type, authorization }: {
    name?: string | null;
    $type?: "Repository" | "Module";
    authorization: string;
}, opts?: Oazapfts.RequestOpts) {
    return oazapfts.ok(oazapfts.fetchJson<{
        status: 200;
        data: ListTagResponse;
    } | {
        status: 422;
        data: HttpValidationError;
    }>(`/v1/tags${QS.query(QS.explode({
        name,
        "type": $type
    }))}`, {
        ...opts,
        headers: oazapfts.mergeHeaders(opts?.headers, {
            authorization
        })
    }));
}
/**
 * General Report Success And Errors
 */
export function generalReportSuccessAndErrorsV1AnalyticsReportsGeneralSuccessAndErrorsGet({ startDate, endDate, authorization }: {
    startDate?: string;
    endDate?: string;
    authorization: string;
}, opts?: Oazapfts.RequestOpts) {
    return oazapfts.ok(oazapfts.fetchJson<{
        status: 200;
        data: GeneralReportSuccessAndErrorsResponses;
    } | {
        status: 400;
        data: HttpErrorResponseRead;
    } | {
        status: 401;
        data: HttpErrorResponseRead;
    } | {
        status: 404;
        data: HttpErrorResponseRead;
    } | {
        status: 422;
    } | {
        status: 500;
        data: HttpErrorResponseRead;
    } | {
        status: 503;
        data: HttpErrorResponseRead;
    }>(`/v1/analytics/reports/general/success-and-errors${QS.query(QS.explode({
        startDate,
        endDate
    }))}`, {
        ...opts,
        headers: oazapfts.mergeHeaders(opts?.headers, {
            authorization
        })
    }));
}
/**
 * Analytics Repository Last Report Status
 */
export function analyticsRepositoryLastReportStatusV1AnalyticsRepositoriesLastReportStatusGet({ pageSize, page, moduleId, executedBy, repositoryUrl, tags, orderBy, orderDirection, startDate, endDate, mode, authorization }: {
    pageSize?: number;
    page?: number;
    moduleId?: string[] | null;
    executedBy?: string[] | null;
    repositoryUrl?: string | null;
    tags?: string[] | null;
    orderBy?: ("repositoryUrl" | "moduleName" | "merged" | "mode" | "reportCreatedAt" | "reportCreatedBy" | "totalFiles" | "totalFilesAnalyzed" | "totalFilesIssues" | "totalFilesChanged" | "totalIssues") | null;
    orderDirection?: ("ASC" | "DESC") | null;
    startDate?: string | null;
    endDate?: string | null;
    mode?: ("scan" | "fix") | null;
    authorization: string;
}, opts?: Oazapfts.RequestOpts) {
    return oazapfts.ok(oazapfts.fetchJson<{
        status: 200;
        data: AnalyticsRepositoryLastReportStatusResponse;
    } | {
        status: 400;
        data: HttpErrorResponseRead;
    } | {
        status: 401;
        data: HttpErrorResponseRead;
    } | {
        status: 404;
        data: HttpErrorResponseRead;
    } | {
        status: 422;
    } | {
        status: 500;
        data: HttpErrorResponseRead;
    } | {
        status: 503;
        data: HttpErrorResponseRead;
    }>(`/v1/analytics/repositories/last-report-status${QS.query(QS.explode({
        pageSize,
        page,
        moduleId,
        executedBy,
        repositoryUrl,
        tags,
        orderBy,
        orderDirection,
        startDate,
        endDate,
        mode
    }))}`, {
        ...opts,
        headers: oazapfts.mergeHeaders(opts?.headers, {
            authorization
        })
    }));
}
/**
 * Analytics Repository Last Report Status Download
 */
export function analyticsRepositoryLastReportStatusDownloadV1AnalyticsRepositoriesLastReportStatusDownloadGet({ moduleId, orderBy, tags, executedBy, repositoryUrl, orderDirection, startDate, endDate, extension, mode, authorization }: {
    moduleId?: string[] | null;
    orderBy?: ("repositoryUrl" | "moduleName" | "mode" | "reportCreatedAt" | "reportCreatedBy" | "totalFiles" | "totalFilesAnalyzed" | "totalFilesIssues" | "totalFilesChanged" | "totalIssues") | null;
    tags?: string[] | null;
    executedBy?: string[] | null;
    repositoryUrl?: string | null;
    orderDirection?: ("ASC" | "DESC") | null;
    startDate?: string;
    endDate?: string;
    extension?: "csv" | "xlsx";
    mode?: ("scan" | "fix") | null;
    authorization: string;
}, opts?: Oazapfts.RequestOpts) {
    return oazapfts.ok(oazapfts.fetchJson<{
        status: 200;
        data: any;
    } | {
        status: 400;
        data: HttpErrorResponseRead;
    } | {
        status: 401;
        data: HttpErrorResponseRead;
    } | {
        status: 404;
        data: HttpErrorResponseRead;
    } | {
        status: 422;
    } | {
        status: 500;
        data: HttpErrorResponseRead;
    } | {
        status: 503;
        data: HttpErrorResponseRead;
    }>(`/v1/analytics/repositories/last-report-status/download${QS.query(QS.explode({
        moduleId,
        orderBy,
        tags,
        executedBy,
        repositoryUrl,
        orderDirection,
        startDate,
        endDate,
        extension,
        mode
    }))}`, {
        ...opts,
        headers: oazapfts.mergeHeaders(opts?.headers, {
            authorization
        })
    }));
}
/**
 * Analytics Repository Usage
 */
export function analyticsRepositoryUsageV1AnalyticsRepositoriesUsageGet({ pageSize, page, moduleId, userId, tags, orderBy, executedBy, repositoryUrl, orderDirection, startDate, endDate, authorization }: {
    pageSize?: number;
    page?: number;
    moduleId?: string[] | null;
    userId?: string | null;
    tags?: string[] | null;
    orderBy?: ("repositoryUrl" | "totalScan" | "totalScanFailure" | "totalScanSuccess" | "totalFix" | "totalFixFailure" | "totalFixSuccess") | null;
    executedBy?: string[] | null;
    repositoryUrl?: string | null;
    orderDirection?: ("ASC" | "DESC") | null;
    startDate?: string;
    endDate?: string;
    authorization: string;
}, opts?: Oazapfts.RequestOpts) {
    return oazapfts.ok(oazapfts.fetchJson<{
        status: 200;
        data: AnalyticsRepositoryUsageResponse;
    } | {
        status: 400;
        data: HttpErrorResponseRead;
    } | {
        status: 401;
        data: HttpErrorResponseRead;
    } | {
        status: 404;
        data: HttpErrorResponseRead;
    } | {
        status: 422;
    } | {
        status: 500;
        data: HttpErrorResponseRead;
    } | {
        status: 503;
        data: HttpErrorResponseRead;
    }>(`/v1/analytics/repositories/usage${QS.query(QS.explode({
        pageSize,
        page,
        moduleId,
        user_id: userId,
        tags,
        orderBy,
        executedBy,
        repositoryUrl,
        orderDirection,
        startDate,
        endDate
    }))}`, {
        ...opts,
        headers: oazapfts.mergeHeaders(opts?.headers, {
            authorization
        })
    }));
}
/**
 * Analytics Repository Usage Download
 */
export function analyticsRepositoryUsageDownloadV1AnalyticsRepositoriesUsageDownloadGet({ moduleId, userId, orderBy, tags, executedBy, repositoryUrl, orderDirection, startDate, endDate, extension, authorization }: {
    moduleId?: string[] | null;
    userId?: string | null;
    orderBy?: ("repositoryUrl" | "totalScan" | "totalScanFailure" | "totalScanSuccess" | "totalFix" | "totalFixFailure" | "totalFixSuccess") | null;
    tags?: string[] | null;
    executedBy?: string[] | null;
    repositoryUrl?: string | null;
    orderDirection?: ("ASC" | "DESC") | null;
    startDate?: string;
    endDate?: string;
    extension?: "csv" | "xlsx";
    authorization: string;
}, opts?: Oazapfts.RequestOpts) {
    return oazapfts.ok(oazapfts.fetchJson<{
        status: 200;
        data: any;
    } | {
        status: 400;
        data: HttpErrorResponseRead;
    } | {
        status: 401;
        data: HttpErrorResponseRead;
    } | {
        status: 404;
        data: HttpErrorResponseRead;
    } | {
        status: 422;
    } | {
        status: 500;
        data: HttpErrorResponseRead;
    } | {
        status: 503;
        data: HttpErrorResponseRead;
    }>(`/v1/analytics/repositories/usage/download${QS.query(QS.explode({
        moduleId,
        user_id: userId,
        orderBy,
        tags,
        executedBy,
        repositoryUrl,
        orderDirection,
        startDate,
        endDate,
        extension
    }))}`, {
        ...opts,
        headers: oazapfts.mergeHeaders(opts?.headers, {
            authorization
        })
    }));
}
/**
 * Analytics User Usage
 */
export function analyticsUserUsageV1AnalyticsUsersUsageGet({ pageSize, page, moduleId, name, reportType, orderBy, orderDirection, startDate, endDate, authorization }: {
    pageSize?: number;
    page?: number;
    moduleId?: string[] | null;
    name?: string | null;
    reportType?: ("repository" | "program_group") | null;
    orderBy?: ("name" | "email" | "totalScan" | "totalScanFailure" | "totalScanSuccess" | "totalFix" | "totalFixFailure" | "totalFixSuccess") | null;
    orderDirection?: ("ASC" | "DESC") | null;
    startDate?: string;
    endDate?: string;
    authorization: string;
}, opts?: Oazapfts.RequestOpts) {
    return oazapfts.ok(oazapfts.fetchJson<{
        status: 200;
        data: AnalyticsUserUsageResponse;
    } | {
        status: 400;
        data: HttpErrorResponseRead;
    } | {
        status: 401;
        data: HttpErrorResponseRead;
    } | {
        status: 404;
        data: HttpErrorResponseRead;
    } | {
        status: 422;
    } | {
        status: 500;
        data: HttpErrorResponseRead;
    } | {
        status: 503;
        data: HttpErrorResponseRead;
    }>(`/v1/analytics/users/usage${QS.query(QS.explode({
        pageSize,
        page,
        moduleId,
        name,
        report_type: reportType,
        orderBy,
        orderDirection,
        startDate,
        endDate
    }))}`, {
        ...opts,
        headers: oazapfts.mergeHeaders(opts?.headers, {
            authorization
        })
    }));
}
/**
 * Analytics User Usage Download
 */
export function analyticsUserUsageDownloadV1AnalyticsUsersUsageDownloadGet({ moduleId, name, reportType, orderBy, orderDirection, startDate, endDate, extension, authorization }: {
    moduleId?: string[] | null;
    name?: string | null;
    reportType?: ("repository" | "program_group") | null;
    orderBy?: ("name" | "email" | "totalScan" | "totalScanFailure" | "totalScanSuccess" | "totalFix" | "totalFixFailure" | "totalFixSuccess") | null;
    orderDirection?: ("ASC" | "DESC") | null;
    startDate?: string;
    endDate?: string;
    extension?: "csv" | "xlsx";
    authorization: string;
}, opts?: Oazapfts.RequestOpts) {
    return oazapfts.ok(oazapfts.fetchJson<{
        status: 200;
        data: any;
    } | {
        status: 400;
        data: HttpErrorResponseRead;
    } | {
        status: 401;
        data: HttpErrorResponseRead;
    } | {
        status: 404;
        data: HttpErrorResponseRead;
    } | {
        status: 422;
    } | {
        status: 500;
        data: HttpErrorResponseRead;
    } | {
        status: 503;
        data: HttpErrorResponseRead;
    }>(`/v1/analytics/users/usage/download${QS.query(QS.explode({
        moduleId,
        name,
        report_type: reportType,
        orderBy,
        orderDirection,
        startDate,
        endDate,
        extension
    }))}`, {
        ...opts,
        headers: oazapfts.mergeHeaders(opts?.headers, {
            authorization
        })
    }));
}
/**
 * Analytics Program Groups Last Report Status Download
 */
export function analyticsProgramGroupsLastReportStatusDownloadV1AnalyticsProgramGroupsLastReportStatusDownloadGet({ moduleId, orderBy, orderDirection, startDate, endDate, reportType, extension, authorization }: {
    moduleId?: string[] | null;
    orderBy?: ("programName" | "moduleName" | "mode" | "reportCreatedAt" | "reportCreatedBy" | "totalFiles" | "totalFilesAnalyzed" | "totalFilesIssues" | "totalFilesChanged" | "totalIssues") | null;
    orderDirection?: ("ASC" | "DESC") | null;
    startDate?: string | null;
    endDate?: string | null;
    reportType?: ("repository" | "program_group") | null;
    extension?: "csv" | "xlsx";
    authorization: string;
}, opts?: Oazapfts.RequestOpts) {
    return oazapfts.ok(oazapfts.fetchJson<{
        status: 200;
        data: any;
    } | {
        status: 400;
        data: HttpErrorResponseRead;
    } | {
        status: 401;
        data: HttpErrorResponseRead;
    } | {
        status: 404;
        data: HttpErrorResponseRead;
    } | {
        status: 422;
    } | {
        status: 500;
        data: HttpErrorResponseRead;
    } | {
        status: 503;
        data: HttpErrorResponseRead;
    }>(`/v1/analytics/program-groups/last-report-status/download${QS.query(QS.explode({
        moduleId,
        orderBy,
        orderDirection,
        startDate,
        endDate,
        report_type: reportType,
        extension
    }))}`, {
        ...opts,
        headers: oazapfts.mergeHeaders(opts?.headers, {
            authorization
        })
    }));
}
/**
 * Analytics Program Groups Last Report Status
 */
export function analyticsProgramGroupsLastReportStatusV1AnalyticsProgramGroupsLastReportStatusGet({ pageSize, page, moduleId, userId, tags, orderBy, executedBy, programName, orderDirection, startDate, endDate, authorization }: {
    pageSize?: number;
    page?: number;
    moduleId?: string[] | null;
    userId?: string | null;
    tags?: string[] | null;
    orderBy?: ("programName" | "moduleName" | "mode" | "reportCreatedAt" | "reportCreatedBy" | "totalFiles" | "totalFilesAnalyzed" | "totalFilesIssues" | "totalFilesChanged" | "totalIssues") | null;
    executedBy?: string[] | null;
    programName?: string | null;
    orderDirection?: ("ASC" | "DESC") | null;
    startDate?: string;
    endDate?: string;
    authorization: string;
}, opts?: Oazapfts.RequestOpts) {
    return oazapfts.ok(oazapfts.fetchJson<{
        status: 200;
        data: AnalyticsProgramGroupsLastReportStatusResponse;
    } | {
        status: 400;
        data: HttpErrorResponseRead;
    } | {
        status: 401;
        data: HttpErrorResponseRead;
    } | {
        status: 404;
        data: HttpErrorResponseRead;
    } | {
        status: 422;
    } | {
        status: 500;
        data: HttpErrorResponseRead;
    } | {
        status: 503;
        data: HttpErrorResponseRead;
    }>(`/v1/analytics/program-groups/last-report-status${QS.query(QS.explode({
        pageSize,
        page,
        moduleId,
        user_id: userId,
        tags,
        orderBy,
        executedBy,
        programName,
        orderDirection,
        startDate,
        endDate
    }))}`, {
        ...opts,
        headers: oazapfts.mergeHeaders(opts?.headers, {
            authorization
        })
    }));
}
/**
 * Analytics Program Groups Usage
 */
export function analyticsProgramGroupsUsageV1AnalyticsProgramGroupsUsageGet({ pageSize, page, executedBy, moduleId, tags, orderBy, orderDirection, programGroupName, startDate, endDate, authorization }: {
    pageSize?: number;
    page?: number;
    executedBy?: string[] | null;
    moduleId?: string[] | null;
    tags?: string[] | null;
    orderBy?: ("name" | "totalScan" | "totalScanFailure" | "totalScanSuccess" | "totalFix" | "totalFixFailure" | "totalFixSuccess") | null;
    orderDirection?: ("ASC" | "DESC") | null;
    programGroupName?: string | null;
    startDate?: string | null;
    endDate?: string | null;
    authorization: string;
}, opts?: Oazapfts.RequestOpts) {
    return oazapfts.ok(oazapfts.fetchJson<{
        status: 200;
        data: AnalyticsProgramGroupsUsageResponse;
    } | {
        status: 400;
        data: HttpErrorResponseRead;
    } | {
        status: 401;
        data: HttpErrorResponseRead;
    } | {
        status: 404;
        data: HttpErrorResponseRead;
    } | {
        status: 422;
    } | {
        status: 500;
        data: HttpErrorResponseRead;
    } | {
        status: 503;
        data: HttpErrorResponseRead;
    }>(`/v1/analytics/program-groups/usage${QS.query(QS.explode({
        pageSize,
        page,
        executedBy,
        moduleId,
        tags,
        orderBy,
        orderDirection,
        programGroupName,
        startDate,
        endDate
    }))}`, {
        ...opts,
        headers: oazapfts.mergeHeaders(opts?.headers, {
            authorization
        })
    }));
}
/**
 * Analytics Program Groups Usage Download
 */
export function analyticsProgramGroupsUsageDownloadV1AnalyticsProgramGroupsUsageDownloadGet({ moduleId, executedBy, tags, orderBy, orderDirection, programGroupName, startDate, endDate, extension, authorization }: {
    moduleId?: string[] | null;
    executedBy?: string[] | null;
    tags?: string[] | null;
    orderBy?: ("name" | "totalScan" | "totalScanFailure" | "totalScanSuccess" | "totalFix" | "totalFixFailure" | "totalFixSuccess") | null;
    orderDirection?: ("ASC" | "DESC") | null;
    programGroupName?: string | null;
    startDate?: string | null;
    endDate?: string | null;
    extension?: "csv" | "xlsx";
    authorization: string;
}, opts?: Oazapfts.RequestOpts) {
    return oazapfts.ok(oazapfts.fetchJson<{
        status: 200;
        data: any;
    } | {
        status: 400;
        data: HttpErrorResponseRead;
    } | {
        status: 401;
        data: HttpErrorResponseRead;
    } | {
        status: 404;
        data: HttpErrorResponseRead;
    } | {
        status: 422;
    } | {
        status: 500;
        data: HttpErrorResponseRead;
    } | {
        status: 503;
        data: HttpErrorResponseRead;
    }>(`/v1/analytics/program-groups/usage/download${QS.query(QS.explode({
        moduleId,
        executedBy,
        tags,
        orderBy,
        orderDirection,
        programGroupName,
        startDate,
        endDate,
        extension
    }))}`, {
        ...opts,
        headers: oazapfts.mergeHeaders(opts?.headers, {
            authorization
        })
    }));
}
/**
 * Analytics Program Groups Details
 */
export function analyticsProgramGroupsDetailsV1AnalyticsProgramGroupsDetailsGet({ pageSize, page, programName, tags, mode, moduleId, programCreatedById, reportCreatedById, startDate, endDate, orderBy, orderDirection, authorization }: {
    pageSize?: number;
    page?: number;
    programName?: string | null;
    tags?: string[] | null;
    mode?: ("scan" | "fix") | null;
    moduleId?: string[] | null;
    programCreatedById?: string[] | null;
    reportCreatedById?: string[] | null;
    startDate?: string;
    endDate?: string;
    orderBy?: ("programGroupName" | "moduleName" | "mode" | "reportCreatedAt" | "reportCreatedByName" | "executionStartedAt" | "executionCompletedAt" | "totalFilesCount" | "analyzedFilesCount" | "issuesFilesCount" | "changedFilesCount" | "issuesCount" | "processingTimeSeconds") | null;
    orderDirection?: ("ASC" | "DESC") | null;
    authorization: string;
}, opts?: Oazapfts.RequestOpts) {
    return oazapfts.ok(oazapfts.fetchJson<{
        status: 200;
        data: AnalyticsProgramGroupsDetailsResponse;
    } | {
        status: 400;
        data: HttpErrorResponseRead;
    } | {
        status: 401;
        data: HttpErrorResponseRead;
    } | {
        status: 404;
        data: HttpErrorResponseRead;
    } | {
        status: 422;
    } | {
        status: 500;
        data: HttpErrorResponseRead;
    } | {
        status: 503;
        data: HttpErrorResponseRead;
    }>(`/v1/analytics/program-groups/details${QS.query(QS.explode({
        pageSize,
        page,
        programName,
        tags,
        mode,
        moduleId,
        programCreatedById,
        reportCreatedById,
        startDate,
        endDate,
        orderBy,
        orderDirection
    }))}`, {
        ...opts,
        headers: oazapfts.mergeHeaders(opts?.headers, {
            authorization
        })
    }));
}
/**
 * Analytics Program Groups Details Download
 */
export function analyticsProgramGroupsDetailsDownloadV1AnalyticsProgramGroupsDetailsDownloadGet({ programName, tags, mode, moduleId, programCreatedById, reportCreatedById, startDate, endDate, orderBy, orderDirection, extension, authorization }: {
    programName?: string | null;
    tags?: string[] | null;
    mode?: ("scan" | "fix") | null;
    moduleId?: string[] | null;
    programCreatedById?: string[] | null;
    reportCreatedById?: string[] | null;
    startDate?: string;
    endDate?: string;
    orderBy?: ("programGroupName" | "moduleName" | "mode" | "reportCreatedAt" | "reportCreatedByName" | "executionStartedAt" | "executionCompletedAt" | "totalFilesCount" | "analyzedFilesCount" | "issuesFilesCount" | "changedFilesCount" | "issuesCount" | "processingTimeSeconds") | null;
    orderDirection?: ("ASC" | "DESC") | null;
    extension?: "csv" | "xlsx";
    authorization: string;
}, opts?: Oazapfts.RequestOpts) {
    return oazapfts.ok(oazapfts.fetchJson<{
        status: 200;
        data: any;
    } | {
        status: 400;
        data: HttpErrorResponseRead;
    } | {
        status: 401;
        data: HttpErrorResponseRead;
    } | {
        status: 404;
        data: HttpErrorResponseRead;
    } | {
        status: 422;
    } | {
        status: 500;
        data: HttpErrorResponseRead;
    } | {
        status: 503;
        data: HttpErrorResponseRead;
    }>(`/v1/analytics/program-groups/details/download${QS.query(QS.explode({
        programName,
        tags,
        mode,
        moduleId,
        programCreatedById,
        reportCreatedById,
        startDate,
        endDate,
        orderBy,
        orderDirection,
        extension
    }))}`, {
        ...opts,
        headers: oazapfts.mergeHeaders(opts?.headers, {
            authorization
        })
    }));
}
/**
 * Analytics Repository Detailed Report
 */
export function analyticsRepositoryDetailedReportV1AnalyticsRepositoriesDetailsGet({ pageSize, page, repositoryUrl, tags, mode, moduleId, repositoryCreatedById, executedBy, startDate, endDate, orderBy, orderDirection, authorization }: {
    pageSize?: number;
    page?: number;
    repositoryUrl?: string | null;
    tags?: string[] | null;
    mode?: ("scan" | "fix") | null;
    moduleId?: string[] | null;
    repositoryCreatedById?: string[] | null;
    executedBy?: string[] | null;
    startDate?: string;
    endDate?: string;
    orderBy?: ("totalFilesCount" | "analyzedFilesCount" | "changedFilesCount" | "issuesCount" | "issuesFilesCount" | "reportCreatedAt" | "repositoryUrl" | "sourceBranch" | "targetBranch" | "executionCompletedAt" | "executionStartedAt" | "moduleName" | "moduleVersion" | "processingTimeSeconds" | "pullRequestLink" | "reportCreatedByEmail" | "reportCreatedByName" | "repositoryCreatedByEmail" | "repositoryCreatedByName") | null;
    orderDirection?: ("ASC" | "DESC") | null;
    authorization: string;
}, opts?: Oazapfts.RequestOpts) {
    return oazapfts.ok(oazapfts.fetchJson<{
        status: 200;
        data: AnalyticsRepositoryDetailedReportResponse;
    } | {
        status: 400;
        data: HttpErrorResponseRead;
    } | {
        status: 401;
        data: HttpErrorResponseRead;
    } | {
        status: 404;
        data: HttpErrorResponseRead;
    } | {
        status: 422;
    } | {
        status: 500;
        data: HttpErrorResponseRead;
    } | {
        status: 503;
        data: HttpErrorResponseRead;
    }>(`/v1/analytics/repositories/details${QS.query(QS.explode({
        pageSize,
        page,
        repositoryUrl,
        tags,
        mode,
        moduleId,
        repositoryCreatedById,
        executedBy,
        startDate,
        endDate,
        orderBy,
        orderDirection
    }))}`, {
        ...opts,
        headers: oazapfts.mergeHeaders(opts?.headers, {
            authorization
        })
    }));
}
/**
 * Analytics Repository Detailed Report Download
 */
export function analyticsRepositoryDetailedReportDownloadV1AnalyticsRepositoriesDetailsDownloadGet({ repositoryUrl, tags, mode, moduleId, repositoryCreatedById, reportCreatedById, startDate, endDate, orderBy, orderDirection, extension, authorization }: {
    repositoryUrl?: string | null;
    tags?: string[] | null;
    mode?: ("scan" | "fix") | null;
    moduleId?: string[] | null;
    repositoryCreatedById?: string[] | null;
    reportCreatedById?: string[] | null;
    startDate?: string;
    endDate?: string;
    orderBy?: ("totalFilesCount" | "analyzedFilesCount" | "changedFilesCount" | "issuesCount" | "issuesFilesCount" | "reportCreatedAt" | "repositoryUrl" | "sourceBranch" | "targetBranch" | "executionCompletedAt" | "executionStartedAt" | "moduleName" | "moduleVersion" | "processingTimeSeconds" | "pullRequestLink" | "reportCreatedByEmail" | "reportCreatedByName" | "repositoryCreatedByEmail" | "repositoryCreatedByName") | null;
    orderDirection?: ("ASC" | "DESC") | null;
    extension?: "csv" | "xlsx";
    authorization: string;
}, opts?: Oazapfts.RequestOpts) {
    return oazapfts.ok(oazapfts.fetchJson<{
        status: 200;
        data: AnalyticsRepositoryDetailedReportResponse;
    } | {
        status: 400;
        data: HttpErrorResponseRead;
    } | {
        status: 401;
        data: HttpErrorResponseRead;
    } | {
        status: 404;
        data: HttpErrorResponseRead;
    } | {
        status: 422;
    } | {
        status: 500;
        data: HttpErrorResponseRead;
    } | {
        status: 503;
        data: HttpErrorResponseRead;
    }>(`/v1/analytics/repositories/details/download${QS.query(QS.explode({
        repositoryUrl,
        tags,
        mode,
        moduleId,
        repositoryCreatedById,
        reportCreatedById,
        startDate,
        endDate,
        orderBy,
        orderDirection,
        extension
    }))}`, {
        ...opts,
        headers: oazapfts.mergeHeaders(opts?.headers, {
            authorization
        })
    }));
}
/**
 * Analytics Program Groups Target Details
 */
export function analyticsProgramGroupsTargetDetailsV1AnalyticsProgramGroupsTargetDetailsGet({ pageSize, page, programName, componentName, tags, mode, moduleId, reportCreatedById, startDate, endDate, orderBy, orderDirection, authorization }: {
    pageSize?: number;
    page?: number;
    programName?: string | null;
    componentName?: string | null;
    tags?: string[] | null;
    mode?: ("scan" | "fix") | null;
    moduleId?: string[] | null;
    reportCreatedById?: string[] | null;
    startDate?: string;
    endDate?: string;
    orderBy?: ("programGroupName" | "moduleName" | "mode" | "reportCreatedAt" | "executionStartedAt" | "executionCompletedAt" | "componentAppName" | "componentName" | "componentType" | "totalIssues" | "changedFilesCount") | null;
    orderDirection?: ("ASC" | "DESC") | null;
    authorization: string;
}, opts?: Oazapfts.RequestOpts) {
    return oazapfts.ok(oazapfts.fetchJson<{
        status: 200;
        data: AnalyticsProgramGroupsTargetDetailsResponse;
    } | {
        status: 400;
        data: HttpErrorResponseRead;
    } | {
        status: 401;
        data: HttpErrorResponseRead;
    } | {
        status: 404;
        data: HttpErrorResponseRead;
    } | {
        status: 422;
    } | {
        status: 500;
        data: HttpErrorResponseRead;
    } | {
        status: 503;
        data: HttpErrorResponseRead;
    }>(`/v1/analytics/program-groups/target-details${QS.query(QS.explode({
        pageSize,
        page,
        programName,
        componentName,
        tags,
        mode,
        moduleId,
        reportCreatedById,
        startDate,
        endDate,
        orderBy,
        orderDirection
    }))}`, {
        ...opts,
        headers: oazapfts.mergeHeaders(opts?.headers, {
            authorization
        })
    }));
}
/**
 * Analytics Program Groups Target Details Download
 */
export function analyticsProgramGroupsTargetDetailsDownloadV1AnalyticsProgramGroupsTargetDetailsDownloadGet({ programName, tags, mode, moduleId, componentName, reportCreatedById, startDate, endDate, orderBy, orderDirection, extension, authorization }: {
    programName?: string | null;
    tags?: string[] | null;
    mode?: ("scan" | "fix") | null;
    moduleId?: string[] | null;
    componentName?: string | null;
    reportCreatedById?: string[] | null;
    startDate?: string;
    endDate?: string;
    orderBy?: ("programGroupName" | "moduleName" | "mode" | "reportCreatedAt" | "executionStartedAt" | "executionCompletedAt" | "componentAppName" | "componentName" | "componentType" | "totalIssues" | "changedFilesCount") | null;
    orderDirection?: ("ASC" | "DESC") | null;
    extension?: "csv" | "xlsx";
    authorization: string;
}, opts?: Oazapfts.RequestOpts) {
    return oazapfts.ok(oazapfts.fetchJson<{
        status: 200;
        data: any;
    } | {
        status: 400;
        data: HttpErrorResponseRead;
    } | {
        status: 401;
        data: HttpErrorResponseRead;
    } | {
        status: 404;
        data: HttpErrorResponseRead;
    } | {
        status: 422;
    } | {
        status: 500;
        data: HttpErrorResponseRead;
    } | {
        status: 503;
        data: HttpErrorResponseRead;
    }>(`/v1/analytics/program-groups/target-details/download${QS.query(QS.explode({
        programName,
        tags,
        mode,
        moduleId,
        componentName,
        reportCreatedById,
        startDate,
        endDate,
        orderBy,
        orderDirection,
        extension
    }))}`, {
        ...opts,
        headers: oazapfts.mergeHeaders(opts?.headers, {
            authorization
        })
    }));
}
/**
 * Analytics Repository Target Details
 */
export function analyticsRepositoryTargetDetailsV1AnalyticsRepositoriesTargetDetailsGet({ pageSize, page, repositoryUrl, targetName, tags, mode, moduleId, reportCreatedById, startDate, endDate, orderBy, orderDirection, authorization }: {
    pageSize?: number;
    page?: number;
    repositoryUrl?: string | null;
    targetName?: string | null;
    tags?: string[] | null;
    mode?: ("scan" | "fix") | null;
    moduleId?: string[] | null;
    reportCreatedById?: string[] | null;
    startDate?: string;
    endDate?: string;
    orderBy?: ("repositoryUrl" | "sourceBranch" | "moduleName" | "targetName" | "targetPath" | "issuesCount" | "reportCreatedAt") | null;
    orderDirection?: ("ASC" | "DESC") | null;
    authorization: string;
}, opts?: Oazapfts.RequestOpts) {
    return oazapfts.ok(oazapfts.fetchJson<{
        status: 200;
        data: AnalyticsRepositoryTargetDetailsResponse;
    } | {
        status: 400;
        data: HttpErrorResponseRead;
    } | {
        status: 401;
        data: HttpErrorResponseRead;
    } | {
        status: 404;
        data: HttpErrorResponseRead;
    } | {
        status: 422;
    } | {
        status: 500;
        data: HttpErrorResponseRead;
    } | {
        status: 503;
        data: HttpErrorResponseRead;
    }>(`/v1/analytics/repositories/target-details${QS.query(QS.explode({
        pageSize,
        page,
        repositoryUrl,
        targetName,
        tags,
        mode,
        moduleId,
        reportCreatedById,
        startDate,
        endDate,
        orderBy,
        orderDirection
    }))}`, {
        ...opts,
        headers: oazapfts.mergeHeaders(opts?.headers, {
            authorization
        })
    }));
}
/**
 * Analytics Repository Target Details Download
 */
export function analyticsRepositoryTargetDetailsDownloadV1AnalyticsRepositoriesTargetDetailsDownloadGet({ repositoryUrl, targetName, tags, mode, moduleId, reportCreatedById, startDate, endDate, orderBy, orderDirection, extension, authorization }: {
    repositoryUrl?: string | null;
    targetName?: string | null;
    tags?: string[] | null;
    mode?: ("scan" | "fix") | null;
    moduleId?: string[] | null;
    reportCreatedById?: string[] | null;
    startDate?: string;
    endDate?: string;
    orderBy?: ("repositoryUrl" | "sourceBranch" | "moduleName" | "targetName" | "targetPath" | "issuesCount" | "reportCreatedAt") | null;
    orderDirection?: ("ASC" | "DESC") | null;
    extension?: "csv" | "xlsx";
    authorization: string;
}, opts?: Oazapfts.RequestOpts) {
    return oazapfts.ok(oazapfts.fetchJson<{
        status: 200;
        data: any;
    } | {
        status: 400;
        data: HttpErrorResponseRead;
    } | {
        status: 401;
        data: HttpErrorResponseRead;
    } | {
        status: 404;
        data: HttpErrorResponseRead;
    } | {
        status: 422;
    } | {
        status: 500;
        data: HttpErrorResponseRead;
    } | {
        status: 503;
        data: HttpErrorResponseRead;
    }>(`/v1/analytics/repositories/target-details/download${QS.query(QS.explode({
        repositoryUrl,
        targetName,
        tags,
        mode,
        moduleId,
        reportCreatedById,
        startDate,
        endDate,
        orderBy,
        orderDirection,
        extension
    }))}`, {
        ...opts,
        headers: oazapfts.mergeHeaders(opts?.headers, {
            authorization
        })
    }));
}
/**
 * Analytics Module Execution Times Report
 */
export function analyticsModuleExecutionTimesReportV1AnalyticsModulesExecutionTimesGet({ pageSize, page, moduleId, version, mode, startDate, endDate, orderBy, orderDirection, authorization }: {
    pageSize?: number;
    page?: number;
    moduleId?: string[] | null;
    version?: string[] | null;
    mode?: ("scan" | "fix") | null;
    startDate?: string;
    endDate?: string;
    orderBy?: ("moduleName" | "moduleVersion" | "queueTimeAvg" | "queueTimeP95" | "executionTimeAvg" | "executionTimeP95" | "totalTimeAvg" | "totalTimeP95" | "executionCount") | null;
    orderDirection?: ("ASC" | "DESC") | null;
    authorization: string;
}, opts?: Oazapfts.RequestOpts) {
    return oazapfts.ok(oazapfts.fetchJson<{
        status: 200;
        data: AnalyticsModuleExecutionTimesResponse;
    } | {
        status: 400;
        data: HttpErrorResponseRead;
    } | {
        status: 401;
        data: HttpErrorResponseRead;
    } | {
        status: 404;
        data: HttpErrorResponseRead;
    } | {
        status: 422;
    } | {
        status: 500;
        data: HttpErrorResponseRead;
    } | {
        status: 503;
        data: HttpErrorResponseRead;
    }>(`/v1/analytics/modules/execution-times${QS.query(QS.explode({
        pageSize,
        page,
        moduleId,
        version,
        mode,
        startDate,
        endDate,
        orderBy,
        orderDirection
    }))}`, {
        ...opts,
        headers: oazapfts.mergeHeaders(opts?.headers, {
            authorization
        })
    }));
}
/**
 * Analytics Module Execution Times Download
 */
export function analyticsModuleExecutionTimesDownloadV1AnalyticsModulesExecutionTimesDownloadGet({ moduleId, version, mode, startDate, endDate, orderBy, orderDirection, extension, authorization }: {
    moduleId?: string[] | null;
    version?: string[] | null;
    mode?: ("scan" | "fix") | null;
    startDate?: string;
    endDate?: string;
    orderBy?: ("moduleName" | "moduleVersion" | "queueTimeAvg" | "queueTimeP95" | "executionTimeAvg" | "executionTimeP95" | "totalTimeAvg" | "totalTimeP95" | "executionCount") | null;
    orderDirection?: ("ASC" | "DESC") | null;
    extension?: "csv" | "xlsx";
    authorization: string;
}, opts?: Oazapfts.RequestOpts) {
    return oazapfts.ok(oazapfts.fetchJson<{
        status: 200;
        data: any;
    } | {
        status: 400;
        data: HttpErrorResponseRead;
    } | {
        status: 401;
        data: HttpErrorResponseRead;
    } | {
        status: 404;
        data: HttpErrorResponseRead;
    } | {
        status: 422;
    } | {
        status: 500;
        data: HttpErrorResponseRead;
    } | {
        status: 503;
        data: HttpErrorResponseRead;
    }>(`/v1/analytics/modules/execution-times/download${QS.query(QS.explode({
        moduleId,
        version,
        mode,
        startDate,
        endDate,
        orderBy,
        orderDirection,
        extension
    }))}`, {
        ...opts,
        headers: oazapfts.mergeHeaders(opts?.headers, {
            authorization
        })
    }));
}
/**
 * List User Service
 */
export function listUserServiceV1UsersGet({ authorization }: {
    authorization: string;
}, opts?: Oazapfts.RequestOpts) {
    return oazapfts.ok(oazapfts.fetchJson<{
        status: 200;
        data: ListUserResponse;
    } | {
        status: 422;
        data: HttpValidationError;
    }>("/v1/users", {
        ...opts,
        headers: oazapfts.mergeHeaders(opts?.headers, {
            authorization
        })
    }));
}
/**
 * Search Repos Scm Service
 */
export function searchReposScmServiceV2ReposSearchScmPost({ authorization, searchReposInScmRequest }: {
    authorization: string;
    searchReposInScmRequest: SearchReposInScmRequest;
}, opts?: Oazapfts.RequestOpts) {
    return oazapfts.ok(oazapfts.fetchJson<{
        status: 200;
        data: SearchReposInScmResponse;
    } | {
        status: 207;
        data: HttpErrorResponseRead;
    } | {
        status: 400;
        data: HttpErrorResponseRead;
    } | {
        status: 401;
        data: HttpErrorResponseRead;
    } | {
        status: 404;
        data: HttpErrorResponseRead;
    } | {
        status: 422;
    } | {
        status: 500;
        data: HttpErrorResponseRead;
    } | {
        status: 503;
        data: HttpErrorResponseRead;
    }>("/v2/repos/search-scm", oazapfts.json({
        ...opts,
        method: "POST",
        body: searchReposInScmRequest,
        headers: oazapfts.mergeHeaders(opts?.headers, {
            authorization
        })
    })));
}
/**
 * Import Repos With Tags Scm Service
 */
export function importReposWithTagsScmServiceV2ReposSearchScmSearchIdPost({ searchId, authorization, importReposScmWithTagsRequest }: {
    searchId: string;
    authorization: string;
    importReposScmWithTagsRequest: ImportReposScmWithTagsRequest;
}, opts?: Oazapfts.RequestOpts) {
    return oazapfts.ok(oazapfts.fetchJson<{
        status: 200;
        data: SaveSearchReposResponse;
    } | {
        status: 400;
        data: HttpErrorResponseRead;
    } | {
        status: 401;
        data: HttpErrorResponseRead;
    } | {
        status: 404;
        data: HttpErrorResponseRead;
    } | {
        status: 422;
    } | {
        status: 500;
        data: HttpErrorResponseRead;
    } | {
        status: 503;
        data: HttpErrorResponseRead;
    }>(`/v2/repos/search-scm/${encodeURIComponent(searchId)}`, oazapfts.json({
        ...opts,
        method: "POST",
        body: importReposScmWithTagsRequest,
        headers: oazapfts.mergeHeaders(opts?.headers, {
            authorization
        })
    })));
}
/**
 * Search Repos Scm V2
 */
export function searchReposScmV2V2ReposSearchScmSearchIdGet({ searchId, authorization }: {
    searchId: string;
    authorization: string;
}, opts?: Oazapfts.RequestOpts) {
    return oazapfts.ok(oazapfts.fetchJson<{
        status: 200;
        data: ResponseSearchRepoByIdResponseRead;
    } | {
        status: 400;
        data: HttpErrorResponseRead;
    } | {
        status: 401;
        data: HttpErrorResponseRead;
    } | {
        status: 404;
        data: HttpErrorResponseRead;
    } | {
        status: 422;
    } | {
        status: 500;
        data: HttpErrorResponseRead;
    } | {
        status: 503;
        data: HttpErrorResponseRead;
    }>(`/v2/repos/search-scm/${encodeURIComponent(searchId)}`, {
        ...opts,
        headers: oazapfts.mergeHeaders(opts?.headers, {
            authorization
        })
    }));
}
/**
 * Download Search Repos Scm V2
 */
export function downloadSearchReposScmV2V2ReposSearchScmSearchIdDownloadGet({ searchId, extension, authorization }: {
    searchId: string;
    extension?: "csv" | "xlsx";
    authorization: string;
}, opts?: Oazapfts.RequestOpts) {
    return oazapfts.ok(oazapfts.fetchJson<{
        status: 200;
        data: any;
    } | {
        status: 400;
        data: HttpErrorResponseRead;
    } | {
        status: 401;
        data: HttpErrorResponseRead;
    } | {
        status: 404;
        data: HttpErrorResponseRead;
    } | {
        status: 422;
    } | {
        status: 500;
        data: HttpErrorResponseRead;
    } | {
        status: 503;
        data: HttpErrorResponseRead;
    }>(`/v2/repos/search-scm/${encodeURIComponent(searchId)}/download${QS.query(QS.explode({
        extension
    }))}`, {
        ...opts,
        headers: oazapfts.mergeHeaders(opts?.headers, {
            authorization
        })
    }));
}
/**
 * Create Repos Batch Service
 */
export function createReposBatchServiceV2ReposBatchPost({ tags, authorization, fastapiCompatV2BodyCreateReposBatchServiceV2ReposBatchPost }: {
    tags?: string[] | null;
    authorization: string;
    fastapiCompatV2BodyCreateReposBatchServiceV2ReposBatchPost: BodyCreateReposBatchServiceV2ReposBatchPost;
}, opts?: Oazapfts.RequestOpts) {
    return oazapfts.ok(oazapfts.fetchJson<{
        status: 200;
        data: CreateRepoBatchResponse;
    } | {
        status: 400;
        data: HttpErrorResponseRead;
    } | {
        status: 401;
        data: HttpErrorResponseRead;
    } | {
        status: 404;
        data: HttpErrorResponseRead;
    } | {
        status: 422;
    } | {
        status: 500;
        data: HttpErrorResponseRead;
    } | {
        status: 503;
        data: HttpErrorResponseRead;
    }>(`/v2/repos/batch${QS.query(QS.explode({
        tags
    }))}`, oazapfts.multipart({
        ...opts,
        method: "POST",
        body: fastapiCompatV2BodyCreateReposBatchServiceV2ReposBatchPost,
        headers: oazapfts.mergeHeaders(opts?.headers, {
            authorization
        })
    })));
}
/**
 * Get Import Result
 */
export function getImportResultV2ReposBatchImportIdGet({ importId, pageSize, page, authorization }: {
    importId: string;
    pageSize?: number;
    page?: number;
    authorization: string;
}, opts?: Oazapfts.RequestOpts) {
    return oazapfts.ok(oazapfts.fetchJson<{
        status: 200;
        data: GetImportResultResponseRead;
    } | {
        status: 400;
        data: HttpErrorResponseRead;
    } | {
        status: 401;
        data: HttpErrorResponseRead;
    } | {
        status: 404;
        data: HttpErrorResponseRead;
    } | {
        status: 422;
    } | {
        status: 500;
        data: HttpErrorResponseRead;
    } | {
        status: 503;
        data: HttpErrorResponseRead;
    }>(`/v2/repos/batch/${encodeURIComponent(importId)}${QS.query(QS.explode({
        pageSize,
        page
    }))}`, {
        ...opts,
        headers: oazapfts.mergeHeaders(opts?.headers, {
            authorization
        })
    }));
}
/**
 * Download Import Result
 */
export function downloadImportResultV2ReposBatchImportIdDownloadGet({ importId, language, status, extension, authorization }: {
    importId: string;
    language?: "ptBr" | "enUs";
    status?: ValidateRepositoryStatus | null;
    extension?: "csv" | "xlsx";
    authorization: string;
}, opts?: Oazapfts.RequestOpts) {
    return oazapfts.ok(oazapfts.fetchJson<{
        status: 200;
        data: any;
    } | {
        status: 400;
        data: HttpErrorResponseRead;
    } | {
        status: 401;
        data: HttpErrorResponseRead;
    } | {
        status: 404;
        data: HttpErrorResponseRead;
    } | {
        status: 422;
    } | {
        status: 500;
        data: HttpErrorResponseRead;
    } | {
        status: 503;
        data: HttpErrorResponseRead;
    }>(`/v2/repos/batch/${encodeURIComponent(importId)}/download${QS.query(QS.explode({
        language,
        status,
        extension
    }))}`, {
        ...opts,
        headers: oazapfts.mergeHeaders(opts?.headers, {
            authorization
        })
    }));
}
/**
 * List Repository Service
 */
export function listRepositoryServiceV2ReposGet({ pageSize, page, lastEvaluatedKey, filter, moduleId, createdByEmail, reportByEmail, orderBy, orderDirection, tags, onlyIds, authorization }: {
    pageSize?: number;
    page?: number;
    lastEvaluatedKey?: string;
    filter?: string;
    moduleId?: string;
    createdByEmail?: string[] | null;
    reportByEmail?: string[] | null;
    orderBy?: "repository_url" | "repository_default_branch" | "report_created_at" | "report_issues_count";
    orderDirection?: "asc" | "desc" | "ASC" | "DESC";
    tags?: string[] | null;
    onlyIds?: boolean;
    authorization: string;
}, opts?: Oazapfts.RequestOpts) {
    return oazapfts.ok(oazapfts.fetchJson<{
        status: 200;
        data: ListRepositoryResponse2;
    } | {
        status: 400;
        data: HttpErrorResponseRead;
    } | {
        status: 401;
        data: HttpErrorResponseRead;
    } | {
        status: 404;
        data: HttpErrorResponseRead;
    } | {
        status: 422;
    } | {
        status: 500;
        data: HttpErrorResponseRead;
    } | {
        status: 503;
        data: HttpErrorResponseRead;
    }>(`/v2/repos${QS.query(QS.explode({
        pageSize,
        page,
        lastEvaluatedKey,
        filter,
        moduleId,
        createdByEmail,
        reportByEmail,
        orderBy,
        orderDirection,
        tags,
        onlyIds
    }))}`, {
        ...opts,
        headers: oazapfts.mergeHeaders(opts?.headers, {
            authorization
        })
    }));
}
/**
 * Create Repository Service
 */
export function createRepositoryServiceV2ReposPost({ authorization, srcPresentationApiRestServicesV2ReposRequestsCreateRepositoryRequestCreateRepositoryRequest }: {
    authorization: string;
    srcPresentationApiRestServicesV2ReposRequestsCreateRepositoryRequestCreateRepositoryRequest: CreateRepositoryRequest2;
}, opts?: Oazapfts.RequestOpts) {
    return oazapfts.ok(oazapfts.fetchJson<{
        status: 200;
        data: CreateRepositoryResponse;
    } | {
        status: 400;
        data: HttpErrorResponseRead;
    } | {
        status: 401;
        data: HttpErrorResponseRead;
    } | {
        status: 404;
        data: HttpErrorResponseRead;
    } | {
        status: 422;
    } | {
        status: 500;
        data: HttpErrorResponseRead;
    } | {
        status: 503;
        data: HttpErrorResponseRead;
    }>("/v2/repos", oazapfts.json({
        ...opts,
        method: "POST",
        body: srcPresentationApiRestServicesV2ReposRequestsCreateRepositoryRequestCreateRepositoryRequest,
        headers: oazapfts.mergeHeaders(opts?.headers, {
            authorization
        })
    })));
}
/**
 * Pat Health Check
 */
export function patHealthCheckV1ScmPatHealthCheckGet({ authorization }: {
    authorization: string;
}, opts?: Oazapfts.RequestOpts) {
    return oazapfts.ok(oazapfts.fetchJson<{
        status: 200;
        data: boolean;
    } | {
        status: 422;
        data: HttpValidationError;
    }>("/v1/scm/pat/health-check", {
        ...opts,
        headers: oazapfts.mergeHeaders(opts?.headers, {
            authorization
        })
    }));
}
/**
 * List Module Execution Batch Activities
 */
export function listModuleExecutionBatchActivitiesV1ActivitiesModuleExecutionBatchGet({ pageSize, page, authorization }: {
    pageSize?: number;
    page?: number;
    authorization: string;
}, opts?: Oazapfts.RequestOpts) {
    return oazapfts.ok(oazapfts.fetchJson<{
        status: 200;
        data: ListModuleExecutionBatchActivitiesResponse;
    } | {
        status: 422;
        data: HttpValidationError;
    }>(`/v1/activities/module-execution-batch${QS.query(QS.explode({
        pageSize,
        page
    }))}`, {
        ...opts,
        headers: oazapfts.mergeHeaders(opts?.headers, {
            authorization
        })
    }));
}
/**
 * Get Module Execution Batch Activities By Id
 */
export function getModuleExecutionBatchActivitiesByIdV1ActivitiesModuleExecutionBatchModuleExecutionBatchIdGet({ moduleExecutionBatchId, authorization }: {
    moduleExecutionBatchId: string;
    authorization: string;
}, opts?: Oazapfts.RequestOpts) {
    return oazapfts.ok(oazapfts.fetchJson<{
        status: 200;
        data: GetModuleExecutionBatchActivitiesById;
    } | {
        status: 422;
        data: HttpValidationError;
    }>(`/v1/activities/module-execution-batch/${encodeURIComponent(moduleExecutionBatchId)}`, {
        ...opts,
        headers: oazapfts.mergeHeaders(opts?.headers, {
            authorization
        })
    }));
}
