// @stack-spot/portal-network/src/api/dataIntegration.ts
/**
 * FastAPI
 * 0.1.0
 * DO NOT MODIFY - This file has been generated using oazapfts.
 * See https://www.npmjs.com/package/oazapfts
 */
import * as Oazapfts from "@oazapfts/runtime";
import * as QS from "@oazapfts/runtime/query";
export const defaults: Oazapfts.Defaults<Oazapfts.CustomHeaders> = {
    headers: {},
    baseUrl: "/",
};
const oazapfts = Oazapfts.runtime(defaults);
export const servers = {};
export type SearchPreProcessing = "multi_query";
export type SearchPostProcessing = "reranker";
export type SearchStrategy = "cosine";
export type SearchMethod = "similarity";
export type SearchKnowledgeObjectRequest = {
    query: string;
    knowledge_source_ids: string[];
    limit?: number;
    threshold?: number | null;
    pre_processing?: SearchPreProcessing | null;
    post_processing?: SearchPostProcessing | null;
    strategy?: SearchStrategy;
    method?: SearchMethod;
    deep_search_ks?: boolean;
};
export type SearchKnowledgeObjectResponse = {
    id: string;
    score: number;
    display_content: string;
    prompt_content: string;
    file_path: string | null;
    ks_id: string;
    ks_name: string;
    ks_slug: string;
    ks_type: string;
};
export type ValidationError = {
    loc: (string | number)[];
    msg: string;
    "type": string;
};
export type HttpValidationError = {
    detail?: ValidationError[];
};
export type SearchCrossAccountKnowledgeObjectRequest = {
    query: string;
    limit?: number;
    threshold?: number | null;
    pre_processing?: SearchPreProcessing | null;
    post_processing?: SearchPostProcessing | null;
    strategy?: SearchStrategy;
    method?: SearchMethod;
};
export type KnowledgeSourceTypeEnum = "snippet" | "api" | "event" | "custom" | "project_file" | "memory";
export type NewKnowledgeSourceRequest = {
    slug: string;
    name: string;
    description: string;
    "type": KnowledgeSourceTypeEnum;
};
export type KnowledgeSourcesWorkspaceForkRequest = {
    ids: string[];
    member_id: string;
};
export type VisibilityLevelEnum = "account" | "personal" | "shared" | "workspace" | "favorite";
export type KnowledgeSourceResponse = {
    id: string;
    slug: string;
    name: string;
    description: string;
    "type": KnowledgeSourceTypeEnum;
    creator: string | null;
    "default": boolean;
    visibility_level: VisibilityLevelEnum;
    model_name: string;
    use_only: boolean;
};
export type KnowledgeSourcesWorkspaceForkResponse = {
    list_ks: KnowledgeSourceResponse[];
    ks_ids_account: string[];
    error_ids: string[];
};
export type KnowledgeSourcesWorkspaceDeleteRequest = {
    ids: string[];
};
export type KnowledgeSourcesWorkspaceListRequest = {
    ids: string[];
    name?: string | null;
};
export type KnowledgeSourceItemResponse = {
    id: string;
    slug: string;
    name: string;
    description: string;
    "type": KnowledgeSourceTypeEnum;
    creator: string | null;
    "default": boolean;
    visibility_level: VisibilityLevelEnum;
    object_count: number;
    use_only: boolean;
};
export type ExceptionType = "DATAINTEGRATION_1000_INVALID_JWT" | "DATAINTEGRATION_1001_DECODE_JWT_ERROR" | "DATAINTEGRATION_1002_KNOWLEDGE_SOURCE_DOES_NOT_EXISTS" | "DATAINTEGRATION_1003_KNOWLEDGE_SOURCE_WRONG_TYPE" | "DATAINTEGRATION_1004_UNSUPPORTED_KNOWLEDGE_SOURCE_TYPE_FOR_UPLOAD" | "DATAINTEGRATION_1005_KNOWLEDGE_OBJECT_NOT_FOUND" | "DATAINTEGRATION_1006_UNSUPPORTED_FILE_TYPE" | "DATAINTEGRATION_1007_KNOWLEDGE_SOURCE_SLUG_ALREADY_EXISTS" | "DATAINTEGRATION_1008_KNOWLEDGE_SOURCE_NOT_FOUND" | "DATAINTEGRATION_1009_KNOWLEDGE_SOURCE_CANNOT_BE_ASSOCIATED" | "DATAINTEGRATION_1010_KNOWLEDGE_SOURCE_NOT_ACCOUNT_VISIBLE" | "DATAINTEGRATION_1011_ACCOUNT_SETTINGS_NOT_FOUND" | "DATAINTEGRATION_1012_SECRET_NOT_FOUND" | "DATAINTEGRATION_1013_STRATEGY_NOT_FOUND" | "DATAINTEGRATION_1014_EMBEDDING_INTEGRATION_ERROR" | "DATAINTEGRATION_1015_KNOWLEDGE_OBJECT_UNEXPECTED_ERROR" | "DATAINTEGRATION_1016_FILE_UPLOAD_DOES_NOT_EXISTS" | "DATAINTEGRATION_1017_INVALID_VISIBILITY_CHANGE" | "DATAINTEGRATION_1018_TOKENS_USAGE_LIMIT_REACHED" | "DATAINTEGRATION_1019_KNOWLEDGE_SOURCE_INVALID_SPLIT_STRATEGY" | "DATAINTEGRATION_1020_FILE_UPLOAD_INVALID_SPLIT_STRATEGY" | "DATAINTEGRATION_1021_FILE_UPLOAD_INDEXED_OR_NOT_SPLITTED" | "DATAINTEGRATION_1022_KNOWLEDGE_GROUP_NOT_FOUND" | "DATAINTEGRATION_1023_FILE_UPLOAD_UNSUPPORTED_TARGET" | "DATAINTEGRATION_1024_FILE_UPLOAD_OPENAPI_SPEC_ERROR" | "DATAINTEGRATION_1025_FILE_UPLOAD_INVALID_SCRIPT_CONTENT" | "DATAINTEGRATION_1026_FILE_UPLOAD_EMPTY_SCRIPT_CONTENT" | "DATAINTEGRATION_1027_FILE_UPLOAD_INVALID_SCRIPT_EXECUTION_STATE_ID" | "DATAINTEGRATION_2000_LLM_INTEGRATION_ERROR" | "DATAINTEGRATION_2001_LLM_RATE_LIMIT_ERROR" | "DATAINTEGRATION_2002_LLM_PROMPT_TOO_LARGE_ERROR" | "DATAINTEGRATION_2003_LLM_FILE_TOO_LARGE_ERROR" | "DATAINTEGRATION_3000_FEATURE_DISABLED_ERROR" | "DATAINTEGRATION_5000_UNEXPECTED_ERROR" | "DATAINTEGRATION_6000_FORBIDDEN" | "DATAINTEGRATION_8001_ACCOUNT_API_ERROR" | "DATAINTEGRATION_4001_FAVORITE_ALREADY_EXISTS" | "DATAINTEGRATION_4002_FAVORITE_NOT_FOUND" | "DATAINTEGRATION_5001_DEFAULT_MODEL_NAME_NOT_FOUND" | "DATAINTEGRATION_9001_CONVERSATION_NOT_FOUND" | "DATAINTEGRATION_9002_INVALID_ID";
export type HttpErrorResponse = {
    "type"?: string;
    status: string;
    code: ExceptionType;
    details: string;
    value?: string | null;
    field?: string | null;
};
export type OrderEnum = "a-to-z" | "z-to-a" | "oldest-first" | "newest-first";
export type KnowledgeSourceSearchRequest = {
    ids: string[];
};
export type KnowledgeSourcePatchesRequest = {
    name: string;
    description: string;
    use_only?: boolean | null;
};
export type KnowledgeSourcesPublishRequest = {
    use_only?: boolean;
};
export type KnowledgeSourcePublishBatchRequest = {
    ids: string[];
};
export type ForkKnowledgeSourceRequest = {
    slug: string;
    name: string;
    description: string;
};
export type KnowledgeObjectResponse = {
    content: string;
    id: string;
    standalone: boolean;
    updated: string;
    checksum: string;
    file_path: string | null;
    programming_language: string | null;
    use_case: string | null;
    attributes: {
        [key: string]: any;
    } | null;
};
export type KnowledgeObjectChecksumResponse = {
    checksum: string;
};
export type KnowledgeSourceLastMessageResponse = {
    id: string;
    prompt_content: string;
    ks_id: string;
};
export type KnowledgeSourceFilesIngestedResponse = {
    id: string;
    path: string;
    checksum: string;
    status?: string;
    created_at: string;
    updated_at: string;
    count_kos: number;
};
export type DeleteIndexedFilesRequest = {
    paths: string[];
};
export type FileUploadType = "PROJECT" | "KNOWLEDGE_SOURCE" | "CONTEXT" | "TOOLKIT";
export type FileUploadIndexType = "RELEVANT" | "FULL_INDEX";
export type NewFileUploadRequest = {
    file_name: string;
    target_id?: string | null;
    target_type: FileUploadType;
    expiration?: number;
    index_type?: FileUploadIndexType | null;
};
export type NewFileUploadFormResponse = {
    id: string;
    url: string;
    form: {
        [key: string]: string;
    };
};
export type FileUploadStatus = "NEW" | "PROCESSING" | "INDEXED" | "SPLITTED" | "EXPIRED" | "ERROR" | "SPLIT_ERROR" | "INDEXED_BUT_CLEANUP_FAILED";
export type SummaryPartResponse = {
    added: number;
    preserved: number;
    removed: number;
    errors: {
        [key: string]: string;
    };
};
export type FileUploadStatusResponse = {
    id: string;
    file_name: string;
    target_id: string;
    target_type: FileUploadType;
    status: FileUploadStatus;
    error_description: string | null;
    summary: SummaryPartResponse | null;
};
export type FileSplitStrategy = "NONE" | "LINES_QUANTITY" | "TOKENS_QUANTITY" | "CHARACTERS_QUANTITY" | "SYNTACTIC" | "ENDPOINT";
export type FileUploadSaveChunksRequest = {
    split_strategy: FileSplitStrategy;
    split_overlap?: number | null;
    split_quantity?: number | null;
};
export type SyncResultDto = {
    added: number;
    preserved: number;
    removed: number;
    errors: {
        [key: string]: string;
    };
};
export type FileUploadUpdateRequest = {
    status: FileUploadStatus;
    summary?: SyncResultDto | null;
    error_description?: string | null;
};
export type FileSplitRequest = {
    split_overlap?: number | null;
    split_quantity?: number | null;
    split_strategy: FileSplitStrategy;
    embed_after_split?: boolean;
    max_split_preview?: number;
};
export type FileMetadataResponse = {
    id: string;
    file_path: string;
    bucket_name: string;
    target_slug: string;
    target_type: FileUploadType;
    status: FileUploadStatus;
    error_description: string | null;
    user_id_headers: string;
};
export type FileUploadException = {
    "type": ExceptionType;
    message: string;
    value?: string | null;
    field?: string | null;
};
export type PresignedUrlsDto = {
    presigned_urls: string[] | null;
    invalid: FileUploadException | null;
};
export type AccountEmbeddingTokenUsage = {
    current_used: number;
    max_embedding_tokens: number;
};
export type ConsumeTokenRequest = {
    token_quantity: number;
};
export type AccountLlmConfigResponse = {
    "type": string | null;
    embedding_max_chunk_size: number;
};
export type GetOrCreateKnowledgeGroupRequest = {
    knowledge_source_id: string;
    file_path: string;
};
export type KnowledgeGroupResponse = {
    id: string;
    knowledge_source_id: string;
    path: string;
    parent: string | null;
    checksum: string;
    created: string;
    updated: string;
    last_modified: string;
    count_kos?: number | null;
};
export type DeleteGroupObjectsRequests = {
    checksums_not_in: string[];
};
export type DeletedCountResponse = {
    count: number;
};
export type UpdateKnowledgeGroupRequest = {
    checksum: Blob;
    last_modified: string;
    parent_zip?: string | null;
};
export type CreateKnowledgeObjectsBatchRequest = {
    knowledge_source_id: string;
    prompt_content: string;
    checksum: string;
    embedding: number[];
    display_content: string;
    model_name: string;
    programming_language?: string | null;
    use_case?: string | null;
    knowledge_group_id?: string | null;
    attributes?: {
        [key: string]: any;
    } | null;
};
export type InternalEmbedTextResponse = {
    vector: number[];
};
export type MinimalConversationResponse = {
    id: string;
    title: string;
    created_at: string;
    updated_at?: string | null;
    agents?: {
        [key: string]: any;
    };
};
export type CreateConversationRequest = {
    title: string;
    conversation_id?: string | null;
    ai_stack_id?: string | null;
    workspace_id?: string | null;
    agents?: string[];
};
export type ConversationHistoryResponse = {
    id: string;
    conversation_id: string;
    user_input: string | null;
    ai_response: string | null;
    agent_id: string | null;
    tools: {
        [key: string]: any;
    } | null;
    agent_info: {
        [key: string]: any;
    } | null;
    llm_response: {
        [key: string]: any;
    } | null;
    updated: string | null;
};
export type ConversationResponse = {
    id: string;
    title: string;
    ai_stack_id: string | null;
    workspace_id: string | null;
    username: string | null;
    agents?: {
        [key: string]: any;
    };
    history?: ConversationHistoryResponse[];
};
export type ConversationUpdateTitleRequest = {
    title: string;
};
export type ConversationHistoryRequest = {
    agent_id: string;
    built_in: boolean;
    user_input: string;
    ai_response: string;
    tools?: {
        [key: string]: any;
    };
    agent_info?: {
        [key: string]: any;
    };
    llm_response?: {
        [key: string]: any;
    };
};
export type SearchConversationHistoryRequest = {
    query: string;
    conversation_ids: string[];
    limit?: number;
    threshold?: number | null;
    strategy?: SearchStrategy;
    method?: SearchMethod;
};
export type SearchConversationHistoryResponse = {
    id: string;
    score: number;
    conversation_id: string;
    user_input: string | null;
    ai_response: string | null;
    agent_id: string | null;
    tools: {
        [key: string]: any;
    } | null;
    agent_info: {
        [key: string]: any;
    } | null;
    created: string | null;
    updated: string | null;
};
export type PaginatedResponseKnowledgeSourceItemResponse = {
    total_pages: number;
    items: KnowledgeSourceItemResponse[];
};
export type PaginatedResponseKnowledgeObjectResponse = {
    total_pages: number;
    items: KnowledgeObjectResponse[];
};
export type UploadScriptRequest = {
    content: string;
};
export type UploadScriptResponse = {
    id: string;
};
export type GetScriptResponse = {
    id: string;
    content: string;
};
export type GetPresignedUrlResponse = {
    url: string;
};
export type CreateScriptExecutionRequest = {
    state: {
        [key: string]: any;
    };
};
export type CreateScriptExecutionResponse = {
    id: string;
};
/**
 * Health Check
 */
export function healthCheckHealthzGet(opts?: Oazapfts.RequestOpts) {
    return oazapfts.ok(oazapfts.fetchJson<{
        status: 200;
        data: any;
    } | {
        status: 404;
    }>("/healthz", {
        ...opts
    }));
}
/**
 * Readiness Check
 */
export function readinessCheckReadyzGet(opts?: Oazapfts.RequestOpts) {
    return oazapfts.ok(oazapfts.fetchJson<{
        status: 200;
        data: any;
    } | {
        status: 404;
    }>("/readyz", {
        ...opts
    }));
}
/**
 * Search Objects
 */
export function searchObjectsV1SearchPost({ xAccountId, xUsername, authorization, searchKnowledgeObjectRequest }: {
    xAccountId?: string | null;
    xUsername?: string | null;
    authorization: string;
    searchKnowledgeObjectRequest: SearchKnowledgeObjectRequest;
}, opts?: Oazapfts.RequestOpts) {
    return oazapfts.ok(oazapfts.fetchJson<{
        status: 200;
        data: SearchKnowledgeObjectResponse[];
    } | {
        status: 404;
    } | {
        status: 422;
        data: HttpValidationError;
    }>("/v1/search", oazapfts.json({
        ...opts,
        method: "POST",
        body: searchKnowledgeObjectRequest,
        headers: oazapfts.mergeHeaders(opts?.headers, {
            "x-account-id": xAccountId,
            "x-username": xUsername,
            authorization
        })
    })));
}
/**
 * Search Cross Account Objects
 */
export function searchCrossAccountObjectsV1SearchCrossAccountPost({ xAccountId, xUsername, authorization, searchCrossAccountKnowledgeObjectRequest }: {
    xAccountId: string;
    xUsername: string;
    authorization: string;
    searchCrossAccountKnowledgeObjectRequest: SearchCrossAccountKnowledgeObjectRequest;
}, opts?: Oazapfts.RequestOpts) {
    return oazapfts.ok(oazapfts.fetchJson<{
        status: 200;
        data: SearchKnowledgeObjectResponse[];
    } | {
        status: 404;
    } | {
        status: 422;
        data: HttpValidationError;
    }>("/v1/search/cross-account", oazapfts.json({
        ...opts,
        method: "POST",
        body: searchCrossAccountKnowledgeObjectRequest,
        headers: oazapfts.mergeHeaders(opts?.headers, {
            "x-account-id": xAccountId,
            "x-username": xUsername,
            authorization
        })
    })));
}
/**
 * Create Ft Knowledge Source
 */
export function createFtKnowledgeSourceV1FtKnowledgeSourcesPost({ newKnowledgeSourceRequest }: {
    newKnowledgeSourceRequest: NewKnowledgeSourceRequest;
}, opts?: Oazapfts.RequestOpts) {
    return oazapfts.ok(oazapfts.fetchJson<{
        status: 204;
    } | {
        status: 404;
    } | {
        status: 422;
        data: HttpValidationError;
    }>("/v1/ft-knowledge-sources", oazapfts.json({
        ...opts,
        method: "POST",
        body: newKnowledgeSourceRequest
    })));
}
/**
 * Delete Ft Knowledge Source
 */
export function deleteFtKnowledgeSourceV1FtKnowledgeSourcesSlugDelete({ slug }: {
    slug: string;
}, opts?: Oazapfts.RequestOpts) {
    return oazapfts.ok(oazapfts.fetchJson<{
        status: 204;
    } | {
        status: 404;
    } | {
        status: 422;
        data: HttpValidationError;
    }>(`/v1/ft-knowledge-sources/${encodeURIComponent(slug)}`, {
        ...opts,
        method: "DELETE"
    }));
}
/**
 * Workspace Fork
 */
export function workspaceForkV1KnowledgeSourcesWorkspaceForkPost({ xAccountId, xUsername, authorization, knowledgeSourcesWorkspaceForkRequest }: {
    xAccountId?: string | null;
    xUsername?: string | null;
    authorization: string;
    knowledgeSourcesWorkspaceForkRequest: KnowledgeSourcesWorkspaceForkRequest;
}, opts?: Oazapfts.RequestOpts) {
    return oazapfts.ok(oazapfts.fetchJson<{
        status: 200;
        data: KnowledgeSourcesWorkspaceForkResponse;
    } | {
        status: 404;
    } | {
        status: 422;
        data: HttpValidationError;
    }>("/v1/knowledge-sources/workspace/fork", oazapfts.json({
        ...opts,
        method: "POST",
        body: knowledgeSourcesWorkspaceForkRequest,
        headers: oazapfts.mergeHeaders(opts?.headers, {
            "x-account-id": xAccountId,
            "x-username": xUsername,
            authorization
        })
    })));
}
/**
 * Workspace Delete Fork
 */
export function workspaceDeleteForkV1KnowledgeSourcesWorkspaceForkDelete({ xAccountId, xUsername, authorization, knowledgeSourcesWorkspaceDeleteRequest }: {
    xAccountId?: string | null;
    xUsername?: string | null;
    authorization: string;
    knowledgeSourcesWorkspaceDeleteRequest: KnowledgeSourcesWorkspaceDeleteRequest;
}, opts?: Oazapfts.RequestOpts) {
    return oazapfts.ok(oazapfts.fetchJson<{
        status: 204;
    } | {
        status: 404;
    } | {
        status: 422;
        data: HttpValidationError;
    }>("/v1/knowledge-sources/workspace/fork", oazapfts.json({
        ...opts,
        method: "DELETE",
        body: knowledgeSourcesWorkspaceDeleteRequest,
        headers: oazapfts.mergeHeaders(opts?.headers, {
            "x-account-id": xAccountId,
            "x-username": xUsername,
            authorization
        })
    })));
}
/**
 * Workspace List
 */
export function workspaceListV1KnowledgeSourcesWorkspaceListPost({ xAccountId, xUsername, authorization, knowledgeSourcesWorkspaceListRequest }: {
    xAccountId?: string | null;
    xUsername?: string | null;
    authorization: string;
    knowledgeSourcesWorkspaceListRequest: KnowledgeSourcesWorkspaceListRequest;
}, opts?: Oazapfts.RequestOpts) {
    return oazapfts.ok(oazapfts.fetchJson<{
        status: 200;
        data: KnowledgeSourceItemResponse[];
    } | {
        status: 404;
    } | {
        status: 422;
        data: HttpValidationError;
    }>("/v1/knowledge-sources/workspace/list", oazapfts.json({
        ...opts,
        method: "POST",
        body: knowledgeSourcesWorkspaceListRequest,
        headers: oazapfts.mergeHeaders(opts?.headers, {
            "x-account-id": xAccountId,
            "x-username": xUsername,
            authorization
        })
    })));
}
/**
 * Create Ks
 */
export function createKsV1KnowledgeSourcesPost({ xAccountId, xUsername, authorization, newKnowledgeSourceRequest }: {
    xAccountId?: string | null;
    xUsername?: string | null;
    authorization: string;
    newKnowledgeSourceRequest: NewKnowledgeSourceRequest;
}, opts?: Oazapfts.RequestOpts) {
    return oazapfts.ok(oazapfts.fetchJson<{
        status: 201;
        data: any;
    } | {
        status: 404;
    } | {
        status: 422;
        data: HttpErrorResponse;
    } | {
        status: 500;
        data: HttpErrorResponse;
    }>("/v1/knowledge-sources", oazapfts.json({
        ...opts,
        method: "POST",
        body: newKnowledgeSourceRequest,
        headers: oazapfts.mergeHeaders(opts?.headers, {
            "x-account-id": xAccountId,
            "x-username": xUsername,
            authorization
        })
    })));
}
/**
 * List Ks
 */
export function listKsV1KnowledgeSourcesGet({ visibility, order, types, xAccountId, xUsername, authorization }: {
    visibility?: VisibilityLevelEnum;
    order?: OrderEnum;
    types?: KnowledgeSourceTypeEnum[];
    xAccountId?: string | null;
    xUsername?: string | null;
    authorization: string;
}, opts?: Oazapfts.RequestOpts) {
    return oazapfts.ok(oazapfts.fetchJson<{
        status: 200;
        data: KnowledgeSourceItemResponse[];
    } | {
        status: 404;
    } | {
        status: 422;
        data: HttpValidationError;
    }>(`/v1/knowledge-sources${QS.query(QS.explode({
        visibility,
        order,
        types
    }))}`, {
        ...opts,
        headers: oazapfts.mergeHeaders(opts?.headers, {
            "x-account-id": xAccountId,
            "x-username": xUsername,
            authorization
        })
    }));
}
/**
 * Search Ks
 */
export function searchKsV1KnowledgeSourcesSearchPost({ xAccountId, xUsername, authorization, knowledgeSourceSearchRequest }: {
    xAccountId?: string | null;
    xUsername?: string | null;
    authorization: string;
    knowledgeSourceSearchRequest: KnowledgeSourceSearchRequest;
}, opts?: Oazapfts.RequestOpts) {
    return oazapfts.ok(oazapfts.fetchJson<{
        status: 200;
        data: KnowledgeSourceResponse[];
    } | {
        status: 404;
    } | {
        status: 422;
        data: HttpValidationError;
    }>("/v1/knowledge-sources/search", oazapfts.json({
        ...opts,
        method: "POST",
        body: knowledgeSourceSearchRequest,
        headers: oazapfts.mergeHeaders(opts?.headers, {
            "x-account-id": xAccountId,
            "x-username": xUsername,
            authorization
        })
    })));
}
/**
 * Find Ks
 */
export function findKsV1KnowledgeSourcesSlugGet({ slug, xAccountId, xUsername, authorization }: {
    slug: string;
    xAccountId?: string | null;
    xUsername?: string | null;
    authorization: string;
}, opts?: Oazapfts.RequestOpts) {
    return oazapfts.ok(oazapfts.fetchJson<{
        status: 200;
        data: KnowledgeSourceResponse;
    } | {
        status: 404;
        data: HttpErrorResponse;
    } | {
        status: 422;
        data: HttpValidationError;
    }>(`/v1/knowledge-sources/${encodeURIComponent(slug)}`, {
        ...opts,
        headers: oazapfts.mergeHeaders(opts?.headers, {
            "x-account-id": xAccountId,
            "x-username": xUsername,
            authorization
        })
    }));
}
/**
 * Update Ks
 */
export function updateKsV1KnowledgeSourcesSlugPatch({ slug, xAccountId, xUsername, authorization, knowledgeSourcePatchesRequest }: {
    slug: string;
    xAccountId?: string | null;
    xUsername?: string | null;
    authorization: string;
    knowledgeSourcePatchesRequest: KnowledgeSourcePatchesRequest;
}, opts?: Oazapfts.RequestOpts) {
    return oazapfts.ok(oazapfts.fetchJson<{
        status: 204;
    } | {
        status: 403;
        data: HttpErrorResponse;
    } | {
        status: 404;
        data: HttpErrorResponse;
    } | {
        status: 422;
        data: HttpValidationError;
    }>(`/v1/knowledge-sources/${encodeURIComponent(slug)}`, oazapfts.json({
        ...opts,
        method: "PATCH",
        body: knowledgeSourcePatchesRequest,
        headers: oazapfts.mergeHeaders(opts?.headers, {
            "x-account-id": xAccountId,
            "x-username": xUsername,
            authorization
        })
    })));
}
/**
 * Delete Ks
 */
export function deleteKsV1KnowledgeSourcesSlugDelete({ slug, xAccountId, xUsername, authorization }: {
    slug: string;
    xAccountId: string;
    xUsername: string;
    authorization: string;
}, opts?: Oazapfts.RequestOpts) {
    return oazapfts.ok(oazapfts.fetchJson<{
        status: 204;
    } | {
        status: 403;
        data: HttpErrorResponse;
    } | {
        status: 404;
        data: HttpErrorResponse;
    } | {
        status: 422;
        data: HttpValidationError;
    } | {
        status: 500;
        data: HttpErrorResponse;
    }>(`/v1/knowledge-sources/${encodeURIComponent(slug)}`, {
        ...opts,
        method: "DELETE",
        headers: oazapfts.mergeHeaders(opts?.headers, {
            "x-account-id": xAccountId,
            "x-username": xUsername,
            authorization
        })
    }));
}
/**
 * Exists Ks
 */
export function existsKsV1KnowledgeSourcesSlugExistsGet({ slug, xAccountId, xUsername, authorization }: {
    slug: string;
    xAccountId?: string | null;
    xUsername?: string | null;
    authorization: string;
}, opts?: Oazapfts.RequestOpts) {
    return oazapfts.ok(oazapfts.fetchJson<{
        status: 204;
    } | {
        status: 404;
        data: HttpErrorResponse;
    } | {
        status: 422;
        data: HttpValidationError;
    }>(`/v1/knowledge-sources/${encodeURIComponent(slug)}/exists`, {
        ...opts,
        headers: oazapfts.mergeHeaders(opts?.headers, {
            "x-account-id": xAccountId,
            "x-username": xUsername,
            authorization
        })
    }));
}
/**
 * Share Ks
 */
export function shareKsV1KnowledgeSourcesSlugSharePost({ slug, xAccountId, xUsername, authorization }: {
    slug: string;
    xAccountId?: string | null;
    xUsername?: string | null;
    authorization: string;
}, opts?: Oazapfts.RequestOpts) {
    return oazapfts.ok(oazapfts.fetchJson<{
        status: 204;
    } | {
        status: 403;
        data: HttpErrorResponse;
    } | {
        status: 404;
        data: HttpErrorResponse;
    } | {
        status: 422;
        data: HttpErrorResponse;
    }>(`/v1/knowledge-sources/${encodeURIComponent(slug)}/share`, {
        ...opts,
        method: "POST",
        headers: oazapfts.mergeHeaders(opts?.headers, {
            "x-account-id": xAccountId,
            "x-username": xUsername,
            authorization
        })
    }));
}
/**
 * Add Favorite
 */
export function addFavoriteV1KnowledgeSourcesSlugFavoritePost({ slug, xAccountId, xUsername, authorization }: {
    slug: string;
    xAccountId?: string | null;
    xUsername?: string | null;
    authorization: string;
}, opts?: Oazapfts.RequestOpts) {
    return oazapfts.ok(oazapfts.fetchJson<{
        status: 204;
    } | {
        status: 403;
        data: HttpErrorResponse;
    } | {
        status: 404;
        data: HttpErrorResponse;
    } | {
        status: 422;
        data: HttpValidationError;
    }>(`/v1/knowledge-sources/${encodeURIComponent(slug)}/favorite`, {
        ...opts,
        method: "POST",
        headers: oazapfts.mergeHeaders(opts?.headers, {
            "x-account-id": xAccountId,
            "x-username": xUsername,
            authorization
        })
    }));
}
/**
 * Delete Favorite
 */
export function deleteFavoriteV1KnowledgeSourcesSlugFavoriteDelete({ slug, xAccountId, xUsername, authorization }: {
    slug: string;
    xAccountId?: string | null;
    xUsername?: string | null;
    authorization: string;
}, opts?: Oazapfts.RequestOpts) {
    return oazapfts.ok(oazapfts.fetchJson<{
        status: 204;
    } | {
        status: 403;
        data: HttpErrorResponse;
    } | {
        status: 404;
        data: HttpErrorResponse;
    } | {
        status: 422;
        data: HttpValidationError;
    }>(`/v1/knowledge-sources/${encodeURIComponent(slug)}/favorite`, {
        ...opts,
        method: "DELETE",
        headers: oazapfts.mergeHeaders(opts?.headers, {
            "x-account-id": xAccountId,
            "x-username": xUsername,
            authorization
        })
    }));
}
/**
 * Publish Ks
 */
export function publishKsV1KnowledgeSourcesSlugPublishPost({ slug, xAccountId, xUsername, authorization, body }: {
    slug: string;
    xAccountId?: string | null;
    xUsername?: string | null;
    authorization: string;
    body?: KnowledgeSourcesPublishRequest | null;
}, opts?: Oazapfts.RequestOpts) {
    return oazapfts.ok(oazapfts.fetchJson<{
        status: 204;
    } | {
        status: 403;
        data: HttpErrorResponse;
    } | {
        status: 404;
        data: HttpErrorResponse;
    } | {
        status: 422;
        data: HttpValidationError;
    }>(`/v1/knowledge-sources/${encodeURIComponent(slug)}/publish`, oazapfts.json({
        ...opts,
        method: "POST",
        body,
        headers: oazapfts.mergeHeaders(opts?.headers, {
            "x-account-id": xAccountId,
            "x-username": xUsername,
            authorization
        })
    })));
}
/**
 * Publish Ks Batch
 */
export function publishKsBatchV1KnowledgeSourcesPublishBatchPost({ xAccountId, xUsername, authorization, knowledgeSourcePublishBatchRequest }: {
    xAccountId?: string | null;
    xUsername?: string | null;
    authorization: string;
    knowledgeSourcePublishBatchRequest: KnowledgeSourcePublishBatchRequest;
}, opts?: Oazapfts.RequestOpts) {
    return oazapfts.ok(oazapfts.fetchJson<{
        status: 204;
    } | {
        status: 403;
        data: HttpErrorResponse;
    } | {
        status: 404;
    } | {
        status: 422;
        data: HttpValidationError;
    }>("/v1/knowledge-sources/publish/batch", oazapfts.json({
        ...opts,
        method: "POST",
        body: knowledgeSourcePublishBatchRequest,
        headers: oazapfts.mergeHeaders(opts?.headers, {
            "x-account-id": xAccountId,
            "x-username": xUsername,
            authorization
        })
    })));
}
/**
 * Fork Ks
 */
export function forkKsV1KnowledgeSourcesSlugForkPost({ slug, xAccountId, xUsername, authorization, forkKnowledgeSourceRequest }: {
    slug: string;
    xAccountId?: string | null;
    xUsername?: string | null;
    authorization: string;
    forkKnowledgeSourceRequest: ForkKnowledgeSourceRequest;
}, opts?: Oazapfts.RequestOpts) {
    return oazapfts.ok(oazapfts.fetchJson<{
        status: 204;
    } | {
        status: 403;
        data: HttpErrorResponse;
    } | {
        status: 404;
        data: HttpErrorResponse;
    } | {
        status: 422;
        data: HttpErrorResponse;
    }>(`/v1/knowledge-sources/${encodeURIComponent(slug)}/fork`, oazapfts.json({
        ...opts,
        method: "POST",
        body: forkKnowledgeSourceRequest,
        headers: oazapfts.mergeHeaders(opts?.headers, {
            "x-account-id": xAccountId,
            "x-username": xUsername,
            authorization
        })
    })));
}
/**
 * List Kos
 */
export function listKosV1KnowledgeSourcesKsSlugObjectsGet({ ksSlug, standalone, filePath, contentLimit, xAccountId, xUsername, authorization }: {
    ksSlug: string;
    standalone?: boolean | null;
    filePath?: string[];
    contentLimit?: number | null;
    xAccountId?: string | null;
    xUsername?: string | null;
    authorization: string;
}, opts?: Oazapfts.RequestOpts) {
    return oazapfts.ok(oazapfts.fetchJson<{
        status: 200;
        data: KnowledgeObjectResponse[];
    } | {
        status: 404;
        data: HttpErrorResponse;
    } | {
        status: 422;
        data: HttpValidationError;
    }>(`/v1/knowledge-sources/${encodeURIComponent(ksSlug)}/objects${QS.query(QS.explode({
        standalone,
        file_path: filePath,
        content_limit: contentLimit
    }))}`, {
        ...opts,
        headers: oazapfts.mergeHeaders(opts?.headers, {
            "x-account-id": xAccountId,
            "x-username": xUsername,
            authorization
        })
    }));
}
/**
 * Reset Kos
 */
export function resetKosV1KnowledgeSourcesKsSlugObjectsDelete({ ksSlug, standalone, xAccountId, xUsername, authorization }: {
    ksSlug: string;
    standalone?: boolean | null;
    xAccountId?: string | null;
    xUsername?: string | null;
    authorization: string;
}, opts?: Oazapfts.RequestOpts) {
    return oazapfts.ok(oazapfts.fetchJson<{
        status: 204;
    } | {
        status: 403;
        data: HttpErrorResponse;
    } | {
        status: 404;
        data: HttpErrorResponse;
    } | {
        status: 422;
        data: HttpValidationError;
    }>(`/v1/knowledge-sources/${encodeURIComponent(ksSlug)}/objects${QS.query(QS.explode({
        standalone
    }))}`, {
        ...opts,
        method: "DELETE",
        headers: oazapfts.mergeHeaders(opts?.headers, {
            "x-account-id": xAccountId,
            "x-username": xUsername,
            authorization
        })
    }));
}
/**
 * List Kos Checksums
 */
export function listKosChecksumsV1KnowledgeSourcesKsSlugObjectsChecksumsGet({ ksSlug, xAccountId, xUsername, authorization }: {
    ksSlug: string;
    xAccountId?: string | null;
    xUsername?: string | null;
    authorization: string;
}, opts?: Oazapfts.RequestOpts) {
    return oazapfts.ok(oazapfts.fetchJson<{
        status: 200;
        data: KnowledgeObjectChecksumResponse[];
    } | {
        status: 404;
        data: HttpErrorResponse;
    } | {
        status: 422;
        data: HttpValidationError;
    }>(`/v1/knowledge-sources/${encodeURIComponent(ksSlug)}/objects/checksums`, {
        ...opts,
        headers: oazapfts.mergeHeaders(opts?.headers, {
            "x-account-id": xAccountId,
            "x-username": xUsername,
            authorization
        })
    }));
}
/**
 * Vectorize Snippet Ks
 */
export function vectorizeSnippetKsV1KnowledgeSourcesSlugSnippetsPost(opts?: Oazapfts.RequestOpts) {
    return oazapfts.ok(oazapfts.fetchJson<{
        status: 201;
        data: any;
    } | {
        status: 404;
    }>(`/v1/knowledge-sources/${encodeURIComponent(slug)}/snippets`, {
        ...opts,
        method: "POST"
    }));
}
/**
 * Vectorize Custom Ks
 */
export function vectorizeCustomKsV1KnowledgeSourcesSlugCustomPost(opts?: Oazapfts.RequestOpts) {
    return oazapfts.ok(oazapfts.fetchJson<{
        status: 201;
        data: any;
    } | {
        status: 404;
    }>(`/v1/knowledge-sources/${encodeURIComponent(slug)}/custom`, {
        ...opts,
        method: "POST"
    }));
}
/**
 * Find Ko By Id
 */
export function findKoByIdV1KnowledgeSourcesKsSlugObjectsKoIdGet({ ksSlug, koId, xAccountId, xUsername, authorization }: {
    ksSlug: string;
    koId: string;
    xAccountId?: string | null;
    xUsername?: string | null;
    authorization: string;
}, opts?: Oazapfts.RequestOpts) {
    return oazapfts.ok(oazapfts.fetchJson<{
        status: 200;
        data: KnowledgeObjectResponse;
    } | {
        status: 404;
        data: HttpErrorResponse;
    } | {
        status: 422;
        data: HttpValidationError;
    }>(`/v1/knowledge-sources/${encodeURIComponent(ksSlug)}/objects/${encodeURIComponent(koId)}`, {
        ...opts,
        headers: oazapfts.mergeHeaders(opts?.headers, {
            "x-account-id": xAccountId,
            "x-username": xUsername,
            authorization
        })
    }));
}
/**
 * Delete Ko By Id
 */
export function deleteKoByIdV1KnowledgeSourcesKsSlugObjectsKoIdDelete({ ksSlug, koId, xAccountId, xUsername, authorization }: {
    ksSlug: string;
    koId: string;
    xAccountId?: string | null;
    xUsername?: string | null;
    authorization: string;
}, opts?: Oazapfts.RequestOpts) {
    return oazapfts.ok(oazapfts.fetchJson<{
        status: 204;
    } | {
        status: 403;
        data: HttpErrorResponse;
    } | {
        status: 404;
        data: HttpErrorResponse;
    } | {
        status: 422;
        data: HttpValidationError;
    }>(`/v1/knowledge-sources/${encodeURIComponent(ksSlug)}/objects/${encodeURIComponent(koId)}`, {
        ...opts,
        method: "DELETE",
        headers: oazapfts.mergeHeaders(opts?.headers, {
            "x-account-id": xAccountId,
            "x-username": xUsername,
            authorization
        })
    }));
}
/**
 * Get Last Message
 */
export function getLastMessageV1KnowledgeSourcesKsSlugLastMessageGet({ ksSlug, xAccountId, xUsername, authorization }: {
    ksSlug: string;
    xAccountId: string;
    xUsername?: string | null;
    authorization: string;
}, opts?: Oazapfts.RequestOpts) {
    return oazapfts.ok(oazapfts.fetchJson<{
        status: 200;
        data: KnowledgeSourceLastMessageResponse[];
    } | {
        status: 404;
        data: HttpErrorResponse;
    } | {
        status: 422;
        data: HttpValidationError;
    }>(`/v1/knowledge-sources/${encodeURIComponent(ksSlug)}/last-message`, {
        ...opts,
        headers: oazapfts.mergeHeaders(opts?.headers, {
            "x-account-id": xAccountId,
            "x-username": xUsername,
            authorization
        })
    }));
}
/**
 * Get Indexed Files
 */
export function getIndexedFilesV1KnowledgeSourcesKsSlugFilesGet({ ksSlug, xAccountId, xUsername, authorization }: {
    ksSlug: string;
    xAccountId?: string | null;
    xUsername?: string | null;
    authorization: string;
}, opts?: Oazapfts.RequestOpts) {
    return oazapfts.ok(oazapfts.fetchJson<{
        status: 200;
        data: KnowledgeSourceFilesIngestedResponse[];
    } | {
        status: 404;
        data: HttpErrorResponse;
    } | {
        status: 422;
        data: HttpValidationError;
    }>(`/v1/knowledge-sources/${encodeURIComponent(ksSlug)}/files`, {
        ...opts,
        headers: oazapfts.mergeHeaders(opts?.headers, {
            "x-account-id": xAccountId,
            "x-username": xUsername,
            authorization
        })
    }));
}
/**
 * Delete Indexed Files
 */
export function deleteIndexedFilesV1KnowledgeSourcesKsSlugFilesDelete({ ksSlug, xAccountId, xUsername, authorization, deleteIndexedFilesRequest }: {
    ksSlug: string;
    xAccountId?: string | null;
    xUsername?: string | null;
    authorization: string;
    deleteIndexedFilesRequest: DeleteIndexedFilesRequest;
}, opts?: Oazapfts.RequestOpts) {
    return oazapfts.ok(oazapfts.fetchJson<{
        status: 204;
    } | {
        status: 403;
        data: HttpErrorResponse;
    } | {
        status: 404;
        data: HttpErrorResponse;
    } | {
        status: 422;
        data: HttpValidationError;
    }>(`/v1/knowledge-sources/${encodeURIComponent(ksSlug)}/files`, oazapfts.json({
        ...opts,
        method: "DELETE",
        body: deleteIndexedFilesRequest,
        headers: oazapfts.mergeHeaders(opts?.headers, {
            "x-account-id": xAccountId,
            "x-username": xUsername,
            authorization
        })
    })));
}
/**
 * Get Upload Form
 */
export function getUploadFormV1FileUploadFormPost({ xAccountId, xUsername, authorization, newFileUploadRequest }: {
    xAccountId?: string | null;
    xUsername?: string | null;
    authorization: string;
    newFileUploadRequest: NewFileUploadRequest;
}, opts?: Oazapfts.RequestOpts) {
    return oazapfts.ok(oazapfts.fetchJson<{
        status: 201;
        data: NewFileUploadFormResponse;
    } | {
        status: 404;
    } | {
        status: 422;
        data: HttpValidationError;
    }>("/v1/file-upload/form", oazapfts.json({
        ...opts,
        method: "POST",
        body: newFileUploadRequest,
        headers: oazapfts.mergeHeaders(opts?.headers, {
            "x-account-id": xAccountId,
            "x-username": xUsername,
            authorization
        })
    })));
}
/**
 * Get File Upload Status
 */
export function getFileUploadStatusV1FileUploadFileUploadIdGet({ fileUploadId, xAccountId, xUsername, authorization }: {
    fileUploadId: string;
    xAccountId?: string | null;
    xUsername?: string | null;
    authorization: string;
}, opts?: Oazapfts.RequestOpts) {
    return oazapfts.ok(oazapfts.fetchJson<{
        status: 200;
        data: FileUploadStatusResponse;
    } | {
        status: 404;
    } | {
        status: 422;
        data: HttpValidationError;
    }>(`/v1/file-upload/${encodeURIComponent(fileUploadId)}`, {
        ...opts,
        headers: oazapfts.mergeHeaders(opts?.headers, {
            "x-account-id": xAccountId,
            "x-username": xUsername,
            authorization
        })
    }));
}
/**
 * Save Chunked Files Deprecated
 */
export function saveChunkedFilesDeprecatedV1FileUploadFileUploadIdPost({ fileUploadId, xAccountId, xUsername, authorization, fileUploadSaveChunksRequest }: {
    fileUploadId: string;
    xAccountId?: string | null;
    xUsername?: string | null;
    authorization: string;
    fileUploadSaveChunksRequest: FileUploadSaveChunksRequest;
}, opts?: Oazapfts.RequestOpts) {
    return oazapfts.ok(oazapfts.fetchJson<{
        status: 202;
        data: any;
    } | {
        status: 404;
    } | {
        status: 422;
        data: HttpValidationError;
    }>(`/v1/file-upload/${encodeURIComponent(fileUploadId)}`, oazapfts.json({
        ...opts,
        method: "POST",
        body: fileUploadSaveChunksRequest,
        headers: oazapfts.mergeHeaders(opts?.headers, {
            "x-account-id": xAccountId,
            "x-username": xUsername,
            authorization
        })
    })));
}
/**
 * Internal Patch File Upload Status
 */
export function internalPatchFileUploadStatusV1FileUploadFileUploadIdPatch({ fileUploadId, xAccountId, xUsername, authorization, fileUploadUpdateRequest }: {
    fileUploadId: string;
    xAccountId: string;
    xUsername?: string | null;
    authorization: string;
    fileUploadUpdateRequest: FileUploadUpdateRequest;
}, opts?: Oazapfts.RequestOpts) {
    return oazapfts.ok(oazapfts.fetchJson<{
        status: 200;
        data: FileUploadStatusResponse;
    } | {
        status: 404;
    } | {
        status: 422;
        data: HttpValidationError;
    }>(`/v1/file-upload/${encodeURIComponent(fileUploadId)}`, oazapfts.json({
        ...opts,
        method: "PATCH",
        body: fileUploadUpdateRequest,
        headers: oazapfts.mergeHeaders(opts?.headers, {
            "x-account-id": xAccountId,
            "x-username": xUsername,
            authorization
        })
    })));
}
/**
 * Save Chunked Files
 */
export function saveChunkedFilesV1FileUploadFileUploadIdKnowledgeObjectsPost({ fileUploadId, xAccountId, xUsername, authorization, fileUploadSaveChunksRequest }: {
    fileUploadId: string;
    xAccountId?: string | null;
    xUsername?: string | null;
    authorization: string;
    fileUploadSaveChunksRequest: FileUploadSaveChunksRequest;
}, opts?: Oazapfts.RequestOpts) {
    return oazapfts.ok(oazapfts.fetchJson<{
        status: 202;
        data: any;
    } | {
        status: 404;
    } | {
        status: 422;
        data: HttpValidationError;
    }>(`/v1/file-upload/${encodeURIComponent(fileUploadId)}/knowledge-objects`, oazapfts.json({
        ...opts,
        method: "POST",
        body: fileUploadSaveChunksRequest,
        headers: oazapfts.mergeHeaders(opts?.headers, {
            "x-account-id": xAccountId,
            "x-username": xUsername,
            authorization
        })
    })));
}
/**
 * Split Uploaded File
 */
export function splitUploadedFileV1FileUploadFileUploadIdSplitPost({ fileUploadId, xAccountId, xUsername, authorization, fileSplitRequest }: {
    fileUploadId: string;
    xAccountId?: string | null;
    xUsername?: string | null;
    authorization: string;
    fileSplitRequest: FileSplitRequest;
}, opts?: Oazapfts.RequestOpts) {
    return oazapfts.ok(oazapfts.fetchJson<{
        status: 202;
        data: any;
    } | {
        status: 404;
    } | {
        status: 422;
        data: HttpValidationError;
    }>(`/v1/file-upload/${encodeURIComponent(fileUploadId)}/split`, oazapfts.json({
        ...opts,
        method: "POST",
        body: fileSplitRequest,
        headers: oazapfts.mergeHeaders(opts?.headers, {
            "x-account-id": xAccountId,
            "x-username": xUsername,
            authorization
        })
    })));
}
/**
 * Split Uploaded File Preview
 */
export function splitUploadedFilePreviewV1FileUploadFileUploadIdSplitPreviewSplitStrategyGet({ fileUploadId, splitStrategy, xAccountId, xUsername, authorization }: {
    fileUploadId: string;
    splitStrategy: FileSplitStrategy;
    xAccountId?: string | null;
    xUsername?: string | null;
    authorization: string;
}, opts?: Oazapfts.RequestOpts) {
    return oazapfts.ok(oazapfts.fetchJson<{
        status: 200;
        data: any;
    } | {
        status: 404;
    } | {
        status: 422;
        data: HttpValidationError;
    }>(`/v1/file-upload/${encodeURIComponent(fileUploadId)}/split/preview/${encodeURIComponent(splitStrategy)}`, {
        ...opts,
        headers: oazapfts.mergeHeaders(opts?.headers, {
            "x-account-id": xAccountId,
            "x-username": xUsername,
            authorization
        })
    }));
}
/**
 * Internal Get File Upload Metadata
 */
export function internalGetFileUploadMetadataV1FileUploadFileUploadIdMetadataGet({ fileUploadId, xAccountId, xUsername, authorization }: {
    fileUploadId: string;
    xAccountId: string;
    xUsername?: string | null;
    authorization: string;
}, opts?: Oazapfts.RequestOpts) {
    return oazapfts.ok(oazapfts.fetchJson<{
        status: 200;
        data: FileMetadataResponse;
    } | {
        status: 404;
    } | {
        status: 422;
        data: HttpValidationError;
    }>(`/v1/file-upload/${encodeURIComponent(fileUploadId)}/metadata`, {
        ...opts,
        headers: oazapfts.mergeHeaders(opts?.headers, {
            "x-account-id": xAccountId,
            "x-username": xUsername,
            authorization
        })
    }));
}
/**
 * Get Presigned Download Url
 */
export function getPresignedDownloadUrlV1FileUploadPresignedUrlGet({ id, expiration, xAccountId, xUsername, authorization }: {
    id?: string[];
    expiration?: number;
    xAccountId?: string | null;
    xUsername?: string | null;
    authorization: string;
}, opts?: Oazapfts.RequestOpts) {
    return oazapfts.ok(oazapfts.fetchJson<{
        status: 200;
        data: PresignedUrlsDto;
    } | {
        status: 404;
    } | {
        status: 422;
        data: HttpValidationError;
    }>(`/v1/file-upload/presigned/url${QS.query(QS.explode({
        id,
        expiration
    }))}`, {
        ...opts,
        headers: oazapfts.mergeHeaders(opts?.headers, {
            "x-account-id": xAccountId,
            "x-username": xUsername,
            authorization
        })
    }));
}
/**
 * Delete Bucket Object
 */
export function deleteBucketObjectV1FileUploadUploadIdDelete({ uploadId, xAccountId, xUsername, authorization }: {
    uploadId: string;
    xAccountId?: string | null;
    xUsername?: string | null;
    authorization: string;
}, opts?: Oazapfts.RequestOpts) {
    return oazapfts.ok(oazapfts.fetchJson<{
        status: 204;
    } | {
        status: 404;
    } | {
        status: 422;
        data: HttpValidationError;
    }>(`/v1/file-upload/${encodeURIComponent(uploadId)}`, {
        ...opts,
        method: "DELETE",
        headers: oazapfts.mergeHeaders(opts?.headers, {
            "x-account-id": xAccountId,
            "x-username": xUsername,
            authorization
        })
    }));
}
/**
 * Internal Get Current Embedding Usage
 */
export function internalGetCurrentEmbeddingUsageV1TokenUsageEmbeddingGet({ xAccountId, xUsername, authorization }: {
    xAccountId: string;
    xUsername?: string | null;
    authorization: string;
}, opts?: Oazapfts.RequestOpts) {
    return oazapfts.ok(oazapfts.fetchJson<{
        status: 200;
        data: AccountEmbeddingTokenUsage;
    } | {
        status: 404;
    } | {
        status: 422;
        data: HttpValidationError;
    }>("/v1/token-usage/embedding", {
        ...opts,
        headers: oazapfts.mergeHeaders(opts?.headers, {
            "x-account-id": xAccountId,
            "x-username": xUsername,
            authorization
        })
    }));
}
/**
 * Internal Consume Token Usage
 */
export function internalConsumeTokenUsageV1TokenUsageConsumePost({ xAccountId, xUsername, authorization, consumeTokenRequest }: {
    xAccountId: string;
    xUsername: string;
    authorization: string;
    consumeTokenRequest: ConsumeTokenRequest;
}, opts?: Oazapfts.RequestOpts) {
    return oazapfts.ok(oazapfts.fetchJson<{
        status: 204;
    } | {
        status: 404;
    } | {
        status: 422;
        data: HttpValidationError;
    }>("/v1/token-usage/consume", oazapfts.json({
        ...opts,
        method: "POST",
        body: consumeTokenRequest,
        headers: oazapfts.mergeHeaders(opts?.headers, {
            "x-account-id": xAccountId,
            "x-username": xUsername,
            authorization
        })
    })));
}
/**
 * Get Upload Form V2
 */
export function getUploadFormV2V2FileUploadFormPost({ xAccountId, xUsername, authorization, newFileUploadRequest }: {
    xAccountId?: string | null;
    xUsername?: string | null;
    authorization: string;
    newFileUploadRequest: NewFileUploadRequest;
}, opts?: Oazapfts.RequestOpts) {
    return oazapfts.ok(oazapfts.fetchJson<{
        status: 201;
        data: NewFileUploadFormResponse;
    } | {
        status: 404;
    } | {
        status: 422;
        data: HttpValidationError;
    }>("/v2/file-upload/form", oazapfts.json({
        ...opts,
        method: "POST",
        body: newFileUploadRequest,
        headers: oazapfts.mergeHeaders(opts?.headers, {
            "x-account-id": xAccountId,
            "x-username": xUsername,
            authorization
        })
    })));
}
/**
 * Internal Get Llm Config
 */
export function internalGetLlmConfigV1AccountsLlmConfigGet({ xAccountId, xUsername, authorization }: {
    xAccountId: string;
    xUsername?: string | null;
    authorization: string;
}, opts?: Oazapfts.RequestOpts) {
    return oazapfts.ok(oazapfts.fetchJson<{
        status: 200;
        data: AccountLlmConfigResponse;
    } | {
        status: 404;
    } | {
        status: 422;
        data: HttpValidationError;
    }>("/v1/accounts/llm-config", {
        ...opts,
        headers: oazapfts.mergeHeaders(opts?.headers, {
            "x-account-id": xAccountId,
            "x-username": xUsername,
            authorization
        })
    }));
}
/**
 * Internal Get Or Create Knowledge Group
 */
export function internalGetOrCreateKnowledgeGroupV1KnowledgeGroupsGetOrCreatePut({ xAccountId, xUsername, authorization, getOrCreateKnowledgeGroupRequest }: {
    xAccountId: string;
    xUsername?: string | null;
    authorization: string;
    getOrCreateKnowledgeGroupRequest: GetOrCreateKnowledgeGroupRequest;
}, opts?: Oazapfts.RequestOpts) {
    return oazapfts.ok(oazapfts.fetchJson<{
        status: 200;
        data: KnowledgeGroupResponse;
    } | {
        status: 404;
    } | {
        status: 422;
        data: HttpValidationError;
    }>("/v1/knowledge-groups/get-or-create", oazapfts.json({
        ...opts,
        method: "PUT",
        body: getOrCreateKnowledgeGroupRequest,
        headers: oazapfts.mergeHeaders(opts?.headers, {
            "x-account-id": xAccountId,
            "x-username": xUsername,
            authorization
        })
    })));
}
/**
 * Internal Delete By Not Checksums
 */
export function internalDeleteByNotChecksumsV1KnowledgeGroupsGroupIdObjectsDelete({ groupId, xAccountId, xUsername, authorization, deleteGroupObjectsRequests }: {
    groupId: string;
    xAccountId: string;
    xUsername?: string | null;
    authorization: string;
    deleteGroupObjectsRequests: DeleteGroupObjectsRequests;
}, opts?: Oazapfts.RequestOpts) {
    return oazapfts.ok(oazapfts.fetchJson<{
        status: 200;
        data: DeletedCountResponse;
    } | {
        status: 404;
    } | {
        status: 422;
        data: HttpValidationError;
    }>(`/v1/knowledge-groups/${encodeURIComponent(groupId)}/objects`, oazapfts.json({
        ...opts,
        method: "DELETE",
        body: deleteGroupObjectsRequests,
        headers: oazapfts.mergeHeaders(opts?.headers, {
            "x-account-id": xAccountId,
            "x-username": xUsername,
            authorization
        })
    })));
}
/**
 * Internal Update Knowledge Group
 */
export function internalUpdateKnowledgeGroupV1KnowledgeGroupsGroupIdPatch({ groupId, xAccountId, xUsername, authorization, updateKnowledgeGroupRequest }: {
    groupId: string;
    xAccountId: string;
    xUsername?: string | null;
    authorization: string;
    updateKnowledgeGroupRequest: UpdateKnowledgeGroupRequest;
}, opts?: Oazapfts.RequestOpts) {
    return oazapfts.ok(oazapfts.fetchJson<{
        status: 200;
        data: any;
    } | {
        status: 404;
    } | {
        status: 422;
        data: HttpValidationError;
    }>(`/v1/knowledge-groups/${encodeURIComponent(groupId)}`, oazapfts.json({
        ...opts,
        method: "PATCH",
        body: updateKnowledgeGroupRequest,
        headers: oazapfts.mergeHeaders(opts?.headers, {
            "x-account-id": xAccountId,
            "x-username": xUsername,
            authorization
        })
    })));
}
/**
 * Internal Create Knowledge Objects Batch
 */
export function internalCreateKnowledgeObjectsBatchV1KnowledgeObjectsBatchPost({ xAccountId, xUsername, authorization, body }: {
    xAccountId: string;
    xUsername?: string | null;
    authorization: string;
    body: CreateKnowledgeObjectsBatchRequest[];
}, opts?: Oazapfts.RequestOpts) {
    return oazapfts.ok(oazapfts.fetchJson<{
        status: 200;
        data: any;
    } | {
        status: 404;
    } | {
        status: 422;
        data: HttpValidationError;
    }>("/v1/knowledge-objects/batch", oazapfts.json({
        ...opts,
        method: "POST",
        body,
        headers: oazapfts.mergeHeaders(opts?.headers, {
            "x-account-id": xAccountId,
            "x-username": xUsername,
            authorization
        })
    })));
}
/**
 * Create Embedding
 */
export function createEmbeddingV1EmbeddingsPost({ xAccountId, xUsername, authorization, body }: {
    xAccountId?: string | null;
    xUsername?: string | null;
    authorization: string;
    body: string;
}, opts?: Oazapfts.RequestOpts) {
    return oazapfts.ok(oazapfts.fetchJson<{
        status: 200;
        data: InternalEmbedTextResponse;
    } | {
        status: 422;
        data: HttpValidationError;
    }>("/v1/embeddings", {
        ...opts,
        method: "POST",
        body,
        headers: oazapfts.mergeHeaders(opts?.headers, {
            "x-account-id": xAccountId,
            "x-username": xUsername,
            authorization
        })
    }));
}
/**
 * List Conversations
 */
export function listConversationsV1ConversationsGet({ size, page, xAccountId, xUsername, authorization }: {
    size?: number;
    page?: number;
    xAccountId?: string | null;
    xUsername?: string | null;
    authorization: string;
}, opts?: Oazapfts.RequestOpts) {
    return oazapfts.ok(oazapfts.fetchJson<{
        status: 200;
        data: MinimalConversationResponse[];
    } | {
        status: 404;
    } | {
        status: 422;
        data: HttpErrorResponse;
    } | {
        status: 500;
        data: HttpErrorResponse;
    }>(`/v1/conversations${QS.query(QS.explode({
        size,
        page
    }))}`, {
        ...opts,
        headers: oazapfts.mergeHeaders(opts?.headers, {
            "x-account-id": xAccountId,
            "x-username": xUsername,
            authorization
        })
    }));
}
/**
 * Create Conversation
 */
export function createConversationV1ConversationsPost({ xAccountId, xUsername, authorization, createConversationRequest }: {
    xAccountId?: string | null;
    xUsername?: string | null;
    authorization: string;
    createConversationRequest: CreateConversationRequest;
}, opts?: Oazapfts.RequestOpts) {
    return oazapfts.ok(oazapfts.fetchJson<{
        status: 201;
        data: MinimalConversationResponse;
    } | {
        status: 404;
    } | {
        status: 422;
        data: HttpErrorResponse;
    } | {
        status: 500;
        data: HttpErrorResponse;
    }>("/v1/conversations", oazapfts.json({
        ...opts,
        method: "POST",
        body: createConversationRequest,
        headers: oazapfts.mergeHeaders(opts?.headers, {
            "x-account-id": xAccountId,
            "x-username": xUsername,
            authorization
        })
    })));
}
/**
 * Get Conversation With History
 */
export function getConversationWithHistoryV1ConversationsConversationIdGet({ conversationId, xAccountId, xUsername, authorization }: {
    conversationId: string;
    xAccountId?: string | null;
    xUsername?: string | null;
    authorization: string;
}, opts?: Oazapfts.RequestOpts) {
    return oazapfts.ok(oazapfts.fetchJson<{
        status: 200;
        data: ConversationResponse;
    } | {
        status: 404;
    } | {
        status: 422;
        data: HttpErrorResponse;
    } | {
        status: 500;
        data: HttpErrorResponse;
    }>(`/v1/conversations/${encodeURIComponent(conversationId)}`, {
        ...opts,
        headers: oazapfts.mergeHeaders(opts?.headers, {
            "x-account-id": xAccountId,
            "x-username": xUsername,
            authorization
        })
    }));
}
/**
 * Update Title
 */
export function updateTitleV1ConversationsConversationIdPatch({ conversationId, xAccountId, xUsername, authorization, conversationUpdateTitleRequest }: {
    conversationId: string;
    xAccountId?: string | null;
    xUsername?: string | null;
    authorization: string;
    conversationUpdateTitleRequest: ConversationUpdateTitleRequest;
}, opts?: Oazapfts.RequestOpts) {
    return oazapfts.ok(oazapfts.fetchJson<{
        status: 200;
        data: ConversationResponse;
    } | {
        status: 404;
    } | {
        status: 422;
        data: HttpErrorResponse;
    } | {
        status: 500;
        data: HttpErrorResponse;
    }>(`/v1/conversations/${encodeURIComponent(conversationId)}`, oazapfts.json({
        ...opts,
        method: "PATCH",
        body: conversationUpdateTitleRequest,
        headers: oazapfts.mergeHeaders(opts?.headers, {
            "x-account-id": xAccountId,
            "x-username": xUsername,
            authorization
        })
    })));
}
/**
 * Delete Conversation
 */
export function deleteConversationV1ConversationsConversationIdDelete({ conversationId, xAccountId, xUsername, authorization }: {
    conversationId: string;
    xAccountId?: string | null;
    xUsername?: string | null;
    authorization: string;
}, opts?: Oazapfts.RequestOpts) {
    return oazapfts.ok(oazapfts.fetchJson<{
        status: 204;
    } | {
        status: 404;
    } | {
        status: 422;
        data: HttpErrorResponse;
    } | {
        status: 500;
        data: HttpErrorResponse;
    }>(`/v1/conversations/${encodeURIComponent(conversationId)}`, {
        ...opts,
        method: "DELETE",
        headers: oazapfts.mergeHeaders(opts?.headers, {
            "x-account-id": xAccountId,
            "x-username": xUsername,
            authorization
        })
    }));
}
/**
 * Download Conversation
 */
export function downloadConversationV1ConversationsConversationIdDownloadGet({ conversationId, xAccountId, xUsername, authorization }: {
    conversationId: string;
    xAccountId?: string | null;
    xUsername?: string | null;
    authorization: string;
}, opts?: Oazapfts.RequestOpts) {
    return oazapfts.ok(oazapfts.fetchJson<{
        status: 200;
        data: any;
    } | {
        status: 404;
    } | {
        status: 422;
        data: HttpErrorResponse;
    } | {
        status: 500;
        data: HttpErrorResponse;
    }>(`/v1/conversations/${encodeURIComponent(conversationId)}/download`, {
        ...opts,
        headers: oazapfts.mergeHeaders(opts?.headers, {
            "x-account-id": xAccountId,
            "x-username": xUsername,
            authorization
        })
    }));
}
/**
 * Add Conversation History
 */
export function addConversationHistoryV1ConversationsConversationIdMessagesPost({ conversationId, xAccountId, xUsername, authorization, conversationHistoryRequest }: {
    conversationId: string;
    xAccountId?: string | null;
    xUsername?: string | null;
    authorization: string;
    conversationHistoryRequest: ConversationHistoryRequest;
}, opts?: Oazapfts.RequestOpts) {
    return oazapfts.ok(oazapfts.fetchJson<{
        status: 201;
        data: any;
    } | {
        status: 404;
    } | {
        status: 422;
        data: HttpErrorResponse;
    } | {
        status: 500;
        data: HttpErrorResponse;
    }>(`/v1/conversations/${encodeURIComponent(conversationId)}/messages`, oazapfts.json({
        ...opts,
        method: "POST",
        body: conversationHistoryRequest,
        headers: oazapfts.mergeHeaders(opts?.headers, {
            "x-account-id": xAccountId,
            "x-username": xUsername,
            authorization
        })
    })));
}
/**
 * Get Last History
 */
export function getLastHistoryV1ConversationsConversationIdMessagesLastGet({ conversationId, xAccountId, xUsername, authorization }: {
    conversationId: string;
    xAccountId?: string | null;
    xUsername?: string | null;
    authorization: string;
}, opts?: Oazapfts.RequestOpts) {
    return oazapfts.ok(oazapfts.fetchJson<{
        status: 200;
        data: ConversationHistoryResponse | null;
    } | {
        status: 404;
    } | {
        status: 422;
        data: HttpErrorResponse;
    } | {
        status: 500;
        data: HttpErrorResponse;
    }>(`/v1/conversations/${encodeURIComponent(conversationId)}/messages/last`, {
        ...opts,
        headers: oazapfts.mergeHeaders(opts?.headers, {
            "x-account-id": xAccountId,
            "x-username": xUsername,
            authorization
        })
    }));
}
/**
 * Search
 */
export function searchV1ConversationsMessagesSearchPost({ xAccountId, xUsername, authorization, searchConversationHistoryRequest }: {
    xAccountId?: string | null;
    xUsername?: string | null;
    authorization: string;
    searchConversationHistoryRequest: SearchConversationHistoryRequest;
}, opts?: Oazapfts.RequestOpts) {
    return oazapfts.ok(oazapfts.fetchJson<{
        status: 200;
        data: SearchConversationHistoryResponse[];
    } | {
        status: 404;
    } | {
        status: 422;
        data: HttpValidationError;
    }>("/v1/conversations/messages/search", oazapfts.json({
        ...opts,
        method: "POST",
        body: searchConversationHistoryRequest,
        headers: oazapfts.mergeHeaders(opts?.headers, {
            "x-account-id": xAccountId,
            "x-username": xUsername,
            authorization
        })
    })));
}
/**
 * List Ks
 */
export function listKsV2KnowledgeSourcesGet({ name, slug, size, page, visibilityList, order, types, xAccountId, xUsername, authorization }: {
    name?: string | null;
    slug?: string | null;
    size?: number;
    page?: number;
    visibilityList?: VisibilityLevelEnum[] | null;
    order?: OrderEnum | null;
    types?: KnowledgeSourceTypeEnum[];
    xAccountId?: string | null;
    xUsername?: string | null;
    authorization: string;
}, opts?: Oazapfts.RequestOpts) {
    return oazapfts.ok(oazapfts.fetchJson<{
        status: 200;
        data: PaginatedResponseKnowledgeSourceItemResponse;
    } | {
        status: 404;
    } | {
        status: 422;
        data: HttpValidationError;
    }>(`/v2/knowledge-sources${QS.query(QS.explode({
        name,
        slug,
        size,
        page,
        visibility_list: visibilityList,
        order,
        types
    }))}`, {
        ...opts,
        headers: oazapfts.mergeHeaders(opts?.headers, {
            "x-account-id": xAccountId,
            "x-username": xUsername,
            authorization
        })
    }));
}
/**
 * List Kos
 */
export function listKosV2KnowledgeSourcesKsSlugObjectsGet({ ksSlug, size, page, order, standalone, filePath, contentLimit, xAccountId, xUsername, authorization }: {
    ksSlug: string;
    size?: number;
    page?: number;
    order?: OrderEnum | null;
    standalone?: boolean | null;
    filePath?: string[];
    contentLimit?: number | null;
    xAccountId?: string | null;
    xUsername?: string | null;
    authorization: string;
}, opts?: Oazapfts.RequestOpts) {
    return oazapfts.ok(oazapfts.fetchJson<{
        status: 200;
        data: PaginatedResponseKnowledgeObjectResponse;
    } | {
        status: 404;
        data: HttpErrorResponse;
    } | {
        status: 422;
        data: HttpValidationError;
    }>(`/v2/knowledge-sources/${encodeURIComponent(ksSlug)}/objects${QS.query(QS.explode({
        size,
        page,
        order,
        standalone,
        file_path: filePath,
        content_limit: contentLimit
    }))}`, {
        ...opts,
        headers: oazapfts.mergeHeaders(opts?.headers, {
            "x-account-id": xAccountId,
            "x-username": xUsername,
            authorization
        })
    }));
}
/**
 * Upload Script
 */
export function uploadScriptV1ScriptUploadPost({ xAccountId, xUsername, authorization, uploadScriptRequest }: {
    xAccountId?: string | null;
    xUsername?: string | null;
    authorization: string;
    uploadScriptRequest: UploadScriptRequest;
}, opts?: Oazapfts.RequestOpts) {
    return oazapfts.ok(oazapfts.fetchJson<{
        status: 201;
        data: UploadScriptResponse;
    } | {
        status: 404;
    } | {
        status: 422;
        data: HttpValidationError;
    }>("/v1/script-upload", oazapfts.json({
        ...opts,
        method: "POST",
        body: uploadScriptRequest,
        headers: oazapfts.mergeHeaders(opts?.headers, {
            "x-account-id": xAccountId,
            "x-username": xUsername,
            authorization
        })
    })));
}
/**
 * Get Script Content
 */
export function getScriptContentV1ScriptUploadScriptIdGet({ scriptId, xAccountId, xUsername, authorization }: {
    scriptId: string;
    xAccountId?: string | null;
    xUsername?: string | null;
    authorization: string;
}, opts?: Oazapfts.RequestOpts) {
    return oazapfts.ok(oazapfts.fetchJson<{
        status: 200;
        data: GetScriptResponse;
    } | {
        status: 404;
    } | {
        status: 422;
        data: HttpValidationError;
    }>(`/v1/script-upload/${encodeURIComponent(scriptId)}`, {
        ...opts,
        headers: oazapfts.mergeHeaders(opts?.headers, {
            "x-account-id": xAccountId,
            "x-username": xUsername,
            authorization
        })
    }));
}
/**
 * Get Script Presigned Url
 */
export function getScriptPresignedUrlV1ScriptUploadScriptIdPresignedUrlGet({ scriptId, xAccountId, xUsername, authorization }: {
    scriptId: string;
    xAccountId?: string | null;
    xUsername?: string | null;
    authorization: string;
}, opts?: Oazapfts.RequestOpts) {
    return oazapfts.ok(oazapfts.fetchJson<{
        status: 200;
        data: GetPresignedUrlResponse;
    } | {
        status: 404;
    } | {
        status: 422;
        data: HttpValidationError;
    }>(`/v1/script-upload/${encodeURIComponent(scriptId)}/presigned/url`, {
        ...opts,
        headers: oazapfts.mergeHeaders(opts?.headers, {
            "x-account-id": xAccountId,
            "x-username": xUsername,
            authorization
        })
    }));
}
/**
 * Create Script Execution State
 */
export function createScriptExecutionStateV1ScriptExecutionExecutionIdStepSlugAttemptPut({ executionId, stepSlug, attempt, xAccountId, xUsername, authorization, createScriptExecutionRequest }: {
    executionId: string;
    stepSlug: string;
    attempt: number;
    xAccountId?: string | null;
    xUsername?: string | null;
    authorization: string;
    createScriptExecutionRequest: CreateScriptExecutionRequest;
}, opts?: Oazapfts.RequestOpts) {
    return oazapfts.ok(oazapfts.fetchJson<{
        status: 201;
        data: CreateScriptExecutionResponse;
    } | {
        status: 404;
    } | {
        status: 422;
        data: HttpValidationError;
    }>(`/v1/script-execution/${encodeURIComponent(executionId)}/${encodeURIComponent(stepSlug)}/${encodeURIComponent(attempt)}`, oazapfts.json({
        ...opts,
        method: "PUT",
        body: createScriptExecutionRequest,
        headers: oazapfts.mergeHeaders(opts?.headers, {
            "x-account-id": xAccountId,
            "x-username": xUsername,
            authorization
        })
    })));
}
/**
 * Get Script Execution State Presigned Url
 */
export function getScriptExecutionStatePresignedUrlV1ScriptExecutionIdPresignedUrlGet({ id, xAccountId, xUsername, authorization }: {
    id: string;
    xAccountId?: string | null;
    xUsername?: string | null;
    authorization: string;
}, opts?: Oazapfts.RequestOpts) {
    return oazapfts.ok(oazapfts.fetchJson<{
        status: 200;
        data: GetPresignedUrlResponse;
    } | {
        status: 404;
    } | {
        status: 422;
        data: HttpValidationError;
    }>(`/v1/script-execution/${encodeURIComponent(id)}/presigned/url`, {
        ...opts,
        headers: oazapfts.mergeHeaders(opts?.headers, {
            "x-account-id": xAccountId,
            "x-username": xUsername,
            authorization
        })
    }));
}
