// @stack-spot/portal-network/src/api/discover.ts
/**
 * OpenAPI definition
 * v0
 * DO NOT MODIFY - This file has been generated using oazapfts.
 * See https://www.npmjs.com/package/oazapfts
 */
import * as Oazapfts from "@oazapfts/runtime";
import * as QS from "@oazapfts/runtime/query";
export const defaults: Oazapfts.Defaults<Oazapfts.CustomHeaders> = {
    headers: {},
    baseUrl: "https://discover-discover-core.dev.stackspot.com",
};
const oazapfts = Oazapfts.runtime(defaults);
export const servers = {
    generatedServerUrl: "https://discover-discover-core.dev.stackspot.com"
};
export type GetOpportunityResponseHypothesisResponse = {
    id: string;
    title: string;
    description: string;
};
export type GetOpportunityResponse = {
    id: string;
    title: string;
    description: string;
    createdBy: string;
    createdAt?: string;
    hypotheses: GetOpportunityResponseHypothesisResponse[];
    updatedBy?: string;
    updatedAt?: string;
    chatId: string;
};
export type UpdateOpportunityRequest = {
    title?: string;
    description?: string;
};
export type UpdateOpportunityResponse = {
    id: string;
    title: string;
    description: string;
    createdBy: string;
    createdAt?: string;
    updatedBy?: string;
    updatedAt?: string;
    chatId: string;
};
export type GetHypothesisResponseDocumentResponse = {
    id: string;
    title: string;
    description: string;
};
export type GetHypothesisResponse = {
    id: string;
    opportunityId: string;
    title: string;
    description: string;
    documents: GetHypothesisResponseDocumentResponse[];
    createdBy: string;
    createdAt?: string;
    updatedBy?: string;
    updatedAt?: string;
    chatId: string;
};
export type UpdateHypothesisRequest = {
    title?: string;
    description?: string;
};
export type UpdateHypothesisResponse = {
    id: string;
    opportunityId: string;
    title: string;
    description: string;
    createdBy: string;
    createdAt?: string;
    updatedBy?: string;
    updatedAt?: string;
    chatId: string;
};
export type GetDocumentResponse = {
    id: string;
    hypothesisId: string;
    title: string;
    description: string;
    createdBy: string;
    createdAt?: string;
    updatedBy?: string;
    updatedAt?: string;
    chatId: string;
};
export type UpdateDocumentRequest = {
    title?: string;
    description?: string;
    content?: string;
};
export type UpdateDocumentResponse = {
    id: string;
    hypothesisId: string;
    title: string;
    description: string;
    createdBy: string;
    createdAt?: string;
    updatedBy?: string;
    updatedAt?: string;
    chatId: string;
};
export type SortResponse = {
    direction: string;
    nullHandling: string;
    ascending: boolean;
    property: string;
    ignoreCase: boolean;
};
export type PageableResponse = {
    offset: number;
    sort: SortResponse[];
    paged: boolean;
    pageSize: number;
    pageNumber: number;
    unpaged: boolean;
};
export type PageResponseGetOpportunityResponse = {
    totalElements: number;
    totalPages: number;
    first: boolean;
    last: boolean;
    size: number;
    content: GetOpportunityResponse[];
    "number": number;
    sort: SortResponse[];
    numberOfElements: number;
    pageable: PageableResponse;
    empty: boolean;
};
export type CreateOpportunityRequest = {
    title?: string;
    description?: string;
    chatId: string;
};
export type CreateOpportunityResponse = {
    id: string;
    title: string;
    description: string;
    createdBy: string;
    createdAt?: string;
    updatedBy?: string;
    updatedAt?: string;
    chatId: string;
};
export type InsightResponseInsightData = {
    id: string;
    title: string;
    summary: string;
    badge: string;
    chartUrl: string;
};
export type InsightResponse = {
    insights: InsightResponseInsightData[];
    generatedAt: string;
};
export type CreateHypothesisRequest = {
    opportunityId?: string;
    title?: string;
    description?: string;
    chatId: string;
};
export type CreateHypothesisResponse = {
    id: string;
    opportunityId: string;
    title: string;
    description: string;
    createdBy: string;
    createdAt?: string;
    updatedBy?: string;
    updatedAt?: string;
    chatId: string;
};
export type CreateDocumentRequest = {
    hypothesisId?: string;
    title?: string;
    description?: string;
    content?: string;
    chatId: string;
};
export type CreateDocumentResponse = {
    id: string;
    hypothesisId: string;
    title: string;
    description: string;
    createdBy: string;
    createdAt?: string;
    updatedBy?: string;
    updatedAt?: string;
    chatId: string;
};
export type MessageRequest = {
    prompt: string;
    conversationId: string;
};
export type SseEmitter = {
    timeout?: number;
};
export type AdminAgentCreateRequest = {
    agentId: string;
    name: string;
    description: string;
    schema: string;
};
export type AdminAgentResponse = {
    id: string;
    name: string;
    description: string;
    active: boolean;
    createdAt: string;
    updatedAt?: string;
};
export type GetInsightResponse = {
    id: string;
    title: string;
    content: string;
    chartUrl: string;
};
export type GetArtifactResponse = {
    "type": "OPPORTUNITY" | "HYPOTHESIS" | "DOCUMENT";
    id: string;
    title: string;
    description: string;
};
export type AdminHelloResponse = {
    message: string;
};
/**
 * Get specific Opportunity
 */
export function getById({ opportunityId }: {
    opportunityId: string;
}, opts?: Oazapfts.RequestOpts) {
    return oazapfts.ok(oazapfts.fetchJson<{
        status: 200;
        data: GetOpportunityResponse;
    }>(`/v2/opportunities/${encodeURIComponent(opportunityId)}`, {
        ...opts
    }));
}
/**
 * Update specific Opportunity
 */
export function updateById({ opportunityId, updateOpportunityRequest }: {
    opportunityId: string;
    updateOpportunityRequest: UpdateOpportunityRequest;
}, opts?: Oazapfts.RequestOpts) {
    return oazapfts.ok(oazapfts.fetchJson<{
        status: 200;
        data: UpdateOpportunityResponse;
    }>(`/v2/opportunities/${encodeURIComponent(opportunityId)}`, oazapfts.json({
        ...opts,
        method: "PUT",
        body: updateOpportunityRequest
    })));
}
/**
 * Delete specific Opportunity
 */
export function deleteById({ opportunityId }: {
    opportunityId: string;
}, opts?: Oazapfts.RequestOpts) {
    return oazapfts.ok(oazapfts.fetchText(`/v2/opportunities/${encodeURIComponent(opportunityId)}`, {
        ...opts,
        method: "DELETE"
    }));
}
/**
 * Get specific Hypothesis
 */
export function getById1({ hypothesisId }: {
    hypothesisId: string;
}, opts?: Oazapfts.RequestOpts) {
    return oazapfts.ok(oazapfts.fetchJson<{
        status: 200;
        data: GetHypothesisResponse;
    }>(`/v2/hypotheses/${encodeURIComponent(hypothesisId)}`, {
        ...opts
    }));
}
/**
 * Update specific Hypothesis
 */
export function updateById1({ hypothesisId, updateHypothesisRequest }: {
    hypothesisId: string;
    updateHypothesisRequest: UpdateHypothesisRequest;
}, opts?: Oazapfts.RequestOpts) {
    return oazapfts.ok(oazapfts.fetchJson<{
        status: 200;
        data: UpdateHypothesisResponse;
    }>(`/v2/hypotheses/${encodeURIComponent(hypothesisId)}`, oazapfts.json({
        ...opts,
        method: "PUT",
        body: updateHypothesisRequest
    })));
}
/**
 * Delete specific Hypothesis
 */
export function deleteById1({ hypothesisId }: {
    hypothesisId: string;
}, opts?: Oazapfts.RequestOpts) {
    return oazapfts.ok(oazapfts.fetchText(`/v2/hypotheses/${encodeURIComponent(hypothesisId)}`, {
        ...opts,
        method: "DELETE"
    }));
}
/**
 * Get specific Document
 */
export function getById2({ documentId }: {
    documentId: string;
}, opts?: Oazapfts.RequestOpts) {
    return oazapfts.ok(oazapfts.fetchJson<{
        status: 200;
        data: GetDocumentResponse;
    }>(`/v2/documents/${encodeURIComponent(documentId)}`, {
        ...opts
    }));
}
/**
 * Update specific Document
 */
export function updateById2({ documentId, updateDocumentRequest }: {
    documentId: string;
    updateDocumentRequest: UpdateDocumentRequest;
}, opts?: Oazapfts.RequestOpts) {
    return oazapfts.ok(oazapfts.fetchJson<{
        status: 200;
        data: UpdateDocumentResponse;
    }>(`/v2/documents/${encodeURIComponent(documentId)}`, oazapfts.json({
        ...opts,
        method: "PUT",
        body: updateDocumentRequest
    })));
}
/**
 * Delete specific Document
 */
export function deleteById2({ documentId }: {
    documentId: string;
}, opts?: Oazapfts.RequestOpts) {
    return oazapfts.ok(oazapfts.fetchText(`/v2/documents/${encodeURIComponent(documentId)}`, {
        ...opts,
        method: "DELETE"
    }));
}
/**
 * Get all Opportunities
 */
export function getAll({ filter, page, size, sort, direction }: {
    filter?: string;
    page?: number;
    size?: number;
    sort?: string;
    direction?: string;
}, opts?: Oazapfts.RequestOpts) {
    return oazapfts.ok(oazapfts.fetchJson<{
        status: 200;
        data: PageResponseGetOpportunityResponse;
    }>(`/v2/opportunities${QS.query(QS.explode({
        filter,
        page,
        size,
        sort,
        direction
    }))}`, {
        ...opts
    }));
}
/**
 * Create new Opportunity
 */
export function create({ createOpportunityRequest }: {
    createOpportunityRequest: CreateOpportunityRequest;
}, opts?: Oazapfts.RequestOpts) {
    return oazapfts.ok(oazapfts.fetchJson<{
        status: 200;
        data: CreateOpportunityResponse;
    }>("/v2/opportunities", oazapfts.json({
        ...opts,
        method: "POST",
        body: createOpportunityRequest
    })));
}
/**
 * Fetch Insights
 */
export function fetchInsights(opts?: Oazapfts.RequestOpts) {
    return oazapfts.ok(oazapfts.fetchJson<{
        status: 200;
        data: InsightResponse;
    }>("/v2/insights", {
        ...opts,
        method: "POST"
    }));
}
/**
 * Refresh Insights
 */
export function refreshInsights(opts?: Oazapfts.RequestOpts) {
    return oazapfts.ok(oazapfts.fetchJson<{
        status: 200;
        data: InsightResponse;
    }>("/v2/insights/refresh", {
        ...opts,
        method: "POST"
    }));
}
/**
 * Get all Hypotheses
 */
export function getAll1({ opportunityId }: {
    opportunityId: string;
}, opts?: Oazapfts.RequestOpts) {
    return oazapfts.ok(oazapfts.fetchJson<{
        status: 200;
        data: GetHypothesisResponse[];
    }>(`/v2/hypotheses${QS.query(QS.explode({
        opportunityId
    }))}`, {
        ...opts
    }));
}
/**
 * Create new Hypothesis
 */
export function create1({ createHypothesisRequest }: {
    createHypothesisRequest: CreateHypothesisRequest;
}, opts?: Oazapfts.RequestOpts) {
    return oazapfts.ok(oazapfts.fetchJson<{
        status: 200;
        data: CreateHypothesisResponse;
    }>("/v2/hypotheses", oazapfts.json({
        ...opts,
        method: "POST",
        body: createHypothesisRequest
    })));
}
/**
 * Get all Documents by Hypothesis
 */
export function getAllByHypothesis({ hypothesisId }: {
    hypothesisId: string;
}, opts?: Oazapfts.RequestOpts) {
    return oazapfts.ok(oazapfts.fetchJson<{
        status: 200;
        data: GetDocumentResponse[];
    }>(`/v2/documents${QS.query(QS.explode({
        hypothesisId
    }))}`, {
        ...opts
    }));
}
/**
 * Create Document
 */
export function create2({ createDocumentRequest }: {
    createDocumentRequest: CreateDocumentRequest;
}, opts?: Oazapfts.RequestOpts) {
    return oazapfts.ok(oazapfts.fetchJson<{
        status: 200;
        data: CreateDocumentResponse;
    }>("/v2/documents", oazapfts.json({
        ...opts,
        method: "POST",
        body: createDocumentRequest
    })));
}
export function chat({ authorization, messageRequest }: {
    authorization: string;
    messageRequest: MessageRequest;
}, opts?: Oazapfts.RequestOpts) {
    return oazapfts.ok(oazapfts.fetchJson<{
        status: 200;
        data: SseEmitter;
    }>("/v2/ai/chat", oazapfts.json({
        ...opts,
        method: "POST",
        body: messageRequest,
        headers: oazapfts.mergeHeaders(opts?.headers, {
            Authorization: authorization
        })
    })));
}
export function createAgent({ adminAgentCreateRequest }: {
    adminAgentCreateRequest: AdminAgentCreateRequest;
}, opts?: Oazapfts.RequestOpts) {
    return oazapfts.ok(oazapfts.fetchJson<{
        status: 200;
        data: AdminAgentResponse;
    }>("/v2/admin/agents", oazapfts.json({
        ...opts,
        method: "POST",
        body: adminAgentCreateRequest
    })));
}
export function deactivateAgent({ schema, agentId }: {
    schema: string;
    agentId: string;
}, opts?: Oazapfts.RequestOpts) {
    return oazapfts.ok(oazapfts.fetchText(`/v2/admin/agents/${encodeURIComponent(schema)}/${encodeURIComponent(agentId)}/deactivate`, {
        ...opts,
        method: "PATCH"
    }));
}
export function activateAgent({ schema, agentId }: {
    schema: string;
    agentId: string;
}, opts?: Oazapfts.RequestOpts) {
    return oazapfts.ok(oazapfts.fetchJson<{
        status: 200;
        data: AdminAgentResponse;
    }>(`/v2/admin/agents/${encodeURIComponent(schema)}/${encodeURIComponent(agentId)}/activate`, {
        ...opts,
        method: "PATCH"
    }));
}
export function getAndCacheData({ key }: {
    key: string;
}, opts?: Oazapfts.RequestOpts) {
    return oazapfts.ok(oazapfts.fetchJson<{
        status: 200;
        data: {
            [key: string]: object;
        };
    }>(`/v2/redis/hello-world-persist-redis/${encodeURIComponent(key)}`, {
        ...opts
    }));
}
/**
 * Get specific Insight
 */
export function getInsightById({ insightId }: {
    insightId: string;
}, opts?: Oazapfts.RequestOpts) {
    return oazapfts.ok(oazapfts.fetchJson<{
        status: 200;
        data: GetInsightResponse;
    }>(`/v2/insights/${encodeURIComponent(insightId)}`, {
        ...opts
    }));
}
/**
 * Get specific Document content
 */
export function getContentById({ documentId }: {
    documentId: string;
}, opts?: Oazapfts.RequestOpts) {
    return oazapfts.ok(oazapfts.fetchJson<{
        status: 200;
        data: string;
    }>(`/v2/documents/${encodeURIComponent(documentId)}/content`, {
        ...opts
    }));
}
/**
 * Get chat artifacts
 */
export function getAll2({ chatId }: {
    chatId: string;
}, opts?: Oazapfts.RequestOpts) {
    return oazapfts.ok(oazapfts.fetchJson<{
        status: 200;
        data: GetArtifactResponse[];
    }>(`/v2/artifacts${QS.query(QS.explode({
        chatId
    }))}`, {
        ...opts
    }));
}
export function hello(opts?: Oazapfts.RequestOpts) {
    return oazapfts.ok(oazapfts.fetchJson<{
        status: 200;
        data: AdminHelloResponse;
    }>("/v2/admin/hello", {
        ...opts
    }));
}
export function listAgents({ schema }: {
    schema: string;
}, opts?: Oazapfts.RequestOpts) {
    return oazapfts.ok(oazapfts.fetchJson<{
        status: 200;
        data: AdminAgentResponse[];
    }>(`/v2/admin/agents/${encodeURIComponent(schema)}`, {
        ...opts
    }));
}
export function deleteAgent({ schema, agentId }: {
    schema: string;
    agentId: string;
}, opts?: Oazapfts.RequestOpts) {
    return oazapfts.ok(oazapfts.fetchText(`/v2/admin/agents/${encodeURIComponent(schema)}/${encodeURIComponent(agentId)}`, {
        ...opts,
        method: "DELETE"
    }));
}
