// @stack-spot/portal-network/src/api/workspace-ai.ts
/**
 * FastAPI
 * 0.1.0
 * DO NOT MODIFY - This file has been generated using oazapfts.
 * See https://www.npmjs.com/package/oazapfts
 */
import * as Oazapfts from "@oazapfts/runtime";
import * as QS from "@oazapfts/runtime/query";
export const defaults: Oazapfts.Defaults<Oazapfts.CustomHeaders> = {
    headers: {},
    baseUrl: "/",
};
const oazapfts = Oazapfts.runtime(defaults);
export const servers = {};
export type NewWorkspaceRequest = {
    name: string;
    slug: string;
    description?: string | null;
    logo?: string | null;
};
export type NewWorkspaceResponse = {
    id: string;
};
export type ValidationError = {
    loc: (string | number)[];
    msg: string;
    "type": string;
};
export type HttpValidationError = {
    detail?: ValidationError[];
};
export type WorkspaceVisibilityLevelEnum = "all" | "favorite";
export type OrderEnum = "a-to-z" | "z-to-a" | "oldest-first" | "newest-first";
export type WorkspaceResponse = {
    id: string;
    slug: string;
    name: string;
    description?: string | null;
    logo?: string | null;
    created_by: string;
};
export type UpdateWorkspaceRequest = {
    name: string;
    description?: string | null;
    logo?: string | null;
};
export type ShareableLinkResponse = {
    id: string;
    workspace_id: string;
    "type": string;
    created_by: string;
};
export type Action = "view" | "edit" | "grant_access";
export type PermissionListRequest = {
    identifier: string;
    action: Action;
};
export type PermissionType = "group" | "member";
export type WorkspacePermissionRequest = {
    permission_list: PermissionListRequest[];
    permission_type: PermissionType;
};
export type SortBy = "name" | "permission";
export type ContentType = "knowledge_source" | "quick_command" | "agent" | "stack" | "spot" | "toolkit";
export type AddWorkspaceContentRequest = {
    content_identifier: string[];
};
export type ListWksContentsResponse = {
    space_name: string;
    space_id: string;
    stacks?: {
        [key: string]: any;
    }[] | null;
    qcs?: {
        [key: string]: any;
    }[] | null;
    kss?: {
        [key: string]: any;
    }[] | null;
    agents?: {
        [key: string]: any;
    }[] | null;
    toolkits?: {
        [key: string]: any;
    }[] | null;
};
export type ActionShared = "VIEW" | "EDIT" | "GRANT_ACESS";
export type PermissionForSharedRequest = {
    userId: string;
    action: ActionShared;
    userName?: string | null;
};
export type PermissionListForSharedRequest = {
    users: PermissionForSharedRequest[];
};
export type ResourceMembersResponse = {
    id: string;
    username: string;
    email: string;
    name: string;
    actions: string[];
};
/**
 * Create Workspace
 */
export function createWorkspaceV1WorkspacesPost({ authorization, xAccountId, newWorkspaceRequest }: {
    authorization: string;
    xAccountId?: string | null;
    newWorkspaceRequest: NewWorkspaceRequest;
}, opts?: Oazapfts.RequestOpts) {
    return oazapfts.ok(oazapfts.fetchJson<{
        status: 201;
        data: NewWorkspaceResponse;
    } | {
        status: 404;
    } | {
        status: 422;
        data: HttpValidationError;
    }>("/v1/workspaces", oazapfts.json({
        ...opts,
        method: "POST",
        body: newWorkspaceRequest,
        headers: oazapfts.mergeHeaders(opts?.headers, {
            authorization,
            "x-account-id": xAccountId
        })
    })));
}
/**
 * List Workspaces
 */
export function listWorkspacesV1WorkspacesGet({ visibility, order, name, authorization, xAccountId }: {
    visibility?: WorkspaceVisibilityLevelEnum;
    order?: OrderEnum;
    name?: string | null;
    authorization: string;
    xAccountId?: string | null;
}, opts?: Oazapfts.RequestOpts) {
    return oazapfts.ok(oazapfts.fetchJson<{
        status: 200;
        data: WorkspaceResponse[];
    } | {
        status: 404;
    } | {
        status: 422;
        data: HttpValidationError;
    }>(`/v1/workspaces${QS.query(QS.explode({
        visibility,
        order,
        name
    }))}`, {
        ...opts,
        headers: oazapfts.mergeHeaders(opts?.headers, {
            authorization,
            "x-account-id": xAccountId
        })
    }));
}
/**
 * Get Workspace By Id
 */
export function getWorkspaceByIdV1WorkspacesIdGet({ id, authorization, xAccountId }: {
    id: string;
    authorization: string;
    xAccountId?: string | null;
}, opts?: Oazapfts.RequestOpts) {
    return oazapfts.ok(oazapfts.fetchJson<{
        status: 200;
        data: WorkspaceResponse;
    } | {
        status: 404;
    } | {
        status: 422;
        data: HttpValidationError;
    }>(`/v1/workspaces/${encodeURIComponent(id)}`, {
        ...opts,
        headers: oazapfts.mergeHeaders(opts?.headers, {
            authorization,
            "x-account-id": xAccountId
        })
    }));
}
/**
 * Update Workspace
 */
export function updateWorkspaceV1WorkspacesWorkspaceIdPatch({ workspaceId, authorization, xAccountId, updateWorkspaceRequest }: {
    workspaceId: string;
    authorization: string;
    xAccountId?: string | null;
    updateWorkspaceRequest: UpdateWorkspaceRequest;
}, opts?: Oazapfts.RequestOpts) {
    return oazapfts.ok(oazapfts.fetchJson<{
        status: 204;
    } | {
        status: 404;
    } | {
        status: 422;
        data: HttpValidationError;
    }>(`/v1/workspaces/${encodeURIComponent(workspaceId)}`, oazapfts.json({
        ...opts,
        method: "PATCH",
        body: updateWorkspaceRequest,
        headers: oazapfts.mergeHeaders(opts?.headers, {
            authorization,
            "x-account-id": xAccountId
        })
    })));
}
/**
 * Delete Workspace
 */
export function deleteWorkspaceV1WorkspacesWorkspaceIdDelete({ workspaceId, authorization, xAccountId }: {
    workspaceId: string;
    authorization: string;
    xAccountId?: string | null;
}, opts?: Oazapfts.RequestOpts) {
    return oazapfts.ok(oazapfts.fetchJson<{
        status: 204;
    } | {
        status: 404;
    } | {
        status: 422;
        data: HttpValidationError;
    }>(`/v1/workspaces/${encodeURIComponent(workspaceId)}`, {
        ...opts,
        method: "DELETE",
        headers: oazapfts.mergeHeaders(opts?.headers, {
            authorization,
            "x-account-id": xAccountId
        })
    }));
}
/**
 * Add Favorite
 */
export function addFavoriteV1WorkspacesWorkspaceIdFavoritePost({ workspaceId, authorization, xAccountId }: {
    workspaceId: string;
    authorization: string;
    xAccountId?: string | null;
}, opts?: Oazapfts.RequestOpts) {
    return oazapfts.ok(oazapfts.fetchJson<{
        status: 204;
    } | {
        status: 404;
    } | {
        status: 422;
        data: HttpValidationError;
    }>(`/v1/workspaces/${encodeURIComponent(workspaceId)}/favorite`, {
        ...opts,
        method: "POST",
        headers: oazapfts.mergeHeaders(opts?.headers, {
            authorization,
            "x-account-id": xAccountId
        })
    }));
}
/**
 * Delete Favorite
 */
export function deleteFavoriteV1WorkspacesWorkspaceIdFavoriteDelete({ workspaceId, authorization, xAccountId }: {
    workspaceId: string;
    authorization: string;
    xAccountId?: string | null;
}, opts?: Oazapfts.RequestOpts) {
    return oazapfts.ok(oazapfts.fetchJson<{
        status: 204;
    } | {
        status: 404;
    } | {
        status: 422;
        data: HttpValidationError;
    }>(`/v1/workspaces/${encodeURIComponent(workspaceId)}/favorite`, {
        ...opts,
        method: "DELETE",
        headers: oazapfts.mergeHeaders(opts?.headers, {
            authorization,
            "x-account-id": xAccountId
        })
    }));
}
/**
 * Create Shareable Link
 */
export function createShareableLinkV1WorkspacesWorkspaceIdSharePost({ workspaceId, authorization, xAccountId }: {
    workspaceId: string;
    authorization: string;
    xAccountId?: string | null;
}, opts?: Oazapfts.RequestOpts) {
    return oazapfts.ok(oazapfts.fetchJson<{
        status: 201;
        data: ShareableLinkResponse;
    } | {
        status: 404;
    } | {
        status: 422;
        data: HttpValidationError;
    }>(`/v1/workspaces/${encodeURIComponent(workspaceId)}/share`, {
        ...opts,
        method: "POST",
        headers: oazapfts.mergeHeaders(opts?.headers, {
            authorization,
            "x-account-id": xAccountId
        })
    }));
}
/**
 * Add Permission
 */
export function addPermissionV1WorkspacesWorkspaceIdPermissionsPatch({ workspaceId, authorization, xAccountId, workspacePermissionRequest }: {
    workspaceId: string;
    authorization: string;
    xAccountId?: string | null;
    workspacePermissionRequest: WorkspacePermissionRequest;
}, opts?: Oazapfts.RequestOpts) {
    return oazapfts.ok(oazapfts.fetchJson<{
        status: 204;
    } | {
        status: 404;
    } | {
        status: 422;
        data: HttpValidationError;
    }>(`/v1/workspaces/${encodeURIComponent(workspaceId)}/permissions`, oazapfts.json({
        ...opts,
        method: "PATCH",
        body: workspacePermissionRequest,
        headers: oazapfts.mergeHeaders(opts?.headers, {
            authorization,
            "x-account-id": xAccountId
        })
    })));
}
/**
 * Remove Permission
 */
export function removePermissionV1WorkspacesWorkspaceIdPermissionsDelete({ workspaceId, authorization, xAccountId, workspacePermissionRequest }: {
    workspaceId: string;
    authorization: string;
    xAccountId?: string | null;
    workspacePermissionRequest: WorkspacePermissionRequest;
}, opts?: Oazapfts.RequestOpts) {
    return oazapfts.ok(oazapfts.fetchJson<{
        status: 204;
    } | {
        status: 404;
    } | {
        status: 422;
        data: HttpValidationError;
    }>(`/v1/workspaces/${encodeURIComponent(workspaceId)}/permissions`, oazapfts.json({
        ...opts,
        method: "DELETE",
        body: workspacePermissionRequest,
        headers: oazapfts.mergeHeaders(opts?.headers, {
            authorization,
            "x-account-id": xAccountId
        })
    })));
}
/**
 * List Groups And Members Permission In Workspace
 */
export function listGroupsAndMembersPermissionInWorkspaceV1WorkspacesWorkspaceIdPermissionsGet({ workspaceId, search, filterBy, filterValue, sortBy, page, size, order, authorization, xAccountId }: {
    workspaceId: string;
    search?: string | null;
    filterBy?: string | null;
    filterValue?: string | null;
    sortBy?: SortBy | null;
    page?: number;
    size?: number;
    order?: OrderEnum;
    authorization: string;
    xAccountId?: string | null;
}, opts?: Oazapfts.RequestOpts) {
    return oazapfts.ok(oazapfts.fetchJson<{
        status: 200;
        data: {
            [key: string]: any;
        };
    } | {
        status: 404;
    } | {
        status: 422;
        data: HttpValidationError;
    }>(`/v1/workspaces/${encodeURIComponent(workspaceId)}/permissions${QS.query(QS.explode({
        search,
        filter_by: filterBy,
        filter_value: filterValue,
        sort_by: sortBy,
        page,
        size,
        order
    }))}`, {
        ...opts,
        headers: oazapfts.mergeHeaders(opts?.headers, {
            authorization,
            "x-account-id": xAccountId
        })
    }));
}
/**
 * Add Content
 */
export function addContentV1WorkspacesWorkspaceIdContentTypePost({ workspaceId, contentType, authorization, xAccountId, addWorkspaceContentRequest }: {
    workspaceId: string;
    contentType: ContentType;
    authorization: string;
    xAccountId?: string | null;
    addWorkspaceContentRequest: AddWorkspaceContentRequest;
}, opts?: Oazapfts.RequestOpts) {
    return oazapfts.ok(oazapfts.fetchJson<{
        status: 200;
        data: {
            [key: string]: any;
        };
    } | {
        status: 404;
    } | {
        status: 422;
        data: HttpValidationError;
    }>(`/v1/workspaces/${encodeURIComponent(workspaceId)}/${encodeURIComponent(contentType)}`, oazapfts.json({
        ...opts,
        method: "POST",
        body: addWorkspaceContentRequest,
        headers: oazapfts.mergeHeaders(opts?.headers, {
            authorization,
            "x-account-id": xAccountId
        })
    })));
}
/**
 * List All By Content Type
 */
export function listAllByContentTypeV1WorkspacesWorkspaceIdContentTypeGet({ workspaceId, contentType, name, authorization, xAccountId }: {
    workspaceId: string;
    contentType: ContentType;
    name?: string | null;
    authorization: string;
    xAccountId?: string | null;
}, opts?: Oazapfts.RequestOpts) {
    return oazapfts.ok(oazapfts.fetchJson<{
        status: 200;
        data: any;
    } | {
        status: 404;
    } | {
        status: 422;
        data: HttpValidationError;
    }>(`/v1/workspaces/${encodeURIComponent(workspaceId)}/${encodeURIComponent(contentType)}${QS.query(QS.explode({
        name
    }))}`, {
        ...opts,
        headers: oazapfts.mergeHeaders(opts?.headers, {
            authorization,
            "x-account-id": xAccountId
        })
    }));
}
/**
 * Remove Content
 */
export function removeContentV1WorkspacesWorkspaceIdContentTypeContentIdDelete({ workspaceId, contentType, contentId, authorization, xAccountId }: {
    workspaceId: string;
    contentType: ContentType;
    contentId: string;
    authorization: string;
    xAccountId?: string | null;
}, opts?: Oazapfts.RequestOpts) {
    return oazapfts.ok(oazapfts.fetchJson<{
        status: 204;
    } | {
        status: 404;
    } | {
        status: 422;
        data: HttpValidationError;
    }>(`/v1/workspaces/${encodeURIComponent(workspaceId)}/${encodeURIComponent(contentType)}/${encodeURIComponent(contentId)}`, {
        ...opts,
        method: "DELETE",
        headers: oazapfts.mergeHeaders(opts?.headers, {
            authorization,
            "x-account-id": xAccountId
        })
    }));
}
/**
 * List All Contents
 */
export function listAllContentsV1WorkspacesContentsContentTypeGet({ contentType, name, authorization, xAccountId }: {
    contentType: ContentType;
    name?: string | null;
    authorization: string;
    xAccountId?: string | null;
}, opts?: Oazapfts.RequestOpts) {
    return oazapfts.ok(oazapfts.fetchJson<{
        status: 200;
        data: ListWksContentsResponse[];
    } | {
        status: 404;
    } | {
        status: 422;
        data: HttpValidationError;
    }>(`/v1/workspaces/contents/${encodeURIComponent(contentType)}${QS.query(QS.explode({
        name
    }))}`, {
        ...opts,
        headers: oazapfts.mergeHeaders(opts?.headers, {
            authorization,
            "x-account-id": xAccountId
        })
    }));
}
/**
 * List All Contents
 */
export function listAllContentsV1WorkspacesWorkspaceIdContentsGet({ workspaceId, name, authorization, xAccountId }: {
    workspaceId: string;
    name?: string | null;
    authorization: string;
    xAccountId?: string | null;
}, opts?: Oazapfts.RequestOpts) {
    return oazapfts.ok(oazapfts.fetchJson<{
        status: 200;
        data: any;
    } | {
        status: 404;
    } | {
        status: 422;
        data: HttpValidationError;
    }>(`/v1/workspaces/${encodeURIComponent(workspaceId)}/contents${QS.query(QS.explode({
        name
    }))}`, {
        ...opts,
        headers: oazapfts.mergeHeaders(opts?.headers, {
            authorization,
            "x-account-id": xAccountId
        })
    }));
}
/**
 * Use Shareable Link
 */
export function useShareableLinkV1ShareShareIdPatch({ shareId, authorization, xAccountId }: {
    shareId: string;
    authorization: string;
    xAccountId?: string | null;
}, opts?: Oazapfts.RequestOpts) {
    return oazapfts.ok(oazapfts.fetchJson<{
        status: 200;
        data: ShareableLinkResponse;
    } | {
        status: 404;
    } | {
        status: 422;
        data: HttpValidationError;
    }>(`/v1/share/${encodeURIComponent(shareId)}`, {
        ...opts,
        method: "PATCH",
        headers: oazapfts.mergeHeaders(opts?.headers, {
            authorization,
            "x-account-id": xAccountId
        })
    }));
}
/**
 * Add Permission
 */
export function addPermissionV1ResourceTypeResourceTypeResourcesResourceIdentifierPatch({ resourceType, resourceIdentifier, authorization, xAccountId, permissionListForSharedRequest }: {
    resourceType: ContentType;
    resourceIdentifier: string;
    authorization: string;
    xAccountId?: string | null;
    permissionListForSharedRequest: PermissionListForSharedRequest;
}, opts?: Oazapfts.RequestOpts) {
    return oazapfts.ok(oazapfts.fetchJson<{
        status: 204;
    } | {
        status: 404;
    } | {
        status: 422;
        data: HttpValidationError;
    }>(`/v1/resource-type/${encodeURIComponent(resourceType)}/resources/${encodeURIComponent(resourceIdentifier)}`, oazapfts.json({
        ...opts,
        method: "PATCH",
        body: permissionListForSharedRequest,
        headers: oazapfts.mergeHeaders(opts?.headers, {
            authorization,
            "x-account-id": xAccountId
        })
    })));
}
/**
 * Remove Permission
 */
export function removePermissionV1ResourceTypeResourceTypeResourcesResourceIdentifierDelete({ resourceType, resourceIdentifier, authorization, xAccountId, permissionListForSharedRequest }: {
    resourceType: ContentType;
    resourceIdentifier: string;
    authorization: string;
    xAccountId?: string | null;
    permissionListForSharedRequest: PermissionListForSharedRequest;
}, opts?: Oazapfts.RequestOpts) {
    return oazapfts.ok(oazapfts.fetchJson<{
        status: 204;
    } | {
        status: 404;
    } | {
        status: 422;
        data: HttpValidationError;
    }>(`/v1/resource-type/${encodeURIComponent(resourceType)}/resources/${encodeURIComponent(resourceIdentifier)}`, oazapfts.json({
        ...opts,
        method: "DELETE",
        body: permissionListForSharedRequest,
        headers: oazapfts.mergeHeaders(opts?.headers, {
            authorization,
            "x-account-id": xAccountId
        })
    })));
}
/**
 * List Members
 */
export function listMembersV1ResourceTypeResourceTypeResourcesResourceIdentifierMembersGet({ resourceType, resourceIdentifier, search, filterBy, filterValue, sortBy, page, size, order, authorization, xAccountId }: {
    resourceType: ContentType;
    resourceIdentifier: string;
    search?: string | null;
    filterBy?: string | null;
    filterValue?: string | null;
    sortBy?: SortBy | null;
    page?: number;
    size?: number;
    order?: OrderEnum;
    authorization: string;
    xAccountId?: string | null;
}, opts?: Oazapfts.RequestOpts) {
    return oazapfts.ok(oazapfts.fetchJson<{
        status: 200;
        data: ResourceMembersResponse[];
    } | {
        status: 404;
    } | {
        status: 422;
        data: HttpValidationError;
    }>(`/v1/resource-type/${encodeURIComponent(resourceType)}/resources/${encodeURIComponent(resourceIdentifier)}/members${QS.query(QS.explode({
        search,
        filter_by: filterBy,
        filter_value: filterValue,
        sort_by: sortBy,
        page,
        size,
        order
    }))}`, {
        ...opts,
        headers: oazapfts.mergeHeaders(opts?.headers, {
            authorization,
            "x-account-id": xAccountId
        })
    }));
}
/**
 * Healthz
 */
export function healthzHealthzGet(opts?: Oazapfts.RequestOpts) {
    return oazapfts.ok(oazapfts.fetchJson<{
        status: 200;
        data: any;
    }>("/healthz", {
        ...opts
    }));
}
/**
 * Base Get
 */
export function baseGetGet(opts?: Oazapfts.RequestOpts) {
    return oazapfts.ok(oazapfts.fetchJson<{
        status: 200;
        data: any;
    }>("/", {
        ...opts
    }));
}
