// @stack-spot/portal-network/src/api/workspaceManager.ts
/**
 * Workspace Manager API
 * v1.0.0
 * DO NOT MODIFY - This file has been generated using oazapfts.
 * See https://www.npmjs.com/package/oazapfts
 */
import * as Oazapfts from "@oazapfts/runtime";
import * as QS from "@oazapfts/runtime/query";
export const defaults: Oazapfts.Defaults<Oazapfts.CustomHeaders> = {
    headers: {},
    baseUrl: "https://workspace-workspace-manager.stg.stackspot.com",
};
const oazapfts = Oazapfts.runtime(defaults);
export const servers = {
    generatedServerUrl: "https://workspace-workspace-manager.stg.stackspot.com"
};
export type WorkspaceVariableResponse = {
    /** Workspace variable name. */
    name: string;
    /** Workspace variable value. */
    value?: string;
    /** Workspace variable description. */
    description: string;
    /** Workspace variable mandate flag. */
    mandate: boolean;
    /** Variable value source. */
    source: string;
    /** Workspace variable creation data */
    createdAt: string;
};
export type UpsertWorkspaceVariableRequest = {
    /** Workspace variable value. */
    value: string;
};
export type ExternalItemsResponse = {
    source: string;
    value: string;
    label?: string;
};
export type InputConditionResponse = {
    variable: string;
    operator: string;
    value?: any;
};
export type ActionConsolidatedAttributeResponse = {
    key?: string;
    value?: any;
    final: boolean;
    label: string;
    "type": string;
    required: boolean;
    help?: string;
    pattern?: string;
    "default"?: any;
    items?: string[];
    externalItems?: ExternalItemsResponse;
    condition?: InputConditionResponse;
    source: "ACCOUNT" | "STACK" | "WORKSPACE";
    readOnly: boolean;
    connectionInterfaceType?: string;
    addQuestion?: string;
    input?: any;
    inputs: any[];
};
export type ComputedInputResponse = {
    name: string;
    expression: string;
};
export type WorkflowActionWithContextResponse = {
    actionId: string;
    versionRange: string;
    name: string;
    stepType?: "CREATE_REPO";
    valuesContext: ActionConsolidatedAttributeResponse[];
    computedInputs: ComputedInputResponse[];
    globalComputedInputs: ComputedInputResponse[];
};
export type WorkflowAccountContextResponse = {
    actionsBefore: WorkflowActionWithContextResponse[];
    actionsAfter: WorkflowActionWithContextResponse[];
};
export type ContextAttribute = {
    key: string;
    value: any;
    final: boolean;
    sourceType?: "ACCOUNT" | "STACK" | "WORKSPACE";
};
export type ContextWorkflowRequest = {
    actionId: string;
    versionRange: string;
    valuesContext: ContextAttribute[];
};
export type WorkflowContextSaveRequest = {
    actionsBefore: ContextWorkflowRequest[];
    actionsAfter: ContextWorkflowRequest[];
};
export type AddStackInWorkspaceRequest = {
    /** Stack version id. */
    stackVersionId: string;
};
export type ConnectionInterfaceAttribute = {
    /** Attribute key. */
    key: string;
    /** Attribute value. */
    value: any;
    /** Indicates if this attribute value is sensitive or not. */
    sensitive: boolean;
    /** Indicates the outuput id for the respective attribute and is used for validation. */
    id: string;
};
export type ConnectionInterfaceAttributesByEnv = {
    /** Environment id. */
    environmentId: string;
    /** List of attributes */
    attributes: ConnectionInterfaceAttribute[];
};
export type ValueByEnvResponse = {
    envName: string;
    value: any;
    mandate: boolean;
    source: string;
};
export type InputValuesContextResponse = {
    name?: string;
    label: string;
    "type": string;
    items?: string[];
    externalItems?: ExternalItemsResponse;
    inputEnv: boolean;
    "default"?: any;
    required?: boolean;
    pattern?: string;
    help?: string;
    valuesByEnv: ValueByEnvResponse[];
    connectionInterfaceType?: string;
    addQuestion?: string;
    input?: any;
    inputs: any[];
    condition?: InputConditionResponse;
    isGlobal?: boolean;
    scope?: string;
};
export type FullInputContextResponse = {
    inputs: InputValuesContextResponse[];
    computedInputs: ComputedInputResponse[];
    globalComputedInputs: ComputedInputResponse[];
};
export type UpsertInputContextRequest = {
    /** Context's environment (omit or null for all in account) */
    envName?: string;
    key: string;
    value: any;
    mandate?: boolean;
};
export type ValueByEnvWorkspaceResponse = {
    envName: string;
    value: any;
    mandate: boolean;
    source: string;
};
export type PluginInputValuesInConsolidatedContextResponse = {
    name: string;
    label: string;
    "type": string;
    inputEnv: boolean;
    required?: boolean;
    "default"?: any;
    pattern?: string;
    help?: string;
    items?: string[];
    itemsValues?: string[];
    externalItems?: ExternalItemsResponse;
    valuesByEnv: ValueByEnvWorkspaceResponse[];
    condition?: InputConditionResponse;
    isGlobal?: boolean;
    connectionInterfaceType?: string;
    addQuestion?: string;
    input?: any;
    inputs: any[];
};
export type UpsertPluginInputContextRequest = {
    /** Context's environment (omit or null for all in account) */
    envName?: string;
    key: string;
    value: any;
    mandate?: boolean;
};
export type UpsertActionInputContextRequest = {
    /** Context's environment (omit or null for all in account) */
    envName?: string;
    key: string;
    value: any;
    mandate?: boolean;
};
export type ValueByEnvAccountResponse = {
    envName: string;
    value: any;
    mandate: boolean;
};
export type PluginInputValuesInAccountContextResponse = {
    name: string;
    label: string;
    "type": string;
    inputEnv: boolean;
    required?: boolean;
    "default"?: any;
    pattern?: string;
    help?: string;
    items?: string[];
    externalItems?: ExternalItemsResponse;
    valuesByEnv: ValueByEnvAccountResponse[];
    condition?: InputConditionResponse;
    isGlobal?: boolean;
    connectionInterfaceType?: string;
    addQuestion?: string;
    input?: any;
    inputs: any[];
};
export type StudioResponse = {
    slug: string;
    name: string;
    imageUrl?: string;
};
export type SimpleStackResponse = {
    stackId: string;
    name: string;
    displayName: string;
    slug: string;
    imageUrl?: string;
    studio: StudioResponse;
};
export type WorkflowAccountSettingsResponse = {
    name: string;
    description: string;
    "type": "CREATE_API" | "CREATE_APP" | "CREATE_INFRA";
    isDefault: boolean;
    stacks: SimpleStackResponse[];
};
export type WorkflowSettingsSaveRequest = {
    name: string;
    description: string;
    "type": "CREATE_API" | "CREATE_APP" | "CREATE_INFRA";
    isDefault: boolean;
    stacksId: string[];
};
export type ActionResponse = {
    actionId: string;
    name: string;
    order: number;
    versionRange: string;
    stepType?: "CREATE_REPO";
};
export type WorkflowActionResponse = {
    actionsBefore: ActionResponse[];
    actionsAfter: ActionResponse[];
};
export type WorkflowActionSimpleRequest = {
    actionId: string;
    order: number;
    versionRange: string;
    stepType?: "CREATE_REPO";
};
export type WorkflowActionsRequest = {
    actionsBefore: WorkflowActionSimpleRequest[];
    actionsAfter: WorkflowActionSimpleRequest[];
};
export type AccountVariableResponse = {
    /** Account variable name */
    name: string;
    /** Account variable value stored */
    value?: string;
    /** Account variable description */
    description: string;
    /** Account variable mandatory flag */
    mandate: boolean;
    /** Account variable creation data */
    createdAt: string;
};
export type UpdateAccountVariableRequest = {
    /** New account variable value */
    value?: string;
    /** New account variable description */
    description: string;
    /** New account variable mandatory flag */
    mandate: boolean;
};
export type RequiresRequest = {
    /** Connections requires name. */
    selected: string;
    /** Connections requires alias. */
    alias: string;
    /** Connections requires type. */
    "type": string;
};
export type GeneratesRequest = {
    /** Connections generates name. */
    selected: string;
    /** Connections generates alias. */
    alias: string;
    /** Connections generates type. */
    "type": string;
};
export type PluginsConnectionsRequest = {
    /** List of connections requires. */
    requires?: RequiresRequest[];
    /** List of connections generates. */
    generates?: GeneratesRequest[];
};
export type PluginRequest = {
    /** Name used in workflow. */
    name: string;
    alias: string;
    pluginVersionId?: string;
    /** Inputs used in workflow. */
    inputs: {
        [key: string]: any;
    };
    /** Inputs sensitives used in workflow. */
    inputsSensitive?: string[];
    /** Inputs Env used in workflow. */
    inputsEnv: {
        [key: string]: {
            [key: string]: any;
        };
    };
    /** Connections used in workflow. */
    connections?: PluginsConnectionsRequest;
};
export type Infra = {
    /** Shared Infrastructure apiId that will be used in case of retry. */
    sharedInfraId?: string;
    stack: string;
    /** Stack used to generate this shared infra. */
    stackVersionId: string;
    /** SharedInfra name. */
    name: string;
    /** SharedInfra repository url. */
    repo?: string;
    /** SharedInfra description. */
    description?: string;
    /** SharedInfra repository base branch. */
    repoBaseBranch: string;
    /** Starter Id used to generate this shared infra. */
    starterId?: string;
    /** Starter used to generate this shared infra. */
    starter: string;
    plugins: PluginRequest[];
};
export type ActionsConnectionsRequest = {
    /** List of connections requires. */
    requires: RequiresRequest[];
};
export type ActionsBeforeRequest = {
    /** Actions before name. */
    name: string;
    /** Actions before inputs. */
    inputs: {
        [key: string]: any;
    };
    /** Actions before inputs sensitive. */
    inputsSensitive?: string[];
    /** Create repo?. */
    repoCreate: boolean;
    /** Actions before connections. */
    connections?: ActionsConnectionsRequest;
};
export type ActionsAfterRequest = {
    /** Actions after name. */
    name: string;
    /** Actions after inputs. */
    inputs: {
        [key: string]: any;
    };
    /** Actions after inputs sensitive. */
    inputsSensitive?: string[];
    /** Create repo?. */
    repoCreate: boolean;
    /** Actions after connections. */
    connections?: ActionsConnectionsRequest;
};
export type CreateSharedRequest = {
    infra: Infra;
    actionsBefore?: ActionsBeforeRequest[];
    actionsAfter?: ActionsAfterRequest[];
    /** Data that will be used in case of retry. */
    frontData?: any;
};
export type CreateShareInfraResponse = {
    sharedInfraId: string;
    workflowId: string;
    logs?: string;
};
export type App = {
    /** Application apiId that will be used in case of retry. */
    applicationId?: string;
    /** Stack used to generate this application. */
    stackVersionId: string;
    /** Starter used to generate this application. */
    starter: string;
    /** Starter Id used to generate this application. */
    starterId: string;
    /** Application name. */
    name: string;
    /** Application repository url. */
    repo?: string;
    /** Application repository base branch. */
    repoBaseBranch: string;
    /** Application description. */
    description?: string;
    plugins: PluginRequest[];
};
export type Definitions = {
    environmentId: string;
    gatewayUrl: string;
    openApiContent: string;
};
export type ApiRequest = {
    /** Name used in create api. */
    name: string;
    /** Description used in create api. */
    description: string;
    connectionInterfaceSlug?: string;
    /** Description used in create api. */
    definitions: Definitions[];
};
export type CreateAppRequest = {
    /** Information for create application. */
    app: App;
    /** Before Actions for execute in workflow . */
    actionsBefore?: ActionsBeforeRequest[];
    /** After Actions for execute in workflow . */
    actionsAfter?: ActionsAfterRequest[];
    /** Information for create api. */
    api?: ApiRequest;
    /** Data that will be used in case of retry. */
    frontData?: any;
};
export type CreateApplicationResponse = {
    applicationId: string;
    apiId?: string;
    workflowId: string;
    logs?: string;
};
export type ContentWorkflowSaveRequest = {
    name: string;
    description: string;
    targetId?: string;
    stackVersionId: string;
    inputs: {
        [key: string]: any;
    };
    frontData?: any;
};
export type ContentWorkflowSaveResponse = {
    targetId: string;
    executionId: string;
};
export type StackVersionLatestAvailableResponse = {
    stackVersionId: string;
    semanticVersion: string;
};
export type WorkspaceStackResponseStackResponse = {
    id: string;
    version: number;
    displayName: string;
    imageUrl?: string;
    slug: string;
    description: string;
    studio: StudioResponse;
    latestVersionAvailable?: StackVersionLatestAvailableResponse;
};
export type WorkspaceStackResponse = {
    stackVersionId: string;
    semanticVersion?: string;
    copyFromStackVersionId?: string;
    status: string;
    hasDeprecatedContent: boolean;
    workspaceId: string;
    accountId: string;
    qualifier: string;
    stack: WorkspaceStackResponseStackResponse;
};
export type RemovePluginAndCreatePrRequest = {
    originBranchName: string;
};
export type ExtraLinkIntegrationResponse = {
    url: string;
    title: string;
    sourceType: "APP_DEPLOY" | "INFRA_DEPLOY";
};
export type ExtraLinkRequest = {
    url: string;
    title: string;
};
export type InfraPluginRequest = {
    pluginVersionId: string;
    stackVersionId: string;
    alias: string;
    inputs: {
        [key: string]: any;
    };
    inputsEnv: {
        [key: string]: {
            [key: string]: any;
        };
    };
    connections: PluginsConnectionsRequest;
};
export type UpdatePluginAndCreatePrRequest = {
    originBranchName: string;
    newBranchName: string;
    stkFilePath: string;
    plugins: InfraPluginRequest[];
};
export type CreateConnectionInterfaceRequest = {
    /** Connection interface type id. */
    typeId: string;
    /** Connection interface id generated by client. */
    connectionInterfaceId: string;
    /** Connection interface owner applications. */
    applicationId?: string;
    /** Connection interface owner shared infra. */
    sharedInfraId?: string;
    /** Version id from plugin that generated this connection interface. */
    sourcePluginVersionId?: string;
    /** Indicates if this connection interface was generated by a plugin or manually. */
    automaticallyGenerated: boolean;
    /** Connection interface attributes. */
    attributesByEnv: ConnectionInterfaceAttributesByEnv[];
};
export type IdResponse = {
    id: string;
};
export type ApplicationDestroyRequest = {
    /** Destroy status. */
    status: "SUCCESS" | "FAILED";
};
export type WorkspaceResponse = {
    /** Workspace id. */
    id: string;
    /** Workspace name */
    name: string;
    /** Workspace description */
    description?: string;
    /** Workspace image */
    imageUrl?: string;
};
export type ImportContextRequest = {
    stackVersionId: string;
    stackVersionReference: string;
};
export type WorkflowActionWithContextRequest = {
    actionId: string;
    order: number;
    versionRange: string;
    stepType?: "CREATE_REPO";
    valuesContext: ContextAttribute[];
};
export type AccountWorkflowCreateRequest = {
    name: string;
    description: string;
    "type": "CREATE_API" | "CREATE_APP" | "CREATE_INFRA";
    isDefault: boolean;
    stacksId: string[];
    actionsBefore: WorkflowActionWithContextRequest[];
    actionsAfter: WorkflowActionWithContextRequest[];
};
export type IdResponseString = {
    id: string;
};
export type PaginatedAccountVariableResponse = {
    /** Current page requested */
    currentPage: number;
    /** Last page available */
    lastPage: number;
    /** Page size requested */
    pageSize: number;
    /** Total items found */
    totalItems: number;
    /** Account variables for current page */
    items: AccountVariableResponse[];
};
export type CreateAccountVariableRequest = {
    /** Account variable name */
    name: string;
    /** Account variable value */
    value?: string;
    /** Account variable description */
    description: string;
    /** Account variable mandatory flag */
    mandate: boolean;
};
export type ManagerRunResponse = {
    environmentName?: string;
    runId: string;
};
export type ConnectionInterfaceAttributeResponse = {
    id?: string;
    key: string;
    value: any;
    sensitive: boolean;
    schema?: any;
};
export type AppliedPluginResponseConnectionInterfacesIdResponse = {
    id: string;
    slug: string;
    connectionInterfaceId: string;
    typeId: string;
    alias?: string;
    typeName: string;
    attributes: ConnectionInterfaceAttributeResponse[];
};
export type AppliedPluginResponseActionsResponseConnectionInterfaceResponse = {
    typeId?: string;
    typeName: string;
    alias: string;
    optional?: boolean;
    source?: string;
};
export type AppliedPluginResponseActionsResponseRequires = {
    connectionInterfaces: AppliedPluginResponseActionsResponseConnectionInterfaceResponse[];
};
export type AppliedPluginResponseActionsResponse = {
    versionId: string;
    qualifier: string;
    description: string;
    slug: string;
    displayName: string;
    requires: AppliedPluginResponseActionsResponseRequires;
};
export type AppliedPluginResponseLink = {
    name: string;
    url: string;
    imageUrl?: string;
};
export type AppliedPluginResponse = {
    alias: string;
    versionId: string;
    qualifier: string;
    slug: string;
    description: string;
    displayName: string;
    "type": string;
    status: string;
    deployStatus: string;
    inCloud?: boolean;
    fork?: boolean;
    iacPath?: string;
    singleUse?: boolean;
    attributes?: {
        [key: string]: any;
    };
    connectionInterfaces?: AppliedPluginResponseConnectionInterfacesIdResponse[];
    actions?: AppliedPluginResponseActionsResponse[];
    links?: AppliedPluginResponseLink[];
    resourcesProvided?: string[];
};
export type SharedInfraAppliedPluginsV2Response = {
    plugins: AppliedPluginResponse[];
};
export type ActivityResponse = {
    title: string;
    owner?: string;
    status: "SUCCESS" | "FAILED" | "RUNNING" | "PENDING" | "SKIPPED" | "CANCELLED" | "NONE" | "DRIFT" | "SUCCEEDED" | "RUNTIME_ERROR" | "USER_ERROR" | "INTERNAL_ERROR" | "ABORTED" | "ABORTING" | "EXTERNAL_ERROR" | "READY_TO_RUN" | "TIMEOUT";
    source: "STACK" | "RUNTIME" | "WORKFLOW" | "SCM";
    "type": "DEPLOY" | "DESTROY" | "ROLLBACK" | "CREATE" | "ACTION" | "CUSTOMER_WORKFLOW";
    action?: "CREATE" | "UPDATE" | "DELETE" | "RUN";
    requestedBy?: string;
    time: number;
    version: string;
    detail: {
        [key: string]: any;
    };
    jobsCount: number;
};
export type PaginatedActivityResponse = {
    items: ActivityResponse[];
    currentPage: number;
    totalItems: number;
    totalPages: number;
};
export type ApplicationAppliedPluginsV2Response = {
    plugins: AppliedPluginResponse[];
};
export type WorkspaceContextV2ResponseWorkspacePluginAttributeContext = {
    /** Key identifier of an input */
    key: string;
    /** Value of an input */
    value: any;
    /** Identifies whether this attribute can be overwritten or not */
    mandate: boolean;
    /** Source */
    source: string;
};
export type WorkspaceContextV2ResponseEnvironment = {
    /** Identifier of a environment */
    environmentId: string;
    /** name of a environment */
    environmentName?: string;
    /** Attributes environments */
    attributes: WorkspaceContextV2ResponseWorkspacePluginAttributeContext[];
};
export type WorkspaceContextV2ResponseConnectionInterfaceRequired = {
    /** Connection interface type required */
    typeName: string;
    /** Identifier of the pluginVersionId that should be used for the connection interface type */
    pluginVersionId?: string;
    alias?: string;
};
export type WorkspaceContextV2ResponsePlugin = {
    /** Identifier of a plugin version */
    pluginVersionId: string;
    /** Plugin slugs name with format `accountSlug/studioSlug/pluginSlug@x.y.z` */
    pluginQualifier: string;
    /** Plugin type */
    pluginType: string;
    /** Status Plugin */
    status: string;
    /** Attributes show by environments */
    environments: WorkspaceContextV2ResponseEnvironment[];
    /** Connection interfaces required */
    connectionInterfaceRequired: WorkspaceContextV2ResponseConnectionInterfaceRequired[];
};
export type WorkspaceContextV2ResponseConnectionInterface = {
    /** Connection interface type */
    typeName: string;
    /** Connection interface id. */
    id?: string;
    /** Connection interface id generated by client. */
    connectionInterfaceId: string;
    /** Automatically generated */
    automaticallyGenerated: boolean;
};
export type WorkspaceContextV2Response = {
    /** Cloud accounts information */
    cloudAccounts: any[];
    /** Context information for each environment and plugin in stack */
    context: WorkspaceContextV2ResponsePlugin[];
    /** Workspace Connection Interfaces information */
    connectionInterfaces: WorkspaceContextV2ResponseConnectionInterface[];
};
export type PluginForAppCreationV2ResponseRequiredConnectionInterfaceByTypeAvailableConnectionInterface = {
    id: string;
    slug: string;
    sourcePluginUri?: string;
};
export type PluginForAppCreationV2ResponseRequiredConnectionInterfaceByTypeConnectionInterfaceInput = {
    name: string;
    "type": "BOOLEAN" | "NUMBER" | "STRING" | "SECRET" | "ARRAY" | "OBJECT";
};
export type PluginForAppCreationV2ResponseRequiredConnectionInterfaceByType = {
    typeId: string;
    typeName: string;
    alias: string;
    available: PluginForAppCreationV2ResponseRequiredConnectionInterfaceByTypeAvailableConnectionInterface[];
    inputDefinitions: PluginForAppCreationV2ResponseRequiredConnectionInterfaceByTypeConnectionInterfaceInput[];
};
export type PluginForAppCreationV2ResponseGeneratesConnectionInterfaceByType = {
    typeId: string;
    typeName: string;
    alias: string;
};
export type PluginForAppCreationV2Response = {
    displayName: string;
    status: string;
    uri: string;
    version: string;
    singleUse?: boolean;
    inputs: PluginInputValuesInConsolidatedContextResponse[];
    computedInputs: ComputedInputResponse[];
    globalComputedInputs: ComputedInputResponse[];
    requiredConnectionInterfacesByType: PluginForAppCreationV2ResponseRequiredConnectionInterfaceByType[];
    generatesConnectionInterfacesByType: PluginForAppCreationV2ResponseGeneratesConnectionInterfaceByType[];
};
export type WorkspacePageResponse = {
    currentPage: number;
    pageSize: number;
    lastPage: number;
    totalItems: number;
    totalPages: number;
    items: WorkspaceResponse[];
};
export type PaginatedWorkspaceVariableResponse = {
    currentPage: number;
    pageSize: number;
    lastPage: number;
    totalItems: number;
    items: WorkspaceVariableResponse[];
};
export type WorkflowCompleteResponse = {
    workflowId: string;
    name: string;
    description: string;
    "type": "CREATE_API" | "CREATE_APP" | "CREATE_INFRA";
    isDefault: boolean;
    actionsBefore: WorkflowActionWithContextResponse[];
    actionsAfter: WorkflowActionWithContextResponse[];
};
export type StackVersionResponse = {
    stackVersionId: string;
    version: number;
    status: string;
    hasDeprecatedContent: boolean;
    published: boolean;
    semanticVersion?: string;
    copyFromStackVersionId?: string;
    hasContext?: boolean;
};
export type WorkspaceStackContextResponseVersionStackResponse = {
    id: string;
    version: number;
    displayName: string;
    imageUrl?: string;
    slug: string;
    description: string;
    studio: StudioResponse;
    versions: StackVersionResponse[];
};
export type WorkspaceStackContextResponse = {
    stackVersionId: string;
    semanticVersion?: string;
    copyFromStackVersionId?: string;
    status: string;
    hasDeprecatedContent: boolean;
    hasVersionsWithContext: boolean;
    workspaceId: string;
    accountId: string;
    qualifier: string;
    hasContext: boolean;
    stack: WorkspaceStackContextResponseVersionStackResponse;
};
export type SimpleStackVersionResponse = {
    stackVersionId: string;
    semanticVersion: string;
};
export type SharedInfraSimpleResponse = {
    id: string;
    name: string;
    createdAt: string;
    stack?: SimpleStackResponse;
    status: string;
    stackVersion?: SimpleStackVersionResponse;
};
export type SharedInfraPageResponse = {
    currentPage: number;
    pageSize: number;
    lastPage: number;
    totalItems: number;
    totalPages: number;
    items: SharedInfraSimpleResponse[];
};
export type PluginConnectionsResponse = {
    "type": string;
    selected: string;
    alias: string;
};
export type Connections = {
    generates: PluginConnectionsResponse[];
    requires: PluginConnectionsResponse[];
};
export type InfraPluginResponse = {
    versionId: string;
    semanticVersion: string;
    slug: string;
    displayName: string;
    alias: string;
    inputs: {
        [key: string]: any;
    };
    inputsEnvs: {
        [key: string]: {
            [key: string]: any;
        };
    };
    connections: Connections;
};
export type StackVersion = {
    versionId: string;
    semanticVersion: string;
    isLatest: boolean;
};
export type InfraPluginStackResponse = {
    stackId: string;
    displayName: string;
    stackVersion: StackVersion;
};
export type AvailablePlugin = {
    pluginVersionId: string;
    semanticVersion: string;
    isDeprecated: boolean;
};
export type AvailableStack = {
    stackVersionId: string;
    semanticVersion: string;
    hasDeprecatedContent: boolean;
    availablePlugins: AvailablePlugin[];
};
export type InfraPluginWithStacksResponse = {
    plugin: InfraPluginResponse;
    stack: InfraPluginStackResponse;
    availableStacks: AvailableStack[];
};
export type UpdateInfraPluginResponse = {
    plugins: InfraPluginWithStacksResponse[];
    stkFilePath: string;
    globalInputs: {
        [key: string]: any;
    };
    globalComputedInputs: {
        [key: string]: any;
    };
};
export type LinksConsolidateResponse = {
    id: string;
    name: string;
    url: string;
    imageUrl?: string;
    readOnly: boolean;
    visible: boolean;
    "type": "STACK" | "PLUGIN" | "APPLICATION" | "SHAREDINFRA" | "WORKSPACE";
};
export type ExtraRepositoriesResponse = {
    name: string;
    description: string;
    provider: string;
    url: string;
};
export type ExtraLinkResponse = {
    url: string;
    title: string;
    sourceType: "APP_DEPLOY" | "INFRA_DEPLOY";
};
export type SharedInfraResponse = {
    /** Shared-infra id. */
    id: string;
    /** Shared-infra name. */
    name: string;
    /** Shared-infra description. */
    description?: string;
    /** Shared-infra repository url. */
    repoUrl?: string;
    /** Shared-infra repository base branch. */
    repoBaseBranch: string;
    /** Stack used to generate this shared-infra. */
    stackVersionId: string;
    /** Starter used to generate this shared-infra. */
    version?: string;
    /** Shared-infra Deploy Status. */
    status?: string;
    /** Shared-infra extra repositories. */
    extraRepositories: ExtraRepositoriesResponse[];
    /** Shared-infra createBy. */
    createdBy: string;
    /** Shared-infra createAt. */
    createdAt: string;
    /** Shared-infra updateBy. */
    updatedBy: string;
    /** Shared-infra updateAt. */
    updatedAt: string;
    /** Extra links */
    extraLinks: ExtraLinkResponse[];
};
export type SimpleWorkspaceResponse = {
    id: string;
    name: string;
    description?: string;
    imageUrl?: string;
};
export type AppDependencyConnInterfaceResponse = {
    name: string;
    "type": string;
    workspaces: SimpleWorkspaceResponse[];
};
export type Dependencies = {
    environmentName: string;
    hasSelfHostedInfra: boolean;
    hasApplication: boolean;
    connectionInterfaces: AppDependencyConnInterfaceResponse[];
};
export type DependenciesResponse = {
    canDelete: boolean;
    hasSelfHostedInfra: boolean;
    hasApplication: boolean;
    dependencies: Dependencies[];
};
export type StackFilterItemResponse = {
    stackId: string;
    displayName: string;
    imageUrl?: string;
    stackVersions: SimpleStackVersionResponse[];
};
export type StackFilterResponse = {
    stacks: StackFilterItemResponse[];
};
export type WorkspacePluginResponseStudio = {
    id: string;
    slug: string;
    name: string;
};
export type WorkspacePluginConnectionsResponse = {
    typeId: string;
    "type": string;
    alias: string;
};
export type WorkspacePluginResponse = {
    stackVersionId: string;
    stackSlug: string;
    stackSemanticVersion: string;
    pluginId: string;
    pluginVersionId: string;
    slug: string;
    displayName: string;
    description: string;
    pluginQualifier: string;
    studio: WorkspacePluginResponseStudio;
    uri: string;
    requiredConnectionInterfacesByType?: WorkspacePluginConnectionsResponse[];
    generatesConnectionInterfacesByType: WorkspacePluginConnectionsResponse[];
};
export type ConnectionInterfaceVisibilityResponse = {
    "type": string;
    ids: string[];
};
export type ConnectionInterfaceAttributesByEnvResponse = {
    environmentId: string;
    envName: string;
    attributes: ConnectionInterfaceAttributeResponse[];
};
export type ConnectionInterfaceDetailsResponse = {
    /** Connection Interface Details id. */
    id: string;
    /** Connection Interface Type id. */
    typeId: string;
    /** Connection Interface id. */
    connectionInterfaceId: string;
    /** Connection Interface visibility. */
    visibility?: ConnectionInterfaceVisibilityResponse;
    /** Source Plugin Version id. */
    sourcePluginVersionId?: string;
    /** Was this Connection Interface automatically generated?. */
    automaticallyGenerated: boolean;
    /** Connection interface attribute's by environment. */
    attributesByEnv: ConnectionInterfaceAttributesByEnvResponse[];
};
export type AvailableConnectionInterfaceResponse = {
    slug: string;
    typeId: string;
    typeName: string;
    automaticallyGenerated: boolean;
};
export type ApplicationSimpleResponse = {
    id: string;
    name: string;
    createdAt: string;
    stack?: SimpleStackResponse;
    status: string;
    stackVersion?: SimpleStackVersionResponse;
};
export type ApplicationPageResponse = {
    currentPage: number;
    pageSize: number;
    lastPage: number;
    totalItems: number;
    totalPages: number;
    items: ApplicationSimpleResponse[];
};
export type ApplicationResponse = {
    /** Application id. */
    id: string;
    /** Application name. */
    name: string;
    /** Application description. */
    description?: string;
    /** Application repository url. */
    repoUrl?: string;
    /** Application repository base branch. */
    repoBaseBranch: string;
    /** Stack used to generate this application. */
    stackVersionId: string;
    /** Starter used to generate this application. */
    starterId: string;
    /** Version of the infrastructure deployed for this application. */
    version?: string;
    /** Application Deploy Version. */
    appVersion?: string;
    /** Application Deploy Status. */
    status?: string;
    /** Application extra repositories. */
    extraRepositories: ExtraRepositoriesResponse[];
    /** Application createdBy */
    createdBy: string;
    /** Application createdAt */
    createdAt: string;
    /** Application updateBy */
    updatedBy: string;
    /** Application updateAt */
    updatedAt: string;
    /** Extra links */
    extraLinks: ExtraLinkResponse[];
};
export type AccountSharedInfraSimpleResponse = {
    id: string;
    name: string;
    createdAt: string;
    stack?: SimpleStackResponse;
    status: string;
    stackVersion?: SimpleStackVersionResponse;
    workspace: SimpleWorkspaceResponse;
};
export type AccountSharedInfraPageResponse = {
    currentPage: number;
    pageSize: number;
    lastPage: number;
    totalItems: number;
    totalPages: number;
    items: AccountSharedInfraSimpleResponse[];
};
export type EnvironmentResponse = {
    id: string;
    name: string;
    description: string;
    isProductive: boolean;
};
export type WorkflowVersionInfoResponse = {
    id: string;
    displayName: string;
    label: string;
    description: string;
    semanticVersion: string;
    targetApp: boolean;
    targetInfra: boolean;
    hasPortalExtension: boolean;
    createdAt: string;
};
export type ExtensionsResponse = {
    id: string;
    name: string;
    slug: string;
};
export type WorkflowForCreationResponse = {
    id: string;
    slug: string;
    "type": string;
    inputs: InputValuesContextResponse[];
    computedInputs: ComputedInputResponse[];
    globalComputedInputs: ComputedInputResponse[];
    versionInfo?: WorkflowVersionInfoResponse;
    portalExtensions: ExtensionsResponse[];
};
export type StackActionInputValuesByEnvInWorkspaceContextResponse = {
    name: string;
    label: string;
    "type": string;
    items?: string[];
    externalItems?: ExternalItemsResponse;
    inputEnv: boolean;
    "default"?: any;
    required?: boolean;
    pattern?: string;
    help?: string;
    valuesByEnv: ValueByEnvWorkspaceResponse[];
    condition?: InputConditionResponse;
    connectionInterfaceType?: string;
    addQuestion?: string;
    input?: any;
    inputs: any[];
};
export type StackActionInputValuesByEnvInAccountContextResponse = {
    name: string;
    label: string;
    "type": string;
    items?: string[];
    externalItems?: ExternalItemsResponse;
    inputEnv: boolean;
    "default"?: any;
    required?: boolean;
    pattern?: string;
    help?: string;
    valuesByEnv: ValueByEnvAccountResponse[];
    condition?: InputConditionResponse;
    connectionInterfaceType?: string;
    addQuestion?: string;
    input?: any;
    inputs: any[];
};
export type AccountApplicationSimpleResponse = {
    id: string;
    name: string;
    createdAt: string;
    stack?: SimpleStackResponse;
    status: string;
    stackVersion?: SimpleStackVersionResponse;
    workspace: SimpleWorkspaceResponse;
};
export type AccountApplicationPageResponse = {
    currentPage: number;
    pageSize: number;
    lastPage: number;
    totalItems: number;
    totalPages: number;
    items: AccountApplicationSimpleResponse[];
};
export type AccountVariableUsageResponse = {
    /** How many plugins are using this account variable */
    pluginsInUse: number;
    /** How many actions are using this account variable */
    actionsInUse: number;
    /** How many contexts are using this account variable */
    contextsInUse: number;
};
export type StackResponse = {
    stackId: string;
    slug: string;
    displayName: string;
    description: string;
    imageUrl?: string;
    existsDraft: boolean;
    createdBy: string;
    createdAt: string;
    lastModificationAt: string;
    studio: StudioResponse;
    tags?: string[];
};
export type StacksByAccountWithWorkflowResponseStackVersion = {
    stackVersionId: string;
    version: number;
    status: string;
    published: boolean;
    publishedBy?: string;
    publishedAt?: string;
    unpublishedBy?: string;
    unpublishedAt?: string;
    semanticVersion?: string;
    copyFromStackVersionId?: string;
};
export type StacksByAccountWithWorkflowResponseStackVersionShortForm = {
    stackVersionId: string;
    version: number;
    published: boolean;
    semanticVersion?: string;
    copyFromStackVersionId?: string;
    status: string;
    hasDeprecatedContent: boolean;
};
export type StacksByAccountWithWorkflowResponseWorkflow = {
    id: string;
    name: string;
    "type": "CREATE_APP" | "CREATE_INFRA" | "CREATE_API";
    isDefault: boolean;
};
export type StacksByAccountWithWorkflowResponse = {
    stack: StackResponse;
    latestVersion: StacksByAccountWithWorkflowResponseStackVersion;
    versions: StacksByAccountWithWorkflowResponseStackVersionShortForm[];
    workflows: StacksByAccountWithWorkflowResponseWorkflow[];
};
export type StackVersionWksResponse = {
    stackVersionId: string;
    version: number;
    status: string;
    hasDeprecatedContent: boolean;
    hasContext?: boolean;
    published: boolean;
    semanticVersion?: string;
    copyFromStackVersionId?: string;
};
export type AccountContextStackResponse = {
    stackVersionId: string;
    hasContext: boolean;
    hasVersionsWithContext: boolean;
    versions: StackVersionWksResponse[];
};
/**
 * Get workspace variable by name
 */
export function workspaceVariableControllerfindByName({ accountId, workspaceId, name }: {
    accountId?: string;
    workspaceId: string;
    name: string;
}, opts?: Oazapfts.RequestOpts) {
    return oazapfts.ok(oazapfts.fetchJson<{
        status: 200;
        data: WorkspaceVariableResponse;
    }>(`/v1/workspaces/${encodeURIComponent(workspaceId)}/variables/${encodeURIComponent(name)}`, {
        ...opts,
        headers: oazapfts.mergeHeaders(opts?.headers, {
            accountId
        })
    }));
}
/**
 * Update workspace variable
 */
export function workspaceVariableControllerupsert({ workspaceId, name, upsertWorkspaceVariableRequest }: {
    workspaceId: string;
    name: string;
    upsertWorkspaceVariableRequest: UpsertWorkspaceVariableRequest;
}, opts?: Oazapfts.RequestOpts) {
    return oazapfts.ok(oazapfts.fetchText(`/v1/workspaces/${encodeURIComponent(workspaceId)}/variables/${encodeURIComponent(name)}`, oazapfts.json({
        ...opts,
        method: "PUT",
        body: upsertWorkspaceVariableRequest
    })));
}
/**
 * Get workspace stack workflow context
 */
export function workflowWorkspaceControllerfindWorkflowStackContext({ workspaceId, stackId, workflowId }: {
    workspaceId: string;
    stackId: string;
    workflowId: string;
}, opts?: Oazapfts.RequestOpts) {
    return oazapfts.ok(oazapfts.fetchJson<{
        status: 200;
        data: WorkflowAccountContextResponse;
    }>(`/v1/workspaces/${encodeURIComponent(workspaceId)}/stacks/${encodeURIComponent(stackId)}/workflows/${encodeURIComponent(workflowId)}/context`, {
        ...opts
    }));
}
/**
 * Save workspace stack workflow context
 */
export function workflowWorkspaceControllersaveWorkflowWorkspaceContext({ workspaceId, stackId, workflowId, workflowContextSaveRequest }: {
    workspaceId: string;
    stackId: string;
    workflowId: string;
    workflowContextSaveRequest: WorkflowContextSaveRequest;
}, opts?: Oazapfts.RequestOpts) {
    return oazapfts.ok(oazapfts.fetchText(`/v1/workspaces/${encodeURIComponent(workspaceId)}/stacks/${encodeURIComponent(stackId)}/workflows/${encodeURIComponent(workflowId)}/context`, oazapfts.json({
        ...opts,
        method: "PUT",
        body: workflowContextSaveRequest
    })));
}
/**
 * Update stacks in a workspace
 */
export function workspaceStackControllerrefreshStacksInWorkspace({ workspaceId, addStackInWorkspaceRequest }: {
    workspaceId: string;
    addStackInWorkspaceRequest: AddStackInWorkspaceRequest;
}, opts?: Oazapfts.RequestOpts) {
    return oazapfts.ok(oazapfts.fetchText(`/v1/workspaces/${encodeURIComponent(workspaceId)}/stacks/-/update`, oazapfts.json({
        ...opts,
        method: "PUT",
        body: addStackInWorkspaceRequest
    })));
}
/**
 * Update workspace connection interface attributes
 */
export function connectionInterfaceControllerupdateConnectionInterfaceAttributes({ workspaceId, connectionInterfaceId, connectionInterfaceAttributesByEnv }: {
    workspaceId: string;
    connectionInterfaceId: string;
    connectionInterfaceAttributesByEnv: ConnectionInterfaceAttributesByEnv;
}, opts?: Oazapfts.RequestOpts) {
    return oazapfts.ok(oazapfts.fetchText(`/v1/workspaces/${encodeURIComponent(workspaceId)}/connection-interfaces/${encodeURIComponent(connectionInterfaceId)}/attributes`, oazapfts.json({
        ...opts,
        method: "PUT",
        body: connectionInterfaceAttributesByEnv
    })));
}
/**
 * List stack workflow inputs in workspace context
 */
export function contextWorkflowControllergetWorkspaceWorkflowInputs({ workspaceId, stackVersionId, workflowVersionId, envName }: {
    workspaceId: string;
    stackVersionId: string;
    workflowVersionId: string;
    envName?: string;
}, opts?: Oazapfts.RequestOpts) {
    return oazapfts.ok(oazapfts.fetchJson<{
        status: 200;
        data: FullInputContextResponse;
    }>(`/v1/contexts/workspaces/${encodeURIComponent(workspaceId)}/stacks/${encodeURIComponent(stackVersionId)}/workflows/${encodeURIComponent(workflowVersionId)}/inputs${QS.query(QS.explode({
        envName
    }))}`, {
        ...opts
    }));
}
/**
 * Update stack workflow inputs in workspace context
 */
export function contextControllerupsertWorkflowWorkspaceContext({ workspaceId, stackVersionId, workflowVersionId, body }: {
    workspaceId: string;
    stackVersionId: string;
    workflowVersionId: string;
    body: UpsertInputContextRequest[];
}, opts?: Oazapfts.RequestOpts) {
    return oazapfts.ok(oazapfts.fetchText(`/v1/contexts/workspaces/${encodeURIComponent(workspaceId)}/stacks/${encodeURIComponent(stackVersionId)}/workflows/${encodeURIComponent(workflowVersionId)}/inputs`, oazapfts.json({
        ...opts,
        method: "PUT",
        body
    })));
}
/**
 * List stack plugin inputs in workspace context
 */
export function contextControllergetConsolidatedPluginInputs({ workspaceId, stackVersionId, pluginVersionId, envName }: {
    workspaceId: string;
    stackVersionId: string;
    pluginVersionId: string;
    envName?: string;
}, opts?: Oazapfts.RequestOpts) {
    return oazapfts.ok(oazapfts.fetchJson<{
        status: 200;
        data: PluginInputValuesInConsolidatedContextResponse[];
    }>(`/v1/contexts/workspaces/${encodeURIComponent(workspaceId)}/stacks/${encodeURIComponent(stackVersionId)}/plugins/${encodeURIComponent(pluginVersionId)}/inputs${QS.query(QS.explode({
        envName
    }))}`, {
        ...opts
    }));
}
/**
 * Update stack plugin inputs in workspace context
 */
export function contextControllerupsertPluginWorkspaceContext({ workspaceId, stackVersionId, pluginVersionId, body }: {
    workspaceId: string;
    stackVersionId: string;
    pluginVersionId: string;
    body: UpsertPluginInputContextRequest[];
}, opts?: Oazapfts.RequestOpts) {
    return oazapfts.ok(oazapfts.fetchText(`/v1/contexts/workspaces/${encodeURIComponent(workspaceId)}/stacks/${encodeURIComponent(stackVersionId)}/plugins/${encodeURIComponent(pluginVersionId)}/inputs`, oazapfts.json({
        ...opts,
        method: "PUT",
        body
    })));
}
/**
 * Update stack action inputs in workspace context
 */
export function contextControllerupsertWorkspaceActionsInputContext({ workspaceId, stackVersionId, actionVersionId, body }: {
    workspaceId: string;
    stackVersionId: string;
    actionVersionId: string;
    body: UpsertActionInputContextRequest[];
}, opts?: Oazapfts.RequestOpts) {
    return oazapfts.ok(oazapfts.fetchText(`/v1/contexts/workspaces/${encodeURIComponent(workspaceId)}/stacks/${encodeURIComponent(stackVersionId)}/actions/${encodeURIComponent(actionVersionId)}/inputs`, oazapfts.json({
        ...opts,
        method: "PUT",
        body
    })));
}
/**
 * List stack workflow inputs in account context
 */
export function contextWorkflowControllergetAccountWorkflowInputs({ stackVersionId, workflowVersionId, envName }: {
    stackVersionId: string;
    workflowVersionId: string;
    envName?: string;
}, opts?: Oazapfts.RequestOpts) {
    return oazapfts.ok(oazapfts.fetchJson<{
        status: 200;
        data: FullInputContextResponse;
    }>(`/v1/contexts/account/stacks/${encodeURIComponent(stackVersionId)}/workflows/${encodeURIComponent(workflowVersionId)}/inputs${QS.query(QS.explode({
        envName
    }))}`, {
        ...opts
    }));
}
/**
 * Update stack workflow inputs in account context
 */
export function contextControllerupsertWorkflowAccountContext({ stackVersionId, workflowVersionId, body }: {
    stackVersionId: string;
    workflowVersionId: string;
    body: UpsertInputContextRequest[];
}, opts?: Oazapfts.RequestOpts) {
    return oazapfts.ok(oazapfts.fetchText(`/v1/contexts/account/stacks/${encodeURIComponent(stackVersionId)}/workflows/${encodeURIComponent(workflowVersionId)}/inputs`, oazapfts.json({
        ...opts,
        method: "PUT",
        body
    })));
}
/**
 * List stack plugin inputs in account context
 */
export function contextControllergetAccountPluginInputs({ stackVersionId, pluginVersionId, envName }: {
    stackVersionId: string;
    pluginVersionId: string;
    envName?: string;
}, opts?: Oazapfts.RequestOpts) {
    return oazapfts.ok(oazapfts.fetchJson<{
        status: 200;
        data: PluginInputValuesInAccountContextResponse[];
    }>(`/v1/contexts/account/stacks/${encodeURIComponent(stackVersionId)}/plugins/${encodeURIComponent(pluginVersionId)}/inputs${QS.query(QS.explode({
        envName
    }))}`, {
        ...opts
    }));
}
/**
 * Update stack plugin inputs in account context
 */
export function contextControllerupsertPluginAccountContext({ stackVersionId, pluginVersionId, body }: {
    stackVersionId: string;
    pluginVersionId: string;
    body: UpsertPluginInputContextRequest[];
}, opts?: Oazapfts.RequestOpts) {
    return oazapfts.ok(oazapfts.fetchText(`/v1/contexts/account/stacks/${encodeURIComponent(stackVersionId)}/plugins/${encodeURIComponent(pluginVersionId)}/inputs`, oazapfts.json({
        ...opts,
        method: "PUT",
        body
    })));
}
/**
 * Update stack action inputs in account context
 */
export function contextControllerupsertAccountActionsInputContext({ stackVersionId, actionVersionId, body }: {
    stackVersionId: string;
    actionVersionId: string;
    body: UpsertActionInputContextRequest[];
}, opts?: Oazapfts.RequestOpts) {
    return oazapfts.ok(oazapfts.fetchText(`/v1/contexts/account/stacks/${encodeURIComponent(stackVersionId)}/actions/${encodeURIComponent(actionVersionId)}/inputs`, oazapfts.json({
        ...opts,
        method: "PUT",
        body
    })));
}
/**
 * Get account workflow settings
 */
export function workflowAccountControllerfindWorkflowSettings({ workflowId }: {
    workflowId: string;
}, opts?: Oazapfts.RequestOpts) {
    return oazapfts.ok(oazapfts.fetchJson<{
        status: 200;
        data: WorkflowAccountSettingsResponse;
    }>(`/v1/account/workflows/${encodeURIComponent(workflowId)}/settings`, {
        ...opts
    }));
}
/**
 * Update account workflow settings
 */
export function workflowAccountControllersaveWorkflowSettings({ workflowId, workflowSettingsSaveRequest }: {
    workflowId: string;
    workflowSettingsSaveRequest: WorkflowSettingsSaveRequest;
}, opts?: Oazapfts.RequestOpts) {
    return oazapfts.ok(oazapfts.fetchText(`/v1/account/workflows/${encodeURIComponent(workflowId)}/settings`, oazapfts.json({
        ...opts,
        method: "PUT",
        body: workflowSettingsSaveRequest
    })));
}
/**
 * List account workflow context
 */
export function workflowAccountControllerfindWorkflowContext({ workflowId }: {
    workflowId: string;
}, opts?: Oazapfts.RequestOpts) {
    return oazapfts.ok(oazapfts.fetchJson<{
        status: 200;
        data: WorkflowAccountContextResponse;
    }>(`/v1/account/workflows/${encodeURIComponent(workflowId)}/context`, {
        ...opts
    }));
}
/**
 * Update account workflow context
 */
export function workflowAccountControllersaveWorkflowContext({ workflowId, workflowContextSaveRequest }: {
    workflowId: string;
    workflowContextSaveRequest: WorkflowContextSaveRequest;
}, opts?: Oazapfts.RequestOpts) {
    return oazapfts.ok(oazapfts.fetchText(`/v1/account/workflows/${encodeURIComponent(workflowId)}/context`, oazapfts.json({
        ...opts,
        method: "PUT",
        body: workflowContextSaveRequest
    })));
}
/**
 * List account workflow actions
 */
export function workflowAccountControllergetWorkflowActions({ workflowId }: {
    workflowId: string;
}, opts?: Oazapfts.RequestOpts) {
    return oazapfts.ok(oazapfts.fetchJson<{
        status: 200;
        data: WorkflowActionResponse;
    }>(`/v1/account/workflows/${encodeURIComponent(workflowId)}/actions`, {
        ...opts
    }));
}
/**
 * Update account workflow actions
 */
export function workflowAccountControllerupdateWorkflowActions({ workflowId, workflowActionsRequest }: {
    workflowId: string;
    workflowActionsRequest: WorkflowActionsRequest;
}, opts?: Oazapfts.RequestOpts) {
    return oazapfts.ok(oazapfts.fetchText(`/v1/account/workflows/${encodeURIComponent(workflowId)}/actions`, oazapfts.json({
        ...opts,
        method: "PUT",
        body: workflowActionsRequest
    })));
}
/**
 * Get account variable by name
 */
export function accountVariableControllerfindByName({ accountId, name }: {
    accountId?: string;
    name: string;
}, opts?: Oazapfts.RequestOpts) {
    return oazapfts.ok(oazapfts.fetchJson<{
        status: 200;
        data: AccountVariableResponse;
    }>(`/v1/account/variables/${encodeURIComponent(name)}`, {
        ...opts,
        headers: oazapfts.mergeHeaders(opts?.headers, {
            accountId
        })
    }));
}
/**
 * Update account variable
 */
export function accountVariableControllerupdate({ name, updateAccountVariableRequest }: {
    name: string;
    updateAccountVariableRequest: UpdateAccountVariableRequest;
}, opts?: Oazapfts.RequestOpts) {
    return oazapfts.ok(oazapfts.fetchText(`/v1/account/variables/${encodeURIComponent(name)}`, oazapfts.json({
        ...opts,
        method: "PUT",
        body: updateAccountVariableRequest
    })));
}
/**
 * Delete account variable
 */
export function accountVariableControllerdelete({ name }: {
    name: string;
}, opts?: Oazapfts.RequestOpts) {
    return oazapfts.ok(oazapfts.fetchText(`/v1/account/variables/${encodeURIComponent(name)}`, {
        ...opts,
        method: "DELETE"
    }));
}
/**
 * Get account stack workflow context
 */
export function workflowStackControllerfindWorkflowStackContext({ stackId, workflowId }: {
    stackId: string;
    workflowId: string;
}, opts?: Oazapfts.RequestOpts) {
    return oazapfts.ok(oazapfts.fetchJson<{
        status: 200;
        data: WorkflowAccountContextResponse;
    }>(`/v1/account/stacks/${encodeURIComponent(stackId)}/workflows/${encodeURIComponent(workflowId)}/context`, {
        ...opts
    }));
}
/**
 * Save account stack workflow context
 */
export function workflowStackControllersaveWorkflowStackContext({ stackId, workflowId, workflowContextSaveRequest }: {
    stackId: string;
    workflowId: string;
    workflowContextSaveRequest: WorkflowContextSaveRequest;
}, opts?: Oazapfts.RequestOpts) {
    return oazapfts.ok(oazapfts.fetchText(`/v1/account/stacks/${encodeURIComponent(stackId)}/workflows/${encodeURIComponent(workflowId)}/context`, oazapfts.json({
        ...opts,
        method: "PUT",
        body: workflowContextSaveRequest
    })));
}
/**
 * Create shared infrastructure in a workspace
 */
export function sharedInfraControllerV2CreateSharedInfra({ workspaceId, createSharedRequest }: {
    workspaceId: string;
    createSharedRequest: CreateSharedRequest;
}, opts?: Oazapfts.RequestOpts) {
    return oazapfts.ok(oazapfts.fetchJson<{
        status: 200;
        data: CreateShareInfraResponse;
    }>(`/v2/workspaces/${encodeURIComponent(workspaceId)}/shared-infra`, oazapfts.json({
        ...opts,
        method: "POST",
        body: createSharedRequest
    })));
}
/**
 * Create application in a workspace
 */
export function applicationControllerV2CreateApp({ workspaceId, createAppRequest }: {
    workspaceId: string;
    createAppRequest: CreateAppRequest;
}, opts?: Oazapfts.RequestOpts) {
    return oazapfts.ok(oazapfts.fetchJson<{
        status: 200;
        data: CreateApplicationResponse;
    }>(`/v2/workspaces/${encodeURIComponent(workspaceId)}/applications`, oazapfts.json({
        ...opts,
        method: "POST",
        body: createAppRequest
    })));
}
/**
 * Create workspace shared infrastructure through workflow
 */
export function contentWorkflowControllercreateSharedInfraThroughWorkflow({ workspaceId, workflowVersionId, contentWorkflowSaveRequest }: {
    workspaceId: string;
    workflowVersionId: string;
    contentWorkflowSaveRequest: ContentWorkflowSaveRequest;
}, opts?: Oazapfts.RequestOpts) {
    return oazapfts.ok(oazapfts.fetchJson<{
        status: 201;
        data: ContentWorkflowSaveResponse;
    }>(`/v1/workspaces/${encodeURIComponent(workspaceId)}/workflows/${encodeURIComponent(workflowVersionId)}/shared-infra`, oazapfts.json({
        ...opts,
        method: "POST",
        body: contentWorkflowSaveRequest
    })));
}
/**
 * Create workspace application through workflow
 */
export function contentWorkflowControllercreateApplicationThroughWorkflow({ workspaceId, workflowVersionId, contentWorkflowSaveRequest }: {
    workspaceId: string;
    workflowVersionId: string;
    contentWorkflowSaveRequest: ContentWorkflowSaveRequest;
}, opts?: Oazapfts.RequestOpts) {
    return oazapfts.ok(oazapfts.fetchJson<{
        status: 201;
        data: ContentWorkflowSaveResponse;
    }>(`/v1/workspaces/${encodeURIComponent(workspaceId)}/workflows/${encodeURIComponent(workflowVersionId)}/applications`, oazapfts.json({
        ...opts,
        method: "POST",
        body: contentWorkflowSaveRequest
    })));
}
/**
 * List workspace stacks
 */
export function workspaceStackControllerlistStacks({ workspaceId }: {
    workspaceId: string;
}, opts?: Oazapfts.RequestOpts) {
    return oazapfts.ok(oazapfts.fetchJson<{
        status: 200;
        data: WorkspaceStackResponse[];
    }>(`/v1/workspaces/${encodeURIComponent(workspaceId)}/stacks`, {
        ...opts
    }));
}
/**
 * Add a stack in a workspace
 */
export function workspaceStackControlleraddStacksInWorkspace({ workspaceId, addStackInWorkspaceRequest }: {
    workspaceId: string;
    addStackInWorkspaceRequest: AddStackInWorkspaceRequest;
}, opts?: Oazapfts.RequestOpts) {
    return oazapfts.ok(oazapfts.fetchText(`/v1/workspaces/${encodeURIComponent(workspaceId)}/stacks`, oazapfts.json({
        ...opts,
        method: "POST",
        body: addStackInWorkspaceRequest
    })));
}
/**
 * Remove workspace shared infrastructure infra plugins and create pull request
 */
export function updateInfrastructureControllerremoveInfraPluginsAndCreatePullRequestBySharedInfra({ workspaceId, sharedInfraId, removePluginAndCreatePrRequest }: {
    workspaceId: string;
    sharedInfraId: string;
    removePluginAndCreatePrRequest: RemovePluginAndCreatePrRequest;
}, opts?: Oazapfts.RequestOpts) {
    return oazapfts.ok(oazapfts.fetchJson<{
        status: 200;
        data: string;
    }>(`/v1/workspaces/${encodeURIComponent(workspaceId)}/shared-infra/${encodeURIComponent(sharedInfraId)}/infra-destroy`, oazapfts.json({
        ...opts,
        method: "POST",
        body: removePluginAndCreatePrRequest
    })));
}
/**
 * List workspace shared infrastructure extra links by environment
 */
export function sharedInfraControllergetSharedInfraExtraLinks({ workspaceId, sharedInfraId, envName }: {
    workspaceId: string;
    sharedInfraId: string;
    envName: string;
}, opts?: Oazapfts.RequestOpts) {
    return oazapfts.ok(oazapfts.fetchJson<{
        status: 200;
        data: ExtraLinkIntegrationResponse[];
    }>(`/v1/workspaces/${encodeURIComponent(workspaceId)}/shared-infra/${encodeURIComponent(sharedInfraId)}/extra-links${QS.query(QS.explode({
        envName
    }))}`, {
        ...opts
    }));
}
/**
 * Update workspace shared infrastructure extra links by environment
 */
export function sharedInfraControllerupdateSharedInfraExtraLinks({ workspaceId, sharedInfraId, envName, body }: {
    workspaceId: string;
    sharedInfraId: string;
    envName: string;
    body: ExtraLinkRequest[];
}, opts?: Oazapfts.RequestOpts) {
    return oazapfts.ok(oazapfts.fetchText(`/v1/workspaces/${encodeURIComponent(workspaceId)}/shared-infra/${encodeURIComponent(sharedInfraId)}/extra-links${QS.query(QS.explode({
        envName
    }))}`, oazapfts.json({
        ...opts,
        method: "POST",
        body
    })));
}
/**
 * Update workspace shared infrastructure infra plugins and create pull request
 */
export function updateInfrastructureControllerupdateInfraPluginsAndCreatePullRequestBySharedInfra({ workspaceId, sharedInfraId, updatePluginAndCreatePrRequest }: {
    workspaceId: string;
    sharedInfraId: string;
    updatePluginAndCreatePrRequest: UpdatePluginAndCreatePrRequest;
}, opts?: Oazapfts.RequestOpts) {
    return oazapfts.ok(oazapfts.fetchJson<{
        status: 200;
        data: string;
    }>(`/v1/workspaces/${encodeURIComponent(workspaceId)}/shared-infra/${encodeURIComponent(sharedInfraId)}/create-pull-request`, oazapfts.json({
        ...opts,
        method: "POST",
        body: updatePluginAndCreatePrRequest
    })));
}
/**
 * Create workspace connection interface
 */
export function connectionInterfaceControllercreateConnectionInterface({ workspaceId, createConnectionInterfaceRequest }: {
    workspaceId: string;
    createConnectionInterfaceRequest: CreateConnectionInterfaceRequest;
}, opts?: Oazapfts.RequestOpts) {
    return oazapfts.ok(oazapfts.fetchJson<{
        status: 201;
        data: IdResponse;
    }>(`/v1/workspaces/${encodeURIComponent(workspaceId)}/connection-interfaces`, oazapfts.json({
        ...opts,
        method: "POST",
        body: createConnectionInterfaceRequest
    })));
}
/**
 * Remove workspace application infra plugins and create pull request
 */
export function updateInfrastructureControllerremoveInfraPluginsAndCreatePullRequestByApplication({ workspaceId, applicationId, removePluginAndCreatePrRequest }: {
    workspaceId: string;
    applicationId: string;
    removePluginAndCreatePrRequest: RemovePluginAndCreatePrRequest;
}, opts?: Oazapfts.RequestOpts) {
    return oazapfts.ok(oazapfts.fetchJson<{
        status: 200;
        data: string;
    }>(`/v1/workspaces/${encodeURIComponent(workspaceId)}/applications/${encodeURIComponent(applicationId)}/infra-destroy`, oazapfts.json({
        ...opts,
        method: "POST",
        body: removePluginAndCreatePrRequest
    })));
}
/**
 * List workspace application extra links by environment
 */
export function applicationControllergetApplicationExtraLinks({ workspaceId, applicationId, envName }: {
    workspaceId: string;
    applicationId: string;
    envName: string;
}, opts?: Oazapfts.RequestOpts) {
    return oazapfts.ok(oazapfts.fetchJson<{
        status: 200;
        data: ExtraLinkIntegrationResponse[];
    }>(`/v1/workspaces/${encodeURIComponent(workspaceId)}/applications/${encodeURIComponent(applicationId)}/extra-links${QS.query(QS.explode({
        envName
    }))}`, {
        ...opts
    }));
}
/**
 * Update workspace application extra links by environment
 */
export function applicationControllerupdateSharedInfraExtraLinks({ workspaceId, applicationId, envName, body }: {
    workspaceId: string;
    applicationId: string;
    envName: string;
    body: ExtraLinkRequest[];
}, opts?: Oazapfts.RequestOpts) {
    return oazapfts.ok(oazapfts.fetchText(`/v1/workspaces/${encodeURIComponent(workspaceId)}/applications/${encodeURIComponent(applicationId)}/extra-links${QS.query(QS.explode({
        envName
    }))}`, oazapfts.json({
        ...opts,
        method: "POST",
        body
    })));
}
/**
 * Register workspace application destroy
 */
export function applicationControllerregistryAppDestroy({ workspaceId, applicationId, environmentId, applicationDestroyRequest }: {
    workspaceId: string;
    applicationId: string;
    environmentId: string;
    applicationDestroyRequest?: ApplicationDestroyRequest;
}, opts?: Oazapfts.RequestOpts) {
    return oazapfts.ok(oazapfts.fetchText(`/v1/workspaces/${encodeURIComponent(workspaceId)}/applications/${encodeURIComponent(applicationId)}/environments/${encodeURIComponent(environmentId)}/app-destroy`, oazapfts.json({
        ...opts,
        method: "POST",
        body: applicationDestroyRequest
    })));
}
/**
 * Update workspace application infra plugins and create pull request
 */
export function updateInfrastructureControllerupdateInfraPluginsAndCreatePullRequestByApplication({ workspaceId, applicationId, updatePluginAndCreatePrRequest }: {
    workspaceId: string;
    applicationId: string;
    updatePluginAndCreatePrRequest: UpdatePluginAndCreatePrRequest;
}, opts?: Oazapfts.RequestOpts) {
    return oazapfts.ok(oazapfts.fetchJson<{
        status: 200;
        data: string;
    }>(`/v1/workspaces/${encodeURIComponent(workspaceId)}/applications/${encodeURIComponent(applicationId)}/create-pull-request`, oazapfts.json({
        ...opts,
        method: "POST",
        body: updatePluginAndCreatePrRequest
    })));
}
/**
 * Get workspaces summary
 */
export function workspaceControllergetWorkspacesSummary({ body }: {
    body: string[];
}, opts?: Oazapfts.RequestOpts) {
    return oazapfts.ok(oazapfts.fetchJson<{
        status: 200;
        data: WorkspaceResponse[];
    }>("/v1/workspaces/summary", oazapfts.json({
        ...opts,
        method: "POST",
        body
    })));
}
/**
 * Import stack context in workspace
 */
export function contextControllerimportContextInWorkspace({ workspaceId, body }: {
    workspaceId: string;
    body: ImportContextRequest[];
}, opts?: Oazapfts.RequestOpts) {
    return oazapfts.ok(oazapfts.fetchText(`/v1/contexts/workspaces/${encodeURIComponent(workspaceId)}/stacks/import-context`, oazapfts.json({
        ...opts,
        method: "POST",
        body
    })));
}
/**
 * Import stack context in account
 */
export function contextControllerimportContextInAccount({ body }: {
    body: ImportContextRequest[];
}, opts?: Oazapfts.RequestOpts) {
    return oazapfts.ok(oazapfts.fetchText("/v1/contexts/account/stacks/import-context", oazapfts.json({
        ...opts,
        method: "POST",
        body
    })));
}
/**
 * Create workflow in an account
 */
export function workflowAccountControllercreateAccountWorkflow({ accountWorkflowCreateRequest }: {
    accountWorkflowCreateRequest: AccountWorkflowCreateRequest;
}, opts?: Oazapfts.RequestOpts) {
    return oazapfts.ok(oazapfts.fetchJson<{
        status: 201;
        data: IdResponseString;
    }>("/v1/account/workflows", oazapfts.json({
        ...opts,
        method: "POST",
        body: accountWorkflowCreateRequest
    })));
}
/**
 * List account variables
 */
export function accountVariableControllerfindAll({ accountId, name, page, size, sortBy, sortDir }: {
    accountId?: string;
    name?: string;
    page?: number;
    size?: number;
    sortBy?: "NAME" | "VALUE" | "CREATED_AT";
    sortDir?: "ASC" | "DESC";
}, opts?: Oazapfts.RequestOpts) {
    return oazapfts.ok(oazapfts.fetchJson<{
        status: 200;
        data: PaginatedAccountVariableResponse;
    }>(`/v1/account/variables${QS.query(QS.explode({
        name,
        page,
        size,
        sortBy,
        sortDir
    }))}`, {
        ...opts,
        headers: oazapfts.mergeHeaders(opts?.headers, {
            accountId
        })
    }));
}
/**
 * Create account variable
 */
export function accountVariableControllercreate({ createAccountVariableRequest }: {
    createAccountVariableRequest: CreateAccountVariableRequest;
}, opts?: Oazapfts.RequestOpts) {
    return oazapfts.ok(oazapfts.fetchText("/v1/account/variables", oazapfts.json({
        ...opts,
        method: "POST",
        body: createAccountVariableRequest
    })));
}
/**
 * Archive application from a workspace
 */
export function managerRunControllerarchiveApplication({ workspaceId, applicationId }: {
    workspaceId: string;
    applicationId: string;
}, opts?: Oazapfts.RequestOpts) {
    return oazapfts.ok(oazapfts.fetchJson<{
        status: 200;
        data: ManagerRunResponse[];
    }>(`/v1/workspaces/${encodeURIComponent(workspaceId)}/applications/${encodeURIComponent(applicationId)}/archive`, {
        ...opts,
        method: "PATCH"
    }));
}
/**
 * List workspace shared infrastructure plugins by environment
 */
export function sharedInfraControllerV2GetSharedInfraAppliedPlugins({ workspaceId, sharedInfraId, envName, pluginType, containsLinks }: {
    workspaceId: string;
    sharedInfraId: string;
    envName: string;
    pluginType?: string;
    containsLinks?: boolean;
}, opts?: Oazapfts.RequestOpts) {
    return oazapfts.ok(oazapfts.fetchJson<{
        status: 200;
        data: SharedInfraAppliedPluginsV2Response;
    }>(`/v2/workspaces/${encodeURIComponent(workspaceId)}/shared-infra/${encodeURIComponent(sharedInfraId)}/environments/${encodeURIComponent(envName)}/plugins${QS.query(QS.explode({
        pluginType,
        containsLinks
    }))}`, {
        ...opts
    }));
}
/**
 * List workspace shared infrastructure activities by environment
 */
export function sharedInfraControllerV2GetSharedInfraActivities({ workspaceId, sharedInfraId, envName, $type, page, size }: {
    workspaceId: string;
    sharedInfraId: string;
    envName: string;
    $type?: "DEPLOY" | "DESTROY" | "ROLLBACK" | "CREATE" | "ACTION" | "CUSTOMER_WORKFLOW";
    page: number;
    size: number;
}, opts?: Oazapfts.RequestOpts) {
    return oazapfts.ok(oazapfts.fetchJson<{
        status: 200;
        data: PaginatedActivityResponse;
    }>(`/v2/workspaces/${encodeURIComponent(workspaceId)}/shared-infra/${encodeURIComponent(sharedInfraId)}/environments/${encodeURIComponent(envName)}/activities${QS.query(QS.explode({
        "type": $type,
        page,
        size
    }))}`, {
        ...opts
    }));
}
/**
 * List workspace application plugins by environment
 */
export function applicationControllerV2GetApplicationAppliedPlugins({ workspaceId, applicationId, envName, pluginType, containsLinks }: {
    workspaceId: string;
    applicationId: string;
    envName: string;
    pluginType?: string;
    containsLinks?: boolean;
}, opts?: Oazapfts.RequestOpts) {
    return oazapfts.ok(oazapfts.fetchJson<{
        status: 200;
        data: ApplicationAppliedPluginsV2Response;
    }>(`/v2/workspaces/${encodeURIComponent(workspaceId)}/applications/${encodeURIComponent(applicationId)}/environments/${encodeURIComponent(envName)}/plugins${QS.query(QS.explode({
        pluginType,
        containsLinks
    }))}`, {
        ...opts
    }));
}
/**
 * List workspace application activities by environment
 */
export function applicationControllerV2GetApplicationActivities({ workspaceId, applicationId, envName, $type, page, size }: {
    workspaceId: string;
    applicationId: string;
    envName: string;
    $type?: "DEPLOY" | "DESTROY" | "ROLLBACK" | "CREATE" | "ACTION" | "CUSTOMER_WORKFLOW";
    page: number;
    size: number;
}, opts?: Oazapfts.RequestOpts) {
    return oazapfts.ok(oazapfts.fetchJson<{
        status: 200;
        data: PaginatedActivityResponse;
    }>(`/v2/workspaces/${encodeURIComponent(workspaceId)}/applications/${encodeURIComponent(applicationId)}/environments/${encodeURIComponent(envName)}/activities${QS.query(QS.explode({
        "type": $type,
        page,
        size
    }))}`, {
        ...opts
    }));
}
/**
 * Get workspace stack context
 */
export function contextStackControllergetWorkspaceContext({ workspaceId, stackVersionId, envName }: {
    workspaceId: string;
    stackVersionId: string;
    envName?: string;
}, opts?: Oazapfts.RequestOpts) {
    return oazapfts.ok(oazapfts.fetchJson<{
        status: 200;
        data: WorkspaceContextV2Response;
    }>(`/v2/contexts/workspaces/${encodeURIComponent(workspaceId)}/stacks/${encodeURIComponent(stackVersionId)}${QS.query(QS.explode({
        envName
    }))}`, {
        ...opts
    }));
}
/**
 * List stack plugin inputs with connection interfaces in workspace context
 */
export function contextgetConsolidatedPluginInputsWithConnectionInterfaces({ workspaceId, stackVersionId, pluginVersionId }: {
    workspaceId: string;
    stackVersionId: string;
    pluginVersionId: string;
}, opts?: Oazapfts.RequestOpts) {
    return oazapfts.ok(oazapfts.fetchJson<{
        status: 200;
        data: PluginForAppCreationV2Response;
    }>(`/v2/contexts/workspaces/${encodeURIComponent(workspaceId)}/stacks/${encodeURIComponent(stackVersionId)}/plugins/${encodeURIComponent(pluginVersionId)}`, {
        ...opts
    }));
}
/**
 * List stack plugin inputs in workspace context
 */
export function contextgetConsolidatedPluginInputs({ workspaceId, stackVersionId, pluginVersionId, envName }: {
    workspaceId: string;
    stackVersionId: string;
    pluginVersionId: string;
    envName?: string;
}, opts?: Oazapfts.RequestOpts) {
    return oazapfts.ok(oazapfts.fetchJson<{
        status: 200;
        data: FullInputContextResponse;
    }>(`/v2/contexts/workspaces/${encodeURIComponent(workspaceId)}/stacks/${encodeURIComponent(stackVersionId)}/plugins/${encodeURIComponent(pluginVersionId)}/inputs${QS.query(QS.explode({
        envName
    }))}`, {
        ...opts
    }));
}
/**
 * List stack action inputs in workspace context
 */
export function contextActionControllergetStackActionInputsInWorkspaceContext({ workspaceId, stackVersionId, actionsVersionId, envName }: {
    workspaceId: string;
    stackVersionId: string;
    actionsVersionId: string;
    envName?: string;
}, opts?: Oazapfts.RequestOpts) {
    return oazapfts.ok(oazapfts.fetchJson<{
        status: 200;
        data: FullInputContextResponse;
    }>(`/v2/contexts/workspaces/${encodeURIComponent(workspaceId)}/stacks/${encodeURIComponent(stackVersionId)}/actions/${encodeURIComponent(actionsVersionId)}/inputs${QS.query(QS.explode({
        envName
    }))}`, {
        ...opts
    }));
}
/**
 * List stack plugin inputs in account context
 */
export function contextgetAccountPluginInputs({ stackVersionId, pluginVersionId, envName }: {
    stackVersionId: string;
    pluginVersionId: string;
    envName?: string;
}, opts?: Oazapfts.RequestOpts) {
    return oazapfts.ok(oazapfts.fetchJson<{
        status: 200;
        data: FullInputContextResponse;
    }>(`/v2/contexts/account/stacks/${encodeURIComponent(stackVersionId)}/plugins/${encodeURIComponent(pluginVersionId)}/inputs${QS.query(QS.explode({
        envName
    }))}`, {
        ...opts
    }));
}
/**
 * List stack action inputs in account context
 */
export function contextActionControllergetStackActionInputsInAccountContext({ stackVersionId, actionsVersionId, envName }: {
    stackVersionId: string;
    actionsVersionId: string;
    envName?: string;
}, opts?: Oazapfts.RequestOpts) {
    return oazapfts.ok(oazapfts.fetchJson<{
        status: 200;
        data: FullInputContextResponse;
    }>(`/v2/contexts/account/stacks/${encodeURIComponent(stackVersionId)}/actions/${encodeURIComponent(actionsVersionId)}/inputs${QS.query(QS.explode({
        envName
    }))}`, {
        ...opts
    }));
}
/**
 * List workspaces
 */
export function workspaceControllergetWorkspaces({ name, aclOnly, accountId, page, size, sortBy, sortDir, tabFilter }: {
    name?: string;
    aclOnly?: boolean;
    accountId?: string;
    page?: number;
    size?: number;
    sortBy?: "NAME" | "DESCRIPTION";
    sortDir?: "ASC" | "DESC";
    tabFilter?: "FAVORITES" | "MY_RESOURCE" | "ALL_RESOURCE";
}, opts?: Oazapfts.RequestOpts) {
    return oazapfts.ok(oazapfts.fetchJson<{
        status: 200;
        data: WorkspacePageResponse;
    }>(`/v1/workspaces${QS.query(QS.explode({
        name,
        aclOnly,
        page,
        size,
        sortBy,
        sortDir,
        tabFilter
    }))}`, {
        ...opts,
        headers: oazapfts.mergeHeaders(opts?.headers, {
            accountId
        })
    }));
}
/**
 * List workspace workflow stacks
 */
export function workflowWorkspaceControllerlistStacks({ workspaceId, workflowId }: {
    workspaceId: string;
    workflowId: string;
}, opts?: Oazapfts.RequestOpts) {
    return oazapfts.ok(oazapfts.fetchJson<{
        status: 200;
        data: SimpleStackResponse[];
    }>(`/v1/workspaces/${encodeURIComponent(workspaceId)}/workflows/${encodeURIComponent(workflowId)}/stacks`, {
        ...opts
    }));
}
/**
 * List workspace variables
 */
export function workspaceVariableControllerfindAll({ workspaceId, accountId, page, size, sortBy, sortDir, mandate, name, showEmptyValues }: {
    workspaceId: string;
    accountId?: string;
    page?: number;
    size?: number;
    sortBy?: "NAME" | "VALUE" | "CREATED_AT";
    sortDir?: "ASC" | "DESC";
    mandate?: boolean;
    name?: string;
    showEmptyValues?: boolean;
}, opts?: Oazapfts.RequestOpts) {
    return oazapfts.ok(oazapfts.fetchJson<{
        status: 200;
        data: PaginatedWorkspaceVariableResponse;
    }>(`/v1/workspaces/${encodeURIComponent(workspaceId)}/variables${QS.query(QS.explode({
        page,
        size,
        sortBy,
        sortDir,
        mandate,
        name,
        showEmptyValues
    }))}`, {
        ...opts,
        headers: oazapfts.mergeHeaders(opts?.headers, {
            accountId
        })
    }));
}
/**
 * Get workspace workflow by stack id and workflow type
 */
export function workflowWorkspaceControllerlistWorkflowByStackIdAndWorkflowType({ workspaceId, stackId, workflowType }: {
    workspaceId: string;
    stackId: string;
    workflowType: "CREATE_API" | "CREATE_APP" | "CREATE_INFRA";
}, opts?: Oazapfts.RequestOpts) {
    return oazapfts.ok(oazapfts.fetchJson<{
        status: 200;
        data: WorkflowCompleteResponse;
    }>(`/v1/workspaces/${encodeURIComponent(workspaceId)}/stacks/${encodeURIComponent(stackId)}/workflows/${encodeURIComponent(workflowType)}`, {
        ...opts
    }));
}
/**
 * List workspace stacks with context
 */
export function workspaceStackControllerlistStacksContext({ workspaceId }: {
    workspaceId: string;
}, opts?: Oazapfts.RequestOpts) {
    return oazapfts.ok(oazapfts.fetchJson<{
        status: 200;
        data: WorkspaceStackContextResponse[];
    }>(`/v1/workspaces/${encodeURIComponent(workspaceId)}/stacks/with-context`, {
        ...opts
    }));
}
/**
 * List workspace shared infrastructures
 */
export function sharedInfraControllergetSharedInfra({ workspaceId, stackId, stackVersionId, name, page, size, sortBy, sortDir }: {
    workspaceId: string;
    stackId?: string;
    stackVersionId?: string;
    name?: string;
    page?: number;
    size?: number;
    sortBy?: "NAME" | "CREATED_AT" | "STACK_VERSION" | "STACK";
    sortDir?: "ASC" | "DESC";
}, opts?: Oazapfts.RequestOpts) {
    return oazapfts.ok(oazapfts.fetchJson<{
        status: 200;
        data: SharedInfraPageResponse;
    }>(`/v1/workspaces/${encodeURIComponent(workspaceId)}/shared-infra${QS.query(QS.explode({
        stackId,
        stackVersionId,
        name,
        page,
        size,
        sortBy,
        sortDir
    }))}`, {
        ...opts
    }));
}
/**
 * List workspace shared infrastructure infra plugins
 */
export function updateInfrastructureControllergetSharedInfraInfraPlugins({ workspaceId, sharedInfraId, branchName }: {
    workspaceId: string;
    sharedInfraId: string;
    branchName: string;
}, opts?: Oazapfts.RequestOpts) {
    return oazapfts.ok(oazapfts.fetchJson<{
        status: 200;
        data: UpdateInfraPluginResponse;
    }>(`/v1/workspaces/${encodeURIComponent(workspaceId)}/shared-infra/${encodeURIComponent(sharedInfraId)}/infra-plugins${QS.query(QS.explode({
        branchName
    }))}`, {
        ...opts
    }));
}
/**
 * List workspace shared infrastructure links by environment
 */
export function sharedInfraControllergetApplicationLinks({ workspaceId, sharedInfraId, envName }: {
    workspaceId: string;
    sharedInfraId: string;
    envName: string;
}, opts?: Oazapfts.RequestOpts) {
    return oazapfts.ok(oazapfts.fetchJson<{
        status: 200;
        data: LinksConsolidateResponse[];
    }>(`/v1/workspaces/${encodeURIComponent(workspaceId)}/shared-infra/${encodeURIComponent(sharedInfraId)}/environments/${encodeURIComponent(envName)}/links`, {
        ...opts
    }));
}
/**
 * Get workspace shared infrastructure details by environment
 */
export function sharedInfraControllergetSharedInfraAppliedPlugins({ workspaceId, sharedInfraId, envName }: {
    workspaceId: string;
    sharedInfraId: string;
    envName: string;
}, opts?: Oazapfts.RequestOpts) {
    return oazapfts.ok(oazapfts.fetchJson<{
        status: 200;
        data: SharedInfraResponse;
    }>(`/v1/workspaces/${encodeURIComponent(workspaceId)}/shared-infra/${encodeURIComponent(sharedInfraId)}/environments/${encodeURIComponent(envName)}/details`, {
        ...opts
    }));
}
/**
 * Get workspace shared infrastructure dependencies
 */
export function sharedInfraControllergetSharedInfraDependencies({ workspaceId, sharedInfraId }: {
    workspaceId: string;
    sharedInfraId: string;
}, opts?: Oazapfts.RequestOpts) {
    return oazapfts.ok(oazapfts.fetchJson<{
        status: 200;
        data: DependenciesResponse;
    }>(`/v1/workspaces/${encodeURIComponent(workspaceId)}/shared-infra/${encodeURIComponent(sharedInfraId)}/dependencies`, {
        ...opts
    }));
}
/**
 * List workspace shared infrastructure branches
 */
export function updateInfrastructureControllergetSharedInfraBranches({ workspaceId, sharedInfraId }: {
    workspaceId: string;
    sharedInfraId: string;
}, opts?: Oazapfts.RequestOpts) {
    return oazapfts.ok(oazapfts.fetchJson<{
        status: 200;
        data: string[];
    }>(`/v1/workspaces/${encodeURIComponent(workspaceId)}/shared-infra/${encodeURIComponent(sharedInfraId)}/branches`, {
        ...opts
    }));
}
/**
 * List workspace shared infrastructure filters
 */
export function sharedInfraControllergetSharedInfraFilters({ workspaceId }: {
    workspaceId: string;
}, opts?: Oazapfts.RequestOpts) {
    return oazapfts.ok(oazapfts.fetchJson<{
        status: 200;
        data: StackFilterResponse;
    }>(`/v1/workspaces/${encodeURIComponent(workspaceId)}/shared-infra/filters`, {
        ...opts
    }));
}
/**
 * List workspace plugins
 */
export function workspacePluginControllerlistPluginsFromWorkspace({ workspaceId, typePlugin }: {
    workspaceId: string;
    typePlugin?: "INFRA" | "APP";
}, opts?: Oazapfts.RequestOpts) {
    return oazapfts.ok(oazapfts.fetchJson<{
        status: 200;
        data: WorkspacePluginResponse[];
    }>(`/v1/workspaces/${encodeURIComponent(workspaceId)}/plugins${QS.query(QS.explode({
        typePlugin
    }))}`, {
        ...opts
    }));
}
/**
 * Get workspace connection interface details
 */
export function connectionInterfaceControllergetConnectionInterfaceDetails({ workspaceId, connectionInterfaceId }: {
    workspaceId: string;
    connectionInterfaceId: string;
}, opts?: Oazapfts.RequestOpts) {
    return oazapfts.ok(oazapfts.fetchJson<{
        status: 200;
        data: ConnectionInterfaceDetailsResponse;
    }>(`/v1/workspaces/${encodeURIComponent(workspaceId)}/connection-interfaces/${encodeURIComponent(connectionInterfaceId)}`, {
        ...opts
    }));
}
/**
 * List workspace available connection interfaces
 */
export function availableConnectionInterfaceControllergetAvailableConnectionInterfaces({ workspaceId, applicationId, typeName, automaticallyGenerated }: {
    workspaceId: string;
    applicationId?: string;
    typeName?: string;
    automaticallyGenerated?: boolean;
}, opts?: Oazapfts.RequestOpts) {
    return oazapfts.ok(oazapfts.fetchJson<{
        status: 200;
        data: AvailableConnectionInterfaceResponse[];
    }>(`/v1/workspaces/${encodeURIComponent(workspaceId)}/available-connection-interfaces${QS.query(QS.explode({
        applicationId,
        typeName,
        automaticallyGenerated
    }))}`, {
        ...opts
    }));
}
/**
 * List workspace applications
 */
export function applicationControllergetApplications({ workspaceId, stackVersionId, stackId, name, page, size, sortBy, sortDir }: {
    workspaceId: string;
    stackVersionId?: string;
    stackId?: string;
    name?: string;
    page?: number;
    size?: number;
    sortBy?: "NAME" | "CREATED_AT" | "STACK_VERSION" | "STACK";
    sortDir?: "ASC" | "DESC";
}, opts?: Oazapfts.RequestOpts) {
    return oazapfts.ok(oazapfts.fetchJson<{
        status: 200;
        data: ApplicationPageResponse;
    }>(`/v1/workspaces/${encodeURIComponent(workspaceId)}/applications${QS.query(QS.explode({
        stackVersionId,
        stackId,
        name,
        page,
        size,
        sortBy,
        sortDir
    }))}`, {
        ...opts
    }));
}
/**
 * List workspace application infra plugins
 */
export function updateInfrastructureControllergetApplicationInfraPlugins({ workspaceId, applicationId, branchName }: {
    workspaceId: string;
    applicationId: string;
    branchName: string;
}, opts?: Oazapfts.RequestOpts) {
    return oazapfts.ok(oazapfts.fetchJson<{
        status: 200;
        data: UpdateInfraPluginResponse;
    }>(`/v1/workspaces/${encodeURIComponent(workspaceId)}/applications/${encodeURIComponent(applicationId)}/infra-plugins${QS.query(QS.explode({
        branchName
    }))}`, {
        ...opts
    }));
}
/**
 * List workspace application links by environment
 */
export function applicationControllergetApplicationLinks({ workspaceId, applicationId, envName }: {
    workspaceId: string;
    applicationId: string;
    envName: string;
}, opts?: Oazapfts.RequestOpts) {
    return oazapfts.ok(oazapfts.fetchJson<{
        status: 200;
        data: LinksConsolidateResponse[];
    }>(`/v1/workspaces/${encodeURIComponent(workspaceId)}/applications/${encodeURIComponent(applicationId)}/environments/${encodeURIComponent(envName)}/links`, {
        ...opts
    }));
}
/**
 * Get workspace application details by environment
 */
export function applicationControllergetApplicationDetails({ workspaceId, applicationId, envName }: {
    workspaceId: string;
    applicationId: string;
    envName: string;
}, opts?: Oazapfts.RequestOpts) {
    return oazapfts.ok(oazapfts.fetchJson<{
        status: 200;
        data: ApplicationResponse;
    }>(`/v1/workspaces/${encodeURIComponent(workspaceId)}/applications/${encodeURIComponent(applicationId)}/environments/${encodeURIComponent(envName)}/details`, {
        ...opts
    }));
}
/**
 * Get workspace application dependencies
 */
export function applicationControllergetApplicationDependencies({ workspaceId, applicationId }: {
    workspaceId: string;
    applicationId: string;
}, opts?: Oazapfts.RequestOpts) {
    return oazapfts.ok(oazapfts.fetchJson<{
        status: 200;
        data: DependenciesResponse;
    }>(`/v1/workspaces/${encodeURIComponent(workspaceId)}/applications/${encodeURIComponent(applicationId)}/dependencies`, {
        ...opts
    }));
}
/**
 * List workspace application branches
 */
export function updateInfrastructureControllergetApplicationBranches({ workspaceId, applicationId }: {
    workspaceId: string;
    applicationId: string;
}, opts?: Oazapfts.RequestOpts) {
    return oazapfts.ok(oazapfts.fetchJson<{
        status: 200;
        data: string[];
    }>(`/v1/workspaces/${encodeURIComponent(workspaceId)}/applications/${encodeURIComponent(applicationId)}/branches`, {
        ...opts
    }));
}
/**
 * List workspace application filters
 */
export function applicationControllergetApplicationFilters({ workspaceId }: {
    workspaceId: string;
}, opts?: Oazapfts.RequestOpts) {
    return oazapfts.ok(oazapfts.fetchJson<{
        status: 200;
        data: StackFilterResponse;
    }>(`/v1/workspaces/${encodeURIComponent(workspaceId)}/applications/filters`, {
        ...opts
    }));
}
/**
 * List account shared infrastructures
 */
export function accountSharedInfraControllergetAccountSharedInfraPage({ page, size, sortBy, sortDir, name, tab, stackVersionId, stackId }: {
    page?: number;
    size?: number;
    sortBy?: "NAME" | "CREATED_AT" | "STACK_VERSION" | "STACK";
    sortDir?: "ASC" | "DESC";
    name?: string;
    tab?: "FAVORITES" | "MY_RESOURCE" | "ALL_RESOURCE";
    stackVersionId?: string;
    stackId?: string;
}, opts?: Oazapfts.RequestOpts) {
    return oazapfts.ok(oazapfts.fetchJson<{
        status: 200;
        data: AccountSharedInfraPageResponse;
    }>(`/v1/shared-infra${QS.query(QS.explode({
        page,
        size,
        sortBy,
        sortDir,
        name,
        tab,
        stackVersionId,
        stackId
    }))}`, {
        ...opts
    }));
}
/**
 * List account shared infrastructure filters
 */
export function accountSharedInfraControllergetAccountSharedInfraFilters(opts?: Oazapfts.RequestOpts) {
    return oazapfts.ok(oazapfts.fetchJson<{
        status: 200;
        data: StackFilterResponse;
    }>("/v1/shared-infra/filters", {
        ...opts
    }));
}
/**
 * List environments by account
 */
export function environmentControllergetEnvironments({ accountId }: {
    accountId?: string;
}, opts?: Oazapfts.RequestOpts) {
    return oazapfts.ok(oazapfts.fetchJson<{
        status: 200;
        data: EnvironmentResponse[];
    }>("/v1/environments", {
        ...opts,
        headers: oazapfts.mergeHeaders(opts?.headers, {
            accountId
        })
    }));
}
/**
 * Get stack workflow consolidated inputs in workspace context
 */
export function contextWorkflowControllergetConsolidatedWorkflowInputs({ workspaceId, stackVersionId, workflowVersionId }: {
    workspaceId: string;
    stackVersionId: string;
    workflowVersionId: string;
}, opts?: Oazapfts.RequestOpts) {
    return oazapfts.ok(oazapfts.fetchJson<{
        status: 200;
        data: WorkflowForCreationResponse;
    }>(`/v1/contexts/workspaces/${encodeURIComponent(workspaceId)}/stacks/${encodeURIComponent(stackVersionId)}/workflows/${encodeURIComponent(workflowVersionId)}`, {
        ...opts
    }));
}
/**
 * Clear stack workflow context in workspace
 */
export function contextControllerclearWorkflowContextInWorkspace({ workspaceId, stackVersionId, workflowVersionId }: {
    workspaceId: string;
    stackVersionId: string;
    workflowVersionId: string;
}, opts?: Oazapfts.RequestOpts) {
    return oazapfts.ok(oazapfts.fetchText(`/v1/contexts/workspaces/${encodeURIComponent(workspaceId)}/stacks/${encodeURIComponent(stackVersionId)}/workflows/${encodeURIComponent(workflowVersionId)}`, {
        ...opts,
        method: "DELETE"
    }));
}
/**
 * List stack action inputs in workspace context
 */
export function contextControllergetStackActionInputsInWorkspaceContext({ workspaceId, stackVersionId, actionsVersionId, envName }: {
    workspaceId: string;
    stackVersionId: string;
    actionsVersionId: string;
    envName?: string;
}, opts?: Oazapfts.RequestOpts) {
    return oazapfts.ok(oazapfts.fetchJson<{
        status: 200;
        data: StackActionInputValuesByEnvInWorkspaceContextResponse[];
    }>(`/v1/contexts/workspaces/${encodeURIComponent(workspaceId)}/stacks/${encodeURIComponent(stackVersionId)}/actions/${encodeURIComponent(actionsVersionId)}/inputs${QS.query(QS.explode({
        envName
    }))}`, {
        ...opts
    }));
}
/**
 * List stack action inputs in account context
 */
export function contextControllergetStackActionInputsInAccountContext({ stackVersionId, actionsVersionId, envName }: {
    stackVersionId: string;
    actionsVersionId: string;
    envName?: string;
}, opts?: Oazapfts.RequestOpts) {
    return oazapfts.ok(oazapfts.fetchJson<{
        status: 200;
        data: StackActionInputValuesByEnvInAccountContextResponse[];
    }>(`/v1/contexts/account/stacks/${encodeURIComponent(stackVersionId)}/actions/${encodeURIComponent(actionsVersionId)}/inputs${QS.query(QS.explode({
        envName
    }))}`, {
        ...opts
    }));
}
/**
 * List account applications
 */
export function accountApplicationControllergetAccountApplicationPage({ page, size, sortBy, sortDir, tab, name, stackVersionId, stackId }: {
    page?: number;
    size?: number;
    sortBy?: "NAME" | "CREATED_AT" | "STACK_VERSION" | "STACK";
    sortDir?: "ASC" | "DESC";
    tab?: "FAVORITES" | "MY_RESOURCE" | "ALL_RESOURCE";
    name?: string;
    stackVersionId?: string;
    stackId?: string;
}, opts?: Oazapfts.RequestOpts) {
    return oazapfts.ok(oazapfts.fetchJson<{
        status: 200;
        data: AccountApplicationPageResponse;
    }>(`/v1/applications${QS.query(QS.explode({
        page,
        size,
        sortBy,
        sortDir,
        tab,
        name,
        stackVersionId,
        stackId
    }))}`, {
        ...opts
    }));
}
/**
 * List account application filters
 */
export function accountApplicationControllergetAccountApplicationsFilters(opts?: Oazapfts.RequestOpts) {
    return oazapfts.ok(oazapfts.fetchJson<{
        status: 200;
        data: StackFilterResponse;
    }>("/v1/applications/filters", {
        ...opts
    }));
}
/**
 * Get account variable usage by name
 */
export function accountVariableControllerusage({ accountId, name }: {
    accountId?: string;
    name: string;
}, opts?: Oazapfts.RequestOpts) {
    return oazapfts.ok(oazapfts.fetchJson<{
        status: 200;
        data: AccountVariableUsageResponse;
    }>(`/v1/account/variables/${encodeURIComponent(name)}/usage`, {
        ...opts,
        headers: oazapfts.mergeHeaders(opts?.headers, {
            accountId
        })
    }));
}
/**
 * List account stacks with workflows
 */
export function accountStackControllerlistStacksByAccountWithWorkflow({ workspaceId }: {
    workspaceId?: string;
}, opts?: Oazapfts.RequestOpts) {
    return oazapfts.ok(oazapfts.fetchJson<{
        status: 200;
        data: StacksByAccountWithWorkflowResponse[];
    }>(`/v1/account/stacks${QS.query(QS.explode({
        workspaceId
    }))}`, {
        ...opts
    }));
}
/**
 * List account workflows by stack id
 */
export function accountStackControllerlistWorkflowsByStack({ stackId }: {
    stackId: string;
}, opts?: Oazapfts.RequestOpts) {
    return oazapfts.ok(oazapfts.fetchJson<{
        status: 200;
        data: WorkflowCompleteResponse[];
    }>(`/v1/account/stacks/${encodeURIComponent(stackId)}/workflows`, {
        ...opts
    }));
}
/**
 * List account most-used stacks in applications and shared infrastructures
 */
export function accountStackControllerlistMostUsedStacksInAppsAndSharedInfras({ accountId }: {
    accountId: string;
}, opts?: Oazapfts.RequestOpts) {
    return oazapfts.ok(oazapfts.fetchJson<{
        status: 200;
        data: string[];
    }>("/v1/account/stacks/most-used", {
        ...opts,
        headers: oazapfts.mergeHeaders(opts?.headers, {
            accountId
        })
    }));
}
/**
 * Get account context by stack version id
 */
export function accountStackControllerlistVersionsWithContext({ stackVersionId }: {
    stackVersionId: string;
}, opts?: Oazapfts.RequestOpts) {
    return oazapfts.ok(oazapfts.fetchJson<{
        status: 200;
        data: AccountContextStackResponse;
    }>(`/v1/account/stackVersionId/${encodeURIComponent(stackVersionId)}`, {
        ...opts
    }));
}
/**
 * Delete workspace
 */
export function workspaceControllerdelete({ workspaceId }: {
    workspaceId: string;
}, opts?: Oazapfts.RequestOpts) {
    return oazapfts.ok(oazapfts.fetchText(`/v1/workspaces/${encodeURIComponent(workspaceId)}`, {
        ...opts,
        method: "DELETE"
    }));
}
/**
 * Delete a stack from a workspace
 */
export function workspaceStackControllerdeleteStacksInWorkspace({ workspaceId, stackVersionId }: {
    workspaceId: string;
    stackVersionId: string;
}, opts?: Oazapfts.RequestOpts) {
    return oazapfts.ok(oazapfts.fetchText(`/v1/workspaces/${encodeURIComponent(workspaceId)}/stacks/${encodeURIComponent(stackVersionId)}`, {
        ...opts,
        method: "DELETE"
    }));
}
/**
 * Delete shared infrastructure from a workspace
 */
export function managerRunControllerdeleteSharedInfra({ workspaceId, sharedInfraId }: {
    workspaceId: string;
    sharedInfraId: string;
}, opts?: Oazapfts.RequestOpts) {
    return oazapfts.ok(oazapfts.fetchText(`/v1/workspaces/${encodeURIComponent(workspaceId)}/shared-infra/${encodeURIComponent(sharedInfraId)}`, {
        ...opts,
        method: "DELETE"
    }));
}
/**
 * Delete application from a workspace
 */
export function managerRunControllerdeleteApplication({ workspaceId, applicationId }: {
    workspaceId: string;
    applicationId: string;
}, opts?: Oazapfts.RequestOpts) {
    return oazapfts.ok(oazapfts.fetchJson<{
        status: 200;
        data: ManagerRunResponse[];
    }>(`/v1/workspaces/${encodeURIComponent(workspaceId)}/applications/${encodeURIComponent(applicationId)}`, {
        ...opts,
        method: "DELETE"
    }));
}
/**
 * Clear stack plugin context in workspace
 */
export function contextControllerclearPluginContextInWorkspace({ workspaceId, stackVersionId, pluginVersionId }: {
    workspaceId: string;
    stackVersionId: string;
    pluginVersionId: string;
}, opts?: Oazapfts.RequestOpts) {
    return oazapfts.ok(oazapfts.fetchText(`/v1/contexts/workspaces/${encodeURIComponent(workspaceId)}/stacks/${encodeURIComponent(stackVersionId)}/plugins/${encodeURIComponent(pluginVersionId)}`, {
        ...opts,
        method: "DELETE"
    }));
}
/**
 * Clear stack action context in workspace
 */
export function contextControllerclearActionContextInWorkspace({ workspaceId, stackVersionId, actionVersionId }: {
    workspaceId: string;
    stackVersionId: string;
    actionVersionId: string;
}, opts?: Oazapfts.RequestOpts) {
    return oazapfts.ok(oazapfts.fetchText(`/v1/contexts/workspaces/${encodeURIComponent(workspaceId)}/stacks/${encodeURIComponent(stackVersionId)}/actions/${encodeURIComponent(actionVersionId)}`, {
        ...opts,
        method: "DELETE"
    }));
}
/**
 * Clear stack workflow context in account
 */
export function contextControllerclearWorkflowContextInAccount({ stackVersionId, workflowVersionId }: {
    stackVersionId: string;
    workflowVersionId: string;
}, opts?: Oazapfts.RequestOpts) {
    return oazapfts.ok(oazapfts.fetchText(`/v1/contexts/account/stacks/${encodeURIComponent(stackVersionId)}/workflows/${encodeURIComponent(workflowVersionId)}`, {
        ...opts,
        method: "DELETE"
    }));
}
/**
 * Clear stack plugin context in account
 */
export function contextControllerclearPluginContextInAccount({ stackVersionId, pluginVersionId }: {
    stackVersionId: string;
    pluginVersionId: string;
}, opts?: Oazapfts.RequestOpts) {
    return oazapfts.ok(oazapfts.fetchText(`/v1/contexts/account/stacks/${encodeURIComponent(stackVersionId)}/plugins/${encodeURIComponent(pluginVersionId)}`, {
        ...opts,
        method: "DELETE"
    }));
}
/**
 * Clear stack action context in account
 */
export function contextControllerclearActionContextInAccount({ stackVersionId, actionVersionId }: {
    stackVersionId: string;
    actionVersionId: string;
}, opts?: Oazapfts.RequestOpts) {
    return oazapfts.ok(oazapfts.fetchText(`/v1/contexts/account/stacks/${encodeURIComponent(stackVersionId)}/actions/${encodeURIComponent(actionVersionId)}`, {
        ...opts,
        method: "DELETE"
    }));
}
