// @stack-spot/portal-network/src/api-addresses.ts
import apisItauJson from './apis-itau.json' assert { type: 'json' }
import apisJson from './apis.json' assert { type: 'json' }
import { NetworkClient } from './network/NetworkClient'
import { Env, Tenant } from './network/types'

type ApisKeys = keyof typeof apisJson;

type ApiConfig = {
  url: Record<Env, string>,
  docs?: string,
}

type PartialApiConfig = {
  url: Partial<Record<Env, string>>,
  docs?: string
}

type Apis = Record<ApisKeys, ApiConfig>;

type PartialApis = Partial<Record<ApisKeys, PartialApiConfig>>;

type ApiAddress = {
  [api: string]: string,
}

let defaultApis: Apis = apisJson;
let overrideApis: PartialApis = {};

const apis: Record<Tenant, Apis> = {
  'stackspot': apisJson,
  'itau': apisItauJson,
}

/**
 * Returns the API catalog for the given tenant.
 *
 * @param tenant The current tenant (e.g., 'stackspot', 'itau').
 * @returns API catalog containing the URLs per environment.
 */
export function getApisByTenant(tenant: Tenant) {
  return apis[tenant]
}

function transformApisToApiAddress(apis: Apis, env: Env): ApiAddress {
  const apiAddress: ApiAddress = {}

  for (const api in apis) {
    const key = api as ApisKeys
    apiAddress[key] = apis[key].url[env]
  }

  return apiAddress
}

/**
 * Searchs the default APIs configuration for the API object that matches the given baseURL.
 * @param baseUrl the base URL to search.
 * @param candidate the URL of the desired API to find.
 * @returns {boolean} true if the URL is found, otherwise false.
 */
function matchesBaseUrl(baseUrl: Record<Env, string>, candidate: Record<Env, string>): boolean {
  const envs: Env[] = ['dev', 'stg', 'prd'];

  return envs.every(e => {
    if (candidate[e]) {
      return baseUrl[e] === candidate[e];
    }
    return true;
  });
}

/**
 * Returns the base URL for the given environment and tenant, applying any overrides if available
 * 
 * @param baseURL the default API base URL.
 * @param targetEnv the target environment
 * @param tenant the current tenant.
 * @returns {string} the base URL for the given environment and tenant, applying any overrides if available.
 */
export const getBaseUrlByTenantWithOverride = (baseUrl: Record<Env, string>, targetEnv: Env = 'prd', tenant: Tenant = 'stackspot'): string => {
  let matchesApiName: ApisKeys | null = null;

  for (const [apiName, config] of Object.entries(defaultApis) as [ApisKeys, ApiConfig][]) {
    if (matchesBaseUrl(baseUrl, config.url)) {
      matchesApiName = apiName;
      break;
    }
  }

  if (!matchesApiName) return ''

  const overrideUrl = overrideApis[matchesApiName]?.url?.[targetEnv]
  if (overrideUrl) {
    return overrideUrl;
  }

  const apis = getApisByTenant(tenant)
  return apis[matchesApiName]?.url?.[targetEnv];
}

/**
 * Sets the APIs addresses for override.
 * @param customApis a JSON object containing APIs address to override.
 */
export const setApisOverride = (customApis: PartialApis) => {
  overrideApis = customApis;
}

/**
 * Sets the default APIs addresses for each environments.
 * @returns {Apis} an object containing APIs URLs grouped by api name and enviroments.
 */
export const getApiAddresses = (): Apis => {
  return defaultApis;
}

export const apiAddresses = () => {
  const env = NetworkClient.getEnv()
  return transformApisToApiAddress(defaultApis, env)
}
