// @stack-spot/portal-network/dist/error/dictionary/agent-tools.js
export const agentToolsDictionary = {
    en: {
        AGENT_1000_INVALID_JWT: 'Invalid JWT token',
        AGENT_1001_DECODE_JWT_ERROR: 'Error decoding JWT token',
        AGENTS_6000_FORBIDDEN_1: 'Insufficient permissions',
        AGENTS_6000_FORBIDDEN_2: 'You do not have permission to access the specified secret.',
        AGENTS_6000_FORBIDDEN_3: 'Toolkit publication is restricted to account secrets or if secrets are not present.',
        AGENT_1002_STRUCT_OUTPUT_WITH_BUILTIN_TOOLS_NOT_PERMITED: 'Structured output cannot be used alongside with builtin tools',
        AGENT_1003_REQUIRED_LLM_PARAM_NOT_PROVIDED: 'Required parameter of llm model was not provided',
        AGENT_1004_INVALID_LLM_PARAM: "The parameter '{param}' is invalid",
        AGENT_1005_INVALID_KNOWLEDGE_SOURCE_ID: 'Knowledge source is invalid',
        AGENT_1006_INVALID_LLM_PARAM_VALUE_TYPE: 'LLM param value is invalid',
        AGENT_1007_LLM_PARAM_VALUE_IS_REQUIRED: 'LLM param value is required',
        AGENT_1008_INVALID_ULID_FOR_MODEL_ID: 'Invalid ulid passed for model id',
        AGENT_1009_INVALID_STRUCTURE_OUTPUT_SCHEMA: 'Structure output schema is invalid. Please, check our documentation or contact our support',
        AGENT_1010_STRUCTURED_OUTPUT_SCHEMA_CONTAINS_INVALID_DATA: 'Your structure output schema contains invalid data. Please, check our documentation or contact our support',
        AGENT_1011_AGENT_WITH_SLUG_ALREADY_EXISTS: 'An agent with this slug already exists',
        AGENT_1013_INVALID_VISIBILITY: 'Visibility is invalid',
        AGENT_1015_AGENT_HAS_PERSONAL_SECRET_TOOL: 'It is not possible to publish this agent, because it has a tool with a personal secret',
        AGENT_2002_INVALID_VISIBILITY_CHANGE: 'Cannot change visibility',
        AGENT_2003_SECRETS_WITH_SAME_TYPE: 'Secrets with the same type are not allowed.',
        TOOL_1200_TOOL_IN_USE: 'You have a tool that is already in use by an Agent and therefore cannot be removed.',
        AGENT_1101_TOOLKIT_IN_USE: 'The toolkit is already in use by an Agent and therefore cannot be removed.',
        AGENT_1101_TOOLKIT_INVALID_CREDENTIAL_TYPE: 'Your toolkit has an invalid type of authorization. Only Client Credentials, and API Key are accepted.',
        AGENT_1102_TOOLKIT_AUTHENTICATION_FAILED: 'An error occurred during authentication in the toolkit.',
        AGENT_1105_TOOLKIT_WITH_PERSONAL_SECRET: 'It is not possible to share the agent, as it has a tool with a personal secret. Remove the tool or replace it with an organization secret.',
        AGENT_1012_AGENT_NOT_FOUND: 'Agent with id was not found',
        AGENT_1014_AGENT_TOOL_NOT_FOUND: 'Agent Tool with id was not found',
        AGENT_1100_TOOLKIT_NOT_FOUND: 'Toolkit with id was not found',
        TOOL_1201_TOOL_NOT_FOUND: 'Tool was not found',
        TOOL_1202_INVALID_SUB_AGENT_ASSIGNMENT: 'Cannot assign sub-agents, this agent is already a sub-agent of another agent.',
        TOOL_1203_INVALID_SUB_AGENT_SELF_ASSIGNMENT: 'Agent cannot be assigned as its own tool. Please select a different agent.',
        TOOL_1204_INVALID_SUB_AGENT_AS_ORCHESTRATOR: 'Cannot assign an orchestrator agent as a sub-agent to another orchestrator. Only single-level agent delegation is supported.',
        AGENT_5000_UNEXPECTED_ERROR: 'An unexpected error occurred. Try again later or contact our support',
        AGENT_2000_ACCOUNT_API_ERROR: 'Error with Account API integration',
        AGENT_2001_IAM_API_ERROR: 'Error with IAM API integration',
    },
    pt: {
        AGENT_1000_INVALID_JWT: 'Token JWT inválido',
        AGENT_1001_DECODE_JWT_ERROR: 'Erro ao decodificar o token JWT',
        AGENTS_6000_FORBIDDEN_1: 'Permissões insuficientes',
        AGENTS_6000_FORBIDDEN_2: 'Você não tem permissão para acessar o segredo especificado.',
        AGENTS_6000_FORBIDDEN_3: 'A publicação do toolkit é restrita a segredos de conta ou se segredos não estiverem presentes.',
        AGENT_1002_STRUCT_OUTPUT_WITH_BUILTIN_TOOLS_NOT_PERMITED: 'A saída estruturada não pode ser usada juntamente com ferramentas internas.',
        AGENT_1003_REQUIRED_LLM_PARAM_NOT_PROVIDED: 'Parâmetro obrigatório do modelo LLM não foi fornecido',
        AGENT_1004_INVALID_LLM_PARAM: "O parâmetro '{param}' é inválido",
        AGENT_1005_INVALID_KNOWLEDGE_SOURCE_ID: 'Fonte de conhecimento inválida',
        AGENT_1006_INVALID_LLM_PARAM_VALUE_TYPE: 'Valor do parâmetro LLM é inválido',
        AGENT_1007_LLM_PARAM_VALUE_IS_REQUIRED: 'Valor do parâmetro LLM é obrigatório',
        AGENT_1008_INVALID_ULID_FOR_MODEL_ID: 'ULID inválido fornecido para o ID do modelo',
        AGENT_1009_INVALID_STRUCTURE_OUTPUT_SCHEMA: 'O schema de saída da estrutura é inválido. Por favor, verifique nossa documentação ou entre em contato com o suporte',
        AGENT_1010_STRUCTURED_OUTPUT_SCHEMA_CONTAINS_INVALID_DATA: 'Seu schema de saída da estrutura contém dados inválidos. Por favor, verifique nossa documentação ou entre em contato com o suporte',
        AGENT_1011_AGENT_WITH_SLUG_ALREADY_EXISTS: 'Já existe um agente com este slug',
        AGENT_1013_INVALID_VISIBILITY: 'Visibilidade inválida',
        AGENT_1015_AGENT_HAS_PERSONAL_SECRET_TOOL: 'Não é possível publicar este agente, pois ele possui uma ferramenta com segredo pessoal',
        AGENT_2002_INVALID_VISIBILITY_CHANGE: 'Não é possível alterar a visibilidade',
        AGENT_2003_SECRETS_WITH_SAME_TYPE: 'Segredos do mesmo tipo não são permitidos.',
        TOOL_1200_TOOL_IN_USE: 'Você possui uma ferramenta que já está em uso por um agente e, portanto, não pode ser removida.',
        AGENT_1101_TOOLKIT_IN_USE: 'O toolkit já está em uso por um agente e, portanto, não pode ser removido.',
        AGENT_1101_TOOLKIT_INVALID_CREDENTIAL_TYPE: 'Seu toolkit possui um tipo de autorização inválido. Apenas Client Credentials e API Key são aceitos.',
        AGENT_1102_TOOLKIT_AUTHENTICATION_FAILED: 'Ocorreu um erro durante a autenticação no toolkit.',
        AGENT_1105_TOOLKIT_WITH_PERSONAL_SECRET: 'Não é possível compartilhar o agente, pois ele possui uma ferramenta com segredo pessoal. Remova a ferramenta ou substitua por um segredo da organização.',
        AGENT_1012_AGENT_NOT_FOUND: 'Agente com o ID informado não foi encontrado',
        AGENT_1014_AGENT_TOOL_NOT_FOUND: 'Ferramenta do agente com o ID informado não foi encontrada',
        AGENT_1100_TOOLKIT_NOT_FOUND: 'Toolkit com o ID informado não foi encontrado',
        TOOL_1201_TOOL_NOT_FOUND: 'Ferramenta não encontrada',
        TOOL_1202_INVALID_SUB_AGENT_ASSIGNMENT: 'Não é possível atribuir subagentes, este agente já é subagente de outro agente.',
        TOOL_1203_INVALID_SUB_AGENT_SELF_ASSIGNMENT: 'O agente não pode ser atribuído como sua própria ferramenta. Por favor, selecione outro agente.',
        TOOL_1204_INVALID_SUB_AGENT_AS_ORCHESTRATOR: 'Não é possível atribuir um agente orquestrador como subagente de outro orquestrador. Apenas delegação de agente em um único nível é suportada.',
        AGENT_5000_UNEXPECTED_ERROR: 'Ocorreu um erro inesperado. Tente novamente mais tarde ou entre em contato com o suporte',
        AGENT_2000_ACCOUNT_API_ERROR: 'Erro na integração com a API de Conta',
        AGENT_2001_IAM_API_ERROR: 'Erro na integração com a API IAM',
    },
};
