// @stack-spot/portal-network/dist/error/dictionary/ai-inference.js
export const inferenceDictionary = {
    en: {
        INFERENCE_2500_AGENT_TOOL_INTEGRATION_ERROR: 'Error integrating with the agent tool.',
        INFERENCE_2501_MODEL_DOES_NOT_EXISTS: 'Model does not exists.',
        INFERENCE_2502_MODEL_RESOURCE_DOES_NOT_EXISTS: 'Model resource does not exists',
        INFERENCE_2503_UNABLE_TO_DELETE_DEFAULT_MODEL_RESOURCE: 'Unable to delete the default model resource.',
        INFERENCE_2504_MODEL_RESOURCE_DEFAULT_UPDATE_NOT_ALLOWED: 'Cannot update default resource {1} to non-default',
        INFERENCE_2505_MODEL_DEFAULT_RESOURCES: 'Cannot deactivate model with default resources',
        INFERENCE_3001_SELF_HOSTED_PROVIDER_WITHOUT_ACCOUNT_ID: 'Self-hosted provider is missing an account ID.',
        INFERENCE_3002_MODEL_NOT_ACTIVE: 'The model is not active.',
        INFERENCE_3003_LLM_PROVIDER_NOT_SUPPORTED: 'The LLM provider is not supported.',
        INFERENCE_3004_INVALID_REQUESTED_MODEL: 'The requested model is invalid.',
    },
    pt: {
        INFERENCE_2500_AGENT_TOOL_INTEGRATION_ERROR: 'Erro ao integrar com a ferramenta do agente.',
        INFERENCE_2501_MODEL_DOES_NOT_EXISTS: 'O modelo especificado não existe.',
        INFERENCE_2502_MODEL_RESOURCE_DOES_NOT_EXISTS: 'O recurso do modelo não existe.',
        INFERENCE_2503_UNABLE_TO_DELETE_DEFAULT_MODEL_RESOURCE: 'Não é possível excluir o recurso padrão do modelo.',
        INFERENCE_2504_MODEL_RESOURCE_DEFAULT_UPDATE_NOT_ALLOWED: 'Não é possível atualizar o recurso padrão {1} para não padrão.',
        INFERENCE_2505_MODEL_DEFAULT_RESOURCES: 'Não é possível desativar o modelo com recursos padrão.',
        INFERENCE_3001_SELF_HOSTED_PROVIDER_WITHOUT_ACCOUNT_ID: 'O provedor auto-hospedado está sem um ID de conta.',
        INFERENCE_3002_MODEL_NOT_ACTIVE: 'O modelo não está ativo.',
        INFERENCE_3003_LLM_PROVIDER_NOT_SUPPORTED: 'O provedor LLM não é suportado.',
        INFERENCE_3004_INVALID_REQUESTED_MODEL: 'O modelo solicitado é inválido.',
    },
};
