// @stack-spot/portal-network/dist/client/agent.js
import { defaults, deleteV1AgentByAgentIdFavorite, getV1AgentByAgentId, getV1Agents, getV1PublicAgentByAgentId, getV1PublicAgents, postV1AgentByAgentIdFavorite, putV1AgentByAgentId, } from '../api/agent.js';
import { StackspotAPIError } from '../error/StackspotAPIError.js';
import { ReactQueryNetworkClient } from '../network/ReactQueryNetworkClient.js';
import { getApiAddresses } from '../api-addresses.js';
export const isAgentDefault = (agentSlug) => agentSlug === 'stk_code_buddy';
class AgentClient extends ReactQueryNetworkClient {
    constructor() {
        super(getApiAddresses().agent.url, defaults);
        /**
         * Gets an agent by id
         */
        Object.defineProperty(this, "agentById", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: this.query({
                name: 'agentById',
                request: (signal, variables) => variables.builtIn
                    ? getV1PublicAgentByAgentId({ ...variables }, { signal })
                    : getV1AgentByAgentId({ ...variables }, { signal }),
            })
        });
        /**
         * List commons agents
         */
        Object.defineProperty(this, "agents", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: this.query(getV1Agents)
        });
        /**
         * Gets a common agent by id
         */
        Object.defineProperty(this, "agent", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: this.query(getV1AgentByAgentId)
        });
        /**
         * List all public agents (built-in)
         */
        Object.defineProperty(this, "publicAgents", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: this.query(getV1PublicAgents)
        });
        /**
         * Gets a public agent by id (built-in)
         */
        Object.defineProperty(this, "publicAgent", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: this.query(getV1PublicAgentByAgentId)
        });
        /**
         * Adds the resource of type Agent to the list of favorites.
         */
        Object.defineProperty(this, "addFavoriteAgent", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: this.mutation(postV1AgentByAgentIdFavorite)
        });
        /**
         * Removes the resource of type Agent from the list of favorites.
         */
        Object.defineProperty(this, "removeFavoriteAgent", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: this.mutation(deleteV1AgentByAgentIdFavorite)
        });
        /**
         * Updates an agent
         */
        Object.defineProperty(this, "updateAgent", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: this.mutation(putV1AgentByAgentId)
        });
    }
    buildStackSpotError(error) {
        const errorData = error.data;
        const message = [errorData?.details];
        Object.keys(errorData?.additionalInformation ?? {}).forEach(k => message.push(`  - ${k}: ${errorData?.additionalInformation?.[k]}`));
        return new StackspotAPIError({
            status: error.status,
            headers: error.headers,
            stack: error.stack,
            code: error.data?.code,
            message: message.join('\n'),
        });
    }
}
export const agentClient = new AgentClient();
