// @stack-spot/portal-network/dist/client/agent-tools.js
import { getApiAddresses } from '../api-addresses.js';
import { addFavoriteV1AgentsAgentIdFavoritePost, createAgentV1AgentsPost, createToolkitToolsV1ToolkitsToolkitIdToolsPost, createToolkitV1ToolkitsPost, defaults, deleteAgentV1AgentsAgentIdDelete, deleteFavoriteV1AgentsAgentIdFavoriteDelete, deleteToolkitToolsV1ToolkitsToolkitIdToolsDelete, deleteToolkitV1ToolkitsToolkitIdDelete, editToolkitToolV1ToolkitsToolkitIdToolsToolIdPut, forkAgentV1AgentsAgentIdForkPost, forkToolkitV1ToolkitsToolkitIdForkPost, getAgentV1AgentsAgentIdGet, getPublicToolKitsV1BuiltinToolkitGet, getToolkitToolV1ToolkitsToolkitIdToolsToolIdGet, getToolkitV1ToolkitsToolkitIdGet, listAgentsUsingToolsV1ToolkitsToolkitIdToolsAgentsPost, listAgentsV1AgentsGet, listAgentsV3AgentsGet, listMcpToolsV1McpToolsToolkitIdGet, listMultiAgentsV1AgentsMultiAgentsGet, listToolkitsV1ToolkitsGet, listToolkitsV2ToolkitsGet, publishAgentV1AgentsAgentIdPublishPost, searchAgentsV1AgentsSearchPost, shareV1AgentsAgentIdSharePost, updateAgentV1AgentsAgentIdPatch, updateToolkitV1ToolkitsToolkitIdPatch } from '../api/agent-tools.js';
import { DefaultAPIError } from '../error/DefaultAPIError.js';
import { agentToolsDictionary } from '../error/dictionary/agent-tools.js';
import { ReactQueryNetworkClient } from '../network/ReactQueryNetworkClient.js';
import { removeAuthorizationParam } from '../utils/remove-authorization-param.js';
import { StreamedArray } from '../utils/StreamedArray.js';
import { workspaceAiClient } from './workspace-ai.js';
const AGENT_DEFAULT_SLUG = 'stk_flex';
const listAgentsV3AgentsWithoutAuthorization = removeAuthorizationParam(listAgentsV3AgentsGet);
class AgentToolsClient extends ReactQueryNetworkClient {
    constructor() {
        super(getApiAddresses()['agent-tools'].url, defaults);
        /**
         * Lists all public toolkits.
         */
        Object.defineProperty(this, "tools", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: this.query(removeAuthorizationParam(getPublicToolKitsV1BuiltinToolkitGet))
        });
        /**
         * Create agent
         */
        Object.defineProperty(this, "createAgent", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: this.mutation(removeAuthorizationParam(createAgentV1AgentsPost))
        });
        /**
         * Delete agent
         */
        Object.defineProperty(this, "deleteAgent", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: this.mutation(removeAuthorizationParam(deleteAgentV1AgentsAgentIdDelete))
        });
        /**
         * Updates an agent
         */
        Object.defineProperty(this, "updateAgent", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: this.mutation(removeAuthorizationParam(updateAgentV1AgentsAgentIdPatch))
        });
        /**
        * Favorite an agent
        */
        Object.defineProperty(this, "favoriteAgent", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: this.mutation(removeAuthorizationParam(addFavoriteV1AgentsAgentIdFavoritePost))
        });
        /**
        * Publish an agent
        */
        Object.defineProperty(this, "publishAgent", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: this.mutation(removeAuthorizationParam(publishAgentV1AgentsAgentIdPublishPost))
        });
        /**
         * List agents
         */
        Object.defineProperty(this, "agents", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: this.infiniteQuery(removeAuthorizationParam(listAgentsV1AgentsGet))
        });
        /**
         * List agents with support for multiple filters
         */
        Object.defineProperty(this, "agentsMultipleFilters", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: this.infiniteQuery(listAgentsV3AgentsWithoutAuthorization, {
                pageParamName: 'filters.page',
                accumulator: 'items',
                getNextPageParam: ({ variables, lastPage, lastPageParam }) => {
                    const size = variables.filters.size ?? 1;
                    const parsedLastPageParam = lastPageParam ?? variables.filters.page ?? 1;
                    return lastPage.items && lastPage.items.length < size ? undefined : parsedLastPageParam + 1;
                },
            })
        });
        /**
         * List agents available to be added to other agents
         */
        Object.defineProperty(this, "availableAgentsForMultiAgentAddition", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: this.query(removeAuthorizationParam(listMultiAgentsV1AgentsMultiAgentsGet))
        });
        /**
         * Gets agent by id
         */
        Object.defineProperty(this, "agent", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: this.query(removeAuthorizationParam(getAgentV1AgentsAgentIdGet))
        });
        /**
         * Gets agents by ids
         */
        Object.defineProperty(this, "agentsByIds", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: this.query(removeAuthorizationParam(searchAgentsV1AgentsSearchPost))
        });
        /**
         * Gets the default agent slug
         */
        Object.defineProperty(this, "agentDefaultSlug", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: AGENT_DEFAULT_SLUG
        });
        /**
         * Gets the default agent
         */
        Object.defineProperty(this, "agentDefault", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: this.query({
                name: 'agentDefault',
                request: async (signal) => {
                    const agentDefault = await listAgentsV1AgentsGet({ visibility: 'built_in', slug: this.agentDefaultSlug, authorization: '' }, { signal });
                    const agentId = agentDefault?.find((agent) => agent.slug === this.agentDefaultSlug)?.id;
                    const agent = agentId ? await this.agent.query({ agentId }) : undefined;
                    return agent;
                },
            })
        });
        /**
         * List agents filtered by visibility.
         */
        Object.defineProperty(this, "allAgents", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: this.query({
                name: 'allAgents',
                request: async (signal, variables) => {
                    const allVisibilities = ['account', 'built_in', 'recently_used', 'favorite', 'personal', 'shared', 'workspace'];
                    const visibilities = variables.visibilities.includes('all')
                        ? allVisibilities
                        : variables.visibilities;
                    const shouldFetchWorkspaceAgents = visibilities.includes('workspace');
                    const workspaceAgentsPromise = shouldFetchWorkspaceAgents
                        ? workspaceAiClient.workspacesContentsByType.query({ contentType: 'agent' })
                        : Promise.resolve([]);
                    const [workspaceAgents, ...agentsByVisibility] = await Promise.all([
                        workspaceAgentsPromise,
                        ...visibilities.map((visibility) => listAgentsV1AgentsGet({ visibility, authorization: '' }, { signal })),
                    ]);
                    const workspaceAgentsWithSpaceName = workspaceAgents.flatMap(({ agents, space_name }) => agents?.map((agent) => ({ ...agent, spaceName: space_name, builtIn: false })));
                    const allAgents = workspaceAgentsWithSpaceName ?? [];
                    agentsByVisibility.forEach(agents => allAgents.push(...agents.map(agent => ({
                        ...agent,
                        builtIn: agent?.visibility_level === 'built_in',
                    }))));
                    return allAgents;
                },
            })
        });
        /**
         * Lists all custom toolkits.
         */
        Object.defineProperty(this, "toolkits", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: this.query(removeAuthorizationParam(listToolkitsV1ToolkitsGet))
        });
        /**
        * Get list of Toolkits V2
        */
        Object.defineProperty(this, "toolkitsV2", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: this.query(removeAuthorizationParam(listToolkitsV2ToolkitsGet))
        });
        /**
         * Lists all tools of an MCP toolkit.
         */
        Object.defineProperty(this, "mcpTools", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: this.query(removeAuthorizationParam(listMcpToolsV1McpToolsToolkitIdGet))
        });
        /**
         * Get a toolkit by Id
         */
        Object.defineProperty(this, "toolkitById", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: this.query(removeAuthorizationParam(getToolkitV1ToolkitsToolkitIdGet))
        });
        /**
         * Create a toolkit
         */
        Object.defineProperty(this, "createToolkit", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: this.mutation(removeAuthorizationParam(createToolkitV1ToolkitsPost))
        });
        /**
         * Update a toolkit
         */
        Object.defineProperty(this, "updateToolkit", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: this.mutation(removeAuthorizationParam(updateToolkitV1ToolkitsToolkitIdPatch))
        });
        /**
         * Delete a toolkit
         */
        Object.defineProperty(this, "deleteToolkit", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: this.mutation(removeAuthorizationParam(deleteToolkitV1ToolkitsToolkitIdDelete))
        });
        /**
         * Get Tool from Toolkit
         */
        Object.defineProperty(this, "toolFromToolkit", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: this.query(removeAuthorizationParam(getToolkitToolV1ToolkitsToolkitIdToolsToolIdGet))
        });
        /**
         * Add tool on Toolkit
         */
        Object.defineProperty(this, "addToolOnToolkit", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: this.mutation(removeAuthorizationParam(createToolkitToolsV1ToolkitsToolkitIdToolsPost))
        });
        /**
         * Edit tool on Toolkit
         */
        Object.defineProperty(this, "editToolOnToolkit", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: this.mutation(removeAuthorizationParam(editToolkitToolV1ToolkitsToolkitIdToolsToolIdPut))
        });
        /**
         * Delete tool from Toolkit
         */
        Object.defineProperty(this, "deleteToolFromToolkit", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: this.mutation(removeAuthorizationParam(deleteToolkitToolsV1ToolkitsToolkitIdToolsDelete))
        });
        /**
         * Previews an uploaded open api file as a stream where each array item is an OpenAPI operation.
         */
        Object.defineProperty(this, "streamUploadedAPIs", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: ({ fileUploadIds, signal }) => {
                const promises = fileUploadIds.map((id) => {
                    const url = `/v1/toolkits/tools/preview/${encodeURIComponent(id)}`;
                    return this.stream(url, { signal });
                });
                return new StreamedArray({ eventsPromises: promises, signal });
            }
        });
        /**
         * Lists the agents using the tools passed as parameter.
         */
        Object.defineProperty(this, "agentsUsingTools", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: this.query(removeAuthorizationParam(listAgentsUsingToolsV1ToolkitsToolkitIdToolsAgentsPost))
        });
        /**
         * Forks the toolkit.
         */
        Object.defineProperty(this, "fork", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: this.mutation(removeAuthorizationParam(forkToolkitV1ToolkitsToolkitIdForkPost))
        });
        /**
         * Share Agent
         */
        Object.defineProperty(this, "shareAgent", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: this.mutation(removeAuthorizationParam(shareV1AgentsAgentIdSharePost))
        });
        /**
         * Fork Agent
         */
        Object.defineProperty(this, "forkAgent", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: this.mutation(removeAuthorizationParam(forkAgentV1AgentsAgentIdForkPost))
        });
        /**
        * Add and agent to favorites
        */
        Object.defineProperty(this, "addFavorite", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: this.mutation(removeAuthorizationParam(addFavoriteV1AgentsAgentIdFavoritePost))
        });
        /**
        * remove an agent from favorites
        */
        Object.defineProperty(this, "removeFavorite", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: this.mutation(removeAuthorizationParam(deleteFavoriteV1AgentsAgentIdFavoriteDelete))
        });
    }
    buildStackSpotError(error) {
        return new DefaultAPIError(error.data, error.status, agentToolsDictionary, error.headers);
    }
}
export const agentToolsClient = new AgentToolsClient();
