// @stack-spot/portal-network/dist/client/data-integration.js
import { getApiAddresses } from '../api-addresses.js';
import { addFavoriteV1KnowledgeSourcesSlugFavoritePost, createKsV1KnowledgeSourcesPost, defaults, deleteFavoriteV1KnowledgeSourcesSlugFavoriteDelete, deleteKoByIdV1KnowledgeSourcesKsSlugObjectsKoIdDelete, findKoByIdV1KnowledgeSourcesKsSlugObjectsKoIdGet, findKsV1KnowledgeSourcesSlugGet, forkKsV1KnowledgeSourcesSlugForkPost, getFileUploadStatusV1FileUploadFileUploadIdGet, getIndexedFilesV1KnowledgeSourcesKsSlugFilesGet, getUploadFormV1FileUploadFormPost, getUploadFormV2V2FileUploadFormPost, listKosV1KnowledgeSourcesKsSlugObjectsGet, listKosV2KnowledgeSourcesKsSlugObjectsGet, listKsV1KnowledgeSourcesGet, listKsV2KnowledgeSourcesGet, publishKsV1KnowledgeSourcesSlugPublishPost, resetKosV1KnowledgeSourcesKsSlugObjectsDelete, saveChunkedFilesDeprecatedV1FileUploadFileUploadIdPost, saveChunkedFilesV1FileUploadFileUploadIdKnowledgeObjectsPost, searchKsV1KnowledgeSourcesSearchPost, searchObjectsV1SearchPost, shareKsV1KnowledgeSourcesSlugSharePost, splitUploadedFilePreviewV1FileUploadFileUploadIdSplitPreviewSplitStrategyGet, splitUploadedFileV1FileUploadFileUploadIdSplitPost, updateKsV1KnowledgeSourcesSlugPatch, } from '../api/dataIntegration.js';
import { DefaultAPIError } from '../error/DefaultAPIError.js';
import { baseDictionary } from '../error/dictionary/base.js';
import { FileUploadError } from '../error/FileUploadError.js';
import { StackspotAPIError } from '../error/StackspotAPIError.js';
import { ReactQueryNetworkClient } from '../network/ReactQueryNetworkClient.js';
import { removeAuthorizationParam } from '../utils/remove-authorization-param.js';
const listKSV2WithoutAuthorization = removeAuthorizationParam(listKsV2KnowledgeSourcesGet);
const listKOV2WithoutAuthorization = removeAuthorizationParam(listKosV2KnowledgeSourcesKsSlugObjectsGet);
class DataIntegrationClient extends ReactQueryNetworkClient {
    constructor() {
        super(getApiAddresses().dataIntegration.url, defaults);
        /**
         * Generates a url to upload the file
         */
        Object.defineProperty(this, "generateUrlToFileUpload", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: this.mutation(removeAuthorizationParam(getUploadFormV2V2FileUploadFormPost))
        });
        /**
         * Get file processing status by id
         */
        Object.defineProperty(this, "getFileProcessingStatusById", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: this.query(removeAuthorizationParam(getFileUploadStatusV1FileUploadFileUploadIdGet))
        });
        /**
         * @deprecated Use saveChunkedKnowledgeObjects instead.
         * Save the splitted knowledge objects
         */
        Object.defineProperty(this, "saveKnowledgeObjects", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: this.mutation(removeAuthorizationParam(saveChunkedFilesDeprecatedV1FileUploadFileUploadIdPost))
        });
        /**
         * Save the splitted knowledge objects
         */
        Object.defineProperty(this, "saveChunkedKnowledgeObjects", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: this.mutation(removeAuthorizationParam(saveChunkedFilesV1FileUploadFileUploadIdKnowledgeObjectsPost))
        });
        /**
         * Splits the file with the passed settings
         */
        Object.defineProperty(this, "splitFile", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: this.mutation(removeAuthorizationParam(splitUploadedFileV1FileUploadFileUploadIdSplitPost))
        });
        /**
         * Generate a preview with splitted objects
         */
        Object.defineProperty(this, "previewFromSplit", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: this.query(removeAuthorizationParam(splitUploadedFilePreviewV1FileUploadFileUploadIdSplitPreviewSplitStrategyGet))
        });
        /**
         * List of knowledge sources
         */
        Object.defineProperty(this, "knowledgeSources", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: this.query(removeAuthorizationParam(listKsV1KnowledgeSourcesGet))
        });
        /**
         * List of knowledge sources with pagination
         */
        Object.defineProperty(this, "knowledgeSourcesV2", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: this.infiniteQuery(listKSV2WithoutAuthorization, { accumulator: 'items' })
        });
        /**
         * Adds the resource of type Knowledge Source to the list of favorites.
         */
        Object.defineProperty(this, "addFavoriteKnowledgeSource", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: this.mutation(removeAuthorizationParam(addFavoriteV1KnowledgeSourcesSlugFavoritePost))
        });
        /**
         * Removes the resource of type Knowledge Source from the list of favorites.
         */
        Object.defineProperty(this, "removeFavoriteKnowledgeSource", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: this.mutation(removeAuthorizationParam(deleteFavoriteV1KnowledgeSourcesSlugFavoriteDelete))
        });
        /**
         * List knowledge objects from a knowledge source
         */
        Object.defineProperty(this, "listKnowledgeObjects", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: this.query(removeAuthorizationParam(listKosV1KnowledgeSourcesKsSlugObjectsGet))
        });
        /**
         * List knowledge objects from a knowledge source with pagination
         */
        Object.defineProperty(this, "listKnowledgeObjectsV2", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: this.infiniteQuery(listKOV2WithoutAuthorization, { accumulator: 'items' })
        });
        /**
         * Get a knowledge object by id
         */
        Object.defineProperty(this, "getKnowledgeObjectById", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: this.query(removeAuthorizationParam(findKoByIdV1KnowledgeSourcesKsSlugObjectsKoIdGet))
        });
        Object.defineProperty(this, "uploadFiles", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: this.mutation({
                name: 'uploadFiles',
                request: (abortSignal, params) => (Promise.all(params.files.map(async (f) => {
                    try {
                        const { id, url, form } = await getUploadFormV2V2FileUploadFormPost({
                            authorization: '',
                            newFileUploadRequest: { file_name: f.name, target_type: params.type, target_id: params.targetId ?? '' },
                        }, { signal: abortSignal });
                        const formData = new FormData();
                        Object.entries(form).forEach(([key, value]) => {
                            formData.append(key, value);
                        });
                        formData.append('file', f);
                        const response = await fetch(url, { method: 'post', body: formData, signal: abortSignal });
                        if (!response.ok)
                            throw new StackspotAPIError({ status: response.status, headers: response.headers, message: 'Failed to upload' });
                        return id;
                    }
                    catch (e) {
                        throw new FileUploadError(f.name, e);
                    }
                }))),
            })
        });
        /**
         * Uploads a KO file
        */
        Object.defineProperty(this, "fileUploadFormKnowledgeObject", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: this.mutation(removeAuthorizationParam(getUploadFormV1FileUploadFormPost))
        });
        /**
         * Create Ks
          */
        Object.defineProperty(this, "createKnowledgeSource", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: this.mutation(removeAuthorizationParam(createKsV1KnowledgeSourcesPost))
        });
        /**
         * Search Ks
          */
        Object.defineProperty(this, "listKnowledgeSourcesDataByIds", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: this.query(removeAuthorizationParam(searchKsV1KnowledgeSourcesSearchPost))
        });
        /**
         * Gets a knowledge source by its slug.
          */
        Object.defineProperty(this, "findKnowledgeSourceBySlug", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: this.query(removeAuthorizationParam(findKsV1KnowledgeSourcesSlugGet))
        });
        /**
         * Search Objects
         */
        Object.defineProperty(this, "searchObjects", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: this.mutation(removeAuthorizationParam(searchObjectsV1SearchPost))
        });
        /**
         * Update Knowledge Source
         */
        Object.defineProperty(this, "updateKnowledgeSource", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: this.mutation(removeAuthorizationParam(updateKsV1KnowledgeSourcesSlugPatch))
        });
        /**
         * Share Knowledge Source
         */
        Object.defineProperty(this, "shareKnowledgeSource", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: this.mutation(removeAuthorizationParam(shareKsV1KnowledgeSourcesSlugSharePost))
        });
        /**
         * Publish Knowledge Source
         */
        Object.defineProperty(this, "publishKnowledgeSource", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: this.mutation(removeAuthorizationParam(publishKsV1KnowledgeSourcesSlugPublishPost))
        });
        /**
         * Fork Knowledge Source
         */
        Object.defineProperty(this, "forkKnowledgeSource", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: this.mutation(removeAuthorizationParam(forkKsV1KnowledgeSourcesSlugForkPost))
        });
        /**
         * Delete Knowledge Objects by Standalone
         */
        Object.defineProperty(this, "deleteKnowledgeObjectsByStandalone", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: this.mutation(removeAuthorizationParam(resetKosV1KnowledgeSourcesKsSlugObjectsDelete))
        });
        /**
         * Delete Knowledge Object from Knowledge Source
         */
        Object.defineProperty(this, "deleteKnowledgeObjectFromKnowledgeSource", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: this.mutation(removeAuthorizationParam(deleteKoByIdV1KnowledgeSourcesKsSlugObjectsKoIdDelete))
        });
        /**
         * Get Indexed Files
         */
        Object.defineProperty(this, "listIndexedProjectFiles", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: this.query(removeAuthorizationParam(getIndexedFilesV1KnowledgeSourcesKsSlugFilesGet))
        });
    }
    buildStackSpotError(error) {
        return new DefaultAPIError(error.data, error.status, baseDictionary, error.headers);
    }
    /**
     * Generate a preview with splitted objects
     */
    streamPreviewFromSplit(fileUploadId, splitStrategy, options) {
        const url = `/v1/file-upload/${fileUploadId}/split/preview/${splitStrategy}`;
        return this.stream(this.resolveURL(url), {
            method: 'GET',
            headers: {
                'Content-Type': 'application/json',
                'Accept': 'text/event-stream',
            },
            signal: options?.signal,
        });
    }
}
export const dataIntegrationClient = new DataIntegrationClient();
