// @stack-spot/portal-network/dist/client/workspace-ai.js
import { addContentV1WorkspacesWorkspaceIdContentTypePost, addFavoriteV1WorkspacesWorkspaceIdFavoritePost, addPermissionV1ResourceTypeResourceTypeResourcesResourceIdentifierPatch, addPermissionV1WorkspacesWorkspaceIdPermissionsPatch, createShareableLinkV1WorkspacesWorkspaceIdSharePost, createWorkspaceV1WorkspacesPost, defaults, deleteFavoriteV1WorkspacesWorkspaceIdFavoriteDelete, deleteWorkspaceV1WorkspacesWorkspaceIdDelete, getWorkspaceByIdV1WorkspacesIdGet, listAllByContentTypeV1WorkspacesWorkspaceIdContentTypeGet, listAllContentsV1WorkspacesContentsContentTypeGet, listGroupsAndMembersPermissionInWorkspaceV1WorkspacesWorkspaceIdPermissionsGet, listMembersV1ResourceTypeResourceTypeResourcesResourceIdentifierMembersGet, listWorkspacesV1WorkspacesGet, removeContentV1WorkspacesWorkspaceIdContentTypeContentIdDelete, removePermissionV1ResourceTypeResourceTypeResourcesResourceIdentifierDelete, removePermissionV1WorkspacesWorkspaceIdPermissionsDelete, updateWorkspaceV1WorkspacesWorkspaceIdPatch, useShareableLinkV1ShareShareIdPatch, } from '../api/workspace-ai.js';
import { ReactQueryNetworkClient } from '../network/ReactQueryNetworkClient.js';
import { removeAuthorizationParam } from '../utils/remove-authorization-param.js';
import { getApiAddresses } from '../api-addresses.js';
import { DefaultAPIError } from '../error/DefaultAPIError.js';
import { wksAiDictionary } from '../error/dictionary/workspace-ai.js';
class WorkspaceAiClient extends ReactQueryNetworkClient {
    constructor() {
        super(getApiAddresses()['workspace-ai'].url, defaults);
        Object.defineProperty(this, "createWorkspaceAi", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: this.mutation(removeAuthorizationParam(createWorkspaceV1WorkspacesPost))
        });
        Object.defineProperty(this, "workspacesAi", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: this.query(removeAuthorizationParam(listWorkspacesV1WorkspacesGet))
        });
        Object.defineProperty(this, "workspaceAi", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: this.query(removeAuthorizationParam(getWorkspaceByIdV1WorkspacesIdGet))
        });
        Object.defineProperty(this, "updateWorkspaceAi", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: this.mutation(removeAuthorizationParam(updateWorkspaceV1WorkspacesWorkspaceIdPatch))
        });
        Object.defineProperty(this, "deleteWorkspaceAi", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: this.mutation(removeAuthorizationParam(deleteWorkspaceV1WorkspacesWorkspaceIdDelete))
        });
        /**
         * Adds a resource to the Spot.
         */
        Object.defineProperty(this, "addResourceTypeToWorkspaceAi", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: this.mutation({
                name: 'addResourceTypeToWorkspaceAi',
                request: (signal, variables) => addContentV1WorkspacesWorkspaceIdContentTypePost({ ...variables, authorization: '' }, { signal }),
                permission: this.createPermissionFunctionFor(addContentV1WorkspacesWorkspaceIdContentTypePost),
            })
        });
        /**
         * Lists all the knowledge sources in a Spot.
         */
        Object.defineProperty(this, "getKSFromWorkspaceAi", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: this.query({
                name: 'KsWorkspaceAi',
                request: (signal, variables) => listAllByContentTypeV1WorkspacesWorkspaceIdContentTypeGet({ ...variables, contentType: 'knowledge_source', authorization: '' }, { signal }),
                permission: this.createPermissionFunctionFor(listAllByContentTypeV1WorkspacesWorkspaceIdContentTypeGet),
            })
        });
        /**
         * Lists all the stacks in a Spot.
         */
        Object.defineProperty(this, "getStackFromWorkspaceAi", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: this.query({
                name: 'StackWorkspaceAi',
                request: (signal, variables) => listAllByContentTypeV1WorkspacesWorkspaceIdContentTypeGet({ ...variables, contentType: 'stack', authorization: '' }, { signal }),
                permission: this.createPermissionFunctionFor(listAllByContentTypeV1WorkspacesWorkspaceIdContentTypeGet),
            })
        });
        /**
         * Lists all the quick commands in a Spot.
         */
        Object.defineProperty(this, "getQCFromWorkspaceAi", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: this.query({
                name: 'QcWorkspaceAi',
                request: (signal, variables) => listAllByContentTypeV1WorkspacesWorkspaceIdContentTypeGet({ ...variables, contentType: 'quick_command', authorization: '' }, { signal }),
                permission: this.createPermissionFunctionFor(listAllByContentTypeV1WorkspacesWorkspaceIdContentTypeGet),
            })
        });
        /**
         * Lists all the agents in a Spot.
         */
        Object.defineProperty(this, "getAgentFromWorkspaceAi", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: this.query({
                name: 'AgentWorkspaceAi',
                request: (signal, variables) => listAllByContentTypeV1WorkspacesWorkspaceIdContentTypeGet({ ...variables, contentType: 'agent', authorization: '' }, { signal }),
                permission: this.createPermissionFunctionFor(listAllByContentTypeV1WorkspacesWorkspaceIdContentTypeGet),
            })
        });
        /**
         * Lists all the toolkits in a Spot.
         */
        Object.defineProperty(this, "getToolkitsFromWorkspaceAi", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: this.query({
                name: 'ToolkitWorkspaceAi',
                request: (signal, variables) => listAllByContentTypeV1WorkspacesWorkspaceIdContentTypeGet({ ...variables, contentType: 'toolkit', authorization: '' }, { signal }),
                permission: this.createPermissionFunctionFor(listAllByContentTypeV1WorkspacesWorkspaceIdContentTypeGet),
            })
        });
        Object.defineProperty(this, "deleteResourceTypeFromWorkspaceAi", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: this.mutation(removeAuthorizationParam(removeContentV1WorkspacesWorkspaceIdContentTypeContentIdDelete))
        });
        Object.defineProperty(this, "addPermissionToWorkspaceAi", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: this.mutation(removeAuthorizationParam(addPermissionV1WorkspacesWorkspaceIdPermissionsPatch))
        });
        Object.defineProperty(this, "removePermissionToWorkspaceAi", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: this.mutation(removeAuthorizationParam(removePermissionV1WorkspacesWorkspaceIdPermissionsDelete))
        });
        /**
         * Get group of permission on workspace AI
         * @param search can be used to search by name
         * @param filter_value is not necessary to use for filtering
         * @param filter_by is not necessary to use for filtering
         */
        // eslint-disable-next-line max-len
        Object.defineProperty(this, "groupsOfWorkspaceAi", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: this.query(removeAuthorizationParam(listGroupsAndMembersPermissionInWorkspaceV1WorkspacesWorkspaceIdPermissionsGet))
        });
        /**
         * Adds the resource of type Workspace AI to the list of favorites.
         */
        Object.defineProperty(this, "addFavoriteWorkspaceAi", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: this.mutation(removeAuthorizationParam(addFavoriteV1WorkspacesWorkspaceIdFavoritePost))
        });
        /**
         * Removes the resource of type Workspace AI from the list of favorites.
         */
        Object.defineProperty(this, "removeFavoriteWorkspaceAi", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: this.mutation(removeAuthorizationParam(deleteFavoriteV1WorkspacesWorkspaceIdFavoriteDelete))
        });
        /**
         * Gets all content from all workspaces to which the user has permissions.
         */
        Object.defineProperty(this, "workspacesContentsByType", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: this.query(removeAuthorizationParam(listAllContentsV1WorkspacesContentsContentTypeGet))
        });
        /**
         * Get a generated link to share a spot
         */
        Object.defineProperty(this, "generateShareSpotLink", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: this.mutation(removeAuthorizationParam(createShareableLinkV1WorkspacesWorkspaceIdSharePost))
        });
        /**
         * Get a shareable link from a spot
         */
        Object.defineProperty(this, "shareableLink", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: this.mutation(removeAuthorizationParam(useShareableLinkV1ShareShareIdPatch))
        });
        /**
         * Add permission to resource
         */
        Object.defineProperty(this, "addPermissionResource", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: this.mutation(removeAuthorizationParam(addPermissionV1ResourceTypeResourceTypeResourcesResourceIdentifierPatch))
        });
        /**
         * Remove permission to resource
         */
        Object.defineProperty(this, "removePermissionResource", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: this.mutation(removeAuthorizationParam(removePermissionV1ResourceTypeResourceTypeResourcesResourceIdentifierDelete))
        });
        /**
         * List members with permission to resource
         */
        Object.defineProperty(this, "listMembersResource", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: this.infiniteQuery(removeAuthorizationParam(listMembersV1ResourceTypeResourceTypeResourcesResourceIdentifierMembersGet))
        });
    }
    buildStackSpotError(error) {
        return new DefaultAPIError(error.data, error.status, wksAiDictionary, error.headers);
    }
}
export const workspaceAiClient = new WorkspaceAiClient();
