// @stack-spot/portal-components/src/components/InfoMaintenanceBanner.tsx
import { Text } from '@citric/core'
import { interpolate, useLanguage, useTranslate } from '@stack-spot/portal-translate'
import { format } from 'date-fns'
import { enUS, ptBR } from 'date-fns/locale'

interface InfoMaintenanceBannerProps {
  date: Date,
  initialHour: string,
  endHour: string,
}

export const InfoMaintenanceBanner = ({ date, initialHour, endHour }: InfoMaintenanceBannerProps) => {
  const t = useTranslate(dictionary)
  const language = useLanguage()
  const locale = language === 'pt' ? ptBR : enUS
  const dateToString = language === 'en' ?
    format(date, "MMMM do',' yyyy", { locale }) :
    format(date, "dd 'de' MMMM 'de' yyyy", { locale })
  return <Text>{interpolate(t.info, dateToString, initialHour, endHour)}</Text>
}

const dictionary = {
  en: {
    info: 'We would like to inform you that StackSpot will undergo maintenance on $0, from $1 (GMT-3) to $2. During this period, this web portal will be temporarily unavailable.',
  },
  pt: {
    info: 'Gostaríamos de informar que a StackSpot passará por manutenção no dia $0, das $1 (GMT-3) às $2. Durante este período, este portal web estará temporariamente indisponível.',
  },
}
