// @stack-spot/portal-components/src/components/Placeholder.tsx
import { Box, Flex, Text } from '@citric/core'
import { AsProp } from '@citric/core/dist/forward-ref-as'
import { SxProperties } from '@citric/core/dist/sx'
import { Card, CardContent } from '@citric/ui'
import { Dictionary, useTranslate } from '@stack-spot/portal-translate'
import { GenericPlaceholder } from '../svg'

interface Props {
  title?: string | React.ReactElement,
  description?: string | React.ReactElement,
  img?: React.ReactElement,
  button?: React.ReactElement | null,
  show?: boolean,
  fullWidth?: boolean,
  sx?: SxProperties,
  sxCard?: SxProperties,
  style?: React.CSSProperties,
  className?: string,
}

interface PropsCallToAction extends Props {
  subDescription?: string,
  titleAs?: AsProp,
  placeholderImageConfig?: { 
    width?: string, 
    height?: string,
    style?: React.CSSProperties,
  },
}

export const Placeholder = ({ title, description, button, img, fullWidth, show = true, sx, style, className }: Props) => {
  if (!show) {
    return null
  }

  return (
    <Card sx={{ ...(fullWidth ? { w: '100%' } : {}) }} style={style} className={className}>
      <Flex alignItems="center" justifyContent="center" py={15} sx={{ gap: '20px', ...sx }}>
        <Box sx={{ justifySelf: 'end' }}>{img ?? <GenericPlaceholder />}</Box>
        <Box>
          {title && <Text as="label" appearance="h4" mb="3" pt={5}>
            {title}
          </Text>}

          <Text mb="3" appearance="body2" colorScheme="light.700">
            {description}
          </Text>

          {button}
        </Box>
      </Flex>
    </Card>
  )
}

export const PlaceholderCallToAction = ({
  title,
  description,
  subDescription,
  img,
  button,
  fullWidth,
  show = true,
  sx,
  sxCard,
  style,
  className,
  titleAs,
  placeholderImageConfig,
}: PropsCallToAction) => {
  if (!show) {
    return null
  }

  return (
    <Card sx={{ w: fullWidth ? '100%' : 'auto', bg: 'transparent', ...sxCard }} style={style} className={className}>
      <CardContent
        as={Flex}
        py={15}
        sx={{ flexDirection: 'column', justifyContent: 'center', alignItems: 'center', gap: '92px', ...sx }}
      >
        <Box sx={{ justifySelf: 'end' }}>{img ?? <GenericPlaceholder {...placeholderImageConfig} />}</Box>

        <Box>
          {title && <Text as={titleAs} appearance="h4" mb="3" aria-live="assertive">
            {title}
          </Text>}

          <Text mb="3" appearance="body2" colorScheme="light.700">
            {description}
          </Text>

          <Text mb="3" appearance="body2" colorScheme="light.700">
            {subDescription}
          </Text>

          {button}
        </Box>
      </CardContent>
    </Card>
  )
}


export const PlaceholderResourceEmptySearch = ({ title, description }: Pick<Props, 'title' | 'description'>) => {
  const t = useTranslate(dictionary)

  return (
    <PlaceholderCallToAction fullWidth={true}
      title={title || t.title} description={description || t.description}
    />
  )
}

const dictionary = {
  en: {
    title: 'Sorry, your search did not match any results.',
    description: 'Please, refine your search and try again.',
  },
  pt: {
    title: 'Desculpe, sua busca não encontrou nenhum resultado.',
    description: 'Por favor, refine sua busca e tente novamente.',
  },
} satisfies Dictionary
