// @stack-spot/portal-components/src/components/Select/CloseItem.tsx
import { IconBox } from '@citric/core'
import { SxProperties } from '@citric/core/dist/sx'
import { TimesMini } from '@citric/icons'
import { IconButton } from '@citric/ui'
import { Dictionary, interpolate, useTranslate } from '@stack-spot/portal-translate'
import { MultiValueRemoveProps, components } from 'react-select'

export const CloseItem = (props: MultiValueRemoveProps<any>) => {
  const t = useTranslate(dictionary)
  return (
    <components.MultiValueRemove {...props}>
      <IconButton
        type="button"
        aria-label={interpolate(t.ariaLabelRemoveItemButton, props.data.value)}
        sx={{
          bg: 'transparent',
          border: 'none',
          px: 0,
          w: 'auto',
          height: 'auto',
          '&:hover, &:focus': { bg: 'transparent' },
        } as SxProperties}>
        <IconBox size="xs" sx={{ px: 0 }}>
          <TimesMini />
        </IconBox>
      </IconButton>
    </components.MultiValueRemove>
  )
}

const dictionary = {
  en: {
    ariaLabelRemoveItemButton: 'Remove item $0',
  },
  pt: {
    ariaLabelRemoveItemButton: 'Remover item $0',
  },
} satisfies Dictionary
