// @stack-spot/portal-components/src/components/Select/CreatableSelect.tsx
import { theme } from '@stack-spot/portal-theme'
import { Dictionary, useTranslate } from '@stack-spot/portal-translate'
import { FC } from 'react'
import { CSSObjectWithLabel } from 'react-select'
import Creatable from 'react-select/creatable'
import { BadgeItem } from './BadgeItem'
import { CreatableSelectOptionType, CreatableValueType } from './types'
import { ClearInput } from './ClearInput'

export interface CreatableSelectProps {
  id?: string,
  name?: string,
  disabled?: boolean,
  isLoading?: boolean,
  isMulti?: boolean,
  options?: CreatableSelectOptionType[],
  className?: string,
  value?: string | string[],
  inputValue?: CreatableValueType,
  onChange: (newValue: CreatableValueType) => void,
}

export const CreatableSelect: FC<CreatableSelectProps> = 
({ disabled, options, inputValue, isMulti, onChange, id, ...props }) => {
  const t = useTranslate(dictionary)

  return (
    <Creatable
      placeholder={t.typeYourOption}
      {...props}
      inputId={id}
      isClearable={true}
      isDisabled={disabled}
      value={inputValue}
      options={options}
      isMulti={isMulti}
      onChange={onChange}
      components={{ 
        MultiValue: BadgeItem, 
        MultiValueRemove: () => <></>, 
        ClearIndicator: ClearInput,
      }}
      noOptionsMessage={() => (t.typeYourOption)}
      formatCreateLabel={(value) => (value)}
      tabSelectsValue={false}
      styles={{
        control: (base, state) => ({
          ...base,
          minHeight: '2.5rem',
          height: 'auto',
          backgroundColor: theme.color.light[300],
          border: `0.063rem solid ${theme.color.light[600]}`,
          borderRadius: '0.25rem',
          fontSize: '0.875rem',
          fontStyle: 'normal',
          fontFeatureSettings: '\'clig\' off, \'liga\' off',
          display: 'flex',
          alignItems: 'center',
          outlineColor: theme.color.light[700],

          '&:hover': {
            border: `0.063rem solid ${theme.color.light[600]}`,
          },

          ...(state.isFocused || state.menuIsOpen ? {
            borderColor: `${theme.color.primary[500]} !important`,
            outline: `0.063rem solid ${theme.color.primary[500]} !important`,
          } : {}),
        } as CSSObjectWithLabel),
        input: () => ({
          color: theme.color.light.contrastText,
        }),
        placeholder: () => ({
          display: 'none',
        }),
        dropdownIndicator: () => ({
          display: 'none',
        }),
        indicatorSeparator: () => ({
          display: 'none',
        }),
        singleValue: (base) => ({
          ...base,
          color: theme.color.light.contrastText,
        } as CSSObjectWithLabel),
        menu: (base) => ({
          ...base,
          backgroundColor: theme.color.light[500],
          border: `0.063rem solid ${theme.color.light[400]}`,
          borderRadius: '0.5rem',
          boxShadow: `0 0 0 0.125rem ${theme.color.gray[600]}`,
          color: theme.color.light.contrastText,
        } as CSSObjectWithLabel),
        multiValue: (base) => ({
          ...base,
          backgroundColor: 'inherit',
          display: 'flex',
          alignItems: 'center',
          justifyItems: 'center',
        }),
        multiValueRemove: (base) => ({
          ...base,
          cursor: 'pointer',
          backgroundColor: 'inherit',
          paddingRight: 0,
          ':hover': {
            backgroundColor: 'inherit',
            color: 'inherit',
          },
        }),
        option: (styles, { isDisabled, isFocused, isSelected }) => ({
          ...styles,
          backgroundColor: isDisabled
            ? undefined
            : isSelected
              ? theme.color.light[700]
              : isFocused
                ? theme.color.light[600]
                : undefined,

          color: isSelected ? theme.color.light[300] : theme.color.light.contrastText,

          cursor: isDisabled ? 'not-allowed' : 'default',

          ':active': {
            ...styles[':active'],
            backgroundColor: !isDisabled
              ? isSelected
                ? theme.color.light.contrastText
                : theme.color.light[600]
              : undefined,
          },

          ':hover': {
            ...styles[':hover'],
            backgroundColor: !isDisabled
              ? isSelected
                ? theme.color.light[700]
                : theme.color.light[600]
              : undefined,
          },
        } as CSSObjectWithLabel),
      }}
    />
  )
}

const dictionary = {
  en: {
    typeYourOption: 'Type your option',
  },
  pt: {
    typeYourOption: 'Insira a sua opção',
  },
} satisfies Dictionary
