// @stack-spot/portal-components/src/components/Select/MultiValue.tsx
import { SxProperties } from '@citric/core/dist/sx'
import { TimesMini } from '@citric/icons'
import { Badge, IconButton } from '@citric/ui'
import { Dictionary, interpolate, useTranslate } from '@stack-spot/portal-translate'
import { MultiValueProps, components } from 'react-select'

export const MultiValue = (props: MultiValueProps) => {
  const t = useTranslate(dictionary)
  return (
    <>
      <components.MultiValue {...props}>
        <Badge appearance="square" className="badge" palette="cyan" sx={
        { 
          mr: 0,
          my: 0,
          '.badge-citric-container:not(:has(> button:first-child))': {
            marginRight: 0,
          },
        } as SxProperties}
        afterElement={
          <components.MultiValueRemove {...props}>
            <IconButton
              type="button"
              aria-label={interpolate(t.ariaLabelRemoveItemButton, props.children)}
              appearance="square"
              colorIcon="cyan.800"
              onClick={(e: React.MouseEvent<HTMLElement, MouseEvent>) => {
                e.stopPropagation()
                props.removeProps?.onClick?.(e as React.MouseEvent<HTMLDivElement, MouseEvent>)
              }}>
              <TimesMini />
            </IconButton>
          </components.MultiValueRemove>
        }>
          {props.children}
        </Badge>
      </components.MultiValue>
    </>
  )
}

const dictionary = {
  en: {
    ariaLabelRemoveItemButton: 'Remove item $0',
  },
  pt: {
    ariaLabelRemoveItemButton: 'Remover item $0',
  },
} satisfies Dictionary
