// @stack-spot/portal-components/src/components/Select/SelectInfiniteScroll.tsx
import { IconBox } from '@citric/core'
import { LoadingCircular } from '@citric/ui'
import InfiniteScroll from 'react-infinite-scroll-component'
import { components, GroupBase, MenuListProps } from 'react-select'
import { SelectSearch, SelectSearchProps } from './SelectSearch'

interface CustomMenuListProps {
  loadMore: () => void,
  hasMore: boolean,
}

const MenuList = <
  OptionType,
  IsMulti extends boolean = false,
  Group extends GroupBase<OptionType> = GroupBase<OptionType>
>(props: MenuListProps<OptionType, IsMulti, Group>) => {
  const { children, selectProps } = props
  const { loadMore, hasMore, options, isLoadingMore } = selectProps as typeof selectProps &
    CustomMenuListProps & { isLoadingMore?: boolean }

  return (
    <InfiniteScroll
      dataLength={options.length}
      next={loadMore}
      hasMore={hasMore}
      scrollableTarget="list-item"
      scrollThreshold={0.8}
      loader={isLoadingMore ? <IconBox w={12} colorBg="light.300" sx={{ alignItems: 'center', justifyContent: 'center' }}>
        <LoadingCircular />
      </IconBox> : null}
      style={{ overflow: 'hidden' }}
    >
      <components.MenuList {...props} innerProps={{ ...props.innerProps, id: 'list-item' }}>{children}</components.MenuList>
    </InfiniteScroll>
  )
}

type Option = { value: string, label: string }

export interface SelectInfiniteScrollProps extends SelectSearchProps {
  options: Option[],
  loadMore: () => void,
  hasMore: boolean,
  isLoadingMore?: boolean,
}

export const SelectInfiniteScroll = ({
  options,
  loadMore,
  hasMore,
  isLoadingMore,
  ...props
}: SelectInfiniteScrollProps) => (
  <SelectSearch
    components={{ MenuList }}
    options={options}
    closeMenuOnSelect={false}
    styles={{
      control: () => ({
        minHeight: '2.5rem',
        height: 'auto',
      }),
      menu: () => ({
        position: 'absolute',
      }),
      menuList: (base) => ({
        ...base,
        maxHeight: 300,
        overflowY: 'auto',
      }),
      input: () => ({
        position: 'relative',
      }),
    }}
    // @ts-ignore: Essas props não existem no tipo, mas são acessíveis via selectProps no MenuList
    loadMore={loadMore}
    hasMore={hasMore}
    isLoadingMore={isLoadingMore}
    {...props}
  />
)
