// @stack-spot/citric-react/src/components/Tabs/types.ts
import { WithColorPalette, WithColorScheme } from '../../types'
import { TabController } from './TabController'

export interface Tab<Key extends string = string> {
  /**
   * A unique identifier for the tab.
   */
  key: Key,
  /**
   * The tab's label. Will be the "key" if not provided.
   */
  label?: string,
  /**
   * The tab's content. This can be a suspended component.
   * 
   * Attention: consider using React's lazy loading to render the tab view.
   */
  content: React.ReactNode,
  /**
   * Whether or not this tab is disabled.
   * 
   * @default false
   */
  disabled?: boolean,
}

export interface BaseTabsProps<Key extends string> extends WithColorScheme, WithColorPalette {
  /**
   * The tabs to render.
   */
  tabs: Tab<Key>[],
  /**
   * The key of the current tab. Use this only if you intend to control the tabs from the outside.
   */
  value?: Key,
  /**
   * What to do when the tab changes. Use this only if you intend to control the tabs from the outside.
   * 
   * @param key the key of the selected tab.
   */
  onChange?: (key: Key) => void,
  /**
   * A tab-controller, useful if you want to control the tabs outside the component context.
   */
  controller?: TabController<Key>,
  /**
   * Instead of each tab occupying only the space it needs, all tabs will have their sizes equally distributed.
   * 
   * @default false
   */
  equallySized?: boolean,
  /**
   * The appearance of Tabs.
   * 
   * @default 'tabs'
   */
  appearance?: 'tabs' | 'toggles',
  /**
   * Allows the use of the tabs component with other elements in the middle of its UI.
   * @param tabSelector the UI for changing the current tab.
   * @param content the content of the selected tab. 
   * @returns the customized UI.
   */
  customRenderer?: (tabSelector: React.ReactElement, content: React.ReactElement) => React.ReactNode,
}

export type TabsProps<Key extends string> = Omit<React.JSX.IntrinsicElements['div'], 'onChange' | 'children'> & BaseTabsProps<Key>
