// @stack-spot/portal-network/src/client/agent-tools.ts
import { HttpError } from '@oazapfts/runtime'
import { getApiAddresses } from '../api-addresses'
import { addFavoriteV1AgentsAgentIdFavoritePost, AgentVisibilityLevelEnum, createAgentV1AgentsPost, createToolkitToolsV1ToolkitsToolkitIdToolsPost, createToolkitV1ToolkitsPost, defaults, deleteAgentV1AgentsAgentIdDelete, deleteFavoriteV1AgentsAgentIdFavoriteDelete, deleteToolkitToolsV1ToolkitsToolkitIdToolsDelete, deleteToolkitV1ToolkitsToolkitIdDelete, editToolkitToolV1ToolkitsToolkitIdToolsToolIdPut, forkAgentV1AgentsAgentIdForkPost, forkToolkitV1ToolkitsToolkitIdForkPost, getAgentV1AgentsAgentIdGet, getPublicToolKitsV1BuiltinToolkitGet, getToolkitToolV1ToolkitsToolkitIdToolsToolIdGet, getToolkitV1ToolkitsToolkitIdGet, listAgentsUsingToolsV1ToolkitsToolkitIdToolsAgentsPost, listAgentsV1AgentsGet, listAgentsV3AgentsGet, listMcpToolsV1McpToolsToolkitIdGet, listMultiAgentsV1AgentsMultiAgentsGet, listToolkitsV1ToolkitsGet, listToolkitsV2ToolkitsGet, publishAgentV1AgentsAgentIdPublishPost, searchAgentsV1AgentsSearchPost, shareV1AgentsAgentIdSharePost, updateAgentV1AgentsAgentIdPatch, updateToolkitV1ToolkitsToolkitIdPatch, VisibilityLevelEnum } from '../api/agent-tools'
import { DefaultAPIError } from '../error/DefaultAPIError'
import { agentToolsDictionary } from '../error/dictionary/agent-tools'
import { StackspotAPIError } from '../error/StackspotAPIError'
import { ReactQueryNetworkClient } from '../network/ReactQueryNetworkClient'
import { FetchEventStream } from '../network/types'
import { removeAuthorizationParam } from '../utils/remove-authorization-param'
import { StreamedArray } from '../utils/StreamedArray'
import { AgentResponseWithBuiltIn, AgentToolsOpenAPIPreview, AgentVisibilityLevel } from './types'
import { workspaceAiClient } from './workspace-ai'

const AGENT_DEFAULT_SLUG = 'stk_flex'
const listAgentsV3AgentsWithoutAuthorization = removeAuthorizationParam(listAgentsV3AgentsGet)

class AgentToolsClient extends ReactQueryNetworkClient {
  constructor() {
    super(getApiAddresses()['agent-tools'].url, defaults)
  }

  protected buildStackSpotError(error: HttpError): StackspotAPIError {
    return new DefaultAPIError(error.data, error.status, agentToolsDictionary, error.headers)
  }

  /**
   * Lists all public toolkits.
   */
  tools = this.query(removeAuthorizationParam(getPublicToolKitsV1BuiltinToolkitGet))

  /**
   * Create agent
   */
  createAgent = this.mutation(removeAuthorizationParam(createAgentV1AgentsPost))

  /**
   * Delete agent
   */
  deleteAgent = this.mutation(removeAuthorizationParam(deleteAgentV1AgentsAgentIdDelete))

  /**
   * Updates an agent
   */
  updateAgent = this.mutation(removeAuthorizationParam(updateAgentV1AgentsAgentIdPatch))

  /**
  * Favorite an agent
  */
  favoriteAgent = this.mutation(removeAuthorizationParam(addFavoriteV1AgentsAgentIdFavoritePost))

  /**
  * Publish an agent
  */
  publishAgent = this.mutation(removeAuthorizationParam(publishAgentV1AgentsAgentIdPublishPost))

  /** 
   * List agents
   */
  agents = this.infiniteQuery(removeAuthorizationParam(listAgentsV1AgentsGet))

  /** 
   * List agents with support for multiple filters
   */
  agentsMultipleFilters = this.infiniteQuery(listAgentsV3AgentsWithoutAuthorization, {
    pageParamName: 'filters.page',
    accumulator: 'items', 
    getNextPageParam: ({ variables, lastPage, lastPageParam }) => {
      const size = variables.filters.size ?? 1
      const parsedLastPageParam = (lastPageParam as number) ?? variables.filters.page ?? 1
      return lastPage.items && lastPage.items.length < size ? undefined : parsedLastPageParam + 1
    },
  })

  /** 
   * List agents available to be added to other agents
   */
  availableAgentsForMultiAgentAddition = this.query(removeAuthorizationParam(listMultiAgentsV1AgentsMultiAgentsGet))

  /** 
   * Gets agent by id
   */
  agent = this.query(removeAuthorizationParam(getAgentV1AgentsAgentIdGet))

  /** 
   * Gets agents by ids
   */
  agentsByIds = this.query(removeAuthorizationParam(searchAgentsV1AgentsSearchPost))

  /** 
   * Gets the default agent slug
   */
  agentDefaultSlug = AGENT_DEFAULT_SLUG

  /** 
   * Gets the default agent
   */
  agentDefault = this.query({
    name: 'agentDefault',
    request: async (signal) => {
      const agentDefault = await listAgentsV1AgentsGet(
        { visibility: 'built_in', slug: this.agentDefaultSlug, authorization: '' }, { signal },
      )

      const agentId = agentDefault?.find((agent) => agent.slug ===  this.agentDefaultSlug)?.id
      const agent = agentId ? await this.agent.query({ agentId }) : undefined
      return agent
    },
  })

  /** 
   * List agents filtered by visibility.
   */
  allAgents = this.query({
    name: 'allAgents',
    request: async (signal, variables: { visibilities: (AgentVisibilityLevel | 'all')[] }) => {
      const allVisibilities = ['account', 'built_in', 'recently_used', 'favorite', 'personal', 'shared', 'workspace'] as const
      const visibilities = variables.visibilities.includes('all')
        ? allVisibilities
        : variables.visibilities as Array<AgentVisibilityLevelEnum | VisibilityLevelEnum>

      const shouldFetchWorkspaceAgents = visibilities.includes('workspace')

      const workspaceAgentsPromise = shouldFetchWorkspaceAgents
        ? workspaceAiClient.workspacesContentsByType.query({ contentType: 'agent' })
        : Promise.resolve([])

      const [workspaceAgents, ...agentsByVisibility] = await Promise.all([
        workspaceAgentsPromise,
        ...visibilities.map((visibility) => listAgentsV1AgentsGet({ visibility, authorization: '' }, { signal })),
      ])

      const workspaceAgentsWithSpaceName = workspaceAgents.flatMap(({ agents, space_name }) =>
        agents?.map((agent) => ({ ...agent, spaceName: space_name, builtIn: false }))) as AgentResponseWithBuiltIn[]

      const allAgents: AgentResponseWithBuiltIn[] = workspaceAgentsWithSpaceName ?? []

      agentsByVisibility.forEach(agents => allAgents.push(...agents.map(agent => ({
        ...agent,
        builtIn: agent?.visibility_level === 'built_in',
      }))))

      return allAgents
    },
  })

  /**
   * Lists all custom toolkits.
   */
  toolkits = this.query(removeAuthorizationParam(listToolkitsV1ToolkitsGet))
  /**
  * Get list of Toolkits V2
  */
  toolkitsV2 = this.query(removeAuthorizationParam(listToolkitsV2ToolkitsGet))
  /**
   * Lists all tools of an MCP toolkit.
   */
  mcpTools = this.query(removeAuthorizationParam(listMcpToolsV1McpToolsToolkitIdGet))
  /**
   * Get a toolkit by Id
   */
  toolkitById = this.query(removeAuthorizationParam(getToolkitV1ToolkitsToolkitIdGet))
  /**
   * Create a toolkit
   */
  createToolkit = this.mutation(removeAuthorizationParam(createToolkitV1ToolkitsPost))
  /**
   * Update a toolkit
   */
  updateToolkit = this.mutation(removeAuthorizationParam(updateToolkitV1ToolkitsToolkitIdPatch))
  /**
   * Delete a toolkit
   */
  deleteToolkit = this.mutation(removeAuthorizationParam(deleteToolkitV1ToolkitsToolkitIdDelete))
  /**
   * Get Tool from Toolkit
   */
  toolFromToolkit = this.query(removeAuthorizationParam(getToolkitToolV1ToolkitsToolkitIdToolsToolIdGet))
  /**
   * Add tool on Toolkit
   */
  addToolOnToolkit = this.mutation(removeAuthorizationParam(createToolkitToolsV1ToolkitsToolkitIdToolsPost))
  /**
   * Edit tool on Toolkit
   */
  editToolOnToolkit = this.mutation(removeAuthorizationParam(editToolkitToolV1ToolkitsToolkitIdToolsToolIdPut))
  /**
   * Delete tool from Toolkit
   */
  deleteToolFromToolkit = this.mutation(removeAuthorizationParam(deleteToolkitToolsV1ToolkitsToolkitIdToolsDelete))
  
  /**
   * Previews an uploaded open api file as a stream where each array item is an OpenAPI operation.
   */
  streamUploadedAPIs = ({ fileUploadIds, signal }: { fileUploadIds: string[], signal?: AbortSignal }) => {
    const promises: Promise<FetchEventStream>[] = fileUploadIds.map((id) => {
      const url = `/v1/toolkits/tools/preview/${encodeURIComponent(id)}`
      return this.stream(url, { signal })
    })
    return new StreamedArray<AgentToolsOpenAPIPreview>({ eventsPromises: promises, signal })
  }
  /**
   * Lists the agents using the tools passed as parameter.
   */
  agentsUsingTools = this.query(removeAuthorizationParam(listAgentsUsingToolsV1ToolkitsToolkitIdToolsAgentsPost))
  /**
   * Forks the toolkit.
   */
  fork = this.mutation(removeAuthorizationParam(forkToolkitV1ToolkitsToolkitIdForkPost))

  /**
   * Share Agent
   */
  shareAgent = this.mutation(removeAuthorizationParam(shareV1AgentsAgentIdSharePost))
  /**
   * Fork Agent
   */
  forkAgent = this.mutation(removeAuthorizationParam(forkAgentV1AgentsAgentIdForkPost))

  /**
  * Add and agent to favorites
  */
  addFavorite = this.mutation(removeAuthorizationParam(addFavoriteV1AgentsAgentIdFavoritePost))

  /**
  * remove an agent from favorites 
  */
  removeFavorite = this.mutation(removeAuthorizationParam(deleteFavoriteV1AgentsAgentIdFavoriteDelete))
}

export const agentToolsClient = new AgentToolsClient()
