// @stack-spot/portal-network/src/client/data-integration.ts
import { HttpError } from '@oazapfts/runtime'
import { getApiAddresses } from '../api-addresses'
import {
  addFavoriteV1KnowledgeSourcesSlugFavoritePost,
  createKsV1KnowledgeSourcesPost,
  defaults,
  deleteFavoriteV1KnowledgeSourcesSlugFavoriteDelete,
  deleteKoByIdV1KnowledgeSourcesKsSlugObjectsKoIdDelete,
  FileUploadType,
  findKoByIdV1KnowledgeSourcesKsSlugObjectsKoIdGet,
  findKsV1KnowledgeSourcesSlugGet,
  forkKsV1KnowledgeSourcesSlugForkPost,
  getFileUploadStatusV1FileUploadFileUploadIdGet,
  getIndexedFilesV1KnowledgeSourcesKsSlugFilesGet,
  getUploadFormV1FileUploadFormPost,
  getUploadFormV2V2FileUploadFormPost,
  listKosV1KnowledgeSourcesKsSlugObjectsGet,
  listKosV2KnowledgeSourcesKsSlugObjectsGet,
  listKsV1KnowledgeSourcesGet,
  listKsV2KnowledgeSourcesGet,
  publishKsV1KnowledgeSourcesSlugPublishPost,
  resetKosV1KnowledgeSourcesKsSlugObjectsDelete,
  saveChunkedFilesDeprecatedV1FileUploadFileUploadIdPost,
  saveChunkedFilesV1FileUploadFileUploadIdKnowledgeObjectsPost,
  searchKsV1KnowledgeSourcesSearchPost,
  searchObjectsV1SearchPost,
  shareKsV1KnowledgeSourcesSlugSharePost,
  splitUploadedFilePreviewV1FileUploadFileUploadIdSplitPreviewSplitStrategyGet,
  splitUploadedFileV1FileUploadFileUploadIdSplitPost,
  updateKsV1KnowledgeSourcesSlugPatch,
} from '../api/dataIntegration'
import { DefaultAPIError } from '../error/DefaultAPIError'
import { baseDictionary } from '../error/dictionary/base'
import { FileUploadError } from '../error/FileUploadError'
import { StackspotAPIError } from '../error/StackspotAPIError'
import { ReactQueryNetworkClient } from '../network/ReactQueryNetworkClient'
import { removeAuthorizationParam } from '../utils/remove-authorization-param'

const listKSV2WithoutAuthorization = removeAuthorizationParam(listKsV2KnowledgeSourcesGet)
const listKOV2WithoutAuthorization = removeAuthorizationParam(listKosV2KnowledgeSourcesKsSlugObjectsGet)


class DataIntegrationClient extends ReactQueryNetworkClient {
  constructor() {
    super(getApiAddresses().dataIntegration.url, defaults)
  }

  protected buildStackSpotError(error: HttpError): StackspotAPIError {
    return new DefaultAPIError(error.data, error.status, baseDictionary, error.headers)
  }
  /**
   * Generates a url to upload the file
   */
  generateUrlToFileUpload = this.mutation(removeAuthorizationParam(getUploadFormV2V2FileUploadFormPost))
  /**
   * Get file processing status by id
   */
  getFileProcessingStatusById = this.query(removeAuthorizationParam(getFileUploadStatusV1FileUploadFileUploadIdGet))
  /**
   * @deprecated Use saveChunkedKnowledgeObjects instead.
   * Save the splitted knowledge objects
   */
  saveKnowledgeObjects = this.mutation(removeAuthorizationParam(saveChunkedFilesDeprecatedV1FileUploadFileUploadIdPost))
  /**
   * Save the splitted knowledge objects
   */
  saveChunkedKnowledgeObjects = this.mutation(removeAuthorizationParam(saveChunkedFilesV1FileUploadFileUploadIdKnowledgeObjectsPost))
  /**
   * Splits the file with the passed settings
   */
  splitFile = this.mutation(removeAuthorizationParam(splitUploadedFileV1FileUploadFileUploadIdSplitPost))
  /**
   * Generate a preview with splitted objects
   */
  streamPreviewFromSplit(
    fileUploadId: string,
    splitStrategy: string,
    options?: { signal?: AbortSignal },
  ) {
    const url = `/v1/file-upload/${fileUploadId}/split/preview/${splitStrategy}`

    return this.stream(this.resolveURL(url), {
      method: 'GET',
      headers: {
        'Content-Type': 'application/json',
        'Accept': 'text/event-stream',
      },
      signal: options?.signal,
    })
  }
  /**
   * Generate a preview with splitted objects
   */
  previewFromSplit = this.query(removeAuthorizationParam(splitUploadedFilePreviewV1FileUploadFileUploadIdSplitPreviewSplitStrategyGet))
  /**
   * List of knowledge sources
   */
  knowledgeSources = this.query(removeAuthorizationParam(listKsV1KnowledgeSourcesGet))
  /**
   * List of knowledge sources with pagination
   */
  knowledgeSourcesV2 = this.infiniteQuery(listKSV2WithoutAuthorization, { accumulator: 'items' })
  /**
   * Adds the resource of type Knowledge Source to the list of favorites.
   */
  addFavoriteKnowledgeSource = this.mutation(removeAuthorizationParam(addFavoriteV1KnowledgeSourcesSlugFavoritePost))
  /**
   * Removes the resource of type Knowledge Source from the list of favorites.
   */
  removeFavoriteKnowledgeSource = this.mutation(removeAuthorizationParam(deleteFavoriteV1KnowledgeSourcesSlugFavoriteDelete))
  /**
   * List knowledge objects from a knowledge source
   */
  listKnowledgeObjects = this.query(removeAuthorizationParam(listKosV1KnowledgeSourcesKsSlugObjectsGet))
  /**
   * List knowledge objects from a knowledge source with pagination
   */
  listKnowledgeObjectsV2 = this.infiniteQuery(listKOV2WithoutAuthorization, { accumulator: 'items' })
  /**
   * Get a knowledge object by id
   */
  getKnowledgeObjectById = this.query(removeAuthorizationParam(findKoByIdV1KnowledgeSourcesKsSlugObjectsKoIdGet))


  uploadFiles = this.mutation({
    name: 'uploadFiles',
    request: (abortSignal, params: { files: File[], type: FileUploadType, targetId?: string }) => (
      Promise.all(params.files.map(async (f) => {
        try {
          const { id, url, form } = await getUploadFormV2V2FileUploadFormPost({
            authorization: '',
            newFileUploadRequest: { file_name: f.name, target_type: params.type, target_id: params.targetId ?? '' },
          }, { signal: abortSignal })
          const formData = new FormData()
          Object.entries(form).forEach(([key, value]) => {
            formData.append(key, value)
          })
          formData.append('file', f)
          const response = await fetch(url, { method: 'post', body: formData, signal: abortSignal })
          if (!response.ok) throw new StackspotAPIError({ status: response.status, headers: response.headers, message: 'Failed to upload' })
          return id
        } catch (e: any) {
          throw new FileUploadError(f.name, e)
        }
      }))
    ),
  })

  /**
   * Uploads a KO file
  */
  fileUploadFormKnowledgeObject = this.mutation(
    removeAuthorizationParam(getUploadFormV1FileUploadFormPost))

  /**
   * Create Ks
    */
  createKnowledgeSource = this.mutation(removeAuthorizationParam(createKsV1KnowledgeSourcesPost))

  /**
   * Search Ks
    */
  listKnowledgeSourcesDataByIds = this.query(removeAuthorizationParam(searchKsV1KnowledgeSourcesSearchPost))

  /**
   * Gets a knowledge source by its slug.
    */
  findKnowledgeSourceBySlug = this.query(removeAuthorizationParam(findKsV1KnowledgeSourcesSlugGet))

  /**
   * Search Objects
   */
  searchObjects = this.mutation(removeAuthorizationParam(searchObjectsV1SearchPost))
  /**
   * Update Knowledge Source
   */
  updateKnowledgeSource = this.mutation(removeAuthorizationParam(updateKsV1KnowledgeSourcesSlugPatch))
  /**
   * Share Knowledge Source
   */
  shareKnowledgeSource = this.mutation(removeAuthorizationParam(shareKsV1KnowledgeSourcesSlugSharePost))
  /**
   * Publish Knowledge Source
   */
  publishKnowledgeSource = this.mutation(removeAuthorizationParam(publishKsV1KnowledgeSourcesSlugPublishPost))
  /**
   * Fork Knowledge Source
   */
  forkKnowledgeSource = this.mutation(removeAuthorizationParam(forkKsV1KnowledgeSourcesSlugForkPost))
  /**
   * Delete Knowledge Objects by Standalone
   */
  deleteKnowledgeObjectsByStandalone = this.mutation(removeAuthorizationParam(resetKosV1KnowledgeSourcesKsSlugObjectsDelete))
  /**
   * Delete Knowledge Object from Knowledge Source
   */
  deleteKnowledgeObjectFromKnowledgeSource = this.mutation(removeAuthorizationParam(deleteKoByIdV1KnowledgeSourcesKsSlugObjectsKoIdDelete))
  /**
   * Get Indexed Files
   */
  listIndexedProjectFiles = this.query(removeAuthorizationParam(getIndexedFilesV1KnowledgeSourcesKsSlugFilesGet))
  
}

export const dataIntegrationClient = new DataIntegrationClient()
