// @stack-spot/portal-network/src/client/workspace-ai.ts
import { HttpError } from '@oazapfts/runtime'
import { AgentResponse } from '../api/agent'
import { CustomToolkitResponse } from '../api/agent-tools'
import { GetAiStackResponse, KnowledgeSourceItemResponse, QuickCommandResponse } from '../api/ai'
import {
  addContentV1WorkspacesWorkspaceIdContentTypePost,
  addFavoriteV1WorkspacesWorkspaceIdFavoritePost,
  addPermissionV1ResourceTypeResourceTypeResourcesResourceIdentifierPatch,
  addPermissionV1WorkspacesWorkspaceIdPermissionsPatch,
  createShareableLinkV1WorkspacesWorkspaceIdSharePost,
  createWorkspaceV1WorkspacesPost,
  defaults,
  deleteFavoriteV1WorkspacesWorkspaceIdFavoriteDelete,
  deleteWorkspaceV1WorkspacesWorkspaceIdDelete,
  getWorkspaceByIdV1WorkspacesIdGet,
  listAllByContentTypeV1WorkspacesWorkspaceIdContentTypeGet,
  listAllContentsV1WorkspacesContentsContentTypeGet,
  listGroupsAndMembersPermissionInWorkspaceV1WorkspacesWorkspaceIdPermissionsGet,
  listMembersV1ResourceTypeResourceTypeResourcesResourceIdentifierMembersGet,
  listWorkspacesV1WorkspacesGet,
  removeContentV1WorkspacesWorkspaceIdContentTypeContentIdDelete,
  removePermissionV1ResourceTypeResourceTypeResourcesResourceIdentifierDelete,
  removePermissionV1WorkspacesWorkspaceIdPermissionsDelete,
  updateWorkspaceV1WorkspacesWorkspaceIdPatch,
  useShareableLinkV1ShareShareIdPatch,
} from '../api/workspace-ai'
import { StackspotAPIError } from '../error/StackspotAPIError'
import { ReactQueryNetworkClient } from '../network/ReactQueryNetworkClient'
import { removeAuthorizationParam } from '../utils/remove-authorization-param'
import { getApiAddresses } from '../api-addresses'
import { DefaultAPIError } from '../error/DefaultAPIError'
import { wksAiDictionary } from '../error/dictionary/workspace-ai'
import { FixedAddResourceToWorkspaceAi, FixedWorkspaceAiPermissions, ReplaceResult } from './types'

class WorkspaceAiClient extends ReactQueryNetworkClient {
  constructor() {
    super(getApiAddresses()['workspace-ai'].url, defaults)
  }

  protected buildStackSpotError(error: HttpError): StackspotAPIError {
    return new DefaultAPIError(error.data, error.status, wksAiDictionary, error.headers)
  }

  createWorkspaceAi = this.mutation(removeAuthorizationParam(createWorkspaceV1WorkspacesPost))
  workspacesAi = this.query(removeAuthorizationParam(listWorkspacesV1WorkspacesGet))
  workspaceAi = this.query(removeAuthorizationParam(getWorkspaceByIdV1WorkspacesIdGet))
  updateWorkspaceAi = this.mutation(removeAuthorizationParam(updateWorkspaceV1WorkspacesWorkspaceIdPatch))
  deleteWorkspaceAi = this.mutation(removeAuthorizationParam(deleteWorkspaceV1WorkspacesWorkspaceIdDelete))

  /**
   * Adds a resource to the Spot.
   */
  addResourceTypeToWorkspaceAi = this.mutation({
    name: 'addResourceTypeToWorkspaceAi',
    request: (
      signal,
      variables: Omit<Parameters<typeof addContentV1WorkspacesWorkspaceIdContentTypePost>[0], 'authorization'>,
    ) =>
      addContentV1WorkspacesWorkspaceIdContentTypePost({ ...variables, authorization: '' },
        { signal }) as unknown as Promise<FixedAddResourceToWorkspaceAi>,
    permission: this.createPermissionFunctionFor(addContentV1WorkspacesWorkspaceIdContentTypePost),
  })

  /**
   * Lists all the knowledge sources in a Spot.
   */
  getKSFromWorkspaceAi = this.query({
    name: 'KsWorkspaceAi',
    request: (
      signal,
      variables: Omit<Parameters<typeof listAllByContentTypeV1WorkspacesWorkspaceIdContentTypeGet>[0],
        'authorization' | 'contentType'>,
    ) =>
      listAllByContentTypeV1WorkspacesWorkspaceIdContentTypeGet({ ...variables, contentType: 'knowledge_source', authorization: '' },
        { signal }) as unknown as Promise<KnowledgeSourceItemResponse[]>,
    permission: this.createPermissionFunctionFor(listAllByContentTypeV1WorkspacesWorkspaceIdContentTypeGet),
  })

  /**
   * Lists all the stacks in a Spot.
   */
  getStackFromWorkspaceAi = this.query({
    name: 'StackWorkspaceAi',
    request: (
      signal,
      variables: Omit<Parameters<typeof listAllByContentTypeV1WorkspacesWorkspaceIdContentTypeGet>[0],
        'authorization' | 'contentType'>,
    ) =>
      listAllByContentTypeV1WorkspacesWorkspaceIdContentTypeGet({ ...variables, contentType: 'stack', authorization: '' },
        { signal }) as unknown as Promise<GetAiStackResponse[]>,
    permission: this.createPermissionFunctionFor(listAllByContentTypeV1WorkspacesWorkspaceIdContentTypeGet),
  })

  /**
   * Lists all the quick commands in a Spot.
   */
  getQCFromWorkspaceAi = this.query({
    name: 'QcWorkspaceAi',
    request: (
      signal,
      variables: Omit<Parameters<typeof listAllByContentTypeV1WorkspacesWorkspaceIdContentTypeGet>[0],
        'authorization' | 'contentType'>,
    ) =>
      listAllByContentTypeV1WorkspacesWorkspaceIdContentTypeGet({ ...variables, contentType: 'quick_command', authorization: '' },
        { signal }) as unknown as Promise<QuickCommandResponse[]>,
    permission: this.createPermissionFunctionFor(listAllByContentTypeV1WorkspacesWorkspaceIdContentTypeGet),
  })

  /**
   * Lists all the agents in a Spot.
   */
  getAgentFromWorkspaceAi = this.query({
    name: 'AgentWorkspaceAi',
    request: (
      signal,
      variables: Omit<Parameters<typeof listAllByContentTypeV1WorkspacesWorkspaceIdContentTypeGet>[0],
        'authorization' | 'contentType'>,
    ) =>
      listAllByContentTypeV1WorkspacesWorkspaceIdContentTypeGet({ ...variables, contentType: 'agent', authorization: '' },
        { signal }) as unknown as Promise<AgentResponse[]>,
    permission: this.createPermissionFunctionFor(listAllByContentTypeV1WorkspacesWorkspaceIdContentTypeGet),
  })

  /**
   * Lists all the toolkits in a Spot.
   */
  getToolkitsFromWorkspaceAi = this.query({
    name: 'ToolkitWorkspaceAi',
    request: (
      signal,
      variables: Omit<Parameters<typeof listAllByContentTypeV1WorkspacesWorkspaceIdContentTypeGet>[0],
        'authorization' | 'contentType'>,
    ) =>
      listAllByContentTypeV1WorkspacesWorkspaceIdContentTypeGet({ ...variables, contentType: 'toolkit', authorization: '' },
        { signal }) as unknown as Promise<CustomToolkitResponse[]>,
    permission: this.createPermissionFunctionFor(listAllByContentTypeV1WorkspacesWorkspaceIdContentTypeGet),
  })

  deleteResourceTypeFromWorkspaceAi = this.mutation(removeAuthorizationParam(
    removeContentV1WorkspacesWorkspaceIdContentTypeContentIdDelete))
  addPermissionToWorkspaceAi = this.mutation(removeAuthorizationParam(addPermissionV1WorkspacesWorkspaceIdPermissionsPatch))
  removePermissionToWorkspaceAi = this.mutation(removeAuthorizationParam(removePermissionV1WorkspacesWorkspaceIdPermissionsDelete))

  /**
   * Get group of permission on workspace AI
   * @param search can be used to search by name
   * @param filter_value is not necessary to use for filtering
   * @param filter_by is not necessary to use for filtering
   */
  // eslint-disable-next-line max-len
  groupsOfWorkspaceAi = this.query(removeAuthorizationParam(listGroupsAndMembersPermissionInWorkspaceV1WorkspacesWorkspaceIdPermissionsGet) as unknown as ReplaceResult<typeof listGroupsAndMembersPermissionInWorkspaceV1WorkspacesWorkspaceIdPermissionsGet, FixedWorkspaceAiPermissions>)

  /**
   * Adds the resource of type Workspace AI to the list of favorites.
   */
  addFavoriteWorkspaceAi = this.mutation(removeAuthorizationParam(addFavoriteV1WorkspacesWorkspaceIdFavoritePost))

  /**
   * Removes the resource of type Workspace AI from the list of favorites.
   */
  removeFavoriteWorkspaceAi = this.mutation(removeAuthorizationParam(deleteFavoriteV1WorkspacesWorkspaceIdFavoriteDelete))

  /**
   * Gets all content from all workspaces to which the user has permissions.
   */
  workspacesContentsByType = this.query(removeAuthorizationParam(listAllContentsV1WorkspacesContentsContentTypeGet))
  /**
   * Get a generated link to share a spot
   */
  generateShareSpotLink = this.mutation(removeAuthorizationParam(createShareableLinkV1WorkspacesWorkspaceIdSharePost))
  /**
   * Get a shareable link from a spot
   */
  shareableLink = this.mutation(removeAuthorizationParam(useShareableLinkV1ShareShareIdPatch))
  /**
   * Add permission to resource
   */
  addPermissionResource = this.mutation(removeAuthorizationParam(addPermissionV1ResourceTypeResourceTypeResourcesResourceIdentifierPatch))
  /**
   * Remove permission to resource
   */
  removePermissionResource =
    this.mutation(removeAuthorizationParam(removePermissionV1ResourceTypeResourceTypeResourcesResourceIdentifierDelete))
  /**
   * List members with permission to resource
   */
  listMembersResource =
    this.infiniteQuery(removeAuthorizationParam(listMembersV1ResourceTypeResourceTypeResourcesResourceIdentifierMembersGet))
}

export const workspaceAiClient = new WorkspaceAiClient()
