// @stack-spot/portal-layout/src/Layout.tsx
import { FadingOverflow } from '@stack-spot/portal-components/FadingOverflow'
import { CSSToCitricAdapter, WithStyle } from '@stack-spot/portal-theme'
import '@stack-spot/portal-theme/dist/theme.css'
import { ReactElement, ReactNode, useEffect } from 'react'
import { overlay } from './LayoutOverlayManager'
import { WelcomeTour } from './WelcomeTour'
import { Header, HeaderProps } from './components/Header'
import { Toaster } from './components/Toaster'
import { ErrorBoundary } from './components/error/ErrorBoundary'
import { DescriptionFn, ErrorHandler, ErrorManager } from './components/error/ErrorManager'
import { SilentErrorBoundary } from './components/error/SilentErrorBoundary'
import { MenuContent } from './components/menu/MenuContent'
import { MenuSections } from './components/menu/MenuSections'
import { MenuProps, MenuPropsWithDynamicContent } from './components/menu/types'
import { elementIds } from './elements'
import './layout.css'

interface Props extends WithStyle {
  /**
   * The config for the menu.
   */
  menu: MenuProps,
  /**
   * The config for the header.
   */
  header: HeaderProps,
  /**
   * The React node to go in the page content.
   */
  children: ReactNode,
  /**
   * A React node to place right after the page.
   */
  extra?: ReactNode,
  /**
   * A function to convert errors into a readable format so they're properly rendered and logged.
   */
  errorDescriptor?: DescriptionFn,
  /**
   * A function to run whenever an error is catch by an error boundary.
   */
  onError?: ErrorHandler,
  /**
   * When true, the sections menu auto-collapses whenever a secondary (contextual) menu appears.
   * @default false
   */
  autoCollapseSectionsMenu?: boolean,
}

interface RawProps extends WithStyle {
  /**
   * The React element to go in the menu sections.
   */
  menuSections?: ReactElement,
  /**
   * The React element to go in the menu content.
   */
  menuContent?: ReactElement,
  /**
   * The React element to go in the menu inner content (third level nav).
   */
  menuInnerContent?: ReactElement,
  /**
   * The React element to go in the header.
   */
  header?: ReactElement,
  /**
   * The React node to go in the page content.
   */
  children?: ReactNode,
  /**
   * A React node to place right after the page.
   */
  extra?: ReactNode,
  /**
   * A function to convert errors into a readable format so they're properly rendered and logged.
   */
  errorDescriptor?: DescriptionFn,
  /**
   * A function to run whenever an error is catch by an error boundary.
   */
  onError?: ErrorHandler,
}

/**
 * Renders the layout with the React elements passed in the props.
 * @param props the component's props {@link RawProps}.
 */
export const RawLayout = (
  { menuSections, menuContent, menuInnerContent, header, children,
    extra, errorDescriptor, onError, className, style }:
    RawProps,
) => {
  const { bottomDialog, modal, rightPanel } = overlay.useOverlays()

  useEffect(() => {
    if (errorDescriptor) ErrorManager.setDescriptionFunction(errorDescriptor)
    if (onError) ErrorManager.setErrorHandler(onError)
  }, [])

  // controls classes in the layout element. This can't be a React prop because we don't want classes to be replaced, we want them to be
  // merged with the previous set of classes.
  useEffect(() => {
    const classList = document.getElementById(elementIds.layout)?.classList
    if (!classList) return
    const hasMenuContent = !!menuContent && !!document.getElementById(elementIds.menuContent)?.children.length
    const hasMenuInnerContent = !!menuInnerContent && !!document.getElementById(elementIds.menuInnerContent)?.children.length

    if (hasMenuContent) {
      classList.add('menu-content-visible')
      if (!classList.contains('menu-manual')) {
        classList.add('menu-compact')
      }
    } else {
      classList.remove('menu-content-visible')
      if (!classList.contains('menu-manual')) {
        classList.remove('menu-compact')
      }
    }

    if (hasMenuInnerContent) {
      classList.add('menu-inner-content-visible')
    } else {
      classList.remove('menu-inner-content-visible')
    }

    classList[menuSections ? 'remove' : 'add']('no-menu-sections')
    if (className) classList.add(...className.split(/\s+/))
  }, [menuContent, menuInnerContent, menuSections, className])

  return (
    <CSSToCitricAdapter>
      <WelcomeTour />
      <div id={elementIds.layout} style={style}>
        {header && <header id={elementIds.header}><SilentErrorBoundary>{header}</SilentErrorBoundary></header>}
        {extra && <SilentErrorBoundary>{extra}</SilentErrorBoundary>}
        <aside id={elementIds.menu}>
          { menuInnerContent &&
            <nav role="menubar" id={elementIds.menuInnerContent}><SilentErrorBoundary>{menuInnerContent}</SilentErrorBoundary></nav>
          }
          <nav role="menubar" id={elementIds.menuContent}><SilentErrorBoundary>{menuContent}</SilentErrorBoundary></nav>
          {menuSections &&
            <FadingOverflow sides={['top', 'bottom']} scroll="arrows">
              <nav role="menubar" id={elementIds.menuSections}>
                <SilentErrorBoundary>{menuSections}</SilentErrorBoundary>
              </nav>
            </FadingOverflow>
          }
        </aside>
        {children && <div id={elementIds.page}>
          <article id={elementIds.content}><ErrorBoundary>{children}</ErrorBoundary></article>
        </div>}
        <div id={elementIds.bottomDialog} role="dialog"><ErrorBoundary>{bottomDialog}</ErrorBoundary></div>
        <div id={elementIds.backdrop}>
          <div id={elementIds.rightPanel} aria-modal role="dialog"><ErrorBoundary>{rightPanel}</ErrorBoundary></div>
          <div id={elementIds.modal} aria-modal role="dialog"><ErrorBoundary>{modal}</ErrorBoundary></div>
        </div>
        <Toaster />
        <div id={elementIds.accessibilityAnnouncer} aria-atomic aria-live="assertive">
        </div>
      </div>
    </CSSToCitricAdapter>
  )
}

const MenuContentRenderer = ({ content }: Required<Pick<Props['menu'], 'content'>>) => {
  const menuContent = typeof content === 'function' ? content() : content
  return <MenuContent {...menuContent} />
}

const hasInnerContent = (menu: MenuProps): menu is MenuPropsWithDynamicContent => 'innerContent' in menu

/**
 * Renders the layout with a menu and header that follow the config objects passed as parameter.
 * @param props the component's props {@link Props}.
 */
export const Layout = ({ menu, header, children, extra, errorDescriptor, onError, className, style }: Props) => (
  <RawLayout
    header={<Header {...header} />}
    menuSections={menu.sections ? <MenuSections {...menu} /> : undefined}
    menuContent={
      menu.content
        ? <MenuContentRenderer key={'contentKey' in menu ? menu.contentKey : undefined} content={menu.content} />
        : undefined
    }
    menuInnerContent={
      hasInnerContent(menu) && menu.innerContent
        ? <MenuContentRenderer key={'contentKey' in menu ? menu.contentKey : undefined} content={menu.innerContent} />
        : undefined
    }
    errorDescriptor={errorDescriptor}
    onError={onError}
    extra={extra}
    className={className}
    style={style}
  >
    {children}
  </RawLayout>
)
