// @stack-spot/portal-observability/src/event-bus-events.ts
import { Session } from '@stack-spot/auth'
import { Event as EventType } from '@stack-spot/portal-network/api/eventBus'
import Bowser from 'bowser'
import { ulid } from 'ulid'

export type Portal =  'AI' | 'EDP' | 'ACCOUNT' | 'PM' | 'ADP'

export interface EventPageDetails {
  pathNameWithParams: string,
  searchParamsWithVariableName?: string,
  searchParams?: string,
  routeParams?: Record<string, string | string[] | undefined>,
}

export const prepareEvent = ({ data, portal, type }: { data: Record<string, any>, portal: Portal, type: string }) => {
  const isoDate = new Date().toISOString()
  const isoDateWithoutMs = isoDate.split('.')[0] + 'Z'

  const event: EventType = {
    id: ulid(),
    source: `/portal/${portal.toLowerCase()}`,
    specversion: '1.0',
    type: `portal.${type}`,
    time: isoDateWithoutMs,
    data,
  }
  return event
}

export const getCommonProps = (userSession?: Session) => {
  if (!userSession) return 
  
  const { browser } = Bowser.parse(window.navigator.userAgent)
  
  return {
    browser: browser.name || '',
    browserVersion: browser.version,
    email: userSession.getTokenData().email,
    os: browser.name,
    company: userSession.getTokenData().tenant,
  }
}

export function prepareRouteDetails(routeDetails: EventPageDetails) {
  const { pathNameWithParams, searchParams, routeParams, searchParamsWithVariableName } = routeDetails

  return {
    entirePath: searchParamsWithVariableName
      ? `${pathNameWithParams}?${searchParamsWithVariableName}`
      : pathNameWithParams,
    queryParams: searchParams,
    routeParams,
    path: pathNameWithParams,
    currentUrl: window.location.href,
  }
}
