// @stack-spot/citric-react/src/components/Accordion.tsx
import { listToClass } from '@stack-spot/portal-theme'
import { Dictionary, useTranslate } from '@stack-spot/portal-translate'
import { useEffect, useMemo, useState } from 'react'
import { applyCSSVariable } from '../utils/css'
import { withRef } from '../utils/react'
import { CitricComponent } from './CitricComponent'

export interface BaseAccordionProps {
  /**
   * A unique id for this accordion.
   */
  id?: string,
  /**
   * The Accordion's appearance. If unset, the appearance is customizable.
   */
  appearance?: 'card',
  /**
   * Controls the expansion state of the accordion.
   */
  expanded?: boolean,
  /**
   * If the expansion is not controlled externally (expanded/onChange), this tells if the Accordion should start opened.
   * @default false
   */
  startExpanded?: boolean,
  /**
   * Controls the expansion state of the accordion. Called when the accordion is expanded or collapsed.
   */
  onChange?: (expanded: boolean) => void,
  /**
   * The header for the accordion, i.e. the part that is always rendered. You can use either a simple title or a custom set of components,
   * in the second case, you may pass a function that receives the "expand/collapse" button as parameter. 
   */
  header: React.ReactNode | ((expandButton: React.ReactElement) => React.ReactNode),
  /**
   * The maximum height for the accordion.
   * 
   * @default 300
   */
  maxHeight?: number,
}

export type AccordionProps = Omit<React.JSX.IntrinsicElements['div'], 'onChange'> & BaseAccordionProps

/**
 * Creates an accordion, showing/hiding the children according to the state of the expand/collapse checkbox.
 * 
 * The accordion must define a max height, which is 300 by default (measured in pixels). Whenever the content surpasses "maxHeight", a 
 * vertical scroll is created. If your content is much shorter than "maxHeight", consider reducing its value for a better animation.
 * 
 * @example
 * A simple accordion, styled like a card:
 * ```
 * <Accordion appearance="card" header="This is the title of the accordion" maxHeight={80}>
 *   <p>This is the content of the accordion</p>
 * </Accordion>
 * ```
 * 
 * A more complex accordion:
 * ```
 * const header = (toggle: React.ReactElement) => (
 *   <Row justifyContent="space-between" bg="cyan.500" border="cyan.700" radius="sm">
 *     <ImageBox><img src="/my-image.png" /></ImageBox>
 *     <Text>Joseph Clinton</Text>
 *     <Text>40 years old</Text>
 *     {toggle}
 *   </Row>
 * )
 * return (
 *   <Accordion header={header}>
 *     <Center>
 *       <img src="/an-image.png" />
 *       <p>This is the content of the accordion</p>
 *     </Center>
 *   </Accordion>
 * )
 * ```
 */
export const Accordion = withRef((
  { id, appearance, expanded, onChange, header, maxHeight = 300, startExpanded, className, style, children, ...props }: AccordionProps,
) => {
  const t = useTranslate(dictionary)
  const initialExpanded = expanded ?? startExpanded ?? false
  const [ariaHidden, setAriaHidden] = useState(!initialExpanded)
  const accordionId = useMemo(() => id || `${Math.random()}`, [id])

  useEffect(() => {
    if (expanded !== undefined) setAriaHidden(!expanded)
  }, [expanded])

  const expandBtn = <input
    type="checkbox"
    defaultChecked={initialExpanded}
    checked={expanded}
    onChange={() => {
      setAriaHidden(!ariaHidden)
      onChange?.(!expanded)
    }}
    aria-label={ariaHidden ? t.open : t.close}
    title={ariaHidden ? t.open : t.close}
    onKeyDown={e => e.key === 'Enter' && (e.target as HTMLElement).click?.()}
    aria-controls={accordionId}
  />
  const headerContent = typeof header === 'function' ? header(expandBtn) : <label>{header}{expandBtn}</label>
  return (
    <CitricComponent
      tag="div"
      component="accordion"
      className={listToClass([appearance, className])}
      style={applyCSSVariable(style, 'max-height', `${maxHeight}px`)}
      {...props}
    >
      <header>{headerContent}</header>
      <section id={accordionId} aria-hidden={ariaHidden} {...(ariaHidden ? { inert: 'true' } : {})}>
        {appearance === 'card' ? <div className="content">{children}</div> : children}
      </section>
    </CitricComponent>
  )
})

const dictionary = {
  en: {
    open: 'Open',
    close: 'Close',
  },
  pt: {
    open: 'Abrir',
    close: 'Fechar',
  },
} satisfies Dictionary
