// @stack-spot/citric-react/src/components/Avatar.tsx
import { listToClass } from '@stack-spot/portal-theme'
import { WithColorPalette } from '../types'
import { withRef } from '../utils/react'
import { CitricComponent } from './CitricComponent'
import { ImageWithFallback } from './ImageWithFallback'

export interface BaseAvatarProps extends WithColorPalette {
  /**
   * @default 'lg'
   */
  size?: 'xxs' | 'xs' | 'sm' | 'md' | 'lg' | 'xl' | 'xxl',
  /**
   * This name will be reduced to a maximum of 2 capital letters (first letter of each word).
   */
  name?: string,
  /**
   * @default 'circle'
   */
  appearance?: 'square' | 'circle',
  /**
   * If instead of the name you want to use an image, this should be the image's url.
   */
  image?: string,
}

export type AvatarProps = React.JSX.IntrinsicElements['div'] & BaseAvatarProps

/**
 * Renders an avatar. If an image is provided it's rendered within the space reserved to the avatar, otherwise, the initial of the names
 * are used.
 * 
 * @example
 * ```
 * <Avatar name="Rick Sanchez" />
 * ```
 */
export const Avatar = withRef(({ size, name, appearance, className, image, ...props }: AvatarProps) => {
  const parts = name?.split(' ') ?? ['']
  const acronym = parts.length >= 2 ? `${parts[0][0]}${parts[1][0]}` : parts[0][0]
  return (
    <CitricComponent tag="div" component="avatar" className={listToClass([size, appearance, className])} {...props}>
      <ImageWithFallback src={image} fallback={acronym} />
    </CitricComponent>
  )
})
