// @stack-spot/citric-react/src/components/AvatarGroup.tsx
import { listToClass } from '@stack-spot/portal-theme'
import { WithColorPalette } from '../types'
import { withRef } from '../utils/react'
import { Avatar } from './Avatar'
import { CitricComponent } from './CitricComponent'

export interface BaseAvatarGroupProps extends WithColorPalette {
  /**
   * @default 'lg'
   */
  size?: 'xxs' | 'xs' | 'sm' | 'md' | 'lg' | 'xl' | 'xxl',
  /**
   * @default 'circle'
   */
  appearance?: 'square' | 'circle',
  /**
   * The avatars in the group. Use "image" for rendering images instead of the avatar.
   */
  items: { name?: string, image?: string }[],
  /**
   * A maximum number of avatars to show in the group.
   **/
  maxItems?: number,
}

export type AvatarGroupProps = React.JSX.IntrinsicElements['div'] & BaseAvatarGroupProps

/**
 * Stacks a group of avatars horizontally.
 * 
 * @example
 * ```
 * <AvatarGroup items={[{ name: 'John Smith' }, { name: 'Mary Anne', image: 'https://profile.com/123.png' }]} />
 * ```
 */
export const AvatarGroup = withRef(({ size, items, appearance, maxItems, className, ...props }: AvatarGroupProps) => {
  const avatars = items.map(({ name, image }) => <Avatar key={name} name={name} image={image} />, [items])
  return (
    <CitricComponent
      tag="div"
      component="avatar-group"
      className={listToClass([size, appearance, className])}
      data-max-items={maxItems}
      {...props}
    >
      {avatars}
    </CitricComponent>
  )
})
