// @stack-spot/citric-react/src/components/ButtonLink.tsx
import { listToClass } from '@stack-spot/portal-theme'
import { Children, useMemo } from 'react'
import { useCitricController } from '../context/hooks'
import { withRef } from '../utils/react'
import { BaseButtonProps } from './Button'
import { asCitricProps, CitricComponent } from './CitricComponent'

export type BaseButtonLinkProps = Omit<BaseButtonProps, 'onClick' | 'loading' | 'type'>
export type ButtonLinkProps = React.JSX.IntrinsicElements['a'] & BaseButtonLinkProps

/**
 * Renders a link with the appearance of a button.
 * 
 * Whenever a link is clicked, the function `onClickLink` of the nearest CitricController is called with the event and the value of the
 * prop `metadata`.
 * 
 * @example
 * ```
 * <ButtonLink href="#">My Button Link</ButtonLink>
 * ```
 */
export const ButtonLink = withRef((
  { appearance, size, feedback, onClick, className, children, metadata, ...props }: ButtonLinkProps,
) => {
  const citric = useCitricController()
  const isAllLowercase = useMemo(
    () => !Children.toArray(children).some(c => typeof c === 'string' && c.toLocaleLowerCase() !== c),
    [children],
  )

  const linkProps = {
    component: 'button',
    className: listToClass([size, appearance, isAllLowercase && 'short-text', className]),
    onClick: (e: React.MouseEvent<HTMLAnchorElement>) => {
      onClick?.(e)
      citric?.onClickLink?.(e, metadata)
    },
    'data-feedback': feedback || undefined,
    ...props,
  } as const

  return citric?.renderLink
    ? citric.renderLink(asCitricProps({ ...linkProps, children }))
    : <CitricComponent tag="a" {...linkProps}>{children}</CitricComponent>
})
