// @stack-spot/citric-react/src/components/Card.tsx
import { listToClass } from '@stack-spot/portal-theme'
import { useCitricController } from '../context/hooks'
import { WithColorScheme, WithStyleShortcuts } from '../types'
import { applyStyles } from '../utils/css'
import { withRef } from '../utils/react'
import { asCitricProps, CitricComponent } from './CitricComponent'

type SupportedTags = 'div' | 'header' | 'ul' | 'ol' | 'dl' | 'li' | 'section' | 'article' | 'a' | 'button' | 'header'

export interface BaseCardProps<T extends SupportedTags = SupportedTags> extends WithColorScheme, Omit<WithStyleShortcuts, 'bg'> {
  /**
   * HTML tag to render.
   * 
   * @default 'div'
   */
  tag?: T,
  onClick?: (event: React.MouseEvent<HTMLDivElement>) => void,
  /**
   * The size of the card.
   * 
   * @default 'md'
   */
  size?: 'xxs' | 'xs' | 'sm' | 'md' | 'lg' | 'xl' | 'xxl',
  /**
   * The background color level. The main color will depend on the "colorScheme", which is "light" by default.
   * 
   * @default 300
   */
  bgLevel?: 300 | 400 | 500 | 600,
  /**
   * The direction of the items in the card.
   * 
   * @default column
   */
  direction?: 'column' | 'row',
}

// applying conditionals to the type in order to increase TS performance. Ideally we'd use React.JSX.IntrinsicElements[T].
export type CardProps<T extends SupportedTags> = (
  T extends 'a' ? React.JSX.IntrinsicElements['a'] : (
    T extends 'button' ? React.JSX.IntrinsicElements['button'] :
      React.JSX.IntrinsicElements['div'])) & BaseCardProps

/**
 * Renders a box with padding, border and background-color. If the card is focusable or has an onClick listener, hover events are also
 * present.
 * 
 * @example
 * ```
 * <Card>The card content</Card>
 * ```
 */
export const Card = withRef(
  function Card<T extends SupportedTags>({ tag = 'div', onClick, size, bgLevel, className, children, direction, ...props }: CardProps<T>) {
    const citric = useCitricController()
    const clickable = onClick || props.tabIndex === 0 || tag === 'button' || tag === 'a'
    const componentProps = {
      tag,
      component: 'card',
      onClick,
      className: listToClass([clickable && 'clickable', size, bgLevel && `bg-${bgLevel}`, direction, className]),
      ...applyStyles(props),
    } as const
    return tag === 'a' && citric?.renderLink
      ? citric.renderLink(asCitricProps(componentProps))
      : <CitricComponent {...componentProps}>{children}</CitricComponent>
  },
)
