// @stack-spot/citric-react/src/components/Checkbox.tsx
import { ControlledInput, WithColorScheme } from '../types'
import { withRef } from '../utils/react'
import { CitricComponent } from './CitricComponent'

export interface BaseCheckboxProps extends WithColorScheme {
  /**
   * "checkbox" for common check boxes, "switch" for toggles.
   * @default 'checkbox' 
   */
  appearance?: 'checkbox' | 'switch',
  value?: boolean,
  onChange?: (value: boolean) => void,
  children?: React.ReactNode,
}

export type CheckboxProps = ControlledInput & BaseCheckboxProps

/**
 * Renders a simple checkbox. If you want to render multiple checkboxes to control the value of multi-selectable field, prefer using the
 * component `CheckboxGroup`.
 * 
 * Attention: the prop `onChange` receives a boolean (current value) instead of the event.
 * 
 * A checkbox can also be a switch! Just change the property "appearance".
 * 
 * @example
 * ```
 * const [value, setValue] = useState(false)
 * return <Checkbox value={value} setValue={setValue}>This is the checkbox label</Checkbox>
 * ```
 */
export const Checkbox = withRef((
  { appearance = 'checkbox', value, onChange, colorScheme, children, className, style, ...props }: CheckboxProps,
) => {
  const handleChange = onChange ? () => onChange(!value) : undefined
  return children ? (
    <CitricComponent tag="label" component={`${appearance}-row`} colorScheme={colorScheme} style={style} className={className}>
      <input type="checkbox" checked={value} onChange={handleChange} {...props} />
      {children}
    </CitricComponent>
  ) : <CitricComponent
    tag="input"
    type="checkbox"
    component={appearance}
    checked={value}
    onChange={handleChange}
    className={className}
    style={style}
    {...props}
  />
})
