// @stack-spot/citric-react/src/components/ErrorBoundary.tsx
import { Component } from 'react'
import { CitricContext } from '../context/CitricContext'
import { ErrorMessage } from './ErrorMessage'

interface State {
  error?: any,
}

interface Props {
  children: React.ReactNode,
  /**
   * Sets a default error string when a string is provided.
   * 
   * Replaces the error component, if a function is provided.
   */
  message?: React.ReactNode | ((error: any) => React.ReactNode),
}

/**
 * An Error Boundary that renders an error feedback instead of its content if any of its children throws.
 * 
 * You can customize the appearance of the error through the function `renderError` of a CitricController.
 * 
 * Errors can be watched through the function `onError` of a CitricController.
 * 
 * Attention: if you're using React Suspense, consider using the component "FallbackBoundary" instead.
 * 
 * @example
 * ```
 * <ErrorBoundary>
 *   {content}
 * </ErrorBoundary>
 * ```
 */
export class ErrorBoundary extends Component<Props, State> {
  static contextType = CitricContext
  declare context: React.ContextType<typeof CitricContext>

  constructor(props: Props) {
    super(props)
    this.state = {}
  }

  static getDerivedStateFromError(error: any) {
    return { error }
  }

  componentDidCatch(error: any, errorInfo: any) {
    this.context?.onError?.(error)
    // eslint-disable-next-line no-console
    console.error(error, errorInfo)
  }

  componentDidUpdate(prevProps: Readonly<Props>) {
    if (this.state.error && this.props.children !== prevProps.children) {
      this.setState({ error: null })
    }
  }

  private renderCustomErrorUI() {
    return this.context?.renderError?.(this.state.error)
  }

  private renderErrorUI() {
    if (typeof this.props.message === 'function') return this.props.message(this.state.error)
    return (this.props.message || !this.context?.renderError)
      ? <ErrorMessage error={this.props.message || this.state.error} />
      : this.renderCustomErrorUI()
  }


  render() {
    return this.state.error ? this.renderErrorUI() : this.props.children
  }
}
