// @stack-spot/citric-react/src/components/Favorite.tsx
import { listToClass } from '@stack-spot/portal-theme'
import { ControlledInput } from '../types'
import { withRef } from '../utils/react'
import { CitricComponent } from './CitricComponent'

type FavoriteAppearance = 'square' | 'circle' | 'text'

export interface BaseFavoriteProps {
  value: boolean | undefined,
  /**
   * Size of the button.
   * 
   * @default 'md'
   */
  size?: 'sm' | 'md' | 'lg' | 'fit-parent',
  /**
   * Button appearance.
   * 
   * @default 'text'
   */
  appearance?: FavoriteAppearance,
  onChange: (value: boolean) => void,
  /**
   * Whether or not to show the loading feedback.
   * 
   * If `loading` is true, this button becomes disabled.
   */
  loading?: boolean,
}

export type FavoriteProps = Omit<ControlledInput, 'size'> & BaseFavoriteProps

/**
 * Renders a checkbox with the appearance of a favorite button. If unchecked, the stroke of a star is rendered, when checked, a filled
 * yellow star is rendered.
 * 
 * You can also set a loading state to indicate the component is busy.
 * 
 * @example
 * ```
 * const { isFavorite, setFavorite, isLoading } = useRemoteFavoriteStatus()
 * return <Favorite value={isFavorite} onChange={setFavorite} loading={isLoading} />
 * ``` 
 */
export const Favorite = withRef(({ value, size, appearance, onChange, loading, disabled, className, ...props }: FavoriteProps) =>
  <CitricComponent
    tag="input"
    type="checkbox"
    component="favorite"
    className={listToClass([appearance, size, className])}
    checked={!!value}
    onChange={() => onChange(!value)}
    aria-busy={loading}
    disabled={loading || disabled}
    {...props}
  />,
)
