// @stack-spot/citric-react/src/components/FormGroup.tsx
import { listToClass } from '@stack-spot/portal-theme'
import { withRef } from '../utils/react'
import { CitricComponent } from './CitricComponent'
import { IconButton } from './IconBox'
import { Text } from './Text'
import { Tooltip } from './Tooltip'

export interface BaseFormGroupProps {
  /**
   * The error string to show, if any. The empty string is equivalent to undefined, i.e. no errors.
   */
  error?: string,
  /**
   * The input's label.
   */
  label?: string,
  /**
   * A help string, to show as a tooltip.
   */
  help?: string,
  /**
   * A warning string, to show right below the field, at all times.
   */
  warning?: string,
}

export type FormGroupProps = React.JSX.IntrinsicElements['div'] & BaseFormGroupProps

/**
 * A form group renders everything a field in a form needs: the input, its label, a tooltip (help), an error and a remark (warning).
 * 
 * Labels, errors, tooltips and warnings are only rendered when they're strings with at least one character.
 * 
 * @example
 * ```
 * <FormGroup label="Birth Date" help="The date when you were born" error="The date you typed is invalid!">
 *   <Input type="date" />
 * </FormGroup>
 * ```
 */
export const FormGroup = withRef(({ error, help, label, warning, className, children, ...props }: FormGroupProps) => (
  <CitricComponent tag="div" component="form-group" className={listToClass([className, error && 'error'])} {...props}>
    <label>
      {help ? (
        <div className="row">
          <Text>{label}</Text>
          <Tooltip content={help}>
            <IconButton icon="InfoCircle" size="xs" />
          </Tooltip>
        </div>
      ): label}
      {children}
      <div className="feedback">{error}</div>
      {warning && <div className="warning">{warning}</div>}
    </label>
  </CitricComponent>
))
