// @stack-spot/citric-react/src/components/ImageWithFallback.tsx
import { listToClass } from '@stack-spot/portal-theme'
import { useEffect, useMemo, useState } from 'react'
import { WithStyleShortcuts } from '../types'
import { getStyleFromProps } from '../utils/css'
import { Skeleton } from './Skeleton'

export interface BaseImageWithFallbackProps extends Pick<WithStyleShortcuts, 'w' | 'h' | 'radius' | 'bg'> {
  /**
   * The react element to fallback to if the image can't be rendered.
   */
  fallback: React.ReactNode,
  /**
   * Whether or not to show a skeleton when the image is still loading.
   * @default false
   */
  showLoading?: boolean,
}

export type ImageWithFallbackProps = JSX.IntrinsicElements['img'] & BaseImageWithFallbackProps

/**
 * Attempts to render an image. If it succeeds, the image is displayed, otherwise, a fallback is rendered.
 * 
 * This is very useful if the value of "src" may be empty or if it may be broken. Instead of rendering nothing or the default browser error,
 * it renders the fallback passed as parameter.
 * 
 * @example
 * ```
 * <ImageWithFallback src={item.image} fallback={<Icon icon="Agent" />} />
 * ```
 */
export const ImageWithFallback = (
  { onLoad, onError, className, style: ogStyle, fallback, showLoading, w, h, radius, bg, ...props }: ImageWithFallbackProps,
) => {
  const [state, setState] = useState<'loading' | 'ready' | 'error'>(props.src ? 'loading' : 'error')
  const isLoading = state === 'loading' && showLoading
  const style = useMemo(() => ({ ...getStyleFromProps({ w, h, radius, bg }), ...ogStyle }), [ogStyle, w, h, radius, bg])

  function handleLoad(e: React.SyntheticEvent<HTMLImageElement, Event>) {
    if (state === 'loading') setState('ready')
    onLoad?.(e)
  }

  function handleError(e: React.SyntheticEvent<HTMLImageElement, Event>) {
    setState('error')
    onError?.(e)
  }

  useEffect(() => {
    if (state !== 'loading') setState(props.src ? 'loading' : 'error')
  }, [props.src])

  if (state === 'error') return <div className={listToClass([className, 'center'])} style={style}>{fallback}</div>
  return <>
    {isLoading && <Skeleton className={className} style={style} />}
    <img
      onLoad={handleLoad}
      onError={handleError}
      style={{ ...style, ...(isLoading ? { pointerEvents: 'none', position: 'absolute', opacity: 0 } : {}) }}
      aria-hidden={isLoading}
      className={className}
      {...props}
    />
  </>
}
