// @stack-spot/citric-react/src/components/ImageBox.tsx
import { listToClass } from '@stack-spot/portal-theme'
import { useCitricController } from '../context/hooks'
import { HTMLTag, WithColorPalette, WithColorScheme } from '../types'
import { withRef } from '../utils/react'
import { asCitricProps, CitricComponent } from './CitricComponent'

type ImageBoxTag = 'a' | 'button' | 'span' | 'div'

export interface BaseImageBoxProps<T extends ImageBoxTag> extends WithColorPalette, WithColorScheme {
  /**
   * The HTML element to render.
   * 
   * @default 'div'
   */
  tag?: T,
  /**
   * The box appearance.
   * 
   * @default 'circle'
   */
  appearance?: 'circle' | 'square',
  /**
   * Size of the box.
   * 
   * - xs: 20px;
   * - sm: 24px;
   * - md: 32px;
   * - lg: 40px.
   * 
   * @default 'sm'
   */
  size?: 'xs' | 'sm' | 'md' | 'lg',
  /**
   * Animated text to show on when the button/link is clicked. This is only valid for buttons and anchors.
   */
  feedback?: string,
  /**
   * Only valid if `tag` is "button" or "a".
   * 
   * Metadata for the general onClick event, set by the CitricController. Useful for creating analytics data.
   * 
   * This only takes effect if there's a CitricController in React's context. The value of `metadata` is passed to the function
   * `onClickButton/onClickLink` of the controller.
   * 
   * @default false
   */
  metadata?: any,
}

export type ImageBoxProps<T extends ImageBoxTag> = HTMLTag[T] & BaseImageBoxProps<T>

/**
 * Renders a wrapper for its child (normally an image). The image will be resized and cropped to fit the container. The image is not cropped
 * if the property "feedback" is set.
 * 
 * Hover and focus effects are applied if the ImageBox is focusable.
 * 
 * This works exactly like an IconBox, but instead of an Icon, it can contain anything.
 * 
 * @example
 * ```
 * <ImageBox><img src="https://images.com/myimage.png" /></ImageBox>
 * ```
 */
export const ImageBox = withRef(
  function ImageBox<T extends ImageBoxTag = 'div'>(
    { tag, appearance, size, className, metadata, onClick, feedback, ...props }: ImageBoxProps<T>,
  ) {
    props['aria-label'] ||= props.title // accessibility
    const citric = useCitricController()

    function handleClick(e: React.MouseEvent<any>) {
      onClick?.(e)
      if (tag === 'button') citric?.onClickButton?.(e, metadata)
      else if (tag === 'a') citric?.onClickLink?.(e, metadata)
    }

    const componentProps = {
      tag: (tag || 'i') as any,
      component: 'icon-box',
      className: listToClass([appearance, size, className]),
      'data-feedback': feedback || undefined,
      onClick: ['button', 'a'].includes(tag ?? '') ? handleClick : onClick,
      ...props,
    } as const

    return tag === 'a' && citric?.renderLink
      ? citric.renderLink(asCitricProps(componentProps))
      : <CitricComponent {...componentProps} />
  },
)

/**
 * A shortcut for `<ImageBox tag="button">`.
 * 
 * Whenever a button is clicked, the function `onClickButton` of the nearest CitricController is called with the event and the value of the
 * prop `metadata`.
 * 
 * @example
 * ```
 * <ImageButton><img src="https://images.com/myimage.png" /></ImageButton>
 * ```
 */
export const ImageButton = withRef(
  function ImageButton(props: Omit<ImageBoxProps<'button'>, 'tag'>) {
    return <ImageBox {...props} tag="button" type={props.type || 'button'} />
  },
)

/**
 * A shortcut for `<ImageBox tag="a">`.
 * 
 * Whenever a link is clicked, the function `onClickLink` of the nearest CitricController is called with the event and the value of the
 * prop `metadata`.
 * 
 * @example
 * ```
 * <ImageLink href="#"><img src="https://images.com/myimage.png" /></ImageButton>
 * ```
 */
export const ImageLink = withRef(
  function ImageLink(props: Omit<ImageBoxProps<'a'>, 'tag'>) {
    return <ImageBox {...props} tag="a" />
  },
)
