// @stack-spot/citric-react/src/components/layout.tsx
import { listToClass } from '@stack-spot/portal-theme'
import { createElement } from 'react'
import { HTMLTag, WithStyleShortcuts } from '../types'
import { applyStyles } from '../utils/css'
import { withRef } from '../utils/react'

export const layout = {
  center: 'center',
  row: 'row',
  column: 'column',
}

export interface BaseLayoutProps<T extends keyof HTMLTag> extends WithStyleShortcuts {
  /**
   * HTML tag to render.
   * 
   * @default 'div'
   */
  tag?: T,
}

export interface BaseRowProps<T extends keyof HTMLTag> extends BaseLayoutProps<T> {
  /**
   * Whether or not to center items horizontally.
   * 
   * @default true
   */
  center?: boolean,
}

export interface BaseColumnProps<T extends keyof HTMLTag> extends BaseLayoutProps<T> {
  /**
   * Whether or not to center items vertically.
   * 
   * @default false
   */
  center?: boolean,
}

export type CenterProps<T extends keyof HTMLTag> = HTMLTag[T] & BaseLayoutProps<T>
export type RowProps<T extends keyof HTMLTag> = HTMLTag[T] & BaseRowProps<T>
export type ColumnProps<T extends keyof HTMLTag> = HTMLTag[T] & BaseColumnProps<T>

/**
 * Centers the content both vertically and horizontally. Elements are placed in a column (flex).
 * 
 * This element accepts many style shortcuts as attributes: bg, fg, border, radius, justifyContent, alignItems, flex, gap, m, mt, mb, ml,
 * mr, p, pt, pb, pl, pr, w, h.
 * 
 * @example
 * ```
 * <Center w="200px" h="200px" gap="10px">
 *   <img src="/an-image.png" />
 *   <Text>This content is centered!</Text>
 * </Center>
 * ```
 */
export const Center = withRef(
  function Center<T extends keyof HTMLTag = 'div'>({ tag, className, ...props }: CenterProps<T>) {
    return createElement(tag || 'div', { className: listToClass(['center', className]), ...applyStyles(props) })
  },
)

/**
 * Creates a row (flex). The elements in this row are horizontally centered unless `center = false`.
 * 
 * @example
 * ```
 * <Row gap={3}>
 *   <Icon icon="Stack" />
 *   <Text>Stacks</Text>
 * </Row>
 * ```
 */
export const Row = withRef(
  function Row<T extends keyof HTMLTag = 'div'>({ tag, className, center, ...props }: RowProps<T>) {
    return createElement(
      tag || 'div',
      { className: listToClass(['row', center === false && 'no-center', className]), ...applyStyles(props) },
    )
  },
)

/**
 * Creates a column (flex). The elements in this row are placed at the start of the horizontal axis, unless `center = true`, in this case,
 * they're centered.
 * 
 * @example
 * ```
 * <Column gap={5}>
 *   <Text>Item 1</Text>
 *   <Text>Item 2</Text>
 *   <Text>Item 3</Text>
 * </Row>
 * ```
 */
export const Column = withRef(
  function Column<T extends keyof HTMLTag = 'div'>({ tag, className, center, ...props }: ColumnProps<T>) {
    return createElement(tag || 'div', { className: listToClass(['column', center && 'center', className]), ...applyStyles(props) })
  },
)
