// @stack-spot/citric-react/src/components/Link.tsx
import { useCitricController } from '../context/hooks'
import { TextAppearance, WithColor } from '../types'
import { applyColor, applyTextAppearance } from '../utils/css'
import { withRef } from '../utils/react'
import { asCitricProps, CitricComponent } from './CitricComponent'

export interface BaseLinkProps extends WithColor {
  /**
   * The text appearance (font and size).
   * By default, it inherits from the parent element.
   */
  appearance?: TextAppearance,
  /**
   * Metadata for the general onClick event, set by the CitricController. Useful for creating analytics data.
   * 
   * This only takes effect if there's a CitricController in React's context. The value of `metadata` is passed to the function
   * `onClickLink` of the controller.
   * 
   * @default false
   */
  metadata?: any,
  /**
   * Whether or not to underline the link.
   * 
   * @default true
   */
  underline?: boolean,
}

export type LinkProps = React.JSX.IntrinsicElements['a'] & BaseLinkProps

/**
 * Renders an html anchor by default, the actual component to render may be set on a CitricController, through the function `renderLink`.
 * 
 * Whenever a link is clicked, the function `onClickLink` of the nearest CitricController is called with the event and the value of the
 * prop `metadata`.
 */
export const Link = withRef(
  ({ appearance, color, style, className, children, onClick, metadata, underline = true, ...props }: LinkProps) => {
    const citric = useCitricController()
    const linkProps = {
      component: 'link',
      style: applyColor({ ...style, textDecoration: underline ? undefined : 'none' }, color),
      className: applyTextAppearance(className, appearance),
      onClick: (e: React.MouseEvent<HTMLAnchorElement>) => {
        onClick?.(e)
        citric?.onClickLink?.(e, metadata ?? false)
      },
      ...props,
    } as const
    return citric?.renderLink
      ? citric.renderLink(asCitricProps({ ...linkProps, children }))
      : <CitricComponent tag="a" {...linkProps}>{children}</CitricComponent>
  },
)
