// @stack-spot/citric-react/src/components/Rating.tsx
import { useCallback } from 'react'
import { withRef } from '../utils/react'
import { CitricComponent } from './CitricComponent'

export type RatingValue = 1 | 2 | 3 | 4 | 5

export interface BaseRatingProps {
  /** 
   * The current value of the rating. 
   */
  value: RatingValue | undefined,
  /** 
   * Callback when the value changes.
   */
  onChange?: (value: RatingValue) => void,
  /**
   * The name of this input.
   */
  name?: string,
  /**
   * Whether the rating inputs are disabled.
   */
  disabled?: boolean,
  /**
   * Whether the rating inputs are readonly.
   */
  readonly?: boolean,
}

export type RatingProps = Omit<React.JSX.IntrinsicElements['div'], 'onChange' | 'children'> & BaseRatingProps

/**
 * Renders a 5 star rating UI. The user can select 1 to 5 stars.
 * 
 * @example
 * ```
 * const [value, setValue] = useState<RatingValue | undefined>()
 * return <Rating value={value} onChange={setValue} />
 * ```
 * In the example above, `RatingValue` is any integer from 1 to 5.
 */
export const Rating = withRef(({ value, onChange, name, disabled, readonly, ...props }: RatingProps) => {
  const handleChange = useCallback((newValue: RatingValue) => {
    if (!disabled && !readonly) {
      onChange?.(newValue)
    }
  }, [disabled, readonly, onChange])

  return (
    <CitricComponent tag="div" component="rating" {...props}>
      <input
        type="radio"
        value="1"
        name={name}
        checked={value === 5}
        onChange={() => handleChange(5)}
        disabled={disabled}
        readOnly={readonly}
      />
      <input
        type="radio"
        value="2"
        name={name}
        checked={value === 4}
        onChange={() => handleChange(4)}
        disabled={disabled}
        readOnly={readonly}
      />
      <input
        type="radio"
        value="3"
        name={name}
        checked={value === 3}
        onChange={() => handleChange(3)}
        disabled={disabled}
        readOnly={readonly}
      />
      <input
        type="radio"
        value="4"
        name={name}
        checked={value === 2}
        onChange={() => handleChange(2)}
        disabled={disabled}
        readOnly={readonly}
      />
      <input
        type="radio"
        value="5"
        name={name}
        checked={value === 1}
        onChange={() => handleChange(1)}
        disabled={disabled}
        readOnly={readonly}
      />
    </CitricComponent>
  )
})
