// @stack-spot/citric-react/src/components/Skeleton.tsx
import { listToClass } from '@stack-spot/portal-theme'
import { withRef } from '../utils/react'
import { CitricComponent } from './CitricComponent'

export interface BaseSkeletonProps {
  /**
   * The skeleton's appearance.
   * 
   * @default 'square'
   */
  appearance?: 'square' | 'circle',
  /**
   * The skeleton's width.
   * 
   * @default "the available width if 'square', 50px if 'circle'".
   */
  width?: string,
  /**
   * The skeleton's width.
   * 
   * @default "32px if 'square', 50px if 'circle'".
   */
  height?: string,
  /**
   * The background color level.
   * 
   * @default 500
   */
  bgLevel?: 400 | 500 | 600, 
}

export type SkeletonProps = React.JSX.IntrinsicElements['div'] & BaseSkeletonProps

/**
 * A loading feedback with a discrete animation. Use this to replace elements on the page that are still loading and will be replaced by
 * the same shape when the loading ends.
 * 
 * Use the property "bgLevel" to manage how light/dark the background is.
 * 
 * @example
 * ```
 * <Skeleton width="100%" height="20px" />
 * ```
 */
export const Skeleton = withRef(({ appearance, className, style, width, height, bgLevel, ...props }: SkeletonProps) =>
  <CitricComponent
    tag="div"
    component="skeleton"
    className={listToClass([appearance, bgLevel && `bg-${bgLevel}`, className])}
    style={{ width, height, ...style }}
    {...props}
  />,
)
