// @stack-spot/citric-react/src/components/Slider.tsx
import { listToClass } from '@stack-spot/portal-theme'
import { ControlledInput, WithColorPalette, WithColorScheme } from '../types'
import { applyCSSVariable } from '../utils/css'
import { withRef } from '../utils/react'
import { CitricComponent } from './CitricComponent'

export interface BaseSliderProps extends WithColorScheme, WithColorPalette {
  value: number,
  /**
   * The minimum value in the numeric range.
   * 
   * @default 0
   */
  min?: number,
  /**
   * The maximum value in the numeric range.
   * 
   * @default 100
   */
  max?: number,
  onChange: (value: number) => void,
  /**
   * When should we render the current value of the input?
   * 
   * @default 'always'
   */
  showValue?: 'always' | 'hover' | 'never',
  /**
   * A function to customize how the value of the input is rendered.
   * 
   * @default "{value}% if 'min' and 'max' are not provided. {value} otherwise."
   */
  renderValue?: (value: number) => string,
}

export type SliderProps =  ControlledInput & BaseSliderProps

/**
 * A UI element where the user can slide a knob over a bar to select a number. By default, a number between 0 and 100 can be selected, use
 * the properties "min" and "max" to change this.
 * 
 * Attention: "onChange" receives the new value (number) instead of the event.
 * 
 * @example
 * ```
 * const [value, setValue] = useState(0)
 * return <Slider value={value} setValue={setValue} />
 * ```
 */
export const Slider = withRef((
  { value, onChange, min, max, style, showValue, renderValue, colorPalette, colorScheme, className, ...props }: SliderProps,
) => {
  const percent = Math.floor((value / ((max ?? 100) - (min ?? 0))) * 100)
  style = applyCSSVariable(style, 'percent', percent)
  const rangeProps = {
    value,
    min,
    max,
    onChange: (e: React.ChangeEvent<HTMLInputElement>) => onChange(parseInt(e.target.value)),
    ...props,
  }  

  return showValue === 'never'
    ? <CitricComponent
      tag="input"
      type="range"
      component="slider"
      style={style}
      className={className}
      colorPalette={colorPalette}
      colorScheme={colorScheme}
      {...rangeProps}
    />
    : (
      <CitricComponent
        tag="div"
        component="labeled-slider"
        style={style}
        colorPalette={colorPalette}
        colorScheme={colorScheme}
        className={listToClass([className, showValue === 'hover' && 'value-on-hover'])}
      >
        <input type="range" {...rangeProps} />
        <div className="value">
          {renderValue ? renderValue(value) : `${value}${min === undefined && max === undefined ? '%' : ''}`}
        </div>
      </CitricComponent>
    )
})
