// @stack-spot/citric-react/src/components/Text.ts
import { listToClass } from '@stack-spot/portal-theme'
import { createElement } from 'react'
import { HTMLTag, TextAppearance, WithColor } from '../types'
import { applyColor, applyTextAppearance } from '../utils/css'
import { withRef } from '../utils/react'

type SupportedTags = 'span' | 'small' | 'p' | `h${1|2|3|4|5|6}` | 'pre'

export interface BaseTextProps<T extends SupportedTags = SupportedTags> extends WithColor {
  /**
   * If "tag" is not defined, it will be based on the value of "appearance".
   * If neither "appearance" or "tag" are defined, the tag will be "p".
   */
  tag?: T,
  /**
   * If "appearance" is not defined, it will be based on the value of "tag".
   * If neither "appearance" or "tag" are defined, the appearance will be inherited from the parent HTML element.
   */
  appearance?: TextAppearance,
  /**
   * Add default margins to the text.
   * 
   * @default false
   */
  showMargins?: boolean,
  /**
   * Shortcut to `style.fontWeight`.
   */
  weight?: React.CSSProperties['fontWeight'],
  /**
   * Shortcut to `style.textAlign`.
   */
  align?: React.CSSProperties['textAlign'],
  /**
   * Shortcut to `style.whiteSpace = 'nowrap'; style.overflow = 'hidden'; style.textOverflow = 'ellipsis'`.
   */
  nowrapEllipsis?: boolean,
}

export type TextProps<T extends SupportedTags> = HTMLTag[T] & BaseTextProps<T>

const tagToAppearance: Record<SupportedTags, TextAppearance | undefined> = {
  h1: 'h1',
  h2: 'h2',
  h3: 'h3',
  h4: 'h4',
  h5: 'h5',
  h6: 'h6',
  span: undefined,
  small: 'microtext1',
  p: 'body2',
  pre: 'code2',
}

const appearanceToTag: Record<TextAppearance, SupportedTags> = {
  h1: 'h1',
  h2: 'h2',
  h3: 'h3',
  h4: 'h4',
  h5: 'h5',
  h6: 'h6',
  body1: 'p',
  body2: 'p',
  code1: 'pre',
  code2: 'pre',
  display1: 'h1',
  microtext1: 'small',
  overheader1: 'p',
  overheader2: 'p',
  subtitle1: 'p',
  subtitle2: 'p',
  subtitle3: 'p',
  subtitle4: 'p',
}

/**
 * Renders a text. The tag can be set by the property "tag". If unset, the tag rendered will be based on the property "appearance". If
 * "appearance" is also unset, the tag "p" is rendered.
 * 
 * Texts have no margin by default, to enable margins, use `showMargins = true`.
 * 
 * If the appearance is not set, it will be inherited from the parent.
 * 
 * A font can be applied to a text without this component, just use `{ font: theme.font.$textAppearance }` in your style.
 * 
 * @example
 * ```
 * <Text appearance="body2">Hello World!</Text>
 * ```
 */
export const Text = withRef(
  function Text<T extends SupportedTags>(
    { tag, appearance, color, showMargins, weight, align, nowrapEllipsis, style, className, children, ...props }: TextProps<T>,
  ) {
    const renderedTag = tag || appearanceToTag[appearance || 'body2']
    const renderedAppearance = appearance || (tag ? tagToAppearance[tag] : undefined)
    if (!appearance && !showMargins) style = { margin: 0, ...style }
    if (weight) style = { fontWeight: weight, ...style }
    if (align) style = { textAlign: align, ...style }
    if (nowrapEllipsis) style = { whiteSpace: 'nowrap', overflow: 'hidden', textOverflow: 'ellipsis', ...style }
    return createElement(
      renderedTag,
      {
        className: applyTextAppearance(listToClass([className, showMargins && 'text-with-margins']), renderedAppearance),
        style: applyColor(style, color),
        ...props,
      },
      children,
    )
  },
)
