// @stack-spot/citric-react/src/components/Tooltip.tsx
import { listToClass } from '@stack-spot/portal-theme'
import { useRef } from 'react'
import { WithColorPalette, WithColorScheme } from '../types'
import { Overlay } from './Overlay'
import { BaseOverlayProps } from './Overlay/types'

export interface BaseTooltipProps extends
  Omit<BaseOverlayProps<'div'>, 'tag' | 'onRenderChild' | 'reference'>, WithColorPalette, WithColorScheme {
  /**
   * Whether or not to show the tooltip's arrow.
   * 
   * @default true
   */
  showArrow?: boolean,
  /**
   * The unique id of the HTML element to be created to show the tooltip. If not provided, a random number will be used.
   */
  tooltipId?: string,
  /**
   * Makes the tooltip text compact, centering it better when it's a single line.
   * 
   * Effectively, it makes the line height equals to 1 and applies a slightly larger padding.
   * 
   * @default true
   */
  compact?: boolean,
  /**
   * The max-width of the tooltip. Defaults to 250px.
   */
  maxWidth?: string,
  /**
   * A space between the tooltip and the element who triggered it.
   * 
   * @default '6px'
   */
  margin?: string,
  /**
   * Whether or not to disable the tooltip, i.e. not show it.
   * 
   * @default false
   */
  disabled?: boolean,
}

export type TooltipProps = Omit<React.JSX.IntrinsicElements['div'], 'content'> & BaseTooltipProps

/**
 * Renders a tooltip overlay for its child. By default, the tooltip is rendered on hover, but this can be changed through the property
 * "triggerOn".
 * 
 * The content of the tooltip can be any React element.
 * 
 * @example
 * ```
 * <Tooltip content="This is some help for you">
 *   <IconButton icon="Question" />
 * </Tooltip>
 * ```
 */
export const Tooltip = ({
  tooltipId, children, showArrow = true, compact = true, attributes, colorScheme, colorPalette, disabled, maxWidth = '250px',
  margin = '6px', ...props
}: TooltipProps) => {
  const id = useRef(tooltipId || `${Math.random()}`)
  return disabled ? children : (
    <Overlay
      attributes={{
        className: listToClass([attributes?.className, showArrow && 'with-arrow', compact && 'compact']),
        style: { margin, maxWidth, lineHeight: typeof props.content === 'string' ? 'normal' : undefined, ...attributes?.style },
        role: 'tooltip',
        id: id.current,
        'data-citric': 'tooltip',
        'data-color-scheme': colorScheme,
        'data-color-palette': colorPalette,
        ...attributes,
      }}
      onRenderChild={e => e.setAttribute('aria-describedby', id.current)}
      {...props}
    >
      {children}
    </Overlay>
  )
}
