// @stack-spot/portal-components/src/components/notification/NotificationPlaceholder.tsx
import { AsProp } from '@citric/core/dist/forward-ref-as'
import { SxProperties } from '@citric/core/dist/sx'
import { Dictionary, useTranslate } from '@stack-spot/portal-translate'
import { PlaceholderCallToAction } from '../Placeholder'

interface Props {
  isSearch?: boolean,
  style?: React.CSSProperties,
  className?: string,
  sx?: SxProperties,
  sxCard?: SxProperties,
  titleAs?: AsProp,
}

export const NotificationPlaceholder = ({ isSearch, sx, sxCard, className, style, titleAs }: Props) => {
  const t = useTranslate(dictionary)

  return <PlaceholderCallToAction
    title={
      isSearch
        ? (<>{t.placeholderSearchTitle}<br />{t.placeholderSearchSubTitle}</>)
        : (<>{t.placeholderTitle}<br />{t.placeholderSubTitle}</>)
    }
    description={
      isSearch
        ? (<>{t.placeholderSearchDescription}<br />{t.placeholderSearchSubDescription}</>)
        : (<>{t.placeholderDescription}<br />{t.placeholderSubDescription}</>)
    }
    fullWidth
    sx={sx}
    sxCard={sxCard}
    className={className}
    style={style}
    titleAs={titleAs}
    placeholderImageConfig={{ width: '200', height: '200' }}
  />
}

const dictionary = {
  en: {
    placeholderTitle: 'You don’t have',
    placeholderSubTitle: 'any notifications yet.',
    placeholderDescription: 'Relevant updates will show up here',
    placeholderSubDescription: 'as soon as they’re available.',
    placeholderSearchTitle: 'No notifications',
    placeholderSearchSubTitle: 'match your filters.',
    placeholderSearchDescription: 'Try changing the filters',
    placeholderSearchSubDescription: 'to see more results.',
  },
  pt: {
    placeholderTitle: 'Você ainda não tem',
    placeholderSubTitle: 'notificações.',
    placeholderDescription: 'Quando houver alguma atualização relevante,',
    placeholderSubDescription: 'ela será exibida aqui.',
    placeholderSearchTitle: 'Nenhuma notificação',
    placeholderSearchSubTitle: 'corresponde aos filtros',
    placeholderSearchDescription: 'Experimente mudar os filtros',
    placeholderSearchSubDescription: 'para encontrar mais notificações.',
  },
} satisfies Dictionary
