// @stack-spot/portal-network/dist/api/account.js
/**
 * OpenAPI definition
 * v0
 * DO NOT MODIFY - This file has been generated using oazapfts.
 * See https://www.npmjs.com/package/oazapfts
 */
import * as Oazapfts from "@oazapfts/runtime";
import * as QS from "@oazapfts/runtime/query";
export const defaults = {
    headers: {},
    baseUrl: "https://account-account-api.dev.stackspot.com",
};
const oazapfts = Oazapfts.runtime(defaults);
export const servers = {
    generatedServerUrl: "https://account-account-api.dev.stackspot.com"
};
/**
 * Retrieves a list of rate limit with the associated service credential.
 */
export function getAllServiceCredentialRateLimit({ size, page, sort, direction, search }, opts) {
    return oazapfts.ok(oazapfts.fetchJson(`/v2/service-credentials/rate-limit${QS.query(QS.explode({
        size,
        page,
        sort,
        direction,
        search
    }))}`, {
        ...opts
    }));
}
/**
 * Updates the rate limit of a service credential.
 */
export function updateServiceCredentialRateLimit({ serviceCredentialRateLimitRequest }, opts) {
    return oazapfts.ok(oazapfts.fetchJson("/v2/service-credentials/rate-limit", oazapfts.json({
        ...opts,
        method: "PUT",
        body: serviceCredentialRateLimitRequest
    })));
}
/**
 * Returns the Feature Flag.
 */
export function getFeatureFlag({ id }, opts) {
    return oazapfts.ok(oazapfts.fetchJson(`/v2/feature-flags/${encodeURIComponent(id)}`, {
        ...opts
    }));
}
/**
 * Update a feature flag by id directly assigned to an account holder
 */
export function updateFeatureFlag({ id, updateFeatureFlagAccountRequest }, opts) {
    return oazapfts.ok(oazapfts.fetchJson(`/v2/feature-flags/${encodeURIComponent(id)}`, oazapfts.json({
        ...opts,
        method: "PUT",
        body: updateFeatureFlagAccountRequest
    })));
}
/**
 * Returns the Feature Flag.
 */
export function getFeatureFlag1({ id }, opts) {
    return oazapfts.ok(oazapfts.fetchJson(`/v2/admin/feature-flags/${encodeURIComponent(id)}`, {
        ...opts
    }));
}
/**
 * Updates an existing Feature Flag with the provided details.
 */
export function update({ id, updateFeatureFlagRequest }, opts) {
    return oazapfts.ok(oazapfts.fetchJson(`/v2/admin/feature-flags/${encodeURIComponent(id)}`, oazapfts.json({
        ...opts,
        method: "PUT",
        body: updateFeatureFlagRequest
    })));
}
/**
 * Deletes an existing Feature Flag with the provided id.
 */
export function deleteV2AdminFeatureFlagsById({ id }, opts) {
    return oazapfts.ok(oazapfts.fetchJson(`/v2/admin/feature-flags/${encodeURIComponent(id)}`, {
        ...opts,
        method: "DELETE"
    }));
}
/**
 * Updates an existing Feature Flag resources with the provided details.
 */
export function updateAccountAssociation({ featureFlagId, accountId, updateFeatureFlagAccountRequest }, opts) {
    return oazapfts.ok(oazapfts.fetchJson(`/v2/admin/feature-flags/${encodeURIComponent(featureFlagId)}/account/${encodeURIComponent(accountId)}`, oazapfts.json({
        ...opts,
        method: "PUT",
        body: updateFeatureFlagAccountRequest
    })));
}
/**
 * Creates a new Feature Flag resource for the given account.
 */
export function createAccountAssociation({ featureFlagId, accountId, createFeatureFlagAccountRequest }, opts) {
    return oazapfts.ok(oazapfts.fetchJson(`/v2/admin/feature-flags/${encodeURIComponent(featureFlagId)}/account/${encodeURIComponent(accountId)}`, oazapfts.json({
        ...opts,
        method: "POST",
        body: createFeatureFlagAccountRequest
    })));
}
/**
 * Deletes Feature Flag resource for the given account.
 */
export function deleteAccountAssociation({ featureFlagId, accountId }, opts) {
    return oazapfts.ok(oazapfts.fetchJson(`/v2/admin/feature-flags/${encodeURIComponent(featureFlagId)}/account/${encodeURIComponent(accountId)}`, {
        ...opts,
        method: "DELETE"
    }));
}
/**
 * Retrieves a list of SCM credentials associated with the current user's account, including secret names and provider details.
 */
export function listScmCredentials(opts) {
    return oazapfts.ok(oazapfts.fetchJson("/v1/users/scm-credentials", {
        ...opts
    }));
}
/**
 * Updates the SCM credentials for a user by saving the new secret information and triggering an update event.
 */
export function scmCredentialUpdate({ userScmInfoRequest }, opts) {
    return oazapfts.ok(oazapfts.fetchJson("/v1/users/scm-credentials", oazapfts.json({
        ...opts,
        method: "PUT",
        body: userScmInfoRequest
    })));
}
/**
 * Saves user SCM credentials by validating the user, storing the credentials securely, and triggering a creation event.
 */
export function scmCredentialSave({ userScmInfoRequest }, opts) {
    return oazapfts.ok(oazapfts.fetchJson("/v1/users/scm-credentials", oazapfts.json({
        ...opts,
        method: "POST",
        body: userScmInfoRequest
    })));
}
/**
 * Deletes the SCM credentials of a user from the repository and secrets manager, and sends a deletion event if the credentials exist.
 */
export function scmDelete(opts) {
    return oazapfts.ok(oazapfts.fetchJson("/v1/users/scm-credentials", {
        ...opts,
        method: "DELETE"
    }));
}
/**
 * Get secret by ID
 */
export function getById({ secretId }, opts) {
    return oazapfts.ok(oazapfts.fetchJson(`/v1/secrets/${encodeURIComponent(secretId)}`, {
        ...opts
    }));
}
/**
 * Update secret information
 */
export function updateSecret({ secretId, updateSecretRequest }, opts) {
    return oazapfts.ok(oazapfts.fetchText(`/v1/secrets/${encodeURIComponent(secretId)}`, oazapfts.json({
        ...opts,
        method: "PUT",
        body: updateSecretRequest
    })));
}
/**
 * Delete secret by ID
 */
export function deleteSecret({ immediate, secretId }, opts) {
    return oazapfts.ok(oazapfts.fetchText(`/v1/secrets/${encodeURIComponent(secretId)}${QS.query(QS.explode({
        immediate
    }))}`, {
        ...opts,
        method: "DELETE"
    }));
}
/**
 * Retrieves a specific scope by its ID for the current user's tenant and returns it.
 */
export function getById1({ scopeId }, opts) {
    return oazapfts.ok(oazapfts.fetchJson(`/v1/scopes/${encodeURIComponent(scopeId)}`, {
        ...opts
    }));
}
/**
 * Updates the details of a specific scope within the tenant using the provided scope ID and update request.
 */
export function updateScope({ scopeId, iamUpdateScopeRequest }, opts) {
    return oazapfts.ok(oazapfts.fetchJson(`/v1/scopes/${encodeURIComponent(scopeId)}`, oazapfts.json({
        ...opts,
        method: "PUT",
        body: iamUpdateScopeRequest
    })));
}
/**
 * Deletes a specific scope identified by its ID within the user's current tenant.
 */
export function deleteScope({ scopeId }, opts) {
    return oazapfts.ok(oazapfts.fetchJson(`/v1/scopes/${encodeURIComponent(scopeId)}`, {
        ...opts,
        method: "DELETE"
    }));
}
/**
 * Retrieves a list of SCM credentials associated with the current user's account.
 */
export function listScmCredentials1(opts) {
    return oazapfts.ok(oazapfts.fetchJson("/v1/scm-credentials", {
        ...opts
    }));
}
/**
 * Updates the SCM credentials for the current user's account, saving the changes and triggering an update event.
 */
export function scmCredentialUpdate1({ accountScmInfoUpdateRequest }, opts) {
    return oazapfts.ok(oazapfts.fetchJson("/v1/scm-credentials", oazapfts.json({
        ...opts,
        method: "PUT",
        body: accountScmInfoUpdateRequest
    })));
}
/**
 * Saves SCM credentials for a user's account.
 */
export function scmCredentialSave1({ accountScmInfoSaveRequest }, opts) {
    return oazapfts.ok(oazapfts.fetchJson("/v1/scm-credentials", oazapfts.json({
        ...opts,
        method: "POST",
        body: accountScmInfoSaveRequest
    })));
}
/**
 * Updates the role of an account.
 */
export function updateAccountRole({ roleId, updateAccountRoleRequest }, opts) {
    return oazapfts.ok(oazapfts.fetchJson(`/v1/roles/${encodeURIComponent(roleId)}`, oazapfts.json({
        ...opts,
        method: "PUT",
        body: updateAccountRoleRequest
    })));
}
/**
 * Deletes a specific account role by its ID.
 */
export function deleteAccountRole({ roleId }, opts) {
    return oazapfts.ok(oazapfts.fetchJson(`/v1/roles/${encodeURIComponent(roleId)}`, {
        ...opts,
        method: "DELETE"
    }));
}
/**
 * Retrieves a of resources and their associated actions with their status compared to a specified role.
 */
export function getResourcesAndActionsWithStatus({ roleId, size, page, sort, direction, search, filterMode, filterBy, filterValue, multiFilterMode, filterIn }, opts) {
    return oazapfts.ok(oazapfts.fetchJson(`/v1/roles/${encodeURIComponent(roleId)}/resources${QS.query(QS.explode({
        size,
        page,
        sort,
        direction,
        search,
        filterMode,
        filterBy,
        filterValue,
        multiFilterMode,
        filterIn
    }))}`, {
        ...opts
    }));
}
/**
 * Updates the role of an account with new resource-action statements.
 */
export function updateRoleWithNewActions({ roleId, body }, opts) {
    return oazapfts.ok(oazapfts.fetchJson(`/v1/roles/${encodeURIComponent(roleId)}/resources`, oazapfts.json({
        ...opts,
        method: "PUT",
        body
    })));
}
/**
 * Updates the actions associated with a specific resource in a role.
 */
export function updateResourceActions({ roleId, resourceId, updateResourceActionsRequest }, opts) {
    return oazapfts.ok(oazapfts.fetchJson(`/v1/roles/${encodeURIComponent(roleId)}/resources/${encodeURIComponent(resourceId)}`, oazapfts.json({
        ...opts,
        method: "PUT",
        body: updateResourceActionsRequest
    })));
}
/**
 * Retrieves account member details by their unique identifier.
 */
export function getMemberById({ memberId }, opts) {
    return oazapfts.ok(oazapfts.fetchJson(`/v1/members/${encodeURIComponent(memberId)}`, {
        ...opts
    }));
}
/**
 * Updates the information of an existing user.
 */
export function updateUser({ memberId, updateUserRequest }, opts) {
    return oazapfts.ok(oazapfts.fetchJson(`/v1/members/${encodeURIComponent(memberId)}`, oazapfts.json({
        ...opts,
        method: "PUT",
        body: updateUserRequest
    })));
}
/**
 * Updates the password of a specified user.
 */
export function updateUserPassword({ memberId, updatePasswordRequest }, opts) {
    return oazapfts.ok(oazapfts.fetchJson(`/v1/members/${encodeURIComponent(memberId)}/password`, oazapfts.json({
        ...opts,
        method: "PUT",
        body: updatePasswordRequest
    })));
}
/**
 * Updates the sales representatives associated with an active account.
 */
export function updateSalesRepresentatives({ accountId, updateSalesRepresentativesRequest }, opts) {
    return oazapfts.ok(oazapfts.fetchJson(`/v1/admin/accounts/${encodeURIComponent(accountId)}/sales-representatives`, oazapfts.json({
        ...opts,
        method: "PUT",
        body: updateSalesRepresentativesRequest
    })));
}
/**
 * Enables an account by updating its status.
 */
export function updateAccountStatus({ accountId, accountStatusRequest }, opts) {
    return oazapfts.ok(oazapfts.fetchJson(`/v1/accounts/${encodeURIComponent(accountId)}/status`, oazapfts.json({
        ...opts,
        method: "PUT",
        body: accountStatusRequest
    })));
}
/**
 * Retrieves detailed SSO information for a specified account.
 */
export function getAccountSso({ id }, opts) {
    return oazapfts.ok(oazapfts.fetchJson(`/v1/accounts/sso/${encodeURIComponent(id)}`, {
        ...opts
    }));
}
/**
 * Updates an existing SSO configuration for a specified ID.
 */
export function updateSso({ id, accountSsoUpdateRequest }, opts) {
    return oazapfts.ok(oazapfts.fetchJson(`/v1/accounts/sso/${encodeURIComponent(id)}`, oazapfts.json({
        ...opts,
        method: "PUT",
        body: accountSsoUpdateRequest
    })));
}
/**
 * Deletes a SSO configuration by its ID.
 */
export function deleteSso({ id }, opts) {
    return oazapfts.ok(oazapfts.fetchJson(`/v1/accounts/sso/${encodeURIComponent(id)}`, {
        ...opts,
        method: "DELETE"
    }));
}
/**
 * Update partially a configured SSO from account
 */
export function partialUpdateSso({ id, accountSsoPartialUpdateRequest }, opts) {
    return oazapfts.ok(oazapfts.fetchJson(`/v1/accounts/sso/${encodeURIComponent(id)}`, oazapfts.json({
        ...opts,
        method: "PATCH",
        body: accountSsoPartialUpdateRequest
    })));
}
/**
 * Retrieves the details of a SSO group mapping by its ID.
 */
export function getGroupMapById({ id, attributeImporterId }, opts) {
    return oazapfts.ok(oazapfts.fetchJson(`/v1/accounts/sso/${encodeURIComponent(id)}/group-mapping/${encodeURIComponent(attributeImporterId)}`, {
        ...opts
    }));
}
/**
 * Updates the SSO group mapping.
 */
export function updateGroupMapping({ id, attributeImporterId, accountGroupMappingUpdateRequest }, opts) {
    return oazapfts.ok(oazapfts.fetchJson(`/v1/accounts/sso/${encodeURIComponent(id)}/group-mapping/${encodeURIComponent(attributeImporterId)}`, oazapfts.json({
        ...opts,
        method: "PUT",
        body: accountGroupMappingUpdateRequest
    })));
}
/**
 * Deletes SSO Group Mapping.
 */
export function deleteGroupMapping({ id, attributeImporterId }, opts) {
    return oazapfts.ok(oazapfts.fetchJson(`/v1/accounts/sso/${encodeURIComponent(id)}/group-mapping/${encodeURIComponent(attributeImporterId)}`, {
        ...opts,
        method: "DELETE"
    }));
}
/**
 * Retrieves a list of service credentials associated with the current user
 */
export function getServiceCredentials({ size, page, sort, direction, search, filterMode, filterBy, filterValue, multiFilterMode, filterIn, name, status }, opts) {
    return oazapfts.ok(oazapfts.fetchJson(`/v2/service-credentials${QS.query(QS.explode({
        size,
        page,
        sort,
        direction,
        search,
        filterMode,
        filterBy,
        filterValue,
        multiFilterMode,
        filterIn,
        name,
        status
    }))}`, {
        ...opts
    }));
}
/**
 * Creates a new service credential for the current user.
 */
export function createServiceCredential({ serviceCredentialCreateRequestV2 }, opts) {
    return oazapfts.ok(oazapfts.fetchJson("/v2/service-credentials", oazapfts.json({
        ...opts,
        method: "POST",
        body: serviceCredentialCreateRequestV2
    })));
}
/**
 * Returns a paged list of Feature Flags associated with the current user's account.
 */
export function getFeatures({ featureFlagPageRequest }, opts) {
    return oazapfts.ok(oazapfts.fetchJson("/v2/feature-flags", oazapfts.json({
        ...opts,
        method: "POST",
        body: featureFlagPageRequest
    })));
}
/**
 * Returns a list of Feature Flags associated with the current user's account.
 */
export function getFeatures1(opts) {
    return oazapfts.ok(oazapfts.fetchJson("/v2/admin/feature-flags", {
        ...opts
    }));
}
/**
 * Creates a new Feature Flag and returns its unique identifier.
 */
export function create({ createFeatureFlagRequest }, opts) {
    return oazapfts.ok(oazapfts.fetchJson("/v2/admin/feature-flags", oazapfts.json({
        ...opts,
        method: "POST",
        body: createFeatureFlagRequest
    })));
}
/**
 * Returns a paged list of Feature Flags associated with the current user's account.
 */
export function getFeatures2({ accountId, featureFlagPageRequest }, opts) {
    return oazapfts.ok(oazapfts.fetchJson(`/v2/admin/feature-flags/account/${encodeURIComponent(accountId)}`, oazapfts.json({
        ...opts,
        method: "POST",
        body: featureFlagPageRequest
    })));
}
/**
 * Retrieves a page of external links for a specified account.
 */
export function getExternalLinks({ accountId, size, page, sort, direction, search, filterMode, filterBy, filterValue, multiFilterMode, filterIn }, opts) {
    return oazapfts.ok(oazapfts.fetchJson(`/v2/admin/accounts/${encodeURIComponent(accountId)}/external-links${QS.query(QS.explode({
        size,
        page,
        sort,
        direction,
        search,
        filterMode,
        filterBy,
        filterValue,
        multiFilterMode,
        filterIn
    }))}`, {
        ...opts
    }));
}
/**
 * Creates a new external link.
 */
export function createExternalLink({ accountId, externalLinkRequest }, opts) {
    return oazapfts.ok(oazapfts.fetchJson(`/v2/admin/accounts/${encodeURIComponent(accountId)}/external-links`, oazapfts.json({
        ...opts,
        method: "POST",
        body: externalLinkRequest
    })));
}
/**
 * Get Service Credentials
 */
export function getServiceCredentials1({ name, status, page, size }, opts) {
    return oazapfts.ok(oazapfts.fetchJson(`/v1/service-credentials${QS.query(QS.explode({
        name,
        status,
        page,
        size
    }))}`, {
        ...opts
    }));
}
/**
 * Create Service Credential
 */
export function createServiceCredential1({ serviceCredentialCreateRequest }, opts) {
    return oazapfts.ok(oazapfts.fetchJson("/v1/service-credentials", oazapfts.json({
        ...opts,
        method: "POST",
        body: serviceCredentialCreateRequest
    })));
}
/**
 * Notify about Service Credential expiration
 */
export function notifyServiceCredentialExpiration({ serviceCredentialExpirationRequest }, opts) {
    return oazapfts.ok(oazapfts.fetchJson("/v1/service-credentials/expiration", oazapfts.json({
        ...opts,
        method: "POST",
        body: serviceCredentialExpirationRequest
    })));
}
/**
 * Find secrets with multiple filters and conditions
 */
export function findSecrets({ size, page, sort, direction, search, filterByScope, filterByType, filterByScopedBy, filterByScopedValue, authorizedOnly }, opts) {
    return oazapfts.ok(oazapfts.fetchJson(`/v1/secrets${QS.query(QS.explode({
        size,
        page,
        sort,
        direction,
        search,
        filterByScope,
        filterByType,
        filterByScopedBy,
        filterByScopedValue,
        authorizedOnly
    }))}`, {
        ...opts
    }));
}
/**
 * Create a new secret
 */
export function createSecret({ createSecretRequest }, opts) {
    return oazapfts.ok(oazapfts.fetchJson("/v1/secrets", oazapfts.json({
        ...opts,
        method: "POST",
        body: createSecretRequest
    })));
}
/**
 * Associate a secret with a resource
 */
export function associateResource({ secretId, resourceSlug, resourceTypeSlug }, opts) {
    return oazapfts.ok(oazapfts.fetchText(`/v1/secrets/${encodeURIComponent(secretId)}/associations/resources/${encodeURIComponent(resourceSlug)}/resource-type/${encodeURIComponent(resourceTypeSlug)}`, {
        ...opts,
        method: "POST"
    }));
}
/**
 * Disassociate a secret with a resource
 */
export function disassociateResource({ secretId, resourceSlug, resourceTypeSlug }, opts) {
    return oazapfts.ok(oazapfts.fetchText(`/v1/secrets/${encodeURIComponent(secretId)}/associations/resources/${encodeURIComponent(resourceSlug)}/resource-type/${encodeURIComponent(resourceTypeSlug)}`, {
        ...opts,
        method: "DELETE"
    }));
}
/**
 * Check if user has access to secret
 */
export function findSecrets1({ checkSecretAccessRequest }, opts) {
    return oazapfts.ok(oazapfts.fetchJson("/v1/secrets/check-access", oazapfts.json({
        ...opts,
        method: "POST",
        body: checkSecretAccessRequest
    })));
}
/**
 * Retrieves a paginated list of scope details for the current user's tenant.
 */
export function findScopes({ page }, opts) {
    return oazapfts.ok(oazapfts.fetchJson(`/v1/scopes${QS.query(QS.explode({
        page
    }))}`, {
        ...opts
    }));
}
/**
 * Creates a new IAM scope for the current user's tenant using the provided request details.
 */
export function createScope({ iamCreateScopeRequest }, opts) {
    return oazapfts.ok(oazapfts.fetchJson("/v1/scopes", oazapfts.json({
        ...opts,
        method: "POST",
        body: iamCreateScopeRequest
    })));
}
/**
 * Retrieves a list of roles available to the current account.
 */
export function getRoles({ size, page, sort, direction, search, filterMode, filterBy, filterValue, multiFilterMode, filterIn }, opts) {
    return oazapfts.ok(oazapfts.fetchJson(`/v1/roles${QS.query(QS.explode({
        size,
        page,
        sort,
        direction,
        search,
        filterMode,
        filterBy,
        filterValue,
        multiFilterMode,
        filterIn
    }))}`, {
        ...opts
    }));
}
/**
 * Creates a new account role.
 */
export function createAccountRole({ createAccountRoleRequest }, opts) {
    return oazapfts.ok(oazapfts.fetchJson("/v1/roles", oazapfts.json({
        ...opts,
        method: "POST",
        body: createAccountRoleRequest
    })));
}
/**
 * Retrieves a list of role members for a specified role.
 */
export function getRoleMembers({ roleId, size, page, sort, direction, search, filterMode, filterBy, filterValue, multiFilterMode, filterIn }, opts) {
    return oazapfts.ok(oazapfts.fetchJson(`/v1/roles/${encodeURIComponent(roleId)}/members${QS.query(QS.explode({
        size,
        page,
        sort,
        direction,
        search,
        filterMode,
        filterBy,
        filterValue,
        multiFilterMode,
        filterIn
    }))}`, {
        ...opts
    }));
}
/**
 * Assigns a specified role to a list of members.
 */
export function addRoleToMember({ roleId, addRoleToMemberRequest }, opts) {
    return oazapfts.ok(oazapfts.fetchJson(`/v1/roles/${encodeURIComponent(roleId)}/members`, oazapfts.json({
        ...opts,
        method: "POST",
        body: addRoleToMemberRequest
    })));
}
/**
 * Retrieves a list of role groups associated with a specific group ID.
 */
export function getRoleGroups({ roleId, size, page, sort, direction, search, filterMode, filterBy, filterValue, multiFilterMode, filterIn }, opts) {
    return oazapfts.ok(oazapfts.fetchJson(`/v1/roles/${encodeURIComponent(roleId)}/groups${QS.query(QS.explode({
        size,
        page,
        sort,
        direction,
        search,
        filterMode,
        filterBy,
        filterValue,
        multiFilterMode,
        filterIn
    }))}`, {
        ...opts
    }));
}
/**
 * Bind a role to a group.
 */
export function bindRoleGroups({ roleId, roleGroupIdsRequest }, opts) {
    return oazapfts.ok(oazapfts.fetchJson(`/v1/roles/${encodeURIComponent(roleId)}/groups`, oazapfts.json({
        ...opts,
        method: "POST",
        body: roleGroupIdsRequest
    })));
}
/**
 * Retrieves a list of resources.
 */
export function getResources({ size, page, sort, direction, search, filterMode, filterBy, filterValue, multiFilterMode, filterIn }, opts) {
    return oazapfts.ok(oazapfts.fetchJson(`/v1/resources${QS.query(QS.explode({
        size,
        page,
        sort,
        direction,
        search,
        filterMode,
        filterBy,
        filterValue,
        multiFilterMode,
        filterIn
    }))}`, {
        ...opts
    }));
}
/**
 * Creates a new resource for the account.
 */
export function createResource({ createResourceRequest }, opts) {
    return oazapfts.ok(oazapfts.fetchJson("/v1/resources", oazapfts.json({
        ...opts,
        method: "POST",
        body: createResourceRequest
    })));
}
/**
 * Retrieves a list of account members for a specified account.
 */
export function getAccountMembers({ size, page, sort, direction, search, filterMode, filterBy, filterValue, multiFilterMode, filterIn }, opts) {
    return oazapfts.ok(oazapfts.fetchJson(`/v1/members${QS.query(QS.explode({
        size,
        page,
        sort,
        direction,
        search,
        filterMode,
        filterBy,
        filterValue,
        multiFilterMode,
        filterIn
    }))}`, {
        ...opts
    }));
}
/**
 * Creates a new user in the account.
 */
export function createUser({ createUserRequest }, opts) {
    return oazapfts.ok(oazapfts.fetchJson("/v1/members", oazapfts.json({
        ...opts,
        method: "POST",
        body: createUserRequest
    })));
}
/**
 * Retrieves a paginated list of roles associated with a specific member.
 */
export function getRoles1({ memberId, size, page, sort, direction, search, filterMode, filterBy, filterValue, multiFilterMode, filterIn }, opts) {
    return oazapfts.ok(oazapfts.fetchJson(`/v1/members/${encodeURIComponent(memberId)}/roles${QS.query(QS.explode({
        size,
        page,
        sort,
        direction,
        search,
        filterMode,
        filterBy,
        filterValue,
        multiFilterMode,
        filterIn
    }))}`, {
        ...opts
    }));
}
/**
 * Binds a member to a list of roles.
 */
export function bindToRoles({ memberId, memberIdsCommonRequest }, opts) {
    return oazapfts.ok(oazapfts.fetchJson(`/v1/members/${encodeURIComponent(memberId)}/roles`, oazapfts.json({
        ...opts,
        method: "POST",
        body: memberIdsCommonRequest
    })));
}
/**
 * Resets the OTP configuration for a specified member.
 */
export function resetOtp({ memberId }, opts) {
    return oazapfts.ok(oazapfts.fetchJson(`/v1/members/${encodeURIComponent(memberId)}/reset-otp`, {
        ...opts,
        method: "POST"
    }));
}
/**
 * Retrieves a list of groups associated with a specific member.
 */
export function getMemberGroups({ memberId, size, page, sort, direction, search, filterMode, filterBy, filterValue, multiFilterMode, filterIn }, opts) {
    return oazapfts.ok(oazapfts.fetchJson(`/v1/members/${encodeURIComponent(memberId)}/groups${QS.query(QS.explode({
        size,
        page,
        sort,
        direction,
        search,
        filterMode,
        filterBy,
        filterValue,
        multiFilterMode,
        filterIn
    }))}`, {
        ...opts
    }));
}
/**
 * Binds a member to a list of groups.
 */
export function bindToGroups({ memberId, memberIdsCommonRequest }, opts) {
    return oazapfts.ok(oazapfts.fetchJson(`/v1/members/${encodeURIComponent(memberId)}/groups`, oazapfts.json({
        ...opts,
        method: "POST",
        body: memberIdsCommonRequest
    })));
}
/**
 * Retrieves a magic link to create a new FIDO key.
 */
export function createFidoMagicLink({ memberId, magicLinkRequest }, opts) {
    return oazapfts.ok(oazapfts.fetchJson(`/v1/members/${encodeURIComponent(memberId)}/fido-credentials/magic-link`, oazapfts.json({
        ...opts,
        method: "POST",
        body: magicLinkRequest
    })));
}
/**
 * Get account member favorite resources
 */
export function listAccountMemberFavorites({ memberId }, opts) {
    return oazapfts.ok(oazapfts.fetchJson(`/v1/members/${encodeURIComponent(memberId)}/favorites`, {
        ...opts
    }));
}
/**
 * Add account member favorite resource
 */
export function addAccountMemberFavorite({ memberId, createFavoriteResourceRequest }, opts) {
    return oazapfts.ok(oazapfts.fetchJson(`/v1/members/${encodeURIComponent(memberId)}/favorites`, oazapfts.json({
        ...opts,
        method: "POST",
        body: createFavoriteResourceRequest
    })));
}
/**
 * Resets the password for a user by generating a magic link and sending it to their email.
 */
export function resetPassword({ memberEmailRequest }, opts) {
    return oazapfts.ok(oazapfts.fetchJson("/v1/members/reset-password", oazapfts.json({
        ...opts,
        method: "POST",
        body: memberEmailRequest
    })));
}
/**
 * Create a personal lead contact
 */
export function personalContact({ personalLeadRequest }, opts) {
    return oazapfts.ok(oazapfts.fetchJson("/v1/leads/personal", oazapfts.json({
        ...opts,
        method: "POST",
        body: personalLeadRequest
    })));
}
/**
 * Create an enterprise lead contact
 */
export function enterpriseContact({ enterpriseLeadRequest }, opts) {
    return oazapfts.ok(oazapfts.fetchJson("/v1/leads/enterprise", oazapfts.json({
        ...opts,
        method: "POST",
        body: enterpriseLeadRequest
    })));
}
/**
 * Get Groups
 */
export function getGroups({ size, page, sort, direction, search, filterMode, filterBy, filterValue, multiFilterMode, filterIn, includeDefaultGroup }, opts) {
    return oazapfts.ok(oazapfts.fetchJson(`/v1/groups${QS.query(QS.explode({
        size,
        page,
        sort,
        direction,
        search,
        filterMode,
        filterBy,
        filterValue,
        multiFilterMode,
        filterIn,
        includeDefaultGroup
    }))}`, {
        ...opts
    }));
}
/**
 * Create Group
 */
export function save({ newGroupRequest }, opts) {
    return oazapfts.ok(oazapfts.fetchJson("/v1/groups", oazapfts.json({
        ...opts,
        method: "POST",
        body: newGroupRequest
    })));
}
/**
 * Get Group members
 */
export function getRoles2({ groupId, size, page, sort, direction, search, filterMode, filterBy, filterValue, multiFilterMode, filterIn }, opts) {
    return oazapfts.ok(oazapfts.fetchJson(`/v1/groups/${encodeURIComponent(groupId)}/roles${QS.query(QS.explode({
        size,
        page,
        sort,
        direction,
        search,
        filterMode,
        filterBy,
        filterValue,
        multiFilterMode,
        filterIn
    }))}`, {
        ...opts
    }));
}
/**
 * Bind Group roles
 */
export function bindRoles({ groupId, groupRoleIdsRequest }, opts) {
    return oazapfts.ok(oazapfts.fetchJson(`/v1/groups/${encodeURIComponent(groupId)}/roles`, oazapfts.json({
        ...opts,
        method: "POST",
        body: groupRoleIdsRequest
    })));
}
/**
 * Get Group Resources
 */
export function getGroupResources({ groupId, size, page, sort, direction, search, filterMode, filterBy, filterValue, multiFilterMode, filterIn }, opts) {
    return oazapfts.ok(oazapfts.fetchJson(`/v1/groups/${encodeURIComponent(groupId)}/resources${QS.query(QS.explode({
        size,
        page,
        sort,
        direction,
        search,
        filterMode,
        filterBy,
        filterValue,
        multiFilterMode,
        filterIn
    }))}`, {
        ...opts
    }));
}
/**
 * Add Multiple Resources to Group
 */
export function addResourcesToGroup({ groupId, addResourcesToGroupRequest }, opts) {
    return oazapfts.ok(oazapfts.fetchJson(`/v1/groups/${encodeURIComponent(groupId)}/resources`, oazapfts.json({
        ...opts,
        method: "POST",
        body: addResourcesToGroupRequest
    })));
}
/**
 * Get Group members
 */
export function getMembers({ groupId, size, page, sort, direction, search, filterMode, filterBy, filterValue, multiFilterMode, filterIn }, opts) {
    return oazapfts.ok(oazapfts.fetchJson(`/v1/groups/${encodeURIComponent(groupId)}/members${QS.query(QS.explode({
        size,
        page,
        sort,
        direction,
        search,
        filterMode,
        filterBy,
        filterValue,
        multiFilterMode,
        filterIn
    }))}`, {
        ...opts
    }));
}
/**
 * Bind Group member
 */
export function bindGroupMembers({ groupId, groupMemberIdsRequest }, opts) {
    return oazapfts.ok(oazapfts.fetchJson(`/v1/groups/${encodeURIComponent(groupId)}/members`, oazapfts.json({
        ...opts,
        method: "POST",
        body: groupMemberIdsRequest
    })));
}
/**
 * List all Account Extensions
 */
export function listExtensions(opts) {
    return oazapfts.ok(oazapfts.fetchJson("/v1/extensions", {
        ...opts
    }));
}
/**
 * Create Account Extension
 */
export function createExtension({ extensionCreateRequest }, opts) {
    return oazapfts.ok(oazapfts.fetchJson("/v1/extensions", oazapfts.json({
        ...opts,
        method: "POST",
        body: extensionCreateRequest
    })));
}
/**
 * List all Account Extension Versions
 */
export function listExtensionVersions({ slug }, opts) {
    return oazapfts.ok(oazapfts.fetchJson(`/v1/extensions/${encodeURIComponent(slug)}/versions`, {
        ...opts
    }));
}
/**
 * Create Account Extension Version
 */
export function createExtensionVersion({ slug, name, body }, opts) {
    return oazapfts.ok(oazapfts.fetchJson(`/v1/extensions/${encodeURIComponent(slug)}/versions${QS.query(QS.explode({
        name
    }))}`, oazapfts.multipart({
        ...opts,
        method: "POST",
        body
    })));
}
/**
 * List all Account Extension Links
 */
export function listExtensionLinks({ workspaceId, $type, extensionSlug, showOnlyActive }, opts) {
    return oazapfts.ok(oazapfts.fetchJson(`/v1/extension-links${QS.query(QS.explode({
        workspaceId,
        "type": $type,
        extensionSlug,
        showOnlyActive
    }))}`, {
        ...opts
    }));
}
/**
 * Create Account Extension Link
 */
export function createExtensionLink({ extensionLinkCreateRequest }, opts) {
    return oazapfts.ok(oazapfts.fetchJson("/v1/extension-links", oazapfts.json({
        ...opts,
        method: "POST",
        body: extensionLinkCreateRequest
    })));
}
/**
 * Receive engagement form webhook
 */
export function receiveWebhook({ provider, body }, opts) {
    return oazapfts.ok(oazapfts.fetchText(`/v1/engagement/forms/webhook/${encodeURIComponent(provider)}`, oazapfts.json({
        ...opts,
        method: "POST",
        body
    })));
}
/**
 * Check if user has answered forms
 */
export function hasUserAnsweredForm({ hasUserAnsweredFormRequest }, opts) {
    return oazapfts.ok(oazapfts.fetchJson("/v1/engagement/forms/answered", oazapfts.json({
        ...opts,
        method: "POST",
        body: hasUserAnsweredFormRequest
    })));
}
/**
 * Send Welcome Email
 */
export function sendWelcomeEmail({ welcomeEmailRequest }, opts) {
    return oazapfts.ok(oazapfts.fetchJson("/v1/emails/welcome", oazapfts.json({
        ...opts,
        method: "POST",
        body: welcomeEmailRequest
    })));
}
/**
 * Send Download Email
 */
export function sendDownloadEmail(opts) {
    return oazapfts.ok(oazapfts.fetchJson("/v1/emails/download", {
        ...opts,
        method: "POST"
    }));
}
/**
 * Send contact email
 */
export function sendContactEmail({ contactEmailRequest }, opts) {
    return oazapfts.ok(oazapfts.fetchJson("/v1/emails/contact", oazapfts.json({
        ...opts,
        method: "POST",
        body: contactEmailRequest
    })));
}
/**
 * Validates a campaign code and returns true if exists and its enabled
 */
export function validateCampaignCode({ code }, opts) {
    return oazapfts.ok(oazapfts.fetchJson(`/v1/campaigns/${encodeURIComponent(code)}/available`, {
        ...opts,
        method: "POST"
    }));
}
/**
 * Get credentials from personal service client (create if not exists one to the logged user).
 */
export function getPersonalClientCredentials(opts) {
    return oazapfts.ok(oazapfts.fetchJson("/v1/authentication/personal-client-credentials", {
        ...opts,
        method: "POST"
    }));
}
/**
 * Get an Access Token generated With Audience on logged user behalf, since previously authorized.
 */
export function getPersonalAccessToken({ apiAuthorization, generateTokenFromPersonalAccountRequest }, opts) {
    return oazapfts.ok(oazapfts.fetchJson("/v1/authentication/personal-access-token", oazapfts.json({
        ...opts,
        method: "POST",
        body: generateTokenFromPersonalAccountRequest,
        headers: oazapfts.mergeHeaders(opts?.headers, {
            "Api-Authorization": apiAuthorization
        })
    })));
}
/**
 * Disable client that generates Personal Access Tokens.
 */
export function disablePersonalAccessTokenGeneration(opts) {
    return oazapfts.ok(oazapfts.fetchJson("/v1/authentication/personal-access-token/unauthorize", {
        ...opts,
        method: "POST"
    }));
}
/**
 * Verify if exists a client created (authorized) to generate Access Token on logged user behalf.
 */
export function personalAccessTokenAuthorization(opts) {
    return oazapfts.ok(oazapfts.fetchJson("/v1/authentication/personal-access-token/authorize", {
        ...opts
    }));
}
/**
 * Authorize creating a client to generate Personal Access Tokens on your behalf.
 */
export function createPersonalAccessToken(opts) {
    return oazapfts.ok(oazapfts.fetchJson("/v1/authentication/personal-access-token/authorize", {
        ...opts,
        method: "POST"
    }));
}
/**
 * Get an Access Token generated on logged user's service account behalf, since previously authorized.
 */
export function getPersonalAccessTokenFromServiceAccountOnly(opts) {
    return oazapfts.ok(oazapfts.fetchJson("/v1/authentication/personal-access-token-sa", {
        ...opts,
        method: "POST"
    }));
}
/**
 * Generate device code.
 */
export function deviceCode({ email, freemiumAccess, codeChallenge, body }, opts) {
    return oazapfts.ok(oazapfts.fetchJson(`/v1/authentication/device-code${QS.query(QS.explode({
        email,
        freemium_access: freemiumAccess,
        code_challenge: codeChallenge
    }))}`, oazapfts.form({
        ...opts,
        method: "POST",
        body
    })));
}
/**
 * Allows to enable or disable trial SSO integrations.
 */
export function updateTrialSsoSettings({ updateTrialSsoSettingsRequest }, opts) {
    return oazapfts.ok(oazapfts.fetchJson("/v1/admin/config/trial/sso", oazapfts.json({
        ...opts,
        method: "POST",
        body: updateTrialSsoSettingsRequest
    })));
}
/**
 * Gets application settings related to secrets.
 */
export function getSecretSettings({ xAccountId }, opts) {
    return oazapfts.ok(oazapfts.fetchJson("/v1/admin/config/secrets", {
        ...opts,
        headers: oazapfts.mergeHeaders(opts?.headers, {
            "x-account-id": xAccountId
        })
    }));
}
/**
 * Manages application settings related to secrets.
 */
export function updateSecretSettings({ xAccountId, updateSecretsSettingsRequest }, opts) {
    return oazapfts.ok(oazapfts.fetchJson("/v1/admin/config/secrets", oazapfts.json({
        ...opts,
        method: "POST",
        body: updateSecretsSettingsRequest,
        headers: oazapfts.mergeHeaders(opts?.headers, {
            "x-account-id": xAccountId
        })
    })));
}
/**
 * Allows to manage catalog integration.
 */
export function updateCatalogSettings({ updateCatalogSettingsRequest }, opts) {
    return oazapfts.ok(oazapfts.fetchJson("/v1/admin/config/catalog", oazapfts.json({
        ...opts,
        method: "POST",
        body: updateCatalogSettingsRequest
    })));
}
/**
 * Gets the Vision One integration parameters for new accounts.
 */
export function getVisionOne(opts) {
    return oazapfts.ok(oazapfts.fetchJson("/v1/admin/config/account-order/vision-one", {
        ...opts
    }));
}
/**
 * Changes the Vision One integration parameters for new accounts.
 */
export function updateVisionOne({ updateVisionOneIntegrationRequest }, opts) {
    return oazapfts.ok(oazapfts.fetchJson("/v1/admin/config/account-order/vision-one", oazapfts.json({
        ...opts,
        method: "POST",
        body: updateVisionOneIntegrationRequest
    })));
}
/**
 * Finds and returns a list of accounts filtered by account type.
 */
export function getAccounts({ size, page, sort, direction, search, filterMode, filterBy, filterValue, multiFilterMode, filterIn }, opts) {
    return oazapfts.ok(oazapfts.fetchJson(`/v1/admin/accounts${QS.query(QS.explode({
        size,
        page,
        sort,
        direction,
        search,
        filterMode,
        filterBy,
        filterValue,
        multiFilterMode,
        filterIn
    }))}`, {
        ...opts
    }));
}
/**
 * Creates a new account order and returns the associated account ID.
 */
export function create1({ createAccountOrderRequest }, opts) {
    return oazapfts.ok(oazapfts.fetchJson("/v1/admin/accounts", oazapfts.json({
        ...opts,
        method: "POST",
        body: createAccountOrderRequest
    })));
}
/**
 * Creates a Trial account.
 */
export function createTrialAccount({ accountTrialCreateRequest }, opts) {
    return oazapfts.ok(oazapfts.fetchJson("/v1/accounts/trial", oazapfts.json({
        ...opts,
        method: "POST",
        body: accountTrialCreateRequest
    })));
}
/**
 * Sets trial account to be removed after expiration.
 */
export function removeTrialAccount(opts) {
    return oazapfts.ok(oazapfts.fetchJson("/v1/accounts/trial", {
        ...opts,
        method: "DELETE"
    }));
}
/**
 * Retrieves a list of SSO group mappings.
 */
export function getAllGroupMapping({ id, size, page, sort, direction, search, filterMode, filterBy, filterValue, multiFilterMode, filterIn }, opts) {
    return oazapfts.ok(oazapfts.fetchJson(`/v1/accounts/sso/${encodeURIComponent(id)}/group-mapping${QS.query(QS.explode({
        size,
        page,
        sort,
        direction,
        search,
        filterMode,
        filterBy,
        filterValue,
        multiFilterMode,
        filterIn
    }))}`, {
        ...opts
    }));
}
/**
 * Creates a new SSO group mapping.
 */
export function createGroupMapping({ id, accountGroupMappingCreateRequest }, opts) {
    return oazapfts.ok(oazapfts.fetchJson(`/v1/accounts/sso/${encodeURIComponent(id)}/group-mapping`, oazapfts.json({
        ...opts,
        method: "POST",
        body: accountGroupMappingCreateRequest
    })));
}
/**
 * Configures new SSO in account.
 */
export function ssoConfigure({ protocol, accountSsoCreateRequest }, opts) {
    return oazapfts.ok(oazapfts.fetchJson(`/v1/accounts/sso/protocol/${encodeURIComponent(protocol)}`, oazapfts.json({
        ...opts,
        method: "POST",
        body: accountSsoCreateRequest
    })));
}
/**
 * Parses an SSO configuration file and returns the parsed data with a callback URL.
 */
export function ssoParseConfigurationFile({ protocol, body }, opts) {
    return oazapfts.ok(oazapfts.fetchJson(`/v1/accounts/sso/protocol/${encodeURIComponent(protocol)}/parser`, oazapfts.multipart({
        ...opts,
        method: "POST",
        body
    })));
}
/**
 * Create an Account Order for Partner
 */
export function createPartner({ createAccountPartnerOrderRequest }, opts) {
    return oazapfts.ok(oazapfts.fetchJson("/v1/account-orders/partners", oazapfts.json({
        ...opts,
        method: "POST",
        body: createAccountPartnerOrderRequest
    })));
}
/**
 * Validate new Partner account data
 */
export function validateNewPartnerData({ validateAccountPartnerDataRequest }, opts) {
    return oazapfts.ok(oazapfts.fetchJson("/v1/account-orders/partners/validations/creation-data", oazapfts.json({
        ...opts,
        method: "POST",
        body: validateAccountPartnerDataRequest
    })));
}
/**
 * Rotates the secret of a service credential.
 */
export function rotateServiceCredentialSecret({ id, rotateServiceCredentialSecretRequest }, opts) {
    return oazapfts.ok(oazapfts.fetchJson(`/v2/service-credentials/${encodeURIComponent(id)}/secret`, oazapfts.json({
        ...opts,
        method: "PATCH",
        body: rotateServiceCredentialSecretRequest
    })));
}
/**
 * Updates the scopes of a service credential
 */
export function updateServiceCredentialScopes({ id, updateServiceCredentialScopesRequest }, opts) {
    return oazapfts.ok(oazapfts.fetchJson(`/v2/service-credentials/${encodeURIComponent(id)}/scopes`, oazapfts.json({
        ...opts,
        method: "PATCH",
        body: updateServiceCredentialScopesRequest
    })));
}
/**
 * Revokes a service account for the current user.
 */
export function revokeServiceCredential({ id }, opts) {
    return oazapfts.ok(oazapfts.fetchJson(`/v2/service-credentials/${encodeURIComponent(id)}/revoke`, {
        ...opts,
        method: "PATCH"
    }));
}
/**
 * Delete a member permissions
 */
export function deleteMemberPermissions({ memberId, body }, opts) {
    return oazapfts.ok(oazapfts.fetchJson(`/v2/members/${encodeURIComponent(memberId)}/permissions`, oazapfts.json({
        ...opts,
        method: "DELETE",
        body
    })));
}
/**
 * Update a member's permissions by assigning resources and actions, creating a default role if necessary.
 */
export function updateMemberPermissions({ memberId, body }, opts) {
    return oazapfts.ok(oazapfts.fetchJson(`/v2/members/${encodeURIComponent(memberId)}/permissions`, oazapfts.json({
        ...opts,
        method: "PATCH",
        body
    })));
}
/**
 * Delete Group permissions
 */
export function deleteGroupPermissions({ groupId, body }, opts) {
    return oazapfts.ok(oazapfts.fetchJson(`/v2/groups/${encodeURIComponent(groupId)}/permissions`, oazapfts.json({
        ...opts,
        method: "DELETE",
        body
    })));
}
/**
 * Update Group permissions
 */
export function updateGroupPermissions({ groupId, body }, opts) {
    return oazapfts.ok(oazapfts.fetchJson(`/v2/groups/${encodeURIComponent(groupId)}/permissions`, oazapfts.json({
        ...opts,
        method: "PATCH",
        body
    })));
}
/**
 * Retrieves an external link by account.
 */
export function getExternalLinkByIdAndAccountId({ externalLinkId, accountId }, opts) {
    return oazapfts.ok(oazapfts.fetchJson(`/v2/admin/accounts/${encodeURIComponent(accountId)}/external-links/${encodeURIComponent(externalLinkId)}`, {
        ...opts
    }));
}
/**
 * Deletes an external link by its unique identifier.
 */
export function deleteExternalLink({ accountId, externalLinkId }, opts) {
    return oazapfts.ok(oazapfts.fetchJson(`/v2/admin/accounts/${encodeURIComponent(accountId)}/external-links/${encodeURIComponent(externalLinkId)}`, {
        ...opts,
        method: "DELETE"
    }));
}
/**
 * Updates an external link by its unique identifier.
 */
export function updateExternalLink({ externalLinkId, accountId, externalLinkRequest }, opts) {
    return oazapfts.ok(oazapfts.fetchJson(`/v2/admin/accounts/${encodeURIComponent(accountId)}/external-links/${encodeURIComponent(externalLinkId)}`, oazapfts.json({
        ...opts,
        method: "PATCH",
        body: externalLinkRequest
    })));
}
/**
 * Associate Group to Service Credential
 */
export function associateGroupToServiceCredential({ id, serviceCredentialAssociateGroupRequest }, opts) {
    return oazapfts.ok(oazapfts.fetchJson(`/v1/service-credentials/${encodeURIComponent(id)}/teams`, oazapfts.json({
        ...opts,
        method: "PATCH",
        body: serviceCredentialAssociateGroupRequest
    })));
}
/**
 * Revoke Service Credential
 */
export function revokeServiceCredential1({ id }, opts) {
    return oazapfts.ok(oazapfts.fetchJson(`/v1/service-credentials/${encodeURIComponent(id)}/revocation`, {
        ...opts,
        method: "PATCH"
    }));
}
/**
 * Enable secret by ID
 */
export function enableSecret({ secretId }, opts) {
    return oazapfts.ok(oazapfts.fetchText(`/v1/secrets/${encodeURIComponent(secretId)}/enable`, {
        ...opts,
        method: "PATCH"
    }));
}
/**
 * Disable secret by ID
 */
export function disableSecret({ secretId }, opts) {
    return oazapfts.ok(oazapfts.fetchText(`/v1/secrets/${encodeURIComponent(secretId)}/disable`, {
        ...opts,
        method: "PATCH"
    }));
}
/**
 * Cancel Secret delete
 */
export function cancelSecretDelete({ secretId }, opts) {
    return oazapfts.ok(oazapfts.fetchText(`/v1/secrets/${encodeURIComponent(secretId)}/cancel-delete`, {
        ...opts,
        method: "PATCH"
    }));
}
/**
 * Migrate Api Key Secrets to new version
 */
export function migrateApiKeySecretsToV2(opts) {
    return oazapfts.ok(oazapfts.fetchText("/v1/secrets/migrate-api-key-secrets-to-v2", {
        ...opts,
        method: "PATCH"
    }));
}
/**
 * Retrieves the preferences of a specific member.
 */
export function getMemberPreferences({ memberId }, opts) {
    return oazapfts.ok(oazapfts.fetchJson(`/v1/members/${encodeURIComponent(memberId)}/preferences`, {
        ...opts
    }));
}
/**
 * Updates the preferences of a specific member.
 */
export function updateMemberPreferences({ memberId, updateMemberPreferencesRequest }, opts) {
    return oazapfts.ok(oazapfts.fetchJson(`/v1/members/${encodeURIComponent(memberId)}/preferences`, oazapfts.json({
        ...opts,
        method: "PATCH",
        body: updateMemberPreferencesRequest
    })));
}
/**
 * Deletes specific permissions from a member's role.
 */
export function deleteMemberPermissions1({ memberId, deleteMemberPermissionsRequest }, opts) {
    return oazapfts.ok(oazapfts.fetchJson(`/v1/members/${encodeURIComponent(memberId)}/permissions`, oazapfts.json({
        ...opts,
        method: "DELETE",
        body: deleteMemberPermissionsRequest
    })));
}
/**
 * Update a member's permissions by assigning resources and actions, creating a default role if necessary.
 */
export function updateMemberPermissions1({ memberId, updateMemberPermissionsRequest }, opts) {
    return oazapfts.ok(oazapfts.fetchJson(`/v1/members/${encodeURIComponent(memberId)}/permissions`, oazapfts.json({
        ...opts,
        method: "PATCH",
        body: updateMemberPermissionsRequest
    })));
}
/**
 * Retrieves a paginated list of Fido user credentials for a specific member.
 */
export function getAllMemberFidoCredentials({ memberId, size, page, sort, direction, search, filterMode, filterBy, filterValue, multiFilterMode, filterIn }, opts) {
    return oazapfts.ok(oazapfts.fetchJson(`/v1/members/${encodeURIComponent(memberId)}/fido-credentials${QS.query(QS.explode({
        size,
        page,
        sort,
        direction,
        search,
        filterMode,
        filterBy,
        filterValue,
        multiFilterMode,
        filterIn
    }))}`, {
        ...opts
    }));
}
/**
 * Deactivates all Fido credentials associated with a specific member.
 */
export function deactivateFidoCredentials({ memberId }, opts) {
    return oazapfts.ok(oazapfts.fetchJson(`/v1/members/${encodeURIComponent(memberId)}/fido-credentials`, {
        ...opts,
        method: "DELETE"
    }));
}
/**
 * Enables Fido credentials for a specified member.
 */
export function enableFidoCredentials({ memberId }, opts) {
    return oazapfts.ok(oazapfts.fetchJson(`/v1/members/${encodeURIComponent(memberId)}/fido-credentials`, {
        ...opts,
        method: "PATCH"
    }));
}
/**
 * Get Group by ID
 */
export function getGroupById({ groupId }, opts) {
    return oazapfts.ok(oazapfts.fetchJson(`/v1/groups/${encodeURIComponent(groupId)}`, {
        ...opts
    }));
}
/**
 * Delete Group
 */
export function delete1({ groupId }, opts) {
    return oazapfts.ok(oazapfts.fetchJson(`/v1/groups/${encodeURIComponent(groupId)}`, {
        ...opts,
        method: "DELETE"
    }));
}
/**
 * Update Group
 */
export function update1({ groupId, updateGroupRequest }, opts) {
    return oazapfts.ok(oazapfts.fetchJson(`/v1/groups/${encodeURIComponent(groupId)}`, oazapfts.json({
        ...opts,
        method: "PATCH",
        body: updateGroupRequest
    })));
}
/**
 * Delete Group permissions
 */
export function deleteGroupPermissions1({ groupId, deleteGroupPermissionsRequest }, opts) {
    return oazapfts.ok(oazapfts.fetchJson(`/v1/groups/${encodeURIComponent(groupId)}/permissions`, oazapfts.json({
        ...opts,
        method: "DELETE",
        body: deleteGroupPermissionsRequest
    })));
}
/**
 * Update Group permissions
 */
export function updateGroupPermissions1({ groupId, updateGroupPermissionsRequest }, opts) {
    return oazapfts.ok(oazapfts.fetchJson(`/v1/groups/${encodeURIComponent(groupId)}/permissions`, oazapfts.json({
        ...opts,
        method: "PATCH",
        body: updateGroupPermissionsRequest
    })));
}
/**
 * Delete Account Extension
 */
export function deleteExtension({ slug }, opts) {
    return oazapfts.ok(oazapfts.fetchJson(`/v1/extensions/${encodeURIComponent(slug)}`, {
        ...opts,
        method: "DELETE"
    }));
}
/**
 * Update Account Extension
 */
export function updateExtension({ slug, extensionUpdateRequest }, opts) {
    return oazapfts.ok(oazapfts.fetchJson(`/v1/extensions/${encodeURIComponent(slug)}`, oazapfts.json({
        ...opts,
        method: "PATCH",
        body: extensionUpdateRequest
    })));
}
/**
 * Delete Account Extension Link
 */
export function deleteExtensionLink({ id }, opts) {
    return oazapfts.ok(oazapfts.fetchJson(`/v1/extension-links/${encodeURIComponent(id)}`, {
        ...opts,
        method: "DELETE"
    }));
}
/**
 * Update Account Extension Link
 */
export function updateExtensionLink({ id, extensionLinkUpdateRequest }, opts) {
    return oazapfts.ok(oazapfts.fetchJson(`/v1/extension-links/${encodeURIComponent(id)}`, oazapfts.json({
        ...opts,
        method: "PATCH",
        body: extensionLinkUpdateRequest
    })));
}
/**
 * Adds a new email domain to an existing account.
 */
export function addNewEmailDomain({ accountId, addNewEmailDomainRequest }, opts) {
    return oazapfts.ok(oazapfts.fetchJson(`/v1/admin/accounts/${encodeURIComponent(accountId)}/domains`, oazapfts.json({
        ...opts,
        method: "PATCH",
        body: addNewEmailDomainRequest
    })));
}
/**
 * Reactivates a Trial account.
 */
export function reactivateTrialAccount(opts) {
    return oazapfts.ok(oazapfts.fetchJson("/v1/accounts/trial/reactivate", {
        ...opts,
        method: "PATCH"
    }));
}
/**
 * Update personal account details
 */
export function updatePersonalAccountDetails({ accountId, accountTrialUpdateRequest }, opts) {
    return oazapfts.ok(oazapfts.fetchJson(`/v1/accounts/trial/personal-details/${encodeURIComponent(accountId)}`, oazapfts.json({
        ...opts,
        method: "PATCH",
        body: accountTrialUpdateRequest
    })));
}
/**
 * Retrieves the SSO attributes configuration for a given SSO ID.
 */
export function ssoGetAttributesConfig({ id }, opts) {
    return oazapfts.ok(oazapfts.fetchJson(`/v1/accounts/sso/${encodeURIComponent(id)}/attributes-configuration`, {
        ...opts
    }));
}
/**
 * Updates the configuration of SSO attributes for a specified SSO.
 */
export function ssoAddAttributes({ id, accountSsoAddAttributesConfigRequest }, opts) {
    return oazapfts.ok(oazapfts.fetchJson(`/v1/accounts/sso/${encodeURIComponent(id)}/attributes-configuration`, oazapfts.json({
        ...opts,
        method: "PATCH",
        body: accountSsoAddAttributesConfigRequest
    })));
}
/**
 * Get account rate limit
 */
export function getAccountRateLimit(opts) {
    return oazapfts.ok(oazapfts.fetchJson("/v1/accounts/rate-limit", {
        ...opts
    }));
}
/**
 * Update Account rate limit
 */
export function updateRateLimit({ updateRateLimitRequest }, opts) {
    return oazapfts.ok(oazapfts.fetchJson("/v1/accounts/rate-limit", oazapfts.json({
        ...opts,
        method: "PATCH",
        body: updateRateLimitRequest
    })));
}
/**
 * Get Partner by account Id
 */
export function getPartnerAccount({ id }, opts) {
    return oazapfts.ok(oazapfts.fetchJson(`/v1/account-partners/${encodeURIComponent(id)}`, {
        ...opts
    }));
}
/**
 * Delete Partner
 */
export function deletePartner({ id }, opts) {
    return oazapfts.ok(oazapfts.fetchJson(`/v1/account-partners/${encodeURIComponent(id)}`, {
        ...opts,
        method: "DELETE"
    }));
}
/**
 * Update Partner Account data.
 */
export function updatePartnerAccountData({ id, accountPartnerAccountDataUpdateRequest }, opts) {
    return oazapfts.ok(oazapfts.fetchJson(`/v1/account-partners/${encodeURIComponent(id)}`, oazapfts.json({
        ...opts,
        method: "PATCH",
        body: accountPartnerAccountDataUpdateRequest
    })));
}
/**
 * Update Partner Account Admin data.
 */
export function updatePartnerAccountAdminData({ id, accountPartnerAdminDataUpdateRequest }, opts) {
    return oazapfts.ok(oazapfts.fetchJson(`/v1/account-partners/${encodeURIComponent(id)}/admin`, oazapfts.json({
        ...opts,
        method: "PATCH",
        body: accountPartnerAdminDataUpdateRequest
    })));
}
/**
 * Retrieves a list of SCM credentials associated with the current user's account, including secret names and provider details.
 */
export function listScmCredentials2({ userId }, opts) {
    return oazapfts.ok(oazapfts.fetchJson(`/v2/users/${encodeURIComponent(userId)}/scm-credentials`, {
        ...opts
    }));
}
/**
 * Retrieves a paginated list of groups associated with a service credential for a given client ID.
 */
export function getServiceCredentialGroups({ id }, opts) {
    return oazapfts.ok(oazapfts.fetchJson(`/v2/service-credentials/${encodeURIComponent(id)}/groups`, {
        ...opts
    }));
}
/**
 * Retrieves an rate limit associated to a given service credential id.
 */
export function getServiceCredentialByIdRateLimit({ clientId }, opts) {
    return oazapfts.ok(oazapfts.fetchJson(`/v2/service-credentials/${encodeURIComponent(clientId)}/rate-limit`, {
        ...opts
    }));
}
/**
 * Get Roles
 */
export function getRoles3({ size, page, sort, direction, search, filterMode, filterBy, filterValue, multiFilterMode, filterIn }, opts) {
    return oazapfts.ok(oazapfts.fetchJson(`/v2/roles${QS.query(QS.explode({
        size,
        page,
        sort,
        direction,
        search,
        filterMode,
        filterBy,
        filterValue,
        multiFilterMode,
        filterIn
    }))}`, {
        ...opts
    }));
}
/**
 * Get all Global Resources and Actions compared to role
 */
export function getResourcesAndActionsWithStatus1({ roleId, size, page, sort, direction, search, filterMode, filterBy, filterValue, multiFilterMode, filterIn }, opts) {
    return oazapfts.ok(oazapfts.fetchJson(`/v2/roles/${encodeURIComponent(roleId)}/resources${QS.query(QS.explode({
        size,
        page,
        sort,
        direction,
        search,
        filterMode,
        filterBy,
        filterValue,
        multiFilterMode,
        filterIn
    }))}`, {
        ...opts
    }));
}
/**
 * Get role members
 */
export function getRoleMembers1({ roleId, size, page, sort, direction, search, filterMode, filterBy, filterValue, multiFilterMode, filterIn }, opts) {
    return oazapfts.ok(oazapfts.fetchJson(`/v2/roles/${encodeURIComponent(roleId)}/members${QS.query(QS.explode({
        size,
        page,
        sort,
        direction,
        search,
        filterMode,
        filterBy,
        filterValue,
        multiFilterMode,
        filterIn
    }))}`, {
        ...opts
    }));
}
/**
 * Get Groups from role
 */
export function getRoleGroups1({ roleId, size, page, sort, direction, search, filterMode, filterBy, filterValue, multiFilterMode, filterIn }, opts) {
    return oazapfts.ok(oazapfts.fetchJson(`/v2/roles/${encodeURIComponent(roleId)}/groups${QS.query(QS.explode({
        size,
        page,
        sort,
        direction,
        search,
        filterMode,
        filterBy,
        filterValue,
        multiFilterMode,
        filterIn
    }))}`, {
        ...opts
    }));
}
/**
 * Retrieves a list of resources.
 */
export function getResources1({ size, page, sort, direction, search, filterMode, filterBy, filterValue, multiFilterMode, filterIn }, opts) {
    return oazapfts.ok(oazapfts.fetchJson(`/v2/resources${QS.query(QS.explode({
        size,
        page,
        sort,
        direction,
        search,
        filterMode,
        filterBy,
        filterValue,
        multiFilterMode,
        filterIn
    }))}`, {
        ...opts
    }));
}
/**
 * Retrieves a list of members for a specified resource.
 */
export function getResourceMembers({ resourceId, size, page, sort, direction, search, filterMode, filterBy, filterValue, multiFilterMode, filterIn }, opts) {
    return oazapfts.ok(oazapfts.fetchJson(`/v2/resources/${encodeURIComponent(resourceId)}/members${QS.query(QS.explode({
        size,
        page,
        sort,
        direction,
        search,
        filterMode,
        filterBy,
        filterValue,
        multiFilterMode,
        filterIn
    }))}`, {
        ...opts
    }));
}
/**
 * Retrieves a list of groups associated with a specified resource.
 */
export function getResourceGroups({ resourceId, size, page, sort, direction, search, filterMode, filterBy, filterValue, multiFilterMode, filterIn }, opts) {
    return oazapfts.ok(oazapfts.fetchJson(`/v2/resources/${encodeURIComponent(resourceId)}/groups${QS.query(QS.explode({
        size,
        page,
        sort,
        direction,
        search,
        filterMode,
        filterBy,
        filterValue,
        multiFilterMode,
        filterIn
    }))}`, {
        ...opts
    }));
}
/**
 * Get Resource Types
 */
export function getResourceTypes({ size, page, sort, direction, search, filterMode, filterBy, filterValue, multiFilterMode, filterIn }, opts) {
    return oazapfts.ok(oazapfts.fetchJson(`/v2/resource-types${QS.query(QS.explode({
        size,
        page,
        sort,
        direction,
        search,
        filterMode,
        filterBy,
        filterValue,
        multiFilterMode,
        filterIn
    }))}`, {
        ...opts
    }));
}
/**
 * Retrieves a list of account members.
 */
export function getAccountMembers1({ size, page, sort, direction, search, filterMode, filterBy, filterValue, multiFilterMode, filterIn }, opts) {
    return oazapfts.ok(oazapfts.fetchJson(`/v2/members${QS.query(QS.explode({
        size,
        page,
        sort,
        direction,
        search,
        filterMode,
        filterBy,
        filterValue,
        multiFilterMode,
        filterIn
    }))}`, {
        ...opts
    }));
}
/**
 * Retrieves a list of roles associated with a specific member.
 */
export function getRoles4({ memberId, size, page, sort, direction, search, filterMode, filterBy, filterValue, multiFilterMode, filterIn }, opts) {
    return oazapfts.ok(oazapfts.fetchJson(`/v2/members/${encodeURIComponent(memberId)}/roles${QS.query(QS.explode({
        size,
        page,
        sort,
        direction,
        search,
        filterMode,
        filterBy,
        filterValue,
        multiFilterMode,
        filterIn
    }))}`, {
        ...opts
    }));
}
/**
 * Retrieves a list of groups for a specified member.
 */
export function getMemberGroups1({ memberId, size, page, sort, direction, search, filterMode, filterBy, filterValue, multiFilterMode, filterIn }, opts) {
    return oazapfts.ok(oazapfts.fetchJson(`/v2/members/${encodeURIComponent(memberId)}/groups${QS.query(QS.explode({
        size,
        page,
        sort,
        direction,
        search,
        filterMode,
        filterBy,
        filterValue,
        multiFilterMode,
        filterIn
    }))}`, {
        ...opts
    }));
}
/**
 * Get Groups
 */
export function getGroups1({ size, page, sort, direction, search, filterMode, filterBy, filterValue, multiFilterMode, filterIn, includeDefaultGroup }, opts) {
    return oazapfts.ok(oazapfts.fetchJson(`/v2/groups${QS.query(QS.explode({
        size,
        page,
        sort,
        direction,
        search,
        filterMode,
        filterBy,
        filterValue,
        multiFilterMode,
        filterIn,
        includeDefaultGroup
    }))}`, {
        ...opts
    }));
}
/**
 * Get Group roles
 */
export function getRoles5({ groupId, size, page, sort, direction, search, filterMode, filterBy, filterValue, multiFilterMode, filterIn }, opts) {
    return oazapfts.ok(oazapfts.fetchJson(`/v2/groups/${encodeURIComponent(groupId)}/roles${QS.query(QS.explode({
        size,
        page,
        sort,
        direction,
        search,
        filterMode,
        filterBy,
        filterValue,
        multiFilterMode,
        filterIn
    }))}`, {
        ...opts
    }));
}
/**
 * Get Group Resources
 */
export function getGroupResources1({ groupId, size, page, sort, direction, search, filterMode, filterBy, filterValue, multiFilterMode, filterIn }, opts) {
    return oazapfts.ok(oazapfts.fetchJson(`/v2/groups/${encodeURIComponent(groupId)}/resources${QS.query(QS.explode({
        size,
        page,
        sort,
        direction,
        search,
        filterMode,
        filterBy,
        filterValue,
        multiFilterMode,
        filterIn
    }))}`, {
        ...opts
    }));
}
/**
 * Get Group members
 */
export function getMembers1({ groupId, size, page, sort, direction, search, filterMode, filterBy, filterValue, multiFilterMode, filterIn }, opts) {
    return oazapfts.ok(oazapfts.fetchJson(`/v2/groups/${encodeURIComponent(groupId)}/members${QS.query(QS.explode({
        size,
        page,
        sort,
        direction,
        search,
        filterMode,
        filterBy,
        filterValue,
        multiFilterMode,
        filterIn
    }))}`, {
        ...opts
    }));
}
/**
 * Get enabled feature flags for an account
 */
export function getEnabledFeatureFlagsForAccount({ queryParams }, opts) {
    return oazapfts.ok(oazapfts.fetchJson(`/v2/feature-flags/available${QS.query(QS.explode({
        queryParams
    }))}`, {
        ...opts
    }));
}
/**
 * Retries a external link by type.
 */
export function getExternalLinksByType({ externalLinkType }, opts) {
    return oazapfts.ok(oazapfts.fetchJson(`/v2/external-links/${encodeURIComponent(externalLinkType)}`, {
        ...opts
    }));
}
/**
 * Verifies if SCM credentials exist for the current user's account and email, throwing an error if not found.
 */
export function isCreatedScmCredentials(opts) {
    return oazapfts.ok(oazapfts.fetchJson("/v1/users/scm-credentials/configuration-status", {
        ...opts
    }));
}
/**
 * Get Accounts
 */
export function getAccounts1({ accountType, slug, id }, opts) {
    return oazapfts.ok(oazapfts.fetchJson(`/v1/stackspot/accounts${QS.query(QS.explode({
        accountType,
        slug,
        id
    }))}`, {
        ...opts
    }));
}
/**
 * Get Account by id
 */
export function getAccount({ id }, opts) {
    return oazapfts.ok(oazapfts.fetchJson(`/v1/stackspot/accounts/${encodeURIComponent(id)}`, {
        ...opts
    }));
}
/**
 * Get Account Members
 */
export function getAccountMembers2({ id, size, page, sort, direction, search, filterMode, filterBy, filterValue, multiFilterMode, filterIn }, opts) {
    return oazapfts.ok(oazapfts.fetchJson(`/v1/stackspot/accounts/${encodeURIComponent(id)}/members${QS.query(QS.explode({
        size,
        page,
        sort,
        direction,
        search,
        filterMode,
        filterBy,
        filterValue,
        multiFilterMode,
        filterIn
    }))}`, {
        ...opts
    }));
}
/**
 * Get Service Credential by ID (Client ID)
 */
export function getServiceCredential({ id }, opts) {
    return oazapfts.ok(oazapfts.fetchJson(`/v1/service-credentials/${encodeURIComponent(id)}`, {
        ...opts
    }));
}
/**
 * List service credential permissions
 */
export function getServiceCredentialPermissions({ id }, opts) {
    return oazapfts.ok(oazapfts.fetchJson(`/v1/service-credentials/${encodeURIComponent(id)}/permissions`, {
        ...opts
    }));
}
/**
 * List service credential groups
 */
export function getServiceCredentialGroups1({ id }, opts) {
    return oazapfts.ok(oazapfts.fetchJson(`/v1/service-credentials/${encodeURIComponent(id)}/groups`, {
        ...opts
    }));
}
/**
 * Retrieve secret value
 */
export function getSecretValue({ xMemberId, xAccountId, xResourceSlug, xResourceTypeSlug, secretId }, opts) {
    return oazapfts.ok(oazapfts.fetchJson(`/v1/secrets/${encodeURIComponent(secretId)}/value`, {
        ...opts,
        headers: oazapfts.mergeHeaders(opts?.headers, {
            "x-member-id": xMemberId,
            "x-account-id": xAccountId,
            "x-resource-slug": xResourceSlug,
            "x-resource-type-slug": xResourceTypeSlug
        })
    }));
}
/**
 * Retrieve the resource associated to the secret
 */
export function findAssociations({ secretId, size, page, sort, direction, search, filterMode, filterBy, filterValue, multiFilterMode, filterIn }, opts) {
    return oazapfts.ok(oazapfts.fetchJson(`/v1/secrets/${encodeURIComponent(secretId)}/associations${QS.query(QS.explode({
        size,
        page,
        sort,
        direction,
        search,
        filterMode,
        filterBy,
        filterValue,
        multiFilterMode,
        filterIn
    }))}`, {
        ...opts
    }));
}
/**
 * Retrieves the SCM provider information for the current user's account.
 */
export function getScmProvider(opts) {
    return oazapfts.ok(oazapfts.fetchJson("/v1/scm-credentials/provider", {
        ...opts
    }));
}
/**
 * Verifies the SCM credentials configuration for the current user's account and returns its status, indicating if the mandate or user-specific configuration exists.
 */
export function isCreatedScmCredentials1(opts) {
    return oazapfts.ok(oazapfts.fetchJson("/v1/scm-credentials/configuration-status", {
        ...opts
    }));
}
/**
 * Retrieves a list of permissions with their status for a specified role.
 */
export function getPermissionsWithStatus({ roleId, size, page, sort, direction, search, filterMode, filterBy, filterValue, multiFilterMode, filterIn }, opts) {
    return oazapfts.ok(oazapfts.fetchJson(`/v1/roles/${encodeURIComponent(roleId)}/permissions${QS.query(QS.explode({
        size,
        page,
        sort,
        direction,
        search,
        filterMode,
        filterBy,
        filterValue,
        multiFilterMode,
        filterIn
    }))}`, {
        ...opts
    }));
}
/**
 * Retrieves a list of members for a specified resource.
 */
export function getResourceMembers1({ resourceId, size, page, sort, direction, search, filterMode, filterBy, filterValue, multiFilterMode, filterIn }, opts) {
    return oazapfts.ok(oazapfts.fetchJson(`/v1/resources/${encodeURIComponent(resourceId)}/members${QS.query(QS.explode({
        size,
        page,
        sort,
        direction,
        search,
        filterMode,
        filterBy,
        filterValue,
        multiFilterMode,
        filterIn
    }))}`, {
        ...opts
    }));
}
/**
 * Retrieves a list of resource groups associated with a specific resource.
 */
export function getResourceGroups1({ resourceId, size, page, sort, direction, search, filterMode, filterBy, filterValue, multiFilterMode, filterIn }, opts) {
    return oazapfts.ok(oazapfts.fetchJson(`/v1/resources/${encodeURIComponent(resourceId)}/groups${QS.query(QS.explode({
        size,
        page,
        sort,
        direction,
        search,
        filterMode,
        filterBy,
        filterValue,
        multiFilterMode,
        filterIn
    }))}`, {
        ...opts
    }));
}
/**
 * Get Resource Types
 */
export function getResourceTypes1({ size, page, sort, direction, search, filterMode, filterBy, filterValue, multiFilterMode, filterIn }, opts) {
    return oazapfts.ok(oazapfts.fetchJson(`/v1/resource-types${QS.query(QS.explode({
        size,
        page,
        sort,
        direction,
        search,
        filterMode,
        filterBy,
        filterValue,
        multiFilterMode,
        filterIn
    }))}`, {
        ...opts
    }));
}
/**
 * Validate permission access
 */
export function getAccess({ resourceType, resource, action, attribute }, opts) {
    return oazapfts.ok(oazapfts.fetchJson(`/v1/permissions/access${QS.query(QS.explode({
        resourceType,
        resource,
        action,
        attribute
    }))}`, {
        ...opts
    }));
}
/**
 * Retrieves a list of resources associated with the specified member.
 */
export function getResources2({ memberId }, opts) {
    return oazapfts.ok(oazapfts.fetchJson(`/v1/members/${encodeURIComponent(memberId)}/resources`, {
        ...opts
    }));
}
/**
 * Get member favorite resources by type
 */
export function listMemberFavoritesByResource({ memberId, resourceType }, opts) {
    return oazapfts.ok(oazapfts.fetchJson(`/v1/members/${encodeURIComponent(memberId)}/favorites/resources/${encodeURIComponent(resourceType)}`, {
        ...opts
    }));
}
/**
 * Returns a list of Feature Flags associated with the current user's account.
 */
export function getFeatures3(opts) {
    return oazapfts.ok(oazapfts.fetchJson("/v1/feature-flags", {
        ...opts
    }));
}
/**
 * Get a Account Extension Version
 */
export function getExtensionVersion({ slug, versionName }, opts) {
    return oazapfts.ok(oazapfts.fetchJson(`/v1/extensions/${encodeURIComponent(slug)}/versions/${encodeURIComponent(versionName)}`, {
        ...opts
    }));
}
/**
 * Delete Account Extension Version
 */
export function deleteExtensionVersion({ slug, versionName }, opts) {
    return oazapfts.ok(oazapfts.fetchJson(`/v1/extensions/${encodeURIComponent(slug)}/versions/${encodeURIComponent(versionName)}`, {
        ...opts,
        method: "DELETE"
    }));
}
/**
 * Get the active Extension Version
 */
export function getActiveExtensionVersion({ slug }, opts) {
    return oazapfts.ok(oazapfts.fetchJson(`/v1/extensions/${encodeURIComponent(slug)}/active`, {
        ...opts
    }));
}
/**
 * Get Authentication Configuration for account
 */
export function getAuthenticationConfig({ tenantSlug }, opts) {
    return oazapfts.ok(oazapfts.fetchJson(`/v1/authentication/config${QS.query(QS.explode({
        tenant_slug: tenantSlug
    }))}`, {
        ...opts
    }));
}
/**
 * Finds and returns account details based on the provided account ID.
 */
export function getAccount1({ accountId }, opts) {
    return oazapfts.ok(oazapfts.fetchJson(`/v1/admin/accounts/${encodeURIComponent(accountId)}`, {
        ...opts
    }));
}
/**
 * Finds and returns account data associated with the provided email.
 */
export function getAccountDataByEmail({ email }, opts) {
    return oazapfts.ok(oazapfts.fetchJson(`/v1/accounts${QS.query(QS.explode({
        email
    }))}`, {
        ...opts
    }));
}
/**
 * Finds the slug associated with an active account by its account ID.
 */
export function getAccountSlug({ accountId }, opts) {
    return oazapfts.ok(oazapfts.fetchJson(`/v1/accounts/${encodeURIComponent(accountId)}`, {
        ...opts
    }));
}
/**
 * Gets the SSO configuration for Freemium.
 */
export function getTrialSsoSettings(opts) {
    return oazapfts.ok(oazapfts.fetchJson("/v1/accounts/trial/sso", {
        ...opts
    }));
}
/**
 * Finds and returns personal account details.
 */
export function getPersonalAccountDetails(opts) {
    return oazapfts.ok(oazapfts.fetchJson("/v1/accounts/trial/personal-details", {
        ...opts
    }));
}
/**
 * Finds and returns personal account expiration data associated with the provided token.
 */
export function getPersonalAccountExpirationData(opts) {
    return oazapfts.ok(oazapfts.fetchJson("/v1/accounts/trial/expiration", {
        ...opts
    }));
}
/**
 * Retrieves a list of SSO for the current account.
 */
export function getAllAccountSso({ size, page, sort, direction, search, filterMode, filterBy, filterValue, multiFilterMode, filterIn }, opts) {
    return oazapfts.ok(oazapfts.fetchJson(`/v1/accounts/sso${QS.query(QS.explode({
        size,
        page,
        sort,
        direction,
        search,
        filterMode,
        filterBy,
        filterValue,
        multiFilterMode,
        filterIn
    }))}`, {
        ...opts
    }));
}
/**
 * Get Partners who are sharing content.
 */
export function getSharingPartners({ name }, opts) {
    return oazapfts.ok(oazapfts.fetchJson(`/v1/account-partners/sharing${QS.query(QS.explode({
        name
    }))}`, {
        ...opts
    }));
}
/**
 * Get Partners with whom it is allowed to share content.
 */
export function getPartnersSharingAllowed({ name }, opts) {
    return oazapfts.ok(oazapfts.fetchJson(`/v1/account-partners/sharing-allowed${QS.query(QS.explode({
        name
    }))}`, {
        ...opts
    }));
}
/**
 * Validate Partner association limit
 */
export function validatePartnerAssociationLimit(opts) {
    return oazapfts.ok(oazapfts.fetchJson("/v1/account-orders/partners/validations/association-limit", {
        ...opts
    }));
}
/**
 * Disassociate Group to Service Credential
 */
export function disassociateGroupToServiceCredential({ id, groupId }, opts) {
    return oazapfts.ok(oazapfts.fetchJson(`/v1/service-credentials/${encodeURIComponent(id)}/teams/${encodeURIComponent(groupId)}`, {
        ...opts,
        method: "DELETE"
    }));
}
/**
 * Removes a specified role from a member.
 */
export function removeRoleFromMember({ roleId, memberId }, opts) {
    return oazapfts.ok(oazapfts.fetchJson(`/v1/roles/${encodeURIComponent(roleId)}/members/${encodeURIComponent(memberId)}`, {
        ...opts,
        method: "DELETE"
    }));
}
/**
 * Deletes a resource from the account.
 */
export function deleteResource({ resourceId }, opts) {
    return oazapfts.ok(oazapfts.fetchJson(`/v1/resources/${encodeURIComponent(resourceId)}`, {
        ...opts,
        method: "DELETE"
    }));
}
/**
 * Deletes the profile image of a member.
 */
export function deleteProfileImage({ memberId }, opts) {
    return oazapfts.ok(oazapfts.fetchJson(`/v1/members/${encodeURIComponent(memberId)}/profile-image`, {
        ...opts,
        method: "DELETE"
    }));
}
/**
 * Deletes a specific FIDO key.
 */
export function deleteFidoKey({ memberId, keyId }, opts) {
    return oazapfts.ok(oazapfts.fetchJson(`/v1/members/${encodeURIComponent(memberId)}/fido-credentials/${encodeURIComponent(keyId)}`, {
        ...opts,
        method: "DELETE"
    }));
}
/**
 * Remove resource from member favorites
 */
export function deleteMemberFavorite({ memberId, resourceType, resourceId }, opts) {
    return oazapfts.ok(oazapfts.fetchJson(`/v1/members/${encodeURIComponent(memberId)}/favorites/${encodeURIComponent(resourceType)}/${encodeURIComponent(resourceId)}`, {
        ...opts,
        method: "DELETE"
    }));
}
/**
 * Delete Group role
 */
export function deleteRole({ groupId, roleId }, opts) {
    return oazapfts.ok(oazapfts.fetchJson(`/v1/groups/${encodeURIComponent(groupId)}/roles/${encodeURIComponent(roleId)}`, {
        ...opts,
        method: "DELETE"
    }));
}
/**
 * Delete Resource from Group
 */
export function deleteResourceFromGroup({ groupId, resourceId }, opts) {
    return oazapfts.ok(oazapfts.fetchJson(`/v1/groups/${encodeURIComponent(groupId)}/resources/${encodeURIComponent(resourceId)}`, {
        ...opts,
        method: "DELETE"
    }));
}
/**
 * Delete Group member
 */
export function deleteMember({ groupId, memberId }, opts) {
    return oazapfts.ok(oazapfts.fetchJson(`/v1/groups/${encodeURIComponent(groupId)}/members/${encodeURIComponent(memberId)}`, {
        ...opts,
        method: "DELETE"
    }));
}
