// @stack-spot/portal-network/dist/api/agent-tools.js
/**
 * genAI Agent Application
 * 0.1.0
 * DO NOT MODIFY - This file has been generated using oazapfts.
 * See https://www.npmjs.com/package/oazapfts
 */
import * as Oazapfts from "@oazapfts/runtime";
import * as QS from "@oazapfts/runtime/query";
export const defaults = {
    headers: {},
    baseUrl: "/",
};
const oazapfts = Oazapfts.runtime(defaults);
export const servers = {};
/**
 * Health Check
 */
export function healthCheckHealthzGet(opts) {
    return oazapfts.ok(oazapfts.fetchJson("/healthz", {
        ...opts
    }));
}
/**
 * Readyz
 */
export function readyzReadyzGet(opts) {
    return oazapfts.ok(oazapfts.fetchJson("/readyz", {
        ...opts
    }));
}
/**
 * Get Public Tool Kits
 */
export function getPublicToolKitsV1BuiltinToolkitGet({ xAccountId, xUsername, xUserId, xUserFullName, authorization }, opts) {
    return oazapfts.ok(oazapfts.fetchJson("/v1/builtin/toolkit", {
        ...opts,
        headers: oazapfts.mergeHeaders(opts?.headers, {
            "x-account-id": xAccountId,
            "x-username": xUsername,
            "x-user-id": xUserId,
            "x-user-full-name": xUserFullName,
            authorization
        })
    }));
}
/**
 * List Toolkits
 */
export function listToolkitsV1ToolkitsGet({ visibility, xAccountId, xUsername, xUserId, xUserFullName, authorization }, opts) {
    return oazapfts.ok(oazapfts.fetchJson(`/v1/toolkits${QS.query(QS.explode({
        visibility
    }))}`, {
        ...opts,
        headers: oazapfts.mergeHeaders(opts?.headers, {
            "x-account-id": xAccountId,
            "x-username": xUsername,
            "x-user-id": xUserId,
            "x-user-full-name": xUserFullName,
            authorization
        })
    }));
}
/**
 * Create Toolkit
 */
export function createToolkitV1ToolkitsPost({ xAccountId, xUsername, xUserId, xUserFullName, authorization, body }, opts) {
    return oazapfts.ok(oazapfts.fetchJson("/v1/toolkits", oazapfts.json({
        ...opts,
        method: "POST",
        body,
        headers: oazapfts.mergeHeaders(opts?.headers, {
            "x-account-id": xAccountId,
            "x-username": xUsername,
            "x-user-id": xUserId,
            "x-user-full-name": xUserFullName,
            authorization
        })
    })));
}
/**
 * Get Toolkit
 */
export function getToolkitV1ToolkitsToolkitIdGet({ toolkitId, xAccountId, xUsername, xUserId, xUserFullName, authorization }, opts) {
    return oazapfts.ok(oazapfts.fetchJson(`/v1/toolkits/${encodeURIComponent(toolkitId)}`, {
        ...opts,
        headers: oazapfts.mergeHeaders(opts?.headers, {
            "x-account-id": xAccountId,
            "x-username": xUsername,
            "x-user-id": xUserId,
            "x-user-full-name": xUserFullName,
            authorization
        })
    }));
}
/**
 * Update Toolkit
 */
export function updateToolkitV1ToolkitsToolkitIdPatch({ toolkitId, xAccountId, xUsername, xUserId, xUserFullName, authorization, body }, opts) {
    return oazapfts.ok(oazapfts.fetchJson(`/v1/toolkits/${encodeURIComponent(toolkitId)}`, oazapfts.json({
        ...opts,
        method: "PATCH",
        body,
        headers: oazapfts.mergeHeaders(opts?.headers, {
            "x-account-id": xAccountId,
            "x-username": xUsername,
            "x-user-id": xUserId,
            "x-user-full-name": xUserFullName,
            authorization
        })
    })));
}
/**
 * Delete Toolkit
 */
export function deleteToolkitV1ToolkitsToolkitIdDelete({ toolkitId, xAccountId, xUsername, xUserId, xUserFullName, authorization }, opts) {
    return oazapfts.ok(oazapfts.fetchJson(`/v1/toolkits/${encodeURIComponent(toolkitId)}`, {
        ...opts,
        method: "DELETE",
        headers: oazapfts.mergeHeaders(opts?.headers, {
            "x-account-id": xAccountId,
            "x-username": xUsername,
            "x-user-id": xUserId,
            "x-user-full-name": xUserFullName,
            authorization
        })
    }));
}
/**
 * Fork Toolkit
 */
export function forkToolkitV1ToolkitsToolkitIdForkPost({ toolkitId, xAccountId, xUsername, xUserId, xUserFullName, authorization, toolkitForkRequest }, opts) {
    return oazapfts.ok(oazapfts.fetchJson(`/v1/toolkits/${encodeURIComponent(toolkitId)}/fork`, oazapfts.json({
        ...opts,
        method: "POST",
        body: toolkitForkRequest,
        headers: oazapfts.mergeHeaders(opts?.headers, {
            "x-account-id": xAccountId,
            "x-username": xUsername,
            "x-user-id": xUserId,
            "x-user-full-name": xUserFullName,
            authorization
        })
    })));
}
/**
 * Share Toolkit
 */
export function shareToolkitV1ToolkitsToolkitIdSharePost({ toolkitId, xAccountId, xUsername, xUserId, xUserFullName, authorization }, opts) {
    return oazapfts.ok(oazapfts.fetchJson(`/v1/toolkits/${encodeURIComponent(toolkitId)}/share`, {
        ...opts,
        method: "POST",
        headers: oazapfts.mergeHeaders(opts?.headers, {
            "x-account-id": xAccountId,
            "x-username": xUsername,
            "x-user-id": xUserId,
            "x-user-full-name": xUserFullName,
            authorization
        })
    }));
}
/**
 * Publish Toolkit
 */
export function publishToolkitV1ToolkitsToolkitIdPublishPost({ toolkitId, xAccountId, xUsername, xUserId, xUserFullName, authorization }, opts) {
    return oazapfts.ok(oazapfts.fetchJson(`/v1/toolkits/${encodeURIComponent(toolkitId)}/publish`, {
        ...opts,
        method: "POST",
        headers: oazapfts.mergeHeaders(opts?.headers, {
            "x-account-id": xAccountId,
            "x-username": xUsername,
            "x-user-id": xUserId,
            "x-user-full-name": xUserFullName,
            authorization
        })
    }));
}
/**
 * Get Toolkit Tool
 */
export function getToolkitToolV1ToolkitsToolkitIdToolsToolIdGet({ toolkitId, toolId, xAccountId, xUsername, xUserId, xUserFullName, authorization }, opts) {
    return oazapfts.ok(oazapfts.fetchJson(`/v1/toolkits/${encodeURIComponent(toolkitId)}/tools/${encodeURIComponent(toolId)}`, {
        ...opts,
        headers: oazapfts.mergeHeaders(opts?.headers, {
            "x-account-id": xAccountId,
            "x-username": xUsername,
            "x-user-id": xUserId,
            "x-user-full-name": xUserFullName,
            authorization
        })
    }));
}
/**
 * Edit Toolkit Tool
 */
export function editToolkitToolV1ToolkitsToolkitIdToolsToolIdPut({ toolkitId, toolId, xAccountId, xUsername, xUserId, xUserFullName, authorization, customToolRequest }, opts) {
    return oazapfts.ok(oazapfts.fetchJson(`/v1/toolkits/${encodeURIComponent(toolkitId)}/tools/${encodeURIComponent(toolId)}`, oazapfts.json({
        ...opts,
        method: "PUT",
        body: customToolRequest,
        headers: oazapfts.mergeHeaders(opts?.headers, {
            "x-account-id": xAccountId,
            "x-username": xUsername,
            "x-user-id": xUserId,
            "x-user-full-name": xUserFullName,
            authorization
        })
    })));
}
/**
 * Create Toolkit Tools
 */
export function createToolkitToolsV1ToolkitsToolkitIdToolsPost({ toolkitId, xAccountId, xUsername, xUserId, xUserFullName, authorization, customToolsRequest }, opts) {
    return oazapfts.ok(oazapfts.fetchJson(`/v1/toolkits/${encodeURIComponent(toolkitId)}/tools`, oazapfts.json({
        ...opts,
        method: "POST",
        body: customToolsRequest,
        headers: oazapfts.mergeHeaders(opts?.headers, {
            "x-account-id": xAccountId,
            "x-username": xUsername,
            "x-user-id": xUserId,
            "x-user-full-name": xUserFullName,
            authorization
        })
    })));
}
/**
 * Delete Toolkit Tools
 */
export function deleteToolkitToolsV1ToolkitsToolkitIdToolsDelete({ toolkitId, xAccountId, xUsername, xUserId, xUserFullName, authorization, deleteToolsRequest }, opts) {
    return oazapfts.ok(oazapfts.fetchJson(`/v1/toolkits/${encodeURIComponent(toolkitId)}/tools`, oazapfts.json({
        ...opts,
        method: "DELETE",
        body: deleteToolsRequest,
        headers: oazapfts.mergeHeaders(opts?.headers, {
            "x-account-id": xAccountId,
            "x-username": xUsername,
            "x-user-id": xUserId,
            "x-user-full-name": xUserFullName,
            authorization
        })
    })));
}
/**
 * Split Uploaded File Preview
 */
export function splitUploadedFilePreviewV1ToolkitsToolsPreviewFileUploadIdGet({ fileUploadId, authorization, xAccountId, xUsername, xUserId, xUserFullName, authorizationHeader }, opts) {
    return oazapfts.ok(oazapfts.fetchJson(`/v1/toolkits/tools/preview/${encodeURIComponent(fileUploadId)}`, {
        ...opts,
        headers: oazapfts.mergeHeaders(opts?.headers, {
            Authorization: authorization,
            "x-account-id": xAccountId,
            "x-username": xUsername,
            "x-user-id": xUserId,
            "x-user-full-name": xUserFullName,
            authorization: authorizationHeader
        })
    }));
}
/**
 * List Agents Using Tools
 */
export function listAgentsUsingToolsV1ToolkitsToolkitIdToolsAgentsPost({ toolkitId, xAccountId, xUsername, xUserId, xUserFullName, authorization, listAgentsUsingToolsRequest }, opts) {
    return oazapfts.ok(oazapfts.fetchJson(`/v1/toolkits/${encodeURIComponent(toolkitId)}/tools/agents`, oazapfts.json({
        ...opts,
        method: "POST",
        body: listAgentsUsingToolsRequest,
        headers: oazapfts.mergeHeaders(opts?.headers, {
            "x-account-id": xAccountId,
            "x-username": xUsername,
            "x-user-id": xUserId,
            "x-user-full-name": xUserFullName,
            authorization
        })
    })));
}
/**
 * Assign Tools
 */
export function assignToolsV1AgentsAgentIdToolsPost({ agentId, isPublic, xAccountId, xUsername, xUserId, xUserFullName, authorization, assignToolsToAgentRequest }, opts) {
    return oazapfts.ok(oazapfts.fetchJson(`/v1/agents/${encodeURIComponent(agentId)}/tools${QS.query(QS.explode({
        is_public: isPublic
    }))}`, oazapfts.json({
        ...opts,
        method: "POST",
        body: assignToolsToAgentRequest,
        headers: oazapfts.mergeHeaders(opts?.headers, {
            "x-account-id": xAccountId,
            "x-username": xUsername,
            "x-user-id": xUserId,
            "x-user-full-name": xUserFullName,
            authorization
        })
    })));
}
/**
 * List Tools
 */
export function listToolsV1AgentsAgentIdToolsGet({ agentId, isPublic, xAccountId, xUsername, xUserId, xUserFullName, authorization }, opts) {
    return oazapfts.ok(oazapfts.fetchJson(`/v1/agents/${encodeURIComponent(agentId)}/tools${QS.query(QS.explode({
        is_public: isPublic
    }))}`, {
        ...opts,
        headers: oazapfts.mergeHeaders(opts?.headers, {
            "x-account-id": xAccountId,
            "x-username": xUsername,
            "x-user-id": xUserId,
            "x-user-full-name": xUserFullName,
            authorization
        })
    }));
}
/**
 * Delete Tools
 */
export function deleteToolsV1AgentsAgentIdToolsDelete({ agentId, isPublic, xAccountId, xUsername, xUserId, xUserFullName, authorization }, opts) {
    return oazapfts.ok(oazapfts.fetchJson(`/v1/agents/${encodeURIComponent(agentId)}/tools${QS.query(QS.explode({
        is_public: isPublic
    }))}`, {
        ...opts,
        method: "DELETE",
        headers: oazapfts.mergeHeaders(opts?.headers, {
            "x-account-id": xAccountId,
            "x-username": xUsername,
            "x-user-id": xUserId,
            "x-user-full-name": xUserFullName,
            authorization
        })
    }));
}
/**
 * List Tools By Agent For Schema
 */
export function listToolsByAgentForSchemaV1AgentsAgentIdToolsSchemaSchemaGet({ agentId, schema, isPublic, xAccountId, xUsername, xUserId, xUserFullName, authorization }, opts) {
    return oazapfts.ok(oazapfts.fetchJson(`/v1/agents/${encodeURIComponent(agentId)}/tools/schema/${encodeURIComponent(schema)}${QS.query(QS.explode({
        is_public: isPublic
    }))}`, {
        ...opts,
        headers: oazapfts.mergeHeaders(opts?.headers, {
            "x-account-id": xAccountId,
            "x-username": xUsername,
            "x-user-id": xUserId,
            "x-user-full-name": xUserFullName,
            authorization
        })
    }));
}
/**
 * Execute Tool
 */
export function executeToolV1AgentsToolsAgentToolIdExecutePost({ agentToolId, isPublic, xAccountId, xUsername, xUserId, xUserFullName, authorization, executeAgentToolRequest }, opts) {
    return oazapfts.ok(oazapfts.fetchJson(`/v1/agents/tools/${encodeURIComponent(agentToolId)}/execute${QS.query(QS.explode({
        is_public: isPublic
    }))}`, oazapfts.json({
        ...opts,
        method: "POST",
        body: executeAgentToolRequest,
        headers: oazapfts.mergeHeaders(opts?.headers, {
            "x-account-id": xAccountId,
            "x-username": xUsername,
            "x-user-id": xUserId,
            "x-user-full-name": xUserFullName,
            authorization
        })
    })));
}
/**
 * List Agents
 */
export function listAgentsV1AgentsGet({ isPublic, name, slug, visibility, size, page, xAccountId, xUsername, xUserId, xUserFullName, authorization }, opts) {
    return oazapfts.ok(oazapfts.fetchJson(`/v1/agents${QS.query(QS.explode({
        is_public: isPublic,
        name,
        slug,
        visibility,
        size,
        page
    }))}`, {
        ...opts,
        headers: oazapfts.mergeHeaders(opts?.headers, {
            "x-account-id": xAccountId,
            "x-username": xUsername,
            "x-user-id": xUserId,
            "x-user-full-name": xUserFullName,
            authorization
        })
    }));
}
/**
 * Create Agent
 */
export function createAgentV1AgentsPost({ isPublic, xAccountId, xUsername, xUserId, xUserFullName, authorization, newAgentRequest }, opts) {
    return oazapfts.ok(oazapfts.fetchJson(`/v1/agents${QS.query(QS.explode({
        is_public: isPublic
    }))}`, oazapfts.json({
        ...opts,
        method: "POST",
        body: newAgentRequest,
        headers: oazapfts.mergeHeaders(opts?.headers, {
            "x-account-id": xAccountId,
            "x-username": xUsername,
            "x-user-id": xUserId,
            "x-user-full-name": xUserFullName,
            authorization
        })
    })));
}
/**
 * Search Agents
 */
export function searchAgentsV1AgentsSearchPost({ xAccountId, xUsername, xUserId, xUserFullName, authorization, searchAgentsRequest }, opts) {
    return oazapfts.ok(oazapfts.fetchJson("/v1/agents/search", oazapfts.json({
        ...opts,
        method: "POST",
        body: searchAgentsRequest,
        headers: oazapfts.mergeHeaders(opts?.headers, {
            "x-account-id": xAccountId,
            "x-username": xUsername,
            "x-user-id": xUserId,
            "x-user-full-name": xUserFullName,
            authorization
        })
    })));
}
/**
 * Workspace Fork
 */
export function workspaceForkV1AgentsWorkspaceForkPost({ xAccountId, xUsername, xUserId, xUserFullName, authorization, workspaceForkRequest }, opts) {
    return oazapfts.ok(oazapfts.fetchJson("/v1/agents/workspace/fork", oazapfts.json({
        ...opts,
        method: "POST",
        body: workspaceForkRequest,
        headers: oazapfts.mergeHeaders(opts?.headers, {
            "x-account-id": xAccountId,
            "x-username": xUsername,
            "x-user-id": xUserId,
            "x-user-full-name": xUserFullName,
            authorization
        })
    })));
}
/**
 * Delete Workspace Fork
 */
export function deleteWorkspaceForkV1AgentsWorkspaceForkDelete({ xAccountId, xUsername, xUserId, xUserFullName, authorization, workspaceDeleteRequest }, opts) {
    return oazapfts.ok(oazapfts.fetchJson("/v1/agents/workspace/fork", oazapfts.json({
        ...opts,
        method: "DELETE",
        body: workspaceDeleteRequest,
        headers: oazapfts.mergeHeaders(opts?.headers, {
            "x-account-id": xAccountId,
            "x-username": xUsername,
            "x-user-id": xUserId,
            "x-user-full-name": xUserFullName,
            authorization
        })
    })));
}
/**
 * List Workspace
 */
export function listWorkspaceV1AgentsWorkspaceListPost({ xAccountId, xUsername, xUserId, xUserFullName, authorization, workspaceListRequest }, opts) {
    return oazapfts.ok(oazapfts.fetchJson("/v1/agents/workspace/list", oazapfts.json({
        ...opts,
        method: "POST",
        body: workspaceListRequest,
        headers: oazapfts.mergeHeaders(opts?.headers, {
            "x-account-id": xAccountId,
            "x-username": xUsername,
            "x-user-id": xUserId,
            "x-user-full-name": xUserFullName,
            authorization
        })
    })));
}
/**
 * List Agent Dependencies
 */
export function listAgentDependenciesV1AgentsDependenciesPost({ xAccountId, xUsername, xUserId, xUserFullName, authorization, agentDependenciesRequest }, opts) {
    return oazapfts.ok(oazapfts.fetchJson("/v1/agents/dependencies", oazapfts.json({
        ...opts,
        method: "POST",
        body: agentDependenciesRequest,
        headers: oazapfts.mergeHeaders(opts?.headers, {
            "x-account-id": xAccountId,
            "x-username": xUsername,
            "x-user-id": xUserId,
            "x-user-full-name": xUserFullName,
            authorization
        })
    })));
}
/**
 * List Multi Agents
 */
export function listMultiAgentsV1AgentsMultiAgentsGet({ name, slug, visibility, size, page, xAccountId, xUsername, xUserId, xUserFullName, authorization }, opts) {
    return oazapfts.ok(oazapfts.fetchJson(`/v1/agents/multi-agents${QS.query(QS.explode({
        name,
        slug,
        visibility,
        size,
        page
    }))}`, {
        ...opts,
        headers: oazapfts.mergeHeaders(opts?.headers, {
            "x-account-id": xAccountId,
            "x-username": xUsername,
            "x-user-id": xUserId,
            "x-user-full-name": xUserFullName,
            authorization
        })
    }));
}
/**
 * Get Agent
 */
export function getAgentV1AgentsAgentIdGet({ agentId, details, xAccountId, xUsername, xUserId, xUserFullName, authorization }, opts) {
    return oazapfts.ok(oazapfts.fetchJson(`/v1/agents/${encodeURIComponent(agentId)}${QS.query(QS.explode({
        details
    }))}`, {
        ...opts,
        headers: oazapfts.mergeHeaders(opts?.headers, {
            "x-account-id": xAccountId,
            "x-username": xUsername,
            "x-user-id": xUserId,
            "x-user-full-name": xUserFullName,
            authorization
        })
    }));
}
/**
 * Update Agent
 */
export function updateAgentV1AgentsAgentIdPatch({ agentId, isPublic, xAccountId, xUsername, xUserId, xUserFullName, authorization, updateAgentRequest }, opts) {
    return oazapfts.ok(oazapfts.fetchJson(`/v1/agents/${encodeURIComponent(agentId)}${QS.query(QS.explode({
        is_public: isPublic
    }))}`, oazapfts.json({
        ...opts,
        method: "PATCH",
        body: updateAgentRequest,
        headers: oazapfts.mergeHeaders(opts?.headers, {
            "x-account-id": xAccountId,
            "x-username": xUsername,
            "x-user-id": xUserId,
            "x-user-full-name": xUserFullName,
            authorization
        })
    })));
}
/**
 * Delete Agent
 */
export function deleteAgentV1AgentsAgentIdDelete({ agentId, isPublic, xAccountId, xUsername, xUserId, xUserFullName, authorization }, opts) {
    return oazapfts.ok(oazapfts.fetchJson(`/v1/agents/${encodeURIComponent(agentId)}${QS.query(QS.explode({
        is_public: isPublic
    }))}`, {
        ...opts,
        method: "DELETE",
        headers: oazapfts.mergeHeaders(opts?.headers, {
            "x-account-id": xAccountId,
            "x-username": xUsername,
            "x-user-id": xUserId,
            "x-user-full-name": xUserFullName,
            authorization
        })
    }));
}
/**
 * Add Favorite
 */
export function addFavoriteV1AgentsAgentIdFavoritePost({ agentId, xAccountId, xUsername, xUserId, xUserFullName, authorization }, opts) {
    return oazapfts.ok(oazapfts.fetchJson(`/v1/agents/${encodeURIComponent(agentId)}/favorite`, {
        ...opts,
        method: "POST",
        headers: oazapfts.mergeHeaders(opts?.headers, {
            "x-account-id": xAccountId,
            "x-username": xUsername,
            "x-user-id": xUserId,
            "x-user-full-name": xUserFullName,
            authorization
        })
    }));
}
/**
 * Delete Favorite
 */
export function deleteFavoriteV1AgentsAgentIdFavoriteDelete({ agentId, xAccountId, xUsername, xUserId, xUserFullName, authorization }, opts) {
    return oazapfts.ok(oazapfts.fetchJson(`/v1/agents/${encodeURIComponent(agentId)}/favorite`, {
        ...opts,
        method: "DELETE",
        headers: oazapfts.mergeHeaders(opts?.headers, {
            "x-account-id": xAccountId,
            "x-username": xUsername,
            "x-user-id": xUserId,
            "x-user-full-name": xUserFullName,
            authorization
        })
    }));
}
/**
 * Fork Agent
 */
export function forkAgentV1AgentsAgentIdForkPost({ agentId, isPublic, xAccountId, xUsername, xUserId, xUserFullName, authorization, forkAgentRequest }, opts) {
    return oazapfts.ok(oazapfts.fetchJson(`/v1/agents/${encodeURIComponent(agentId)}/fork${QS.query(QS.explode({
        is_public: isPublic
    }))}`, oazapfts.json({
        ...opts,
        method: "POST",
        body: forkAgentRequest,
        headers: oazapfts.mergeHeaders(opts?.headers, {
            "x-account-id": xAccountId,
            "x-username": xUsername,
            "x-user-id": xUserId,
            "x-user-full-name": xUserFullName,
            authorization
        })
    })));
}
/**
 * Publish Agent
 */
export function publishAgentV1AgentsAgentIdPublishPost({ agentId, xAccountId, xUsername, xUserId, xUserFullName, authorization, publishAgentRequest }, opts) {
    return oazapfts.ok(oazapfts.fetchJson(`/v1/agents/${encodeURIComponent(agentId)}/publish`, oazapfts.json({
        ...opts,
        method: "POST",
        body: publishAgentRequest,
        headers: oazapfts.mergeHeaders(opts?.headers, {
            "x-account-id": xAccountId,
            "x-username": xUsername,
            "x-user-id": xUserId,
            "x-user-full-name": xUserFullName,
            authorization
        })
    })));
}
/**
 * Share
 */
export function shareV1AgentsAgentIdSharePost({ agentId, xAccountId, xUsername, xUserId, xUserFullName, authorization }, opts) {
    return oazapfts.ok(oazapfts.fetchJson(`/v1/agents/${encodeURIComponent(agentId)}/share`, {
        ...opts,
        method: "POST",
        headers: oazapfts.mergeHeaders(opts?.headers, {
            "x-account-id": xAccountId,
            "x-username": xUsername,
            "x-user-id": xUserId,
            "x-user-full-name": xUserFullName,
            authorization
        })
    }));
}
/**
 * Disassociate Ks From All Agents
 */
export function disassociateKsFromAllAgentsV1AgentsKnowledgeSourceKsIdDelete({ ksId, xAccountId, xUsername, xUserId, xUserFullName, authorization }, opts) {
    return oazapfts.ok(oazapfts.fetchJson(`/v1/agents/knowledge-source/${encodeURIComponent(ksId)}`, {
        ...opts,
        method: "DELETE",
        headers: oazapfts.mergeHeaders(opts?.headers, {
            "x-account-id": xAccountId,
            "x-username": xUsername,
            "x-user-id": xUserId,
            "x-user-full-name": xUserFullName,
            authorization
        })
    }));
}
/**
 * Get Agent By Ks Id
 */
export function getAgentByKsIdV1AgentsKnowledgeSourceKsIdGet({ ksId, xAccountId, xUsername, xUserId, xUserFullName, authorization }, opts) {
    return oazapfts.ok(oazapfts.fetchJson(`/v1/agents/knowledge-source/${encodeURIComponent(ksId)}`, {
        ...opts,
        headers: oazapfts.mergeHeaders(opts?.headers, {
            "x-account-id": xAccountId,
            "x-username": xUsername,
            "x-user-id": xUserId,
            "x-user-full-name": xUserFullName,
            authorization
        })
    }));
}
/**
 * Migrate Agent Avatar By Id
 */
export function migrateAgentAvatarByIdV1AgentsAgentIdMigrateAvatarPost({ agentId, xAccountId, xUsername, xUserId, xUserFullName, authorization }, opts) {
    return oazapfts.ok(oazapfts.fetchJson(`/v1/agents/${encodeURIComponent(agentId)}/migrate-avatar`, {
        ...opts,
        method: "POST",
        headers: oazapfts.mergeHeaders(opts?.headers, {
            "x-account-id": xAccountId,
            "x-username": xUsername,
            "x-user-id": xUserId,
            "x-user-full-name": xUserFullName,
            authorization
        })
    }));
}
/**
 * List Agents
 */
export function listAgentsV2AgentsGet({ name, slug, visibility, size, page, xAccountId, xUsername, xUserId, xUserFullName, authorization }, opts) {
    return oazapfts.ok(oazapfts.fetchJson(`/v2/agents${QS.query(QS.explode({
        name,
        slug,
        visibility,
        size,
        page
    }))}`, {
        ...opts,
        headers: oazapfts.mergeHeaders(opts?.headers, {
            "x-account-id": xAccountId,
            "x-username": xUsername,
            "x-user-id": xUserId,
            "x-user-full-name": xUserFullName,
            authorization
        })
    }));
}
/**
 * Search Agents
 */
export function searchAgentsV2AgentsSearchPost({ xAccountId, xUsername, xUserId, xUserFullName, authorization, searchAgentsRequest }, opts) {
    return oazapfts.ok(oazapfts.fetchJson("/v2/agents/search", oazapfts.json({
        ...opts,
        method: "POST",
        body: searchAgentsRequest,
        headers: oazapfts.mergeHeaders(opts?.headers, {
            "x-account-id": xAccountId,
            "x-username": xUsername,
            "x-user-id": xUserId,
            "x-user-full-name": xUserFullName,
            authorization
        })
    })));
}
/**
 * List Agents
 */
export function listAgentsV3AgentsGet({ filters, isPublic, xAccountId, xUsername, xUserId, xUserFullName, authorization }, opts) {
    return oazapfts.ok(oazapfts.fetchJson(`/v3/agents${QS.query(QS.explode({
        filters,
        is_public: isPublic
    }))}`, {
        ...opts,
        headers: oazapfts.mergeHeaders(opts?.headers, {
            "x-account-id": xAccountId,
            "x-username": xUsername,
            "x-user-id": xUserId,
            "x-user-full-name": xUserFullName,
            authorization
        })
    }));
}
/**
 * Internal List Toolkits By Ids
 */
export function internalListToolkitsByIdsV1SpotToolkitsPost({ xAccountId, xUsername, xUserId, xUserFullName, authorization, internalListToolkitsRequest }, opts) {
    return oazapfts.ok(oazapfts.fetchJson("/v1/spot/toolkits", oazapfts.json({
        ...opts,
        method: "POST",
        body: internalListToolkitsRequest,
        headers: oazapfts.mergeHeaders(opts?.headers, {
            "x-account-id": xAccountId,
            "x-username": xUsername,
            "x-user-id": xUserId,
            "x-user-full-name": xUserFullName,
            authorization
        })
    })));
}
/**
 * Internal Delete Toolkits By Ids
 */
export function internalDeleteToolkitsByIdsV1SpotToolkitsDelete({ xAccountId, xUsername, xUserId, xUserFullName, authorization, internalDeleteToolkitsRequest }, opts) {
    return oazapfts.ok(oazapfts.fetchJson("/v1/spot/toolkits", oazapfts.json({
        ...opts,
        method: "DELETE",
        body: internalDeleteToolkitsRequest,
        headers: oazapfts.mergeHeaders(opts?.headers, {
            "x-account-id": xAccountId,
            "x-username": xUsername,
            "x-user-id": xUserId,
            "x-user-full-name": xUserFullName,
            authorization
        })
    })));
}
/**
 * Internal Fork Toolkits By Ids
 */
export function internalForkToolkitsByIdsV1SpotToolkitsForkPost({ xAccountId, xUsername, xUserId, xUserFullName, authorization, internalToolkitForkRequest }, opts) {
    return oazapfts.ok(oazapfts.fetchJson("/v1/spot/toolkits/fork", oazapfts.json({
        ...opts,
        method: "POST",
        body: internalToolkitForkRequest,
        headers: oazapfts.mergeHeaders(opts?.headers, {
            "x-account-id": xAccountId,
            "x-username": xUsername,
            "x-user-id": xUserId,
            "x-user-full-name": xUserFullName,
            authorization
        })
    })));
}
/**
 * List Toolkits
 */
export function listToolkitsV2ToolkitsGet({ name, visibilityList, size, page, order, xAccountId, xUsername, xUserId, xUserFullName, authorization }, opts) {
    return oazapfts.ok(oazapfts.fetchJson(`/v2/toolkits${QS.query(QS.explode({
        name,
        visibility_list: visibilityList,
        size,
        page,
        order
    }))}`, {
        ...opts,
        headers: oazapfts.mergeHeaders(opts?.headers, {
            "x-account-id": xAccountId,
            "x-username": xUsername,
            "x-user-id": xUserId,
            "x-user-full-name": xUserFullName,
            authorization
        })
    }));
}
/**
 * List Mcp Tools
 */
export function listMcpToolsV1McpToolsToolkitIdGet({ toolkitId, cacheControl, xAccountId, xUsername, xUserId, xUserFullName, authorization }, opts) {
    return oazapfts.ok(oazapfts.fetchJson(`/v1/mcp-tools/${encodeURIComponent(toolkitId)}`, {
        ...opts,
        headers: oazapfts.mergeHeaders(opts?.headers, {
            "cache-control": cacheControl,
            "x-account-id": xAccountId,
            "x-username": xUsername,
            "x-user-id": xUserId,
            "x-user-full-name": xUserFullName,
            authorization
        })
    }));
}
