// @stack-spot/portal-network/dist/api/ai.js
/**
 * FastAPI
 * 0.1.0
 * DO NOT MODIFY - This file has been generated using oazapfts.
 * See https://www.npmjs.com/package/oazapfts
 */
import * as Oazapfts from "@oazapfts/runtime";
import * as QS from "@oazapfts/runtime/query";
export const defaults = {
    headers: {},
    baseUrl: "/",
};
const oazapfts = Oazapfts.runtime(defaults);
export const servers = {};
/**
 * Metrics
 */
export function metricsMetricsGet(opts) {
    return oazapfts.ok(oazapfts.fetchJson("/metrics", {
        ...opts
    }));
}
/**
 * Workspace Fork
 */
export function workspaceForkV1AiStacksWorkspaceForkPost({ authorization, xAccountId, xMemberId, xUsername, aiStackWorkspaceForkRequest }, opts) {
    return oazapfts.ok(oazapfts.fetchJson("/v1/ai-stacks/workspace/fork", oazapfts.json({
        ...opts,
        method: "POST",
        body: aiStackWorkspaceForkRequest,
        headers: oazapfts.mergeHeaders(opts?.headers, {
            authorization,
            "x-account-id": xAccountId,
            "x-member-id": xMemberId,
            "x-username": xUsername
        })
    })));
}
/**
 * Workspace Delete Fork
 */
export function workspaceDeleteForkV1AiStacksWorkspaceForkDelete({ authorization, xAccountId, xMemberId, xUsername, aiStackWorkspaceDeleteRequest }, opts) {
    return oazapfts.ok(oazapfts.fetchJson("/v1/ai-stacks/workspace/fork", oazapfts.json({
        ...opts,
        method: "DELETE",
        body: aiStackWorkspaceDeleteRequest,
        headers: oazapfts.mergeHeaders(opts?.headers, {
            authorization,
            "x-account-id": xAccountId,
            "x-member-id": xMemberId,
            "x-username": xUsername
        })
    })));
}
/**
 * Workspace List
 */
export function workspaceListV1AiStacksWorkspaceListPost({ authorization, xAccountId, xMemberId, xUsername, aiStackWorkspaceListRequest }, opts) {
    return oazapfts.ok(oazapfts.fetchJson("/v1/ai-stacks/workspace/list", oazapfts.json({
        ...opts,
        method: "POST",
        body: aiStackWorkspaceListRequest,
        headers: oazapfts.mergeHeaders(opts?.headers, {
            authorization,
            "x-account-id": xAccountId,
            "x-member-id": xMemberId,
            "x-username": xUsername
        })
    })));
}
/**
 * List Ai Stacks
 */
export function listAiStacksV1AiStacksGet({ visibility, authorization, xAccountId, xMemberId, xUsername }, opts) {
    return oazapfts.ok(oazapfts.fetchJson(`/v1/ai-stacks${QS.query(QS.explode({
        visibility
    }))}`, {
        ...opts,
        headers: oazapfts.mergeHeaders(opts?.headers, {
            authorization,
            "x-account-id": xAccountId,
            "x-member-id": xMemberId,
            "x-username": xUsername
        })
    }));
}
/**
 * Create Ai Stack
 */
export function createAiStackV1AiStacksPost({ authorization, xAccountId, xMemberId, xUsername, newAiStackRequest }, opts) {
    return oazapfts.ok(oazapfts.fetchJson("/v1/ai-stacks", oazapfts.json({
        ...opts,
        method: "POST",
        body: newAiStackRequest,
        headers: oazapfts.mergeHeaders(opts?.headers, {
            authorization,
            "x-account-id": xAccountId,
            "x-member-id": xMemberId,
            "x-username": xUsername
        })
    })));
}
/**
 * Update Ai Stack
 */
export function updateAiStackV1AiStacksStackIdPatch({ stackId, authorization, xAccountId, xMemberId, xUsername, updateAiStackRequest }, opts) {
    return oazapfts.ok(oazapfts.fetchJson(`/v1/ai-stacks/${encodeURIComponent(stackId)}`, oazapfts.json({
        ...opts,
        method: "PATCH",
        body: updateAiStackRequest,
        headers: oazapfts.mergeHeaders(opts?.headers, {
            authorization,
            "x-account-id": xAccountId,
            "x-member-id": xMemberId,
            "x-username": xUsername
        })
    })));
}
/**
 * Get Ai Stack
 */
export function getAiStackV1AiStacksStackIdGet({ stackId, authorization, xAccountId, xMemberId, xUsername }, opts) {
    return oazapfts.ok(oazapfts.fetchJson(`/v1/ai-stacks/${encodeURIComponent(stackId)}`, {
        ...opts,
        headers: oazapfts.mergeHeaders(opts?.headers, {
            authorization,
            "x-account-id": xAccountId,
            "x-member-id": xMemberId,
            "x-username": xUsername
        })
    }));
}
/**
 * Remove Ai Stack
 */
export function removeAiStackV1AiStacksStackIdDelete({ stackId, authorization, xAccountId, xMemberId, xUsername }, opts) {
    return oazapfts.ok(oazapfts.fetchJson(`/v1/ai-stacks/${encodeURIComponent(stackId)}`, {
        ...opts,
        method: "DELETE",
        headers: oazapfts.mergeHeaders(opts?.headers, {
            authorization,
            "x-account-id": xAccountId,
            "x-member-id": xMemberId,
            "x-username": xUsername
        })
    }));
}
/**
 * Get Ai Stack Exists
 */
export function getAiStackExistsV1AiStacksStackNameExistsGet({ stackName, authorization, xAccountId, xMemberId, xUsername }, opts) {
    return oazapfts.ok(oazapfts.fetchJson(`/v1/ai-stacks/${encodeURIComponent(stackName)}/exists`, {
        ...opts,
        headers: oazapfts.mergeHeaders(opts?.headers, {
            authorization,
            "x-account-id": xAccountId,
            "x-member-id": xMemberId,
            "x-username": xUsername
        })
    }));
}
/**
 * Add Favorite
 */
export function addFavoriteV1AiStacksStackIdFavoritePost({ stackId, authorization, xAccountId, xMemberId, xUsername }, opts) {
    return oazapfts.ok(oazapfts.fetchJson(`/v1/ai-stacks/${encodeURIComponent(stackId)}/favorite`, {
        ...opts,
        method: "POST",
        headers: oazapfts.mergeHeaders(opts?.headers, {
            authorization,
            "x-account-id": xAccountId,
            "x-member-id": xMemberId,
            "x-username": xUsername
        })
    }));
}
/**
 * Delete Favorite
 */
export function deleteFavoriteV1AiStacksStackIdFavoriteDelete({ stackId, authorization, xAccountId, xMemberId, xUsername }, opts) {
    return oazapfts.ok(oazapfts.fetchJson(`/v1/ai-stacks/${encodeURIComponent(stackId)}/favorite`, {
        ...opts,
        method: "DELETE",
        headers: oazapfts.mergeHeaders(opts?.headers, {
            authorization,
            "x-account-id": xAccountId,
            "x-member-id": xMemberId,
            "x-username": xUsername
        })
    }));
}
/**
 * Fork
 */
export function forkV1AiStacksStackIdForkPost({ stackId, authorization, xAccountId, xMemberId, xUsername, aiStackForkRequest }, opts) {
    return oazapfts.ok(oazapfts.fetchJson(`/v1/ai-stacks/${encodeURIComponent(stackId)}/fork`, oazapfts.json({
        ...opts,
        method: "POST",
        body: aiStackForkRequest,
        headers: oazapfts.mergeHeaders(opts?.headers, {
            authorization,
            "x-account-id": xAccountId,
            "x-member-id": xMemberId,
            "x-username": xUsername
        })
    })));
}
/**
 * Share
 */
export function shareV1AiStacksStackIdSharePost({ stackId, authorization, xAccountId, xMemberId, xUsername }, opts) {
    return oazapfts.ok(oazapfts.fetchJson(`/v1/ai-stacks/${encodeURIComponent(stackId)}/share`, {
        ...opts,
        method: "POST",
        headers: oazapfts.mergeHeaders(opts?.headers, {
            authorization,
            "x-account-id": xAccountId,
            "x-member-id": xMemberId,
            "x-username": xUsername
        })
    }));
}
/**
 * Publish
 */
export function publishV1AiStacksStackIdPublishPost({ stackId, authorization, xAccountId, xMemberId, xUsername, aiStackPublishRequest }, opts) {
    return oazapfts.ok(oazapfts.fetchJson(`/v1/ai-stacks/${encodeURIComponent(stackId)}/publish`, oazapfts.json({
        ...opts,
        method: "POST",
        body: aiStackPublishRequest,
        headers: oazapfts.mergeHeaders(opts?.headers, {
            authorization,
            "x-account-id": xAccountId,
            "x-member-id": xMemberId,
            "x-username": xUsername
        })
    })));
}
/**
 * Project Files
 */
export function projectFilesV1ProjectFilesPost({ authorization, newProjectFilesRequest }, opts) {
    return oazapfts.ok(oazapfts.fetchJson("/v1/project-files", oazapfts.json({
        ...opts,
        method: "POST",
        body: newProjectFilesRequest,
        headers: oazapfts.mergeHeaders(opts?.headers, {
            authorization
        })
    })));
}
/**
 * Quick Actions
 */
export function quickActionsV1QuickActionsPost({ authorization, xAccountId, xMemberId, xUsername, quickActionsRequest }, opts) {
    return oazapfts.ok(oazapfts.fetchJson("/v1/quick-actions", oazapfts.json({
        ...opts,
        method: "POST",
        body: quickActionsRequest,
        headers: oazapfts.mergeHeaders(opts?.headers, {
            authorization,
            "x-account-id": xAccountId,
            "x-member-id": xMemberId,
            "x-username": xUsername
        })
    })));
}
/**
 * Dev Assistant
 */
export function devAssistantV1ChatPost(opts) {
    return oazapfts.ok(oazapfts.fetchJson("/v1/chat", {
        ...opts,
        method: "POST"
    }));
}
/**
 * Autocomplete V1
 */
export function autocompleteV1V1AutocompletePost({ authorization, xAccountId, xMemberId, xUsername, autoCompleteRequest }, opts) {
    return oazapfts.ok(oazapfts.fetchJson("/v1/autocomplete", oazapfts.json({
        ...opts,
        method: "POST",
        body: autoCompleteRequest,
        headers: oazapfts.mergeHeaders(opts?.headers, {
            authorization,
            "x-account-id": xAccountId,
            "x-member-id": xMemberId,
            "x-username": xUsername
        })
    })));
}
/**
 * Post Event
 */
export function postEventV1EventsPost({ authorization, xAccountId, xMemberId, xUsername, body }, opts) {
    return oazapfts.ok(oazapfts.fetchJson("/v1/events", oazapfts.json({
        ...opts,
        method: "POST",
        body,
        headers: oazapfts.mergeHeaders(opts?.headers, {
            authorization,
            "x-account-id": xAccountId,
            "x-member-id": xMemberId,
            "x-username": xUsername
        })
    })));
}
/**
 * Get Upload Form
 */
export function getUploadFormV1FileUploadFormPost({ authorization, newFileUploadUrlRequest }, opts) {
    return oazapfts.ok(oazapfts.fetchJson("/v1/file-upload/form", oazapfts.json({
        ...opts,
        method: "POST",
        body: newFileUploadUrlRequest,
        headers: oazapfts.mergeHeaders(opts?.headers, {
            authorization
        })
    })));
}
/**
 * Get File Upload Status
 */
export function getFileUploadStatusV1FileUploadFileUploadIdGet({ fileUploadId, authorization }, opts) {
    return oazapfts.ok(oazapfts.fetchJson(`/v1/file-upload/${encodeURIComponent(fileUploadId)}`, {
        ...opts,
        headers: oazapfts.mergeHeaders(opts?.headers, {
            authorization
        })
    }));
}
/**
 * Report Health
 */
export function reportHealthHealthzGet(opts) {
    return oazapfts.ok(oazapfts.fetchJson("/healthz", {
        ...opts
    }));
}
/**
 * Create Knowledge Source
 */
export function createKnowledgeSourceV1KnowledgeSourcesPost({ authorization, newKnowledgeSourceRequest }, opts) {
    return oazapfts.ok(oazapfts.fetchJson("/v1/knowledge-sources", oazapfts.json({
        ...opts,
        method: "POST",
        body: newKnowledgeSourceRequest,
        headers: oazapfts.mergeHeaders(opts?.headers, {
            authorization
        })
    })));
}
/**
 * List Knowledge Sources
 */
export function listKnowledgeSourcesV1KnowledgeSourcesGet({ visibility, order, $default, types, authorization, xAccountId, xMemberId, xUsername }, opts) {
    return oazapfts.ok(oazapfts.fetchJson(`/v1/knowledge-sources${QS.query(QS.explode({
        visibility,
        order,
        "default": $default,
        types
    }))}`, {
        ...opts,
        headers: oazapfts.mergeHeaders(opts?.headers, {
            authorization,
            "x-account-id": xAccountId,
            "x-member-id": xMemberId,
            "x-username": xUsername
        })
    }));
}
/**
 * Find Knowledge Source
 */
export function findKnowledgeSourceV1KnowledgeSourcesSlugGet({ slug, authorization }, opts) {
    return oazapfts.ok(oazapfts.fetchJson(`/v1/knowledge-sources/${encodeURIComponent(slug)}`, {
        ...opts,
        headers: oazapfts.mergeHeaders(opts?.headers, {
            authorization
        })
    }));
}
/**
 * Update Knowledge Source
 */
export function updateKnowledgeSourceV1KnowledgeSourcesSlugPatch({ slug, authorization, knowledgeSourcePatchesRequest }, opts) {
    return oazapfts.ok(oazapfts.fetchJson(`/v1/knowledge-sources/${encodeURIComponent(slug)}`, oazapfts.json({
        ...opts,
        method: "PATCH",
        body: knowledgeSourcePatchesRequest,
        headers: oazapfts.mergeHeaders(opts?.headers, {
            authorization
        })
    })));
}
/**
 * Delete Knowledge Source
 */
export function deleteKnowledgeSourceV1KnowledgeSourcesSlugDelete({ slug, authorization, xAccountId, xMemberId, xUsername }, opts) {
    return oazapfts.ok(oazapfts.fetchJson(`/v1/knowledge-sources/${encodeURIComponent(slug)}`, {
        ...opts,
        method: "DELETE",
        headers: oazapfts.mergeHeaders(opts?.headers, {
            authorization,
            "x-account-id": xAccountId,
            "x-member-id": xMemberId,
            "x-username": xUsername
        })
    }));
}
/**
 * Exists Knowledge Source
 */
export function existsKnowledgeSourceV1KnowledgeSourcesSlugExistsGet({ slug, authorization }, opts) {
    return oazapfts.ok(oazapfts.fetchJson(`/v1/knowledge-sources/${encodeURIComponent(slug)}/exists`, {
        ...opts,
        headers: oazapfts.mergeHeaders(opts?.headers, {
            authorization
        })
    }));
}
/**
 * Search
 */
export function searchV1KnowledgeSourcesSlugSimilaritySearchGet({ slug, q, size, authorization, xAccountId, xMemberId, xUsername }, opts) {
    return oazapfts.ok(oazapfts.fetchJson(`/v1/knowledge-sources/${encodeURIComponent(slug)}/similarity-search${QS.query(QS.explode({
        q,
        size
    }))}`, {
        ...opts,
        headers: oazapfts.mergeHeaders(opts?.headers, {
            authorization,
            "x-account-id": xAccountId,
            "x-member-id": xMemberId,
            "x-username": xUsername
        })
    }));
}
/**
 * Share Knowledge Source
 */
export function shareKnowledgeSourceV1KnowledgeSourcesSlugSharePost({ slug, authorization }, opts) {
    return oazapfts.ok(oazapfts.fetchJson(`/v1/knowledge-sources/${encodeURIComponent(slug)}/share`, {
        ...opts,
        method: "POST",
        headers: oazapfts.mergeHeaders(opts?.headers, {
            authorization
        })
    }));
}
/**
 * Publish Knowledge Source
 */
export function publishKnowledgeSourceV1KnowledgeSourcesSlugPublishPost({ slug, authorization, knowledgeSourcesPublishRequest }, opts) {
    return oazapfts.ok(oazapfts.fetchJson(`/v1/knowledge-sources/${encodeURIComponent(slug)}/publish`, oazapfts.json({
        ...opts,
        method: "POST",
        body: knowledgeSourcesPublishRequest,
        headers: oazapfts.mergeHeaders(opts?.headers, {
            authorization
        })
    })));
}
/**
 * Publish Knowledge Source In Batch
 */
export function publishKnowledgeSourceInBatchV1KnowledgeSourcesPublishBatchPost({ authorization, knowSourcePublishBatchRequest }, opts) {
    return oazapfts.ok(oazapfts.fetchJson("/v1/knowledge-sources/publish/batch", oazapfts.json({
        ...opts,
        method: "POST",
        body: knowSourcePublishBatchRequest,
        headers: oazapfts.mergeHeaders(opts?.headers, {
            authorization
        })
    })));
}
/**
 * Fork Knowledge Source
 */
export function forkKnowledgeSourceV1KnowledgeSourcesSlugForkPost({ slug, authorization, forkKnowledgeSourceRequest }, opts) {
    return oazapfts.ok(oazapfts.fetchJson(`/v1/knowledge-sources/${encodeURIComponent(slug)}/fork`, oazapfts.json({
        ...opts,
        method: "POST",
        body: forkKnowledgeSourceRequest,
        headers: oazapfts.mergeHeaders(opts?.headers, {
            authorization
        })
    })));
}
/**
 * List Knowledge Objects
 */
export function listKnowledgeObjectsV1KnowledgeSourcesSlugObjectsGet({ slug, standalone, authorization }, opts) {
    return oazapfts.ok(oazapfts.fetchJson(`/v1/knowledge-sources/${encodeURIComponent(slug)}/objects${QS.query(QS.explode({
        standalone
    }))}`, {
        ...opts,
        headers: oazapfts.mergeHeaders(opts?.headers, {
            authorization
        })
    }));
}
/**
 * Reset Knowledge Objects
 */
export function resetKnowledgeObjectsV1KnowledgeSourcesSlugObjectsDelete({ slug, standalone, authorization }, opts) {
    return oazapfts.ok(oazapfts.fetchJson(`/v1/knowledge-sources/${encodeURIComponent(slug)}/objects${QS.query(QS.explode({
        standalone
    }))}`, {
        ...opts,
        method: "DELETE",
        headers: oazapfts.mergeHeaders(opts?.headers, {
            authorization
        })
    }));
}
/**
 * Find Knowledge Object By Custom Id
 */
export function findKnowledgeObjectByCustomIdV1KnowledgeSourcesSlugObjectsCustomIdGet({ slug, customId, authorization }, opts) {
    return oazapfts.ok(oazapfts.fetchJson(`/v1/knowledge-sources/${encodeURIComponent(slug)}/objects/${encodeURIComponent(customId)}`, {
        ...opts,
        headers: oazapfts.mergeHeaders(opts?.headers, {
            authorization
        })
    }));
}
/**
 * Delete Knowledge Object By Custom Id
 */
export function deleteKnowledgeObjectByCustomIdV1KnowledgeSourcesSlugObjectsCustomIdDelete({ customId, slug, authorization }, opts) {
    return oazapfts.ok(oazapfts.fetchJson(`/v1/knowledge-sources/${encodeURIComponent(slug)}/objects/${encodeURIComponent(customId)}`, {
        ...opts,
        method: "DELETE",
        headers: oazapfts.mergeHeaders(opts?.headers, {
            authorization
        })
    }));
}
/**
 * Find Snippet Doc By Custom Id
 */
export function findSnippetDocByCustomIdV1KnowledgeSourcesSlugSnippetsIdGet({ slug, id, authorization }, opts) {
    return oazapfts.ok(oazapfts.fetchJson(`/v1/knowledge-sources/${encodeURIComponent(slug)}/snippets/${encodeURIComponent(id)}`, {
        ...opts,
        headers: oazapfts.mergeHeaders(opts?.headers, {
            authorization
        })
    }));
}
/**
 * Vectorize Snippet Knowledge Source
 */
export function vectorizeSnippetKnowledgeSourceV1KnowledgeSourcesSlugSnippetsPost({ slug, authorization, xAccountId, xMemberId, xUsername, snippetKnowledgeSourceRequest }, opts) {
    return oazapfts.ok(oazapfts.fetchJson(`/v1/knowledge-sources/${encodeURIComponent(slug)}/snippets`, oazapfts.json({
        ...opts,
        method: "POST",
        body: snippetKnowledgeSourceRequest,
        headers: oazapfts.mergeHeaders(opts?.headers, {
            authorization,
            "x-account-id": xAccountId,
            "x-member-id": xMemberId,
            "x-username": xUsername
        })
    })));
}
/**
 * Find Api Doc By Custom Id
 */
export function findApiDocByCustomIdV1KnowledgeSourcesSlugApisIdGet({ slug, id, authorization }, opts) {
    return oazapfts.ok(oazapfts.fetchJson(`/v1/knowledge-sources/${encodeURIComponent(slug)}/apis/${encodeURIComponent(id)}`, {
        ...opts,
        headers: oazapfts.mergeHeaders(opts?.headers, {
            authorization
        })
    }));
}
/**
 * Find Event Doc By Custom Id
 */
export function findEventDocByCustomIdV1KnowledgeSourcesSlugEventsIdGet({ slug, id, authorization }, opts) {
    return oazapfts.ok(oazapfts.fetchJson(`/v1/knowledge-sources/${encodeURIComponent(slug)}/events/${encodeURIComponent(id)}`, {
        ...opts,
        headers: oazapfts.mergeHeaders(opts?.headers, {
            authorization
        })
    }));
}
/**
 * Vectorize Event Knowledge Source
 */
export function vectorizeEventKnowledgeSourceV1KnowledgeSourcesSlugEventsPost({ slug, authorization, xAccountId, xMemberId, xUsername, body }, opts) {
    return oazapfts.ok(oazapfts.fetchJson(`/v1/knowledge-sources/${encodeURIComponent(slug)}/events`, oazapfts.json({
        ...opts,
        method: "POST",
        body,
        headers: oazapfts.mergeHeaders(opts?.headers, {
            authorization,
            "x-account-id": xAccountId,
            "x-member-id": xMemberId,
            "x-username": xUsername
        })
    })));
}
/**
 * Find Custom Doc By Custom Id
 */
export function findCustomDocByCustomIdV1KnowledgeSourcesSlugCustomIdGet({ slug, id, authorization }, opts) {
    return oazapfts.ok(oazapfts.fetchJson(`/v1/knowledge-sources/${encodeURIComponent(slug)}/custom/${encodeURIComponent(id)}`, {
        ...opts,
        headers: oazapfts.mergeHeaders(opts?.headers, {
            authorization
        })
    }));
}
/**
 * Vectorize Custom Knowledge Source
 */
export function vectorizeCustomKnowledgeSourceV1KnowledgeSourcesSlugCustomPost({ slug, authorization, xAccountId, xMemberId, xUsername, customKnowledgeSourceRequest }, opts) {
    return oazapfts.ok(oazapfts.fetchJson(`/v1/knowledge-sources/${encodeURIComponent(slug)}/custom`, oazapfts.json({
        ...opts,
        method: "POST",
        body: customKnowledgeSourceRequest,
        headers: oazapfts.mergeHeaders(opts?.headers, {
            authorization,
            "x-account-id": xAccountId,
            "x-member-id": xMemberId,
            "x-username": xUsername
        })
    })));
}
/**
 * Sync Embedding Model
 */
export function syncEmbeddingModelV1KnowledgeSourcesSyncPost(opts) {
    return oazapfts.ok(oazapfts.fetchJson("/v1/knowledge-sources/sync", {
        ...opts,
        method: "POST"
    }));
}
/**
 * Search Knowledge Sources
 */
export function searchKnowledgeSourcesV1KnowledgeSourcesSearchPost({ authorization, searchKnowledgeSourcesRequest }, opts) {
    return oazapfts.ok(oazapfts.fetchJson("/v1/knowledge-sources/search", oazapfts.json({
        ...opts,
        method: "POST",
        body: searchKnowledgeSourcesRequest,
        headers: oazapfts.mergeHeaders(opts?.headers, {
            authorization
        })
    })));
}
/**
 * Change Llm
 */
export function changeLlmV1AccountsLlmPatch({ authorization, xAccountId, xMemberId, xUsername, accountSettingsChangeLlmRequest }, opts) {
    return oazapfts.ok(oazapfts.fetchJson("/v1/accounts/llm", oazapfts.json({
        ...opts,
        method: "PATCH",
        body: accountSettingsChangeLlmRequest,
        headers: oazapfts.mergeHeaders(opts?.headers, {
            authorization,
            "x-account-id": xAccountId,
            "x-member-id": xMemberId,
            "x-username": xUsername
        })
    })));
}
/**
 * Change Limit
 */
export function changeLimitV1AccountsTokenLimitsPut({ authorization, xAccountId, xMemberId, xUsername, accountSettingsChangeLimitRequest }, opts) {
    return oazapfts.ok(oazapfts.fetchJson("/v1/accounts/token-limits", oazapfts.json({
        ...opts,
        method: "PUT",
        body: accountSettingsChangeLimitRequest,
        headers: oazapfts.mergeHeaders(opts?.headers, {
            authorization,
            "x-account-id": xAccountId,
            "x-member-id": xMemberId,
            "x-username": xUsername
        })
    })));
}
/**
 * Reset Limit
 */
export function resetLimitV1AccountsTokenLimitsDelete({ authorization, xAccountId, xMemberId, xUsername }, opts) {
    return oazapfts.ok(oazapfts.fetchJson("/v1/accounts/token-limits", {
        ...opts,
        method: "DELETE",
        headers: oazapfts.mergeHeaders(opts?.headers, {
            authorization,
            "x-account-id": xAccountId,
            "x-member-id": xMemberId,
            "x-username": xUsername
        })
    }));
}
/**
 * Change External Rqc
 */
export function changeExternalRqcV1AccountsExternalRqcPatch({ authorization, xAccountId, xMemberId, xUsername, accountSettingsChangeErqcRequest }, opts) {
    return oazapfts.ok(oazapfts.fetchJson("/v1/accounts/external-rqc", oazapfts.json({
        ...opts,
        method: "PATCH",
        body: accountSettingsChangeErqcRequest,
        headers: oazapfts.mergeHeaders(opts?.headers, {
            authorization,
            "x-account-id": xAccountId,
            "x-member-id": xMemberId,
            "x-username": xUsername
        })
    })));
}
/**
 * Get External Configs
 */
export function getExternalConfigsV1AccountsExternalConfigGet(opts) {
    return oazapfts.ok(oazapfts.fetchJson("/v1/accounts/external-config", {
        ...opts
    }));
}
/**
 * Change External Configs
 */
export function changeExternalConfigsV1AccountsExternalConfigPatch({ body }, opts) {
    return oazapfts.ok(oazapfts.fetchJson("/v1/accounts/external-config", oazapfts.json({
        ...opts,
        method: "PATCH",
        body
    })));
}
/**
 * Tokens Usage
 */
export function tokensUsageV1AccountsTokensUsageGet({ authorization, xAccountId, xMemberId, xUsername }, opts) {
    return oazapfts.ok(oazapfts.fetchJson("/v1/accounts/tokens-usage", {
        ...opts,
        headers: oazapfts.mergeHeaders(opts?.headers, {
            authorization,
            "x-account-id": xAccountId,
            "x-member-id": xMemberId,
            "x-username": xUsername
        })
    }));
}
/**
 * Total
 */
export function totalV1TokensUsageTotalGet({ authorization, xAccountId, xMemberId, xUsername }, opts) {
    return oazapfts.ok(oazapfts.fetchJson("/v1/tokens-usage/total", {
        ...opts,
        headers: oazapfts.mergeHeaders(opts?.headers, {
            authorization,
            "x-account-id": xAccountId,
            "x-member-id": xMemberId,
            "x-username": xUsername
        })
    }));
}
/**
 * Current
 */
export function currentV1TokensUsageCurrentGet({ year, month, authorization, xAccountId, xMemberId, xUsername }, opts) {
    return oazapfts.ok(oazapfts.fetchJson(`/v1/tokens-usage/current${QS.query(QS.explode({
        year,
        month
    }))}`, {
        ...opts,
        headers: oazapfts.mergeHeaders(opts?.headers, {
            authorization,
            "x-account-id": xAccountId,
            "x-member-id": xMemberId,
            "x-username": xUsername
        })
    }));
}
/**
 * Monthly
 */
export function monthlyV1TokensUsageMonthlyGet({ year, authorization, xAccountId, xMemberId, xUsername }, opts) {
    return oazapfts.ok(oazapfts.fetchJson(`/v1/tokens-usage/monthly${QS.query(QS.explode({
        year
    }))}`, {
        ...opts,
        headers: oazapfts.mergeHeaders(opts?.headers, {
            authorization,
            "x-account-id": xAccountId,
            "x-member-id": xMemberId,
            "x-username": xUsername
        })
    }));
}
/**
 * Top Users
 */
export function topUsersV1TokensUsageTopUsersGet({ year, month, authorization, xAccountId, xMemberId, xUsername }, opts) {
    return oazapfts.ok(oazapfts.fetchJson(`/v1/tokens-usage/top-users${QS.query(QS.explode({
        year,
        month
    }))}`, {
        ...opts,
        headers: oazapfts.mergeHeaders(opts?.headers, {
            authorization,
            "x-account-id": xAccountId,
            "x-member-id": xMemberId,
            "x-username": xUsername
        })
    }));
}
/**
 * Add Association
 */
export function addAssociationV1WorkspaceWorkspaceIdPost({ workspaceId, authorization, xAccountId, xMemberId, xUsername, addWorkspaceKnowledgeSourceRequest }, opts) {
    return oazapfts.ok(oazapfts.fetchJson(`/v1/workspace/${encodeURIComponent(workspaceId)}`, oazapfts.json({
        ...opts,
        method: "POST",
        body: addWorkspaceKnowledgeSourceRequest,
        headers: oazapfts.mergeHeaders(opts?.headers, {
            authorization,
            "x-account-id": xAccountId,
            "x-member-id": xMemberId,
            "x-username": xUsername
        })
    })));
}
/**
 * List Association
 */
export function listAssociationV1WorkspaceWorkspaceIdGet({ workspaceId, authorization, xAccountId, xMemberId, xUsername }, opts) {
    return oazapfts.ok(oazapfts.fetchJson(`/v1/workspace/${encodeURIComponent(workspaceId)}`, {
        ...opts,
        headers: oazapfts.mergeHeaders(opts?.headers, {
            authorization,
            "x-account-id": xAccountId,
            "x-member-id": xMemberId,
            "x-username": xUsername
        })
    }));
}
/**
 * Delete Association
 */
export function deleteAssociationV1WorkspaceWorkspaceIdKnowledgeSourceKnowledgeSourceSlugDelete({ workspaceId, knowledgeSourceSlug, authorization, xAccountId, xMemberId, xUsername }, opts) {
    return oazapfts.ok(oazapfts.fetchJson(`/v1/workspace/${encodeURIComponent(workspaceId)}/knowledge_source/${encodeURIComponent(knowledgeSourceSlug)}`, {
        ...opts,
        method: "DELETE",
        headers: oazapfts.mergeHeaders(opts?.headers, {
            authorization,
            "x-account-id": xAccountId,
            "x-member-id": xMemberId,
            "x-username": xUsername
        })
    }));
}
/**
 * Workspace Fork
 */
export function workspaceForkV1QuickCommandsWorkspaceForkPost({ authorization, xAccountId, xMemberId, xUsername, quickCommandWorkspaceForkRequest }, opts) {
    return oazapfts.ok(oazapfts.fetchJson("/v1/quick-commands/workspace/fork", oazapfts.json({
        ...opts,
        method: "POST",
        body: quickCommandWorkspaceForkRequest,
        headers: oazapfts.mergeHeaders(opts?.headers, {
            authorization,
            "x-account-id": xAccountId,
            "x-member-id": xMemberId,
            "x-username": xUsername
        })
    })));
}
/**
 * Workspace Delete Fork
 */
export function workspaceDeleteForkV1QuickCommandsWorkspaceForkDelete({ authorization, xAccountId, xMemberId, xUsername, quickCommandWorkspaceDeleteRequest }, opts) {
    return oazapfts.ok(oazapfts.fetchJson("/v1/quick-commands/workspace/fork", oazapfts.json({
        ...opts,
        method: "DELETE",
        body: quickCommandWorkspaceDeleteRequest,
        headers: oazapfts.mergeHeaders(opts?.headers, {
            authorization,
            "x-account-id": xAccountId,
            "x-member-id": xMemberId,
            "x-username": xUsername
        })
    })));
}
/**
 * Workspace List
 */
export function workspaceListV1QuickCommandsWorkspaceListPost({ authorization, xAccountId, xMemberId, xUsername, quickCommandWorkspaceListRequest }, opts) {
    return oazapfts.ok(oazapfts.fetchJson("/v1/quick-commands/workspace/list", oazapfts.json({
        ...opts,
        method: "POST",
        body: quickCommandWorkspaceListRequest,
        headers: oazapfts.mergeHeaders(opts?.headers, {
            authorization,
            "x-account-id": xAccountId,
            "x-member-id": xMemberId,
            "x-username": xUsername
        })
    })));
}
/**
 * Create Quick Command
 */
export function createQuickCommandV1QuickCommandsPost({ authorization, xAccountId, xMemberId, xUsername, quickCommandsCreateRequest }, opts) {
    return oazapfts.ok(oazapfts.fetchJson("/v1/quick-commands", oazapfts.json({
        ...opts,
        method: "POST",
        body: quickCommandsCreateRequest,
        headers: oazapfts.mergeHeaders(opts?.headers, {
            authorization,
            "x-account-id": xAccountId,
            "x-member-id": xMemberId,
            "x-username": xUsername
        })
    })));
}
/**
 * List All
 */
export function listAllV1QuickCommandsAllGet({ visibility, order, types, authorization, xAccountId, xMemberId, xUsername }, opts) {
    return oazapfts.ok(oazapfts.fetchJson(`/v1/quick-commands/all${QS.query(QS.explode({
        visibility,
        order,
        types
    }))}`, {
        ...opts,
        headers: oazapfts.mergeHeaders(opts?.headers, {
            authorization,
            "x-account-id": xAccountId,
            "x-member-id": xMemberId,
            "x-username": xUsername
        })
    }));
}
/**
 * Update Quick Command
 */
export function updateQuickCommandV1QuickCommandsSlugPatch({ slug, authorization, xAccountId, xMemberId, xUsername, quickCommandsUpdateRequest }, opts) {
    return oazapfts.ok(oazapfts.fetchJson(`/v1/quick-commands/${encodeURIComponent(slug)}`, oazapfts.json({
        ...opts,
        method: "PATCH",
        body: quickCommandsUpdateRequest,
        headers: oazapfts.mergeHeaders(opts?.headers, {
            authorization,
            "x-account-id": xAccountId,
            "x-member-id": xMemberId,
            "x-username": xUsername
        })
    })));
}
/**
 * Get Quick Command
 */
export function getQuickCommandV1QuickCommandsSlugGet({ slug, findAgents, authorization, xAccountId, xMemberId, xUsername }, opts) {
    return oazapfts.ok(oazapfts.fetchJson(`/v1/quick-commands/${encodeURIComponent(slug)}${QS.query(QS.explode({
        find_agents: findAgents
    }))}`, {
        ...opts,
        headers: oazapfts.mergeHeaders(opts?.headers, {
            authorization,
            "x-account-id": xAccountId,
            "x-member-id": xMemberId,
            "x-username": xUsername
        })
    }));
}
/**
 * Delete Quick Command
 */
export function deleteQuickCommandV1QuickCommandsSlugDelete({ slug, authorization, xAccountId, xMemberId, xUsername }, opts) {
    return oazapfts.ok(oazapfts.fetchJson(`/v1/quick-commands/${encodeURIComponent(slug)}`, {
        ...opts,
        method: "DELETE",
        headers: oazapfts.mergeHeaders(opts?.headers, {
            authorization,
            "x-account-id": xAccountId,
            "x-member-id": xMemberId,
            "x-username": xUsername
        })
    }));
}
/**
 * Add Favorite
 */
export function addFavoriteV1QuickCommandsSlugFavoritePost({ slug, authorization, xAccountId, xMemberId, xUsername }, opts) {
    return oazapfts.ok(oazapfts.fetchJson(`/v1/quick-commands/${encodeURIComponent(slug)}/favorite`, {
        ...opts,
        method: "POST",
        headers: oazapfts.mergeHeaders(opts?.headers, {
            authorization,
            "x-account-id": xAccountId,
            "x-member-id": xMemberId,
            "x-username": xUsername
        })
    }));
}
/**
 * Delete Favorite
 */
export function deleteFavoriteV1QuickCommandsSlugFavoriteDelete({ slug, authorization, xAccountId, xMemberId, xUsername }, opts) {
    return oazapfts.ok(oazapfts.fetchJson(`/v1/quick-commands/${encodeURIComponent(slug)}/favorite`, {
        ...opts,
        method: "DELETE",
        headers: oazapfts.mergeHeaders(opts?.headers, {
            authorization,
            "x-account-id": xAccountId,
            "x-member-id": xMemberId,
            "x-username": xUsername
        })
    }));
}
/**
 * Share
 */
export function shareV1QuickCommandsSlugSharePost({ slug, authorization, xAccountId, xMemberId, xUsername }, opts) {
    return oazapfts.ok(oazapfts.fetchJson(`/v1/quick-commands/${encodeURIComponent(slug)}/share`, {
        ...opts,
        method: "POST",
        headers: oazapfts.mergeHeaders(opts?.headers, {
            authorization,
            "x-account-id": xAccountId,
            "x-member-id": xMemberId,
            "x-username": xUsername
        })
    }));
}
/**
 * Publish
 */
export function publishV1QuickCommandsSlugPublishPost({ slug, authorization, xAccountId, xMemberId, xUsername, quickCommandPublishRequest }, opts) {
    return oazapfts.ok(oazapfts.fetchJson(`/v1/quick-commands/${encodeURIComponent(slug)}/publish`, oazapfts.json({
        ...opts,
        method: "POST",
        body: quickCommandPublishRequest,
        headers: oazapfts.mergeHeaders(opts?.headers, {
            authorization,
            "x-account-id": xAccountId,
            "x-member-id": xMemberId,
            "x-username": xUsername
        })
    })));
}
/**
 * Fork
 */
export function forkV1QuickCommandsSlugForkPost({ slug, authorization, xAccountId, xMemberId, xUsername, quickCommandsMakeACopyRequest }, opts) {
    return oazapfts.ok(oazapfts.fetchJson(`/v1/quick-commands/${encodeURIComponent(slug)}/fork`, oazapfts.json({
        ...opts,
        method: "POST",
        body: quickCommandsMakeACopyRequest,
        headers: oazapfts.mergeHeaders(opts?.headers, {
            authorization,
            "x-account-id": xAccountId,
            "x-member-id": xMemberId,
            "x-username": xUsername
        })
    })));
}
/**
 * Check Quick Command Exists
 */
export function checkQuickCommandExistsV1QuickCommandsSlugExistsGet({ slug, authorization, xAccountId, xMemberId, xUsername }, opts) {
    return oazapfts.ok(oazapfts.fetchJson(`/v1/quick-commands/${encodeURIComponent(slug)}/exists`, {
        ...opts,
        headers: oazapfts.mergeHeaders(opts?.headers, {
            authorization,
            "x-account-id": xAccountId,
            "x-member-id": xMemberId,
            "x-username": xUsername
        })
    }));
}
/**
 * Get Quick Command By Ks Slug
 */
export function getQuickCommandByKsSlugV1QuickCommandsKnowledgeSourcesSlugGet({ slug, authorization, xAccountId, xMemberId, xUsername }, opts) {
    return oazapfts.ok(oazapfts.fetchJson(`/v1/quick-commands/knowledge-sources/${encodeURIComponent(slug)}`, {
        ...opts,
        headers: oazapfts.mergeHeaders(opts?.headers, {
            authorization,
            "x-account-id": xAccountId,
            "x-member-id": xMemberId,
            "x-username": xUsername
        })
    }));
}
/**
 * Get Quick Commands By Agent Id
 */
export function getQuickCommandsByAgentIdV1QuickCommandsAgentsAgentIdGet({ agentId, authorization, xAccountId, xMemberId, xUsername }, opts) {
    return oazapfts.ok(oazapfts.fetchJson(`/v1/quick-commands/agents/${encodeURIComponent(agentId)}`, {
        ...opts,
        headers: oazapfts.mergeHeaders(opts?.headers, {
            authorization,
            "x-account-id": xAccountId,
            "x-member-id": xMemberId,
            "x-username": xUsername
        })
    }));
}
/**
 * Disassociate Agent
 */
export function disassociateAgentV1QuickCommandsAgentsAgentIdDelete({ agentId, authorization, xAccountId, xMemberId, xUsername }, opts) {
    return oazapfts.ok(oazapfts.fetchJson(`/v1/quick-commands/agents/${encodeURIComponent(agentId)}`, {
        ...opts,
        method: "DELETE",
        headers: oazapfts.mergeHeaders(opts?.headers, {
            authorization,
            "x-account-id": xAccountId,
            "x-member-id": xMemberId,
            "x-username": xUsername
        })
    }));
}
/**
 * List By Workspace Id
 */
export function listByWorkspaceIdV1QuickCommandsWorkspacesWorkspaceIdGet({ workspaceId, authorization, xAccountId, xMemberId, xUsername }, opts) {
    return oazapfts.ok(oazapfts.fetchJson(`/v1/quick-commands/workspaces/${encodeURIComponent(workspaceId)}`, {
        ...opts,
        headers: oazapfts.mergeHeaders(opts?.headers, {
            authorization,
            "x-account-id": xAccountId,
            "x-member-id": xMemberId,
            "x-username": xUsername
        })
    }));
}
/**
 * Format Fetch Step
 */
export function formatFetchStepV1QuickCommandsSlugStepsStepSlugFetchFormatPost({ slug, stepSlug, authorization, xAccountId, xMemberId, xUsername, quickCommandsExecutionRequest }, opts) {
    return oazapfts.ok(oazapfts.fetchJson(`/v1/quick-commands/${encodeURIComponent(slug)}/steps/${encodeURIComponent(stepSlug)}/fetch_format`, oazapfts.json({
        ...opts,
        method: "POST",
        body: quickCommandsExecutionRequest,
        headers: oazapfts.mergeHeaders(opts?.headers, {
            authorization,
            "x-account-id": xAccountId,
            "x-member-id": xMemberId,
            "x-username": xUsername
        })
    })));
}
/**
 * Format Result
 */
export function formatResultV1QuickCommandsSlugResultFormatPost({ slug, authorization, xAccountId, xMemberId, xUsername, quickCommandsExecutionRequest }, opts) {
    return oazapfts.ok(oazapfts.fetchJson(`/v1/quick-commands/${encodeURIComponent(slug)}/result_format`, oazapfts.json({
        ...opts,
        method: "POST",
        body: quickCommandsExecutionRequest,
        headers: oazapfts.mergeHeaders(opts?.headers, {
            authorization,
            "x-account-id": xAccountId,
            "x-member-id": xMemberId,
            "x-username": xUsername
        })
    })));
}
/**
 * Calculates the next route for a ROUTER step
 */
export function calculateNextStepV1QuickCommandsSlugStepsStepSlugCalculateNextStepPost({ slug, stepSlug, authorization, xAccountId, xMemberId, xUsername, quickCommandEvaluateStepRouterRequest }, opts) {
    return oazapfts.ok(oazapfts.fetchJson(`/v1/quick-commands/${encodeURIComponent(slug)}/steps/${encodeURIComponent(stepSlug)}/calculate_next_step`, oazapfts.json({
        ...opts,
        method: "POST",
        body: quickCommandEvaluateStepRouterRequest,
        headers: oazapfts.mergeHeaders(opts?.headers, {
            authorization,
            "x-account-id": xAccountId,
            "x-member-id": xMemberId,
            "x-username": xUsername
        })
    })));
}
/**
 * Add Workspace
 */
export function addWorkspaceV1QuickCommandsSlugWorkspacesWorkspaceIdAddPost({ slug, workspaceId, authorization, xAccountId, xMemberId, xUsername }, opts) {
    return oazapfts.ok(oazapfts.fetchJson(`/v1/quick-commands/${encodeURIComponent(slug)}/workspaces/${encodeURIComponent(workspaceId)}/add`, {
        ...opts,
        method: "POST",
        headers: oazapfts.mergeHeaders(opts?.headers, {
            authorization,
            "x-account-id": xAccountId,
            "x-member-id": xMemberId,
            "x-username": xUsername
        })
    }));
}
/**
 * Remove Workspace
 */
export function removeWorkspaceV1QuickCommandsSlugWorkspacesWorkspaceIdRemoveDelete({ slug, workspaceId, authorization, xAccountId, xMemberId, xUsername }, opts) {
    return oazapfts.ok(oazapfts.fetchJson(`/v1/quick-commands/${encodeURIComponent(slug)}/workspaces/${encodeURIComponent(workspaceId)}/remove`, {
        ...opts,
        method: "DELETE",
        headers: oazapfts.mergeHeaders(opts?.headers, {
            authorization,
            "x-account-id": xAccountId,
            "x-member-id": xMemberId,
            "x-username": xUsername
        })
    }));
}
/**
 * Create Execution
 */
export function createExecutionV1QuickCommandsCreateExecutionSlugPost({ slug, conversationId, quickCommandCreateRequest }, opts) {
    return oazapfts.ok(oazapfts.fetchJson(`/v1/quick-commands/create-execution/${encodeURIComponent(slug)}${QS.query(QS.explode({
        conversation_id: conversationId
    }))}`, oazapfts.json({
        ...opts,
        method: "POST",
        body: quickCommandCreateRequest
    })));
}
/**
 * Triggers asynchronous execution of a Remote Quick Command
 */
export function executeV1QuickCommandsExecuteExecutionIdPost({ executionId, authorization, xAccountId, xMemberId, xUsername, quickCommandExecutionRequest }, opts) {
    return oazapfts.ok(oazapfts.fetchJson(`/v1/quick-commands/execute/${encodeURIComponent(executionId)}`, oazapfts.json({
        ...opts,
        method: "POST",
        body: quickCommandExecutionRequest,
        headers: oazapfts.mergeHeaders(opts?.headers, {
            authorization,
            "x-account-id": xAccountId,
            "x-member-id": xMemberId,
            "x-username": xUsername
        })
    })));
}
/**
 * Callback
 */
export function callbackV1QuickCommandsCallbackExecutionIdGet({ executionId }, opts) {
    return oazapfts.ok(oazapfts.fetchJson(`/v1/quick-commands/callback/${encodeURIComponent(executionId)}`, {
        ...opts
    }));
}
/**
 * Run Fetch Step
 */
export function runFetchStepV1QuickCommandsSlugStepsStepSlugFetchRunPost({ slug, stepSlug, authorization, xAccountId, xMemberId, xUsername, quickCommandsExecutionRequest }, opts) {
    return oazapfts.ok(oazapfts.fetchJson(`/v1/quick-commands/${encodeURIComponent(slug)}/steps/${encodeURIComponent(stepSlug)}/fetch_run`, oazapfts.json({
        ...opts,
        method: "POST",
        body: quickCommandsExecutionRequest,
        headers: oazapfts.mergeHeaders(opts?.headers, {
            authorization,
            "x-account-id": xAccountId,
            "x-member-id": xMemberId,
            "x-username": xUsername
        })
    })));
}
/**
 * Starts the execution of a step with a user defined script
 */
export function startScriptStepV1QuickCommandsSlugStepsStepSlugStartScriptPost({ slug, stepSlug, quickCommandStartScriptRequest }, opts) {
    return oazapfts.ok(oazapfts.fetchJson(`/v1/quick-commands/${encodeURIComponent(slug)}/steps/${encodeURIComponent(stepSlug)}/start-script`, oazapfts.json({
        ...opts,
        method: "POST",
        body: quickCommandStartScriptRequest
    })));
}
/**
 * Get the status of a script execution step
 */
export function getScriptExecutionStatusV1QuickCommandsScriptExecutionsScriptExecutionIdGet({ scriptExecutionId }, opts) {
    return oazapfts.ok(oazapfts.fetchJson(`/v1/quick-commands/script-executions/${encodeURIComponent(scriptExecutionId)}`, {
        ...opts
    }));
}
/**
 * List Conversations
 */
export function listConversationsV1ConversationsGet({ size, page, authorization, xAccountId, xMemberId, xUsername }, opts) {
    return oazapfts.ok(oazapfts.fetchJson(`/v1/conversations${QS.query(QS.explode({
        size,
        page
    }))}`, {
        ...opts,
        headers: oazapfts.mergeHeaders(opts?.headers, {
            authorization,
            "x-account-id": xAccountId,
            "x-member-id": xMemberId,
            "x-username": xUsername
        })
    }));
}
/**
 * Conversation History
 */
export function conversationHistoryV1ConversationsConversationIdGet({ conversationId, authorization, xAccountId, xMemberId, xUsername }, opts) {
    return oazapfts.ok(oazapfts.fetchJson(`/v1/conversations/${encodeURIComponent(conversationId)}`, {
        ...opts,
        headers: oazapfts.mergeHeaders(opts?.headers, {
            authorization,
            "x-account-id": xAccountId,
            "x-member-id": xMemberId,
            "x-username": xUsername
        })
    }));
}
/**
 * Update Title
 */
export function updateTitleV1ConversationsConversationIdPatch({ conversationId, authorization, xAccountId, xMemberId, xUsername, conversationUpdateTitleRequest }, opts) {
    return oazapfts.ok(oazapfts.fetchJson(`/v1/conversations/${encodeURIComponent(conversationId)}`, oazapfts.json({
        ...opts,
        method: "PATCH",
        body: conversationUpdateTitleRequest,
        headers: oazapfts.mergeHeaders(opts?.headers, {
            authorization,
            "x-account-id": xAccountId,
            "x-member-id": xMemberId,
            "x-username": xUsername
        })
    })));
}
/**
 * Delete Conversation
 */
export function deleteConversationV1ConversationsConversationIdDelete({ conversationId, authorization, xAccountId, xMemberId, xUsername }, opts) {
    return oazapfts.ok(oazapfts.fetchJson(`/v1/conversations/${encodeURIComponent(conversationId)}`, {
        ...opts,
        method: "DELETE",
        headers: oazapfts.mergeHeaders(opts?.headers, {
            authorization,
            "x-account-id": xAccountId,
            "x-member-id": xMemberId,
            "x-username": xUsername
        })
    }));
}
/**
 * Download Conversation
 */
export function downloadConversationV1ConversationsConversationIdDownloadGet({ conversationId, authorization, xAccountId, xMemberId, xUsername }, opts) {
    return oazapfts.ok(oazapfts.fetchJson(`/v1/conversations/${encodeURIComponent(conversationId)}/download`, {
        ...opts,
        headers: oazapfts.mergeHeaders(opts?.headers, {
            authorization,
            "x-account-id": xAccountId,
            "x-member-id": xMemberId,
            "x-username": xUsername
        })
    }));
}
/**
 * Create Knowledge Source
 */
export function createKnowledgeSourceV1DefaultKnowledgeSourcesPost({ authorization, newKnowledgeSourceRequest }, opts) {
    return oazapfts.ok(oazapfts.fetchJson("/v1/default-knowledge-sources", oazapfts.json({
        ...opts,
        method: "POST",
        body: newKnowledgeSourceRequest,
        headers: oazapfts.mergeHeaders(opts?.headers, {
            authorization
        })
    })));
}
/**
 * Delete Knowledge Source
 */
export function deleteKnowledgeSourceV1DefaultKnowledgeSourcesSlugDelete({ slug, authorization }, opts) {
    return oazapfts.ok(oazapfts.fetchJson(`/v1/default-knowledge-sources/${encodeURIComponent(slug)}`, {
        ...opts,
        method: "DELETE",
        headers: oazapfts.mergeHeaders(opts?.headers, {
            authorization
        })
    }));
}
/**
 * Upload Knowledge Objects Zip
 */
export function uploadKnowledgeObjectsZipV1DefaultKnowledgeSourcesSlugObjectsBatchPost({ slug, autoDelete, bodyUploadKnowledgeObjectsZipV1DefaultKnowledgeSourcesSlugObjectsBatchPost }, opts) {
    return oazapfts.ok(oazapfts.fetchJson(`/v1/default-knowledge-sources/${encodeURIComponent(slug)}/objects/batch${QS.query(QS.explode({
        "auto-delete": autoDelete
    }))}`, oazapfts.multipart({
        ...opts,
        method: "POST",
        body: bodyUploadKnowledgeObjectsZipV1DefaultKnowledgeSourcesSlugObjectsBatchPost
    })));
}
/**
 * Find Knowledge Source Dependencies
 */
export function findKnowledgeSourceDependenciesV1KnowledgeSourcesSlugDependenciesGet({ slug, authorization, xAccountId, xMemberId, xUsername }, opts) {
    return oazapfts.ok(oazapfts.fetchJson(`/v1/knowledge-sources/${encodeURIComponent(slug)}/dependencies`, {
        ...opts,
        headers: oazapfts.mergeHeaders(opts?.headers, {
            authorization,
            "x-account-id": xAccountId,
            "x-member-id": xMemberId,
            "x-username": xUsername
        })
    }));
}
/**
 * Get Flags
 */
export function getFlagsV1FlagsGet(opts) {
    return oazapfts.ok(oazapfts.fetchJson("/v1/flags", {
        ...opts
    }));
}
/**
 * Get Content Dependencies
 */
export function getContentDependenciesV1ContentContentTypeContentIdDependenciesGet({ contentType, contentId, authorization, xAccountId, xMemberId, xUsername }, opts) {
    return oazapfts.ok(oazapfts.fetchJson(`/v1/content/${encodeURIComponent(contentType)}/${encodeURIComponent(contentId)}/dependencies`, {
        ...opts,
        headers: oazapfts.mergeHeaders(opts?.headers, {
            authorization,
            "x-account-id": xAccountId,
            "x-member-id": xMemberId,
            "x-username": xUsername
        })
    }));
}
/**
 * Tokens Daily Usage
 */
export function tokensDailyUsageV1AnalyticsTokensDailyUsageGet({ month, year, accountSlug }, opts) {
    return oazapfts.ok(oazapfts.fetchJson(`/v1/analytics/tokens/daily-usage${QS.query(QS.explode({
        month,
        year,
        account_slug: accountSlug
    }))}`, {
        ...opts
    }));
}
/**
 * Tokens By User
 */
export function tokensByUserV1AnalyticsTokensByUserGet({ month, year, page, pageSize }, opts) {
    return oazapfts.ok(oazapfts.fetchJson(`/v1/analytics/tokens/by-user${QS.query(QS.explode({
        month,
        year,
        page,
        page_size: pageSize
    }))}`, {
        ...opts
    }));
}
/**
 * Get Reviews By Resource
 */
export function getReviewsByResourceV1ResourcesResourceTypeSlugResourceSlugReviewsGet({ resourceSlug, resourceTypeSlug, size, page, sort, direction, search, filterMode, filterBy, filterValue, multiFilterMode, filterIn, authorization, xAccountId, xMemberId, xUsername }, opts) {
    return oazapfts.ok(oazapfts.fetchJson(`/v1/resources/${encodeURIComponent(resourceTypeSlug)}/${encodeURIComponent(resourceSlug)}/reviews${QS.query(QS.explode({
        size,
        page,
        sort,
        direction,
        search,
        filterMode,
        filterBy,
        filterValue,
        multiFilterMode,
        filterIn
    }))}`, {
        ...opts,
        headers: oazapfts.mergeHeaders(opts?.headers, {
            authorization,
            "x-account-id": xAccountId,
            "x-member-id": xMemberId,
            "x-username": xUsername
        })
    }));
}
/**
 * Create Resource Review
 */
export function createResourceReviewV1ResourcesResourceTypeSlugResourceSlugReviewsPost({ resourceSlug, resourceTypeSlug, authorization, xAccountId, xMemberId, xUsername, resourceReviewRequest }, opts) {
    return oazapfts.ok(oazapfts.fetchJson(`/v1/resources/${encodeURIComponent(resourceTypeSlug)}/${encodeURIComponent(resourceSlug)}/reviews`, oazapfts.json({
        ...opts,
        method: "POST",
        body: resourceReviewRequest,
        headers: oazapfts.mergeHeaders(opts?.headers, {
            authorization,
            "x-account-id": xAccountId,
            "x-member-id": xMemberId,
            "x-username": xUsername
        })
    })));
}
/**
 * Create Answer For Review
 */
export function createAnswerForReviewV1ResourcesResourceTypeSlugResourceSlugReviewsReviewIdAnswersPost({ reviewId, authorization, xAccountId, xMemberId, xUsername, reviewAnswer }, opts) {
    return oazapfts.ok(oazapfts.fetchJson(`/v1/resources/${encodeURIComponent(resourceTypeSlug)}/${encodeURIComponent(resourceSlug)}/reviews/${encodeURIComponent(reviewId)}/answers`, oazapfts.json({
        ...opts,
        method: "POST",
        body: reviewAnswer,
        headers: oazapfts.mergeHeaders(opts?.headers, {
            authorization,
            "x-account-id": xAccountId,
            "x-member-id": xMemberId,
            "x-username": xUsername
        })
    })));
}
/**
 * Update Resource Review
 */
export function updateResourceReviewV1ResourcesResourceTypeSlugResourceSlugReviewsReviewIdPatch({ reviewId, authorization, xAccountId, xMemberId, xUsername, resourceReviewRequest }, opts) {
    return oazapfts.ok(oazapfts.fetchJson(`/v1/resources/${encodeURIComponent(resourceTypeSlug)}/${encodeURIComponent(resourceSlug)}/reviews/${encodeURIComponent(reviewId)}`, oazapfts.json({
        ...opts,
        method: "PATCH",
        body: resourceReviewRequest,
        headers: oazapfts.mergeHeaders(opts?.headers, {
            authorization,
            "x-account-id": xAccountId,
            "x-member-id": xMemberId,
            "x-username": xUsername
        })
    })));
}
/**
 * Delete Resource Review
 */
export function deleteResourceReviewV1ResourcesResourceTypeSlugResourceSlugReviewsReviewIdDelete({ reviewId, authorization, xAccountId, xMemberId, xUsername }, opts) {
    return oazapfts.ok(oazapfts.fetchJson(`/v1/resources/${encodeURIComponent(resourceTypeSlug)}/${encodeURIComponent(resourceSlug)}/reviews/${encodeURIComponent(reviewId)}`, {
        ...opts,
        method: "DELETE",
        headers: oazapfts.mergeHeaders(opts?.headers, {
            authorization,
            "x-account-id": xAccountId,
            "x-member-id": xMemberId,
            "x-username": xUsername
        })
    }));
}
/**
 * Update Review Comment
 */
export function updateReviewCommentV1ResourcesResourceTypeSlugResourceSlugReviewsReviewIdAnswersAnswerIdPatch({ reviewId, answerId, authorization, xAccountId, xMemberId, xUsername, reviewAnswer }, opts) {
    return oazapfts.ok(oazapfts.fetchJson(`/v1/resources/${encodeURIComponent(resourceTypeSlug)}/${encodeURIComponent(resourceSlug)}/reviews/${encodeURIComponent(reviewId)}/answers/${encodeURIComponent(answerId)}`, oazapfts.json({
        ...opts,
        method: "PATCH",
        body: reviewAnswer,
        headers: oazapfts.mergeHeaders(opts?.headers, {
            authorization,
            "x-account-id": xAccountId,
            "x-member-id": xMemberId,
            "x-username": xUsername
        })
    })));
}
/**
 * Delete Review Comment
 */
export function deleteReviewCommentV1ResourcesResourceTypeSlugResourceSlugReviewsReviewIdAnswersAnswerIdDelete({ reviewId, answerId, authorization, xAccountId, xMemberId, xUsername }, opts) {
    return oazapfts.ok(oazapfts.fetchJson(`/v1/resources/${encodeURIComponent(resourceTypeSlug)}/${encodeURIComponent(resourceSlug)}/reviews/${encodeURIComponent(reviewId)}/answers/${encodeURIComponent(answerId)}`, {
        ...opts,
        method: "DELETE",
        headers: oazapfts.mergeHeaders(opts?.headers, {
            authorization,
            "x-account-id": xAccountId,
            "x-member-id": xMemberId,
            "x-username": xUsername
        })
    }));
}
/**
 * Dev Assistant V2
 */
export function devAssistantV2V2ChatPost({ accept }, opts) {
    return oazapfts.ok(oazapfts.fetchJson("/v2/chat", {
        ...opts,
        method: "POST",
        headers: oazapfts.mergeHeaders(opts?.headers, {
            accept
        })
    }));
}
/**
 * Quick Commands Run V2
 */
export function quickCommandsRunV2V2QuickCommandsSlugStepsStepSlugRunPost({ slug, stepSlug, authorization, xAccountId, xMemberId, xUsername, quickCommandsExecutionRequest }, opts) {
    return oazapfts.ok(oazapfts.fetchJson(`/v2/quick-commands/${encodeURIComponent(slug)}/steps/${encodeURIComponent(stepSlug)}/run`, oazapfts.json({
        ...opts,
        method: "POST",
        body: quickCommandsExecutionRequest,
        headers: oazapfts.mergeHeaders(opts?.headers, {
            authorization,
            "x-account-id": xAccountId,
            "x-member-id": xMemberId,
            "x-username": xUsername
        })
    })));
}
/**
 * List All
 */
export function listAllV2QuickCommandsAllGet({ visibility, order, types, authorization, xAccountId, xMemberId, xUsername }, opts) {
    return oazapfts.ok(oazapfts.fetchJson(`/v2/quick-commands/all${QS.query(QS.explode({
        visibility,
        order,
        types
    }))}`, {
        ...opts,
        headers: oazapfts.mergeHeaders(opts?.headers, {
            authorization,
            "x-account-id": xAccountId,
            "x-member-id": xMemberId,
            "x-username": xUsername
        })
    }));
}
/**
 * Dev Assistant V3
 */
export function devAssistantV3V3ChatPost({ authorization, xAccountId, xMemberId, xUsername, chatRequest }, opts) {
    return oazapfts.ok(oazapfts.fetchJson("/v3/chat", oazapfts.json({
        ...opts,
        method: "POST",
        body: chatRequest,
        headers: oazapfts.mergeHeaders(opts?.headers, {
            authorization,
            "x-account-id": xAccountId,
            "x-member-id": xMemberId,
            "x-username": xUsername
        })
    })));
}
