// @stack-spot/portal-network/dist/api/cloudPlatformHorizon.js
/**
 * PlatformHorizonApplication
 * 0.0.2
 * DO NOT MODIFY - This file has been generated using oazapfts.
 * See https://www.npmjs.com/package/oazapfts
 */
import * as Oazapfts from "@oazapfts/runtime";
import * as QS from "@oazapfts/runtime/query";
export const defaults = {
    headers: {},
    baseUrl: "https://cloud-platform-horizon.dev.stackspot.com",
};
const oazapfts = Oazapfts.runtime(defaults);
export const servers = {
    generatedServerUrl: "https://cloud-platform-horizon.dev.stackspot.com"
};
/**
 * Retrieves the organization's icon image.
 */
export function getOrganizationIcon({ organizationId }, opts) {
    return oazapfts.ok(oazapfts.fetchJson(`/v1/organizations/${encodeURIComponent(organizationId)}/icon`, {
        ...opts
    }));
}
/**
 * Uploads or update the organization's icon.
 */
export function createOrUpdateOrganizationIcon({ organizationId, body }, opts) {
    return oazapfts.ok(oazapfts.fetchJson(`/v1/organizations/${encodeURIComponent(organizationId)}/icon`, oazapfts.multipart({
        ...opts,
        method: "PUT",
        body
    })));
}
/**
 * Uploads the organization's icon.
 */
export function createOrganizationIcon({ organizationId, body }, opts) {
    return oazapfts.ok(oazapfts.fetchJson(`/v1/organizations/${encodeURIComponent(organizationId)}/icon`, oazapfts.multipart({
        ...opts,
        method: "POST",
        body
    })));
}
/**
 * List deployment contexts
 */
export function list({ environment, runtimeId, deployTemplateId, workspaceId, isDefault }, opts) {
    return oazapfts.ok(oazapfts.fetchJson(`/v1/deployment-contexts${QS.query(QS.explode({
        environment,
        runtimeId,
        deployTemplateId,
        workspaceId,
        isDefault
    }))}`, {
        ...opts
    }));
}
/**
 * Updates a deployment context
 */
export function update({ deploymentContextManifestV1 }, opts) {
    return oazapfts.ok(oazapfts.fetchJson("/v1/deployment-contexts", oazapfts.json({
        ...opts,
        method: "PUT",
        body: deploymentContextManifestV1
    })));
}
/**
 * Create a Deployment Context
 */
export function create({ deploymentContextManifestV1 }, opts) {
    return oazapfts.ok(oazapfts.fetchJson("/v1/deployment-contexts", oazapfts.json({
        ...opts,
        method: "POST",
        body: deploymentContextManifestV1
    })));
}
/**
 * List all tenants in the stackspot account
 */
export function listTenants({ xAccountId, page, perPage }, opts) {
    return oazapfts.ok(oazapfts.fetchJson(`/v1/tenants${QS.query(QS.explode({
        page,
        perPage
    }))}`, {
        ...opts,
        headers: oazapfts.mergeHeaders(opts?.headers, {
            "X-Account-Id": xAccountId
        })
    }));
}
/**
 * Creates a new tenant.
 */
export function createTenant({ tenantManifestV1 }, opts) {
    return oazapfts.ok(oazapfts.fetchJson("/v1/tenants", oazapfts.json({
        ...opts,
        method: "POST",
        body: tenantManifestV1
    })));
}
/**
 * Returns paginated runtimes.
 */
export function listRuntimes({ xAccountId, page, perPage }, opts) {
    return oazapfts.ok(oazapfts.fetchJson(`/v1/runtimes${QS.query(QS.explode({
        page,
        perPage
    }))}`, {
        ...opts,
        headers: oazapfts.mergeHeaders(opts?.headers, {
            "X-Account-Id": xAccountId
        })
    }));
}
/**
 * Creates a new runtime.
 */
export function createRuntime({ runtimeManifestV1 }, opts) {
    return oazapfts.ok(oazapfts.fetchJson("/v1/runtimes", oazapfts.json({
        ...opts,
        method: "POST",
        body: runtimeManifestV1
    })));
}
/**
 * Retrieves the Stackspot account organization
 */
export function getOrganizations({ xAccountId, page, perPage }, opts) {
    return oazapfts.ok(oazapfts.fetchJson(`/v1/organizations${QS.query(QS.explode({
        page,
        perPage
    }))}`, {
        ...opts,
        headers: oazapfts.mergeHeaders(opts?.headers, {
            "X-Account-Id": xAccountId
        })
    }));
}
/**
 * Creates a new organization.
 */
export function createOrganization({ organizationManifestV1 }, opts) {
    return oazapfts.ok(oazapfts.fetchJson("/v1/organizations", oazapfts.json({
        ...opts,
        method: "POST",
        body: organizationManifestV1
    })));
}
/**
 * Returns paginated tenants.
 */
export function listTenants1({ organizationId, page, perPage }, opts) {
    return oazapfts.ok(oazapfts.fetchJson(`/v1/organizations/${encodeURIComponent(organizationId)}/tenants${QS.query(QS.explode({
        page,
        perPage
    }))}`, {
        ...opts
    }));
}
/**
 * Creates a new tenant.
 */
export function createTenant1({ organizationId, tenantManifestV1 }, opts) {
    return oazapfts.ok(oazapfts.fetchJson(`/v1/organizations/${encodeURIComponent(organizationId)}/tenants`, oazapfts.json({
        ...opts,
        method: "POST",
        body: tenantManifestV1
    })));
}
/**
 * Returns paginated runtimes.
 */
export function listRuntimes1({ organizationId, tenantId, page, perPage }, opts) {
    return oazapfts.ok(oazapfts.fetchJson(`/v1/organizations/${encodeURIComponent(organizationId)}/tenants/${encodeURIComponent(tenantId)}/runtimes${QS.query(QS.explode({
        page,
        perPage
    }))}`, {
        ...opts
    }));
}
/**
 * Creates a new runtime.
 */
export function createRuntime1({ organizationId, tenantId, runtimeManifestV1 }, opts) {
    return oazapfts.ok(oazapfts.fetchJson(`/v1/organizations/${encodeURIComponent(organizationId)}/tenants/${encodeURIComponent(tenantId)}/runtimes`, oazapfts.json({
        ...opts,
        method: "POST",
        body: runtimeManifestV1
    })));
}
/**
 *
 *             Requires one of the following permissions:
 *               - cloud_platform:*:view
 *               - cloud_platform:*:support
 *
 */
export function listApplicationDeployments({ organizationId, tenantId, runtimeId }, opts) {
    return oazapfts.ok(oazapfts.fetchJson(`/v1/organizations/${encodeURIComponent(organizationId)}/tenants/${encodeURIComponent(tenantId)}/runtimes/${encodeURIComponent(runtimeId)}/applications/deployments`, {
        ...opts
    }));
}
/**
 * Apply manifesto resource yaml.
 */
export function deploy({ organizationId, tenantId, runtimeId, applicationDeploymentManifestV1 }, opts) {
    return oazapfts.ok(oazapfts.fetchJson(`/v1/organizations/${encodeURIComponent(organizationId)}/tenants/${encodeURIComponent(tenantId)}/runtimes/${encodeURIComponent(runtimeId)}/applications/deployments`, oazapfts.json({
        ...opts,
        method: "POST",
        body: applicationDeploymentManifestV1
    })));
}
/**
 * Execute a rollout action on an application deployment.
 */
export function postApplicationDeploymentRolloutAction({ organizationId, tenantId, runtimeId, deploymentId, action }, opts) {
    return oazapfts.ok(oazapfts.fetchJson(`/v1/organizations/${encodeURIComponent(organizationId)}/tenants/${encodeURIComponent(tenantId)}/runtimes/${encodeURIComponent(runtimeId)}/applications/deployments/${encodeURIComponent(deploymentId)}/rollout/action/${encodeURIComponent(action)}`, {
        ...opts,
        method: "POST"
    }));
}
/**
 * Associate a runtime with a workspace's environment
 */
export function associateWithWorkspaceEnvironment({ xAccountId, organizationId, tenantId, associateRuntimeWithWorkspaceEnvironmentRequest }, opts) {
    return oazapfts.ok(oazapfts.fetchJson(`/v1/organizations/${encodeURIComponent(organizationId)}/tenants/${encodeURIComponent(tenantId)}/runtimes/associations`, oazapfts.json({
        ...opts,
        method: "POST",
        body: associateRuntimeWithWorkspaceEnvironmentRequest,
        headers: oazapfts.mergeHeaders(opts?.headers, {
            "X-Account-Id": xAccountId
        })
    })));
}
/**
 * List DeployTemplate's
 */
export function listDeployTemplate({ organizationId }, opts) {
    return oazapfts.ok(oazapfts.fetchJson(`/v1/organizations/${encodeURIComponent(organizationId)}/deploy-templates`, {
        ...opts
    }));
}
/**
 * Create DeployTemplate
 */
export function createDeployTemplate({ organizationId, deployTemplateManifestV1 }, opts) {
    return oazapfts.ok(oazapfts.fetchJson(`/v1/organizations/${encodeURIComponent(organizationId)}/deploy-templates`, oazapfts.json({
        ...opts,
        method: "POST",
        body: deployTemplateManifestV1
    })));
}
/**
 * List Addons
 */
export function listAddon({ organizationId }, opts) {
    return oazapfts.ok(oazapfts.fetchJson(`/v1/organizations/${encodeURIComponent(organizationId)}/addons`, {
        ...opts
    }));
}
/**
 * Create Addon
 */
export function createAddon({ organizationId, addonManifestV1 }, opts) {
    return oazapfts.ok(oazapfts.fetchJson(`/v1/organizations/${encodeURIComponent(organizationId)}/addons`, oazapfts.json({
        ...opts,
        method: "POST",
        body: addonManifestV1
    })));
}
/**
 * List application deployments
 */
export function listApplicationDeployments1(opts) {
    return oazapfts.ok(oazapfts.fetchJson("/v1/applications/deployments", {
        ...opts
    }));
}
/**
 * Apply manifesto resource yaml.
 */
export function deploy1({ applicationDeploymentManifestV1 }, opts) {
    return oazapfts.ok(oazapfts.fetchJson("/v1/applications/deployments", oazapfts.json({
        ...opts,
        method: "POST",
        body: applicationDeploymentManifestV1
    })));
}
/**
 * Retrieves the tenant details.
 */
export function getTenant({ id }, opts) {
    return oazapfts.ok(oazapfts.fetchJson(`/v1/tenants/${encodeURIComponent(id)}`, {
        ...opts
    }));
}
/**
 * Retrieves the runtime details.
 */
export function getRuntime({ id }, opts) {
    return oazapfts.ok(oazapfts.fetchJson(`/v1/runtimes/${encodeURIComponent(id)}`, {
        ...opts
    }));
}
export function listAssociations({ xAccountId, workspaceId, envId, applicationId }, opts) {
    return oazapfts.ok(oazapfts.fetchJson(`/v1/runtimes/associations/workspaces${QS.query(QS.explode({
        workspaceId,
        envId,
        applicationId
    }))}`, {
        ...opts,
        headers: oazapfts.mergeHeaders(opts?.headers, {
            "X-Account-Id": xAccountId
        })
    }));
}
/**
 * Retrieves associate a runtime with a workspace's environment
 */
export function getAssociationWorkspaceEnvironmentByRuntimeId({ xAccountId, organizationId, tenantId, runtimeId }, opts) {
    return oazapfts.ok(oazapfts.fetchJson(`/v1/organizations/${encodeURIComponent(organizationId)}/tenants/${encodeURIComponent(tenantId)}/runtimes/${encodeURIComponent(runtimeId)}/associations`, {
        ...opts,
        headers: oazapfts.mergeHeaders(opts?.headers, {
            "X-Account-Id": xAccountId
        })
    }));
}
/**
 *
 *             Requires one of the following permissions:
 *               - cloud_platform:*:view
 *               - cloud_platform:*:support
 *
 */
export function findApplicationDeploymentById({ organizationId, tenantId, runtimeId, deploymentId }, opts) {
    return oazapfts.ok(oazapfts.fetchJson(`/v1/organizations/${encodeURIComponent(organizationId)}/tenants/${encodeURIComponent(tenantId)}/runtimes/${encodeURIComponent(runtimeId)}/applications/deployments/${encodeURIComponent(deploymentId)}`, {
        ...opts
    }));
}
/**
 * Delete application deployment
 */
export function deleteApplicationDeployment({ organizationId, tenantId, runtimeId, deploymentId }, opts) {
    return oazapfts.ok(oazapfts.fetchJson(`/v1/organizations/${encodeURIComponent(organizationId)}/tenants/${encodeURIComponent(tenantId)}/runtimes/${encodeURIComponent(runtimeId)}/applications/deployments/${encodeURIComponent(deploymentId)}`, {
        ...opts,
        method: "DELETE"
    }));
}
/**
 * Get application deployment rollout.
 */
export function listApplicationDeploymentRollouts({ organizationId, tenantId, runtimeId, deploymentId }, opts) {
    return oazapfts.ok(oazapfts.fetchJson(`/v1/organizations/${encodeURIComponent(organizationId)}/tenants/${encodeURIComponent(tenantId)}/runtimes/${encodeURIComponent(runtimeId)}/applications/deployments/${encodeURIComponent(deploymentId)}/rollouts`, {
        ...opts
    }));
}
/**
 * Get application deployment logs.
 */
export function getApplicationDeploymentLogs({ organizationId, tenantId, runtimeId, deploymentId, numberOfLines }, opts) {
    return oazapfts.ok(oazapfts.fetchJson(`/v1/organizations/${encodeURIComponent(organizationId)}/tenants/${encodeURIComponent(tenantId)}/runtimes/${encodeURIComponent(runtimeId)}/applications/deployments/${encodeURIComponent(deploymentId)}/logs${QS.query(QS.explode({
        numberOfLines
    }))}`, {
        ...opts
    }));
}
/**
 * Get application deployment health.
 */
export function getApplicationDeploymentHealth({ organizationId, tenantId, runtimeId, deploymentId }, opts) {
    return oazapfts.ok(oazapfts.fetchJson(`/v1/organizations/${encodeURIComponent(organizationId)}/tenants/${encodeURIComponent(tenantId)}/runtimes/${encodeURIComponent(runtimeId)}/applications/deployments/${encodeURIComponent(deploymentId)}/health`, {
        ...opts
    }));
}
/**
 * Retrieves the runtime details.
 */
export function getRuntime1({ organizationId, tenantId, id }, opts) {
    return oazapfts.ok(oazapfts.fetchJson(`/v1/organizations/${encodeURIComponent(organizationId)}/tenants/${encodeURIComponent(tenantId)}/runtimes/${encodeURIComponent(id)}`, {
        ...opts
    }));
}
/**
 * Retrieves the tenant details.
 */
export function getTenant1({ organizationId, id }, opts) {
    return oazapfts.ok(oazapfts.fetchJson(`/v1/organizations/${encodeURIComponent(organizationId)}/tenants/${encodeURIComponent(id)}`, {
        ...opts
    }));
}
/**
 * Get DeployTemplate
 */
export function getDeployTemplate({ organizationId, deployTemplateId }, opts) {
    return oazapfts.ok(oazapfts.fetchJson(`/v1/organizations/${encodeURIComponent(organizationId)}/deploy-templates/${encodeURIComponent(deployTemplateId)}`, {
        ...opts
    }));
}
/**
 * Delete DeployTemplate
 */
export function deleteDeployTemplate({ organizationId, deployTemplateId }, opts) {
    return oazapfts.ok(oazapfts.fetchJson(`/v1/organizations/${encodeURIComponent(organizationId)}/deploy-templates/${encodeURIComponent(deployTemplateId)}`, {
        ...opts,
        method: "DELETE"
    }));
}
/**
 * Get Addon
 */
export function getAddon({ organizationId, addonId }, opts) {
    return oazapfts.ok(oazapfts.fetchJson(`/v1/organizations/${encodeURIComponent(organizationId)}/addons/${encodeURIComponent(addonId)}`, {
        ...opts
    }));
}
/**
 * Retrieves the organization by ID
 */
export function getOrganizationById({ organizationId }, opts) {
    return oazapfts.ok(oazapfts.fetchJson(`/v1/organizations/${encodeURIComponent(organizationId)}`, {
        ...opts
    }));
}
/**
 * Downloads the bootstrap module for creating a new organization.
 */
export function downloadBootstrapModule({ version }, opts) {
    return oazapfts.ok(oazapfts.fetchJson(`/v1/organizations/bootstrap${QS.query(QS.explode({
        version
    }))}`, {
        ...opts
    }));
}
export function getTenantExample(opts) {
    return oazapfts.ok(oazapfts.fetchJson("/v1/examples/tenant", {
        ...opts
    }));
}
export function getRuntimeExample(opts) {
    return oazapfts.ok(oazapfts.fetchJson("/v1/examples/runtime", {
        ...opts
    }));
}
export function getApplicationDeploymentExample(opts) {
    return oazapfts.ok(oazapfts.fetchJson("/v1/examples/applicationDeployment", {
        ...opts
    }));
}
/**
 * Get deployment context by ID
 */
export function getV1DeploymentContextsById({ id }, opts) {
    return oazapfts.ok(oazapfts.fetchJson(`/v1/deployment-contexts/${encodeURIComponent(id)}`, {
        ...opts
    }));
}
/**
 * Delete deployment context by ID
 */
export function deleteV1DeploymentContextsById({ id }, opts) {
    return oazapfts.ok(oazapfts.fetchJson(`/v1/deployment-contexts/${encodeURIComponent(id)}`, {
        ...opts,
        method: "DELETE"
    }));
}
/**
 * Get application deployment by ID
 */
export function findApplicationDeploymentById1({ deploymentId }, opts) {
    return oazapfts.ok(oazapfts.fetchJson(`/v1/applications/deployments/${encodeURIComponent(deploymentId)}`, {
        ...opts
    }));
}
/**
 * Get application deployment logs.
 */
export function getApplicationDeploymentLogs1({ deploymentId, numberOfLines }, opts) {
    return oazapfts.ok(oazapfts.fetchJson(`/v1/applications/deployments/${encodeURIComponent(deploymentId)}/logs${QS.query(QS.explode({
        numberOfLines
    }))}`, {
        ...opts
    }));
}
/**
 * Get application deployment health.
 */
export function getApplicationDeploymentHealth1({ deploymentId }, opts) {
    return oazapfts.ok(oazapfts.fetchJson(`/v1/applications/deployments/${encodeURIComponent(deploymentId)}/health`, {
        ...opts
    }));
}
