// @stack-spot/portal-network/dist/api/cloudPlatform.js
/**
 * OpenAPI definition
 * v0
 * DO NOT MODIFY - This file has been generated using oazapfts.
 * See https://www.npmjs.com/package/oazapfts
 */
import * as Oazapfts from "@oazapfts/runtime";
import * as QS from "@oazapfts/runtime/query";
export const defaults = {
    headers: {},
    baseUrl: "https://cloud-cloud-platform-api.dev.stackspot.com",
};
const oazapfts = Oazapfts.runtime(defaults);
export const servers = {
    generatedServerUrl: "https://cloud-cloud-platform-api.dev.stackspot.com"
};
export function putVpnTags({ authorization, xAccountId, foundationId, vpnId, body }, opts) {
    return oazapfts.ok(oazapfts.fetchJson(`/v1/foundations/${encodeURIComponent(foundationId)}/vpns/${encodeURIComponent(vpnId)}/tags`, oazapfts.json({
        ...opts,
        method: "PUT",
        body,
        headers: oazapfts.mergeHeaders(opts?.headers, {
            Authorization: authorization,
            "x-account-id": xAccountId
        })
    })));
}
export function putProjectTags({ authorization, xAccountId, foundationId, projectId, body }, opts) {
    return oazapfts.ok(oazapfts.fetchJson(`/v1/foundations/${encodeURIComponent(foundationId)}/projects/${encodeURIComponent(projectId)}/tags`, oazapfts.json({
        ...opts,
        method: "PUT",
        body,
        headers: oazapfts.mergeHeaders(opts?.headers, {
            Authorization: authorization,
            "x-account-id": xAccountId
        })
    })));
}
export function putNetworkTags({ authorization, xAccountId, foundationId, networkId, body }, opts) {
    return oazapfts.ok(oazapfts.fetchJson(`/v1/foundations/${encodeURIComponent(foundationId)}/networks/${encodeURIComponent(networkId)}/tags`, oazapfts.json({
        ...opts,
        method: "PUT",
        body,
        headers: oazapfts.mergeHeaders(opts?.headers, {
            Authorization: authorization,
            "x-account-id": xAccountId
        })
    })));
}
export function getFolderTags({ authorization, xAccountId, foundationId, folderId }, opts) {
    return oazapfts.ok(oazapfts.fetchJson(`/v1/foundations/${encodeURIComponent(foundationId)}/folders/${encodeURIComponent(folderId)}/tags`, {
        ...opts,
        headers: oazapfts.mergeHeaders(opts?.headers, {
            Authorization: authorization,
            "x-account-id": xAccountId
        })
    }));
}
export function putFolderTags({ authorization, xAccountId, foundationId, folderId, body }, opts) {
    return oazapfts.ok(oazapfts.fetchJson(`/v1/foundations/${encodeURIComponent(foundationId)}/folders/${encodeURIComponent(folderId)}/tags`, oazapfts.json({
        ...opts,
        method: "PUT",
        body,
        headers: oazapfts.mergeHeaders(opts?.headers, {
            Authorization: authorization,
            "x-account-id": xAccountId
        })
    })));
}
export function putDnsZoneTags({ authorization, xAccountId, foundationId, dnsZoneId, body }, opts) {
    return oazapfts.ok(oazapfts.fetchJson(`/v1/foundations/${encodeURIComponent(foundationId)}/dns-zones/${encodeURIComponent(dnsZoneId)}/tags`, oazapfts.json({
        ...opts,
        method: "PUT",
        body,
        headers: oazapfts.mergeHeaders(opts?.headers, {
            Authorization: authorization,
            "x-account-id": xAccountId
        })
    })));
}
export function putDnsRecordRecords({ authorization, xAccountId, foundationId, dnsRecordId, body }, opts) {
    return oazapfts.ok(oazapfts.fetchJson(`/v1/foundations/${encodeURIComponent(foundationId)}/dns-records/${encodeURIComponent(dnsRecordId)}/records`, oazapfts.json({
        ...opts,
        method: "PUT",
        body,
        headers: oazapfts.mergeHeaders(opts?.headers, {
            Authorization: authorization,
            "x-account-id": xAccountId
        })
    })));
}
export function putCidrTags({ authorization, xAccountId, foundationId, cidrId, body }, opts) {
    return oazapfts.ok(oazapfts.fetchJson(`/v1/foundations/${encodeURIComponent(foundationId)}/cidrs/${encodeURIComponent(cidrId)}/tags`, oazapfts.json({
        ...opts,
        method: "PUT",
        body,
        headers: oazapfts.mergeHeaders(opts?.headers, {
            Authorization: authorization,
            "x-account-id": xAccountId
        })
    })));
}
export function putCertificateTags({ authorization, xAccountId, foundationId, certificateId, body }, opts) {
    return oazapfts.ok(oazapfts.fetchJson(`/v1/foundations/${encodeURIComponent(foundationId)}/certificates/${encodeURIComponent(certificateId)}/tags`, oazapfts.json({
        ...opts,
        method: "PUT",
        body,
        headers: oazapfts.mergeHeaders(opts?.headers, {
            Authorization: authorization,
            "x-account-id": xAccountId
        })
    })));
}
export function listFoundations({ authorization, xAccountId }, opts) {
    return oazapfts.ok(oazapfts.fetchJson("/v1/foundations", {
        ...opts,
        headers: oazapfts.mergeHeaders(opts?.headers, {
            Authorization: authorization,
            "x-account-id": xAccountId
        })
    }));
}
export function createFoundation({ authorization, xAccountId, createFoundationRequest }, opts) {
    return oazapfts.ok(oazapfts.fetchJson("/v1/foundations", oazapfts.json({
        ...opts,
        method: "POST",
        body: createFoundationRequest,
        headers: oazapfts.mergeHeaders(opts?.headers, {
            Authorization: authorization,
            "x-account-id": xAccountId
        })
    })));
}
export function listVpns({ authorization, xAccountId, foundationId }, opts) {
    return oazapfts.ok(oazapfts.fetchJson(`/v1/foundations/${encodeURIComponent(foundationId)}/vpns`, {
        ...opts,
        headers: oazapfts.mergeHeaders(opts?.headers, {
            Authorization: authorization,
            "x-account-id": xAccountId
        })
    }));
}
export function createVpn({ authorization, xAccountId, foundationId, createVpnRequest }, opts) {
    return oazapfts.ok(oazapfts.fetchJson(`/v1/foundations/${encodeURIComponent(foundationId)}/vpns`, oazapfts.json({
        ...opts,
        method: "POST",
        body: createVpnRequest,
        headers: oazapfts.mergeHeaders(opts?.headers, {
            Authorization: authorization,
            "x-account-id": xAccountId
        })
    })));
}
export function listTenant({ authorization, xAccountId, foundationId }, opts) {
    return oazapfts.ok(oazapfts.fetchJson(`/v1/foundations/${encodeURIComponent(foundationId)}/tenants`, {
        ...opts,
        headers: oazapfts.mergeHeaders(opts?.headers, {
            Authorization: authorization,
            "x-account-id": xAccountId
        })
    }));
}
export function createTenant({ authorization, xAccountId, foundationId, createTenantRequest }, opts) {
    return oazapfts.ok(oazapfts.fetchJson(`/v1/foundations/${encodeURIComponent(foundationId)}/tenants`, oazapfts.json({
        ...opts,
        method: "POST",
        body: createTenantRequest,
        headers: oazapfts.mergeHeaders(opts?.headers, {
            Authorization: authorization,
            "x-account-id": xAccountId
        })
    })));
}
export function listRuntime({ authorization, xAccountId, foundationId, tenantId, projectId }, opts) {
    return oazapfts.ok(oazapfts.fetchJson(`/v1/foundations/${encodeURIComponent(foundationId)}/runtimes${QS.query(QS.explode({
        tenantId,
        projectId
    }))}`, {
        ...opts,
        headers: oazapfts.mergeHeaders(opts?.headers, {
            Authorization: authorization,
            "x-account-id": xAccountId
        })
    }));
}
export function createRuntime({ authorization, xAccountId, foundationId, createRuntimeRequest }, opts) {
    return oazapfts.ok(oazapfts.fetchJson(`/v1/foundations/${encodeURIComponent(foundationId)}/runtimes`, oazapfts.json({
        ...opts,
        method: "POST",
        body: createRuntimeRequest,
        headers: oazapfts.mergeHeaders(opts?.headers, {
            Authorization: authorization,
            "x-account-id": xAccountId
        })
    })));
}
export function listProject({ authorization, xAccountId, foundationId, parentFolderId }, opts) {
    return oazapfts.ok(oazapfts.fetchJson(`/v1/foundations/${encodeURIComponent(foundationId)}/projects${QS.query(QS.explode({
        parentFolderId
    }))}`, {
        ...opts,
        headers: oazapfts.mergeHeaders(opts?.headers, {
            Authorization: authorization,
            "x-account-id": xAccountId
        })
    }));
}
export function createProject({ authorization, xAccountId, foundationId, createProjectRequest }, opts) {
    return oazapfts.ok(oazapfts.fetchJson(`/v1/foundations/${encodeURIComponent(foundationId)}/projects`, oazapfts.json({
        ...opts,
        method: "POST",
        body: createProjectRequest,
        headers: oazapfts.mergeHeaders(opts?.headers, {
            Authorization: authorization,
            "x-account-id": xAccountId
        })
    })));
}
export function listNetwork({ authorization, xAccountId, foundationId, projectId }, opts) {
    return oazapfts.ok(oazapfts.fetchJson(`/v1/foundations/${encodeURIComponent(foundationId)}/networks${QS.query(QS.explode({
        projectId
    }))}`, {
        ...opts,
        headers: oazapfts.mergeHeaders(opts?.headers, {
            Authorization: authorization,
            "x-account-id": xAccountId
        })
    }));
}
export function createNetwork({ authorization, xAccountId, foundationId, createNetworkRequest }, opts) {
    return oazapfts.ok(oazapfts.fetchJson(`/v1/foundations/${encodeURIComponent(foundationId)}/networks`, oazapfts.json({
        ...opts,
        method: "POST",
        body: createNetworkRequest,
        headers: oazapfts.mergeHeaders(opts?.headers, {
            Authorization: authorization,
            "x-account-id": xAccountId
        })
    })));
}
export function getFolder({ authorization, xAccountId, foundationId, folderId }, opts) {
    return oazapfts.ok(oazapfts.fetchJson(`/v1/foundations/${encodeURIComponent(foundationId)}/folders${QS.query(QS.explode({
        folderId
    }))}`, {
        ...opts,
        headers: oazapfts.mergeHeaders(opts?.headers, {
            Authorization: authorization,
            "x-account-id": xAccountId
        })
    }));
}
export function createFolder({ authorization, xAccountId, foundationId, createFolderRequest }, opts) {
    return oazapfts.ok(oazapfts.fetchJson(`/v1/foundations/${encodeURIComponent(foundationId)}/folders`, oazapfts.json({
        ...opts,
        method: "POST",
        body: createFolderRequest,
        headers: oazapfts.mergeHeaders(opts?.headers, {
            Authorization: authorization,
            "x-account-id": xAccountId
        })
    })));
}
export function listDnsZone({ authorization, xAccountId, foundationId, projectId, privacy }, opts) {
    return oazapfts.ok(oazapfts.fetchJson(`/v1/foundations/${encodeURIComponent(foundationId)}/dns-zones${QS.query(QS.explode({
        projectId,
        privacy
    }))}`, {
        ...opts,
        headers: oazapfts.mergeHeaders(opts?.headers, {
            Authorization: authorization,
            "x-account-id": xAccountId
        })
    }));
}
export function createDnsZone({ authorization, xAccountId, foundationId, createDnsZoneRequest }, opts) {
    return oazapfts.ok(oazapfts.fetchJson(`/v1/foundations/${encodeURIComponent(foundationId)}/dns-zones`, oazapfts.json({
        ...opts,
        method: "POST",
        body: createDnsZoneRequest,
        headers: oazapfts.mergeHeaders(opts?.headers, {
            Authorization: authorization,
            "x-account-id": xAccountId
        })
    })));
}
export function listDnsRecord({ authorization, xAccountId, foundationId, projectId, dnsZoneId }, opts) {
    return oazapfts.ok(oazapfts.fetchJson(`/v1/foundations/${encodeURIComponent(foundationId)}/dns-records${QS.query(QS.explode({
        projectId,
        dnsZoneId
    }))}`, {
        ...opts,
        headers: oazapfts.mergeHeaders(opts?.headers, {
            Authorization: authorization,
            "x-account-id": xAccountId
        })
    }));
}
export function createDnsRecord({ authorization, xAccountId, foundationId, createDnsRecordRequest }, opts) {
    return oazapfts.ok(oazapfts.fetchJson(`/v1/foundations/${encodeURIComponent(foundationId)}/dns-records`, oazapfts.json({
        ...opts,
        method: "POST",
        body: createDnsRecordRequest,
        headers: oazapfts.mergeHeaders(opts?.headers, {
            Authorization: authorization,
            "x-account-id": xAccountId
        })
    })));
}
export function listCidr({ authorization, xAccountId, foundationId }, opts) {
    return oazapfts.ok(oazapfts.fetchJson(`/v1/foundations/${encodeURIComponent(foundationId)}/cidrs`, {
        ...opts,
        headers: oazapfts.mergeHeaders(opts?.headers, {
            Authorization: authorization,
            "x-account-id": xAccountId
        })
    }));
}
export function createCidr({ authorization, xAccountId, foundationId, createCidrRequest }, opts) {
    return oazapfts.ok(oazapfts.fetchJson(`/v1/foundations/${encodeURIComponent(foundationId)}/cidrs`, oazapfts.json({
        ...opts,
        method: "POST",
        body: createCidrRequest,
        headers: oazapfts.mergeHeaders(opts?.headers, {
            Authorization: authorization,
            "x-account-id": xAccountId
        })
    })));
}
export function listCertificates({ authorization, xAccountId, foundationId, forInbound }, opts) {
    return oazapfts.ok(oazapfts.fetchJson(`/v1/foundations/${encodeURIComponent(foundationId)}/certificates${QS.query(QS.explode({
        forInbound
    }))}`, {
        ...opts,
        headers: oazapfts.mergeHeaders(opts?.headers, {
            Authorization: authorization,
            "x-account-id": xAccountId
        })
    }));
}
export function createCertificate({ authorization, xAccountId, foundationId, body }, opts) {
    return oazapfts.ok(oazapfts.fetchJson(`/v1/foundations/${encodeURIComponent(foundationId)}/certificates`, oazapfts.json({
        ...opts,
        method: "POST",
        body,
        headers: oazapfts.mergeHeaders(opts?.headers, {
            Authorization: authorization,
            "x-account-id": xAccountId
        })
    })));
}
export function getFoundation({ authorization, xAccountId, foundationId }, opts) {
    return oazapfts.ok(oazapfts.fetchJson(`/v1/foundations/${encodeURIComponent(foundationId)}`, {
        ...opts,
        headers: oazapfts.mergeHeaders(opts?.headers, {
            Authorization: authorization,
            "x-account-id": xAccountId
        })
    }));
}
export function getVpn({ authorization, xAccountId, foundationId, vpnId }, opts) {
    return oazapfts.ok(oazapfts.fetchJson(`/v1/foundations/${encodeURIComponent(foundationId)}/vpns/${encodeURIComponent(vpnId)}`, {
        ...opts,
        headers: oazapfts.mergeHeaders(opts?.headers, {
            Authorization: authorization,
            "x-account-id": xAccountId
        })
    }));
}
export function deleteVpn({ authorization, xAccountId, foundationId, vpnId }, opts) {
    return oazapfts.ok(oazapfts.fetchText(`/v1/foundations/${encodeURIComponent(foundationId)}/vpns/${encodeURIComponent(vpnId)}`, {
        ...opts,
        method: "DELETE",
        headers: oazapfts.mergeHeaders(opts?.headers, {
            Authorization: authorization,
            "x-account-id": xAccountId
        })
    }));
}
export function getVpnConfiguration({ authorization, xAccountId, foundationId, vpnId }, opts) {
    return oazapfts.ok(oazapfts.fetchJson(`/v1/foundations/${encodeURIComponent(foundationId)}/vpns/${encodeURIComponent(vpnId)}/configuration`, {
        ...opts,
        headers: oazapfts.mergeHeaders(opts?.headers, {
            Authorization: authorization,
            "x-account-id": xAccountId
        })
    }));
}
export function getTenant({ authorization, xAccountId, foundationId, tenantId }, opts) {
    return oazapfts.ok(oazapfts.fetchJson(`/v1/foundations/${encodeURIComponent(foundationId)}/tenants/${encodeURIComponent(tenantId)}`, {
        ...opts,
        headers: oazapfts.mergeHeaders(opts?.headers, {
            Authorization: authorization,
            "x-account-id": xAccountId
        })
    }));
}
export function getRuntime({ authorization, xAccountId, foundationId, runtimeId }, opts) {
    return oazapfts.ok(oazapfts.fetchJson(`/v1/foundations/${encodeURIComponent(foundationId)}/runtimes/${encodeURIComponent(runtimeId)}`, {
        ...opts,
        headers: oazapfts.mergeHeaders(opts?.headers, {
            Authorization: authorization,
            "x-account-id": xAccountId
        })
    }));
}
export function getProject({ authorization, xAccountId, foundationId, projectId }, opts) {
    return oazapfts.ok(oazapfts.fetchJson(`/v1/foundations/${encodeURIComponent(foundationId)}/projects/${encodeURIComponent(projectId)}`, {
        ...opts,
        headers: oazapfts.mergeHeaders(opts?.headers, {
            Authorization: authorization,
            "x-account-id": xAccountId
        })
    }));
}
export function deleteProject({ authorization, xAccountId, foundationId, projectId }, opts) {
    return oazapfts.ok(oazapfts.fetchText(`/v1/foundations/${encodeURIComponent(foundationId)}/projects/${encodeURIComponent(projectId)}`, {
        ...opts,
        method: "DELETE",
        headers: oazapfts.mergeHeaders(opts?.headers, {
            Authorization: authorization,
            "x-account-id": xAccountId
        })
    }));
}
export function getNetwork({ authorization, xAccountId, foundationId, networkId }, opts) {
    return oazapfts.ok(oazapfts.fetchJson(`/v1/foundations/${encodeURIComponent(foundationId)}/networks/${encodeURIComponent(networkId)}`, {
        ...opts,
        headers: oazapfts.mergeHeaders(opts?.headers, {
            Authorization: authorization,
            "x-account-id": xAccountId
        })
    }));
}
export function deleteNetwork({ authorization, xAccountId, foundationId, networkId }, opts) {
    return oazapfts.ok(oazapfts.fetchText(`/v1/foundations/${encodeURIComponent(foundationId)}/networks/${encodeURIComponent(networkId)}`, {
        ...opts,
        method: "DELETE",
        headers: oazapfts.mergeHeaders(opts?.headers, {
            Authorization: authorization,
            "x-account-id": xAccountId
        })
    }));
}
export function getDnsZone({ authorization, xAccountId, foundationId, dnsZoneId }, opts) {
    return oazapfts.ok(oazapfts.fetchJson(`/v1/foundations/${encodeURIComponent(foundationId)}/dns-zones/${encodeURIComponent(dnsZoneId)}`, {
        ...opts,
        headers: oazapfts.mergeHeaders(opts?.headers, {
            Authorization: authorization,
            "x-account-id": xAccountId
        })
    }));
}
export function deleteDnsZone({ authorization, xAccountId, foundationId, dnsZoneId }, opts) {
    return oazapfts.ok(oazapfts.fetchText(`/v1/foundations/${encodeURIComponent(foundationId)}/dns-zones/${encodeURIComponent(dnsZoneId)}`, {
        ...opts,
        method: "DELETE",
        headers: oazapfts.mergeHeaders(opts?.headers, {
            Authorization: authorization,
            "x-account-id": xAccountId
        })
    }));
}
export function getDnsRecord({ authorization, xAccountId, foundationId, dnsRecordId }, opts) {
    return oazapfts.ok(oazapfts.fetchJson(`/v1/foundations/${encodeURIComponent(foundationId)}/dns-records/${encodeURIComponent(dnsRecordId)}`, {
        ...opts,
        headers: oazapfts.mergeHeaders(opts?.headers, {
            Authorization: authorization,
            "x-account-id": xAccountId
        })
    }));
}
export function deleteDnsRecord({ authorization, xAccountId, foundationId, dnsRecordId }, opts) {
    return oazapfts.ok(oazapfts.fetchText(`/v1/foundations/${encodeURIComponent(foundationId)}/dns-records/${encodeURIComponent(dnsRecordId)}`, {
        ...opts,
        method: "DELETE",
        headers: oazapfts.mergeHeaders(opts?.headers, {
            Authorization: authorization,
            "x-account-id": xAccountId
        })
    }));
}
export function getCidr({ authorization, xAccountId, foundationId, cidrId }, opts) {
    return oazapfts.ok(oazapfts.fetchJson(`/v1/foundations/${encodeURIComponent(foundationId)}/cidrs/${encodeURIComponent(cidrId)}`, {
        ...opts,
        headers: oazapfts.mergeHeaders(opts?.headers, {
            Authorization: authorization,
            "x-account-id": xAccountId
        })
    }));
}
export function deleteCidr({ authorization, xAccountId, foundationId, cidrId }, opts) {
    return oazapfts.ok(oazapfts.fetchText(`/v1/foundations/${encodeURIComponent(foundationId)}/cidrs/${encodeURIComponent(cidrId)}`, {
        ...opts,
        method: "DELETE",
        headers: oazapfts.mergeHeaders(opts?.headers, {
            Authorization: authorization,
            "x-account-id": xAccountId
        })
    }));
}
export function getCertificate({ authorization, xAccountId, foundationId, certificateId }, opts) {
    return oazapfts.ok(oazapfts.fetchJson(`/v1/foundations/${encodeURIComponent(foundationId)}/certificates/${encodeURIComponent(certificateId)}`, {
        ...opts,
        headers: oazapfts.mergeHeaders(opts?.headers, {
            Authorization: authorization,
            "x-account-id": xAccountId
        })
    }));
}
export function deleteCertificate({ authorization, xAccountId, foundationId, certificateId }, opts) {
    return oazapfts.ok(oazapfts.fetchText(`/v1/foundations/${encodeURIComponent(foundationId)}/certificates/${encodeURIComponent(certificateId)}`, {
        ...opts,
        method: "DELETE",
        headers: oazapfts.mergeHeaders(opts?.headers, {
            Authorization: authorization,
            "x-account-id": xAccountId
        })
    }));
}
export function getProvisionAvailability({ cloudProvider }, opts) {
    return oazapfts.ok(oazapfts.fetchJson(`/v1/foundations/provision-avaliability${QS.query(QS.explode({
        cloudProvider
    }))}`, {
        ...opts
    }));
}
export function providers(opts) {
    return oazapfts.ok(oazapfts.fetchJson("/v1/foundations/providers", {
        ...opts
    }));
}
export function deleteFolder({ authorization, xAccountId, foundationId, folderId }, opts) {
    return oazapfts.ok(oazapfts.fetchText(`/v1/foundations/${encodeURIComponent(foundationId)}/folders/${encodeURIComponent(folderId)}`, {
        ...opts,
        method: "DELETE",
        headers: oazapfts.mergeHeaders(opts?.headers, {
            Authorization: authorization,
            "x-account-id": xAccountId
        })
    }));
}
