// @stack-spot/portal-network/dist/api/codeShift.js
/**
 * Code shift API
 * 0.1.0
 * DO NOT MODIFY - This file has been generated using oazapfts.
 * See https://www.npmjs.com/package/oazapfts
 */
import * as Oazapfts from "@oazapfts/runtime";
import * as QS from "@oazapfts/runtime/query";
export const defaults = {
    headers: {},
    baseUrl: "/",
};
const oazapfts = Oazapfts.runtime(defaults);
export const servers = {};
/**
 * Check Role Route
 */
export function checkRoleRouteV1RolesRoleGet({ role, authorization }, opts) {
    return oazapfts.ok(oazapfts.fetchJson(`/v1/roles/${encodeURIComponent(role)}`, {
        ...opts,
        headers: oazapfts.mergeHeaders(opts?.headers, {
            authorization
        })
    }));
}
/**
 * Module Favorite Service Add
 */
export function moduleFavoriteServiceAddV1ModulesModuleIdFavoritesPost({ moduleId, authorization }, opts) {
    return oazapfts.ok(oazapfts.fetchJson(`/v1/modules/${encodeURIComponent(moduleId)}/favorites`, {
        ...opts,
        method: "POST",
        headers: oazapfts.mergeHeaders(opts?.headers, {
            authorization
        })
    }));
}
/**
 * Module Favorite Service Delete
 */
export function moduleFavoriteServiceDeleteV1ModulesModuleIdFavoritesDelete({ moduleId, authorization }, opts) {
    return oazapfts.ok(oazapfts.fetchJson(`/v1/modules/${encodeURIComponent(moduleId)}/favorites`, {
        ...opts,
        method: "DELETE",
        headers: oazapfts.mergeHeaders(opts?.headers, {
            authorization
        })
    }));
}
/**
 * Create Module Service
 */
export function createModuleServiceV1ModulesPost({ authorization, xAccountSlug, createModuleRequest }, opts) {
    return oazapfts.ok(oazapfts.fetchJson("/v1/modules", oazapfts.json({
        ...opts,
        method: "POST",
        body: createModuleRequest,
        headers: oazapfts.mergeHeaders(opts?.headers, {
            authorization,
            "x-account-slug": xAccountSlug
        })
    })));
}
/**
 * List Modules Service
 */
export function listModulesServiceV1ModulesGet({ pageSize, moduleType, page, lastEvaluatedKey, sharedWith, createdBy, visibility, name, tags, xAccountSlug, authorization }, opts) {
    return oazapfts.ok(oazapfts.fetchJson(`/v1/modules${QS.query(QS.explode({
        pageSize,
        moduleType,
        page,
        lastEvaluatedKey,
        sharedWith,
        createdBy,
        visibility,
        name,
        tags
    }))}`, {
        ...opts,
        headers: oazapfts.mergeHeaders(opts?.headers, {
            "x-account-slug": xAccountSlug,
            authorization
        })
    }));
}
/**
 * Delete Module Service
 */
export function deleteModuleServiceV1ModulesModuleIdDelete({ moduleId, authorization, xAccountSlug }, opts) {
    return oazapfts.ok(oazapfts.fetchJson(`/v1/modules/${encodeURIComponent(moduleId)}`, {
        ...opts,
        method: "DELETE",
        headers: oazapfts.mergeHeaders(opts?.headers, {
            authorization,
            "x-account-slug": xAccountSlug
        })
    }));
}
/**
 * Update Module Service
 */
export function updateModuleServiceV1ModulesModuleIdPut({ moduleId, authorization, xAccountSlug, putModuleRequest }, opts) {
    return oazapfts.ok(oazapfts.fetchJson(`/v1/modules/${encodeURIComponent(moduleId)}`, oazapfts.json({
        ...opts,
        method: "PUT",
        body: putModuleRequest,
        headers: oazapfts.mergeHeaders(opts?.headers, {
            authorization,
            "x-account-slug": xAccountSlug
        })
    })));
}
/**
 * Get Module
 */
export function getModuleV1ModulesModuleIdGet({ moduleId, authorization, xAccountSlug }, opts) {
    return oazapfts.ok(oazapfts.fetchJson(`/v1/modules/${encodeURIComponent(moduleId)}`, {
        ...opts,
        headers: oazapfts.mergeHeaders(opts?.headers, {
            authorization,
            "x-account-slug": xAccountSlug
        })
    }));
}
/**
 * Get Module Inputs
 */
export function getModuleInputsV1ModulesModuleIdInputsGet({ moduleId, authorization }, opts) {
    return oazapfts.ok(oazapfts.fetchJson(`/v1/modules/${encodeURIComponent(moduleId)}/inputs`, {
        ...opts,
        headers: oazapfts.mergeHeaders(opts?.headers, {
            authorization
        })
    }));
}
/**
 * Get Module Docs
 */
export function getModuleDocsV1ModulesModuleIdDocsGet({ moduleId, language, authorization }, opts) {
    return oazapfts.ok(oazapfts.fetchJson(`/v1/modules/${encodeURIComponent(moduleId)}/docs${QS.query(QS.explode({
        language
    }))}`, {
        ...opts,
        headers: oazapfts.mergeHeaders(opts?.headers, {
            authorization
        })
    }));
}
/**
 * Dispatch Module Service
 */
export function dispatchModuleServiceV1ModulesDispatchesPost({ batchName, authorization, body }, opts) {
    return oazapfts.ok(oazapfts.fetchJson(`/v1/modules/dispatches${QS.query(QS.explode({
        batchName
    }))}`, oazapfts.json({
        ...opts,
        method: "POST",
        body,
        headers: oazapfts.mergeHeaders(opts?.headers, {
            authorization
        })
    })));
}
/**
 * Put Report
 */
export function putReportV1ReportsReportIdPut({ reportId, authorization, putReportRequest }, opts) {
    return oazapfts.ok(oazapfts.fetchJson(`/v1/reports/${encodeURIComponent(reportId)}`, oazapfts.json({
        ...opts,
        method: "PUT",
        body: putReportRequest,
        headers: oazapfts.mergeHeaders(opts?.headers, {
            authorization
        })
    })));
}
/**
 * Get Report
 */
export function getReportV1ReportsReportIdGet({ reportId, authorization }, opts) {
    return oazapfts.ok(oazapfts.fetchJson(`/v1/reports/${encodeURIComponent(reportId)}`, {
        ...opts,
        headers: oazapfts.mergeHeaders(opts?.headers, {
            authorization
        })
    }));
}
/**
 * Get Report Pull Request Content
 */
export function getReportPullRequestContentV1ReportsReportIdPullRequestGet({ reportId, authorization }, opts) {
    return oazapfts.ok(oazapfts.fetchJson(`/v1/reports/${encodeURIComponent(reportId)}/pull-request`, {
        ...opts,
        headers: oazapfts.mergeHeaders(opts?.headers, {
            authorization
        })
    }));
}
/**
 * Download Report
 */
export function downloadReportV1ReportsReportIdDownloadGet({ reportId, extension, authorization }, opts) {
    return oazapfts.ok(oazapfts.fetchJson(`/v1/reports/${encodeURIComponent(reportId)}/download${QS.query(QS.explode({
        extension
    }))}`, {
        ...opts,
        headers: oazapfts.mergeHeaders(opts?.headers, {
            authorization
        })
    }));
}
/**
 * Put Customer Rating Report
 */
export function putCustomerRatingReportV1ReportsReportIdCustomerRatingPut({ reportId, authorization, putCustomerRatingReportRequest }, opts) {
    return oazapfts.ok(oazapfts.fetchJson(`/v1/reports/${encodeURIComponent(reportId)}/customer-rating`, oazapfts.json({
        ...opts,
        method: "PUT",
        body: putCustomerRatingReportRequest,
        headers: oazapfts.mergeHeaders(opts?.headers, {
            authorization
        })
    })));
}
/**
 * Search Repos Scm Service
 */
export function searchReposScmServiceV1ReposSearchScmGet({ repoName, authorization }, opts) {
    return oazapfts.ok(oazapfts.fetchJson(`/v1/repos/search-scm${QS.query(QS.explode({
        repoName
    }))}`, {
        ...opts,
        headers: oazapfts.mergeHeaders(opts?.headers, {
            authorization
        })
    }));
}
/**
 * Get Status Search Repos Scm Service
 */
export function getStatusSearchReposScmServiceV1ReposSearchScmSearchRepoIdStatusGet({ searchRepoId, authorization }, opts) {
    return oazapfts.ok(oazapfts.fetchJson(`/v1/repos/search-scm/${encodeURIComponent(searchRepoId)}/status`, {
        ...opts,
        headers: oazapfts.mergeHeaders(opts?.headers, {
            authorization
        })
    }));
}
/**
 * Download Search Repos Scm Service
 */
export function downloadSearchReposScmServiceV1ReposSearchScmSearchRepoIdDownloadGet({ searchRepoId, extension, authorization }, opts) {
    return oazapfts.ok(oazapfts.fetchJson(`/v1/repos/search-scm/${encodeURIComponent(searchRepoId)}/download${QS.query(QS.explode({
        extension
    }))}`, {
        ...opts,
        headers: oazapfts.mergeHeaders(opts?.headers, {
            authorization
        })
    }));
}
/**
 * List Repository Report
 */
export function listRepositoryReportV1ReposRepositoryIdReportsGet({ repositoryId, pageSize, page, lastEvaluatedKey, authorization }, opts) {
    return oazapfts.ok(oazapfts.fetchJson(`/v1/repos/${encodeURIComponent(repositoryId)}/reports${QS.query(QS.explode({
        pageSize,
        page,
        lastEvaluatedKey
    }))}`, {
        ...opts,
        headers: oazapfts.mergeHeaders(opts?.headers, {
            authorization
        })
    }));
}
/**
 * Create Repos Batch Service
 */
export function createReposBatchServiceV1ReposBatchPost({ tags, authorization, fastapiCompatV2BodyCreateReposBatchServiceV1ReposBatchPost }, opts) {
    return oazapfts.ok(oazapfts.fetchJson(`/v1/repos/batch${QS.query(QS.explode({
        tags
    }))}`, oazapfts.multipart({
        ...opts,
        method: "POST",
        body: fastapiCompatV2BodyCreateReposBatchServiceV1ReposBatchPost,
        headers: oazapfts.mergeHeaders(opts?.headers, {
            authorization
        })
    })));
}
/**
 * Create Repository Service
 */
export function createRepositoryServiceV1ReposPost({ authorization, srcPresentationApiRestServicesV1ReposRequestsCreateRepositoryRequestCreateRepositoryRequest }, opts) {
    return oazapfts.ok(oazapfts.fetchJson("/v1/repos", oazapfts.json({
        ...opts,
        method: "POST",
        body: srcPresentationApiRestServicesV1ReposRequestsCreateRepositoryRequestCreateRepositoryRequest,
        headers: oazapfts.mergeHeaders(opts?.headers, {
            authorization
        })
    })));
}
/**
 * List Repository Service
 */
export function listRepositoryServiceV1ReposGet({ pageSize, page, lastEvaluatedKey, filter, moduleId, createdByEmail, reportByEmail, orderBy, orderDirection, tags, authorization }, opts) {
    return oazapfts.ok(oazapfts.fetchJson(`/v1/repos${QS.query(QS.explode({
        pageSize,
        page,
        lastEvaluatedKey,
        filter,
        moduleId,
        createdByEmail,
        reportByEmail,
        orderBy,
        orderDirection,
        tags
    }))}`, {
        ...opts,
        headers: oazapfts.mergeHeaders(opts?.headers, {
            authorization
        })
    }));
}
/**
 * List Repository Download Service
 */
export function listRepositoryDownloadServiceV1ReposDownloadGet({ filter, moduleId, createdByEmail, reportByEmail, orderBy, orderDirection, tags, extension, authorization }, opts) {
    return oazapfts.ok(oazapfts.fetchJson(`/v1/repos/download${QS.query(QS.explode({
        filter,
        moduleId,
        createdByEmail,
        reportByEmail,
        orderBy,
        orderDirection,
        tags,
        extension
    }))}`, {
        ...opts,
        headers: oazapfts.mergeHeaders(opts?.headers, {
            authorization
        })
    }));
}
/**
 * List Branches Service
 */
export function listBranchesServiceV1ReposBranchesGet({ repositoryUrl, authorization }, opts) {
    return oazapfts.ok(oazapfts.fetchJson(`/v1/repos/branches${QS.query(QS.explode({
        repositoryUrl
    }))}`, {
        ...opts,
        headers: oazapfts.mergeHeaders(opts?.headers, {
            authorization
        })
    }));
}
/**
 * Get Repository By Id Service
 */
export function getRepositoryByIdServiceV1ReposRepositoryIdGet({ repositoryId, authorization }, opts) {
    return oazapfts.ok(oazapfts.fetchJson(`/v1/repos/${encodeURIComponent(repositoryId)}`, {
        ...opts,
        headers: oazapfts.mergeHeaders(opts?.headers, {
            authorization
        })
    }));
}
/**
 * Delete Repository Service
 */
export function deleteRepositoryServiceV1ReposRepositoryIdDelete({ repositoryId, authorization }, opts) {
    return oazapfts.ok(oazapfts.fetchJson(`/v1/repos/${encodeURIComponent(repositoryId)}`, {
        ...opts,
        method: "DELETE",
        headers: oazapfts.mergeHeaders(opts?.headers, {
            authorization
        })
    }));
}
/**
 * Update Repository Service
 */
export function updateRepositoryServiceV1ReposRepositoryIdPut({ repositoryId, authorization, putRepositoryRequest }, opts) {
    return oazapfts.ok(oazapfts.fetchJson(`/v1/repos/${encodeURIComponent(repositoryId)}`, oazapfts.json({
        ...opts,
        method: "PUT",
        body: putRepositoryRequest,
        headers: oazapfts.mergeHeaders(opts?.headers, {
            authorization
        })
    })));
}
/**
 * Validate Scm Url Service
 */
export function validateScmUrlServiceV1ReposValidateScmUrlPost({ raiseConflict, authorization, validateScmUrlRequest }, opts) {
    return oazapfts.ok(oazapfts.fetchJson(`/v1/repos/validate-scm-url${QS.query(QS.explode({
        raiseConflict
    }))}`, oazapfts.json({
        ...opts,
        method: "POST",
        body: validateScmUrlRequest,
        headers: oazapfts.mergeHeaders(opts?.headers, {
            authorization
        })
    })));
}
/**
 * Get Account Settings
 */
export function getAccountSettingsV1SettingsGet({ authorization }, opts) {
    return oazapfts.ok(oazapfts.fetchJson("/v1/settings", {
        ...opts,
        headers: oazapfts.mergeHeaders(opts?.headers, {
            authorization
        })
    }));
}
/**
 * Create Account Settings
 */
export function createAccountSettingsV1SettingsPut({ authorization, createAccountSettingsRequest }, opts) {
    return oazapfts.ok(oazapfts.fetchJson("/v1/settings", oazapfts.json({
        ...opts,
        method: "PUT",
        body: createAccountSettingsRequest,
        headers: oazapfts.mergeHeaders(opts?.headers, {
            authorization
        })
    })));
}
/**
 * Create Integration Service
 */
export function createIntegrationServiceV1IntegrationsPost({ authorization, createIntegrationRequest }, opts) {
    return oazapfts.ok(oazapfts.fetchJson("/v1/integrations", oazapfts.json({
        ...opts,
        method: "POST",
        body: createIntegrationRequest,
        headers: oazapfts.mergeHeaders(opts?.headers, {
            authorization
        })
    })));
}
/**
 * List Integration Service
 */
export function listIntegrationServiceV1IntegrationsGet({ pageSize, page, name, authorization }, opts) {
    return oazapfts.ok(oazapfts.fetchJson(`/v1/integrations${QS.query(QS.explode({
        pageSize,
        page,
        name
    }))}`, {
        ...opts,
        headers: oazapfts.mergeHeaders(opts?.headers, {
            authorization
        })
    }));
}
/**
 * Get Integration By Id Service
 */
export function getIntegrationByIdServiceV1IntegrationsIntegrationIdGet({ integrationId, authorization }, opts) {
    return oazapfts.ok(oazapfts.fetchJson(`/v1/integrations/${encodeURIComponent(integrationId)}`, {
        ...opts,
        headers: oazapfts.mergeHeaders(opts?.headers, {
            authorization
        })
    }));
}
/**
 * Update Integration Service
 */
export function updateIntegrationServiceV1IntegrationsIntegrationIdPut({ integrationId, authorization, putIntegrationRequest }, opts) {
    return oazapfts.ok(oazapfts.fetchJson(`/v1/integrations/${encodeURIComponent(integrationId)}`, oazapfts.json({
        ...opts,
        method: "PUT",
        body: putIntegrationRequest,
        headers: oazapfts.mergeHeaders(opts?.headers, {
            authorization
        })
    })));
}
/**
 * Delete Integration Service
 */
export function deleteIntegrationServiceV1IntegrationsIntegrationIdDelete({ integrationId, authorization }, opts) {
    return oazapfts.ok(oazapfts.fetchJson(`/v1/integrations/${encodeURIComponent(integrationId)}`, {
        ...opts,
        method: "DELETE",
        headers: oazapfts.mergeHeaders(opts?.headers, {
            authorization
        })
    }));
}
/**
 * Create Program Group Service
 */
export function createProgramGroupServiceV1ProgramGroupsPost({ name, tags, integrationId, authorization, fastapiCompatV2BodyCreateProgramGroupServiceV1ProgramGroupsPost }, opts) {
    return oazapfts.ok(oazapfts.fetchJson(`/v1/program-groups${QS.query(QS.explode({
        name,
        tags,
        integrationId
    }))}`, oazapfts.multipart({
        ...opts,
        method: "POST",
        body: fastapiCompatV2BodyCreateProgramGroupServiceV1ProgramGroupsPost,
        headers: oazapfts.mergeHeaders(opts?.headers, {
            authorization
        })
    })));
}
/**
 * List Program Group Service
 */
export function listProgramGroupServiceV1ProgramGroupsGet({ moduleId, createdByEmail, reportByEmail, tags, name, pageSize, page, orderBy, orderDirection, integrationId, authorization }, opts) {
    return oazapfts.ok(oazapfts.fetchJson(`/v1/program-groups${QS.query(QS.explode({
        moduleId,
        createdByEmail,
        reportByEmail,
        tags,
        name,
        pageSize,
        page,
        orderBy,
        orderDirection,
        integrationId
    }))}`, {
        ...opts,
        headers: oazapfts.mergeHeaders(opts?.headers, {
            authorization
        })
    }));
}
/**
 * List Program Group Download Service
 */
export function listProgramGroupDownloadServiceV1ProgramGroupsDownloadGet({ moduleId, createdByEmail, reportByEmail, tags, name, orderBy, orderDirection, integrationId, extension, authorization }, opts) {
    return oazapfts.ok(oazapfts.fetchJson(`/v1/program-groups/download${QS.query(QS.explode({
        moduleId,
        createdByEmail,
        reportByEmail,
        tags,
        name,
        orderBy,
        orderDirection,
        integrationId,
        extension
    }))}`, {
        ...opts,
        headers: oazapfts.mergeHeaders(opts?.headers, {
            authorization
        })
    }));
}
/**
 * Get Program Group By Id Service
 */
export function getProgramGroupByIdServiceV1ProgramGroupsProgramGroupIdGet({ programGroupId, authorization }, opts) {
    return oazapfts.ok(oazapfts.fetchJson(`/v1/program-groups/${encodeURIComponent(programGroupId)}`, {
        ...opts,
        headers: oazapfts.mergeHeaders(opts?.headers, {
            authorization
        })
    }));
}
/**
 * Update Program Group Service
 */
export function updateProgramGroupServiceV1ProgramGroupsProgramGroupIdPut({ programGroupId, name, tags, integrationId, authorization, fastapiCompatV2BodyUpdateProgramGroupServiceV1ProgramGroupsProgramGroupIdPut }, opts) {
    return oazapfts.ok(oazapfts.fetchJson(`/v1/program-groups/${encodeURIComponent(programGroupId)}${QS.query(QS.explode({
        name,
        tags,
        integrationId
    }))}`, oazapfts.multipart({
        ...opts,
        method: "PUT",
        body: fastapiCompatV2BodyUpdateProgramGroupServiceV1ProgramGroupsProgramGroupIdPut,
        headers: oazapfts.mergeHeaders(opts?.headers, {
            authorization
        })
    })));
}
/**
 * Delete Program Group Service
 */
export function deleteProgramGroupServiceV1ProgramGroupsProgramGroupIdDelete({ programGroupId, authorization }, opts) {
    return oazapfts.ok(oazapfts.fetchJson(`/v1/program-groups/${encodeURIComponent(programGroupId)}`, {
        ...opts,
        method: "DELETE",
        headers: oazapfts.mergeHeaders(opts?.headers, {
            authorization
        })
    }));
}
/**
 * Update Program Group Components Service
 */
export function updateProgramGroupComponentsServiceV1ProgramGroupsProgramGroupIdComponentsPut({ programGroupId, authorization, body }, opts) {
    return oazapfts.ok(oazapfts.fetchJson(`/v1/program-groups/${encodeURIComponent(programGroupId)}/components`, oazapfts.json({
        ...opts,
        method: "PUT",
        body,
        headers: oazapfts.mergeHeaders(opts?.headers, {
            authorization
        })
    })));
}
/**
 * List Program Group Report Service
 */
export function listProgramGroupReportServiceV1ProgramGroupsProgramGroupIdReportsGet({ programGroupId, pageSize, page, lastEvaluatedKey, authorization }, opts) {
    return oazapfts.ok(oazapfts.fetchJson(`/v1/program-groups/${encodeURIComponent(programGroupId)}/reports${QS.query(QS.explode({
        pageSize,
        page,
        lastEvaluatedKey
    }))}`, {
        ...opts,
        headers: oazapfts.mergeHeaders(opts?.headers, {
            authorization
        })
    }));
}
/**
 * List Tags Service
 */
export function listTagsServiceV1TagsGet({ name, $type, authorization }, opts) {
    return oazapfts.ok(oazapfts.fetchJson(`/v1/tags${QS.query(QS.explode({
        name,
        "type": $type
    }))}`, {
        ...opts,
        headers: oazapfts.mergeHeaders(opts?.headers, {
            authorization
        })
    }));
}
/**
 * General Report Success And Errors
 */
export function generalReportSuccessAndErrorsV1AnalyticsReportsGeneralSuccessAndErrorsGet({ startDate, endDate, authorization }, opts) {
    return oazapfts.ok(oazapfts.fetchJson(`/v1/analytics/reports/general/success-and-errors${QS.query(QS.explode({
        startDate,
        endDate
    }))}`, {
        ...opts,
        headers: oazapfts.mergeHeaders(opts?.headers, {
            authorization
        })
    }));
}
/**
 * Analytics Repository Last Report Status
 */
export function analyticsRepositoryLastReportStatusV1AnalyticsRepositoriesLastReportStatusGet({ pageSize, page, moduleId, executedBy, repositoryUrl, tags, orderBy, orderDirection, startDate, endDate, mode, authorization }, opts) {
    return oazapfts.ok(oazapfts.fetchJson(`/v1/analytics/repositories/last-report-status${QS.query(QS.explode({
        pageSize,
        page,
        moduleId,
        executedBy,
        repositoryUrl,
        tags,
        orderBy,
        orderDirection,
        startDate,
        endDate,
        mode
    }))}`, {
        ...opts,
        headers: oazapfts.mergeHeaders(opts?.headers, {
            authorization
        })
    }));
}
/**
 * Analytics Repository Last Report Status Download
 */
export function analyticsRepositoryLastReportStatusDownloadV1AnalyticsRepositoriesLastReportStatusDownloadGet({ moduleId, orderBy, tags, executedBy, repositoryUrl, orderDirection, startDate, endDate, extension, mode, authorization }, opts) {
    return oazapfts.ok(oazapfts.fetchJson(`/v1/analytics/repositories/last-report-status/download${QS.query(QS.explode({
        moduleId,
        orderBy,
        tags,
        executedBy,
        repositoryUrl,
        orderDirection,
        startDate,
        endDate,
        extension,
        mode
    }))}`, {
        ...opts,
        headers: oazapfts.mergeHeaders(opts?.headers, {
            authorization
        })
    }));
}
/**
 * Analytics Repository Usage
 */
export function analyticsRepositoryUsageV1AnalyticsRepositoriesUsageGet({ pageSize, page, moduleId, userId, tags, orderBy, executedBy, repositoryUrl, orderDirection, startDate, endDate, authorization }, opts) {
    return oazapfts.ok(oazapfts.fetchJson(`/v1/analytics/repositories/usage${QS.query(QS.explode({
        pageSize,
        page,
        moduleId,
        user_id: userId,
        tags,
        orderBy,
        executedBy,
        repositoryUrl,
        orderDirection,
        startDate,
        endDate
    }))}`, {
        ...opts,
        headers: oazapfts.mergeHeaders(opts?.headers, {
            authorization
        })
    }));
}
/**
 * Analytics Repository Usage Download
 */
export function analyticsRepositoryUsageDownloadV1AnalyticsRepositoriesUsageDownloadGet({ moduleId, userId, orderBy, tags, executedBy, repositoryUrl, orderDirection, startDate, endDate, extension, authorization }, opts) {
    return oazapfts.ok(oazapfts.fetchJson(`/v1/analytics/repositories/usage/download${QS.query(QS.explode({
        moduleId,
        user_id: userId,
        orderBy,
        tags,
        executedBy,
        repositoryUrl,
        orderDirection,
        startDate,
        endDate,
        extension
    }))}`, {
        ...opts,
        headers: oazapfts.mergeHeaders(opts?.headers, {
            authorization
        })
    }));
}
/**
 * Analytics User Usage
 */
export function analyticsUserUsageV1AnalyticsUsersUsageGet({ pageSize, page, moduleId, name, reportType, orderBy, orderDirection, startDate, endDate, authorization }, opts) {
    return oazapfts.ok(oazapfts.fetchJson(`/v1/analytics/users/usage${QS.query(QS.explode({
        pageSize,
        page,
        moduleId,
        name,
        report_type: reportType,
        orderBy,
        orderDirection,
        startDate,
        endDate
    }))}`, {
        ...opts,
        headers: oazapfts.mergeHeaders(opts?.headers, {
            authorization
        })
    }));
}
/**
 * Analytics User Usage Download
 */
export function analyticsUserUsageDownloadV1AnalyticsUsersUsageDownloadGet({ moduleId, name, reportType, orderBy, orderDirection, startDate, endDate, extension, authorization }, opts) {
    return oazapfts.ok(oazapfts.fetchJson(`/v1/analytics/users/usage/download${QS.query(QS.explode({
        moduleId,
        name,
        report_type: reportType,
        orderBy,
        orderDirection,
        startDate,
        endDate,
        extension
    }))}`, {
        ...opts,
        headers: oazapfts.mergeHeaders(opts?.headers, {
            authorization
        })
    }));
}
/**
 * Analytics Program Groups Last Report Status Download
 */
export function analyticsProgramGroupsLastReportStatusDownloadV1AnalyticsProgramGroupsLastReportStatusDownloadGet({ moduleId, orderBy, orderDirection, startDate, endDate, reportType, extension, authorization }, opts) {
    return oazapfts.ok(oazapfts.fetchJson(`/v1/analytics/program-groups/last-report-status/download${QS.query(QS.explode({
        moduleId,
        orderBy,
        orderDirection,
        startDate,
        endDate,
        report_type: reportType,
        extension
    }))}`, {
        ...opts,
        headers: oazapfts.mergeHeaders(opts?.headers, {
            authorization
        })
    }));
}
/**
 * Analytics Program Groups Last Report Status
 */
export function analyticsProgramGroupsLastReportStatusV1AnalyticsProgramGroupsLastReportStatusGet({ pageSize, page, moduleId, userId, tags, orderBy, executedBy, programName, orderDirection, startDate, endDate, authorization }, opts) {
    return oazapfts.ok(oazapfts.fetchJson(`/v1/analytics/program-groups/last-report-status${QS.query(QS.explode({
        pageSize,
        page,
        moduleId,
        user_id: userId,
        tags,
        orderBy,
        executedBy,
        programName,
        orderDirection,
        startDate,
        endDate
    }))}`, {
        ...opts,
        headers: oazapfts.mergeHeaders(opts?.headers, {
            authorization
        })
    }));
}
/**
 * Analytics Program Groups Usage
 */
export function analyticsProgramGroupsUsageV1AnalyticsProgramGroupsUsageGet({ pageSize, page, executedBy, moduleId, tags, orderBy, orderDirection, programGroupName, startDate, endDate, authorization }, opts) {
    return oazapfts.ok(oazapfts.fetchJson(`/v1/analytics/program-groups/usage${QS.query(QS.explode({
        pageSize,
        page,
        executedBy,
        moduleId,
        tags,
        orderBy,
        orderDirection,
        programGroupName,
        startDate,
        endDate
    }))}`, {
        ...opts,
        headers: oazapfts.mergeHeaders(opts?.headers, {
            authorization
        })
    }));
}
/**
 * Analytics Program Groups Usage Download
 */
export function analyticsProgramGroupsUsageDownloadV1AnalyticsProgramGroupsUsageDownloadGet({ moduleId, executedBy, tags, orderBy, orderDirection, programGroupName, startDate, endDate, extension, authorization }, opts) {
    return oazapfts.ok(oazapfts.fetchJson(`/v1/analytics/program-groups/usage/download${QS.query(QS.explode({
        moduleId,
        executedBy,
        tags,
        orderBy,
        orderDirection,
        programGroupName,
        startDate,
        endDate,
        extension
    }))}`, {
        ...opts,
        headers: oazapfts.mergeHeaders(opts?.headers, {
            authorization
        })
    }));
}
/**
 * Analytics Program Groups Details
 */
export function analyticsProgramGroupsDetailsV1AnalyticsProgramGroupsDetailsGet({ pageSize, page, programName, tags, mode, moduleId, programCreatedById, reportCreatedById, startDate, endDate, orderBy, orderDirection, authorization }, opts) {
    return oazapfts.ok(oazapfts.fetchJson(`/v1/analytics/program-groups/details${QS.query(QS.explode({
        pageSize,
        page,
        programName,
        tags,
        mode,
        moduleId,
        programCreatedById,
        reportCreatedById,
        startDate,
        endDate,
        orderBy,
        orderDirection
    }))}`, {
        ...opts,
        headers: oazapfts.mergeHeaders(opts?.headers, {
            authorization
        })
    }));
}
/**
 * Analytics Program Groups Details Download
 */
export function analyticsProgramGroupsDetailsDownloadV1AnalyticsProgramGroupsDetailsDownloadGet({ programName, tags, mode, moduleId, programCreatedById, reportCreatedById, startDate, endDate, orderBy, orderDirection, extension, authorization }, opts) {
    return oazapfts.ok(oazapfts.fetchJson(`/v1/analytics/program-groups/details/download${QS.query(QS.explode({
        programName,
        tags,
        mode,
        moduleId,
        programCreatedById,
        reportCreatedById,
        startDate,
        endDate,
        orderBy,
        orderDirection,
        extension
    }))}`, {
        ...opts,
        headers: oazapfts.mergeHeaders(opts?.headers, {
            authorization
        })
    }));
}
/**
 * Analytics Repository Detailed Report
 */
export function analyticsRepositoryDetailedReportV1AnalyticsRepositoriesDetailsGet({ pageSize, page, repositoryUrl, tags, mode, moduleId, repositoryCreatedById, executedBy, startDate, endDate, orderBy, orderDirection, authorization }, opts) {
    return oazapfts.ok(oazapfts.fetchJson(`/v1/analytics/repositories/details${QS.query(QS.explode({
        pageSize,
        page,
        repositoryUrl,
        tags,
        mode,
        moduleId,
        repositoryCreatedById,
        executedBy,
        startDate,
        endDate,
        orderBy,
        orderDirection
    }))}`, {
        ...opts,
        headers: oazapfts.mergeHeaders(opts?.headers, {
            authorization
        })
    }));
}
/**
 * Analytics Repository Detailed Report Download
 */
export function analyticsRepositoryDetailedReportDownloadV1AnalyticsRepositoriesDetailsDownloadGet({ repositoryUrl, tags, mode, moduleId, repositoryCreatedById, reportCreatedById, startDate, endDate, orderBy, orderDirection, extension, authorization }, opts) {
    return oazapfts.ok(oazapfts.fetchJson(`/v1/analytics/repositories/details/download${QS.query(QS.explode({
        repositoryUrl,
        tags,
        mode,
        moduleId,
        repositoryCreatedById,
        reportCreatedById,
        startDate,
        endDate,
        orderBy,
        orderDirection,
        extension
    }))}`, {
        ...opts,
        headers: oazapfts.mergeHeaders(opts?.headers, {
            authorization
        })
    }));
}
/**
 * Analytics Program Groups Target Details
 */
export function analyticsProgramGroupsTargetDetailsV1AnalyticsProgramGroupsTargetDetailsGet({ pageSize, page, programName, componentName, tags, mode, moduleId, reportCreatedById, startDate, endDate, orderBy, orderDirection, authorization }, opts) {
    return oazapfts.ok(oazapfts.fetchJson(`/v1/analytics/program-groups/target-details${QS.query(QS.explode({
        pageSize,
        page,
        programName,
        componentName,
        tags,
        mode,
        moduleId,
        reportCreatedById,
        startDate,
        endDate,
        orderBy,
        orderDirection
    }))}`, {
        ...opts,
        headers: oazapfts.mergeHeaders(opts?.headers, {
            authorization
        })
    }));
}
/**
 * Analytics Program Groups Target Details Download
 */
export function analyticsProgramGroupsTargetDetailsDownloadV1AnalyticsProgramGroupsTargetDetailsDownloadGet({ programName, tags, mode, moduleId, componentName, reportCreatedById, startDate, endDate, orderBy, orderDirection, extension, authorization }, opts) {
    return oazapfts.ok(oazapfts.fetchJson(`/v1/analytics/program-groups/target-details/download${QS.query(QS.explode({
        programName,
        tags,
        mode,
        moduleId,
        componentName,
        reportCreatedById,
        startDate,
        endDate,
        orderBy,
        orderDirection,
        extension
    }))}`, {
        ...opts,
        headers: oazapfts.mergeHeaders(opts?.headers, {
            authorization
        })
    }));
}
/**
 * Analytics Repository Target Details
 */
export function analyticsRepositoryTargetDetailsV1AnalyticsRepositoriesTargetDetailsGet({ pageSize, page, repositoryUrl, targetName, tags, mode, moduleId, reportCreatedById, startDate, endDate, orderBy, orderDirection, authorization }, opts) {
    return oazapfts.ok(oazapfts.fetchJson(`/v1/analytics/repositories/target-details${QS.query(QS.explode({
        pageSize,
        page,
        repositoryUrl,
        targetName,
        tags,
        mode,
        moduleId,
        reportCreatedById,
        startDate,
        endDate,
        orderBy,
        orderDirection
    }))}`, {
        ...opts,
        headers: oazapfts.mergeHeaders(opts?.headers, {
            authorization
        })
    }));
}
/**
 * Analytics Repository Target Details Download
 */
export function analyticsRepositoryTargetDetailsDownloadV1AnalyticsRepositoriesTargetDetailsDownloadGet({ repositoryUrl, targetName, tags, mode, moduleId, reportCreatedById, startDate, endDate, orderBy, orderDirection, extension, authorization }, opts) {
    return oazapfts.ok(oazapfts.fetchJson(`/v1/analytics/repositories/target-details/download${QS.query(QS.explode({
        repositoryUrl,
        targetName,
        tags,
        mode,
        moduleId,
        reportCreatedById,
        startDate,
        endDate,
        orderBy,
        orderDirection,
        extension
    }))}`, {
        ...opts,
        headers: oazapfts.mergeHeaders(opts?.headers, {
            authorization
        })
    }));
}
/**
 * Analytics Module Execution Times Report
 */
export function analyticsModuleExecutionTimesReportV1AnalyticsModulesExecutionTimesGet({ pageSize, page, moduleId, version, mode, startDate, endDate, orderBy, orderDirection, authorization }, opts) {
    return oazapfts.ok(oazapfts.fetchJson(`/v1/analytics/modules/execution-times${QS.query(QS.explode({
        pageSize,
        page,
        moduleId,
        version,
        mode,
        startDate,
        endDate,
        orderBy,
        orderDirection
    }))}`, {
        ...opts,
        headers: oazapfts.mergeHeaders(opts?.headers, {
            authorization
        })
    }));
}
/**
 * Analytics Module Execution Times Download
 */
export function analyticsModuleExecutionTimesDownloadV1AnalyticsModulesExecutionTimesDownloadGet({ moduleId, version, mode, startDate, endDate, orderBy, orderDirection, extension, authorization }, opts) {
    return oazapfts.ok(oazapfts.fetchJson(`/v1/analytics/modules/execution-times/download${QS.query(QS.explode({
        moduleId,
        version,
        mode,
        startDate,
        endDate,
        orderBy,
        orderDirection,
        extension
    }))}`, {
        ...opts,
        headers: oazapfts.mergeHeaders(opts?.headers, {
            authorization
        })
    }));
}
/**
 * List User Service
 */
export function listUserServiceV1UsersGet({ authorization }, opts) {
    return oazapfts.ok(oazapfts.fetchJson("/v1/users", {
        ...opts,
        headers: oazapfts.mergeHeaders(opts?.headers, {
            authorization
        })
    }));
}
/**
 * Search Repos Scm Service
 */
export function searchReposScmServiceV2ReposSearchScmPost({ authorization, searchReposInScmRequest }, opts) {
    return oazapfts.ok(oazapfts.fetchJson("/v2/repos/search-scm", oazapfts.json({
        ...opts,
        method: "POST",
        body: searchReposInScmRequest,
        headers: oazapfts.mergeHeaders(opts?.headers, {
            authorization
        })
    })));
}
/**
 * Import Repos With Tags Scm Service
 */
export function importReposWithTagsScmServiceV2ReposSearchScmSearchIdPost({ searchId, authorization, importReposScmWithTagsRequest }, opts) {
    return oazapfts.ok(oazapfts.fetchJson(`/v2/repos/search-scm/${encodeURIComponent(searchId)}`, oazapfts.json({
        ...opts,
        method: "POST",
        body: importReposScmWithTagsRequest,
        headers: oazapfts.mergeHeaders(opts?.headers, {
            authorization
        })
    })));
}
/**
 * Search Repos Scm V2
 */
export function searchReposScmV2V2ReposSearchScmSearchIdGet({ searchId, authorization }, opts) {
    return oazapfts.ok(oazapfts.fetchJson(`/v2/repos/search-scm/${encodeURIComponent(searchId)}`, {
        ...opts,
        headers: oazapfts.mergeHeaders(opts?.headers, {
            authorization
        })
    }));
}
/**
 * Download Search Repos Scm V2
 */
export function downloadSearchReposScmV2V2ReposSearchScmSearchIdDownloadGet({ searchId, extension, authorization }, opts) {
    return oazapfts.ok(oazapfts.fetchJson(`/v2/repos/search-scm/${encodeURIComponent(searchId)}/download${QS.query(QS.explode({
        extension
    }))}`, {
        ...opts,
        headers: oazapfts.mergeHeaders(opts?.headers, {
            authorization
        })
    }));
}
/**
 * Create Repos Batch Service
 */
export function createReposBatchServiceV2ReposBatchPost({ tags, authorization, fastapiCompatV2BodyCreateReposBatchServiceV2ReposBatchPost }, opts) {
    return oazapfts.ok(oazapfts.fetchJson(`/v2/repos/batch${QS.query(QS.explode({
        tags
    }))}`, oazapfts.multipart({
        ...opts,
        method: "POST",
        body: fastapiCompatV2BodyCreateReposBatchServiceV2ReposBatchPost,
        headers: oazapfts.mergeHeaders(opts?.headers, {
            authorization
        })
    })));
}
/**
 * Get Import Result
 */
export function getImportResultV2ReposBatchImportIdGet({ importId, pageSize, page, authorization }, opts) {
    return oazapfts.ok(oazapfts.fetchJson(`/v2/repos/batch/${encodeURIComponent(importId)}${QS.query(QS.explode({
        pageSize,
        page
    }))}`, {
        ...opts,
        headers: oazapfts.mergeHeaders(opts?.headers, {
            authorization
        })
    }));
}
/**
 * Download Import Result
 */
export function downloadImportResultV2ReposBatchImportIdDownloadGet({ importId, language, status, extension, authorization }, opts) {
    return oazapfts.ok(oazapfts.fetchJson(`/v2/repos/batch/${encodeURIComponent(importId)}/download${QS.query(QS.explode({
        language,
        status,
        extension
    }))}`, {
        ...opts,
        headers: oazapfts.mergeHeaders(opts?.headers, {
            authorization
        })
    }));
}
/**
 * List Repository Service
 */
export function listRepositoryServiceV2ReposGet({ pageSize, page, lastEvaluatedKey, filter, moduleId, createdByEmail, reportByEmail, orderBy, orderDirection, tags, onlyIds, authorization }, opts) {
    return oazapfts.ok(oazapfts.fetchJson(`/v2/repos${QS.query(QS.explode({
        pageSize,
        page,
        lastEvaluatedKey,
        filter,
        moduleId,
        createdByEmail,
        reportByEmail,
        orderBy,
        orderDirection,
        tags,
        onlyIds
    }))}`, {
        ...opts,
        headers: oazapfts.mergeHeaders(opts?.headers, {
            authorization
        })
    }));
}
/**
 * Create Repository Service
 */
export function createRepositoryServiceV2ReposPost({ authorization, srcPresentationApiRestServicesV2ReposRequestsCreateRepositoryRequestCreateRepositoryRequest }, opts) {
    return oazapfts.ok(oazapfts.fetchJson("/v2/repos", oazapfts.json({
        ...opts,
        method: "POST",
        body: srcPresentationApiRestServicesV2ReposRequestsCreateRepositoryRequestCreateRepositoryRequest,
        headers: oazapfts.mergeHeaders(opts?.headers, {
            authorization
        })
    })));
}
/**
 * Pat Health Check
 */
export function patHealthCheckV1ScmPatHealthCheckGet({ authorization }, opts) {
    return oazapfts.ok(oazapfts.fetchJson("/v1/scm/pat/health-check", {
        ...opts,
        headers: oazapfts.mergeHeaders(opts?.headers, {
            authorization
        })
    }));
}
/**
 * List Module Execution Batch Activities
 */
export function listModuleExecutionBatchActivitiesV1ActivitiesModuleExecutionBatchGet({ pageSize, page, authorization }, opts) {
    return oazapfts.ok(oazapfts.fetchJson(`/v1/activities/module-execution-batch${QS.query(QS.explode({
        pageSize,
        page
    }))}`, {
        ...opts,
        headers: oazapfts.mergeHeaders(opts?.headers, {
            authorization
        })
    }));
}
/**
 * Get Module Execution Batch Activities By Id
 */
export function getModuleExecutionBatchActivitiesByIdV1ActivitiesModuleExecutionBatchModuleExecutionBatchIdGet({ moduleExecutionBatchId, authorization }, opts) {
    return oazapfts.ok(oazapfts.fetchJson(`/v1/activities/module-execution-batch/${encodeURIComponent(moduleExecutionBatchId)}`, {
        ...opts,
        headers: oazapfts.mergeHeaders(opts?.headers, {
            authorization
        })
    }));
}
