// @stack-spot/portal-network/dist/api/dataIntegration.js
/**
 * FastAPI
 * 0.1.0
 * DO NOT MODIFY - This file has been generated using oazapfts.
 * See https://www.npmjs.com/package/oazapfts
 */
import * as Oazapfts from "@oazapfts/runtime";
import * as QS from "@oazapfts/runtime/query";
export const defaults = {
    headers: {},
    baseUrl: "/",
};
const oazapfts = Oazapfts.runtime(defaults);
export const servers = {};
/**
 * Health Check
 */
export function healthCheckHealthzGet(opts) {
    return oazapfts.ok(oazapfts.fetchJson("/healthz", {
        ...opts
    }));
}
/**
 * Readiness Check
 */
export function readinessCheckReadyzGet(opts) {
    return oazapfts.ok(oazapfts.fetchJson("/readyz", {
        ...opts
    }));
}
/**
 * Search Objects
 */
export function searchObjectsV1SearchPost({ xAccountId, xUsername, authorization, searchKnowledgeObjectRequest }, opts) {
    return oazapfts.ok(oazapfts.fetchJson("/v1/search", oazapfts.json({
        ...opts,
        method: "POST",
        body: searchKnowledgeObjectRequest,
        headers: oazapfts.mergeHeaders(opts?.headers, {
            "x-account-id": xAccountId,
            "x-username": xUsername,
            authorization
        })
    })));
}
/**
 * Search Cross Account Objects
 */
export function searchCrossAccountObjectsV1SearchCrossAccountPost({ xAccountId, xUsername, authorization, searchCrossAccountKnowledgeObjectRequest }, opts) {
    return oazapfts.ok(oazapfts.fetchJson("/v1/search/cross-account", oazapfts.json({
        ...opts,
        method: "POST",
        body: searchCrossAccountKnowledgeObjectRequest,
        headers: oazapfts.mergeHeaders(opts?.headers, {
            "x-account-id": xAccountId,
            "x-username": xUsername,
            authorization
        })
    })));
}
/**
 * Create Ft Knowledge Source
 */
export function createFtKnowledgeSourceV1FtKnowledgeSourcesPost({ newKnowledgeSourceRequest }, opts) {
    return oazapfts.ok(oazapfts.fetchJson("/v1/ft-knowledge-sources", oazapfts.json({
        ...opts,
        method: "POST",
        body: newKnowledgeSourceRequest
    })));
}
/**
 * Delete Ft Knowledge Source
 */
export function deleteFtKnowledgeSourceV1FtKnowledgeSourcesSlugDelete({ slug }, opts) {
    return oazapfts.ok(oazapfts.fetchJson(`/v1/ft-knowledge-sources/${encodeURIComponent(slug)}`, {
        ...opts,
        method: "DELETE"
    }));
}
/**
 * Workspace Fork
 */
export function workspaceForkV1KnowledgeSourcesWorkspaceForkPost({ xAccountId, xUsername, authorization, knowledgeSourcesWorkspaceForkRequest }, opts) {
    return oazapfts.ok(oazapfts.fetchJson("/v1/knowledge-sources/workspace/fork", oazapfts.json({
        ...opts,
        method: "POST",
        body: knowledgeSourcesWorkspaceForkRequest,
        headers: oazapfts.mergeHeaders(opts?.headers, {
            "x-account-id": xAccountId,
            "x-username": xUsername,
            authorization
        })
    })));
}
/**
 * Workspace Delete Fork
 */
export function workspaceDeleteForkV1KnowledgeSourcesWorkspaceForkDelete({ xAccountId, xUsername, authorization, knowledgeSourcesWorkspaceDeleteRequest }, opts) {
    return oazapfts.ok(oazapfts.fetchJson("/v1/knowledge-sources/workspace/fork", oazapfts.json({
        ...opts,
        method: "DELETE",
        body: knowledgeSourcesWorkspaceDeleteRequest,
        headers: oazapfts.mergeHeaders(opts?.headers, {
            "x-account-id": xAccountId,
            "x-username": xUsername,
            authorization
        })
    })));
}
/**
 * Workspace List
 */
export function workspaceListV1KnowledgeSourcesWorkspaceListPost({ xAccountId, xUsername, authorization, knowledgeSourcesWorkspaceListRequest }, opts) {
    return oazapfts.ok(oazapfts.fetchJson("/v1/knowledge-sources/workspace/list", oazapfts.json({
        ...opts,
        method: "POST",
        body: knowledgeSourcesWorkspaceListRequest,
        headers: oazapfts.mergeHeaders(opts?.headers, {
            "x-account-id": xAccountId,
            "x-username": xUsername,
            authorization
        })
    })));
}
/**
 * Create Ks
 */
export function createKsV1KnowledgeSourcesPost({ xAccountId, xUsername, authorization, newKnowledgeSourceRequest }, opts) {
    return oazapfts.ok(oazapfts.fetchJson("/v1/knowledge-sources", oazapfts.json({
        ...opts,
        method: "POST",
        body: newKnowledgeSourceRequest,
        headers: oazapfts.mergeHeaders(opts?.headers, {
            "x-account-id": xAccountId,
            "x-username": xUsername,
            authorization
        })
    })));
}
/**
 * List Ks
 */
export function listKsV1KnowledgeSourcesGet({ visibility, order, types, xAccountId, xUsername, authorization }, opts) {
    return oazapfts.ok(oazapfts.fetchJson(`/v1/knowledge-sources${QS.query(QS.explode({
        visibility,
        order,
        types
    }))}`, {
        ...opts,
        headers: oazapfts.mergeHeaders(opts?.headers, {
            "x-account-id": xAccountId,
            "x-username": xUsername,
            authorization
        })
    }));
}
/**
 * Search Ks
 */
export function searchKsV1KnowledgeSourcesSearchPost({ xAccountId, xUsername, authorization, knowledgeSourceSearchRequest }, opts) {
    return oazapfts.ok(oazapfts.fetchJson("/v1/knowledge-sources/search", oazapfts.json({
        ...opts,
        method: "POST",
        body: knowledgeSourceSearchRequest,
        headers: oazapfts.mergeHeaders(opts?.headers, {
            "x-account-id": xAccountId,
            "x-username": xUsername,
            authorization
        })
    })));
}
/**
 * Find Ks
 */
export function findKsV1KnowledgeSourcesSlugGet({ slug, xAccountId, xUsername, authorization }, opts) {
    return oazapfts.ok(oazapfts.fetchJson(`/v1/knowledge-sources/${encodeURIComponent(slug)}`, {
        ...opts,
        headers: oazapfts.mergeHeaders(opts?.headers, {
            "x-account-id": xAccountId,
            "x-username": xUsername,
            authorization
        })
    }));
}
/**
 * Update Ks
 */
export function updateKsV1KnowledgeSourcesSlugPatch({ slug, xAccountId, xUsername, authorization, knowledgeSourcePatchesRequest }, opts) {
    return oazapfts.ok(oazapfts.fetchJson(`/v1/knowledge-sources/${encodeURIComponent(slug)}`, oazapfts.json({
        ...opts,
        method: "PATCH",
        body: knowledgeSourcePatchesRequest,
        headers: oazapfts.mergeHeaders(opts?.headers, {
            "x-account-id": xAccountId,
            "x-username": xUsername,
            authorization
        })
    })));
}
/**
 * Delete Ks
 */
export function deleteKsV1KnowledgeSourcesSlugDelete({ slug, xAccountId, xUsername, authorization }, opts) {
    return oazapfts.ok(oazapfts.fetchJson(`/v1/knowledge-sources/${encodeURIComponent(slug)}`, {
        ...opts,
        method: "DELETE",
        headers: oazapfts.mergeHeaders(opts?.headers, {
            "x-account-id": xAccountId,
            "x-username": xUsername,
            authorization
        })
    }));
}
/**
 * Exists Ks
 */
export function existsKsV1KnowledgeSourcesSlugExistsGet({ slug, xAccountId, xUsername, authorization }, opts) {
    return oazapfts.ok(oazapfts.fetchJson(`/v1/knowledge-sources/${encodeURIComponent(slug)}/exists`, {
        ...opts,
        headers: oazapfts.mergeHeaders(opts?.headers, {
            "x-account-id": xAccountId,
            "x-username": xUsername,
            authorization
        })
    }));
}
/**
 * Share Ks
 */
export function shareKsV1KnowledgeSourcesSlugSharePost({ slug, xAccountId, xUsername, authorization }, opts) {
    return oazapfts.ok(oazapfts.fetchJson(`/v1/knowledge-sources/${encodeURIComponent(slug)}/share`, {
        ...opts,
        method: "POST",
        headers: oazapfts.mergeHeaders(opts?.headers, {
            "x-account-id": xAccountId,
            "x-username": xUsername,
            authorization
        })
    }));
}
/**
 * Add Favorite
 */
export function addFavoriteV1KnowledgeSourcesSlugFavoritePost({ slug, xAccountId, xUsername, authorization }, opts) {
    return oazapfts.ok(oazapfts.fetchJson(`/v1/knowledge-sources/${encodeURIComponent(slug)}/favorite`, {
        ...opts,
        method: "POST",
        headers: oazapfts.mergeHeaders(opts?.headers, {
            "x-account-id": xAccountId,
            "x-username": xUsername,
            authorization
        })
    }));
}
/**
 * Delete Favorite
 */
export function deleteFavoriteV1KnowledgeSourcesSlugFavoriteDelete({ slug, xAccountId, xUsername, authorization }, opts) {
    return oazapfts.ok(oazapfts.fetchJson(`/v1/knowledge-sources/${encodeURIComponent(slug)}/favorite`, {
        ...opts,
        method: "DELETE",
        headers: oazapfts.mergeHeaders(opts?.headers, {
            "x-account-id": xAccountId,
            "x-username": xUsername,
            authorization
        })
    }));
}
/**
 * Publish Ks
 */
export function publishKsV1KnowledgeSourcesSlugPublishPost({ slug, xAccountId, xUsername, authorization, body }, opts) {
    return oazapfts.ok(oazapfts.fetchJson(`/v1/knowledge-sources/${encodeURIComponent(slug)}/publish`, oazapfts.json({
        ...opts,
        method: "POST",
        body,
        headers: oazapfts.mergeHeaders(opts?.headers, {
            "x-account-id": xAccountId,
            "x-username": xUsername,
            authorization
        })
    })));
}
/**
 * Publish Ks Batch
 */
export function publishKsBatchV1KnowledgeSourcesPublishBatchPost({ xAccountId, xUsername, authorization, knowledgeSourcePublishBatchRequest }, opts) {
    return oazapfts.ok(oazapfts.fetchJson("/v1/knowledge-sources/publish/batch", oazapfts.json({
        ...opts,
        method: "POST",
        body: knowledgeSourcePublishBatchRequest,
        headers: oazapfts.mergeHeaders(opts?.headers, {
            "x-account-id": xAccountId,
            "x-username": xUsername,
            authorization
        })
    })));
}
/**
 * Fork Ks
 */
export function forkKsV1KnowledgeSourcesSlugForkPost({ slug, xAccountId, xUsername, authorization, forkKnowledgeSourceRequest }, opts) {
    return oazapfts.ok(oazapfts.fetchJson(`/v1/knowledge-sources/${encodeURIComponent(slug)}/fork`, oazapfts.json({
        ...opts,
        method: "POST",
        body: forkKnowledgeSourceRequest,
        headers: oazapfts.mergeHeaders(opts?.headers, {
            "x-account-id": xAccountId,
            "x-username": xUsername,
            authorization
        })
    })));
}
/**
 * List Kos
 */
export function listKosV1KnowledgeSourcesKsSlugObjectsGet({ ksSlug, standalone, filePath, contentLimit, xAccountId, xUsername, authorization }, opts) {
    return oazapfts.ok(oazapfts.fetchJson(`/v1/knowledge-sources/${encodeURIComponent(ksSlug)}/objects${QS.query(QS.explode({
        standalone,
        file_path: filePath,
        content_limit: contentLimit
    }))}`, {
        ...opts,
        headers: oazapfts.mergeHeaders(opts?.headers, {
            "x-account-id": xAccountId,
            "x-username": xUsername,
            authorization
        })
    }));
}
/**
 * Reset Kos
 */
export function resetKosV1KnowledgeSourcesKsSlugObjectsDelete({ ksSlug, standalone, xAccountId, xUsername, authorization }, opts) {
    return oazapfts.ok(oazapfts.fetchJson(`/v1/knowledge-sources/${encodeURIComponent(ksSlug)}/objects${QS.query(QS.explode({
        standalone
    }))}`, {
        ...opts,
        method: "DELETE",
        headers: oazapfts.mergeHeaders(opts?.headers, {
            "x-account-id": xAccountId,
            "x-username": xUsername,
            authorization
        })
    }));
}
/**
 * List Kos Checksums
 */
export function listKosChecksumsV1KnowledgeSourcesKsSlugObjectsChecksumsGet({ ksSlug, xAccountId, xUsername, authorization }, opts) {
    return oazapfts.ok(oazapfts.fetchJson(`/v1/knowledge-sources/${encodeURIComponent(ksSlug)}/objects/checksums`, {
        ...opts,
        headers: oazapfts.mergeHeaders(opts?.headers, {
            "x-account-id": xAccountId,
            "x-username": xUsername,
            authorization
        })
    }));
}
/**
 * Vectorize Snippet Ks
 */
export function vectorizeSnippetKsV1KnowledgeSourcesSlugSnippetsPost(opts) {
    return oazapfts.ok(oazapfts.fetchJson(`/v1/knowledge-sources/${encodeURIComponent(slug)}/snippets`, {
        ...opts,
        method: "POST"
    }));
}
/**
 * Vectorize Custom Ks
 */
export function vectorizeCustomKsV1KnowledgeSourcesSlugCustomPost(opts) {
    return oazapfts.ok(oazapfts.fetchJson(`/v1/knowledge-sources/${encodeURIComponent(slug)}/custom`, {
        ...opts,
        method: "POST"
    }));
}
/**
 * Find Ko By Id
 */
export function findKoByIdV1KnowledgeSourcesKsSlugObjectsKoIdGet({ ksSlug, koId, xAccountId, xUsername, authorization }, opts) {
    return oazapfts.ok(oazapfts.fetchJson(`/v1/knowledge-sources/${encodeURIComponent(ksSlug)}/objects/${encodeURIComponent(koId)}`, {
        ...opts,
        headers: oazapfts.mergeHeaders(opts?.headers, {
            "x-account-id": xAccountId,
            "x-username": xUsername,
            authorization
        })
    }));
}
/**
 * Delete Ko By Id
 */
export function deleteKoByIdV1KnowledgeSourcesKsSlugObjectsKoIdDelete({ ksSlug, koId, xAccountId, xUsername, authorization }, opts) {
    return oazapfts.ok(oazapfts.fetchJson(`/v1/knowledge-sources/${encodeURIComponent(ksSlug)}/objects/${encodeURIComponent(koId)}`, {
        ...opts,
        method: "DELETE",
        headers: oazapfts.mergeHeaders(opts?.headers, {
            "x-account-id": xAccountId,
            "x-username": xUsername,
            authorization
        })
    }));
}
/**
 * Get Last Message
 */
export function getLastMessageV1KnowledgeSourcesKsSlugLastMessageGet({ ksSlug, xAccountId, xUsername, authorization }, opts) {
    return oazapfts.ok(oazapfts.fetchJson(`/v1/knowledge-sources/${encodeURIComponent(ksSlug)}/last-message`, {
        ...opts,
        headers: oazapfts.mergeHeaders(opts?.headers, {
            "x-account-id": xAccountId,
            "x-username": xUsername,
            authorization
        })
    }));
}
/**
 * Get Indexed Files
 */
export function getIndexedFilesV1KnowledgeSourcesKsSlugFilesGet({ ksSlug, xAccountId, xUsername, authorization }, opts) {
    return oazapfts.ok(oazapfts.fetchJson(`/v1/knowledge-sources/${encodeURIComponent(ksSlug)}/files`, {
        ...opts,
        headers: oazapfts.mergeHeaders(opts?.headers, {
            "x-account-id": xAccountId,
            "x-username": xUsername,
            authorization
        })
    }));
}
/**
 * Delete Indexed Files
 */
export function deleteIndexedFilesV1KnowledgeSourcesKsSlugFilesDelete({ ksSlug, xAccountId, xUsername, authorization, deleteIndexedFilesRequest }, opts) {
    return oazapfts.ok(oazapfts.fetchJson(`/v1/knowledge-sources/${encodeURIComponent(ksSlug)}/files`, oazapfts.json({
        ...opts,
        method: "DELETE",
        body: deleteIndexedFilesRequest,
        headers: oazapfts.mergeHeaders(opts?.headers, {
            "x-account-id": xAccountId,
            "x-username": xUsername,
            authorization
        })
    })));
}
/**
 * Get Upload Form
 */
export function getUploadFormV1FileUploadFormPost({ xAccountId, xUsername, authorization, newFileUploadRequest }, opts) {
    return oazapfts.ok(oazapfts.fetchJson("/v1/file-upload/form", oazapfts.json({
        ...opts,
        method: "POST",
        body: newFileUploadRequest,
        headers: oazapfts.mergeHeaders(opts?.headers, {
            "x-account-id": xAccountId,
            "x-username": xUsername,
            authorization
        })
    })));
}
/**
 * Get File Upload Status
 */
export function getFileUploadStatusV1FileUploadFileUploadIdGet({ fileUploadId, xAccountId, xUsername, authorization }, opts) {
    return oazapfts.ok(oazapfts.fetchJson(`/v1/file-upload/${encodeURIComponent(fileUploadId)}`, {
        ...opts,
        headers: oazapfts.mergeHeaders(opts?.headers, {
            "x-account-id": xAccountId,
            "x-username": xUsername,
            authorization
        })
    }));
}
/**
 * Save Chunked Files Deprecated
 */
export function saveChunkedFilesDeprecatedV1FileUploadFileUploadIdPost({ fileUploadId, xAccountId, xUsername, authorization, fileUploadSaveChunksRequest }, opts) {
    return oazapfts.ok(oazapfts.fetchJson(`/v1/file-upload/${encodeURIComponent(fileUploadId)}`, oazapfts.json({
        ...opts,
        method: "POST",
        body: fileUploadSaveChunksRequest,
        headers: oazapfts.mergeHeaders(opts?.headers, {
            "x-account-id": xAccountId,
            "x-username": xUsername,
            authorization
        })
    })));
}
/**
 * Internal Patch File Upload Status
 */
export function internalPatchFileUploadStatusV1FileUploadFileUploadIdPatch({ fileUploadId, xAccountId, xUsername, authorization, fileUploadUpdateRequest }, opts) {
    return oazapfts.ok(oazapfts.fetchJson(`/v1/file-upload/${encodeURIComponent(fileUploadId)}`, oazapfts.json({
        ...opts,
        method: "PATCH",
        body: fileUploadUpdateRequest,
        headers: oazapfts.mergeHeaders(opts?.headers, {
            "x-account-id": xAccountId,
            "x-username": xUsername,
            authorization
        })
    })));
}
/**
 * Save Chunked Files
 */
export function saveChunkedFilesV1FileUploadFileUploadIdKnowledgeObjectsPost({ fileUploadId, xAccountId, xUsername, authorization, fileUploadSaveChunksRequest }, opts) {
    return oazapfts.ok(oazapfts.fetchJson(`/v1/file-upload/${encodeURIComponent(fileUploadId)}/knowledge-objects`, oazapfts.json({
        ...opts,
        method: "POST",
        body: fileUploadSaveChunksRequest,
        headers: oazapfts.mergeHeaders(opts?.headers, {
            "x-account-id": xAccountId,
            "x-username": xUsername,
            authorization
        })
    })));
}
/**
 * Split Uploaded File
 */
export function splitUploadedFileV1FileUploadFileUploadIdSplitPost({ fileUploadId, xAccountId, xUsername, authorization, fileSplitRequest }, opts) {
    return oazapfts.ok(oazapfts.fetchJson(`/v1/file-upload/${encodeURIComponent(fileUploadId)}/split`, oazapfts.json({
        ...opts,
        method: "POST",
        body: fileSplitRequest,
        headers: oazapfts.mergeHeaders(opts?.headers, {
            "x-account-id": xAccountId,
            "x-username": xUsername,
            authorization
        })
    })));
}
/**
 * Split Uploaded File Preview
 */
export function splitUploadedFilePreviewV1FileUploadFileUploadIdSplitPreviewSplitStrategyGet({ fileUploadId, splitStrategy, xAccountId, xUsername, authorization }, opts) {
    return oazapfts.ok(oazapfts.fetchJson(`/v1/file-upload/${encodeURIComponent(fileUploadId)}/split/preview/${encodeURIComponent(splitStrategy)}`, {
        ...opts,
        headers: oazapfts.mergeHeaders(opts?.headers, {
            "x-account-id": xAccountId,
            "x-username": xUsername,
            authorization
        })
    }));
}
/**
 * Internal Get File Upload Metadata
 */
export function internalGetFileUploadMetadataV1FileUploadFileUploadIdMetadataGet({ fileUploadId, xAccountId, xUsername, authorization }, opts) {
    return oazapfts.ok(oazapfts.fetchJson(`/v1/file-upload/${encodeURIComponent(fileUploadId)}/metadata`, {
        ...opts,
        headers: oazapfts.mergeHeaders(opts?.headers, {
            "x-account-id": xAccountId,
            "x-username": xUsername,
            authorization
        })
    }));
}
/**
 * Get Presigned Download Url
 */
export function getPresignedDownloadUrlV1FileUploadPresignedUrlGet({ id, expiration, xAccountId, xUsername, authorization }, opts) {
    return oazapfts.ok(oazapfts.fetchJson(`/v1/file-upload/presigned/url${QS.query(QS.explode({
        id,
        expiration
    }))}`, {
        ...opts,
        headers: oazapfts.mergeHeaders(opts?.headers, {
            "x-account-id": xAccountId,
            "x-username": xUsername,
            authorization
        })
    }));
}
/**
 * Delete Bucket Object
 */
export function deleteBucketObjectV1FileUploadUploadIdDelete({ uploadId, xAccountId, xUsername, authorization }, opts) {
    return oazapfts.ok(oazapfts.fetchJson(`/v1/file-upload/${encodeURIComponent(uploadId)}`, {
        ...opts,
        method: "DELETE",
        headers: oazapfts.mergeHeaders(opts?.headers, {
            "x-account-id": xAccountId,
            "x-username": xUsername,
            authorization
        })
    }));
}
/**
 * Internal Get Current Embedding Usage
 */
export function internalGetCurrentEmbeddingUsageV1TokenUsageEmbeddingGet({ xAccountId, xUsername, authorization }, opts) {
    return oazapfts.ok(oazapfts.fetchJson("/v1/token-usage/embedding", {
        ...opts,
        headers: oazapfts.mergeHeaders(opts?.headers, {
            "x-account-id": xAccountId,
            "x-username": xUsername,
            authorization
        })
    }));
}
/**
 * Internal Consume Token Usage
 */
export function internalConsumeTokenUsageV1TokenUsageConsumePost({ xAccountId, xUsername, authorization, consumeTokenRequest }, opts) {
    return oazapfts.ok(oazapfts.fetchJson("/v1/token-usage/consume", oazapfts.json({
        ...opts,
        method: "POST",
        body: consumeTokenRequest,
        headers: oazapfts.mergeHeaders(opts?.headers, {
            "x-account-id": xAccountId,
            "x-username": xUsername,
            authorization
        })
    })));
}
/**
 * Get Upload Form V2
 */
export function getUploadFormV2V2FileUploadFormPost({ xAccountId, xUsername, authorization, newFileUploadRequest }, opts) {
    return oazapfts.ok(oazapfts.fetchJson("/v2/file-upload/form", oazapfts.json({
        ...opts,
        method: "POST",
        body: newFileUploadRequest,
        headers: oazapfts.mergeHeaders(opts?.headers, {
            "x-account-id": xAccountId,
            "x-username": xUsername,
            authorization
        })
    })));
}
/**
 * Internal Get Llm Config
 */
export function internalGetLlmConfigV1AccountsLlmConfigGet({ xAccountId, xUsername, authorization }, opts) {
    return oazapfts.ok(oazapfts.fetchJson("/v1/accounts/llm-config", {
        ...opts,
        headers: oazapfts.mergeHeaders(opts?.headers, {
            "x-account-id": xAccountId,
            "x-username": xUsername,
            authorization
        })
    }));
}
/**
 * Internal Get Or Create Knowledge Group
 */
export function internalGetOrCreateKnowledgeGroupV1KnowledgeGroupsGetOrCreatePut({ xAccountId, xUsername, authorization, getOrCreateKnowledgeGroupRequest }, opts) {
    return oazapfts.ok(oazapfts.fetchJson("/v1/knowledge-groups/get-or-create", oazapfts.json({
        ...opts,
        method: "PUT",
        body: getOrCreateKnowledgeGroupRequest,
        headers: oazapfts.mergeHeaders(opts?.headers, {
            "x-account-id": xAccountId,
            "x-username": xUsername,
            authorization
        })
    })));
}
/**
 * Internal Delete By Not Checksums
 */
export function internalDeleteByNotChecksumsV1KnowledgeGroupsGroupIdObjectsDelete({ groupId, xAccountId, xUsername, authorization, deleteGroupObjectsRequests }, opts) {
    return oazapfts.ok(oazapfts.fetchJson(`/v1/knowledge-groups/${encodeURIComponent(groupId)}/objects`, oazapfts.json({
        ...opts,
        method: "DELETE",
        body: deleteGroupObjectsRequests,
        headers: oazapfts.mergeHeaders(opts?.headers, {
            "x-account-id": xAccountId,
            "x-username": xUsername,
            authorization
        })
    })));
}
/**
 * Internal Update Knowledge Group
 */
export function internalUpdateKnowledgeGroupV1KnowledgeGroupsGroupIdPatch({ groupId, xAccountId, xUsername, authorization, updateKnowledgeGroupRequest }, opts) {
    return oazapfts.ok(oazapfts.fetchJson(`/v1/knowledge-groups/${encodeURIComponent(groupId)}`, oazapfts.json({
        ...opts,
        method: "PATCH",
        body: updateKnowledgeGroupRequest,
        headers: oazapfts.mergeHeaders(opts?.headers, {
            "x-account-id": xAccountId,
            "x-username": xUsername,
            authorization
        })
    })));
}
/**
 * Internal Create Knowledge Objects Batch
 */
export function internalCreateKnowledgeObjectsBatchV1KnowledgeObjectsBatchPost({ xAccountId, xUsername, authorization, body }, opts) {
    return oazapfts.ok(oazapfts.fetchJson("/v1/knowledge-objects/batch", oazapfts.json({
        ...opts,
        method: "POST",
        body,
        headers: oazapfts.mergeHeaders(opts?.headers, {
            "x-account-id": xAccountId,
            "x-username": xUsername,
            authorization
        })
    })));
}
/**
 * Create Embedding
 */
export function createEmbeddingV1EmbeddingsPost({ xAccountId, xUsername, authorization, body }, opts) {
    return oazapfts.ok(oazapfts.fetchJson("/v1/embeddings", {
        ...opts,
        method: "POST",
        body,
        headers: oazapfts.mergeHeaders(opts?.headers, {
            "x-account-id": xAccountId,
            "x-username": xUsername,
            authorization
        })
    }));
}
/**
 * List Conversations
 */
export function listConversationsV1ConversationsGet({ size, page, xAccountId, xUsername, authorization }, opts) {
    return oazapfts.ok(oazapfts.fetchJson(`/v1/conversations${QS.query(QS.explode({
        size,
        page
    }))}`, {
        ...opts,
        headers: oazapfts.mergeHeaders(opts?.headers, {
            "x-account-id": xAccountId,
            "x-username": xUsername,
            authorization
        })
    }));
}
/**
 * Create Conversation
 */
export function createConversationV1ConversationsPost({ xAccountId, xUsername, authorization, createConversationRequest }, opts) {
    return oazapfts.ok(oazapfts.fetchJson("/v1/conversations", oazapfts.json({
        ...opts,
        method: "POST",
        body: createConversationRequest,
        headers: oazapfts.mergeHeaders(opts?.headers, {
            "x-account-id": xAccountId,
            "x-username": xUsername,
            authorization
        })
    })));
}
/**
 * Get Conversation With History
 */
export function getConversationWithHistoryV1ConversationsConversationIdGet({ conversationId, xAccountId, xUsername, authorization }, opts) {
    return oazapfts.ok(oazapfts.fetchJson(`/v1/conversations/${encodeURIComponent(conversationId)}`, {
        ...opts,
        headers: oazapfts.mergeHeaders(opts?.headers, {
            "x-account-id": xAccountId,
            "x-username": xUsername,
            authorization
        })
    }));
}
/**
 * Update Title
 */
export function updateTitleV1ConversationsConversationIdPatch({ conversationId, xAccountId, xUsername, authorization, conversationUpdateTitleRequest }, opts) {
    return oazapfts.ok(oazapfts.fetchJson(`/v1/conversations/${encodeURIComponent(conversationId)}`, oazapfts.json({
        ...opts,
        method: "PATCH",
        body: conversationUpdateTitleRequest,
        headers: oazapfts.mergeHeaders(opts?.headers, {
            "x-account-id": xAccountId,
            "x-username": xUsername,
            authorization
        })
    })));
}
/**
 * Delete Conversation
 */
export function deleteConversationV1ConversationsConversationIdDelete({ conversationId, xAccountId, xUsername, authorization }, opts) {
    return oazapfts.ok(oazapfts.fetchJson(`/v1/conversations/${encodeURIComponent(conversationId)}`, {
        ...opts,
        method: "DELETE",
        headers: oazapfts.mergeHeaders(opts?.headers, {
            "x-account-id": xAccountId,
            "x-username": xUsername,
            authorization
        })
    }));
}
/**
 * Download Conversation
 */
export function downloadConversationV1ConversationsConversationIdDownloadGet({ conversationId, xAccountId, xUsername, authorization }, opts) {
    return oazapfts.ok(oazapfts.fetchJson(`/v1/conversations/${encodeURIComponent(conversationId)}/download`, {
        ...opts,
        headers: oazapfts.mergeHeaders(opts?.headers, {
            "x-account-id": xAccountId,
            "x-username": xUsername,
            authorization
        })
    }));
}
/**
 * Add Conversation History
 */
export function addConversationHistoryV1ConversationsConversationIdMessagesPost({ conversationId, xAccountId, xUsername, authorization, conversationHistoryRequest }, opts) {
    return oazapfts.ok(oazapfts.fetchJson(`/v1/conversations/${encodeURIComponent(conversationId)}/messages`, oazapfts.json({
        ...opts,
        method: "POST",
        body: conversationHistoryRequest,
        headers: oazapfts.mergeHeaders(opts?.headers, {
            "x-account-id": xAccountId,
            "x-username": xUsername,
            authorization
        })
    })));
}
/**
 * Get Last History
 */
export function getLastHistoryV1ConversationsConversationIdMessagesLastGet({ conversationId, xAccountId, xUsername, authorization }, opts) {
    return oazapfts.ok(oazapfts.fetchJson(`/v1/conversations/${encodeURIComponent(conversationId)}/messages/last`, {
        ...opts,
        headers: oazapfts.mergeHeaders(opts?.headers, {
            "x-account-id": xAccountId,
            "x-username": xUsername,
            authorization
        })
    }));
}
/**
 * Search
 */
export function searchV1ConversationsMessagesSearchPost({ xAccountId, xUsername, authorization, searchConversationHistoryRequest }, opts) {
    return oazapfts.ok(oazapfts.fetchJson("/v1/conversations/messages/search", oazapfts.json({
        ...opts,
        method: "POST",
        body: searchConversationHistoryRequest,
        headers: oazapfts.mergeHeaders(opts?.headers, {
            "x-account-id": xAccountId,
            "x-username": xUsername,
            authorization
        })
    })));
}
/**
 * List Ks
 */
export function listKsV2KnowledgeSourcesGet({ name, slug, size, page, visibilityList, order, types, xAccountId, xUsername, authorization }, opts) {
    return oazapfts.ok(oazapfts.fetchJson(`/v2/knowledge-sources${QS.query(QS.explode({
        name,
        slug,
        size,
        page,
        visibility_list: visibilityList,
        order,
        types
    }))}`, {
        ...opts,
        headers: oazapfts.mergeHeaders(opts?.headers, {
            "x-account-id": xAccountId,
            "x-username": xUsername,
            authorization
        })
    }));
}
/**
 * List Kos
 */
export function listKosV2KnowledgeSourcesKsSlugObjectsGet({ ksSlug, size, page, order, standalone, filePath, contentLimit, xAccountId, xUsername, authorization }, opts) {
    return oazapfts.ok(oazapfts.fetchJson(`/v2/knowledge-sources/${encodeURIComponent(ksSlug)}/objects${QS.query(QS.explode({
        size,
        page,
        order,
        standalone,
        file_path: filePath,
        content_limit: contentLimit
    }))}`, {
        ...opts,
        headers: oazapfts.mergeHeaders(opts?.headers, {
            "x-account-id": xAccountId,
            "x-username": xUsername,
            authorization
        })
    }));
}
/**
 * Upload Script
 */
export function uploadScriptV1ScriptUploadPost({ xAccountId, xUsername, authorization, uploadScriptRequest }, opts) {
    return oazapfts.ok(oazapfts.fetchJson("/v1/script-upload", oazapfts.json({
        ...opts,
        method: "POST",
        body: uploadScriptRequest,
        headers: oazapfts.mergeHeaders(opts?.headers, {
            "x-account-id": xAccountId,
            "x-username": xUsername,
            authorization
        })
    })));
}
/**
 * Get Script Content
 */
export function getScriptContentV1ScriptUploadScriptIdGet({ scriptId, xAccountId, xUsername, authorization }, opts) {
    return oazapfts.ok(oazapfts.fetchJson(`/v1/script-upload/${encodeURIComponent(scriptId)}`, {
        ...opts,
        headers: oazapfts.mergeHeaders(opts?.headers, {
            "x-account-id": xAccountId,
            "x-username": xUsername,
            authorization
        })
    }));
}
/**
 * Get Script Presigned Url
 */
export function getScriptPresignedUrlV1ScriptUploadScriptIdPresignedUrlGet({ scriptId, xAccountId, xUsername, authorization }, opts) {
    return oazapfts.ok(oazapfts.fetchJson(`/v1/script-upload/${encodeURIComponent(scriptId)}/presigned/url`, {
        ...opts,
        headers: oazapfts.mergeHeaders(opts?.headers, {
            "x-account-id": xAccountId,
            "x-username": xUsername,
            authorization
        })
    }));
}
/**
 * Create Script Execution State
 */
export function createScriptExecutionStateV1ScriptExecutionExecutionIdStepSlugAttemptPut({ executionId, stepSlug, attempt, xAccountId, xUsername, authorization, createScriptExecutionRequest }, opts) {
    return oazapfts.ok(oazapfts.fetchJson(`/v1/script-execution/${encodeURIComponent(executionId)}/${encodeURIComponent(stepSlug)}/${encodeURIComponent(attempt)}`, oazapfts.json({
        ...opts,
        method: "PUT",
        body: createScriptExecutionRequest,
        headers: oazapfts.mergeHeaders(opts?.headers, {
            "x-account-id": xAccountId,
            "x-username": xUsername,
            authorization
        })
    })));
}
/**
 * Get Script Execution State Presigned Url
 */
export function getScriptExecutionStatePresignedUrlV1ScriptExecutionIdPresignedUrlGet({ id, xAccountId, xUsername, authorization }, opts) {
    return oazapfts.ok(oazapfts.fetchJson(`/v1/script-execution/${encodeURIComponent(id)}/presigned/url`, {
        ...opts,
        headers: oazapfts.mergeHeaders(opts?.headers, {
            "x-account-id": xAccountId,
            "x-username": xUsername,
            authorization
        })
    }));
}
