// @stack-spot/portal-network/dist/api/discover.js
/**
 * OpenAPI definition
 * v0
 * DO NOT MODIFY - This file has been generated using oazapfts.
 * See https://www.npmjs.com/package/oazapfts
 */
import * as Oazapfts from "@oazapfts/runtime";
import * as QS from "@oazapfts/runtime/query";
export const defaults = {
    headers: {},
    baseUrl: "https://discover-discover-core.dev.stackspot.com",
};
const oazapfts = Oazapfts.runtime(defaults);
export const servers = {
    generatedServerUrl: "https://discover-discover-core.dev.stackspot.com"
};
/**
 * Get specific Opportunity
 */
export function getById({ opportunityId }, opts) {
    return oazapfts.ok(oazapfts.fetchJson(`/v2/opportunities/${encodeURIComponent(opportunityId)}`, {
        ...opts
    }));
}
/**
 * Update specific Opportunity
 */
export function updateById({ opportunityId, updateOpportunityRequest }, opts) {
    return oazapfts.ok(oazapfts.fetchJson(`/v2/opportunities/${encodeURIComponent(opportunityId)}`, oazapfts.json({
        ...opts,
        method: "PUT",
        body: updateOpportunityRequest
    })));
}
/**
 * Delete specific Opportunity
 */
export function deleteById({ opportunityId }, opts) {
    return oazapfts.ok(oazapfts.fetchText(`/v2/opportunities/${encodeURIComponent(opportunityId)}`, {
        ...opts,
        method: "DELETE"
    }));
}
/**
 * Get specific Hypothesis
 */
export function getById1({ hypothesisId }, opts) {
    return oazapfts.ok(oazapfts.fetchJson(`/v2/hypotheses/${encodeURIComponent(hypothesisId)}`, {
        ...opts
    }));
}
/**
 * Update specific Hypothesis
 */
export function updateById1({ hypothesisId, updateHypothesisRequest }, opts) {
    return oazapfts.ok(oazapfts.fetchJson(`/v2/hypotheses/${encodeURIComponent(hypothesisId)}`, oazapfts.json({
        ...opts,
        method: "PUT",
        body: updateHypothesisRequest
    })));
}
/**
 * Delete specific Hypothesis
 */
export function deleteById1({ hypothesisId }, opts) {
    return oazapfts.ok(oazapfts.fetchText(`/v2/hypotheses/${encodeURIComponent(hypothesisId)}`, {
        ...opts,
        method: "DELETE"
    }));
}
/**
 * Get specific Document
 */
export function getById2({ documentId }, opts) {
    return oazapfts.ok(oazapfts.fetchJson(`/v2/documents/${encodeURIComponent(documentId)}`, {
        ...opts
    }));
}
/**
 * Update specific Document
 */
export function updateById2({ documentId, updateDocumentRequest }, opts) {
    return oazapfts.ok(oazapfts.fetchJson(`/v2/documents/${encodeURIComponent(documentId)}`, oazapfts.json({
        ...opts,
        method: "PUT",
        body: updateDocumentRequest
    })));
}
/**
 * Delete specific Document
 */
export function deleteById2({ documentId }, opts) {
    return oazapfts.ok(oazapfts.fetchText(`/v2/documents/${encodeURIComponent(documentId)}`, {
        ...opts,
        method: "DELETE"
    }));
}
/**
 * Get all Opportunities
 */
export function getAll({ filter, page, size, sort, direction }, opts) {
    return oazapfts.ok(oazapfts.fetchJson(`/v2/opportunities${QS.query(QS.explode({
        filter,
        page,
        size,
        sort,
        direction
    }))}`, {
        ...opts
    }));
}
/**
 * Create new Opportunity
 */
export function create({ createOpportunityRequest }, opts) {
    return oazapfts.ok(oazapfts.fetchJson("/v2/opportunities", oazapfts.json({
        ...opts,
        method: "POST",
        body: createOpportunityRequest
    })));
}
/**
 * Fetch Insights
 */
export function fetchInsights(opts) {
    return oazapfts.ok(oazapfts.fetchJson("/v2/insights", {
        ...opts,
        method: "POST"
    }));
}
/**
 * Refresh Insights
 */
export function refreshInsights(opts) {
    return oazapfts.ok(oazapfts.fetchJson("/v2/insights/refresh", {
        ...opts,
        method: "POST"
    }));
}
/**
 * Get all Hypotheses
 */
export function getAll1({ opportunityId }, opts) {
    return oazapfts.ok(oazapfts.fetchJson(`/v2/hypotheses${QS.query(QS.explode({
        opportunityId
    }))}`, {
        ...opts
    }));
}
/**
 * Create new Hypothesis
 */
export function create1({ createHypothesisRequest }, opts) {
    return oazapfts.ok(oazapfts.fetchJson("/v2/hypotheses", oazapfts.json({
        ...opts,
        method: "POST",
        body: createHypothesisRequest
    })));
}
/**
 * Get all Documents by Hypothesis
 */
export function getAllByHypothesis({ hypothesisId }, opts) {
    return oazapfts.ok(oazapfts.fetchJson(`/v2/documents${QS.query(QS.explode({
        hypothesisId
    }))}`, {
        ...opts
    }));
}
/**
 * Create Document
 */
export function create2({ createDocumentRequest }, opts) {
    return oazapfts.ok(oazapfts.fetchJson("/v2/documents", oazapfts.json({
        ...opts,
        method: "POST",
        body: createDocumentRequest
    })));
}
export function chat({ authorization, messageRequest }, opts) {
    return oazapfts.ok(oazapfts.fetchJson("/v2/ai/chat", oazapfts.json({
        ...opts,
        method: "POST",
        body: messageRequest,
        headers: oazapfts.mergeHeaders(opts?.headers, {
            Authorization: authorization
        })
    })));
}
export function createAgent({ adminAgentCreateRequest }, opts) {
    return oazapfts.ok(oazapfts.fetchJson("/v2/admin/agents", oazapfts.json({
        ...opts,
        method: "POST",
        body: adminAgentCreateRequest
    })));
}
export function deactivateAgent({ schema, agentId }, opts) {
    return oazapfts.ok(oazapfts.fetchText(`/v2/admin/agents/${encodeURIComponent(schema)}/${encodeURIComponent(agentId)}/deactivate`, {
        ...opts,
        method: "PATCH"
    }));
}
export function activateAgent({ schema, agentId }, opts) {
    return oazapfts.ok(oazapfts.fetchJson(`/v2/admin/agents/${encodeURIComponent(schema)}/${encodeURIComponent(agentId)}/activate`, {
        ...opts,
        method: "PATCH"
    }));
}
export function getAndCacheData({ key }, opts) {
    return oazapfts.ok(oazapfts.fetchJson(`/v2/redis/hello-world-persist-redis/${encodeURIComponent(key)}`, {
        ...opts
    }));
}
/**
 * Get specific Insight
 */
export function getInsightById({ insightId }, opts) {
    return oazapfts.ok(oazapfts.fetchJson(`/v2/insights/${encodeURIComponent(insightId)}`, {
        ...opts
    }));
}
/**
 * Get specific Document content
 */
export function getContentById({ documentId }, opts) {
    return oazapfts.ok(oazapfts.fetchJson(`/v2/documents/${encodeURIComponent(documentId)}/content`, {
        ...opts
    }));
}
/**
 * Get chat artifacts
 */
export function getAll2({ chatId }, opts) {
    return oazapfts.ok(oazapfts.fetchJson(`/v2/artifacts${QS.query(QS.explode({
        chatId
    }))}`, {
        ...opts
    }));
}
export function hello(opts) {
    return oazapfts.ok(oazapfts.fetchJson("/v2/admin/hello", {
        ...opts
    }));
}
export function listAgents({ schema }, opts) {
    return oazapfts.ok(oazapfts.fetchJson(`/v2/admin/agents/${encodeURIComponent(schema)}`, {
        ...opts
    }));
}
export function deleteAgent({ schema, agentId }, opts) {
    return oazapfts.ok(oazapfts.fetchText(`/v2/admin/agents/${encodeURIComponent(schema)}/${encodeURIComponent(agentId)}`, {
        ...opts,
        method: "DELETE"
    }));
}
