// @stack-spot/portal-network/dist/api/genAiInference.js
/**
 * FastAPI
 * 0.1.0
 * DO NOT MODIFY - This file has been generated using oazapfts.
 * See https://www.npmjs.com/package/oazapfts
 */
import * as Oazapfts from "@oazapfts/runtime";
import * as QS from "@oazapfts/runtime/query";
export const defaults = {
    headers: {},
    baseUrl: "/",
};
const oazapfts = Oazapfts.runtime(defaults);
export const servers = {};
/**
 * Health Check
 */
export function healthCheckHealthzGet(opts) {
    return oazapfts.ok(oazapfts.fetchJson("/healthz", {
        ...opts
    }));
}
/**
 * Readiness Check
 */
export function readinessCheckReadyzGet(opts) {
    return oazapfts.ok(oazapfts.fetchJson("/readyz", {
        ...opts
    }));
}
/**
 * Handle conversation requests
 */
export function converseV0ConversationPost({ xAccountId, authorization, sendConversation }, opts) {
    return oazapfts.ok(oazapfts.fetchJson("/v0/conversation", oazapfts.json({
        ...opts,
        method: "POST",
        body: sendConversation,
        headers: oazapfts.mergeHeaders(opts?.headers, {
            "x-account-id": xAccountId,
            authorization
        })
    })));
}
/**
 * Add Self Hosted Model
 */
export function addSelfHostedModelV1LlmModelsPost({ xAccountId, authorization, createLlmModelRequest }, opts) {
    return oazapfts.ok(oazapfts.fetchJson("/v1/llm/models", oazapfts.json({
        ...opts,
        method: "POST",
        body: createLlmModelRequest,
        headers: oazapfts.mergeHeaders(opts?.headers, {
            "x-account-id": xAccountId,
            authorization
        })
    })));
}
/**
 * List Models
 */
export function listModelsV1LlmModelsGet({ active, resource, $default, displayName, selfHosted, orderBy, orderDir, page, pageSize, xAccountId, authorization }, opts) {
    return oazapfts.ok(oazapfts.fetchJson(`/v1/llm/models${QS.query(QS.explode({
        active,
        resource,
        "default": $default,
        display_name: displayName,
        self_hosted: selfHosted,
        order_by: orderBy,
        order_dir: orderDir,
        page,
        page_size: pageSize
    }))}`, {
        ...opts,
        headers: oazapfts.mergeHeaders(opts?.headers, {
            "x-account-id": xAccountId,
            authorization
        })
    }));
}
/**
 * Get Model
 */
export function getModelV1LlmModelsModelIdGet({ modelId, xAccountId, authorization }, opts) {
    return oazapfts.ok(oazapfts.fetchJson(`/v1/llm/models/${encodeURIComponent(modelId)}`, {
        ...opts,
        headers: oazapfts.mergeHeaders(opts?.headers, {
            "x-account-id": xAccountId,
            authorization
        })
    }));
}
/**
 * Toggle Model Status
 */
export function toggleModelStatusV1LlmModelsModelIdPatch({ modelId, xAccountId, authorization, toggleModelStatusRequest }, opts) {
    return oazapfts.ok(oazapfts.fetchJson(`/v1/llm/models/${encodeURIComponent(modelId)}`, oazapfts.json({
        ...opts,
        method: "PATCH",
        body: toggleModelStatusRequest,
        headers: oazapfts.mergeHeaders(opts?.headers, {
            "x-account-id": xAccountId,
            authorization
        })
    })));
}
/**
 * Update
 */
export function updateV1LlmModelsModelIdPut({ modelId, xAccountId, authorization, updateLlmModelRequest }, opts) {
    return oazapfts.ok(oazapfts.fetchJson(`/v1/llm/models/${encodeURIComponent(modelId)}`, oazapfts.json({
        ...opts,
        method: "PUT",
        body: updateLlmModelRequest,
        headers: oazapfts.mergeHeaders(opts?.headers, {
            "x-account-id": xAccountId,
            authorization
        })
    })));
}
/**
 * Delete
 */
export function deleteV1LlmModelsModelIdDelete({ modelId, xAccountId, authorization }, opts) {
    return oazapfts.ok(oazapfts.fetchJson(`/v1/llm/models/${encodeURIComponent(modelId)}`, {
        ...opts,
        method: "DELETE",
        headers: oazapfts.mergeHeaders(opts?.headers, {
            "x-account-id": xAccountId,
            authorization
        })
    }));
}
/**
 * Get Models List
 */
export function getModelsListV1LlmModelsListPost({ xAccountId, authorization, getModelByIdsRequest }, opts) {
    return oazapfts.ok(oazapfts.fetchJson("/v1/llm/models-list", oazapfts.json({
        ...opts,
        method: "POST",
        body: getModelByIdsRequest,
        headers: oazapfts.mergeHeaders(opts?.headers, {
            "x-account-id": xAccountId,
            authorization
        })
    })));
}
/**
 * Save Or Update Model Resources
 */
export function saveOrUpdateModelResourcesV1LlmModelsModelIdResourcesPut({ modelId, xAccountId, authorization, body }, opts) {
    return oazapfts.ok(oazapfts.fetchJson(`/v1/llm/models/${encodeURIComponent(modelId)}/resources`, oazapfts.json({
        ...opts,
        method: "PUT",
        body,
        headers: oazapfts.mergeHeaders(opts?.headers, {
            "x-account-id": xAccountId,
            authorization
        })
    })));
}
/**
 * Delete Model Resources
 */
export function deleteModelResourcesV1LlmResourcesResourceIdDelete({ resourceId, xAccountId, authorization }, opts) {
    return oazapfts.ok(oazapfts.fetchJson(`/v1/llm/resources/${encodeURIComponent(resourceId)}`, {
        ...opts,
        method: "DELETE",
        headers: oazapfts.mergeHeaders(opts?.headers, {
            "x-account-id": xAccountId,
            authorization
        })
    }));
}
/**
 * List Llm Providers
 */
export function listLlmProvidersV1LlmProvidersGet({ acceptsSelfHosted, xAccountId, authorization }, opts) {
    return oazapfts.ok(oazapfts.fetchJson(`/v1/llm/providers${QS.query(QS.explode({
        accepts_self_hosted: acceptsSelfHosted
    }))}`, {
        ...opts,
        headers: oazapfts.mergeHeaders(opts?.headers, {
            "x-account-id": xAccountId,
            authorization
        })
    }));
}
/**
 * Handle completions requests
 */
export function createCompletionsV1ChatCompletionsPost({ xAccountId, authorization, xRequestOrigin, xConversationId, xMessageId, xQcExecutionId, xQcSlug, xPlatformVersion, chatCompletion }, opts) {
    return oazapfts.ok(oazapfts.fetchJson("/v1/chat/completions", oazapfts.json({
        ...opts,
        method: "POST",
        body: chatCompletion,
        headers: oazapfts.mergeHeaders(opts?.headers, {
            "x-account-id": xAccountId,
            authorization,
            "x-request-origin": xRequestOrigin,
            "x-conversation-id": xConversationId,
            "x-message-id": xMessageId,
            "x-qc-execution-id": xQcExecutionId,
            "x-qc-slug": xQcSlug,
            "x-platform-version": xPlatformVersion
        })
    })));
}
/**
 * Agent Chat
 */
export function agentChatV1AgentAgentIdChatPost({ agentId, xAccountId, authorization, chatRequest }, opts) {
    return oazapfts.ok(oazapfts.fetchJson(`/v1/agent/${encodeURIComponent(agentId)}/chat`, oazapfts.json({
        ...opts,
        method: "POST",
        body: chatRequest,
        headers: oazapfts.mergeHeaders(opts?.headers, {
            "x-account-id": xAccountId,
            authorization
        })
    })));
}
