// @stack-spot/portal-network/dist/api/workspace-ai.js
/**
 * FastAPI
 * 0.1.0
 * DO NOT MODIFY - This file has been generated using oazapfts.
 * See https://www.npmjs.com/package/oazapfts
 */
import * as Oazapfts from "@oazapfts/runtime";
import * as QS from "@oazapfts/runtime/query";
export const defaults = {
    headers: {},
    baseUrl: "/",
};
const oazapfts = Oazapfts.runtime(defaults);
export const servers = {};
/**
 * Create Workspace
 */
export function createWorkspaceV1WorkspacesPost({ authorization, xAccountId, newWorkspaceRequest }, opts) {
    return oazapfts.ok(oazapfts.fetchJson("/v1/workspaces", oazapfts.json({
        ...opts,
        method: "POST",
        body: newWorkspaceRequest,
        headers: oazapfts.mergeHeaders(opts?.headers, {
            authorization,
            "x-account-id": xAccountId
        })
    })));
}
/**
 * List Workspaces
 */
export function listWorkspacesV1WorkspacesGet({ visibility, order, name, authorization, xAccountId }, opts) {
    return oazapfts.ok(oazapfts.fetchJson(`/v1/workspaces${QS.query(QS.explode({
        visibility,
        order,
        name
    }))}`, {
        ...opts,
        headers: oazapfts.mergeHeaders(opts?.headers, {
            authorization,
            "x-account-id": xAccountId
        })
    }));
}
/**
 * Get Workspace By Id
 */
export function getWorkspaceByIdV1WorkspacesIdGet({ id, authorization, xAccountId }, opts) {
    return oazapfts.ok(oazapfts.fetchJson(`/v1/workspaces/${encodeURIComponent(id)}`, {
        ...opts,
        headers: oazapfts.mergeHeaders(opts?.headers, {
            authorization,
            "x-account-id": xAccountId
        })
    }));
}
/**
 * Update Workspace
 */
export function updateWorkspaceV1WorkspacesWorkspaceIdPatch({ workspaceId, authorization, xAccountId, updateWorkspaceRequest }, opts) {
    return oazapfts.ok(oazapfts.fetchJson(`/v1/workspaces/${encodeURIComponent(workspaceId)}`, oazapfts.json({
        ...opts,
        method: "PATCH",
        body: updateWorkspaceRequest,
        headers: oazapfts.mergeHeaders(opts?.headers, {
            authorization,
            "x-account-id": xAccountId
        })
    })));
}
/**
 * Delete Workspace
 */
export function deleteWorkspaceV1WorkspacesWorkspaceIdDelete({ workspaceId, authorization, xAccountId }, opts) {
    return oazapfts.ok(oazapfts.fetchJson(`/v1/workspaces/${encodeURIComponent(workspaceId)}`, {
        ...opts,
        method: "DELETE",
        headers: oazapfts.mergeHeaders(opts?.headers, {
            authorization,
            "x-account-id": xAccountId
        })
    }));
}
/**
 * Add Favorite
 */
export function addFavoriteV1WorkspacesWorkspaceIdFavoritePost({ workspaceId, authorization, xAccountId }, opts) {
    return oazapfts.ok(oazapfts.fetchJson(`/v1/workspaces/${encodeURIComponent(workspaceId)}/favorite`, {
        ...opts,
        method: "POST",
        headers: oazapfts.mergeHeaders(opts?.headers, {
            authorization,
            "x-account-id": xAccountId
        })
    }));
}
/**
 * Delete Favorite
 */
export function deleteFavoriteV1WorkspacesWorkspaceIdFavoriteDelete({ workspaceId, authorization, xAccountId }, opts) {
    return oazapfts.ok(oazapfts.fetchJson(`/v1/workspaces/${encodeURIComponent(workspaceId)}/favorite`, {
        ...opts,
        method: "DELETE",
        headers: oazapfts.mergeHeaders(opts?.headers, {
            authorization,
            "x-account-id": xAccountId
        })
    }));
}
/**
 * Create Shareable Link
 */
export function createShareableLinkV1WorkspacesWorkspaceIdSharePost({ workspaceId, authorization, xAccountId }, opts) {
    return oazapfts.ok(oazapfts.fetchJson(`/v1/workspaces/${encodeURIComponent(workspaceId)}/share`, {
        ...opts,
        method: "POST",
        headers: oazapfts.mergeHeaders(opts?.headers, {
            authorization,
            "x-account-id": xAccountId
        })
    }));
}
/**
 * Add Permission
 */
export function addPermissionV1WorkspacesWorkspaceIdPermissionsPatch({ workspaceId, authorization, xAccountId, workspacePermissionRequest }, opts) {
    return oazapfts.ok(oazapfts.fetchJson(`/v1/workspaces/${encodeURIComponent(workspaceId)}/permissions`, oazapfts.json({
        ...opts,
        method: "PATCH",
        body: workspacePermissionRequest,
        headers: oazapfts.mergeHeaders(opts?.headers, {
            authorization,
            "x-account-id": xAccountId
        })
    })));
}
/**
 * Remove Permission
 */
export function removePermissionV1WorkspacesWorkspaceIdPermissionsDelete({ workspaceId, authorization, xAccountId, workspacePermissionRequest }, opts) {
    return oazapfts.ok(oazapfts.fetchJson(`/v1/workspaces/${encodeURIComponent(workspaceId)}/permissions`, oazapfts.json({
        ...opts,
        method: "DELETE",
        body: workspacePermissionRequest,
        headers: oazapfts.mergeHeaders(opts?.headers, {
            authorization,
            "x-account-id": xAccountId
        })
    })));
}
/**
 * List Groups And Members Permission In Workspace
 */
export function listGroupsAndMembersPermissionInWorkspaceV1WorkspacesWorkspaceIdPermissionsGet({ workspaceId, search, filterBy, filterValue, sortBy, page, size, order, authorization, xAccountId }, opts) {
    return oazapfts.ok(oazapfts.fetchJson(`/v1/workspaces/${encodeURIComponent(workspaceId)}/permissions${QS.query(QS.explode({
        search,
        filter_by: filterBy,
        filter_value: filterValue,
        sort_by: sortBy,
        page,
        size,
        order
    }))}`, {
        ...opts,
        headers: oazapfts.mergeHeaders(opts?.headers, {
            authorization,
            "x-account-id": xAccountId
        })
    }));
}
/**
 * Add Content
 */
export function addContentV1WorkspacesWorkspaceIdContentTypePost({ workspaceId, contentType, authorization, xAccountId, addWorkspaceContentRequest }, opts) {
    return oazapfts.ok(oazapfts.fetchJson(`/v1/workspaces/${encodeURIComponent(workspaceId)}/${encodeURIComponent(contentType)}`, oazapfts.json({
        ...opts,
        method: "POST",
        body: addWorkspaceContentRequest,
        headers: oazapfts.mergeHeaders(opts?.headers, {
            authorization,
            "x-account-id": xAccountId
        })
    })));
}
/**
 * List All By Content Type
 */
export function listAllByContentTypeV1WorkspacesWorkspaceIdContentTypeGet({ workspaceId, contentType, name, authorization, xAccountId }, opts) {
    return oazapfts.ok(oazapfts.fetchJson(`/v1/workspaces/${encodeURIComponent(workspaceId)}/${encodeURIComponent(contentType)}${QS.query(QS.explode({
        name
    }))}`, {
        ...opts,
        headers: oazapfts.mergeHeaders(opts?.headers, {
            authorization,
            "x-account-id": xAccountId
        })
    }));
}
/**
 * Remove Content
 */
export function removeContentV1WorkspacesWorkspaceIdContentTypeContentIdDelete({ workspaceId, contentType, contentId, authorization, xAccountId }, opts) {
    return oazapfts.ok(oazapfts.fetchJson(`/v1/workspaces/${encodeURIComponent(workspaceId)}/${encodeURIComponent(contentType)}/${encodeURIComponent(contentId)}`, {
        ...opts,
        method: "DELETE",
        headers: oazapfts.mergeHeaders(opts?.headers, {
            authorization,
            "x-account-id": xAccountId
        })
    }));
}
/**
 * List All Contents
 */
export function listAllContentsV1WorkspacesContentsContentTypeGet({ contentType, name, authorization, xAccountId }, opts) {
    return oazapfts.ok(oazapfts.fetchJson(`/v1/workspaces/contents/${encodeURIComponent(contentType)}${QS.query(QS.explode({
        name
    }))}`, {
        ...opts,
        headers: oazapfts.mergeHeaders(opts?.headers, {
            authorization,
            "x-account-id": xAccountId
        })
    }));
}
/**
 * List All Contents
 */
export function listAllContentsV1WorkspacesWorkspaceIdContentsGet({ workspaceId, name, authorization, xAccountId }, opts) {
    return oazapfts.ok(oazapfts.fetchJson(`/v1/workspaces/${encodeURIComponent(workspaceId)}/contents${QS.query(QS.explode({
        name
    }))}`, {
        ...opts,
        headers: oazapfts.mergeHeaders(opts?.headers, {
            authorization,
            "x-account-id": xAccountId
        })
    }));
}
/**
 * Use Shareable Link
 */
export function useShareableLinkV1ShareShareIdPatch({ shareId, authorization, xAccountId }, opts) {
    return oazapfts.ok(oazapfts.fetchJson(`/v1/share/${encodeURIComponent(shareId)}`, {
        ...opts,
        method: "PATCH",
        headers: oazapfts.mergeHeaders(opts?.headers, {
            authorization,
            "x-account-id": xAccountId
        })
    }));
}
/**
 * Add Permission
 */
export function addPermissionV1ResourceTypeResourceTypeResourcesResourceIdentifierPatch({ resourceType, resourceIdentifier, authorization, xAccountId, permissionListForSharedRequest }, opts) {
    return oazapfts.ok(oazapfts.fetchJson(`/v1/resource-type/${encodeURIComponent(resourceType)}/resources/${encodeURIComponent(resourceIdentifier)}`, oazapfts.json({
        ...opts,
        method: "PATCH",
        body: permissionListForSharedRequest,
        headers: oazapfts.mergeHeaders(opts?.headers, {
            authorization,
            "x-account-id": xAccountId
        })
    })));
}
/**
 * Remove Permission
 */
export function removePermissionV1ResourceTypeResourceTypeResourcesResourceIdentifierDelete({ resourceType, resourceIdentifier, authorization, xAccountId, permissionListForSharedRequest }, opts) {
    return oazapfts.ok(oazapfts.fetchJson(`/v1/resource-type/${encodeURIComponent(resourceType)}/resources/${encodeURIComponent(resourceIdentifier)}`, oazapfts.json({
        ...opts,
        method: "DELETE",
        body: permissionListForSharedRequest,
        headers: oazapfts.mergeHeaders(opts?.headers, {
            authorization,
            "x-account-id": xAccountId
        })
    })));
}
/**
 * List Members
 */
export function listMembersV1ResourceTypeResourceTypeResourcesResourceIdentifierMembersGet({ resourceType, resourceIdentifier, search, filterBy, filterValue, sortBy, page, size, order, authorization, xAccountId }, opts) {
    return oazapfts.ok(oazapfts.fetchJson(`/v1/resource-type/${encodeURIComponent(resourceType)}/resources/${encodeURIComponent(resourceIdentifier)}/members${QS.query(QS.explode({
        search,
        filter_by: filterBy,
        filter_value: filterValue,
        sort_by: sortBy,
        page,
        size,
        order
    }))}`, {
        ...opts,
        headers: oazapfts.mergeHeaders(opts?.headers, {
            authorization,
            "x-account-id": xAccountId
        })
    }));
}
/**
 * Healthz
 */
export function healthzHealthzGet(opts) {
    return oazapfts.ok(oazapfts.fetchJson("/healthz", {
        ...opts
    }));
}
/**
 * Base Get
 */
export function baseGetGet(opts) {
    return oazapfts.ok(oazapfts.fetchJson("/", {
        ...opts
    }));
}
